Fix # (see https://help.github.com/en/articles/closing-issues-using-keywords)

Describe changes proposed in this pull request:
- a
- b

# Checks
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). We can fix this during merge by using a squash+merge if necessary
- [ ] Has tests or has a separate issue that describes the types of test that should be created. If no test is included it should explicitly be mentioned in the PR why there is no test.
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!
- [ ] Make sure your PR has one of the labels defined in https://github.com/cBioPortal/cbioportal/blob/master/.github/release-drafter.yml

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [Giphy CAPTURE](https://giphy.com/apps/giphycapture) or [Peek](https://github.com/phw/peek)

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). It can help to figure out who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them either through slack or by assigning them as a reviewer on the PR
Fix # (see https://help.github.com/en/articles/closing-issues-using-keywords)

Describe changes proposed in this pull request:
- a
- b

# Checks
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). We can fix this during merge by using a squash+merge if necessary
- [ ] Has tests or has a separate issue that describes the types of test that should be created. If no test is included it should explicitly be mentioned in the PR why there is no test.
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!
- [ ] Make sure your PR has one of the labels defined in https://github.com/cBioPortal/cbioportal/blob/master/.github/release-drafter.yml

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [Giphy CAPTURE](https://giphy.com/apps/giphycapture) or [Peek](https://github.com/phw/peek)

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). It can help to figure out who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them either through slack or by assigning them as a reviewer on the PR
# cBioPortal Community Code of Conduct

cBioPortal follows the [CNCF Code of Conduct](https://github.com/cncf/foundation/blob/master/code-of-conduct.md).

Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting <conduct@cbioportal.org>.
# How to contribute

Thank you for your interest in contributing to cBioPortal! This document provides a brief set of guidelines for contributing.

# Who are you?

We are curious to learn more about you! We would love to help you get started! The contributors in our community all have different backgrounds. For instance some people have:

1. Engineering experience but no to little knowledge of cancer genomics
2. Knowledge about cancer genomics but no to little engineering experience
3. No engineering nor cancer genomics experience but an eagerness to contribute

if you feel like you don't fall into any of these categories, please reach out so you can help us update the above list üôÇ! Note that there are many contributions that can be made to an open source commmunity without coding a single line of code. You can reach us through our [public slack channel](https://slack.cbioportal.org).

# Join the Slack!

Come and chat with us at https://slack.cbioportal.org üëã

# Making a code contribution

The cBioPortal currently uses a "fork and pull" model for collaborative software development.

From the [GitHub Help Page of Using Pull Requests](https://help.github.com/articles/using-pull-requests/):

"The fork & pull model lets anyone fork an existing repository and push changes to their personal fork without requiring access be granted to the source repository. The changes must then be pulled into the source repository by the project maintainer. This model reduces the amount of friction for new contributors and is popular with open source projects because it allows people to work independently without upfront coordination."

## Getting Started

### Programming
 * Make sure you have a [GitHub account](https://github.com/signup/free).
 * Familiarize yourself with the [project documentation](https://docs.cbioportal.org), including the [Feature Development Guide](https://docs.cbioportal.org/4.-development/feature-development-guide), the [Architecture docs](https://docs.cbioportal.org/2.1-deployment/architecture-overview), the [backend code organization](docs/Backend-Code-Organization.md) and [backend development guidelines](docs/Backend-Development-Guidelines.md).
 * Find a [good first issue](https://github.com/cBioPortal/cbioportal/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22good+first+issue%22) to start with
 * Check if the issue will require frontend or backend changes. If it is for the frontend look at how to set up the [frontend repo](https://github.com/cbioPortal/cbioportal-frontend/) instead
 * Fork the cbioportal or cbioportal-frontend project on GitHub depending on what your working on.  For general instructions on forking a GitHub project, see [Forking a Repo](https://help.github.com/articles/fork-a-repo/) and [Syncing a fork](https://help.github.com/articles/syncing-a-fork/).
 * Reach out on slack or our [Google user group](https://groups.google.com/g/cbioportal) if you run into any issues

### Documentation
If you'd like to improve our documentation, have a look at https://docs.cbioportal.org. At the bottom of each page is an "edit page" button to make changes. You can use the GitHub UI to edit the pages and submit them.

## Contributing Code Changes via a Pull Request

Once you have forked the repo, you need to create your code contributions within a new branch of your forked repo.  For general background on creating and managing branches within GitHub, see:  [Git Branching and Merging](https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging).

* To begin, create a topic branch from where you want to base your work.
 * For any change that requires database migrations, this will be the **rc branch**. For all other changes, this will be the **master branch**. For additional details, see [Branches within cBioPortal](#branches-within-cbioportal) below.

You usually create a branch like so:

```
git checkout master
git checkout -b [name_of_your_new_branch]
```

You then usually commit code changes, and push your branch back to GitHub like so:

```git push origin [name_of_your_new_branch]```

A few tips:

* Make commits in logical/cohesive units.
* Make sure you have added the necessary tests for your changes.
* Run _all_ tests to assure nothing else was accidentally broken in the java part (data loading and front-end parts are tested by other scripts in github actions). This is done by running:  ```mvn integration-test```.

When you are ready to submit your pull-request:

* Push your branch to your GitHub project.
* Open the pull request to the branch you've based your work on

For more details on submitting a pull-request, please see:  [GitHub Guide to Collaborating with issues and pull requests](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests).

## Branches within cBioPortal

To figure out where your first pull request might go, it helps to have an understanding of cBioPortal's branching model. The cBioPortal currently maintains three branches in both the [frontend](https://github.com/cbioportal/cbioportal-frontend) and [backend repo](https://github.com/cbioportal/cbioportal):

 * **master**:  this reflects what will be released with our next weekly release (https://github.com/cBioPortal/cbioportal/releases). For the [frontend repo](https://github.com/cbioportal/cbioportal-frontend) this branch is automatically deployed to production. On the backend it is deployed at least once a week. New features, bug fixes and documentation updates can go here. Only if the feature requires a database migration it should go to **rc**.
  * **rc**:  release candidate branch, this branch contains new changes that require a database migration. It is deployed to https://rc.cbioportal.org. Once it's ready for more thorough product review a new branch is created with the name **release-x.y.z**. This way people can still continue to submit new changes to **rc**, while more thorough testing takes place of the **release-x.y.z** branch.
  *  **release-x.y.z**: this branch contains changes that require a database migration. It will be merged to master after thorough product review on https://beta.cbioportal.org.
  
We try to continuously merge new changes from `master` to `release-x.y.z`, and subsequently from the `release-x.y.z` branch to `rc` such that everybody is working on the latest code. Keep in mind though that occasionally there are conflicts that need to be resolved before we can merge. If you're working on e.g. the rc branch, you can check whether all the changes in master are in rc like this: https://github.com/cBioPortal/cbioportal/compare/rc...master. If a particular change you are waiting for is not there, one can help creating a pull request that merges these changes in. Try e.g. (if origin points to the cbioportal repo): `git fetch origin && git checkout origin/rc -b merge-master-to-rc && git merge origin/master`. Then resolve conflicts and push the branch `merge-master-to-rc` to your repo and open the PR.

### Getting your changes reviewed

Once you've submitted your pull request, you want
other members of the development community to review
whether integrating your change will cause problems
for any users or the maintainability of the software.

If you have an idea who might be able to spot such issues
in the parts of the code and functionality affected by your changes,
notify them by requesting a review using the **Reviewers** menu
to the right of the summary you just wrote
and/or `@`-mentioning them in a comment. Or reaching out them on [slack](https://slack.cbioportal.org).

Reviewers may request you to rephrase or adjust things
before they allow the changes to be integrated.
If they do, commit the amendments as new, separate changes,
to allow the reviewers to see what changed since they last read your code.
Do not overwrite previously-reviewed commits with
ones that include additional changes (by `--amend`ing or squashing)
until the reviewers approve.
Reviewers may request you to squash such amendment commits afterwards,
or offer to push rewritten versions of your commits themselves.

## Pull Request Reviewers Guide
If someone requests your review on a pull request,
read the title and description and assign any other collaborators
who would want to know about the proposed change.

Decide whether you think that your input is needed,
and that the PR should wait for your further review before being merged.
If not, un-assign yourself as a reviewer and leave a comment.

Here we describe the guidelines for the reviewer. Always follow the checks in
general, then follow the other checks that apply:

### General
- Double check all the things in the **Checks** section of the Pull Request.
  Remind the submitter if any of them are not fulfilled
- Are the test cases spanning a decent amount of scenarios? It is the
  submitters as well as the reviewers responsibility to not let any errors
  sneak into the portal.

Bug fixes:

- Should the bug that causes the issue be added as a test case?

New features:

- If this is a new feature make sure the proposed changes are in line with the
  current planning of cBioPortal e.g. is the right API used, is this in line
  with current refactoring efforts.

### Backend
New features:

- Is the new persistence stack used?

### Frontend
New features:

- What APIs are used to get the data? Is the REST API used?
- Should this be a separate library in a separate repo or should it be part of cBioPortal?
- Are dependencies properly listed? Ideally in a package.json
- How is the package included in cBioPortal?

### Devops
New features:

- Does the configuration style follow the config guidelines? That is compile.
- Runtime (Spring) goes in `application.properties`. Default values should be in `GlobalProperties.java`.
- Non-stable configuration should be done through war overlays.
- Is the configuration tested as part of the CI tests? It's not a necessity but be
  aware that untested configuration will be tough to maintain.
- Is there documentation on the proposed changes?

## Additional Resources

* [cBioPortal Issue Tracker](https://github.com/cBioPortal/cbioportal/issues)
* [General GitHub documentation](http://help.github.com/)
* [GitHub Pull Request documentation](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests)
# Tools for development

In this folder is some additional configuration that can be useful for local development. None of this should be deployed directly to production

# Set up keycloak for cBioPortal >v6

Requirements:
- System runs docker (including docker compose)

  1. Run from the root of the repository:

```
cd dev
docker compose up -d 
```

2. (Option 1) Apply SAML2 configuration to _security.properties_ in cBioPortal:

```properties
authenticate=saml
spring.security.saml2.relyingparty.registration.keycloak.assertingparty.metadata-uri=http://localhost:8084/realms/cbio/protocol/saml/descriptor
spring.security.saml2.relyingparty.registration.keycloak.assertingparty.entity-id=http://localhost:8084/realms/cbio
spring.security.saml2.relyingparty.registration.keycloak.entity-id=cbioportal
spring.security.saml2.relyingparty.registration.keycloak.signing.credentials[0].certificate-location=classpath:/dev/security/signing-cert.pem
spring.security.saml2.relyingparty.registration.keycloak.signing.credentials[0].private-key-location=classpath:/dev/security/signing-key.pem
```

3. (Option 2) Apply OIDC configuration to _security.properties_ in cBioPortal:

```properties
authenticate=oauth2
spring.security.oauth2.client.registration.keycloak.redirect-uri=http://localhost:8080/login/oauth2/code/keycloak
spring.security.oauth2.client.registration.keycloak.client-name=cbioportal_oauth2
spring.security.oauth2.client.registration.keycloak.client-id=cbioportal_oauth2
spring.security.oauth2.client.registration.keycloak.client-secret=client_secret
spring.security.oauth2.client.registration.keycloak.authorization-grant-type=authorization_code
spring.security.oauth2.client.registration.keycloak.client-authentication-method=client_secret_post
spring.security.oauth2.client.registration.keycloak.scope=openid,email,roles
spring.security.oauth2.client.provider.keycloak.issuer-uri=http://localhost:8084/realms/cbio
spring.security.oauth2.client.provider.keycloak.user-name-attribute=email
```

4. Set the following in _application.properties_:

```properties
persistence.cache_type=no-cache
session.service.url=http://localhost:5000/api/sessions/my_portal/

spring.datasource.url=jdbc:mysql://localhost:3306/cbioportal?useSSL=false&allowPublicKeyRetrieval=true
spring.datasource.username=cbio_user
spring.datasource.password=somepassword
spring.jpa.database-platform=org.hibernate.dialect.MySQL5InnoDBDialect
spring.datasource.driver-class-name=com.mysql.jdbc.Driver
```

4. Start cBioPortal application on port 8080. The login credentials are `testuser:P@assword1`.

‚ö†Ô∏è Warning: Do not use this directly for production use as it takes several shortcuts to get a quick keycloak instance up.
See [../docs/docker/README.md](../docs/docker/README.md)
# File Not Found

<script>
window.redirectMap = {
    "5.1-data-loading/data-loading/file-formats":"file-formats",
    "2.1-deployment/architecture-overview":"architecture-overview",
    "6.-web-api-and-clients/api-and-api-clients":"web-api-and-api-clients",
    "2.2-authorization-and-authentication/authenticating-users-via-tokens":"deployment/authorization-and-authentication/authenticating-users-via-tokens",
    "2.4-integration-with-other-webservices/oncokb-data-access":"deployment/integration-with-other-webservices/oncokb-data-access",
    "4.-development/release-procedure#a-note-on-versioning":"release-procedure",
    "2.1.2-deploy-without-docker/deploying":"deployment/deploy-without-docker/deploying",
    "2.2-authorization-and-authentication/authenticating-users-via-tokens":"deployment/authorization-and-authentication/authenticating-users-via-tokens",
    "1.-general/news":"news",
    "1.-general/faq":"user-guide/faq",
    "1.-general/about-us":"about-us",
    "1.-general/rfc-list":"rfc-list",
    "2.1-deployment/architecture-overview":"architecture-overview",
    "2.1-deployment/hardware-requirements":"hardware-requirements",
    "2.1.1-deploy-with-docker-recommended/docker":"deployment/docker",
    "2.1.1-deploy-with-docker-recommended/import_data":"deployment/docker/import_data",
    "2.1.1-deploy-with-docker-recommended/example_commands":"deployment/docker/example_commands",
    "2.1.1-deploy-with-docker-recommended/using-keycloak":"deployment/docker/using-keycloak",
    "2.1.2-deploy-without-docker/software-requirements":"deployment/deploy-without-docker/software-requirements",
    "2.1.2-deploy-without-docker/pre-build-steps":"deployment/deploy-without-docker/pre-build-steps",
    "2.1.2-deploy-without-docker/build-from-source":"deployment/deploy-without-docker/build-from-source",
    "2.1.2-deploy-without-docker/import-the-seed-database":"deployment/deploy-without-docker/import-the-seed-database",
    "2.1.2-deploy-without-docker/deploying":"deployment/deploy-without-docker/deploying",
    "2.1.2-deploy-without-docker/load-sample-cancer-study":"deployment/deploy-without-docker/load-sample-cancer-study",
    "2.2-authorization-and-authentication/user-authorization":"deployment/authorization-and-authentication/user-authorization",
    "2.2-authorization-and-authentication/authenticating-users-via-saml":"deployment/authorization-and-authentication/authenticating-users-via-saml",
    "2.2-authorization-and-authentication/authenticating-users-via-ldap":"deployment/authorization-and-authentication/authenticating-users-via-ldap",
    "2.2-authorization-and-authentication/authenticating-and-authorizing-users-via-keycloak":"deployment/authorization-and-authentication/authenticating-and-authorizing-users-via-keycloak",
    "2.2-authorization-and-authentication/authenticating-users-via-tokens":"deployment/authorization-and-authentication/authenticating-users-via-tokens",
    "2.3-customization/customizing-your-instance-of-cbioportal":"deployment/customization/customizing-your-instance-of-cbioportal",
    "2.3-customization/portal.properties-reference":"deployment/customization/application.properties-reference/",
    "deployment/customization/portal.properties-reference":"deployment/customization/application.properties-reference/",
    "2.3-customization/caching":"deployment/customization/caching",
    "2.3-customization/layout":"deployment/customization/studyview/#how-does-the-study-view-organize-the-charts",
    "2.3-customization/priority":"deployment/customization/studyview/#study-view-customization-with-priority-data",
    "2.4-integration-with-other-webservices/oncokb-data-access":"deployment/integration-with-other-webservices/oncokb-data-access",
    "3.-cbioportal-maintenance/updating-your-cbioportal-installation":"updating-your-cbioportal-installation",
    "3.-cbioportal-maintenance/updating-gene-and-gene_alias-tables":"updating-gene-and-gene_alias-tables",
    "4.-development/feature-development-guide":"feature-development-guide",
    "4.-development/cbioportal-er-diagram":"cbioportal-er-diagram",
    "4.-development/build-different-frontend":"build-different-frontend",
    "4.-development/session-service-working":"ession-service-working",
    "4.-development/manual-test-cases":"manual-test-cases",
    "4.-development/release-procedure":"release-procedure",
    "4.-development/deployment-procedure":"deployment-procedure",
    "5.1-data-loading/data-loading":"data-loading",
    "5.2-datasets/downloads":"downloads",
    "6.-web-api-and-clients/api-and-api-clients":"web-api-and-clients",
    "7.-user-guide/user-guide":"user-guide",
    "5.1-data-loading/data-loading/using-the-dataset-validator":"using-the-dataset-validator",
    "5.1-data-loading/data-loading/using-the-metaimport-script":"using-the-metaimport-script",
    "5.1-data-loading/data-loading/file-formats/z-score-normalization-script":"z-score-normalization-script",
    "5.1-data-loading/data-loading/data-loading-maintaining-studies":"data-loading-maintaining-studies",
    "5.1-data-loading/data-loading/data-loading-for-developers":"data-loading-for-developers",
    "5.1-data-loading/data-loading/data-loading-tips-and-best-practices":"data-loading-tips-and-best-practices",
    "5.1-data-loading/data-loading/import-gene-sets":"import-gene-sets",
    "5.1-data-loading/data-loading/import-gene-panels":"import-gene-panels",
};

let newLocation = null;
Object.keys(redirectMap).forEach(function(path){
    if ((new RegExp(path + "$").test(window.location.pathname))) {
        let root = document.querySelector("[name=turbo-root]")?.content || "";
        root = root.replace("/","");
        newLocation = `${window.location.protocol}//${window.location.hostname}${window.location.port ? ":" : ""}${window.location.port}/${root}/${window.redirectMap[path]}${window.location.hash}`; 
    } 

});

if (newLocation) {
    window.location.href = newLocation;
} else {
    document.getElementById("file-not-found").innerHTML = ` 
    <h1>Document not found</h1>
    <p>We recently changed our link structure and you may have a defunct url.  Please search for your topic above.</p> 
    `;
    document.getElementById("file-not-found").style.display = "block";
}
</script>




<h2 id=""></h2>




The cBioPortal for Cancer Genomics was originally developed at [Memorial Sloan Kettering Cancer Center](https://www.mskcc.org/) (MSK). The [public cBioPortal site](https://www.cbioportal.org) is hosted by the [Center for Molecular Oncology](https://www.mskcc.org/research/molecular-oncology) at MSK. The cBioPortal software is now available under an open source license via [GitHub](https://github.com/cBioPortal/). The software is now developed and maintained by a multi-institutional team, consisting of MSK, the Dana Farber Cancer Institute, Princess Margaret Cancer Centre in Toronto, Children's Hospital of Philadelphia, Caris Life Sciences, [The Hyve](https://thehyve.nl) and [SE4BIO](https://www.se4.bio/) in the Netherlands, and Bilkent University in Ankara, Turkey.

## Memorial Sloan Kettering Cancer Center
* Benjamin Gross
* S Onur Sumer
* Hongxin Zhang
* Ritika Kundra
* Ino de Bruijn
* Robert Sheridan
* Angelica Ochoa
* Aaron Lisman
* Manda Wilson
* Avery Wang
* Ramyasree Madupuri
* Gaofei Zhao
* Xiang Li
* Bryan Lai
* Calla Chennault
* Anusha Satravada
* Rima AlHamad
* Nikolaus Schultz

## Dana-Farber Cancer Institute
* Ethan Cerami
* Tali Mazor
* Jeremy Easton-Marks
* Zhaoyuan (Ryan) Fu
* Augustin Luna
* James Lindsay
* Chris Sander

## Princess Margaret Cancer Centre, Toronto
* Prasanna Jagannathan
* Trevor Pugh

## Children's Hospital of Philadelphia
* Charles Haynes
* David Higgins
* Allison Heath
* John Maris
* Adam Resnick
* Miguel Brown

## Caris Life Sciences
* Jianjiong Gao
* Priti Kumari
* Karthik Kalletla

## The Hyve
* Oleguer Plantalech
* Pim van Nierop
* Sander Rodenburg
* Bas Leenknegt
* Elena G Lara
* Jessica Singh
* Matthijs Pon
* Tim Kuijpers
* Mirella Kalafati
* Sjoerd van Hagen

## SE4BIO
* Pieter Lukasse
* Ruslan Forostianov

## Bilkent University
* Ugur Dogrusoz
* Yusuf Ziya Ozgul

## Alumni
* Adam Abeshouse
* B Arman Aksoy
* M Salih Altun
* Istemi Bahceci
* Caitlin Byrne
* Hsiao-Wei Chen
* Ersin Ciftci
* Fred Criscuolo
* Leonard Dervishi
* Gideon Dresdner
* Andy Dufilie
* Catherine Del Vecchio Fitz
* Olivier Elemento
* Ziya Erkoc
* Arthur Goldberg
* Zachary Heins
* Michael Heuer
* Anders Jacobsen
* Peter Kok
* Erik Larsson
* Dong Li
* Divya Madala
* Tamba Monrose
* Irina Pulyakhina
* Pichai Raman
* M Furkan Sahin
* Kaan Sancak
* Alexandros Sigaras
* Luke Sikina
* Jing Su
* Yichao Sun
* Sander Tan
* Jiaojiao Wang
* Stuart Watt
* James Xu
* Dionne Zaal
* Kelsey Zhu
* Fedde Schaeffer
* Riza Nugraha
* Ngoc Nguyen
* Diana Baiceanu
* Paul van Dijk
* Kees van Bochove

## Funding for the cBioPortal for Cancer Genomics is or has been provided by:

### Current:
* NCI, through [ITCR grant NCI-U24CA274633](https://itcr.nci.nih.gov/funded-project/cbioportal-cancer-genomics) and [HTAN grant NCI-U24CA233243](https://humantumoratlas.org/)
* Marie-Jos√© and Henry R. Kravis Center for Molecular Oncology at MSK
* Dana Farber Cancer Institute
* American Association for Cancer Research through AACR Project GENIE
* Prostate Cancer Foundation
* The Cholangiocarcinoma Foundation
* Robertson Foundation

### Past:
* NCI, through [ITCR grant NCI-U24CA220457](https://itcr.nci.nih.gov/funded-project/cbioportal-cancer-genomics)
* Stand Up 2 Cancer
* The Ben & Catherine Ivy Foundation
* NCI, as a [TCGA Genome Data Analysis Center (GDAC)](https://tcga.cancer.gov/wwd/program/research_network/gdac.asp) (NCI-U24CA143840)
* NCRR, as the [National Resource for Network Biology (NRNB)](https://nrnb.org/) Research Resource (RR 031228-02)
* Starr Cancer Consortium
* Breast Cancer Research Foundation
* Adenoid Cystic Carcinoma Research Foundation
* POETIC Consortium
* Parker Institute for Cancer Immunotherapy
# Architecture Overview
cBioPortal consists of the following components:

- [backend](https://github.com/cBioPortal/cbioportal)
    - MySQL database
    - REST API written in Java Spring
    - Redis cache for storing frequently used queries (optional)
- [validator](https://github.com/cBioPortal/cbioportal/tree/master/core/src/main/scripts/importer)
  checks file formats before importing data into the database
- [frontend](https://github.com/cBioPortal/cbioportal-frontend)
  built with React, Mobx and Bootstrap
- [session service](https://github.com/cBioPortal/session-service) for storing
  user saved data such as virtual studies and groups
     - REST API written in Java Spring enabling retrieval and writing to the database
     - MongoDB database
- cBioPortal also uses the APIs from various [external services](#external-services) to provide more information about a variant

## Backend

The [backend](https://github.com/cBioPortal/cbioportal) is written in Java and
connects to a MySQL database to serve a REST API following the OpenAPI
specification (https://www.cbioportal.org/api/). Note that the repo where this
lives in (https://github.com/cBioPortal/cbioportal) also contains Java classes
to import data as well as the validator. The backend can be configured to
connect to a Redis cache to store database query results for improved performance.

The backend is organized as a multi-module Maven project.
See [cBioPortal backend code organization](Backend-Code-Organization.md).

## Validator
The
[validator](https://github.com/cBioPortal/cbioportal/tree/master/core/src/main/scripts/importer)
checks file formats before importing data into the database. There is a wrapper
script `metaImport.py` that validates the data and subsequently calls the
relevant Java classes to import the data.

## Session Service

The [session service](https://github.com/cBioPortal/session-service) is used
for storing user saved data such as virtual studies and groups. See the
[tutorials](https://www.cbioportal.org/tutorials) section to read more about
these features. Session service is a Java app that serves a REST API backed by
a Mongo database. The session service is served as a proxy through the
cBioPortal backend REST API. The backend is therefore the only component that
needs to be able to connect to it. The frontend does not connect to it
directly. 

## Frontend
The frontend is a single page app built with React, Mobx and Bootstrap. The
data gets pulled from the backend REST API. The frontend is by default included
with the backend so no extra setup is required. 

## External Services
cBioPortal uses the APIs from several external services to provide more
information about a variant:

- [OncoKB](#oncokb)
- [CIVIC](#civic)
- [Genome Nexus](#genome-nexus)
- [G2S](#g2s)

For privacy concerns see the section: [A note on privacy](#a-note-on-privacy).

### OncoKB
[OncoKB](https://www.oncokb.org) is a precision oncology knowledge base that
contains information about the effects and treatment implications of specific
cancer gene alterations. See the section [OncoKB Data Access](/deployment/integration-with-other-webservices/OncoKB-Data-Access.md)
for how to configure external OncoKB service.

### CIVIC
[CIVIC](https://civicdb.org) is a community-edited forum for discussion and
interpretation of peer-reviewed publications pertaining to the clinical
relevance of variants (or biomarker alterations) in cancer. For information on
how to deploy this service yourself see:
https://github.com/griffithlab/civic-server. It is also possible to disable
showing CIVIC in cBioPortal by setting `show.civic=false` in the
`application.properties` (See [application.properties reference](/deployment/customization/application.properties-Reference.md#civic-integration)).

### Genome Nexus
[Genome Nexus](https://www.genomenexus.org) is a comprehensive one-stop
resource for fast, automated and high-throughput annotation and interpretation
of genetic variants in cancer. For information on how to deploy this service
yourself see: https://github.com/genome-nexus/genome-nexus. For more
information on the various annotation sources and versions provided by Genome
Nexus see: https://docs.genomenexus.org/annotation-sources.

### G2S
[G2S (Genome to Structure)](https://g2s.genomenexus.org) maps genomic variants
to 3D structures. cBioPortal uses it on the mutations tab to show the variants
on a 3D structure. For information on how to deploy this service yourself see:
https://github.com/genome-nexus/g2s.

### A note on privacy

cBioPortal calls these services with variant information from the cBioPortal
database. It however does not send over information that links a variant to a
particular sample or patient. If this is a concern for your use case we recommmend
to deploy your own versions of these services. See the sections above to
linkouts for instructions on how to do this.
# cBioPortal Backend Code Organization

## Maven Project

The backend code is structured as a [Maven](https://maven.apache.org/index.html)
project. Included in the repository are [pom.xml files](https://maven.apache.org/pom.html#What_is_the_POM)
which define the code components and modules which make up the backend. The pom.xml
files also list dependencies on external open source libraries, and dependencies between cBioPortal Maven modules.
They also give instructions for how to build and package code components into a
deployable format.

### Refactoring

The code base is undergoing a long term refactoring effort which is nearing completion.

#### Web service refactoring

The original data distribution service (still in operation) was a custom servlet called
[webservice.do](https://github.com/cBioPortal/cbioportal/blob/v3.1.5/core/src/main/java/org/mskcc/cbio/portal/servlet/WebService.java)
This servlet takes a "cmd" argument which selects the category of data requested.
More details can be seen on the cBioPortal [webAPI](https://www.cbioportal.org/webAPI) tab.

An example request:
`curl -X GET "https://www.cbioportal.org/webservice.do?cmd=getGeneticProfiles&cancer_study_id=brca_tcga"`

The new data distribution approach follows a [RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer) approach,
mapping various URL paths to data services, allowing the specification of database
entities within the path as helpful.

An example request:
`curl -X GET "https://www.cbioportal.org/api/studies/brca_tcga/molecular-profiles"`

#### Presentation / frontend refactoring

The original presentation code was written as a set of java service pages ([JSP](https://en.wikipedia.org/wiki/JavaServer_Pages)),
connected to servlets which populated needed data for page rendering into request attributes
within the backend server. You can see examples of the setting of these request attributes in
[this example servlet](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/servlet/QueryBuilder.java)

This has largely been superceded and expanded on by a new [React](https://reactjs.org/)
javascript application located in a separate
[GitHub repository](https://github.com/cBioPortal/cbioportal-frontend), which maintains a
browser data store populated from the new RESTful Web API. Although the new frontend
project is not dependent on the original custom servlets for data delivery, we may continue
to use the servlet / JSP delivery path to embed configuration data in the initial HTML page
in order to allow us to optimize loading times. Some of this logic has already been
migrated out of the core module and moved to the portal module.

### Current Maven Modules

#### The core module

The [core](https://github.com/cBioPortal/cbioportal/blob/master/core) module contains the oldest code in the code base. Much of
the backend web data services logic has been re-implemented and expanded in
other backend maven modules, such as [model](https://github.com/cBioPortal/cbioportal/blob/master/model),
[persistence-mybatis](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-mybatis),
[service](https://github.com/cBioPortal/cbioportal/blob/master/service), and [web](https://github.com/cBioPortal/cbioportal/blob/master/web). We plan to discontinue the legacy
webservice.do data servlet and fully transition to the new data services API in
the near future. The core module will then be purged of much of the legacy
code. Any surviving functionality (such as the handling of global configuration
properties) may be relocated into new modules. It is therefore important that
no new code features be introduced which are dependent on the core module
functionality.

#### The central stack modules : web, service, persistence

Three maven modules make up the central stack of the new web API implementation.

* [web](https://github.com/cBioPortal/cbioportal/blob/master/web) : define Web API request handlers, map endpoints, select services
* [service](https://github.com/cBioPortal/cbioportal/blob/master/service) : call persistence module or utils, apply business logic
* [persistence](https://github.com/cBioPortal/cbioportal/blob/master/persistence) : retrieve data from database. Two submodules:
  * [persistence-api](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api) : declaration of repository classes / functions, caching markup
  * [persistence-mybatis](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-mybatis) : implementation of persistence-api using [MyBatis](https://mybatis.org/mybatis-3/) mappers

#### Other modules

* [model](https://github.com/cBioPortal/cbioportal/blob/master/model) : data model POJO classes (used throughout the stack)
* [security](https://github.com/cBioPortal/cbioportal/blob/master/security) : user authentication methods and authorization, request filters 
* [business](https://github.com/cBioPortal/cbioportal/blob/master/business) : a legacy refactor of webservice.do, has active dataSource definition
* [portal](https://github.com/cBioPortal/cbioportal/blob/master/portal) : web application packaging and launch
* [scripts](https://github.com/cBioPortal/cbioportal/blob/master/scripts) : data import tool packaging (from core module scripts package)
* [db_scripts](https://github.com/cBioPortal/cbioportal/blob/master/db_scripts) : installation and migration scripts for the database

#### External code modules

* [cbioportal-frontend](https://github.com/cBioPortal/cbioportal-frontend) : a React application using MobX and TypeScript
* [session-service](https://github.com/cBioPortal/session-service) : an external session key/query specifier storage system

**cbioportal-frontend** is packaged in the web application as a default frontend
implementation, but the source of the frontend code can also be directed to an external
source host and be deployed independently of the backend web application. See
[details](Deployment-Procedure.md)

**session-service** is imported into the web module in order to set up a proxy service
which receives and forwards requests for saved cBioPortal sessions to a separate system
providing this storage (using a document based database) The code is needed for handling
modeled types such as VirtualStudy and Session

### cBioPortal Module Dependencies

![Module Dependencies](images/maven-module-dependencies.png)

_A module is directly dependent on all modules which touch it from below (touching only at a corner
does not count). Dependencies are transitive : web is dependent on service, service
is dependent on persistence-mybatis, therefore web is also dependent on
persistence-mybatis. During packaging, all dependencies are made available to the
dependent package for use. Avoid creating new dependencies between modules.
**cyclical dependencies will cause errors during build**_

Module | Dependencies
------ | ------------
portal | core, _cbioportal-frontend_
scripts | core
core | web 
web | business, service, _session-service_
security-spring | service
service | persistence-mybatis
persistence-mybatis | persistence-api, db-scripts
persistence-api | model
business | model
db-scripts |
model |
# cBioPortal Development Guidelines

1. Avoid introducing new functionality in the core module or creating new dependencies on the existing core module code.
1. One exception to the previous guideline is that new data import functions should be added to the core module scripts package.
1. Do not expand the business module.
1. Place all new web API controllers in the web module.
1. Make abstract data-driven web controllers part of the default/public interface (tag them with @PublicApi).
1. Make Special purpose or visualization-driven web controllers part of the internal interface (tag them with @InternalApi).
1. Do not include business logic in the web module handler functions. Limit processing to argument examination and service module method selection.
1. Test new data-driven web controllers for proper behavior on a portal deployment which requires user authentication and authorities.
1. Do not call persistence module functions directly from the web module. Create pass-through service layer functions instead.
1. Locate database query code in the persistence modules, and follow existing patterns.
1. Consider the tradeoffs between using database query constructs to accomplish business logic requirements versus writing service layer java code.
# Create and Publish Virtual Study

A [Virtual Study](./user-guide/faq.md#what-is-a-virtual-study) defines a subset or a combination of samples from one or more studies in the system.

*Note*: To publish or un-publish a virtual study, your cBioPortal instance must be configured with  `session.endpoint.publisher-api-key` in the `application.properties`.

## Create Virtual Study

To create a virtual study in cBioPortal, follow these steps:

1. Define the desired filters on the study or multiple studies summary page.
2. Click the button with the bookmark icon (ÔÄÆ) in the top right corner of the screen.
3. Provide a title and description, then click the Save button. You will see a link that looks like:

```
https://<cbioportal_host>/study?id=<virtual_study_id>
```

4. Save the virtual study link or ID if you want to publish it.

If you are logged in, this virtual study will appear in the `My Virtual Studies` section on the landing page.
You can always find the ID of the virtual study from the URL of the page that opens after clicking on it.

## Publish Virtual Study

To publish a virtual study, you need to supply the publisher API key in the `X-PUBLISHER-API-KEY` header.

Here is a curl command to publish a virtual study:
```shell
curl \
  -X POST \
  -H 'X-PUBLISHER-API-KEY: <session.endpoint.publisher-api-key>' \
  -v 'http://<cbioportal_host>/api/public_virtual_studies/<virtual_study_id>'
```
The published virtual study will appear under the `Public Virtual Studies` section (next to the `My Virtual Studies` section) on the landing page for all users of cBioPortal.

While publishing, you can specify the PubMed ID (`pmid`) and `typeOfCancerId` of the virtual study using the following command:
```shell
curl \
  -X POST \
  -H 'X-PUBLISHER-API-KEY: <session.endpoint.publisher-api-key>' \
  -v 'http://<cbioportal_host>/api/public_virtual_studies/<virtual_study_id>?pmid=<pmid>&typeOfCancerId=<code>'
```

The type of cancer code should match the known types of cancers in the cBioPortal database.
If the type of cancer is specified, the published virtual study will appear under the respective cancer section on the landing page.
Specifying the `pmid` enables a link to the PubMed page of the study.

## Un-publish Virtual Study

To un-publish a virtual study, you need to supply the publisher API key in the `X-PUBLISHER-API-KEY` header.
After un-publishing, virtual study will no longer be displayed in the `Public Virtual Studies` section on the landing page.
However, it reappears in the `My Virtual Studies` section for the owner.

Here is the command to un-publish a virtual study:
```shell
curl \
  -X DELETE \
  -H 'X-PUBLISHER-API-KEY: <session.endpoint.publisher-api-key>' \
  -v 'http://<cbioportal_host>/api/public_virtual_studies/<virtual_study_id>'
```# Importing single data files for development
In some cases, for example during development, it may be useful to import a single data file into an existing study. To import one data file at a time, you can use the following command. Note that this process will not validate the data.

This can be done by running `cbioportalImporter.py` from `<cbioportal_source_folder>/core/src/main/scripts/importer/`.

## Requirements

This script requires `$PORTAL_HOME` to point to the folder containing your
cBioPortal configuration. This can be done with:
```
export PORTAL_HOME=<cbioportal_configuration_folder>
```

## Workflow

First, if your cancer type does not yet exist, you need to create it:
```
./cbioportalImporter.py -c import-cancer-type -data <path to study directory>/<name of data file>
```

Next, create the study using
```
./cbioportalImporter.py -c import-study -meta <path to study directory>/<name of meta file>
```
The meta file has to contain the study information.

Now you can import your data file(s): 
```
./cbioportalImporter.py -c import-study-data -meta <path to study directory>/<name of meta file> -data <path to study directory>/<name of data file> 
```
:warning: 
Your first data file should always be the clinical data!

Finally, after you've imported all data, import your case lists:
```
./cbioportalImporter.py -c import-case-list -meta <path to study directory>/<path to case lists>
```

#### Example:
```
./cbioportalImporter.py -c import-cancer-type -data /data/brca_small/data_cancer_type.txt
./cbioportalImporter.py -c import-study -meta /data/brca_small/meta_study.txt
./cbioportalImporter.py -c import-study-data -meta /data/brca_small/meta_clinical.txt -data /data/brca_small/data_clinical.txt
./cbioportalImporter.py -c import-study-data -meta /data/brca_small/meta_expression.txt -data /data/brca_small/data_expression.txt
./cbioportalImporter.py -c import-case-list -meta /data/brca_small/case_lists
```
# Data loading: Importing without validation and deleting studies
For data curators and developers cbioportalImporter.py is available. This script can import data regardless of validation results. If data format is incorrect, the importer may stop with an error or crash, or leave the database in an inconsistent state.

This script can also be used to delete studies.

- [Requirements](#requirements)
- [Importing a study without validation](#importing-a-study-without-validation)
- [Deleting a study](#deleting-a-study)

## Requirements
This script requires `$PORTAL_HOME` to point to the folder containing your
cBioPortal configuration. This can be done with:
```
export PORTAL_HOME=<cbioportal_configuration_folder>
```

The script itself can be found in `<cbioportal_source_folder>/core/src/main/scripts/importer`.

## Importing a study without validation 
To import a study without validation, run: 
```
./cbioportalImporter.py -s <path to study directory>
```

For example:
```
./cbioportalImporter.py -s ../../../test/scripts/test_data/study_es_0/
```

## Deleting a study
To remove a study, run: 
```
./cbioportalImporter.py -c remove-study -meta <path to study directory>/meta_study.txt
```
The `meta_study.txt` file should contain the study ID in `cancer_study_identifier: ` of the study you would like to remove.

For example:
```
./cbioportalImporter.py -c remove-study -meta /data/brca_small/meta_study.txt
```

If you have the Cancer Study Id of the study, or studies you want to remove, you can also use:
```
./cbioportalImporter.py -c remove-study -id study1_id
```
Where `study1_id` is the Cancer Study Id of the study you would like to remove.

You can also remove multiple studies at once by passing the Cancer Study Ids separated by commas:
```
./cbioportalImporter.py -c remove-study -id study1_id,study2_id,study3_id
```
Where `study1_id`, `study2_id` and `study3_id` are the Cancer Study IDs of the studies you would like to remove.
Here we describe some Tips and Best Practices.

## Running GISTIC 2.0
To generate discrete copy number data file you may need to run GISTIC 2.0. GISTIC 2.0 can be [installed](https://www.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=216&p=t) or run online using the GISTIC 2.0 module on [GenePattern](https://cloud.genepattern.org). 
Running GISTIC 2.0 requires two input files: 

1. A segmentation file, which contains the segmented data 
2. A marker file, which identifies the marker names and positions of the markers in the original dataset (before segmentation). 

In some cases the marker file may not be available. Your can create one as follows:
Using your segmentation file, create a line for each start and end position. E.g. if your seg file contains

<table>
<tr>
<td>Sample</td><td>Chrom</td><td>Start</td><td>Stop</td><td>#Mark</td><td>Seg.CN</td>
</tr>
<tr>
<td>S1</td><td>1</td><td>61735</td><td>77473</td><td>10</td><td>-0.1234 </td>
</tr>
</table>

In your markerfile this becomes

<table>
<tr>
<td>Marker Name</td><td>Chrom</td><td>Pos</td>
</tr>
<tr>
<td>S1</td><td>1</td><td>61735</td>
</tr>
<tr>
<td>S1</td><td>1</td><td>77473</td>
</tr>
</table>


## Effect of cBioPortal instance on validation
When validating data, you can decide against which server to validate your data with the -u flag. The selected server can have a significant effect on the validation results in the following ways:

1. Genes may or may not be available on a specific server
2. Clinical data and its description may vary per server
3. ...

It is advised to use the server where you plan on upload your data as validation server. 
# Data Loading
## Introduction
This page is the starting point for data loading. The [General Overview](#general-overview) section below contains all the required steps to get you started. 

## General Overview
Getting your study data into cBioPortal requires four steps:

1. [Setting up the validator](#setting-up-the-validator)
2. [Preparing your study data](#preparing-study-data)
3. [Validating your study data](#validating-your-study-data)
4. [Loading your study data](#loading-data)

## Setting up the validator
### Installation

If you have a git clone of cBioPortal, the relevant scripts can be found in the folder: `<your_cbioportal_dir>/core/src/main/scripts/importer`

### Dependencies
The scripts run in Python 3.4 or newer, and they require the modules `requests` and `pyyaml`.
You can use this command to install those modules:
```console
$ sudo python3 -m pip install requests pyyaml
```

If you want the scripts to be able to generate html reports (recommended way for reading the validation errors, if any), then you will also need to install `Jinja2`. You can use this command:
```console
$ sudo python3 -m pip install Jinja2
```

## Preparing Study Data 
A study to be loaded in cBioPortal can basically consist of a directory where all the data files are located. 
Each *data* file needs a *meta* file that refers to it and both files need to comply to the format required for the specific data type. The format and fields expected for each file are documented in the [File Formats page](File-Formats.md). Below is an example of the files in such a directory.

```
dir
|-meta_study.txt
|-meta_cancer_type.txt -> cancer_type.txt
|-meta_clinical.txt -> data_clinical.txt
|-meta_[expression|mutations|CNA|etc] -> data_[expression|mutations|CNA|etc]
```
#### Rules
There are just a few rules to follow:
- meta_study, meta_clinical and respective clinical data file are the only **mandatory** files.
- cancer type files can be mandatory if the study is referring to a cancer type that does not yet exist in the DB.
- meta files can be named anything, as long as it starts or ends with name 'meta'. E.g. meta_test, meta.test, test.meta are all fine; metal_test and metastudy are wrong.
- data files can be named anything and are referenced by a property `data_filename` set in the meta file. 

## Validating your study data
Once all files are in place and follow the proper format, you can [validate your files using the dataset validator script](/Using-the-dataset-validator.md). 

The validation can be run standalone, but it is also integrated into the [metaImport script](/Using-the-metaImport-script.md), which validates the data and then loads it if validation succeeds. 

## Loading Data
To load the data into cBioPortal, the [metaImport script](/Using-the-metaImport-script.md) has to be used. This script first validates the data and, if validation succeeds, loads the data. 

## Removing a Study
To remove a study, the [cbioportalImporter script](/Data-Loading-Maintaining-Studies.md#deleting-a-study) can be used.

## Example studies
Examples for the different types of data are available on the [File Formats](/File-Formats.md) page. The Provisional TCGA studies, downloadable from the [Data Sets section](https://www.cbioportal.org/datasets) are complete studies that can be used as reference when creating data files.

## Public Virtual Studies
If your new study data is a subset or a combination of existing studies in the system, consider using [Public Virtual Studies](./Create-And-Publish-Virtual-Study.md) instead of duplicating data. # Authenticating and Authorizing Users via Keycloak

**‚ö†Ô∏è This documentation is for keycloak <v20, see related [ticket](https://github.com/cBioPortal/cbioportal/issues/10360) ‚ö†Ô∏è**

## Introduction

The cBioPortal includes support for Keycloak authentication. Keycloak can function as an Identity Provider (IDP) for cBioPortal.
This document explains why you might find Keycloak authentication useful for storing your user login information outside
the cBioPortal database. It also shows you how to configure Keycloak to communicate with your instance of cBioPortal using
SAML (Security Assertion Markup Language).

Please note that configuring your local instance to use Keycloak authentication requires a Keycloak server to be set up. For details on how to set up a Keycloak server, please read online document at <https://www.keycloak.org/guides#server>.

This document focuses mainly on the steps to configure Keycloak for **authenticating** cBioPortal users.

To skip to the authorization section see: [authorization with Keycloak](#authorization-with-keycloak). Or continue reading to learn how to integrate Keycloak with cBioPortal.

### Why Keycloak?

Keycloak is an [open source identity and access management solution](https://www.keycloak.org/docs/latest/server_admin/index.html). It has a built-in RDBM system to store login information. It can help build a security layer on top of the cBioPortal web application.

Keycloak boils down to three simple terms:
* **realm**: A realm secures and manages security metadata for a set of users, application, and registered auth clients.
* **client**: Clients are entities that can request authentication of a user within a realm.
* **role**: Roles identify a type or category of user. Keycloak often assigns access and permissions to specific roles rather than individual users for a fine-grained access control.

Keycloak offers three types of roles:
* Realm-level roles are in global namespace shared by all clients.
* Client roles have basically a namespace dedicated to a client.
* A composite role is a role that has one or more additional roles associated with it.


### How does Keycloak SAML Authentication work?

Keycloak supports both OpenID-Connect and SAML authentication. When you use SAML authentication, the Keycloak server exchanges XML documents with a web application. XML signatures and encryption are then used to verify requests from the application.

## Configure Keycloak to authenticate your cbioportal instance
1. Log in to your Keycloak Identity Provider, e.g. <http://localhost:8080/auth>, as an admin user. :warning: when setting this up on something else than localhost (e.g. production), you will need to use/enable https on your Keycloak server. For simplicity, the rest of the documentation below continues on http://localhost.
2. Hover over the top-left‚Äìcorner drop down menu (titled ‚Äò**Master**‚Äô) to create a new realm.
![](/images/previews/add-realm.png)
Please note if you are logged in the master realm, this drop-down menu lists all the realms created. The last entry of this drop-down menu is always **Add Realm**. Click this to add a realm. Then type '_cbioportal_' in the name field and click the **Create** button.
3. To create a SAML client, go to the **Clients** item in the left menu. On this page, click the **Create** button on the right. This will bring you to the **Add Client** page.
    * Enter a **Client ID** for the client, e.g. '_cbioportal_', this will be the expected `issuer` value in SAML requests sent by the application.
    * Select _saml_ in the **Client Protocol** drop down box.
    * Enter `http://localhost:8081/saml` in the **Client SAML Endpoint** textbox, this is the URL that the Keycloak server will send SAML requests and responses to. Then click the **Save** button; this will take you to the client page below.

![](/images/previews/edit-client.png)

4. Choose _email_ as your **Name ID Format**.
5. Next, enter a pattern for **Valid Redirect URIs** that Keycloak can use upon a successful authentication, e.g. `http://localhost:8081/*`. :information_source: notice that you can add multiple URLs in this field. You could use this in some cases to support
the URLs with and without port (e.g. if tomcat is running on port `80` and you want to allow both `http://localhost:80/*` and `http://localhost/*` as redirect URLs).
6. Set **Force POST Binding** and **Front Channel Logout** to _OFF_ and **Force Name ID Format** to _ON_.
7. Expand the subsection **Fine Grain SAML Endpoint Configuration** and set **Logout Service POST Binding URL** to `http://localhost:8081/saml/logout`.
8. Leave everything else as it is and click **Save**.

### Map SAML Assertion Attributes
To specify attributes included in the SAML assertion, simply click on the **Mappers** tab, and add mappers using the **Add Builtin** and **Create** buttons.
Make sure you add at least:
- the built-in User Property mapper named _X500 email_ and
- a _Role list_-type attribute (keep the default word _Role_ as its **Role attribute name**).
- a _User Property_-type attribute with the name _username_. Use _username_ for the attributes **Property**, **Name**, and **SAML Attribute Name**.

![](/images/previews/add-mappers.png)

Edit the email attribute to use the word _email_ as the **SAML Attribute Name**.
![](/images/previews/edit-email-attribute.png)

Finally, head to the **Scope** tab for the client and switch off
**Full Scope Allowed**, to ensure that only those roles relevant to a
particular cBioPortal instance are listed in assertions sent to the
instance, and not any other roles tracked in Keycloak.

### Export configuration for cBioPortal

There are two known ways to download the keycloak configuration (aka IDP SSO Descriptor) file for cBioPortal.

#### I. Download via link

The file can be fetched by the following url:

```
http(s)://{KEYCLOAK-URL}/auth/realms/{REALM-NAME}/protocol/saml/descriptor
```

For example:

```
curl -o client-tailored-saml-idp-metadata.xml "http://localhost:8081/auth/realms/cbioportal/protocol/saml/descriptor"
```

**Note:** if you use https protocol with self-signed protocol you need to add `--insecure` option to the above curl command.

#### II. Export via GUI (legacy)
1. Next, navigate to the **Installation** tab for the same client.
2. Select _SAML Metadata IDPSSODescriptor_ as the Format Option and click the **Download** button.
‚ö†Ô∏è This GUI option has been removed from the newer versions of Keycloak.
![](/images/previews/download-IDPSSODescriptor-file.png)

After you've downloaded the XML file with one of the above ways, move it to `portal/src/main/resources/` if you're compiling cBioPortal yourself or if you're using the Docker container, mount the file in the `/cbioportal-webapp` folder with `-v /path/to/client-tailored-saml-idp-metadata.xml:/cbioportal-webapp/WEB-INF/classes/client-tailored-saml-idp-metadata.xml`.

## Create a signing key for cBioPortal

Use the Java '`keytool`' command to generate keystore, as described
[here](Authenticating-Users-via-SAML.md#creating-a-keystore)
on the page about SAML in cBioPortal:

```
keytool -genkey -alias secure-key -keyalg RSA -keystore samlKeystore.jks
```

**Important:** The validity of this keystore is **90 days**. You can change the default
value by adding the `-validity` parameter and the number of days (e.g. `-validity 200`
for 200 days). If the keystore expires, then **'invalid requester'** errors are thrown.

Install the generated JKS file to `portal/src/main/resources/` if you're compiling cBioPortal yourself or if you're using the Docker container, mount the file in the `/cbioportal-webapp` folder with `-v /path/to/samlKeystore.jks:/cbioportal-webapp/WEB-INF/classes/samlKeystore.jks`.

Import the key's certificate into Keycloak, so that Keycloak knows that it can trust the holder of this
key. To do that, head to the **SAML Keys** tab in the keycloak admin screen about the `cbioportal` client and:
1. Click the **Import** button.
2. Select the _JKS_ archive format.
3. Specify the key alias _secure-key_.
4. Type the store password _apollo1_ (not the private key password, as Keycloak only
needs to know the certificate)
5. Select the file you just installed.

**Important:** Keycloak may not give an indication of successful
completion, but when navigating to the **SAML Keys** tab again you
should now see the certificate and no private key.

## Modifying configuration

1. Within the application.properties file , make sure that this line is present:
```
    app.name=cbioportal
```

2. Then, modify the properties under the comment `# authentication`. In particular, see the options listed in the example below:

```properties
    filter_groups_by_appname=false
    saml.sp.metadata.entityid=cbioportal
    saml.sp.metadata.wantassertionsigned=true
    saml.idp.metadata.location=classpath:/client-tailored-saml-idp-metadata.xml
    saml.idp.metadata.entityid=http://localhost:8080/auth/realms/cbioportal
    saml.keystore.location=classpath:/samlKeystore.jks
    saml.keystore.password=apollo1
    saml.keystore.private-key.key=secure-key
    saml.keystore.private-key.password=apollo2
    saml.keystore.default-key=secure-key
    saml.idp.comm.binding.settings=defaultBinding
    saml.idp.comm.binding.type=
    saml.idp.metadata.attribute.email=email
    saml.idp.metadata.attribute.userName=username
    saml.idp.metadata.attribute.role=Role
    saml.custom.userservice.class=org.cbioportal.security.spring.authentication.keycloak.SAMLUserDetailsServiceImpl
    # global logout (as opposed to local logout):
    saml.logout.local=false
    saml.logout.url=/
```

## Obtain user identities

### Optional: create users in Keycloak

To create a user, click on **Users** in the left menu bar. This menu
option brings you to the user list page. On the right side of the
empty user list, you should see an **Add User** button. Click that to
start creating your new user.

![](/images/previews/add-user.png)

### Optional: integrate company-wide authentication services

Keycloak can read credentials from existing user databases, for
instance over LDAP. This is referred to as _user federation_. Keycloak
can also allow authentication by an external login form altogether
using a protocol such as SAML, it calls this _identity brokering_. In
either case, Keycloak acts as a proxy between your user directory and
cBioPortal, deciding which authorities to grant when telling
cBioPortal that the user has authenticated.

Please refer to the Keycloak documentation on
[user federation](https://www.keycloak.org/docs/latest/server_admin/index.html#_user-storage-federation)
and
[identity brokering](https://www.keycloak.org/docs/latest/server_admin/index.html#_identity_broker)
for more information on how to integrate Keycloak with your local LDAP
or SAML service.

#### Federate LDAP/AD user directories

Some notes on user federation using LDAP/Active Directory:

By specifying the **Vendor** of your LDAP server, Keycloak will choose
sensible defaults for the required objectClasses and attributes of
your user entries. Apart from the Dedicated Name of the tree in which
to search for users and the DN and password that Keycloak should use
to bind to the server, make sure to specify the following **Custom
User LDAP Filter** to ensure that only user entries that have an email
address are considered:

```
(mail=*)
```

When using LDAP to load users from your institute's user directory,
you will most likely want Keycloak to refrain from trying to
synchronise changes in user details back to the central directory. You
should set **Edit Mode** to `READ_ONLY`, as the alternative `UNSYNCED`
would mean that users can be changed in the Keycloak database once
imported from LDAP, and start diverging. Also disable **Sync
Registrations** unless you want Keycloak to add new users to the LDAP
store.

Do turn on **Import Users** to make Keycloak remember users after the
first login, if you want to be able to assign non-default roles. If
the LDAP tree holding your users is large and you do not want to
import all users into Keycloak, make sure to disable **Periodic Full
Sync** and **Periodic Changed Users Sync**.

## Authorization with Keycloak

### Create roles to authorize cBioPortal users

The roles you assign to users will be used to tell cBioPortal which
studies a user is allowed to see.

To create a role, head to the **Roles** tab that is displayed along
the top while configuring the `cbioportal` client ‚Äì this tab is _not_
the link of the same name in the left sidebar.  Click the **Add
Role** button. Enter a name (e.g.  `brca_tcga_pub`) and description
for the role and hit the **Save** button.

![](/images/previews/add-role-for-study.png)

**Note:** if `filter_groups_by_appname` is set to `false` as specified above, the `Role Name` has to match with an id of the study you would give access to by assigning this role. Otherwise, if `filter_groups_by_appname` is set to `true` (**DEFAULT**), you have to add the application name (`app.name`) followed by the colon as a prefix to the study id. e.g. `cbioportal:brca_tcga_pub`

#### Groups

Keycloak allows you to create Groups for easy mapping of multiple
studies to multiple users. One can, for example, make a Keycloak group
with name `PUBLIC_STUDIES` and add all the individual Keycloak roles
corresponding to public studies to this group. It is also possible to
configure a group to be "default" in Keycloak, meaning new users are
automatically added to this group when logging in for the first time.

Alternatively, the Keycloak roles can correspond to the **groups** specified
in the [metadata files of studies](/File-Formats.md#cancer-study) instead
of corresponding to individual **study identifiers**. Although this will
result in less roles that need to be added and maintained in Keycloak,
it does result in group configuration being spread over both Keycloak
and meta study files.

### Assign roles to users

Next, assign roles to users. Head to **Users** in the left sidebar,
find a user (users from external providers should be known to Keycloak
after they have logged in for the first time), click the ID and
open the **Role Mappings** tab for that user. Select the
_cbioportal_ client in the dropdown under **Client Roles**, and use
the **Available Roles** selection and its **Add selected** button to
assign client roles to this user.

![](/images/previews/assign-role-to-user.png)

To automatically assign roles to all users when Keycloak first sees
them, the **Roles** pane accessed from the left sidebar has a
**Default Roles** tab. The interface for assigning roles here is much
the same as the one for assigning roles to individual users.

### Doing a Test Run

Rebuild the WAR file and follow the [Deployment with authentication
steps](/deployment/deploy-without-docker/Deploying.md#required-login) using `authenticate=saml`.

Then, go to:  [http://localhost:8081/](http://localhost:8081/).

If all goes well, the following should happen:

* You will be redirected to the Keycloak Login Page.
* After authenticating, you will be redirected back to your local instance of cBioPortal.

If this does not happen, see the Troubleshooting Tips  below.

### Add client for OAuth2 token-based data access

With cBioPortal instances that require user authentication the API can be queried when including a data access token in the request header (see [Authenticating Users via Tokens](/deployment/authorization-and-authentication/Authenticating-Users-via-Tokens.md)). KeyCloak can be configured as an OAuth2 authentication provider that distributes data access tokens to users and validates these tokens when used while querying the API. This feature is enabled by creating a `cbioportal_api` OpenID Connect client that has access to the user roles defined in the `cbioportal` SAML client.

The step below were verified to work with Keycloak versions 4.8.3.Final and 8.0.2.

1. Create a client with name `cbioportal_api`. Set _Client Protocol_ to `openid-connect`.

![](/images/previews/oauth2_client_1.png)

2. On the configuration page of `cbioportal_api` client apply the following settings:

#### Settings tab

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| Access Type      | confidential |  |
| Standard Flow Enabled      | ON      |   (default value) |
| Enabled      | ON      |   (default value) |
| Consent Required      | OFF        |   (default value) |
| Client Protocol      | openid-connect   |   (default value) |
| Access Type | confidential      |     (default value) |
| Standard Flow Enabled | ON      |     (default value) |
| Implicit Flow Enabled | OFF      |     (default value) |
| Direct Access Grants Enabled | OFF      |    |
| Service Accounts Enabled | OFF      |  (default value)   |
| Authorization Enabled | OFF      |  (default value)  |
| Valid Redirect URIs | _url_/api/data-access-token/oauth2  |  _url_ refers to base url of cBioPortal instance |

![](/images/previews/oauth2_client_3.png)

#### Credentials tab

Select `Client Id and Secret`. Take notice of the value of _Secret_ the secret field. This secret should be added to `application.properties` file of the cBioPortal backend.

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| Client Authenticator     | Client Id and Secret |   (default value) |

![](/images/previews/oauth2_client_6.png)

#### Client Scopes tab

Keep only scopes `roles` and `offline_access` (remove all others).

![](/images/previews/oauth2_client_2.png)

#### Mapper tab

Create a new _Audience_ mapper with name `cbioportal_api_audience`. This value will be used by the cBioPortal backend during validation of access tokens.

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| Name       | cbioportal_api_audience |  |
| Mapper Type     | Audience      |   |
| Included Client Audience      | cbioportal_api      |    |
| Add to ID token      | OFF        |  (default value)  |
| Add to access token  | ON      |   (default value)  |

![](/images/previews/oauth2_client_4.png)

#### Scope tab

Enable _Full Scope_. This setting will include the user roles defined in the `cbioportal` SAML client in access tokens distributed by the `cbioportal_api` client.

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| Full Scope Allowed       | ON | (default value) |

![](/images/previews/oauth2_client_5.png)

3. Add these parameters to `application.properties` of the cBioPortal backend.

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| dat.method       | oauth2 |  |
| dat.oauth2.clientId       | cbioportal_api |  |
| dat.oauth2.clientSecret    | ?      | see _Secret_ field in the _Credentials_ tab  |
| dat.oauth2.accessTokenUri     | _keycloak_url_/auth/realms/cbioportal/protocol/openid-connect/token      |   _keycloak_url_ refers to URL of the KeyCloak server from perspective of the cBioPortal instance |
| dat.oauth2.jwkUrl  | _keycloak_url_/auth/realms/cbioportal/protocol/openid-connect/certs      |   _keycloak_url_ refers to URL of the KeyCloak server from perspective of the cBioPortal instance |
| dat.oauth2.issuer      | _keycloak_url_/auth/realms/cbioportal        |  _keycloak_url_ refers to URL of the KeyCloak server from perspective of the browser |
| dat.oauth2.userAuthorizationUri  | _keycloak_url_/auth/realms/cbioportal/protocol/openid-connect/auth      |  _keycloak_url_ refers to URL of the KeyCloak server from perspective of the browser |
| dat.oauth2.redirectUri  | _cbioportal_url_/api/data-access-token/oauth2 | _cbioportal_url_ is url up to _/api_ path |
| dat.oauth2.jwtRolesPath  | '::'-separated path to array with user roles in JWT token returned by Keycloak | example: _resource_access::cbioportal::roles_ |

More information on configuration of the cBioPortal backend can be found in [Authenticating Users via Tokens](/deployment/authorization-and-authentication/Authenticating-Users-via-Tokens.md).

### Troubleshooting

#### Logging

Getting this to work requires many steps, and can be a bit tricky.  If you get stuck or get an obscure error message, your best bet is to turn on all DEBUG logging.
This can be done via `src/main/resources/logback.xml`. See [logback.DEBUG.EXAMPLE.xml](./logback.DEBUG.EXAMPLE.xml) file for an example of how to configure debug levels for cbioportal.

Then, rebuild the WAR, redeploy, and try to authenticate again.  Your log file will then include hundreds of SAML-specific messages, even the full XML of each SAML message, and this should help you debug the error.

If you're using the Docker container, mount the file instead with `-v ./logback.xml:/cbioportal-webapp/WEB-INF/classes/logback.xml`.

#### Determining jwtRolesPath for OAuth2 Token
By default user-roles are extracted from path `resource_access::cbioportal::roles` in the JWT json. Changes to the configuration of roles at the realm and client level in Keycloak instance can alter this path and must be set acordingly with the `dat.oauth2.jwtRolesPath` property in the `application.properties` file. 

To check the the roles path, go into the `Client Scopes` tab inside KeyCloak. Enter the `Evaluate` section, select a test user, and click `Evaluate`. In the section below, select the `Generated Access Token` tab to examine the JWT structure. 

![](/images/previews/oauth2_client_7.png)

A sample JWT might look like this:
```
{
  "exp": 1234567891,
  "iat": 1234567892,
  "jti": "transient-id",
  "iss": "issuer",
  "sub": "subject",
  "typ": "Bearer",
  "session_state": "sessionstate",
  "acr": "1",
  "realm_access": {
    "roles": [
      "all"
    ]
  },
  "scope": "openid"
}
```
The `jwtRolesPath` in this case would be `realm_access::roles`. Double check this against the `jwtRolesPath` value set in `application.properties`.

To connect cBioPortal to an external user database such as Active Directory will require the installation of [Keycloak](http://www.keycloak.org).  Please read the Wiki page on [Authenticating and Authorizing Users via Keycloak](Authenticating-and-Authorizing-Users-via-keycloak.md) for information on how to connect the cBioPortal with Keycloak.  You can also read how to connected Keycloak to Active Directory via LDAP on the [User Storage Federation](https://www.keycloak.org/docs/latest/server_admin/#_user-storage-federation) webpage of the Keycloak website.  
# Introduction

The cBioPortal includes support for SAML (Security Assertion Markup Language).  This document explains why you might find SAML useful, and how to configure SAML within your own instance of cBioPortal.

Please note that configuring your local instance to support SAML requires many steps.  This includes configuration changes and a small amount of debugging.  If you follow the steps below, you should be up and running relatively quickly, but be forewarned that you may have do a few trial runs to get everything working.

In the documentation below, we also provide details on how to perform SAML authentication via a commercial company:  [OneLogin](https://www.onelogin.com/).  OneLogin provides a free tier for testing out SAML authentication, and is one of the easier options to get a complete SAML workflow set-up.  Once you have OneLogin working, you should then have enough information to transition to your final authentication service.

## What is SAML?

SAML is an open standard that enables one to more easily add an authentication service on top of any existing web application.  For the full definition, see the [SAML Wikipedia entry](https://en.wikipedia.org/wiki/Security_Assertion_Markup_Language).

In its simplest terms, SAML boils down to four terms:

* **identity provider**:  this is a web-based service that stores user names and passwords, and provides a login form for users to authenticate.  Ideally, it also provides easy methods to add / edit / delete users, and also provides methods for users to reset their password.  In the documentation below, OneLogin.com serves as the identity provider.

* **service provider**:  any web site or web application that provides a service, but should only be available to authenticated and authorized users.  In the documentation below, the cBioPortal is the service provider.

* **authentication**:  a means of verifying that a user is who they purport to be.  Authentication is performed by the identify provider, by extracting the user name and password provided in a login form, and matching this with information stored in a database. When authentication is enabled, multiple cancer studies can be stored within a single instance of cBioPortal while providing fine-grained control over which users can access which studies.  Authorization is implemented within the core cBioPortal code, and *not* the identify provider.

## Why is SAML Relevant to cBioPortal?

The cBioPortal code has no means of storing user name and passwords and no means of directly authenticating users.  If you want to restrict access to your instance of cBioPortal, you therefore have to consider an external authentication service.  SAML is one means of doing so, and your larger institution may already provide SAML support.  For example, at Sloan Kettering and Dana-Farber, users of the internal cBioPortal instances login with their regular credentials via SAML.  This greatly simplifies user management.

# Setting up an Identity Provider

As noted above, we provide details on how to perform SAML authentication via a commercial company:  [OneLogin](https://www.onelogin.com/).  
If you already have an IDP set up, you can skip this part and go to [Configuring SAML within cBioPortal](#configuring-saml-within-cbioportal).

OneLogin provides a free tier for testing out SAML authentication, and is one of the easier options to get a complete SAML workflow set-up.  Once you have OneLogin working, you should then have enough information to transition to your final authentication service.  As you follow the steps below, the following link may be helpful: [How to Use the OneLogin SAML Test Connector](https://support.onelogin.com/hc/en-us/articles/202673944-How-to-Use-the-OneLogin-SAML-Test-Connector).

To get started:

* [Register a new OneLogin.com Account](https://www.onelogin.com/signup?ref=floating-sidebar)

## Setting up a SAML Test Connector

* [Login to OneLogin.com](https://app.onelogin.com/login).
* Under Apps, Select Add Apps.
* Search for SAML.
* Select the option labeled: OneLogin SAML Test (IdP w/attr).
* "SAVE" the app, then select the Configuration Tab.

* Under the Configuration Tab for OneLogin SAML Test (IdP w/attr), paste the following fields (this is assuming you are testing everything via localhost).

    * Audience: cbioportal
    * Recipient: http://localhost:8080/saml/SSO
    * ACS (Consumer) URL Validator*:  ^http:\/\/localhost\:8080\/cbioportal\/saml\/SSO$
    * ACS (Consumer) URL*:  http://localhost:8080/saml/SSO

![](/images/previews/onelogin-config.png)


* Add at least the parameters: 
    * Email (Attribute)
    * Email (SAML NameID)

![](/images/previews/onelogin-config-parameters.png)

* Find your user in the "Users" menu

![](/images/previews/onelogin-users-search.png)

* Link the SAML app to your user (click "New app" on the **+** icon found on the top right of the "Applications" table to do this - see screenshot below): 

![](/images/previews/onelogin-add-app.png)

* Configure these **email** parameters in the Users menu:

![](/images/previews/onelogin-users.png)
 


## Downloading the SAML Test Connector Meta Data

* Go to the SSO Tab within OneLogin SAML Test (IdP), find the field labeled:  Issuer URL.  Copy this URL and download it's contents.  This is an XML file that describes the identity provider.

![](https://s3.amazonaws.com/f.cl.ly/items/1R0v0L3P1U2E23202V1d/Image%202015-05-24%20at%2010.07.56%20PM.png)

then, move this XML file to:

    portal/src/main/resources/

You should now be all set with OneLogin.com.  Next, you need to configure your instance of cBioPortal.

# Configuring SAML within cBioPortal

## Creating a KeyStore

In order to use SAML, you must create a [Java Keystore](https://docs.oracle.com/javase/7/docs/api/java/security/KeyStore.html).  

This can be done via the Java `keytool` command, which is bundled with Java.

Type the following:

    keytool -genkey -alias secure-key -keyalg RSA -keystore samlKeystore.jks

This will create a Java keystore for a key called:  `secure-key` and place the keystore in a file named `samlKeystore.jks`.  You will be prompted for:

* keystore password (required, for example:  apollo1)
* your name, organization and location (optional)
* key password for `secure-key` (required, for example apollo2)

When you are done, copy `samlKeystore.jsk` to the correct location:
    
    mv samlKeystore.jks portal/src/main/resources/

If you need to export the public certificate associated within your keystore, run:

    keytool -export -keystore samlKeystore.jks -alias secure-key -file cBioPortal.cer

##### HTTPS and Tomcat

:warning: If you already have an official (non-self-signed) SSL certificate, and need to get your site 
running on HTTPS directly from Tomcat, then you need to import your certificate into the keystore instead. 
See [this Tomcat documentation page](https://tomcat.apache.org/tomcat-8.0-doc/ssl-howto.html) for more details.

:warning: An extra warning for when configuring HTTPS for Tomcat: use the same password for 
both keystore and secure-key. This seems to be an extra restriction by Tomcat.


## Modifying configuration

Within application.properties, make sure that:

    app.name=cbioportal

Then, modify the section labeled `authentication`. See SAML parameters shown in example below:

    saml.sp.metadata.entityid=cbioportal
    saml.sp.metadata.wantassertionsigned=true
    saml.idp.metadata.location=classpath:/onelogin_metadata_620035.xml
    saml.idp.metadata.entityid=https://app.onelogin.com/saml/metadata/620035
    saml.keystore.location=classpath:/samlKeystore.jks
    saml.keystore.password=apollo1
    saml.keystore.private-key.key=secure-key
    saml.keystore.private-key.password=apollo2
    saml.keystore.default-key=secure-key
    saml.idp.comm.binding.settings=defaultBinding
    saml.idp.comm.binding.type=
    saml.idp.metadata.attribute.email=User.email
    saml.idp.metadata.attribute.userName=User.name
    saml.custom.userservice.class=org.cbioportal.security.spring.authentication.saml.SAMLUserDetailsServiceImpl
    # global logout (as opposed to local logout):
    saml.logout.local=false
    saml.logout.url=/

Please note that you will have to modify all the above to match your own settings. `saml.idp.comm.binding.type` can be left empty if `saml.idp.comm.binding.settings=defaultBinding`. The `saml.logout.*` settings above reflect the settings of an IDP that supports Single Logout (hopefully the default in most cases - more details in section below).

In the case that you are running cBioPortal behind a reverse proxy that handles the SSL certificates (such as nginx or traefik), you will have to also specify `saml.sp.metadata.entitybaseurl`. This should point to `https://host.example.come:443`. This setting is required such that cBioPortal uses the Spring SAML library appropriately for creating redirects back into cBioPortal.

In addition there is a known bug where redirect from the cBioPortal instance always goes over http instead of https (https://github.com/cBioPortal/cbioportal/issues/6342). To get around this issue you can pass the full URL including https to the `webapp-runnner.jar` command with e.g. `--proxy-base-url https://mycbioportalinstance.org`.

### Custom scenarios

:information_source: Some settings may need to be adjusted to non-default values, depending on your IDP. For example, if your 
IDP required HTTP-GET requests instead of HTTP-POST, you need to set these properties as such:
 
    saml.idp.comm.binding.settings=specificBinding
    saml.idp.comm.binding.type=bindings:HTTP-Redirect

If you need a very different parsing of the SAML tokens than what is done at `org.cbioportal.security.spring.authentication.saml.SAMLUserDetailsServiceImpl`, you can point the `saml.custom.userservice.class` to your own implementation: 

    saml.custom.userservice.class=<your_package.your_class_name>

:warning: The properties `saml.idp.metadata.attribute.email`, and `saml.idp.metadata.attribute.userName` can also vary per IDP. It is important to set these correctly since these are a required field by the cBioPortal SAML parser (that is, if `org.cbioportal.security.spring.authentication.saml.SAMLUserDetailsServiceImpl` is chosen for property `saml.custom.userservice.class`). 

:warning: Some IDPs like to provide their own logout page (e.g. when they don't support the custom SAML Single Logout protocol). For this you can adjust the  
`saml.logout.url` property to a custom URL provided by the IDP. Also set the `saml.logout.local=true` property in this case to indicate that global logout (or Single Logout) is not supported by IDP:

    # local logout followed by a redirect to a global logout page:
    saml.logout.local=true
    saml.logout.url=<idp specific logout URL, e.g. https://idp.logoutpage.com >
    

:warning: Some IDPs (e.g. Azure Active Directory) cache user data for more than 2 hours causing cbioportal to complain that the authentication statement is too old to be used. You can fix this problem by setting `forceAuthN` to true. Below is an example how you can do this with the properties. You can choose any binding type you like. `bindings:HTTP-Redirect` is given just as an example.

```
    saml.idp.comm.binding.settings=specificBinding
    saml.idp.comm.binding.type=bindings:HTTP-Redirect
    saml.idp.comm.binding.force-auth-n=true
```

## More customizations

If your IDP does not have the flexibility of sending the specific credential fields expected by our 
default "user details parsers" implementation (i.e. `security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/saml/SAMLUserDetailsServiceImpl.java` 
expects field `mail` to be present in the SAML credential), then please let us know via a [new 
issue at our issue tracking system](https://github.com/cBioPortal/cbioportal/issues/new), so we can 
evaluate whether this is a scenario we would like to support in the default code. You can also consider 
adding your own version of the `SAMLUserDetailsService` class. 

## Authorizing Users

Next, please read the Wiki page on [User Authorization](User-Authorization.md), and add user rights for a single user.


## Configuring the Login.html Page (not applicable to most external IDPs)

The login page is configurable via the `application.properties` properties `skin.authorization_message` and `skin.login.saml.registration_htm`. 
For example in `skin.authorization_message` you can be set to something like this:

```
skin.authorization_message= Welcome to this portal. Access to this portal is available to authorized test users at YOUR ORG.  [<a href="https://thehyve.nl/">Request Access</a>].
```

and `skin.login.saml.registration_htm` can be set to:

```
skin.login.saml.registration_htm=Sign in via XXXX
```

You can also set a standard text in `skin.login.contact_html` that will appear in case of problems: 

```
skin.login.contact_html=If you think you have received this message in error, please contact us at <a style="color:#FF0000" href="mailto:cbioportal-access@your.org">cbioportal-access@your.org</a>
```


## Doing a Test Run

You are now ready to go.

Rebuild the WAR file and follow the [Deployment with authentication
steps](/deployment/deploy-without-docker/Deploying.md#required-login) using `authenticate=saml`.

Then, go to:  [http://localhost:8080/](http://localhost:8080/).

If all goes well, the following should happen:

* You will be redirected to the OneLogin Login Page.
* After authenticating, you will be redirected back to your local instance of cBioPortal.

If this does not happen, see the Troubleshooting Tips  below.

## Troubleshooting Tips 

### Logging

Getting this to work requires many steps, and can be a bit tricky.  If you get stuck or get an obscure error message,
your best bet is to turn on all DEBUG logging.  This can be done via `src/main/resources/logback.xml`.  For example:

```
<root level="debug">
    <appender-ref ref="STDOUT" />
    <appender-ref ref="FILE" />
</root>

<logger name="org.mskcc" level="debug">
    <appender-ref ref="STDOUT" />
    <appender-ref ref="FILE" />
</logger>

<logger name="org.cbioportal.security" level="debug">
    <appender-ref ref="STDOUT" />
    <appender-ref ref="FILE" />
</logger>
```

Then, rebuild the WAR, redeploy, and try to authenticate again.  Your log file will then include hundreds of SAML-specific messages, even the full XML of each SAML message, and this should help you debug the error.

### Seeing the SAML messages

Another tool we can use to troubleshoot is SAML tracer (https://addons.mozilla.org/en-US/firefox/addon/saml-tracer/ ). You can add this to Firefox and it will give you an extra menu item in "Tools". Go through the loging steps and you will see the SAML messages that are sent by the IDP. 

### Obtaining the Service Provider Meta Data File

By default, the portal will automatically generate a Service Provider (SP) Meta Data File.  You may need to provide this file to your Identity Provider (IP).

You can access the Service Provider Meta Data File via a URL such as:

[http://localhost:8080/saml/metadata](http://localhost:8080/saml/metadata)

# Authenticating Users via Tokens

The cBioPortal provides access to data within its database through the [cBioPortal web service](https://www.cbioportal.org/api/swagger-ui.html). While most users interact with this web service indirectly while visiting cBioPortal web pages, it is possible to access this web service directly. Direct access to the web service is appealing for those users with an interest in mining or analyzing cBioPortal data beyond the confines of a web browser, such as from ‚ÄòR‚Äô or Python clients.

If your instance of the cBioPortal will not require [user authorization](User-Authorization.md) (login), direct access to the web service is available without any additional server configuration or client side prerequisites. However, if your instance of the cBioPortal will require login, additional data access token server configuration is necessary before direct access to the web service is available to your users. Additionally, prior to web service access, clients will need to obtain a data access token and present this token with each web service request.

This document contains information for both cBioPortal administrators and users. For administrators, this document describes how to configure an instance of the cBioPortal with data access token support. For users, this document provides details about how to obtain a data access token and present it with a web service request (see [Using Data Access Tokens](Authenticating-Users-via-Tokens.md#using-data-access-tokens)).

## Configuring Data Access Tokens within cBioPortal

### Implementations

The cBioPortal codebase comes with three different data access token implementations, each of which come with strengths and weaknesses.

The first implementation based on [JSON Web Token](https://jwt.io/introduction/) (JWT) technology, is appealing because the data access token contains all the information required to validate the token during a web service request. When a data access token is generated by the cBioPortal, its expiration date is inserted into the token. The data access token is then signed and encrypted before presenting it to the user. When the data access token is presented in a subsequent web service request, the cBioPortal is able to inspect the token and determine its validity before satisfying the request. The downside to this implementation is that since the data access token is self-validating, no information about the token is stored in the database. This means there is no way to revoke a data access token in the event that it has been compromised or limit the number of data access tokens per user.

The second implementation based on [Universally unique identifier](https://en.wikipedia.org/wiki/Universally\_unique\_identifier) (UUID) technology, allows for greater control over the data access token at the cost of additional database support. Unlike JWT based data access tokens which are self-validating, a UUID based token is a string representation of a unique number. When a data access token is generated by the cBioPortal, a record is inserted into the `data_access_tokens` database table:

```
mysql> describe data_access_tokens;
+------------+--------------+------+-----+---------+-------+
| Field      | Type         | Null | Key | Default | Extra |
+------------+--------------+------+-----+---------+-------+
| TOKEN      | varchar(50)  | NO   | PRI | NULL    |       |
| USERNAME   | varchar(255) | NO   | MUL | NULL    |       |
| EXPIRATION | datetime     | NO   |     | NULL    |       |
| CREATION   | datetime     | YES  |     | NULL    |       |
+------------+--------------+------+-----+---------+-------+
4 rows in set (0.00 sec)
```

In addition to storing the token (the unique number), its creation and expiration dates along with the associated username is also stored prior to presenting it to the user. When the data access token is presented in a subsequent web service request, the cBioPortal compares the token presented in the request against this table for validity. If the token exists in the table and has not yet expired, the request will be satisfied. The creation date is used by the cBioPortal when a user requests a new data access token. Based on the server configuration, if the number of outstanding data access tokens together with this new one exceeds the number of allowable data access tokens per user, the cBioPortal may refuse to present a new one or it may invalidate the oldest one before presenting the new one to the user.

A third implementation makes use of an external OAuth2-authentication provider and can integrate with single-sign on (SSO) solutions. The OAuth2 provider provides users with an **offline token** that does not allow for direct access to protected cBioPortal resources. Instead, the offline token is passed to the OAuth2 authorization server that returns an **access token**, a short-lived permission to access cBioPortal resources. The access token contains up-to-date user permissions (the studies that the user is permitted to view). This implementation ensures that any changes of user permissions registered at the authentication provider are effective immediately and do not rely on the expiration date of the token of the user. At any moment OAuth2 offline tokens can be revoked by the Authentication provider on a per-user basis.

A step-by-step guide to configure KeyCloak to provide OAuth2 client functionality can be found in [Authenticating and Authorizing Users via keycloak](Authenticating-and-Authorizing-Users-via-keycloak.md#add-client-for-oauth2-token-based-data-access).

### Modifying Configuration

The following properties must be present in application.properties in order to allow direct access to the cBioPortal web service when login is required:

**Property**: dat.method (required)

* **Description**: The desired data access token implementation.
* **Permissible Values**: jwt, uuid, oauth2, none
* **Default Value**: none

**Property**: dat.filter\_user\_role (optional)

* **Description**: This property determines users access in token generation. If present, this role will be checked in user roles before generating a token.
* **Permissible Values**: A string value.
* **Default Value**: none

**Property**: dat.unauth\_users (optional, not used for dat.method = oauth2)

* **Description**: A list of users that should not be allowed to download a data access token.
* **Permissible Values**: A comma-delimited list of valid user names as found in the users table.
* **Default Value**: anonymousUser

**Property**: dat.ttl\_seconds (optional, not used for dat.method = oauth2)

* **Description**: The time in seconds between token creation and token expiration.
* **Permissible Values**: An integer value greater than zero.
* **Default Value**: 2592000 (30 days)

**Property**: dat.jwt.secret\_key (required only when dat.method = jwt)

* **Description**: The signature used to sign JWT based data access tokens during creation. This signature is used in subsequent web service requests to validate the request. This property is used only when dat.method = jwt. Keys used with the HS256 algorithm MUST have a size >= 256 bits. See https://tools.ietf.org/html/rfc7518#section-3.2 for more information.
* **Permissible Values**: A string value.
* **Default Value**: none

**Property**: dat.uuid.max\_number\_per\_user (required only when dat.method = uuid)

* **Description**: This property determines the number of data access tokens that can be allocated to a single user at any given time. If a user requests a data access token that puts them over this limit, the cBioPortal will present a new token to the user and it may invalidate the oldest one before presenting the new one to the user. This property is used only when dat.method = uuid.
* **Permissible Values**: An integer value greater than zero.
* **Default Value**: 1

**Property**: dat.oauth2.clientId (required only when dat.method = oauth2)

* **Description**: Identifier of the OAuth2 client of the authentication provider.
* **Permissible Values**: string
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.clientSecret (required only when dat.method = oauth2)

* **Description**: Secret of the OAuth2 client of the authentication provider used for back channel communication between cBioPortal and the provider. This value is specified by the authentication provider.
* **Permissible Values**: string
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.issuer (required only when dat.method = oauth2)

* **Description**: Name of the issuer of the tokens. It is used to assert the correct origin of the token. This value is specified by the authentication provider.
* **Permissible Values**: string
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.userAuthorizationUri (required only when dat.method = oauth2)

* **Description**: Url of the user login page of the OAuth2 authorization provider. According to OAuth2 specification this endpoint must end with `/auth`.
* **Permissible Values**: url
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.accessTokenUri (required only when dat.method = oauth2)

* **Description**: Url for retrieval of tokens from the OAuth2 authorization provider. According to OAuth2 specification this endpoint must end with `/token`.
* **Permissible Values**: url
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.jwkUrl (required only when dat.method = oauth2)

* **Description**: Url for retrieval of the public key used by the OAuth2 authorization provider to sign tokens. According to OAuth2 specification this endpoint must end with `/certs`.
* **Permissible Values**: url
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.jwtRolesPath (required only when dat.method = oauth2)

* **Description**: `::`-separated path to array with user roles in JWT token returned by Keycloak.
* **Permissible Values**: `::`-separated strings
* **Default Value**: resource\_access::cbioportal::roles

## Using Data Access Tokens

To directly access the web service of a cBioPortal instance which requires user login, clients will need to obtain a data access token and present this token with each web service request. To obtain a data access token, login to the cBioPortal and then choose ‚ÄúData Access Token‚Äù from the ‚ÄúLogged in as‚Ä¶‚Äù dropdown menu located on the top-right section of the webpage (**note this menu option will not be present if the instance of the cBioPortal has not be configured to support data access tokens**):

![](/images/previews/dat-download-menu.png)

This will bring you to a section of the Web API page where you can download your token using the "Download Token" button.

![](/images/previews/download-dat-button.png)

The downloaded data access token file will contain the token, and, for `uuid` and `jwt` token methods, a creation and expiration date:

```
token: 63efa81c-2490-4e15-9d1c-fb6e8e50e35d
creation_date: 2018-11-28T20:23:55.241Z
expiration_date: 2019-05-27T20:23:55.241Z
```

To directly access the web service, include the token as an **Authorization** header to the web service request:

```
Authorization: Bearer 63efa81c-2490-4e15-9d1c-fb6e8e50e35d
```

Here is an example curl command showing a request to GET /api/studies with the proper **Authorization** header set:

```
curl -X GET "https://www.cbioportal.org/api/studies" -H "accept: application/json" -H "Authorization: Bearer 63efa81c-2490-4e15-9d1c-fb6e8e50e35d"
```

If a web service request is made without an **Authorization** header or if the token within the header is not valid, the web service will not satisfy the request and will return with an HTTP Status 401. Additionally, there are a handful of properties which determine the behavior of cBioPortal data access tokens (see [Configuring Data Access Tokens within cBioPortal](Authenticating-Users-via-Tokens.md#configuring-data-access-tokens-within-cbioportal)). For example, a data access token may have an expiration date of 30 days after its creation date or 300 days. Similarly, there may be a limit to the number of data access tokens a user may have at any one time. If this limit is reached a new data access token may or may not be downloadable. Please check with the local cBioPortal system administrator for more information about how data access token behavior has been configured.
# User Authorization

This step is only required if you intend on running an instance of the portal that supports user authorization.

Two tables need to be populated in order to support user authorization.

## Table:  users

This table contains all the users that have authorized access to the instance of the portal.  The table requires a user's email address, name, and integer flag indicating if the account is enabled.

```
mysql> describe users;
+---------+--------------+------+-----+---------+-------+
| Field   | Type         | Null | Key | Default | Extra |
+---------+--------------+------+-----+---------+-------+
| EMAIL   | varchar(128) | NO   | PRI | NULL    |       |
| NAME    | varchar(255) | NO   |     | NULL    |       |
| ENABLED | tinyint(1)   | NO   |     | NULL    |       |
+---------+--------------+------+-----+---------+-------+
3 rows in set (0.00 sec)
```

An example entry would be:

```
mysql> select * from users where email = "john.smith@gmail.com";
+--------------------------+----------------+---------+
| EMAIL                    | NAME           | ENABLED |
+--------------------------+----------------+---------+
| john.smith@gmail.com     | John Smith     |       1 | 
+--------------------------+----------------+---------+
1 row in set (0.00 sec)
```

Note, if the ENABLED value is set to 0, the user will be able to login to the portal, but will see no studies.

You need to add users via MySQL directly.  For example:

```
INSERT INTO cbioportal.users (EMAIL, NAME, ENABLED) 
    VALUES ('john.smith@gmail.com', 'John Smith', 1);
```

## Table:  authorities

This table contains the list of cancer studies that each user is authorized to view.  The table requires a user email address and an authority (e.g. cancer study) granted to the user.

```
mysql> describe authorities;
+-----------+--------------+------+-----+---------+-------+
| Field     | Type         | Null | Key | Default | Extra |
+-----------+--------------+------+-----+---------+-------+
| EMAIL     | varchar(128) | NO   |     | NULL    |       | 
| AUTHORITY | varchar(50)  | NO   |     | NULL    |       | 
+-----------+--------------+------+-----+---------+-------+
2 rows in set (0.00 sec)
```

Some example entries would be:

```
mysql> select * from authorities where email = "john.smith@gmail.com";
+--------------------------+---------------------------+
| EMAIL                    | AUTHORITY                 |
+--------------------------+---------------------------+
| john.smith@gmail.com     | cbioportal:CANCER_STUDY_1 | 
| john.smith@gmail.com     | cbioportal:CANCER_STUDY_2 | 
| john.smith@gmail.com     | cbioportal:CANCER_STUDY_3 | 
+--------------------------+---------------------------+
5 rows in set (0.00 sec)
```
The value in the EMAIL column should be the same email address contained in the USER table.  

The value in the AUTHORITY column is made of two parts:

* The first part is the name of your portal instance.  This name should also match the `app.name` property found in the `application.properties` file.  
* Following a colon delimiter, the second part is the [cancer_study_identifier](../../File-Formats.md#cancer-study) of the cancer study this user has rights to access. 

**If the user has rights to all available cancer studies, a single entry with the keyword `app.name:` + "ALL" is sufficient (so e.g. "cbioportal:ALL").**

You need to add users via MySQL directly.  For example:

```
mysql> INSERT INTO cbioportal.authorities (EMAIL, AUTHORITY) VALUES
    ('john.smith@gmail.com', 'cbioportal:CANCER_STUDY_1');
```
**Important Note:**  The cancer study identifier is *not* CASE sEnsitive. So it can be UPPER CASE, or just how it is stored in the `cancer_study` table.
Changes to these tables become effective the *next* time the user logs in. 

## Using groups

It is also possible to define groups and assign multiple studies and users to a group. You can add a group name to the `cancer_study` table `GROUPS` column. This same group name can be used in the `AUTHORITY` column of the `authorities` table mentioned above. 

### Example: 

We want to create the group "TEST_GROUP1" and assign two existing studies to it and give our user 'john.smith@gmail.com' access to this group of studies. Steps:

1- Find your studies in table `cancer_study`
```
mysql> select CANCER_STUDY_ID, CANCER_STUDY_IDENTIFIER,GROUPS from cancer_study where CANCER_STUDY_ID in (93,94);
+-----------------+-------------------------+-------------+
| CANCER_STUDY_ID | CANCER_STUDY_IDENTIFIER | GROUPS      |
+-----------------+-------------------------+-------------+
|              93 | acc_tcga                | GROUPB      |
|              94 | brca_tcga               |             |
+-----------------+-------------------------+-------------+
```
2- Update the `GROUPS` field, adding your "TEST_GROUP1" to it. :warning: This is a `;` separated column, so if you want a study to be part of multiple groups, separate them with `;`.  
```
mysql> update cancer_study set GROUPS='TEST_GROUP1' where CANCER_STUDY_ID = 94;
```

If `GROUPS` already has a value (like for study 93 in example above) then add ";TEST_GROUP1" to ensure existing groups are not ovewritten.
```
mysql> update cancer_study set GROUPS=concat(GROUPS,';TEST_GROUP1') where CANCER_STUDY_ID = 93;
```
3- Check the result:
```
mysql> select CANCER_STUDY_ID, CANCER_STUDY_IDENTIFIER,GROUPS from cancer_study where CANCER_STUDY_ID in (93,94);
+-----------------+-------------------------+--------------------+
| CANCER_STUDY_ID | CANCER_STUDY_IDENTIFIER | GROUPS             |
+-----------------+-------------------------+--------------------+
|              93 | acc_tcga                | GROUPB;TEST_GROUP1 |
|              94 | brca_tcga               | TEST_GROUP1        |
+-----------------+-------------------------+--------------------+
```
4- Add the group to user 'john.smith@gmail.com', using `app.name:` + "TEST_GROUP1" like so:
```
mysql> INSERT INTO cbioportal.authorities (EMAIL, AUTHORITY) VALUES
    ('john.smith@gmail.com', 'cbioportal:TEST_GROUP1');
```

After **next login**, the user 'john.smith@gmail.com' will have access to these two studies. 


## Configuring PUBLIC studies

To enable a set of public studies that should be visible to all users, without the need to configure this for each user in the `authorities` and `users` tables, you can set the property `always_show_study_group` in **application.properties** file. For example, you can set:

```
always_show_study_group=PUBLIC
```

This will enable the word "PUBLIC" to be used in the column `GROUPS` of the table `cancer_study` to indicate which studies should be always shown to *any authenticated user*, regardless of authorization configurations.

### Example:

To reuse the example table above, let's assume the property `always_show_study_group` is set as indicated above and the `cancer_study` table contents are set to the following: 

```
+-----------------+-------------------------+--------------------+
| CANCER_STUDY_ID | CANCER_STUDY_IDENTIFIER | GROUPS             |
+-----------------+-------------------------+--------------------+
|              93 | acc_tcga                | GROUPB;TEST_GROUP1 |
|              94 | brca_tcga               | TEST_GROUP1;PUBLIC |
+-----------------+-------------------------+--------------------+
```

In this case, the study `brca_tcga` will be visible to *any authenticated user* while the study `acc_tcga` will be visible only to users configured to be part of `GROUPB` or `TEST_GROUP1`
# More application.properties Settings

This page describes the main properties within application.properties.

## Database Settings

```
db.user=
db.password=
db.connection_string=
db.driver=[this is the name of your JDBC driver, e.g., com.mysql.jdbc.Driver]
```

The format of the `db.connection_string` is:
```
jdbc:mysql://<host>:<port>/<database name>?<parameter1>&<parameter2>&<parameter...>
```

For example:

```
jdbc:mysql://localhost:3306/cbiodb?zeroDateTimeBehavior=convertToNull&useSSL=false
```

:warning: The fields `db.host` and `db.portal_db_name` and `db.use_ssl` are deprecated. It is required to configure the database connection using
the `db.connection_string` instead.

`db.tomcat_resource_name` is required in order to work with the tomcat database connection pool and should have the default value jdbc/cbioportal in order to work correctly with the your WAR file.

```
db.tomcat_resource_name=jdbc/cbioportal
```

## cBioPortal Customization

### Hide tabs (pages)

Settings controlling which tabs (pages) to hide. Set them to `false` if you want to hide those tabs, otherwise set the properties to `true`.

```
skin.show_data_tab=
skin.show_web_api_tab=
skin.show_r_matlab_tab=
skin.show_tutorials_tab=
skin.show_faqs_tab=
skin.show_news_tab=
skin.show_tools_tab=
skin.show_about_tab=
```

**Note:** `skin.show_tools_tab` refers to the `Visualize Your Data` tab, while `skin.show_data_tab` refers to the `Data Sets` tab.

### Cross Cancer Study Query Default

The cross cancer study query default is a list of studies used when querying one or more genes and not specifying a specific study or list of studies. There are two ways in which the default cross cancer study list is used:

1. When using the linkout links without a study e.g. `/ln?q=TP53:MUT`. Those links are used mostly used to allow for easy linking to particular queries. One can't get those links using the cBioPortal user interface itself, they are only mentioned in the documentation of the Web API (https://www.cbioportal.org/webAPI).
2. In the quick search when querying for a gene. Quick search is disabled by default. It is a beta feature. See the [quick search documentation](application.properties-Reference.md#quick-search-beta).

The configuration is set with the following if you have session service enabled:

```
default_cross_cancer_study_session_id=
```

The title will be pulled from the virtual study. Make sure to create a `virtual_study` with studies that everybody has access to and don't use a `main_session` id.

If session service is disabled one can use the following instead:

```
# query this comma separated list of studies
default_cross_cancer_study_list=
default_cross_cancer_study_list_name=
```

### Quick Search (BETA)

![Quick search example](/images/previews/quick\_search\_example.png)

Enable or disable the quick search with the following:

```
# Enable/Disable quick search (default is false)
quick_search.enabled=true
```

The default studies queried when searching for a single gene is defined with the `default_cross_cancer_study_session_id` or `default_cross_cancer_study_list` properties as described in the [cross cancer study query default section](application.properties-Reference.md#cross-cancer-study-query-default).

### Hide sections in the right navigation bar

Settings controlling what to show in the right navigation bar. Set them to `false` if you want to hide those sections, otherwise set the properties to `true`.

```
#Cancer Studies section:
skin.right_nav.show_data_sets=

#Example Queries section:
skin.right_nav.show_examples=

#Testimonials section:
skin.right_nav.show_testimonials=

#What's New section
skin.right_nav.show_whats_new=

#Twitter feed in What's New section
skin.right_nav.show_twitter=
```

#Interactive tours section
skin.right_nav.show_web_tours=

### Control the content of specific sections

Setting controlling the blurb: you can add any HTML code here that you want to visualize. This will be shown between the cBioPortal menu and the Query selector in the main page.

```
skin.blurb=
```

Setting controlling the citation below the blurb: you can add any HTML code here that you want to visualize. If the field is left empty, this HTML code will be shown: `Please cite: <a href="https://cancerdiscovery.aacrjournals.org/content/2/5/401.abstract" target="_blank">Cerami et al., 2012</a> &amp; <a href="https://www.ncbi.nlm.nih.gov/pubmed/23550210" target="_blank">Gao et al., 2013</a>`).

```
skin.citation_rule_text=
```

Setting controlling the footer: you can add any HTML code here that you want to visualize. If the field is left empty, the default footer (from www.cbioportal.org) will be shown.

```
skin.footer=
```

Setting controlling listing the development channels in the site's footer section

```
skin.footer_show_dev=
```

Settings controlling the "What's New" blurb in the right navigation bar: you can add any HTML code here that you want to visualize. If the field is left empty, the Twitter timeline will be shown (as long as `skin.right_nav.show_whats_new` is `true`, otherwise this section will not be displayed).

```
skin.right_nav.whats_new_blurb=
```

Add custom logos to the left or right side of the header section. Place here the full name of the logo file (e.g. `logo.png`). This file should be present in `$PORTAL_HOME/portal/images/`.

```
skin.left_logo=
skin.right_logo=
```

Prevent users from saving data by removing all Download tabs and download and copy-to-clipboard controls.
```
skin.hide_download_controls=
```



### Control default setting for filtering of genes in mutation and CNA tables of patient view

Different samples of a patient may have been analyzed with different gene panels. In patient view mutations and discrete CNA's can be filtered based on whether the gene of respective mutations/CNA's was profiled in all samples of the patient (mutations profiled in `all samples`), or not (mutations profiled in `any sample`). Setting this field to `true` will make patient view select the `all samples` filter at startup. When set to false or left blank the patient view will default to the `any samples` filter setting.

```
skin.patientview.filter_genes_profiled_all_samples=
```

### Control default settings of the VAF line chart in the genomic evolution tab of patient view
If you want to enable log scale and sequential mode by default, set this property to `true`:
```
vaf.log_scale.default=true|false
vaf.sequential_mode.default=true|false
```

### Control unauthorized studies to be displayed on the home page

By default, on an authenticated portal the home page will only show studies for which the current user is authorized. By setting the _skin.home\_page.show\_unauthorized\_studies_ property to _true_ the home page will also show unauthorized studies. The unauthorized studies will appear greyed out and cannot be selected for downstream analysis in Results View or Study View.

```
skin.home_page.show_unauthorized_studies=
```

If _show\_unauthorized\_studies_ feature has been enabled, a message (template) can be defined by the property below that informs the user of insufficient permissions. This message will appear inside a tooltip when hovering the lock icon next to the study name on the Study Selection page. This message may contain placeholders for study-specific information derived from [study tags data](/File-Formats.md#study-tags-file). The information in the study tags JSON file can be accessed using Json Path placeholders. For example `{$.Owner.email}` points to member `{Owner: {email: "me@myself.org"}}`. For the studies that don't have this information available in the study tags, the default message "The study is unauthorized. You need to request access." will be displayed. In addition to the study tags information, the cancer study identifier can be included in the message using {$.studyId} placeholder (does not have to be present in study tags file). For example: skin.home\_page.unauthorized\_studies\_global\_message=You do not have access to this study. You can request access with {$.Owner.email} (please mention the '{$.studyId}' study identifier).

```
skin.home_page.unauthorized_studies_global_message=
```

### Show badge with reference genome
In instances with hg19 and hg38 studies you can show the reference genome in the home page next to the number of samples. This can be done setting this property to `true` (`false` by default):

```
skin.home_page.show_reference_genome=
```

### Control the appearance of the settings menu in study view and group comparison that controls custom annotation-based filtering

A settings menu that allows the user to filter alterations in study view and group comparison may be used when [custom driver annotations](/File-Formats.md#custom-driver-annotations) were loaded for the study or studies displayed in these sections. This menu will only appear, when setting the property _skin.show\_settings\_menu_ to _true_.

```
skin.show_settings_menu=
```

### Hide logout button

When the user is logged-in, a button with logout and data access token options is shown at the right side of the header section. This button can be hidden by setting the property \_skin.hide\_logout\_button to _true_ (default is _false_). Hiding the logout button will remove the ability to perform a manual logout and download data access tokens by the user.

```
skin.hide_logout_button=
```

### Namespace columns visible in Mutation Table by default

Namespace columns are custom columns in the MAF file that can be shown in Mutation Table components. By setting this property to _true_ these columns will be visible when the mutation table is created. By default, the namespace column is hidden and can be made visible using the column selection menu.

```
skin.mutation_table.namespace_column.show_by_default=
```

### Default visible columns on init in Mutation, Copy-Number and Structural Variant Tables

Define the columns that are going to be visible in the Mutation, Copy-Number and Structural Variant Tables on the Patient View and the Mutation Table in the Results View. The `skin.mutation_table.namespace_column.show_by_default` takes precedence over these settings for namespace columns.
```
skin.results_view.mutation_table.columns.show_on_init=
skin.patient_view.mutation_table.columns.show_on_init=
skin.patient_view.copy_number_table.columns.show_on_init=
skin.patient_view.structural_variant_table.columns.show_on_init=
```

### Default sort columns on Mutation, Copy-Number and Structural Variant Tables

Define the column that are going to sort be default in the Mutation, Copy-Number and Structural Variant Tables on the Patient View and the Mutation Table in the Results View.
Column name should be exactly the same as shown in tables.
```
skin.results_view.tables.default_sort_column=
skin.patient_view.tables.default_sort_column=
```

### Define custom sample type colors
Define the colors of custom sample types in the patient view using a json object with for each sample type a color:
```
skin.patient_view.custom_sample_type_colors_json=classpath:/skin-patient-view-custom-sample-type-colors.json
```
Example of json file contents:
```json
{
    "Primary": "green",
    "Biopsy 3": "#00c040ff"
}
```

### Choose the display name for authenticated users

By default the display name for authenticated users is email, but it can be changed for the user name:

```
skin.user_display_name=username
```

### Hide p- and q-values in survival types table

```
# Show/hide p- and q-values in survival types table (default is true)
survival.show_p_q_values_in_survival_type_table=
```

### Set the initial x-axis limit for the survival plot

By default, the initial x-axis limit for a survival plot is the time of the latest event in the data. With this setting, you can make the initial x-axis limit be smaller than that. ![Survival x-axis limit examples](/images/previews/survival\_x\_axis\_limit.png)

```
# Set initial x-axis limit for survival plot (by default, initial limit will be the latest event in the data)
survival.initial_x_axis_limit=
```

### Display installation map

This setting specifies the URL for the iframes of a given portal's installation map on the homepage and standalone page. By default, the installation map configuration is commented out with the iframes and standalone page hidden. The configuration below is used for https://www.cbioportal.org/.

```
installation_map_url=https://installationmap.netlify.app/
```

To set up an installation map instance, one may consult the source code for the installation map [here](https://github.com/cbioportal/installation-map).


### Show structural variants table on study view

The structural variants table widget on study view allows users to define cohorts based on gene-orientation specific structural variant data format (see [here](/File-Formats.md#structural-variant-data)).
The structural variants table widget supports cohort selection based on _gene1_, _gene2_ and _gene1/gene2_ orientation specific genomic events.
This property enables the structural variants table widget on Study View.

:warning: Although _gene1_ and _gene2_ specific queries may be used to investigate up- and downstream fusion partners, respectively, the validity of this
depends on supports for this interpretation in the underlying data.

```
skin.study_view.show_sv_table=true
```


## Ensembl transcript lookup URL

The Mutations tab contains various links, redirecting the user to external information resources regarding the displayed transcript. The Ensembl template URL can be customized by modifying the property:

```
ensembl.transcript_url=
```

The default setting is `https://ensembl.org/homo_sapiens/Transcript/Summary?t=<%= transcriptId %>`. The `<%= transcriptId %>` is substituted by the frontend code into respective transcript ID.

## Segment File URL

This is a root URL to where segment files can be found. This is used when you want to provide segment file viewing via external tools such as [IGV](https://www.broadinstitute.org/igv/).

```
segfile.url=
```

## Bitly API Username and Key

The following properties are used to provide shortened bookmarks to the cBioPortal:

```
bitly.user=
bitly.api_key=
```

To obtain a bitly username and key, first register at: https://bitly.com/

Then, go to: https://bitly.com/a/your\_api\_key

**Note:** If you are developing on a local machine, and using localhost, the bitly URL shortening service will not work. This is because bitly will not shorten URLs for localhost. Once you deploy to your final server, the issue should go away.

## Google Analytics

When the following property is defined, Google Analytics will track site usage.

:warning: In contrast to what this property's name suggests, this should be the **GA Universal Analytics ID** (UA-XXXXX-YY), not to be confused with the new [GA-4 ID, introduced in 2020](https://support.google.com/analytics/answer/10089681). A Universal Analytics GA property can be created at [Google Analytics](https://analytics.google.com/analytics/web/provision/), create a new GA property, and make sure to click **Show advanced options** and switch on **Create a Universal Analytics property**. After creating a Universal Analytics property, the tracking ID can be found at **Admin**, column **property**, under **Property Settings**

:warning: Adblockers may block site tracking by GA

```
google_analytics_profile_id
```

## Password Authentication

The portal supports password authentication via Google+. Before you start you need to setup a google account that will own the authentication API. Follow https://developers.google.com/identity/sign-in/web/devconsole-project to get clientID and secret. Fill it in application.properties:

```
googleplus.consumer.key=195047654890-499gl89hj65j8d2eorqe0jvjnfaxcln0.apps.googleusercontent.com
googleplus.consumer.secret=2jCfg4SPWdGfXF44WC588dK
```

(note: these are just examples, you need to get your own) You will also need to go to "Google+ API" and click Enable button. In case of problems make sure to enable DEBUG logging for org.springframework.social and org.springframework.security.web.authentication.

To activate password authentication follow the [Deployment with authentication steps](/deployment/deploy-without-docker/Deploying.md#required-login) and set `authenticate=googleplus`.

In addition, set this property in `application.properties`:

```
app.name=cbioportal
```

app.name should be set to the name of the portal instance referenced in the "AUTHORITY" column of the "AUTHORITIES" table. See the [User Authorization](/deployment/authorization-and-authentication/User-Authorization.md) for more information.

## OncoKB integration

OncoKB integration can be turned on or off with the following property (default: true):

```
show.oncokb=true|false
```

A private token is required to access the OncoKB Data (for details see the section [OncoKB Data Access](/deployment/integration-with-other-webservices/OncoKB-Data-Access.md)):

```
oncokb.token=
```

"cBioPortal>=" driver annotation sources in the settings menu of Results View can be hidden by turning off the following property (default: true):

```
show.cbioportal=true|false
```

"COSMIC>=" driver annotation sources in the settings menu of Results View can be hidden by turning off the following property (default: true):

```
show.cosmic=true|false
```

## CIViC integration

CIViC integration can be turned on or off with the following property (default: true):

```
show.civic=true|false
```

The CIViC API url is set to https://civic.genome.wustl.edu/api/ by default. It can be overridden using the following property:

```
civic.url=
```

## Genome Nexus Integration

Genome Nexus provides annotations of mutations in cBioPortal. The mutations tab relies heavily on the Genome Nexus API, therefore that tab won't work well without it. By default cBioPortal will use the public Genome Nexus API, such that no extra configuration is necessary.

### Genome Build

Genome Nexus supports both GRCh37 and GRCh38, but support for the latter is limited. Several annotation sources served by Genome Nexus might not have official GRCh38 support yet i.e. [OncoKB](https://www.oncokb.org), [CIViC](https://civicdb.org), [Cancer Hotspots](https://www.cancerhotspots.org), [My Cancer Genome](https://www.mycancergenome.org) and [3D structures](https://g2s.genomenexus.org). Although most of the time the canonical transcript for a gene will be the same between GRCh37 and GRCh38 there might be some that cause issues. In addition the complete integration of cBioPortal with Genome Nexus' GRCh38 is not complete yet. That is cBioPortal currently only connects to one Genome Nexus API by default (the GRCh37 one), so it's not possible to have multiple genome builds in one instance of cBioPortal and get the correct annotations from Genome Nexus for both. Currently only the [mutation mapper tool page](https://www.cbioportal.org/mutation\_mapper) is able to handle both.

### Properties

By default the Genome Nexus API url is set to https://v1.genomenexus.org/, which uses GRCh37. It can be overridden using the following property:

```
genomenexus.url=
```

Genome Nexus provides a set of mappings from Hugo genes names to Ensembl transcript IDs. There are two mappings: `mskcc` and `uniprot`. 
You can read more about the difference between those in the [Mutation Data Annotation Section](./../../mutation-data-transcript-annotation.md). The default is currently `uniprot`, but we recommend new installers to use `mskcc` and people with older installations to consider migrating. The property
can be changed with:

```
genomenexus.isoform_override_source=mskcc
```

The mutation mapper tool page can annotate GRCh38 coordinates. By default it uses https://grch38.genomenexus.org. It can be overridden by setting:

```
genomenexus.url.grch38=
```

The GRCh38 annotation in mutation mapper can be hidden by setting `show.mutation_mappert_tool.grch38=false`, by default it's set to `true`.

## MDACC Heatmap Integration

MDACC Heatmap integration (button in OncoPrint heatmap dropdown and tab on Study page can be turned on or off by setting the following property:

```
show.mdacc.heatmap=true
```

## OncoPrint

The default view in OncoPrint ("patient" or "sample") can be set with the following option. The default is "patient".

```
oncoprint.defaultview=sample
```

Configuration of tracks that will be visible by default in the oncoprint. It points to a JSON file on the classpath.

```
oncoprint.clinical_tracks.config_json=classpath:/oncoprint-default-tracks.json
```

## Custom annotation of driver and passenger mutations

cBioPortal supports 2 formats to add custom annotations for driver and passenger mutations.

1. **cbp\_driver**: This will define whether a mutation is a driver or not.
2. **cbp\_driver\_tiers**: This can be used to define multiple classes of driver mutations.

These data formats are described in the [cBioPortal MAF specifications](/File-Formats.md#extended-maf-format).

**Enabling custom annotations in the OncoPrint**

To enable functionality for one or both types of custom annotations, enter values for the following properties. These values will appear in the OncoPrint's "Mutation color" menu, Patient View's (mutation, CNA, SV) tables, Results View's mutation table, and Group Comparison View's mutation table.

```
oncoprint.custom_driver_annotation.binary.menu_label=Custom Driver
oncoprint.custom_driver_annotation.binary.menu_description=Custom driver tiers
oncoprint.custom_driver_annotation.tiers.menu_label=Custom Driver Tiers
oncoprint.custom_driver_annotation.tiers.menu_description=Custom driver tiers
```

**Automatic selection of OncoKB, hotspots and custom annotations**

OncoKB and Hotspots are by default automatically selected as annotation source, if `show.oncokb` and `show.hotspots` are set to `true`. To add automatic selection of custom driver or custom driver tiers annotations, set the respective property to `true`. Default is `false`.

```
oncoprint.custom_driver_annotation.binary.default=true|false
oncoprint.custom_driver_annotation.tiers.default=true|false
```

If you want to disable the automatic selection of OncoKB and hotspots as annotation source, set these properties to `false`:

```
oncoprint.oncokb.default=true|false
oncoprint.hotspots.default=true|false
```

If you want to enable oncoprint heatmap clustering by default, set this property to `true`:
```
oncoprint.clustered.default=true|false
```

**Automatic hiding of variants of unknown significance (VUS)**

By default, the selection box to hide VUS mutations is unchecked. If you want to automatically hide VUS, set this property to `true`. Default is `false`.

```
oncoprint.hide_vus.default=true|false
```

## Gene sets used for gene querying

To change the gene sets used for gene querying, create a JSON file and add gene sets, following the format specified in the examples below. Set the path to this file (e.g. `file:/cbioportal/custom_gene_sets.json`) in the following property and restart Tomcat to apply the update. The default gene sets will be replaced by the ones in `custom_gene_sets.json`.

```
querypage.setsofgenes.location=file:/<path>
```

### Example with gene names

In this example, two gene sets will appear in the query page, under the names "Prostate Cancer: AR Signaling" and "Prostate Cancer: AR and steroid synthesis enzymes".

```
[{
	"id": "Prostate Cancer: AR Signaling",
	"genes": ["SOX9", "RAN", "TNK2", "EP300", "PXN", "NCOA2", "AR", "NRIP1", "NCOR1", "NCOR2"]
}, {
	"id": "Prostate Cancer: AR and steroid synthesis enzymes",
	"genes": ["AKR1C3", "AR", "CYB5A", "CYP11A1", "CYP11B1", "CYP11B2", "CYP17A1", "CYP19A1", "CYP21A2", "HSD17B1", "HSD17B10", "HSD17B11", "HSD17B12", "HSD17B13", "HSD17B14", "HSD17B2", "HSD17B3", "HSD17B4", "HSD17B6", "HSD17B7", "HSD17B8", "HSD3B1", "HSD3B2", "HSD3B7", "RDH5", "SHBG", "SRD5A1", "SRD5A2", "SRD5A3", "STAR"]
}]
```

### Example with specific alterations

In this example, only one gene set will appear in the query page, under the name "Genes with alterations", which will add the different genetic alterations stated below in the query box.

```
[{
	"id": "Genes with alterations",
	"genes": ["TP53: MUT=R273C;", "KRAS: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;"]
}]
```

### Example with merged gene tracks

In this example, only one gene set will appear in the query page, under the name "BRCA genes test", containing the merged gene track called "BRCA genes".

```
[{
	"id": "BRCA genes test",
	"genes": ["[\\\"BRCA genes\\\" BRCA1: MUT=E1258D;", "BRCA2: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;]"]
}]
```

This gene set will add the following in the query box:

```
"BRCA genes" BRCA1: MUT=E1258D; BRCA2: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;
```

## Cache Settings

cBioPortal is supported on the backend with Ehcache or Redis. These caches are configurable from within application.properties through the following properties.

The cache type is set using `persistence.cache_type`. Valid values are `no-cache`, `redis` (redis), `ehache-heap` (ehcache heap-only), `ehache-disk` (ehcache disk-only), and `ehache-hybrid` (ehcache disk + heap). By default, `persistence.cache_type` is set to `no-cache` which disables the cache. When the cache is disabled, no responses will be stored in the cache.

:warning: the 'redis' caching option will likely cause a conflict when installing the portal in a Tomcat installation which uses redisson for session management. If you plan to deploy cbioportal to such a system, avoid the 'redis' caching option for `persistence.cache_type` and be sure to build cbioportal.war with the maven option `-Dexclude-redisson` (see [Building with Maven](/deployment/deploy-without-docker/Build-from-Source.md#building-with-maven)).

```
persistence.cache_type=[no-cache or ehache-heap or ehcache-disk or ehcache-hybrid or redis]
```

Logged metrics and additional information such as cache size and cached keys are available through an optional endpoint. The optional endpoint is turned off by default but can be turned on by setting `cache.statistics_endpoint_enabled` to true.

```
cache.statistics_endpoint_enabled=false[true or false]
```

The cache statistics endpoint is hidden on the api page; users must directly access the URL to view the response. The cache statistics endpoint can be accessed in the following ways.

For a list of all keys in the cache:

```
/api/[name of cache]/keysInCache
```

For a list of counts of keys in cache per repository class:

```
/api/[name of cache]/keyCountsPerClass
```

For general statistics about the cache such as memory usage (not currently implemented for Redis):

```
/api/cacheStatistics
```

**WARNING**: It must be noted that since cache statistics endpoint returns data on cache keys, the endpoint may expose otherwise hidden database query parameters such as sample identifiers, study names, etc. Generally, it is recommended that the endpoint only be turned on during cache-related development for testing. Deployers of a protected portal where users only have authorities to a subset of studies should carefully consider whether or not to turn on the cache statistics endpoint, as it does not filter the results.

For more information on how caching is implemented in cBioPortal refer to the [Caching](/deployment/customization/Caching.md) documentation.

### Redis

To cache with Redis set `persistence.cache_type` to `redis`.

To setup the Redis cache servers the following properties are required:

```
redis.name=unique_cbioportal_instance_name
redis.leader_address=redis://{host/servicename}:6379
redis.follower_address=redis://{host/servicename}:6379
redis.database=0
redis.password=password
```

If you are running one redis instance for multiple instances of cBioPortal, one can use the properties `redis.name` and `redis.database` to avoid clashes. If you are running only one instance of cBioPortal any value for name/database will do.

There are also some optional parameters:

`redis.clear_on_startup`: If `true`, the caches will clear on startup. This is important to do to avoid reading old study data from the cache. You may want to turn it off and clear redis yourself if you are running in a clustered environments, as you'll have frequent restarts that do not require you to clear the redis cache.\
`redis.ttl_mins`: The time to live of items in the general cache, in minutes. The default value is 10000, or just under 7 days.

For more information on Redis, refer to the official documentation [here](https://redis.io/documentation)

### Ehcache

To cache with Ehcache set `persistence.cache_type` to `ehache-heap` (ehcache heap-only), `ehache-disk` (ehcache disk-only), or `ehache-hybrid` (ehcache disk + heap).

Ehcache initializes caches using a template found in an Ehcache xml configuration file. When caching is enabled, set `ehcache.xml_configuration` to the name of the Ehcache xml configuration file. The default provided is `ehcache.xml`; to change the cache template, directly edit this file. Alternatively, you can create your own Ehcache xml configuration file, place it under `/persistence/persistence-api/src/main/resources/` and set `ehcache.xml_configuration` to `/[Ehcache xml configuration filename]`.

```
ehcache.xml_configuration=
```

If the cache is configured to use disk resources, users must make a directory available and set it with the `ehcache.persistence_path` property. Ehcache will create separate directories under the provided path for each cache defined in the ehcache.xml\_configuration file.

```
ehcache.persistence_path=[location on the disk filesystem where Ehcache can write the cache to /tmp/]
```

Cache size must be set for heap and/or disk depending on which are in use; Ehcache requires disk size to be greater than heap size in a hybrid configuration. Zero is not a supported size and will cause an exception. Units are in megabytes. Default values are provided. The general repository cache is specified to use 1024MB of heap and 4096MB of disk. The static repository cache is specified to use 30MB of heap and 32MB of disk. For installations with increased traffic or data, cache sizes can be increased to further improve performance.

```
ehcache.general_repository_cache.max_mega_bytes_heap=
ehcache.general_repository_cache.max_mega_bytes_local_disk=

ehcache.static_repository_cache_one.max_mega_bytes_heap=
ehcache.static_repository_cache_one.max_mega_bytes_local_disk=
```

For more information on Ehcache, refer to the official documentation [here](https://www.ehcache.org/documentation/3.7/index.html)

## Evict caches with the /api/cache endpoint

`DELETE` http requests to the `/api/cache` endpoint will flush the cBioPortal caches, and serves as an alternative to restarting the cBioPortal application.

By default the endpoint is disabled. The endpoint can be enabled by setting:

```
cache.endpoint.enabled=true
```

Access to the endpoint is not regulated by the configured user authorization mechanism. Instead, an API key should be passed with the `X-API-KEY` header. The accepted value for the API key can be configured by setting (for example):

```
cache.endpoint.api-key=7d70fecb-cda8-490f-9ea2-ef874b6512f4
```

## Delegate user-authorization cache to Spring-managed cache

For evaluation fo user permissions cBioPortal uses a user-authorization cache that is populated at startup. By setting the `cache.cache-map-utils.spring-managed` property to _true_ this cache will be managed by the Spring caching solution such as EHCache or Redis. For more extended information, see [here](Caching.md#user-authorization-cache)

```
cache.cache-map-utils.spring-managed=false
```

## Enable GSVA functionality

[GSVA functionality](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#gene-set-data) can be enabled by uncommenting this line (and making sure it is set to `true`):

```
skin.show_gsva=true
```

## Set default thresholds for geneset hierarchy

```
skin.geneset_hierarchy.default_gsva_score=0.3
skin.geneset_hierarchy.default_p_value=0.02
```

## Collapses the tree widget of the geneset hierarchy dialog on initialization

By default, the tree is expanded (property value is `false`).

```
skin.geneset_hierarchy.collapse_by_default = true
```

## Cross study expression and protein data


By default we hide expression data for multi-study queries as they are 
usually not normalized across studies. For the public cBioPortal for instance, only TCGA Pancan Atlas studies expression data has been normalized.  

If you know the expression data in your instance is comparable, or is comparable for a subset of
studies, you can configure a rule as follows. 

The value of this property can be boolean (true|false) or a javascript function
which executes at runtime and is passed the list of study objects being queried
by the user and evaluates whether expression data can be safely displayed.

```
// a function that accepts the users's selected studies and
// returns whether or not to allow expression data from the involved studies to be mixed
enable_cross_study_expression = (selectedStudies)=>{ [your logic] return true|false }
```

```
// boolean
enable_cross_study_expression = true|false
```

## Combined Study View Summary Limits
### Background
A limit is added to prevent poor performance of Study View when selecting too large sample numbers.
### Properties
* `studyview.max_samples_selected`: Limit is disabled when not set

### Behavior
When these limits are exceeded the "Explore Selected Studies" button will be disabled on the Study View Page.

## Request Body Compression

### Background

Some REST endpoints that the cBioPortal frontend uses have request bodies that scale as your dataset increases. In portals where users commonly query more than 100,000 samples, we found that some of these request bodies could get as large as 20 Mb. These large request bodies pose a significant problem for users with poor upload speeds - some users experienced upload times of more than five minutes for these requests. Request body compression is our temporary solution to this problem. When this feature is toggled on, we compress the request bodies of a few problematic endpoints.

### Properties

There are two `portal.property` values related to this feature:

* `enable_request_body_gzip_compression`: when `true`, the feature will be enabled.
* `request_gzip_body_size_bytes`: the maximum allowable unzipped request body in bytes. Defaults to 80000000 (80 Mb).

### Behavior

* This is a nonbreaking change. Any consumers of the cBioPortal API you have that send requests with uncompressed request bodies will continue to work, regardless of whether you turn this feature on or off.
* If you turn this feature on, the cBioPortal API will now be able to handle any request with a gzipped request body, provided:
  * It is a POST request.
  * It has a `Content-Encoding: gzip` header.

### Reasons to Enable This Feature

* You have studies with tens of thousands of samples.
* You have users with poor upload speeds (< 1mb up).

### Reasons to Disable This Feature

* It is harder to debug gzipped requests
  * Chrome's `copy request as CURL` will not work.
  * The compressed request body is not human-readable.
* It is a potential vector for denial of memory attacks.
  * Any request that has a body that takes significantly more space in memory than it does in the request body is potentially dangerous. We try to mitigate this by limiting the size of the unzipped request body via the `request_gzip_body_size_bytes` property, but at a fundamental level, this is still a concern.
  * Along these lines, if you do enable this feature, setting `request_gzip_body_size_bytes` to an arbitrarily large number would be unwise.
* This is not a cure-all for performance issues
  * Most requests the cBioPortal makes do not have large request bodies, so most requests will not be compressed, and will see no performance improvement.
  * Users with good upload speeds will see minimal performance improvements, as their upload speed is not a bottleneck.

# DataSets Tab (Study Download Links)
### Background
The DataSets tab has the ability to create a ``download`` button that allows users to quickly download "raw" public studies.
### Properties
* ``study_download_url`` : when set, the feature will be enabled
### Behavior
For private instances that want to replicate the``public-portal`` they first must set up their studies 
they want available for download in a similar format to what is described in the Example section below.
The studies are located on the ``public-portal`` at  `https://cbioportal-datahub.s3.amazonaws.com/`. 
Then there is a ``study_list.json`` defined that list the studies that can be downloaded.
The studies to be downloaded need to be compressed with the extension ``tar.gz``

### Example
* We have set ``study_download_url`` property to `https://cbioportal-datahub.s3.amazonaws.com/`
* ``study_list.json`` resides ``https://cbioportal-datahub.s3.amazonaws.com/study_list.json``
  * ``[ "acbc_mskcc_2015", "acc_2019"]`` Example of contents
* ``acbc_mskcc_2015.tar.gz`` resides `https://cbioportal-datahub.s3.amazonaws.com/acbc_mskcc_2015.tar.gz`

# Prioritized studies on study selector view 
By default, the studies loaded into a local cBioPortal instance are organized based on their cancer type (i.e. Breast >> Other).
```
priority_studies=
```
The value of this variable will create a custom category with studies on the top of the study selector view. The format for the string should be category1#study1a,study1b,study1c;category2#study2 (e.g., PanCancer Studies#msk_impact_2017), where the ``category`` can be any string and the ``study`` should be the study ID of the required uploaded study. 

## Study Tag functionality
Study Tags allow portal maintainers to define miscellaneous descriptive meta data to studies, which will be shown to users in tooltips and are also searchable. This feature
is on by default but can be disabled using the following property.
```
//boolean
enable_study_tags=true|false
```
# Backend Caching

cBioPortal provides the option of caching information on the backend to improve performance. Without caching, every time
a request is received by the backend, a query is sent to the database system for information, and the returned data is
processed to construct a response. This may lead to performance issues as the entire process can be rather costly,
especially for queries on larger studies. With caching turned on, query responses can be taken directly from the cache
if they have already been constructed. They would only be constructed for the initial query.

## Cache Configuration

The portal is configured to use Ehcache or Redis for backend caching. Ehcache supports a hybrid (disk + heap),
disk-only, and heap-only mode. Redis stores the cache in memory and periodically writes the updated data to disk. Cache
configuration is specified inside `application.properties`(more
information [here](/deployment/customization/application.properties-Reference.md#cache-settings)).

## Creating additional caches

The default configuration initializes two separate caches; however, you may wish to introduce new caches for different
datatypes. Please see the [Redis](#redis) and [Ehcache](#ehcache) sections to see how to set up a new cache in whichever
system you are using.

### Redis

Cache initialization is handled inside
the [CustomRedisCachingProvider](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api/src/main/java/org/cbioportal/persistence/util/CustomRedisCachingProvider.java)
. To create additional caches (e.g creating a cache specifically for clinical data), new code must be added to
the `CustomRedisCachingProvider`.

Within the `CustomRedisCachingProvider`, create your new cache using the `CacheManager`. The appName must be prepended
to your cache name.

```
manager.createCache(appName + "ClinicalDataCache", config);
```

You also need to create a new cache resolver
in [RedisConfig.java](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api/src/main/java/org/cbioportal/persistence/config/RedisConfig.java):

```
@Bean
public CacheResolver generalRepositoryCacheResolver() {
    return new NamedCacheResolver(cacheManager(), ${redis.name:cbioportal} + "GeneralRepositoryCache");
}
```

The `@Cacheable` annotation must also be added (or adjusted) to function declarations to indicate which functions are to
be cached. Those might look like this example:

``` 
@Cacheable(cacheResolver = "clinicalDataCacheResolver", condition = "@cacheEnabledConfig.getEnabled()")
public String getDataFromClinicalDataRepository(String param) {}
```

For more information on linking caches to functions, refer to the
documentation [here](https://spring.io/guides/gs/caching/).

### Ehcache

Within the `CustomEhcachingProvider`, initialize a new `ResourcePoolsBuilder` for the new cache and set the resources
accordingly.

```
ResourcePoolsBuilder clinicalDataCacheResourcePoolsBuilder = ResourcePoolsBuilder.newResourcePoolsBuilder();
clinicalDataCacheResourcePoolsBuilder = clinicalDataCacheResourcePoolsBuilder.heap(clinicalDataCacheHeapSize, MemoryUnit.MB);
clinicalDataCacheResourcePoolsBuilder = clinicalDataCacheResourcePoolsBuilder.disk(clinicalDataCacheDiskSize, MemoryUnit.MB);
```

After initialzing the `ResourcePoolsBuilder`, create a CacheConfiguration for the new cache using the
new `ResourcePoolsBuilder` just created.

```
CacheConfiguration<Object, Object> clinicalDataCacheConfiguration = xmlConfiguration.newCacheConfigurationBuilderFromTemplate("RepositoryCacheTemplate", 
Object.class, Object.class, clinicalDataCacheResourcePoolsBuilder)
.withSizeOfMaxObjectGraph(Long.MAX_VALUE)
.withSizeOfMaxObjectSize(Long.MAX_VALUE, MemoryUnit.B)
.build();
```

Finally, add the new `CacheConfiguration` to the map of managed caches with a name for the cache.

```
caches.put("ClinicalDataCache", ClinicalDataCacheConfiguration);
```

You also need to create a new cache resolver
in [EhCacheConfig.java](https://github.com/cBioPortal/cbioportal/blob/master/persistence-api/src/main/java/org/cbioportal/persistence/config/EhCacheConfig.java):

```
@Bean
public NamedCacheResolver generalRepositoryCacheResolver() {
    return new NamedCacheResolver(cacheManager(), "GeneralRepositoryCache");
}
```

The `@Cacheable` annotation must also be added (or adjusted) to function declarations to indicate which functions are to
be cached. Those might look like this example:

``` 
@Cacheable(cacheResolver = "clinicalDataCacheResolver", condition = "@cacheEnabledConfig.getEnabled()")
public String getDataFromClinicalDataRepository(String param) {}
```

Additionally, new properties for setting cache sizes should be added to `application.properties` and loaded into
the [CustomEhcachingProvider](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api/src/main/java/org/cbioportal/persistence/util/CustomEhcachingProvider.java)
. Alternatively, values may be hardcoded directly
inside [CustomEhcachingProvider](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api/src/main/java/org/cbioportal/persistence/util/CustomEhcachingProvider.java)
.

For more information on cache templates and the Ehcache xml configuration file, refer to the
documentation [here](https://www.ehcache.org/documentation/3.7/xml.html).

For more information on linking caches to functions, refer to the
documentation [here](https://spring.io/guides/gs/caching/).


## User-authorization cache

In addition to the above-mentioned Spring-managed caches, cBioPortal maintains a separate cache that holds references to
sample lists, molecular profiles and cancer studies. This user-authorization cache is used to establish
whether a user has access to the data of a particular sample list or molecular profile based on study-level permissions.

By default, the user-authorization cache is implemented as a HashMap that is populated when cBioPortal is started. This
implementation allows for very fast response times of user-permission evaluation.

The user-authorization cache can be delegated to the Spring-managed caching solution by setting the [cache.cache-map-utils.spring-managed](application.properties-Reference.md#cache-settings)
to _true_. Depending on the implementation, this may add a delay to any data request that is caused by the additional consultation
of the external cache. This configuration should only be used where a central caching solution is required or no
instance/container-specific local caches are allowed. For example, cache eviction via the `api/cache` endpoint in a Kubernetes
deployment of cBioPortal where multiple pods/containers that represent a single cBioPortal instance is possible with a
Spring-managed user-authorization cache because a call to this endpoint in a single pod/container invalidates Redis caches
for the entire deployment thereby preventing inconsistent state of user-authorization caches between pods.


## Cache eviction

When the database is updated (e.g new studies loaded, existing study updated, new gene panel imported) the caches of a
cBioPortal instance should be updated. One way is to restart the cBioPortal spring application. When using Redis, this
will work only when `redis.clear_on_startup` is not set to _false_.

Alternatively, caches can be cleared (a.k.a. _evicted_) by calling the _/api/cache_ endpoint. Advantage of the cache
eviction end point is that user-sessions remain undisturbed since the portal instance is not restarted. By default the
cache eviction enpoint is disabled and can be enabled by setting `cache.endpoint.enabled` to _true_. The endpoint is
secured by a secret API key that can be customized with the `cache.endpoint.api-key` property. Caches are evicted by
making a DELETE request to the endoint while passing the API key in the `X-API-KEY` header. When using _curl_ use the
following command (replace the API key for the value configured in _application.properties_):

```
curl -X DELETE http://my-portal-url.org/api/cache  -H 'X-API-KEY: fd15f1ae-66f2-4b8a-8d54-fb899b03557e'
```

### Cache eviction after cancer study updates

When a study is added, deleted or updated, a more selective cache eviction strategy is possible, where only affected
cached data is evicted. This more selective cache eviction is triggered by calling the `/api/cache/{studyId}` endpoint
where _{studyId}_ is the _cancer_study_identifier_ stated in the meta-study.txt file. When using _curl_ use the
following command after update of a study with study identifier _my_cancer_study_ (replace the API key for the value
configured in _application.properties_):

```
curl -X DELETE http://my-portal-url.org/api/cache/my_cancer_study  -H 'X-API-KEY: fd15f1ae-66f2-4b8a-8d54-fb899b03557e'
```

:important: This endpoint can ony be used when adding/deleting/updating a study. When data related to gene panels or
gene sets is updated, all caches should be evicted with a call to `/api/cache`.

#### How does study-specific cache eviction work?

##### Structure of cache keys

The caching solutions integrated with cBioPortal (EHCache and Redis) store data as key-value pairs. Each key represents
a method call signature that contains the Java class name, the method name and a serialized representation of all method
arguments. For instance this is the key for a call to the CancerTypMyBatisRepository.getAllCancerTypes() method with
arguments _[ SUMMARY, 10000000, 0, null, ASC]_:

```
CancerTypeMyBatisRepository_getAllCancerTypes_"SUMMARY"_10000000_0_null_"ASC"
```

Cached data that relates to a specific study can be recognized by the occurrence of the cancer study identifier anywhere
in the method arguments. The study identifier can occur in the method arguments because it is passed as argument itself,
like here for a study with identifier _study_es_0_:

```
SignificantCopyNumberRegionMyBatisRepository_getSignificantCopyNumberRegions_"study_es_0"_"SUMMARY"_null_null_null_null
```

Alternatively the study identifier is present as the prefix of referenced study entities. For example, this is the
request for all molecular profiles:

```
MolecularProfileMyBatisRepository_getMolecularProfiles_["study_es_0_mrna_median_Zscores","study_es_0_log2CNA","study_es_0_fusion","study_es_0_treatment_ec50","study_es_0_mutational_signature","study_es_0_gsva_pvalues","study_es_0_mutations","study_es_0_methylation_hm27","study_es_0_gistic","study_es_0_generic_assay_patient_test","study_es_0_treatment_ic50","study_es_0_mrna","study_es_0_gsva_scores"]_"SUMMARY""
```

#### Cache eviction rules

When a study is added, deleted or updated, all caches are evicted where the respective key meets any of these
requirements:

1. The key contains the cancer study identifier of the study that is added, deleted or updated.
2. The key does not contain the cancer study identifier of any study present in the database.

The rationale behind rule 1. is that when a key references data for the affected study it points to potentially outdated
data and its associated cache should be evicted. The rationale behind rule 2. is that any key that does not reference
data for any study potentially points to data derived from all studies in the database, including the affected study,
and its associated cache should be evicted. Because not every key without study identifiers necessarily points to study
related data, this rule is overly broad. At the moment of this writing we were unable to implement reliable methods that
would further specify such keys. This might be a start-off point for future optimizations.


# Customizing your instance of cBioPortal

cBioPortal contains a number of properties that allow you to customize your cBioPortal instance. 

This page focuses on the skin properties, which allow you to
customize the web page cosmetics, such as custom images,
texts, which tabs are visible, etc.
Nearly all the skins properties have defaults, which can be overwritten by
changing the `application.properties` file located in `$PORTAL_HOME`--see
[the deployment manual](/deployment/deploy-without-docker/Deploying.md). If your cBioPortal instance
does not yet contain a `application.properties` file, you can copy
`src/main/resources/application.properties.EXAMPLE` and edit it.

Below you can find the complete list of all the available skin properties.

# cBioSkins Properties

**Please note!**  If you are reading this on the GitHub web site, you may need to scroll right to see all the columns in the table below.

<table>
   <tbody>
		<tr>
         <th>Name</th>
         <th>Description</th>
         <th>Default</th>
         <th>Type Allowed</th>
		</tr>
		<tr>
			<td>skin.authorization_message</td>
			<td>sets the authorization message that appears after an unsuccessful login</td>
			<td>Access to this portal is only available to authorized users at Memorial Sloan Kettering Cancer Center...</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.blurb</td>
			<td>sets the main text on the index page</td>
			<td>Text for public cBioPortal.org.</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.example_study_queries</td>
			<td>\n-separated list of study query suggestions displayed when clicking the ‚Äòselect a cancer study' search box</td>
			<td>tcga\ntcga -provisional\ntcga -moratorium\ntcga OR icgc\n-"cell line"\nprostate mskcc\nesophageal OR stomach\nserous\nbreast</td>
			<td>text separated by the escape sequence \n</td>
		</tr>
		<tr>
			<td>skin.data_sets_footer</td>
			<td>sets the text that is shown below the number of samples, after clicking on the "DATA SETS" tab in the header.</td>
			<td>The portal currently contains data from the following cancer genomics studies.  The table below lists the number of available samples per data type and tumor.</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.data_sets_header</td>
			<td>sets the text that is shown below the Data Sets title, after clicking on the "DATA SETS" tab in the header. By adding &lt;NUM_CANCER_STUDIES&gt; to the text, the number of cancer studies is inserted</td>
			<td>Data sets of TCGA studies were downloaded from Broad Firehose (https://gdac.broadinstitute.org) and updated monthly. In some studies, data sets were from the TCGA working groups directly.</td>
			<td>Any HTML text</td>
		</tr>
      <tr>
         <td>skin.documentation.about</td>
         <td>set the about page which can be accessed via ABOUT in the header. Please see the cBioportal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>About-Us.md</td>
         <td>text</td>
      </tr>
      <tr>
         <td>skin.documentation.baseurl</td>
         <td>set the base url for the documentation. Please see the cBioPortal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>https://github.com/cBioPortal/cbioportal/wiki/</td>
         <td>text</td>
      </tr>
      <tr>
         <td>skin.documentation.faq</td>
         <td>set the faq page which can be accessed via FAQ in the header. Please see the cBioPortal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>user-guide/faq.md</td>
         <td>text</td>
      </tr>
      <tr>
         <td>skin.documentation.markdown</td>
         <td>set whether the documentation files are in markdown. Please see the cBioPortal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>true</td>
         <td>text</td>
      </tr>
      <tr>
         <td>skin.documentation.news</td>
         <td>set the news page which can be accessed via NEWS in the header. Please see the cBioPortal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>News.md</td>
         <td>text</td>
      </tr>
		<tr>
			<td>skin.email_contact</td>
			<td>sets the contact email address for Questions and feedback, located in the footer.</td>
			<td>cbioportal at googlegroups dot com</td>
			<td>Specify as name at somewhere dot extension or as name@somewhere.extension</td>
		</tr>
      <tr>
         <td>skin.examples_right_column_html</td>
         <td>set the Examples Queries section in the right navigation bar</td>
         <td>&lt;ul&gt;&lt;li&gt;&lt;a href="patient?studyId=ucec_tcga_pub&caseId=TCGA-BK-A0CC"&gt;Patient view of an endometrial cancer case&lt;/a&gt;&lt;/li&gt;&lt;/ul&gt;</td>
         <td>Any HTML text</td>
      </tr>
		<tr>
			<td>skin.footer</td>
			<td>sets the footer text. The text will be appended to the already existing cBioPortal link</td>
			<td>cBioPortal public instance footer.</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.login.contact_html</td>
			<td>sets the feedback message which appears after a failed login attempt</td>
			<td>If you think you have received this message in error, please contact us at...</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.login.saml.registration_html</td>
			<td>sets the button text for saml login</td>
			<td>Sign in with MSK</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.left_logo</td>
			<td>sets the left logo in the header. Logo should be placed in the images directory or in a subdirectory of the images directory. If placed in a subfolder, the skin.left_logo value has to contain the folder, e.g. skin.right_logo = myFolder/myImage.jpg</td>
			<td>cBioPortal Logo</td>
			<td>text</td>
		</tr>
		<tr>
			<td>skin.right_logo</td>
			<td>sets the right institute logo in the header. Logo should be placed in the images directory or in a subdirectory of the images directory. If placed in a subfolder, the skin.right_logo value has to contain the folder, e.g. skin.right_logo = myFolder/myImage.jpg</td>
			<td>no logo</td>
			<td>text</td>
		</tr>
		<tr>
			<td>skin.right_nav.show_data_sets</td>
			<td>show the "Data Sets" section in the right navigation bar</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.right_nav.show_examples</td>
			<td>show the "Example Queries" section in the right navigation bar</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.right_nav.show_testimonials</td>
			<td>show the "What People are Saying" section in the right navigation bar</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.right_nav.whats_new_blurb</td>
			<td>set the content of the "What‚Äôs New" section in the right navigation bar</td>
			<td>Text for public cBioPortal.org.</td>
			<td>Any HTML text</td>
		</tr>
        <tr>
			<td>skin.right_nav.show_web_tours</td>
			<td>set the "Interactive Tours" section in the right navigation bar</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_about_tab</td>
			<td>show the "ABOUT" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_data_tab</td>
			<td>show the "DATA SETS" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_faqs_tab</td>
			<td>show the "FAQ" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_news_tab</td>
			<td>show the "NEWS" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_r_matlab_tab</td>
			<td>show the "R/MATLAB" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_tools_tab</td>
			<td>show the "TOOLS" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_tutorials_tab</td>
			<td>show the "TUTORIALS" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_web_api_tab</td>
			<td>show the "WEB API" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.tag_line_image</td>
			<td>sets the tag line image in the header. The image should be placed in the images directory or in a subdirectory of the images directory. If placed in a subdirectory, the skin.tag_line_image value has to contain the folder, e.g. skin.tag_line_image = myFolder/myImage.jpg</td>
			<td>tag_line.png</td>
			<td>text</td>
		</tr>
		<tr>
			<td>skin.title</td>
			<td>sets the title of the site / name of the browser tab. E.g. when navigating to About Us with the default setting, the title becomes "cBioPortal for Cancer Genomics::About Us"</td>
			<td>cBioPortal for Cancer Genomics</td>
			<td>text</td>
		</tr>
		<tr>
			<td>skin.show_gsva</td>
			<td>enables gsva to query</td>
			<td>false</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.geneset_hierarchy.default_p_value</td>
			<td>modifies default value for p_value threshold</td>
			<td>0.05</td>
			<td>0.05</td>
		</tr>
		<tr>
			<td>skin.geneset_hierarchy.default_gsva_score</td>
			<td>modifies default value for gsva_score threshold</td>
			<td>0.5</td>
			<td>0.5</td>
		</tr>
		<tr>
			<td>skin.geneset_hierarchy.collapse_by_default</td>
			<td>collapses the tree widget of the gsva hierarchy dialog on init</td>
			<td>false</td>
			<td>true / false</td>
		</tr>
		<tr>
            <td>skin.query.max_tree_depth</td>
            <td>sets the maximum number of subcategories shown in the query component hierarchy before each study. E.g. when set to 0, the hierarchy is flat, meaning only the study elements show in the component, without any tissue or cancer type sorting.</td>
            <td>3</td>
            <td>integer</td>
        </tr>
		<tr>
            <td>skin.patientview.filter_genes_profiled_all_samples</td>
            <td>sets default setting for the genes filter in patient view to only show mutations for genes that were profiled for mutations or CNA's in all samples of that patient. If unset, patient view will show mutations/CNA's for genes that were profiled in any sample.</td>
            <td>false</td>
            <td>true / false</td>
        </tr>
		<tr>
            <td>skin.hide_download_controls</td>
            <td>controls download options in UI. **true**: hides all download options. **false**:  shows all download options. **data**: disable data download but not figure download</td>
            <td>false</td>
            <td>true / false / data</td>
        </tr>
		<tr>
            <td>download_group</td>
            <td>controls download options in UI for each user. If present and user is authenticated, this value is checked against user roles.
If the download_group is present in user groups then download options are shown in UI, else it fallback to **skin.hide_download_controls**</td>
            <td></td>
            <td>text</td>
        </tr>
        <tr>
            <td>skin.show_settings_menu</td>
            <td>controls the appearance of the settings menu in study view and group comparison that controls annotation-based filtering.</td>
            <td>false</td>
            <td>true / false</td>
        </tr>
        <tr>
            <td>skin.hide_logout_button</td>
            <td>controls the appearance the logout button on a portal with user authentication.</td>
            <td>false</td>
            <td>true / false</td>
        </tr>
        <tr>
            <td>google_analytics_profile_id</td>
            <td>enables google analaytics tracking on your site</td>
            <td>disabled</td>
            <td>string</td>
        </tr>
   </tbody>
</table>

# Customization when using Docker
To add images or other files to the Docker image one can mount them inside the `cbioportal-webapp` folder. For example to use a custom logo for `skin.right_logo`. Mount the image inside the`cbioportal-webapp` folder with `-v /path/to/custom_logo.png:/cbioportal-webapp/images/custom_logo.png` and set `skin.right_logo=custom_logo.png` in your properties file.

# Creating your own local news/about or FAQ page

Creating you own local news/about or FAQ page involves three steps. For example, to create your own news page:

1. Copy `<cbioportal_source_folder>/portal/src/main/webapp/content/news.html` to `news_XXXX.html`
2. Modify `news_XXXX.html` as needed.
3. Edit the `application.properties` file and change the `skin.documentation.news` property, giving it the name of your news HTML component.

You may need to modify additional settings depending on whether you're e.g. using markdown pages. Please refer to the [cBioPortal Documentation](#cbioportal-documentation-settings) for more information.

# cBioPortal Documentation Settings
cBioPortal supports internal as well as external pages and supports html and markdown pages. 

## Using External Pages
cBioPortal can retrieve external pages and use them as news/faq/about page. For example, if your news page is stored on github wiki and reachable via https://github.com/cBioPortal/cbioportal/wiki/News.md use

```
skin.documentation.baseurl=https://github.com/cBioPortal/cbioportal/wiki/
skin.documentation.news=News.md
```

## Using Internal Pages
cBioPortal can use pages stored in (a subdirectory of) your content directory as news/faq/about page. For example, if your news page is called myNews.md and stored in content/myFolder, use 

```
skin.documentation.baseurl=/
skin.documentation.news=myFolder/myNews.md
```

## Markdown
By default, cBioPortal assumes all documentation is in markdown. If your documentation is not in markdown, please use

```
skin.documentation.markdown=false
```

If your documentation contains a relative link, cBioPortal will assume it uses the same format and, when clicked, will try to interpret the page as such. If your documentation contains an absolute link, cBioPortal will open it in a new window when clicked.
Please be aware that the links may be case-sensitive! E.g. https://github.com/cBioPortal/cbioportal/wiki/News.md is not the same as https://github.com/cBioPortal/cbioportal/wiki/news.md

This file has been depreciated and information about customizing cBioPortal can be found in [application.properties documentation](/deployment/customization/application.properties-Reference.md).# Application Properties

The following are the properties for configuring authentication and authorization in the application.

## Authentication Configuration

### General Authentication Settings

```properties
# authentication (available options: [false, oauth2, optional_oauth2, saml])
authenticate=false
```
### OAUTH2
#### NOTE for Custom Authorization (validate users via db) 
```properties
authenticate=oauth2
authorization=true
```
#### Google OAuth2 Client/Login Configuration

#### Example of utilizing google client for oAuth2 (Authentication)
```properties
spring.security.oauth2.client.registration.google.clientId=
spring.security.oauth2.client.registration.google.clientSecret=
spring.security.oauth2.client.provider.google.user-name-attribute=email
```

### Microsoft OAUTH2 Multi-tenant Client/Login Config
#### Example with Utilizing AzureAD for oAuth2
```properties
spring.security.oauth2.client.registration.azure.authorization-grant-type=authorization_code
spring.security.oauth2.client.registration.azure.client-id=<client-id>
spring.security.oauth2.client.registration.azure.clientSecret=<secret>
spring.security.oauth2.client.provider.azure.user-name-attribute=email
spring.security.oauth2.client.registration.azure.redirect-uri=http://localhost:8080/login/oauth2/code/azure
spring.security.oauth2.client.provider.azure.authorization-uri=https://login.microsoftonline.com/common/oauth2/v2.0/authorize
spring.security.oauth2.client.provider.azure.token-uri=https://login.microsoftonline.com/common/oauth2/v2.0/token
spring.security.oauth2.client.provider.azure.jwk-set-uri=https://login.microsoftonline.com/common/discovery/v2.0/keys
spring.security.oauth2.client.registration.azure.scope=openid,profile,email
```
#### Custom OAUTH2 Client Configuration

```properties
# For OIDC clients the issuer-uri is sufficient to autoconfigure the provider (via .well-known endpoint)
spring.security.oauth2.client.provider.cbio-idp.issuer-uri=http://localhost:8080/realms/cbioportal
spring.security.oauth2.client.provider.cbio-idp.user-name-attribute=email
# Required Scopes [openid, email, roles] 
spring.security.oauth2.client.registration.cbio-idp.scope=openid,email,roles
spring.security.oauth2.client.registration.cbio-idp.client-id=
spring.security.oauth2.client.registration.cbio-idp.client-secret=
```

##### Configuring the individual settings below is not recommended.
```properties
# spring.security.oauth2.client.provider.cbio-idp.authorization-uri=
# spring.security.oauth2.client.provider.cbio-idp.token-uri=
# TODO update docs, the user info endpoint must expose the roles !!
# spring.security.oauth2.client.provider.cbio-idp.user-info-uri=
# spring.security.oauth2.client.provider.cbio-idp.jwk-set-uri=
# spring.security.oauth2.client.provider.cbio-idp.logout-uri= # NOTE: this is not an official property.
# TODO Can be authorization_code, ...
#spring.security.oauth2.client.registration.cbio-idp.authorization-grant-type=
# TODO Can be client_secret_post, ...
#spring.security.oauth2.client.registration.cbio-idp.client-authentication-method=
#spring.security.oauth2.client.registration.cbio-idp.redirect-uri=<server-url>/login/oauth2/authorization/<client-registration-id (cbio-idp in this case)>
```

### SAML Configuration
#### Example to generate cert and key 
```shell
openssl req -newkey rsa:2048 -nodes -keyout local.key -x509 -days 365 -out local.crt
```

```properties

# For SAML 2.0
## SAML settings
# TODO add options for auto- and manual config to docs
# TODO add to docs: metadata-uri can be both URL or metadata xml file
spring.security.saml2.relyingparty.registration.cbio-saml-idp.assertingparty.metadata-uri=classpath:/client-tailored-saml-idp-metadata.xml
#spring.security.saml2.relyingparty.registration.cbio-saml-idp.assertingparty.metadata-uri=http://localhost:8080/realms/cbioportal/protocol/saml/descriptor
spring.security.saml2.relyingparty.registration.cbio-saml-idp.entity-id=cbioportal-saml
spring.security.saml2.relyingparty.registration.cbio-saml-idp.signing.credentials[0].certificate-location=classpath:/local.crt
spring.security.saml2.relyingparty.registration.cbio-saml-idp.signing.credentials[0].private-key-location=classpath:/local.key
spring.security.saml2.relyingparty.registration.cbio-saml-idp.singlelogout.binding=POST
# TODO add to docs (in minutes; default 1)
spring.security.oauth2.allowed-clock-skew=
```
### Data Access Token Settings

```properties

## data access token settings
## Resource Server issuer-uri for Data Access Token
#spring.security.oauth2.resourceserver.jwt.issuer-uri=http://localhost:8080/realms/cbioportal
# TODO: Currently Dat.method has only been tested with oauth2
dat.unauth_users=
dat.method=oauth2
dat.ttl_seconds=2592000
dat.uuid.max_number_per_user=1
dat.jwt.secret_key=
dat.filter_user_role=

# OAuth2 token data access settings (If using OAuth2 for Login can copy setting here)
## TODO: Reuse OAUTH2 Spring settings defined above
dat.oauth2.clientId=<client-id>
dat.oauth2.clientSecret=<client-secret>
dat.oauth2.issuer=<token-issuer>
dat.oauth2.accessTokenUri=<authorization-server-url>/.../token
dat.oauth2.userAuthorizationUri=<authorization-server-url>/.../auth
dat.oauth2.jwkUrl=<authorization-server-url>/.../certs
dat.oauth2.redirectUri=<cbioportal-url>/.../api/data-access-token/oauth2
spring.security.oauth2.resourceserver.jwt.issuer-uri=http://localhost:8081/realms/cbioportal

```
### Authorization Configuration/Study View Settings

```properties

## Authorization
## study view settings
## always show studies with this group
always_show_study_group=PUBLIC
## Should the permissions for groups and users be filtered by this instance's app.name?
## (true means the system only handles "CBIOPORTAL:someGroupPermission" groups, false means "someGroupPermission" works)
filter_groups_by_appname=false
# Can disable authorization
security.method_authorization_enabled=true
```

### CORS Configuration
To Enable CORS set the allowed-origins urls. (comma delimited list)
To enable all origins use *
```properties
security.cors.allowed-origins=*
##Or http://localhost:8080,http://localhost:8081
```# Study View Customization

## How does the study view organize the charts

Study view page is fully responsive, it will try to fit as many charts as possible based on the browser's width and height.

The layout of a chart is determined mainly based on priority. Higher priority will promote chart closer to the left-top.

In order to improve the layout, we added a layout algorithm layer. The study view page is using grid layout. All charts will be put into 2-dimensional systems. For example, pie chart, by default, takes 1 block and bar chart uses two blocks. All charts will be placed from left to right, top to bottom. In order to prevent misalignment, we promote small charts to fit into the space.

For logged-in(authenticated) users, charts layout is saved to users profile i.e, whenever user tries to re-visits the same url, previously saved layout will be loaded.
Please note for optimal performance, the portal will initially display 20 clinical charts by default. These charts are prioritized from high to low. Additional clinical charts are initially hidden but can be easily selected by users through the 'Add charts' dropdown menu.

## Study View Customization with Priority Data
Example of study view in public portal: https://www.cbioportal.org/study?id=acc_tcga,lgg_tcga#summary

## Priorities

The priority system is represented with a final score.
The higher the final (numeric) score, the higher priority assigned.

The default score is 1.

To promote certain chart in study view, please increase priority in the database to a certain number.
The higher the score, the higher priority it will be displayed in the study view.
If you want to hide chart, please set the priority to 0.
For combination chart, as long as one of the clinical attributes has been set to 0, it will be hidden.

To disable the chart, set the priority to -1.(Currently disables charts for single clinical attributes only)

Currently, we preassigned priority to few charts, but as long as you assign a priority in the database except than 1, these preassigned priorities will be overwritten.

| Chart name(clinical attribute ID)                          	| Frontend default priority 	| Additional Info                                   |
|------------------------------------------------------------	|---------------------------	|-------------------------------------------------	|
| CANCER_TYPE                                                	| 3000                      	|                                                 	|
| CANCER_TYPE_DETAILED                                       	| 2000                      	|                                                 	|
| Overall Survival Plot                                      	| 400                       	| This is combination of OS_MONTH and OS_STATUS   	|
| Disease Free Survival Plot                                 	| 300                       	| This is combination of DFS_MONTH and DFS_STATUS 	|
| Mutation Count vs. Fraction of Genome Altered Density Plot 	| 200                       	|                                                 	|
| Mutated Genes Table                                        	| 90                        	|                                                 	|
| CNA Genes Table                                            	| 80                        	|                                                 	|
| Cancer Studies                                             	| 70                        	|                                                 	|
| Number of Samples Per Patient                                 | 40                        	|                                                 	|
| Mutation Count Bar Chart                                   	| 30                        	|                                                 	|
| CNA Bar Chart                                              	| 20                        	|                                                 	|
| GENDER, SEX                                                	| 9                         	|                                                 	|
| AGE                                                        	| 8                         	|                                                 	|
# Building from Source

## Building with Maven

To compile the cBioPortal source code, move into the root directory and run the following maven command:

```
mvn -DskipTests clean install
```

Note: cBioPortal 6.X requires Java 21
# Deploying the Web Application

## Before running cbioportal backend

You will need to update the src/main/resources/applications.properties to include your DB connection information.

The configuration defined in `application.properties` can also be passed as command line arguments. The priority of property loading is as follows:

1. `-D` command line parameters overrides all
2. `src/main/resources/application.properties`
3. `application.properties` supplied at compile time
4. Defaults defined in code

Note that the `authenticate` property is currently required to be set as a command line argument, it won't work when set in `application.properties` (See issue [#6109](https://github.com/cBioPortal/cbioportal/issues/6109)).

Some scripts require a `${PORTAL_HOME}/application.properties` file, so it is best to define the properties there.

For more information about the `application.properties` file, see the [reference](/deployment/customization/Customizing-your-instance-of-cBioPortal.md) page.


## Run the cbioportal backend

To run the backend execute the following commabd

```
java -jar target/cbioportal-exec.jar
```

This runs the app in the foreground. If a port is already in use it will raise an error mentioning that. To change the port use the `--server.port` flag.



There are three main ways to run the portal: without authentication, with optional login, and with required login. All of them require the cBioPortal session service to be running.

### Without authentication

In this mode users are able to use the portal, but they won't be able to save their own virtual studies and groups.

```
java -jar target/cbioportal-exec.jar -Dauthenticate=false
```

### With authentication

To configure the authentication and authorization please consult the [Authorization](./../authorization-and-authentication/User-Authorization.md) and [Authentication](./../authorization-and-authentication/Authenticating-and-Authorizing-Users-via-keycloak.md) Sections.
# Importing the Seed Database

The next step is to populate your cBioPortal instance with all the required background data sets. This includes for example gene data, ID mappings, and network interactions. Rather than importing each of these data sets individually, we have provided a simple "seed" database that you can import directly.

## Download the cBioPortal Seed Database

A cBioPortal seed database for human can be found on [the datahub page](https://github.com/cBioPortal/datahub/blob/master/seedDB/README.md). If you are looking for mouse, check [this link](https://github.com/cBioPortal/datahub/blob/master/seedDB_mouse/README.md).

After download, the files can be unzipped by entering the following command:

    gunzip *.sql.gz

## Import the cBioPortal Seed Database

**Important:** Before importing, make sure that you have [followed the pre-build steps](Pre-Build-Steps.md) for creating the `cbioportal` database (see section "Create the cBioPortal MySQL Databases and User").

1. Import the database schema (/db-scripts/src/main/resources/cgds.sql):

    ```
    mysql --user=cbio_user --password=somepassword cbioportal < cgds.sql
    ```

    Note that this may currently fail when using the default character encoding on MySQL 8.0 (`utf8mb4`); this is why MySQL 5.7 (which uses `latin1`) is recommended.

2. Import the main part of the seed database:

    ```
    mysql --user=cbio_user --password=somepassword cbioportal < seed-cbioportal_RefGenome_vX.Y.Z.sql
    ```

    **Important:** Replace `seed-cbioportal_RefGenome_vX.Y.Z.sql` with the downloaded version of the seed database, such as `seed-cbioportal_hg19_v2.3.1.sql` or `seed-cbioportal_mm10_v2.3.1.sql`.

3. (Human only) Import the Protein Data Bank (PDB) part of the seed database. This will enable the visualization of PDB structures in the mutation tab. Loading this file takes more time than loading the previous files, and is optional for users that do not require PDB structures.

    ```
    mysql --user=cbio_user --password=somepassword cbioportal < seed-cbioportal_hg19_vX.Y.Z_only-pdb.sql
    ```
    **Important:** Replace `seed-cbioportal_hg19_vX.Y.Z_only-pdb.sql` with the downloaded version of the PDB database, such as `seed-cbioportal_hg19_v2.3.1_only-pdb.sql`.

4. (optional : support for microRNA genomic profiles) Import constructed gene table records for microRNA genomic profiles. Currently, cBioPortal supports the combined display of copy number alterations (generally reported for microRNA precursors) and expression (generally reported for microRNA mature forms) by adding gene table records which represent the combination of microRNA precursor and microRNA mature form. Appropriate aliases are added to the gene_alias table so that both the name of the precursor and the name of the mature form are recognized references to the combination.

This involves downloading the cBioPortal Core code located [here](https://github.com/cBioPortal/cbioportal-core). 

After the code has been successfully configured and built, you can import the needed microRNA records by running the following command from the cBioPortal core directory:

    java -cp scripts/target/scripts-*.jar org.mskcc.cbio.portal.scripts.ImportGeneData -microrna core/src/main/resources/micrornas.tsv
    

**Important:** Please be aware of the version of the seed database. In the [README on datahub](https://github.com/cbioportal/datahub/blob/master/seedDB/README.md), we stated which version of cBioPortal is compatible with the current seed database.

If the database is older than what cBioPortal is expecting, the system will ask you (during startup or data loading) to migrate the database to a newer version. The migration process is described [here](/Updating-your-cBioPortal-installation.md#running-the-migration-script).
# Loading a Sample Study

Once you have confirmed that the cBioPortal server is installed,
you are ready to import data. Importing a sample study is recommended
to verify that everything is working correctly.

cBioPortal Core has a [small dummy study, `study_es_0`](https://https://github.com/cBioPortal/cbioportal-core/tree/main/src/test/scripts/test_data/study_es_0), which contains all datatypes supported by cBioPortal. This document describes how to import the prerequisites for the sample study and how to import the study itself.


## Download and Build cBioPortal Core

```
	git clone https://github.com/cBioPortal/cbioportal-core.git
	cd cbioportal-core
	git checkout main
	mvn -DskipTests clean install
```



## Set the PORTAL_HOME environment variable

Most cBioPortal command-line tools, including the data loading pipeline,
expect the environment variable `$PORTAL_HOME` to point to a folder
containing the `application.properties` configuration file,
as explained during [the previous step](./Deploying.md).

Configure your shell to keep the variable set to the right folder.
On GNU/Linux and macOS this usually means appending a line
like the following to `.bash_profile` in your home directory:

```
export PORTAL_HOME=/Users/johndoe/cbioportal
```

## Import Gene Panel for Sample Study

The sample gene panel has to be imported before gene panel study data can be added to the database.

```
cd cbioportal-core/src/main/resources/scripts 
./importGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel1.txt
./importGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel2.txt
```

After loading gene panels into the database, please restart Tomcat or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information)
so that the validator can retrieve gene panel information from the cBioPortal API.

More details to load your own gene panel and gene set data can be found here: [Import Gene Panels](/Import-Gene-Panels.md).

## Validating the Sample Study

First it's useful to validate the study `study_es_0`, to check if the data is formatted correctly.

To do so, go to the importer folder: 

```
cd cbioprtal-core/src/main/scripts/importer
```

and then run the following command:

```
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -n
```

If all goes well, you should see the final output message:

```
Validation of study succeeded with warnings.
```

## Importing the Sample Study

To import the sample study:

```
cd cbioportal-core/src/main/scripts/importer
```

and then run the following command:

```
./metaImport.py -s ../../../test/scripts/test_data/study_es_0/ -n -o
```

You will see a series of output messages, hopefully ending with a status message like this:

```
Done.
Total time:  7742 ms
```

After loading the study data, please restart the app  or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information).
# Pre-Build Steps

## Get the Latest Code

Make sure that you have cloned the last code, and make sure you are on the `master` branch:
```
	git clone https://github.com/cBioPortal/cbioportal.git
	cd cbioportal
	git checkout master
```

## Prepare the log4j.properties File

This file configures logging for the portal.
An example file is available within GitHub:
```
    cd src/main/resources
    cp logback.xml.EXAMPLE logback.xml
```

If you don't create your own `logback.xml`, maven will copy the EXAMPLE file to that location when it builds.
If `logback.xml` already exists, it will just use that. This allows us to give you a working, versioned
log config, which you can then override easily.

To modify the logging during tests the same EXAMPLE file can be copied to the relevant test resources folder.

## Prepare the global configuration file

The portal is configured using a global configuration file, `application.properties`. An example file is available in the `src/main/resources` folder. Use it as a template to create your own:

```
cd src/main/resources
cp application.properties.EXAMPLE application.properties
```

For more information about the `application.properties` file, see the [reference](/deployment/customization/Customizing-your-instance-of-cBioPortal.md) page.

## Create the cBioPortal MySQL Databases and User

You must create a `cbioportal` database and a `cgds_test` database within MySQL, and a user account with rights to access both databases.  This is done via the `mysql` shell.
```
    > mysql -u root -p
    Enter password: ********

    Welcome to the MySQL monitor.  Commands end with ; or \g.
    Your MySQL connection id is 64
    Server version: 5.6.23 MySQL Community Server (GPL)

    Copyright (c) 2000, 2015, Oracle and/or its affiliates. All rights reserved.

    mysql> create database cbioportal;
    Query OK, 1 row affected (0.00 sec)

    mysql> create database cgds_test;
    Query OK, 1 row affected (0.00 sec)

    mysql> CREATE USER 'cbio'@'localhost' IDENTIFIED BY 'P@ssword1';
    Query OK, 0 rows affected (0.00 sec)

    mysql> GRANT ALL ON cbioportal.* TO 'cbio'@'localhost';
    Query OK, 0 rows affected (0.00 sec)

    mysql> GRANT ALL ON cgds_test.* TO 'cbio'@'localhost';
    Query OK, 0 rows affected (0.00 sec)

    mysql>  flush privileges;
    Query OK, 0 rows affected (0.00 sec)
```
# Software Requirements

This page describes various system software required to run the cBioPortal.

## MySQL

The cBioPortal software should run properly on MySQL version 5.7.x. Versions higher than 5.7.x can cause an issue
while loading the database schema. Minor versions lower than 5.7.x will cause issues with persistent cache invalidation.
The software can be found and downloaded from the [MySQL website](https://www.mysql.com/).

On Ubuntu:  ```sudo apt-get install mysql-server```

## MongoDB

The session service uses MongoDB 3.6.6

## Java

cBioPortal requires Java 12 and above.  The software can be found and download from the [Oracle](https://www.oracle.com/us/technologies/java/overview/index.html) website.

On Ubuntu:  ```sudo apt-get install default-jdk```

## Apache Maven

The cBioPortal source code is an [Apache Maven](https://maven.apache.org/) driven project.  The software needs to be downloaded and installed prior to building the application from source code.  It can be found on the [Apache Maven](https://maven.apache.org/download.cgi) website. We are currently using version 3.9.5.

On Ubuntu:  ```sudo apt-get install maven```

## Git

You will need a git client to download the cBioPortal source code.

On Ubuntu:  ```sudo apt-get install git```
### Importing gene panel ###

Use this command to import a gene panel. Specify the gene panel file by replacing 
`<path_to_genepanel_file>` with the absolute path to the gene panel file. Another option is to add the 
gene panel files in `./study` which is mounted inside the container on `/study/.

```shell
docker-compose run \
    -v <path_to_genepanel_file>:/gene_panels/gene_panel.txt:ro \
    cbioportal \
    bash -c 'cd /cbioportal/core/src/main/scripts/ && ./importGenePanel.pl --data /gene_panels/gene_panel.txt'
```

### Importing data ###

Use this command to validate a dataset. Add the study to the `./study` folder. The
command will connect to the web API of the container `cbioportal-container`, and import 
the study in its associated database. Make sure to replace `<path_to_report_folder>` with 
the absolute path were the html report of the validation will be saved.

```shell
docker-compose run \
    -v "<path_to_report_folder>:/report" \
    cbioportal \
    metaImport.py -u http://cbioportal:8080 -s /study/name_of_study --html=/report/report.html
```
:warning: after importing a study, remember to restart `cbioportal-container`
to see the study on the home page. Run `docker-compose restart cbioportal`.

#### Using cached portal side-data ####

In some setups the data validation step may not have direct access to the web API, for instance when the web API is only accessible to authenticated browser sessions. You can use this command to generate a cached folder of files that the validation script can use instead. Make sure to replace `<path_to_portalinfo>` with the absolute path where the cached folder is going to be generated.

```shell
docker-compose run \
    -v "<path_to_portalinfo>/portalinfo:/portalinfo" \
    -w /cbioportal/core/src/main/scripts \
    cbioportal \
    ./dumpPortalInfo.pl /portalinfo
```

Then, grant the validation/loading command access to this folder and tell the script it to use it instead of the API:

```shell
docker-compose run \
    -v "<path_to_report_folder>:/report" \
    -v "<path_to_portalinfo>/portalinfo:/portalinfo:ro" \
    cbioportal \
    metaImport.py -p /portalinfo -s /study --html=/report/report.html
```

### Inspecting or adjusting the database ###

```shell
docker-compose run cbioportal_database \
    sh -c 'mysql -hcbioportal_database -u"$MYSQL_USER" -p"$MYSQL_PASSWORD" "$MYSQL_DATABASE"'
```

### Deleting a study ###

To remove a study, run:

```shell
docker-compose run \
    cbioportal \
    cbioportalImporter.py -c remove-study -id study_id
```

Where `study_id` is the `cancer_study_identifier` of the study you would like to remove.
# Import data with Docker

### Import data instructions

This is an example to import a sample study: `study_es_0`. When trying to import other studies, please follow the same routine:

1. import gene panels (if applicable, studies without gene panels are assumed to be whole exome/genome)
2. import study data

#### Step 1 - Import gene panels

To import gene panels for your study, please reference the example commands in [this file](example\_commands.md#importing-gene-panel)

These are the commands for importing `study_es_0` gene panels (`data_gene_panel_testpanel1` and `data_gene_panel_testpanel2`):

```shell
docker-compose run \
    cbioportal \
    bash -c 'cd /cbioportal/core/src/main/scripts/ && ./importGenePanel.pl --data /cbioportal/core/src/test/scripts/test_data/study_es_0/data_gene_panel_testpanel1.txt'
```

```shell
docker-compose run \
    cbioportal \
    bash -c 'cd /cbioportal/core/src/main/scripts/ && ./importGenePanel.pl --data /cbioportal/core/src/test/scripts/test_data/study_es_0/data_gene_panel_testpanel2.txt'
```

#### Step 2 - Import data

To import data for your study, please reference the example commands in [this file](example\_commands.md#importing-data)

Command for importing `study_es_0` data:

```shell
docker-compose run cbioportal metaImport.py -u http://cbioportal:8080 -s /cbioportal/core/src/test/scripts/test_data/study_es_0 -o
```

:warning: after importing a study, remember to restart `cbioportal` to see the study on the home page. Run `docker-compose restart cbioportal`.

You have now imported the test study `study_es_0`. Note that this study is included inside the cbioportal container. The process for adding a study that is outside of the container is similar. Just make sure to add the data files in the `./study` folder. This folder is mounted as `/study/` inside of the container.

### Frequently Asked Questions

#### Gene panel ID is not in database

If you see an error like this when you importing the data:\
`ERROR: data_gene_panel_matrix.txt: lines [2, 3, 4, (10 more)]: Gene panel ID is not in database. Please import this gene panel before loading study data.; values encountered: ['TESTPANEL1', 'TESTPANEL2']`

please follow the first step to import gene panels (e.g. import `data_gene_panel_testpanel1` and `data_gene_panel_testpanel2` for `study_es_0`), then try to import the data again.

#### Error occurred during validation step

Please make sure the seed database was correctly imported.

#### Study imported correctly, but got error when trying to query something

Remember to restart the `cbioportal` after data imported.

```shell
docker-compose restart cbioportal
```

#### Import GRCh38 data

If you are importing GRCh38 data, please remember to set the `reference_genome: hg38` field in the `meta_study.txt` file. See also [cancer study metadata](/File-Formats.md).
#### Notes for non-Linux systems

##### Docker for Mac/Windows (newer versions)
Make sure to assign enough memory to Docker when using Docker for Windows (Windows 10 Pro 64-bit) or Docker for Mac (Mac OS X Yosemite 10.10.3 or above). In Mac OS X this can be set when clicked on the Docker icon -> Preferences... -> Adjust the Memory slider. By default it's set to 2 GB, which is too low and causes problems when loading multiple studies.

##### Docker-machine (older versions)
Because the Docker Engine daemon uses Linux-specific kernel features, you can‚Äôt run Docker Engine natively in Windows or Mac OS X. In versions of these systems that do not support the newer lightweight virtualisation technologies mentioned above, you must instead use the Docker Machine command, `docker-machine`. This creates and attaches to a small Linux VM on your machine, which hosts Docker Engine.

The Docker Quickstart Terminal in the Docker Toolbox will automatically create a default VM for you (`docker-machine create`), boot it up (`docker-machine start`) and set environment variables in the running shell to transparently forward the docker commands to the VM (`eval $(docker-machine env)`). Do note however, that forwarded ports in the docker commands will pertain to the VM and not your Windows/OS X system. The local cBioPortal and MySQL servers will not be available on `localhost` or `127.0.0.1`, but on the address printed by the command `docker-machine ip`, unless you configure VirtualBox to further forward the port to the host system.
# Deploy with Docker

## Prerequisites

Docker provides a way to run applications securely isolated in a container, packaged with all its dependencies and libraries.
To learn more on Docker, kindly refer here: [Docker overview](https://docs.docker.com/engine/docker-overview/).

Make sure that you have the latest version of Docker installed on your machine. [Get latest version](https://www.docker.com/products/overview#/install_the_platform)

[Notes for non-Linux systems](notes-for-non-linux.md)

## Usage instructions
In this example we use [Docker Compose](https://docs.docker.com/compose/) to spin up all the different required containers/services for cBioPortal.

### Quick Start

```
git clone https://github.com/cBioPortal/cbioportal-docker-compose.git
cd cbioportal-docker-compose
./init.sh
docker compose up
```

You should now be able to see the cBioPortal website at http://localhost:8080

Import studies with:

```
docker compose run cbioportal metaImport.py -u http://cbioportal:8080 -s study/lgg_ucsf_2014/ -o
```

Clear persistent data volumes with:

```
docker compose down -v
```

### Comprehensive Start

#### Step 1 - Run Docker Compose

Download the git repo that has the Docker compose file and go to the root of that folder:

```
git clone https://github.com/cBioPortal/cbioportal-docker-compose.git
cd cbioportal-docker-compose
```

Then download all necessary files (seed data, example config and example study from datahub) with the init script:
```
./init.sh
```

Then run:

```
docker compose up
```

This will start all four containers (services) defined [here](https://github.com/cBioPortal/cbioportal-docker-compose/blob/master/docker-compose.yml). That is:

- the mysql database, which holds most of the cBioPortal data
- the cBioPortal Java web app, this serves the React frontend as well as the REST API
- the session service Java web app. This service has a REST API and stores session information (e.g. what genes are being queried) and user specific data (e.g. saved cohorts) in a separate mongo database
- the mongo database that persists the data for the session service

It will take a few minutes the first time to import the seed database and perform migrations if necessary. Each container outputs logs to the terminal. For each log you'll see the name of the container that outputs it (e.g. `cbioportal_container` or `cbioportal_session_database_container`). If all is well you won't see any significant errors (maybe some warnings, that's fine to ignore). If all went well you should be able to visit the cBioPortal homepage on http://localhost:8080. You'll notice that no studies are shown on the homepage yet:

<img width="1414" alt="Screen Shot 2022-01-24 at 2 10 10 PM" src="https://user-images.githubusercontent.com/1334004/150848276-dec9551f-6b90-470f-bb59-e7754829fc83.png">


Go to the next step to see how to import studies.

##### Notes on detached mode

If you prefer to run the services in detached mode (i.e. not logging everything to your terminal), you can run

```
docker compose up -d
```

In this mode, you'll have to check the logs of each container manually using e.g.:

```
docker logs -f cbioportal_container
```

You can list all containers running on your system with

```
docker ps -a
```

#### Step 2 - Import Studies
To import studies you can run:

```
docker compose run cbioportal metaImport.py -u http://cbioportal:8080 -s study/lgg_ucsf_2014/ -o
```
This will import the [lgg_ucsf_2014 study](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014) into your local database. It will take a few minutes to import. After importing, restart the cbioportal web container:

```
docker compose restart cbioportal
```

or 

All public studies can be downloaded from https://www.cbioportal.org/datasets, or https://github.com/cBioPortal/datahub/. You can add any of them to the `./study` folder and import them. There's also a script (`./study/init.sh`) to download multiple studies. You can set `DATAHUB_STUDIES` to any public study id (e.g. `lgg_ucsf_2014`) and run `./init.sh`.

##### Notes on restarting
To avoid having to restart one can alternatively hit an API endpoint. To do so, call the `/api/cache` endpoint with a `DELETE` http-request (see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information):

```
curl -x DELETE -H "X-API-KEY: my-secret-api-key-value" http://localhost:8080/api/cache
```

The value of the API key is configured in the _application.properties_ file. You can visit http://localhost:8080 again and you should be able to see the new study.

#### Step 3 - Customize your application.properties file ###

The properties file can be found in `./config/application.properties`. Which was set up when running `init.sh`.

This properties file allows you to customize your instance of cBioPortal with e.g. custom logos, or point the cBioPortal container to e.g. use an external mysql database. See the [properties](/deployment/customization/Customizing-your-instance-of-cBioPortal.md) documentation for a comprehensive overview.

If you would like to enable OncoKB see [OncoKB data access](/deployment/integration-with-other-webservices/OncoKB-Data-Access.md) for 
how to obtain a data access token. After obtaining a valid token use:

#### Step 4 - Customize cBioPortal setup
To read more about the various ways to use authentication and parameters for running the cBioPortal web app see the relevant [backend deployment documentation](/deployment/customization/Customizing-your-instance-of-cBioPortal.md).

On server systems that can easily spare 4 GiB or more of memory, set the `-Xms`
and `-Xmx` options to the same number. This should increase performance of
certain memory-intensive web services such as computing the data for the
co-expression tab. If you are using MacOS or Windows, make sure to take a look
at [these notes](notes-for-non-linux.md) to allocate more memory for the
virtual machine in which all Docker processes are running.

## More commands ##
For documentation on how to import a study, see [this tutorial](import_data.md)
For more uses of the cBioPortal image, see [this file](example_commands.md)

To Dockerize a Keycloak authentication service alongside cBioPortal,
see [this file](using-keycloak.md).

## Uninstalling cBioPortal ##
```
docker compose down -v --rmi all
```
# Authenticating and Authorizing Users using Keycloak in Docker

This guide describes a way to Dockerise Keycloak along with cBioPortal, for authentication.

First, create an isolated network in which the Keycloak and MySQL servers can talk to one another.

```shell
docker network create kcnet
```

Run a MySQL database in which Keycloak can store its data. This database server will not be addressable from outside the Docker network. Replace `<path_to_database>` with the absolute path where the folder `kcdb-files` will be placed. This folder is used by the database to store its files.

```shell
docker run -d --restart=always \
    --name=kcdb \
    --net=kcnet \
    -v "<path_to_database>/kcdb-files:/var/lib/mysql" \
    -e MYSQL_DATABASE=keycloak \
    -e MYSQL_USER=keycloak \
    -e MYSQL_PASSWORD=password \
    -e MYSQL_ROOT_PASSWORD=root_password \
    mysql:5.7
```

Then run the actual Keycloak server, using [this image](https://hub.docker.com/r/jboss/keycloak/) available from Docker Hub. This will by default connect to the database using the (non-root) credentials in the example above. The server will be accessible to the outside world on port 8180, so make sure to choose a strong administrator password.

The command below uses the default values for `MYSQL_DATABASE`, `MYSQL_USER` and `MYSQL_PASSWORD` (listed in the command above). If you wish to change these credentials, specify them in the command below. For instance, if `MYSQL_USER` in the database container is `user`, you need to add `-e MYSQL_USER=user`.

```
docker run -d --restart=always \
    --name=cbiokc \
    --net=kcnet \
    -p 8180:8080 \
    -e DB_VENDOR=mysql \
    -e DB_ADDR=kcdb \
    -e KEYCLOAK_USER=admin \
    -e "KEYCLOAK_PASSWORD=<admin_password_here>" \
    jboss/keycloak:4.8.3.Final
```

Finally, configure Keycloak and cBioPortal as explained in the [Keycloak documentation](./../authorization-and-authentication/Authenticating-and-Authorizing-Users-via-keycloak.md). Remember to specify port 8180 for the Keycloak server, wherever the guide says 8080.

After configuring Keycloak, set up cBioPortal containers [as specified in the documentation](./). Make sure to update the `-Dauthenticate` in the [docker-compose file](https://github.com/cBioPortal/cbioportal-docker-compose/blob/5da068f0eb9b4f42db52ab5e91321b26a1826d7a/docker-compose.yml#L20) to `-Dauthenticate=saml`.
# Introduction

Using OncoKB does not require a data access token. But the instance you are connecting to only includes biological information by default. 
If you want to include tumor type summary, therapeutic levels and more, please consider obtaining a license from OncoKB.

# How to obtain an OncoKB license
1. Please review OncoKB [terms](https://www.oncokb.org/terms) 
2. Please request for [data access](https://www.oncokb.org/dataAccess)
3. You can find your token in your [Account Settings](https://www.oncokb.org/account/settings) after login.

# Set up cBioPortal to include full OncoKB content
Following properties can be edited in the `application.properties` file or set in system variables if you are using docker. 
- `show.oncokb` should be set to `true`
- `oncokb.token` should be set to a valid OncoKB access token value
- `oncokb.public_api.url` should be set to `https://www.oncokb.org/api/v1`

Thank you for supporting future OncoKB development.

# Include MSI-H and TMB-H annotation
If you want to include the MSI-H and TMB-H annotation on patient view, please follow the instruction to [import required clinical data](/File-Formats.md#clinical-data).  
For MSI-H, a clinical attribute MSI_TYPE with value Instable is required.  
For TMB-H, a clinical attribute TMB_SCORE with value >=10 is required.  

# Disable OncoKB Service

Please set `show.oncokb` to `false` in `application.properties` or in system variables if you are using docker.
# Integration with Other Webservices

# Deployment

Private instances of cBioPortal are maintained by institutions and companies [around the world](https://www.cbioportal.org/installations).

An instance can be deployed using Docker **(recommended)** or by building and deploying from source. The source code of cBioPortal is [available](https://github.com/cBioPortal/cbioportal) on GitHub under the terms of [Affero GPL V3](https://www.gnu.org/licenses/agpl-3.0.en.html).

This section contains instructions for both of these paths.

Please note that installing a local version requires system administration skills; for example, installing and configuring Tomcat and MySQL. With limited resources, we cannot provide technical support on system administration.
# Build with different frontend versions

Maven will build cBioPortal with a cBioPortal-frontend version and git repository url as determined by respectively the _frontend.version_ and _frontend.groupId_ parameters in the root POM.xml.

```
  <properties>
    <frontend.version>v2.1.0_/frontend.version>
    <frontend.groupId>com.github.cbioportal</frontend.groupId>
    ...
```
To build cBioPortal with a different frontend version different values for  _frontend.version_ and _frontend.groupId_ parameters can be specified as part of the maven install command. For example:

```
mvn clean -DskipTests install -Dfrontend.version=93d9cbcbf007ff620ab51ef5af5927a0eb1ebed4 -Dfrontend.groupId=com.github.thehyve
```

Remarks:
- The _frontend.version_ parameter allows release tags (e.g. 'v2.1.0') and commit sha-hashes (e.g., '93d9cbcb').
- The _frontend.groupId_ is a reversed, dot-separated derivative of the git url. Git repository location _github.com/cbioportal_ is represented by the _com.github.cbioportal_ groupId.# cBioPortal ER Diagram
[cBioPortal ER Diagram - PDF Version](https://github.com/cBioPortal/cbioportal/blob/master/db-scripts/src/main/resources/cbioportal-er-diagram.pdf)

![cBioPortal ER Diagram](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/db-scripts/src/main/resources/cbioportal-er-diagram.png)
# cBioPortal Database Versioning

We follow the following logic when deciding how/when to increment the version of cBioPortal database. It's a complete modification of semantic versioning (MAJOR.MINOR.PATCH) more suitable for our purposes:

**MAJOR** : A non-backward compatible significant change in the database. Which requires the maintainer to reload and re-import all studies in the database entirely.

**MINOR** : Including deleting, renaming tables or columns, or changing constraints

Delete a table

```SQL
DROP TABLE t ;
```

Drop column c from the table

```SQL
ALTER TABLE t DROP COLUMN c ;
```

Add a constraint

```SQL
ALTER TABLE t ADD constraint;
```

Drop a constraint

```SQL
ALTER TABLE t DROP constraint;
```

Rename a table from t1 to t2

```SQL
ALTER TABLE t1 RENAME TO t2;
```

Rename column c1 to c2

```SQL
ALTER TABLE t1 RENAME c1 TO c2 ;
```

**PATCH** : Changes that don't change existing database schemes but add new tables or columns, manipulating data.

Create a new table

```SQL
CREATE TABLE t (
     id INT PRIMARY KEY,
);
```

Add a new column to the table

```SQL
ALTER TABLE t ADD column;
```

Remove all data in a table

```SQL
TRUNCATE TABLE t;
```

Delete data from a table:

```SQL
DELETE from TABLE t;
```

Update data in a table:

```SQL
UPDATE TABLE t SET ...;
```

Insert multiple rows into a table

```SQL
INSERT INTO t(column_list)
VALUES (value_list), 
       (value_list), ‚Ä¶;
```
# Deployment Procedure
This describes our internal deployment procedure. Shared publicly, in case it
may be of use. Instructions on how to deploy cBioPortal can be found elsewhere,
see e.g. [Deploying the web application](/deployment/deploy-without-docker/Deploying.md) and [Deploy using
Docker](/deployment/docker/).

We deploy the master branch of backend and the master branch of frontend to
production. The public portal (https://www.cbioportal.org) runs on AWS EKS. The configuration 
can be found in the knowledgesystems repo:

https://github.com/knowledgesystems/knowledgesystems-k8s-deployment 

Other internal MSK portals run on AWS EKS as well.

The frontend and backend can be upgraded independently. We have the following
events that can require a new deployment:

- New **frontend** commit in master
- New **backend** commit in master

## New frontend commit in master
Currently we are auto-deploying the frontend master branch to netlify:
https://frontend.cbioportal.org. So any change should be automatically
built and deployed to the relevant portals if the frontend configuration has
been set up properly. Do note that the current build time for the frontend
project is ~15 minutes or so. To see what frontend commit is deployed, check
`window.FRONTEND_COMMIT` in the console of the browser.

### Public Portal Frontend URL
The public portal is on AWS and running inside a Kubernetes cluster.
The URL that it gets the frontend version from is here:

https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/search?q=-Dfrontend.url&unscoped_q=-Dfrontend.url

This should be a URL pointing to netlify.

### Internal Portal Frontend URL
For the internally runnning portals the frontend.url is defined in the
application.properties file in the mercurial portal-configuration repo. If set up
correctly, this should point to a file on both dashi and dashi2 that in turn
points to a netlify frontend URL. The reason we have a separate file with the
URL in it is that it allows us to update the frontend URL without redeploying
the backend.

## New backend commit in master
A new backend commit usually also means a new frontend change is necessary. For
this reason the following sections assume that's the case.

### Public Portal Backend Upgrade
Once the backend repo has been tagged on github, a docker image gets build on Docker Hub automatically. It can take ~5 min before the image is available. You can check here what the status of the builds is: https://github.com/cBioPortal/cbioportal/actions?query=workflow%3A%22Docker+Image+CI%22.


After that, if you have access to the kubernetes cluster you can change the image in the configuration of the kubernetes cluster:

https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/public-eks/cbioportal-prod/cbioportal_spring_boot.yaml

point this line, to the new tag on docker hub e.g.:

```
image: cbioportal/cbioportal:6.0.2-web-shenandoah
```

Make sure it is an image with the postfix `-web-shenandoah`. This is the image that only has the web part of cBioPortal and uses the shenandoah garbage collector.

Also remove the `-Dfrontend.url` parameter such that the frontend version inside the war will be used:

```
"-Dfrontend.url=https://frontend.cbioportal.org"
```

Then running this command applies the changes to the cluster:

```
kubectl apply -f public-eks/cbioportal-prod/cbioportal_spring_boot.yaml
```

You can keep track of what's happening by looking at the pods:

```
kubectl get po
```

If you have the watch command installed you can also use that to see the output
of this every 2s:

```
watch kubectl get po
```

Another thing to look at is the events:

```
kubectl get events --sort-by='{.lastTimestamp}'
```

If there are any issues, point the image back to what it was, set
`-Dfrontend.url` and run `kubectl apply -f filename` again.

If everything went ok, you can re-enable auto deployment on netlify, set
`-Dfrontend.url` in the kubernetes file and run `kubectl apply -f filename`
again.

Make sure to commit your changes to the knowledgesystems-k8s-deployment repo
and push them to the main repo, so that other people making changes to the
kubernetes config will be using the latest version. 


## Upgrading Related Backend Components
Backend upgrades involving the database schema, DAO classes, etc. require updates to databases and importers. CBioPortal has multiple MySQL databases (all using AWS RDS) backing different portals. Similarly, there are multiple importers responsible for loading portal-specific data. Every database must be manually migrated on an individual basis; all importers/data fetchers can be updated simultaneously through an existing deployment script.

Before upgrading, make sure to turn off import jobs in the crontab and alert the backend pipelines team (Avery, Angelica, Rob, Manda). 

To access the crontab, log in to pipelines, log in as cbioportal_importer: `sudo su - cbioportal_importer`, and run `crontab -e`. Comment out any lines that run import jobs, save, and exit. Make sure to uncomment these lines once the upgrade (database and importers) is complete. Lines that need to be commented out will be under the `Import Jobs` section, shown [here](https://github.com/knowledgesystems/cmo-pipelines/blob/942de83c0f9a731e301151d10dad73744cd9c9a0/import-scripts/mycrontab#L4).

## Updating Databases
First, make sure there is a backup of the database being migrated. 
If there is not a weekly dump, backup the database being migrated using mysqldump. This process may take awhile depending on the size of the database. 

```
mysqldump -u <user> -h <host> -p <database name> | gzip > <database_name>_`date +%Y%m%d_%H%M`.sql.gz 
```
    
The second step is to migrate the database. Make sure that the migration script is the same version as the deployed cBioPortal website. It is recommended to first test the migration script manually line-by-line in a copy of the existing database. This will catch any data-related bugs that might not be captured by the python migration script. After testing is successful, migrate the production databases following these steps [here](/Updating-your-cBioPortal-installation.md#running-the-migration-script). 

These are all cBioPortal databases and their locations:

| Website  | Database | Location |
| ------------- | ------------- | ------------- |
| cbioportal.mskcc.org  | cgds_gdac  | pipelines |
| cbioportal.org  | cgds_public  | AWS |
| genie.cbioportal.org | cgds_genie | AWS | 
| triage.cbioportal.org | cgds_triage | pipelines |

To obtain information such as usernames, passwords, hostnames - ask Avery, Angelica, Rob, Manda, and Ino. 

## Updating Importers/Data Fetchers
Importers (code found [here](https://github.com/knowledgesystems/pipelines)) and data fetchers (code found [here](https://github.com/knowledgesystems/cmo-pipelines)) use code from the cBioPortal codebase. The cbioportal dependency is packaged with the genome-nexus-annotation-pipeline and specified in the pipelines importer pom.

The following steps are used during releases/updates to build new importers with the **most-up-to-date** cBioPortal and genome-nexus-annotation-pipeline code. All steps should be performed on the pipelines machine. 

1. Set the jitpack hash [here](https://github.com/genome-nexus/genome-nexus-annotation-pipeline/blob/9510299395986653d9e9b672a38d472e52e7625b/pom.xml#L71) in the genome-nexus-annotation-pipeline codebase to the most recent cbioportal/cbioportal commit hash in master.

2. Merge this change into genome-nexus-annotation-pipeline/master.

3. Set the commit hash [here](https://github.com/knowledgesystems/pipelines/blob/f6c52bbda86b3929222d42c9bc84581fd6333fb4/pom.xml#L76) in the pipelines codebase to the most most recent genome-nexus/genome-nexus-annotation-pipeline commit hash **(after merge specfied in step 2)**. Also ensure the db version in the pom [here](https://github.com/knowledgesystems/pipelines/blob/f6c52bbda86b3929222d42c9bc84581fd6333fb4/pom.xml#L76) matches the db schema version in the cbioportal codebase. 

4. Merge this change into pipelines/master.

5. Set the commit hash [here](https://github.com/knowledgesystems/cmo-pipelines/blob/e740c9fa3d409ab75988e7a157682733e261fca5/cvr/pom.xml#L70) in the cmo-pipelines codebase to the most recent genome-nexus/genome-nexus-annotation-pipeline commit hash **(after merge specified in step 2)**

6. Merge this change into cmo-pipelines/master

7. Run the deployment wrapper script. See details [here](/Deployment-Procedure.md#deployment-script).  

8. Verify new importers/data fetchers have been placed in `/data/portal-cron/lib` by checking timestamps.
```
ls -tlra /data/portal-cron/lib
```

## Deployment Script
The wrapper script is found on pipelines here:
`/data/portal-cron/git-repos/pipelines-configuration/build-importer-jars/buildproductionjars.sh`.

Run `git pull` to pull in any updates to the build script.

The wrapper script takes two arguments:
1. --cbioportal-git-hash (required): Set to the cBioPortal commit hash being used in the pipelines build (hash specified in **step 1** of [updating importers](#updating-importersdata-fetchers)). This must match because the build copies out resource files (e.g application-context-business.xml) from the cbioportal codebase. 
2. --skip-deployment (optional): Set to true to skip auto-deployment to `/data/portal-cron/lib`. Built jars will be found in `/data/portal-cron/git-repos/pipelines-configuration/build-importer-jars/` and can be be manually moved.

The wrapper script will automatically backup the importers/data-fetchers to `/data/portal-cron/lib/backup`.
# Documentation site

This documentation site is created using [https://retype.com/](https://retype.com/), a static site generator based on markdown.

Visit their site for installation instructions and a guide
on how
- Develop, build and test the site locally
- Markdown syntax
- Deploy to Github pages using Github actions.

### Navigation
Please note that the navigation for the site is defined in docs/SUMMARY.md.
This is a deprecated form of configuration which is not documented in Retype.
# Feature Development Guide

This is a guide for developers that are implementing a new feature.

## Before Implementation

As a first step it is important to determine how complex the proposed feature is. Incremental improvements on existing features are often easier to accomplish and require input from fewer people. Most minor changes can be submitted as a Pull Request. If the proposed feature would require one or more days of work it makes sense to connect on [slack](https://slack.cbioportal.org) to discuss the idea. For more complex new features that require weeks of work or more, it is best to get input from several people in the cBioPortal community, including people with a deep understanding of the cBioPortal product and its users as well as the engineers that write the software. In that case we often start out with a Request For Comments document that describes the feature in more detail, see [our list of RFCs ](../RFC-List.md)for some examples. The community can then help guide the feature development in the right direction.

During this process you will most likely receive some pointers which part of the stack you will be editing (see [Architecture Overview](../Architecture-Overview.md)). This will be helpful when actually starting your implementation and figuring out how to set up your development environment. For many features it is not necessary to understand all parts of the stack, so seeking out advice on this is highly recommended.

Before you start implementing a more complex feature, ideally many of these things are clear:

* Who can you contact for help?
* Who will be helping to review the code?
* What part of the stack will you work on?

Gold stars if you already start thinking about:

* How do we release the feature incrementally?
* When is feature development done?

See more thoughts about these topics further below

## Starting Implementation

Once you are ready to start implementing, the first thing is to set up the development environment. We strive to make this as easy as possible, but it can often still take some time so definitely reach out if you run into issues. If you haven't submitted a Pull Request to cBioPortal before, it might make sense to look at some [good first issues](https://github.com/cBioPortal/cbioportal/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) before starting with your feature. This will help to get some familiarity with the process of proposing a change, getting it reviewed, making edits and getting it to production. Don't pick anything to complicated for a first issue, it could be as simple as fixing some typos in the `README`.

## During Implementation

The most important part during implementation is communication. Continue getting feedback as your implementation evolves. One of the best ways to do this is to fully integrate into the development team while you work on a feature. Anybody is welcome to join our weekly planning meeting (Tuesdays 11AM-12PM Eastern Time) and our daily standups 2.30-3PM Eastern Time. Please reach out on [slack](https://slack.cbioportal.org) to get an invite. If these times are not ideal or you're working on the feature more sporadically then it's totally fine to skip them.

## Plan to release to production early

Don't wait until the feature is fully finished to get feedback from the product team and the engineering team. Think about ways we can release a portion of the feature to production without finishing the entire thing. We have found feature flags to work well here. Instead of using long running feature branches we try to add a configuration property that allows us to turn the feature on or off. That way portions of the code can be released to production early on. We want to avoid working on some piece of code for more than a week or so without being able to release it. For instance: if one is trying to add some new tab on the Patient View Page, one could start with adding the on/off configuration switch for this tab. That could be released to production relatively quickly.

## Regression Testing

Make sure to think about ways to incorporate testing for your feature. We have an extensive suite of unit, integration and end to end tests including automated browser testing that mimics user interaction. Adding some regression tests will make sure the feature won't break with new versions of cBioPortal.

## When is feature development done?

There are many stages in feature development:

* Design
* Implementation
* Review
* Production Deployment
* Production Usage Monitoring

The process is hardly ever a linear line and it can move back and forth between any stage. This is expected and one of the reasons why time estimations are notoriously hard. An additional note is that feature development usually isn't done at the moment it gets merged to the main branch and deployed, but rather only after a few weeks of using it in production and not identifying any new issues.
# Manual test cases

Linked below is a Google Doc with some manual test cases you might want to look at when developing, and deploying new changes. Note that this list is a work in progress and subject to change.

<https://docs.google.com/document/d/1IniMNrrSEGsPRtkaU3Vbcm6sMgcOhytBurIRg0SqXgA/edit?usp=sharing>
# Development

cBioPortal is an open source project hosted on [Github](https://www.github.com/cbioportal/cbioportal).  The core team will
consider pull requests from any source.

The following section describes how to set up a development environment and start contributing.

Note that cBioPortal is split into a backend project (api) and a frontend project which are kept in separate repositories.

- [Frontend](http://www.github.com/cbioportal/cbioportal-frontend)
  (Typescript, React, Mobx)
- [Backend](http://www.github.com/cbioportal/cbioportal)
  (Java, Spring, MySQL, MongoDB)

If you are interested in coordinating the development of new features, please contact cbioportal@cbioportal.org or reach out on https://slack.cbioportal.org.
# Release Procedure

We have release procedures for the following scenarios:

1. [cBioPortal community release of code already in production](Release-Procedure.md#cbioportal-community-release-of-code-already-in-production)
2. [Release with database migration](Release-Procedure.md#release-with-database-migration)

## cBioPortal community release of code already in production

We often run code in production that is not ready yet for use by the wider cBioPortal community. The frontend gets deployed to production after every merge to master. The backend gets updated every Tuesday before our community call (occasionally more frequently too if issues are identified). We tag whatever is in the master and frontend repo and put a "Pre-Release" indication on it. After a month of stable usage in production, one of the tags will get the "latest" indication on the GitHub. These are the steps we follow to create a release:

1. Create a new frontend tag. The releases can be found here: https://github.com/cBioPortal/cbioportal-frontend/releases. A draft of the release notes are automatically generated by https://github.com/marketplace/actions/release-drafter. If there are pull requests in the `Changes` section i.e. they have not been labeled with one of the labels defined [here](https://github.com/cBioPortal/cbioportal-frontend/blob/master/.github/release-drafter.yml). Try to label them and trigger a rerun by committing something to the master branch. Alternatively you can manually put them in a particular section. Note that our goal is to have automated release notes, so it would be great if you could send a PR to update the [release-drafter.yml](https://github.com/cBioPortal/cbioportal-frontend/blob/master/.github/release-drafter.yml) in case you find certain PRs don't fit in a particular section or a section should be altered. Look at other release notes for inspiration: https://github.com/cBioPortal/cbioportal-frontend/releases. You can save your work as a draft if necessary.
2.  Once the frontend code is tagged, create a pull request to the backend repo where the frontend version is incremented in `portal/pom.xml`:

    ```
       <groupId>com.github.cbioportal</groupId>
       <artifactId>cbioportal-frontend</artifactId>
       <version>CHANGE_THIS_TO_THE_FRONTEND_TAG</version>
    ```
3.  Once that PR is merged, one can create a tag for the backend repo with the same tag as the frontend repo. You should see a draft from release drafter similar to the frontend in the backend repo: https://github.com/cBioPortal/cbioportal/releases. The idea is to create one set of release notes in the backend repo that is a combination of the frontend and backend notes. To make the hyperlinks from the frontend repo work in the backend repo you can copy the frontend release notes raw markdown and run the following one liner to convert the links:

    ```
    pbpaste | sed 's|(#\(....\)|([cbioportal-frontend#\1](https://github.com/cBioPortal/cbioportal-frontend/pull/\1)|g'
    ```

    then put them in the right sections following same style as other releases: https://github.com/cBioPortal/cbioportal-frontend/releases.
4. Create a news item with a link to your carefully crafted release notes. Highlight a few major changes that could be interesting to users of cBioPortal ideally with a screenshot similar to: https://github.com/cBioPortal/cbioportal/pull/6914/files?short\_path=6f95322#diff-6f953229832059bab3fe229d4af08b52 (in the files changed section, you can click on view rich diff to see the converted markdown).

## Backend Hotfixes
Occasionally there are bugs identified in the backend running in production. Creating an entire new release can be cumbersome so in this case, we create a new release branch instead. E.g. if the current release is `6.0.2`, the new branch would be `release-6.0.3`. Any fixes are then merged to this branch and if it looks good, they are deployed to production. On the following Tuesday whatever's in this `release-x` branch get merged to master and included in that Tuesday's release. 

## Release with database migration

For releases with database migrations, we increase the MINOR number in MAJOR.MINOR.PATCH. For those releases we have a separate branch (see https://github.com/cBioPortal/cbioportal/blob/master/CONTRIBUTING.md#branches-within-cbioportal), which needs to be merged to master on both backend and frontend:

1. Make sure no auto deployment is running for frontend from netlify
2. Merge frontend release-x.y.z branch to frontend master
3. Follow same procedure as for [a PATCH release](Release-Procedure.md#cbioportal-community-release-of-code-already-in-production), but instead of having a separate PR to update the frontend (step 2) one can add it to the already existing backend branch release-x.y.z and open the PR from there to backend's master. This is merely for convenience to avoid having to create another branch just to update the frontend version.

## A note on versioning

### From pre-release to official release

On the [GitHub Release Page](https://github.com/cBioPortal/cbioportal/releases) you will see that some releases have the pre-release indication whereas others do not. In general we make a new pre-release release every week. We test it out in production (https://cbioportal.org) for one month and if no new critical issues are identified we make it an official release. Occasionally we make a new official release in less than a month's time if we identify a critical issue in the previous release.

### cBioPortal Software Version Numbers

We follow the following logic when deciding how/when to increment the version of cBioPortal. It's a complete modification of semantic versioning (MAJOR.MINOR.PATCH) more suitable for our purposes:

MAJOR : A big change in how cBioPortal works. We changed the major version from 1 to 2 when we completely moved from using JSPs to a Single Page App written in React calling a REST service. Another example: we changed from 2 to 3 when we made session-service a requirement.

MINOR : Changes that require a database migration, an upgrade to another cBioPortal component (e.g. [session-service](session-service-working.md)) or anything that could require additional effort for a deployer of cBioPortal (e.g. see [transcript change](https://github.com/cBioPortal/cbioportal-frontend/pull/4350)).

PATCH : Changes that don't require database migrations. Could be new features as well as bug fixes to either frontend, backend or both.

See also: https://github.com/cBioPortal/cbioportal/releases

### cBioPortal Database Version Numbers

cBioPortal database version numbers are different version numbers than the software version numbers, it's only updated when developers make database scheme changes, please see [cBioPortal Database Versioning](Database-Versioning.md)

## User announcements

The following is a provisional system of alerting user to new features and announcements. It would probably be better for these messages to be configurable on an instance level by installers. For now, the following suffices.

Top banner: see sample configuration in src/shared/components/userMessager/UserMessage.tsx

For a beacon and associated dialog message, use this as a model, where child of InfoBeacon component is any component that will be shown when InfoBeacon is moused over and conditions are met

```
<If condition={!isWebdriver()}>
  <InfoBeacon
      top={-15}
      right={45}
      interaction={'mouseover'}
      color={'green'}
      id={'groupComparison1'}
 >
 <WrappedTour/>
 </InfoBeacon>
</If>
```
# Session Service

## 1. Steps to add new Session type through session-service

1. Define new type at [SessionType.java](https://github.com/cBioPortal/session-service/blob/master/src/main/java/org/cbioportal/session\_service/domain/SessionType.java).
2. Add necessary tests if required.
3. Update [session-service](https://github.com/cBioPortal/session-service#valid-type) documentation.

## 2. cBioPortal Backend

### 2.1 configuring session-service

Here are the properties that needs to be set

```
WARNING: do not use session service with -Dauthenticate=false
 either use authentication or change to -Dauthenticate=noauthsessionservice
session.service.url=
#if basic authentication is enabled on session service one should set:
session.service.user=
session.service.password=
```

#### session.service.url format

```
session.service.url: http://[host]:[port]/[session_service_app]/api/sessions/[portal_instance]/
```

#### example

```
session.service.url: http://localhost:8080/session_service/api/sessions/public_portal/
```

### 2.2 Updates to code

1. Update session-service dependency version in [pom.xml](https://github.com/cBioPortal/cbioportal/blob/master/pom.xml).
2. Add/Update api's in [SessionServiceController.java](https://github.com/cBioPortal/cbioportal/blob/master/web/src/main/java/org/cbioportal/web/SessionServiceController.java) to support new session type.
3. Sometimes we might need to defined model for session. Check how session class is extended in [VirtualStudy](https://github.com/cBioPortal/cbioportal/blob/master/web/src/main/java/org/cbioportal/web/parameter/VirtualStudy.java) and how it is consumed in [SessionServiceController.java](https://github.com/cBioPortal/cbioportal/blob/master/web/src/main/java/org/cbioportal/web/SessionServiceController.java).

## 3. cBioPortal Frontend

1. Client code for cbio session service is in https://github.com/cBioPortal/cbioportal-frontend/blob/master/src/shared/api/sessionServiceAPI.ts. Update it accordingly when the new api added in cBioPortal backend SessionServiceController.java

## 4. Local development

### 4.1 Expose database port

If [docker-compose.yml](https://github.com/cBioPortal/session-service/blob/master/docker-compose.yml) is used for running session service and if you like to access mongodb directly then uncomment this block of [code](https://github.com/cBioPortal/session-service/blob/master/docker-compose.yml#L21)

### 4.2 Test results with cURL in terminal

After you have local cBioPortal instance running, you can use cURL to test session-service endpoints.

**Example for GET methods:**

```
curl 'http://localhost:8080/api/session/custom_gene_list' \
  -H 'Cookie: COOCKIE_COPIED_FROM_BROWSER_REQUEST'
```

```
curl 'http://localhost:8080/api/session/custom_gene_list' \
  -H 'Cookie: _ga=GA1.1.984976088.1616537420; _ga_F1EX2C1M11=GS1.1.1616537420.1.1.1616538749.0; JSESSIONID=E7498ECA52F30588CAA1B34F420122DA'
```

**Example for POST methods:**

```
curl 'http://localhost:8080/api/session/custom_gene_list/save' \
  -H 'Content-Type: application/json' \
  -H 'Accept: */*' \
  -H 'Cookie: COOCKIE_COPIED_FROM_BROWSER_REQUEST' \
  --data-raw '{"name":"test","description":"test description","geneList":["GENE1","GENE2"]}'
```

```
curl 'http://localhost:8080/api/session/custom_gene_list/save' \
  -H 'Content-Type: application/json' \
  -H 'Accept: */*' \
  -H 'Cookie: _ga=GA1.1.984976088.1616537420; _ga_F1EX2C1M11=GS1.1.1616537420.1.1.1616538749.0; JSESSIONID=04457E173F769F7BCF9072CD5480308F' \
  --data-raw '{"name":"test","description":"test description","geneList":["GENE1","GENE2"]}'
```

Note on **Cookie**: You can log in to your account in your local cBioPortal instance, after login, copy cookies from requests in network tab. See [login configuration](/deployment/deploy-without-docker/Deploying.md#optional-login) for more information.
# Downloads

This page describes the various files available for download. The first section is targeted towards [users](Downloads.md#user-downloads) of cBioPortal. The second section towards [maintainers](Downloads.md#instance-maintainer-downloads) of cBioPortal instances.

### User Downloads

There are several ways in which one can download data from cBioPortal including manual and programmatic approaches. See options outlined below.

#### Datasets Page

A zip file for each study on [cbioportal.org](https://www.cbioportal.org) can be download from the [Datasets Page](https://www.cbioportal.org/datasets). One can also use the R client [cBioPortalData](/API-and-API-Clients.md#r-client) to programmatically download all of these files.

#### Datahub

The files for each study are also available from our [datahub repository](https://github.com/cBioPortal/datahub). This is basically the extracted version of the zip files in the [Datasets Page](https://www.cbioportal.org/datasets). Note that this is a git LFS repo so if you are familiar with git you might prefer using this option.

#### API and API Clients

Besides downloading all the study data one can also request slices of the data using the API. A slice of the data could e.g. be "give me all the mutation data for one patient" or "get me all EGFR mutations for a particular group of samples". There are API clients available in a variety of languages including bash, R and Python. See for more information the [API documentation](/API-and-API-Clients.md#api-and-api-clients).

### Instance Maintainer Downloads

As an instance maintainer of cBioPortal there are a variety of files that might be helpful. See below.

#### Study staging files

Staging files for the studies on cbioportal.org can be download from the [Datasets Page](https://www.cbioportal.org/datasets). These studies can be validated and loaded in a local cBioPortal instances using the [validator and importer](Data-Loading.md). Any issues with a downloaded study can be reported on [cBioPortal DataHub](https://github.com/cBioPortal/datahub/).

**Example studies**

TCGA Provisional studies often contain many different data types. These are excellent examples to use as reference when creating your own staging files. A detailed description on supported data types can be found in the [File Formats documentation](/File-Formats.md).

#### Complete cBioPortal database

A MySQL database dump of the complete cbioportal.org database can be found here: http://download.cbioportal.org/mysql-snapshots/public-portal-dump.latest.sql.gz

#### Seed Database

The seed database is a MySQL dump for seeding a new instance of the cBioPortal. Instructions for loading the seed database can be found [here](/deployment/deploy-without-docker/Import-the-Seed-Database.md). The seed database for human can be downloaded from [cBioPortal Datahub](https://github.com/cBioPortal/datahub/tree/master/seedDB). A mouse version can be found [here](https://github.com/cBioPortal/datahub/tree/master/seedDB\_mouse).
* [Introduction](#introduction)
* [Formats](#formats)
    * [Cancer Study](#cancer-study)
    * [Cancer Type](#cancer-type)
    * [Clinical Data](#clinical-data)
    * [Discrete Copy Number Data](#discrete-copy-number-data)
    * [Continuous Copy Number Data](#continuous-copy-number-data)
    * [Segmented Data](#segmented-data)
    * [Expression Data](#expression-data)
    * [Mutation Data](#mutation-data)
    * [Structural Variant Data](#structural-variant-data)
    * [Fusion Data](#fusion-data) (DEPRECATED)
    * [Methylation Data](#methylation-data)
    * [Protein level Data](#protein-level-data)
    * [Case Lists](#case-lists)
    * [Timeline Data](#timeline-data)
    * [GISTIC 2.0 Data](#gistic-20-data)
    * [Mutsig Data](#mutsig-data)
    * [Gene Panel Data](#gene-panel-data)
    * [Gene Set Data](#gene-set-data)
    * [Study Tags file](#study-tags-file)
    * [Generic Assay](#generic-assay)
        * [Arm Level CNA Data](#arm-level-cna-data)
        * [Mutational Signature Data](#mutational-signature-data)
    * [Resource Data](#resource-data)
    * [Custom namespace columns](#custom-namespace-columns)

# Introduction

This page describes the file formats that cancer study data should assume in order to be successfully imported into the database.  Unless otherwise noted, all data files are in tabular-TSV (tab separated value) format and have an associated metadata file which is in a multiline record format.  The metadata and data files should follow a [few rules documented at the Data Loading page](Data-Loading.md#preparing-study-data).

# Formats

## Cancer Study

As described in the [Data Loading tool](Data-Loading.md) page, the following file is needed to describe the cancer study:

### Meta file
This file contains metadata about the cancer study. The file contains the following fields:

1. **type_of_cancer**: The cancer type abbreviation, e.g., "brca".  This should be the same cancer type as specified in the `meta_cancer_type.txt` file, if available. The type can be "mixed" for studies with multiple cancer types.
2. **cancer_study_identifier**: A string used to uniquely identify this cancer study within the database, e.g., "brca_joneslab_2013".  
3. **name**: The name of the cancer study, e.g., "Breast Cancer (Jones Lab 2013)".
4. **description**: A description of the cancer study, e.g., "Comprehensive profiling of 103 breast cancer samples. Generated by the Jones Lab 2013". This description may contain one or more URLs to relevant information.
5. **citation (Optional)**: A relevant citation, e.g., "TCGA, Nature 2012".
6. **pmid (Optional)**: One or more relevant pubmed ids (comma separated without whitespace). If used, the field citation has to be filled, too.
7. **groups (Optional)**: When using an authenticating cBioPortal, lists the user-groups that are allowed access to this study. Multiple groups are separated with a semicolon ";". The study will be invisible to users not in _at least one_ of the listed groups, as if it wasn't loaded at all. e.g., "PUBLIC;GDAC;SU2C-PI3K". see [User-Authorization](/deployment/authorization-and-authentication/User-Authorization.md) for more information on groups
8. **add_global_case_list (Optional)**: set to 'true' if you would like the "All samples" case list to be generated automatically for you. See also [Case lists](#case-lists).
9. **tags_file (Optional)**: the file name containing custom study tags for the [study tags](#study-tags-file).
10. **reference_genome (Optional)**: the study reference genome (e.g. `hg19`, `hg38`). Without specifying this property, the study will be assigned to the reference genome specified in `application.properties` (property `ucsc.build`).

### Example  
An example meta_study.txt file would be:
```
type_of_cancer: brca
cancer_study_identifier: brca_joneslab_2013
name: Breast Cancer (Jones Lab 2013)
description: Comprehensive profiling of 103 breast cancer samples. Generated by the Jones Lab 2013.
add_global_case_list: true
```

## Cancer Type
If the type_of_cancer specified in the meta_study.txt does not yet exist in the `type_of_cancer` database table, a `meta_cancer_type.txt` file is also mandatory.

### Meta file
The file is comprised of the following fields:

1. **genetic_alteration_type**: CANCER_TYPE
2. **datatype**: CANCER_TYPE
3. **data_filename**: your datafile

### Example
An example meta_cancer_type.txt file would be:
```
genetic_alteration_type: CANCER_TYPE
datatype: CANCER_TYPE
data_filename: cancer_type.txt
```

### Data file
The file is comprised of the following columns in the order specified:

1. **type_of_cancer**: The cancer type abbreviation, e.g., "brca".
2. **name**: The name of the cancer type, e.g., "Breast Invasive Carcinoma".
3. **dedicated_color**: CSS color name of the color associated with this cancer study, e.g., "HotPink". See [this list](https://www.w3.org/TR/css3-color/#svg-color) for supported names, and follow the [awareness ribbons](https://en.wikipedia.org/wiki/List_of_awareness_ribbons) color schema. This color is associated with the cancer study on various web pages within the cBioPortal.
4. **parent_type_of_cancer**:  The `type_of_cancer` field of the cancer type of which this is a subtype, e.g., "Breast". :information_source: : you can set parent to `tissue`, which is the reserved word to place the given cancer type at "root" level in the "studies oncotree" that will be generated in the homepage (aka query page) of the portal. 

### Example
An example record would be:
```
brca<TAB>Breast Invasive Carcinoma<TAB>HotPink<TAB>Breast
```

## Clinical Data

The clinical data is used to capture both clinical attributes and the mapping between patient and sample ids. The software supports multiple samples per patient.  

As of March 2016, the clinical file is split into a patient clinical file and a sample clinical file. The *sample* file is required, whereas the *patient* file is optional. cBioPortal has specific functionality for a core set of patient and sample columns, but can also display custom columns (see section ["Custom columns in clinical data"](#custom-columns-in-clinical-data)).

### Meta files
The two clinical metadata files (or just one metadata file if you choose to leave the *patient* file out) have to contain the following fields:

1. **cancer_study_identifier**: same value specified in meta_study.txt
2. **genetic_alteration_type**: CLINICAL
3. **datatype**: PATIENT_ATTRIBUTES or SAMPLE_ATTRIBUTES
4. **data_filename**: your datafile

### Examples
An example metadata file, e.g. named meta_clinical_sample.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: CLINICAL
datatype: SAMPLE_ATTRIBUTES
data_filename: data_clinical_sample.txt
```

An example metadata file, e.g. named meta_clinical_patient.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: CLINICAL
datatype: PATIENT_ATTRIBUTES
data_filename: data_clinical_patient.txt
```

### Data files
For both patients and samples, the clinical data file is a two dimensional matrix with multiple clinical attributes. When the attributes are defined in the *patient* file they are considered to be patient attributes; when they are defined in the *sample* file they are considered to be sample attributes. 

The first four rows of the clinical data file contain tab-delimited metadata about the clinical attributes. These rows **have to start with a '#' symbol**. Each of these four rows contain different type of information regarding each of the attributes that are defined in the fifth row:

- Row 1: **The attribute Display Names**: The display name for each clinical attribute
- Row 2: **The attribute Descriptions**: Long(er) description of each clinical attribute
- Row 3: **The attribute Datatype**: The datatype of each clinical attribute (must be one of:  STRING, NUMBER, BOOLEAN)
- Row 4: **The attribute Priority**: A number which indicates the importance of each attribute.  In the future, higher priority attributes will appear in more prominent places than lower priority ones on relevant pages (such as the [Study View](https://www.cbioportal.org/study?id=brca_tcga)). A higher number indicates a higher priority.
    ```
    To promote certain chart in study view, please increase priority to a certain number. The higher the score, the higher priority it will be displayed in the study view.
    If you want to hide chart, please set the priority to 0. For combination chart, as long as one of the clinical attribute has been set to 0, it will be hidden.
    
    Currently, we preassigned priority to few charts, but as long as you assign a priority except than 1, these preassigned priorities will be overwritten.
    
    CANCER_TYPE: 3000, CANCER_TYPE_DETAILED: 2000,
    Overall survival plot: 400 (This is combination of OS_MONTH and OS_STATUS) 
    Disease Free Survival Plot: 300 (This is combination of DFS_MONTH and DFS_STATUS) 
    Mutation Count vs. CNA Scatter Plot: 200,
    Mutated Genes Table: 90, CNA Genes Table: 80, study_id: 70, # of Samples Per Patient: 40,
    With Mutation Data Pie Chart: 60, With CNA Data Pie Chart: 50, 
    Mutation Count Bar Chart: 30, CNA Bar Chart: 20,
    GENDER: 9, SEX: 9, AGE: 8
    ```
    
    Please note: 
    Priority is not the sole factor determining which chart will be displayed first.
    A layout algorithm in study view also makes a minor adjustment on the layout.
    The algorithm tries to fit all charts into a 2 by 2 matrix (Mutated Genes Table occupies 2 by 2 space).
    When a chart can not be fitted in the first matrix, the second matrixed will be generated. 
    And the second matrix will have lower priority than the first one. 
    If later chart can fit into the first matrix, then its priority will be promoted.
    
    Please see [here](/deployment/customization/Studyview.md) for more detailed information about how study view utilize priority and how the layout is calculated based on priority.
- Row 5: **The attribute name for the database**: This name should be in upper case.
- Row 6: This is the first row that contains actual data.


### Example clinical header
Below is an example of the first 4 rows with the respective metadata for the attributes defined in the 5th row. 
```
#Patient Identifier<TAB>Overall Survival Status<TAB>Overall Survival (Months)<TAB>Disease Free Status<TAB>Disease Free (Months)<TAB>...
#Patient identifier<TAB>Overall survival status<TAB>Overall survival in months since diagnosis<TAB>Disease free status<TAB>Disease free in months since treatment<TAB>...
#STRING<TAB>STRING<TAB>NUMBER<TAB>STRING<TAB>NUMBER<TAB>...
#1<TAB>1<TAB>1<TAB>1<TAB>1<TAB>
PATIENT_ID<TAB>OS_STATUS<TAB>OS_MONTHS<TAB>DFS_STATUS<TAB>DFS_MONTHS<TAB>...
....
data - see examples below
....
```

### Clinical patient columns
The file containing the patient attributes has one **required** column:
- **PATIENT_ID (required)**: a unique patient ID. This field allows only numbers, letters, points, underscores and hyphens.

The following columns are used by the study view as well as the patient view. In the [study view](https://www.cbioportal.org/study?id=brca_tcga) they are used to create the survival plots. In the patient view they are used to add information to the [header](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P05).

**Note on survival plots**: to generate the survival plots successfully, the columns are required to be in pairs, which means the file should have a pair of columns that have the same prefix but ending with `_STATUS` and `_MONTHS` individually. For example, `PFS_STATUS` and `PFS_MONTHS` are a valid pair of columns that can generate the survival plots.

**Note on survival status value**: the value of survival status must prefixed with `0:` or `1:`. Value with prefix `0:` means that no event (e.g. `LIVING`, `DiseaseFree`). Value with prefix `1:` means that an event occurred (e.g. `DECEASED`, `Recurred/Progressed`). 

- **OS_STATUS**:  Overall patient survival status
    - Possible values: 1:DECEASED, 0:LIVING
    - In the patient view, 0:LIVING creates a green label, 1:DECEASED a red label.
- **OS_MONTHS**:  Overall survival in months since initial diagnosis
- **DFS_STATUS**: Disease free status since initial treatment
    - Possible values: 0:DiseaseFree, 1:Recurred/Progressed
    - In the patient view, 0:DiseaseFree creates a green label, 1:Recurred/Progressed a red label.
- **DFS_MONTHS**: Disease free (months) since initial treatment

These columns, when provided, add additional information to the patient description in the header:
- **PATIENT_DISPLAY_NAME**: Patient display name (string)
- **GENDER** or **SEX**: Gender or sex of the patient (string)
- **AGE**: Age at which the condition or disease was first diagnosed, in years (number)
- **TUMOR_SITE**

Custom attributes:
- **Custom Clinical Attribute Headers**: Any other custom attribute can be added as well. See section ["Custom columns in clinical data"](#custom-columns-in-clinical-data).

### Example *patient* data file
```
#Patient Identifier<TAB>Overall Survival Status<TAB>Overall Survival (Months)<TAB>Disease Free Status<TAB>Disease Free (Months)<TAB>...
#Patient identifier<TAB>Overall survival status<TAB>Overall survival in months since diagnosis<TAB>Disease free status<TAB>Disease free in months since treatment<TAB>...
#STRING<TAB>STRING<TAB>NUMBER<TAB>STRING<TAB>NUMBER<TAB>...
#1<TAB>1<TAB>1<TAB>1<TAB>1<TAB>
PATIENT_ID<TAB>OS_STATUS<TAB>OS_MONTHS<TAB>DFS_STATUS<TAB>DFS_MONTHS<TAB>...
PATIENT_ID_1<TAB>1:DECEASED<TAB>17.97<TAB>1:Recurred/Progressed<TAB>30.98<TAB>...
PATIENT_ID_2<TAB>0:LIVING<TAB>63.01<TAB>0:DiseaseFree<TAB>63.01<TAB>...
...
```

### Clinical sample columns
The file containing the sample attributes has two **required** columns:
- **PATIENT_ID (required)**: A patient ID. This field can only contain numbers, letters, points, underscores and hyphens.
- **SAMPLE_ID (required)**: A sample ID. This field can only contain numbers, letters, points, underscores and hyphens.

By adding `PATIENT_ID` here, cBioPortal will map the given sample to this patient. This enables one to associate multiple samples to one patient. For example, a single patient may have had multiple biopsies, each of which has been genomically profiled. See [this example for a patient with multiple samples](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P04).

The following columns are required for the pan-cancer summary statistics tab ([example](https://www.cbioportal.org/index.do?cancer_study_id=msk_impact_2017&Z_SCORE_THRESHOLD=2&RPPA_SCORE_THRESHOLD=2&data_priority=0&case_set_id=msk_impact_2017_cnaseq&gene_list=BRAF&geneset_list=+&tab_index=tab_visualize&Action=Submit&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna)).
- **CANCER_TYPE**: Cancer Type
- **CANCER_TYPE_DETAILED**: Cancer Type Detailed, a sub-type of the specified CANCER_TYPE

The following columns affect the header of the patient view by adding text to the samples in the header:
- **SAMPLE_DISPLAY_NAME**: displayed in addition to the ID
- **SAMPLE_CLASS**
- **METASTATIC_SITE** or **PRIMARY_SITE**: Override TUMOR_SITE (patient level attribute) depending on sample type

The following columns additionally affect the [Timeline data](#timeline-data) visualization:
- **OTHER_SAMPLE_ID**: OTHER_SAMPLE_ID is no longer supported. Please replace this column header with SAMPLE_ID.   
- **SAMPLE_TYPE**, **TUMOR_TISSUE_SITE** or **TUMOR_TYPE**: gives sample icon in the timeline a color.
    - If set to `recurrence`, `recurred`, `progression` or `progressed`: orange
    - If set to `metastatic` or `metastasis`: red
    - If set to `primary` or otherwise: black

Custom attributes:
- **Custom Clinical Attribute Headers**: Any other custom attribute can be added as well. See section ["Custom columns in clinical data"](#custom-columns-in-clinical-data).

### Example sample data file
```
#Patient Identifier<TAB>Sample Identifier<TAB>Subtype<TAB>...
#Patient identifier<TAB>Sample Identifier<TAB>Subtype description<TAB>...
#STRING<TAB>STRING<TAB>STRING<TAB>...
#1<TAB>1<TAB>1<TAB>...
PATIENT_ID<TAB>SAMPLE_ID<TAB>SUBTYPE<TAB>...
PATIENT_ID_1<TAB>SAMPLE_ID_1<TAB>basal-like<TAB>...
PATIENT_ID_2<TAB>SAMPLE_ID_2<TAB>Her2 enriched<TAB>...
...
```

### Columns with specific functionality
These columns can be in either the patient or sample file.
- **CANCER_TYPE**: Overrides study wide cancer type
- **CANCER_TYPE_DETAILED**
- **KNOWN_MOLECULAR_CLASSIFIER**
- **GLEASON_SCORE**: Radical prostatectomy Gleason score for prostate cancer
- **HISTOLOGY**
- **TUMOR_STAGE_2009**
- **TUMOR_GRADE**
- **ETS_RAF_SPINK1_STATUS**
- **TMPRSS2_ERG_FUSION_STATUS**
- **ERG_FUSION_ACGH**
- **SERUM_PSA**
- **DRIVER_MUTATIONS**

### Custom columns in clinical data
cBioPortal supports custom columns with clinical data in either the patient or sample file. They should follow the previously described 5-row header format. Be sure to provide the correct `Datatype`, for optimal search, sorting, filtering (in [clinical data tab](https://www.cbioportal.org/study?id=brca_tcga#clinical)) and visualization.

The Clinical Data Dictionary from MSKCC is used to normalize clinical data, and should be followed to make the clinical data comparable between studies. This dictionary provides a definition whether an attribute should be defined on the patient or sample level, as well as provides a name, description and datatype. The data curator can choose to ignore these proposed definitions, but not following this dictionary might make comparing data between studies more difficult. It should however not break any cBioPortal functionality. See GET /api/ at [https://oncotree.mskcc.org/cdd/swagger-ui.html#/](https://oncotree.mskcc.org/cdd/swagger-ui.html#/) for the data dictionary of all known clinical attributes.

### Banned column names
`MUTATION_COUNT` and `FRACTION_GENOME_ALTERED` are auto populated clinical attributes, and should therefore not be present in clinical data files.

## Discrete Copy Number Data
The discrete copy number data file contain values that would be derived from copy-number analysis algorithms like [GISTIC 2.0](https://www.ncbi.nlm.nih.gov/sites/entrez?term=18077431) or [RAE](https://www.ncbi.nlm.nih.gov/sites/entrez?term=18784837). GISTIC 2.0 can be [installed](https://www.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=216&p=t) or run online using the GISTIC 2.0 module on [GenePattern](https://cloud.genepattern.org). For some help on using GISTIC 2.0, check the [Data Loading: Tips and Best Practices](Data-Loading-Tips-and-Best-Practices.md) page. When loading case list data, the `_cna` case list is required. See the [case list section](#case-lists).

### Wide vs Long format
For CNA data two formats are supported: the wide, and the long format:
- **Wide format**: a matrix, where each row is a gene, and each column is a sample
- **Long format**: not a matrix, each row is a gene-sample combination; this makes the file longer

### Wide format

#### Meta file 
The meta file is comprised of the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: COPY_NUMBER_ALTERATION
3. **datatype**: `DISCRETE`
4. **stable_id**: gistic, cna, cna_rae or cna_consensus
5. **show_profile_in_analysis_tab**: true
6. **profile_name**: A name for the discrete copy number data, e.g., "Putative copy-number alterations from GISTIC"
7. **profile_description**: A description of the copy number data, e.g., "Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification."
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id
10. **pd_annotations_filename (Optional)**: name of [custom driver annotations file](File-Formats.md#custom-driver-annotations-file)

##### Example
An example metadata file could be named meta_cna.txt and its contents could be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: COPY_NUMBER_ALTERATION
datatype: DISCRETE
stable_id: gistic
show_profile_in_analysis_tab: true
profile_name: Putative copy-number alterations from GISTIC
profile_description: Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification.
data_filename: data_cna.txt
pd_annotations_filename: data_cna_pd_annotations.txt
```

#### Data file
For each gene (row) in the data file, the following columns are required in the order specified:

One or both of:
- ***Hugo_Symbol***: A [HUGO](https://www.genenames.org/) gene symbol.
- ***Entrez_Gene_Id***: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier.

And:
- An additional column for each sample in the dataset using the sample id as the column header.

For each gene-sample combination, a copy number level is specified:
- "-2" is a deep loss, possibly a homozygous deletion
- "-1" is a single-copy loss (heterozygous deletion)
- "0" is diploid
- "1" indicates a low-level gain
- "2" is a high-level amplification.

##### Example
An example data file which includes the required column header would look like:
```
Hugo_Symbol<TAB>Entrez_Gene_Id<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
ACAP3<TAB>116983<TAB>0<TAB>-1<TAB>...
AGRN<TAB>375790<TAB>2<TAB>0<TAB>...
...
...
`````

### Long format

#### Meta file 
The meta file of **wide format** is comprised of the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: COPY_NUMBER_ALTERATION
3. **datatype**: `DISCRETE_LONG`
   Note: It will end up as datatype `DISCRETE` in the database, because the LONG data format is only relevant while importing. 
4. **stable_id**: gistic, cna, cna_rae or cna_consensus
5. **show_profile_in_analysis_tab**: true
6. **profile_name**: A name for the discrete copy number data, e.g., "Putative copy-number alterations from GISTIC"
7. **profile_description**: A description of the copy number data, e.g., "Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification."
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id
10. **namespaces (Optional)**: Comma-delimited list of `namespaces` to import. 

##### Example
An example metadata file could be named meta_cna.txt and its contents could be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: COPY_NUMBER_ALTERATION
datatype: DISCRETE_LONG
stable_id: gistic
show_profile_in_analysis_tab: true
profile_name: Putative copy-number alterations from GISTIC
profile_description: Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification.
data_filename: data_cna.txt
namespaces: MyNamespace,MyNamespace2
```

#### Data file
Each row contains a row-sample combination. Custom driver annotations are added as columns to the data file, just like custom namespace columns.

##### Example
An example data file which includes the required column header would look like:
```
Hugo_Symbol	Entrez_Gene_Id	Sample_Id	Value	cbp_driver	cbp_driver_annotation	cbp_driver_tiers	cbp_driver_tiers_annotation	MyNamespace.column1
ACAP3	116983	TCGA-A2-A04U-01	2	Putative_Passenger	Test passenger	Class 2	Class annotation	value1
...
```

#### Adding your own discrete copy number columns
Additional columns can be added to the discrete copy number **long** data file. In this way, the portal will parse and store your own CNA fields in the database.

See [Custom namespace columns](#custom-namespace-columns) for more information on adding custom columns to data files.

### Custom driver annotations file

Custom driver annotations can be defined for discrete copy number data. These annotations can be used to complement or replace default driver annotation resources OncoKB and HotSpots.
Custom driver annotations can be placed in a separate file that is referenced by the `pd_annotations_file` field of the meta file. The annotation file can hold the following columns:

1. **Hugo_Symbol (Optional)**: A [HUGO](https://www.genenames.org/) gene symbol. Required when column `Entrez_Gene_Id` is not present.
2. **Entrez_Gene_Id (Optional)**: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier. Required when column `Hugo_Symbol` is not present.
3. **SAMPLE_ID**: A sample ID. This field can only contain numbers, letters, points, underscores and hyphens.
4. **cbp_driver (Optional)**: "Putative_Driver", "Putative_Passenger", "Unknown", "NA" or "" (empty value). This field must be present if the cbp_driver_annotation is also present in the MAF file. 
5. **cbp_driver_annotation (Optional)**: Description field for the cbp_driver value (limited to 80 characters). This field must be present if the cbp_driver is also present in the MAF file. This field is free text. Example values for this field are: "Pathogenic" or "VUS".
6. **cbp_driver_tiers (Optional)**: Free label/category that marks the mutation as a putative driver such as "Driver", "Highly actionable", "Potential drug target". . This field must be present if the cbp_driver_tiers_annotation is also present in the MAF file. In the OncoPrint view's Mutation Color dropdown menu, these tiers are ordered alphabetically. This field is free text and limited to 20 characters. For mutations without a custom annotation, leave the field blank or type "NA".
7. **cbp_driver_tiers_annotation (Optional)**: Description field for the cbp_driver_tiers value (limited to 80 characters). This field must be present if the cbp_driver_tiers is also present in the MAF file. This field can not be present when the cbp_driver_tiers field is not present. 

All genes referenced in the custom driver annotation file must be present in the data file for discrete copy number alterations.

The `cbp_driver` column flags the mutation as either driver or passenger. In cBioPortal, passenger mutations are also known as variants of unknown significance (VUS). The `cbp_driver_tiers` column assigns an annotation tier to the mutation, such as "Driver", "Highly actionable" or "Potential drug target". When a tier is selected, mutations with that annotation are highlighted as driver. Both types of custom annotations contain a second column with the suffix `_annotation`, to add a description. This is displayed in the tooltip that appears when hovering over the sample's custom annotation icon in the OncoPrint view.

You can learn more about configuring these annotations in the [application.properties documentation](/deployment/customization/application.properties-Reference.md#custom-annotation-of-driver-and-passenger-mutations). When properly configured, the customized annotations appear in the "Mutation Color" menu of the OncoPrint view: \
![schreenshot mutation color menu](images/screenshot-mutation-color-menu.png) 

### Example
An example data file which includes the required column header would look like:
```
SAMPLE_ID<TAB>Hugo_Symbol<TAB>Entrez_Gene_Id<TAB>cbp_driver<TAB>cbp_driver_annotation<TAB>cbp_driver_tiers<TAB>cbp_driver_tiers_annotation<TAB>...
TCGA-BH-A0E6-01<TAB>GENEA<TAB>116983<TAB>Putative_Driver<TAB>see: PMID:12345678<TAB>Highly actionable<TAB>Per decision 01/01/2020<TAB>
TCGA-BH-A0E6-01<TAB>GENEB<TAB>375790<TAB>Putative_Passenger<TAB>see: PMID:12345678<TAB><TAB><TAB>
...
`````

### GISTIC 2.0 Format
GISTIC 2.0 outputs a tabular file similarly formatted to the cBioPortal format, called `<prefix>_all_thresholded.by_genes.txt`.
In this file the gene symbol is found in the `Gene Symbol` column, while Entrez gene IDs are in the `Gene ID` or 
`Locus ID` column. Please rename `Gene Symbol` to `Hugo_Symbol` and `Gene ID` or `Locus ID` to `Entrez_Gene_Id`. The 
`Cytoband` column can be kept in the table, but note that these values are ignored in cBioPortal. cBioPortal uses 
cytoband annotations from the `map_location` column in NCBI's `Homo_sapiens.gene_info.gz` when loading genes into 
the seed database.

## Continuous Copy Number Data

### Meta file
The continuous copy number metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: COPY_NUMBER_ALTERATION.
3. **datatype**: CONTINUOUS
4. **stable_id**: linear_CNA 
5. **show_profile_in_analysis_tab**: false.
6. **profile_name**: A name for the copy number data, e.g., "copy-number values".
7. **profile_description**: A description of the copy number data, e.g., "copy-number values for each gene (from Affymetrix SNP6).".
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id

cBioPortal also supports log2 copy number data. If your data is in log2, change the following fields:

3. **datatype**: LOG2-VALUE
4. **stable_id**: log2CNA 

### Example
An example metadata file, e.g. meta_log2_cna.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: COPY_NUMBER_ALTERATION
datatype: LOG2-VALUE
stable_id: log2CNA
show_profile_in_analysis_tab: false
profile_description: Log2 copy-number values for each gene (from Affymetrix SNP6).
profile_name: Log2 copy-number values
data_filename: data_log2_cna.txt
```

### Data file

The log2 copy number data file follows the same format as expression data files.  See [Expression Data](#expression-data) for a description of the expression data file format.

### GISTIC 2.0 Format
GISTIC 2.0 outputs a tabular file similarly formatted to the cBioPortal format, called `<prefix>_all_data_by_genes.txt`. 
In this file the gene symbol is found in the `Gene Symbol` column, while Entrez gene IDs are in the `Gene ID` or 
`Locus ID` column. Please rename `Gene Symbol` to `Hugo_Symbol` and `Gene ID` or `Locus ID` to `Entrez_Gene_Id`. The 
`Cytoband` column can be kept in the table, but note that these values are ignored in cBioPortal. cBioPortal uses 
cytoband annotations from the `map_location` column in NCBI's `Homo_sapiens.gene_info.gz` when loading genes into 
the seed database.


## Segmented Data

A SEG file (segmented data; .seg or .cbs) is a tab-delimited text file that lists loci and associated numeric values. The segmented data file format is the output of the Circular Binary Segmentation algorithm (Olshen et al., 2004). This Segment data enables the 'CNA' lane in the Genomic overview of the Patient view (as [can be seen in this example](https://www.cbioportal.org/patient?sampleId=TCGA-BH-A0E6-01&studyId=brca_tcga)). 

### Meta file
The segmented metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: COPY_NUMBER_ALTERATION
3. **datatype**: SEG
4. **reference_genome_id**: Reference genome version. Supported values: "hg19"
5. **description**: A description of the segmented data, e.g., "Segment data for the XYZ cancer study.".
6. **data_filename**: your datafile

### Example:
An example metadata file, e.g. meta_cna_hg19_seg.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: COPY_NUMBER_ALTERATION
datatype: SEG
reference_genome_id: hg19
description: Somatic CNA data (copy number ratio from tumor samples minus ratio from matched normals) from TCGA.
data_filename: data_cna_hg19.seg
```

### Data file

The first row contains column headings and each subsequent row contains a locus and an associated numeric value. See also the [Broad IGV page on this format](https://www.broadinstitute.org/software/igv/SEG).

### Example:

An example data file which includes the required column header would look like:
```
ID<TAB>chrom<TAB>loc.start<TAB>loc.end<TAB>num.mark<TAB>seg.mean
SAMPLE_ID_1<TAB>1<TAB>3208470<TAB>245880329<TAB>128923<TAB>0.0025
SAMPLE_ID_2<TAB>2<TAB>474222<TAB>5505492<TAB>2639<TAB>-0.0112
SAMPLE_ID_2<TAB>2<TAB>5506070<TAB>5506204<TAB>2<TAB>-1.5012
SAMPLE_ID_2<TAB>2<TAB>5512374<TAB>159004775<TAB>80678<TAB>-0.0013
...
...
```

## Expression Data

An expression data file is a two dimensional matrix with a gene per row and a sample per column.  For each gene-sample pair, a real number represents the gene expression in that sample.  

### Meta file

The expression metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: MRNA_EXPRESSION
3. **datatype**: CONTINUOUS, DISCRETE or Z-SCORE
4. **stable_id**: see table below.
5. **source_stable_id**: Required when both conditions are true: (1) `datatype` = `Z-SCORE` and (2) this study contains [GSVA data](#gene-set-data). Should contain `stable_id` of the expression file for which this [Z-SCORE file is the statistic](#z-score-instructions).
5. **show_profile_in_analysis_tab**: false (you can set to **true** if Z-SCORE to enable it in the oncoprint, for example).
6. **profile_name**: A name for the expression data, e.g., "mRNA expression (microarray)".
7. **profile_description**: A description of the expression data, e.g., "Expression levels (Agilent microarray).".
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id

### Supported stable_id values for MRNA_EXPRESSION
For historical reasons, cBioPortal expects the `stable_id` to be one of those listed in the following static set.
The stable_id for continuous RNA-seq data has two options: `rna_seq_mrna` or `rna_seq_v2_mrna`. These options were added to distinguish between two different TCGA pipelines, which perform different types of normalization (RPKM and RSEM). However, for custom datasets either one of these `stable_id` can be chosen.

<table>
<thead>
<tr><th>datatype</th><th>stable_id</th><th>description</th></tr>
</thead>
<tr><td>CONTINUOUS</td><td>mrna_U133</td><td>Affymetrix U133 Array</td></tr>
<tr><td>Z-SCORE</td><td>mrna_U133_Zscores</td><td>Affymetrix U133 Array</td></tr>
<tr><td>Z-SCORE</td><td>rna_seq_mrna_median_Zscores</td><td>RNA-seq data</td></tr>
<tr><td>Z-SCORE</td><td>mrna_median_Zscores</td><td>mRNA data</td></tr>
<tr><td>CONTINUOUS</td><td>rna_seq_mrna</td><td>RNA-seq data</td></tr>
<tr><td>CONTINUOUS</td><td>rna_seq_v2_mrna</td><td>RNA-seq data</td></tr>
<tr><td>Z-SCORE</td><td>rna_seq_v2_mrna_median_Zscores</td><td>RNA-seq data</td></tr>
<tr><td>CONTINUOUS</td><td>mirna</td><td>MicroRNA data</td></tr>
<tr><td>Z-SCORE</td><td>mirna_median_Zscores</td><td>MicroRNA data</td></tr>
<tr><td>Z-SCORE</td><td>mrna_merged_median_Zscores</td><td>?</td></tr>
<tr><td>CONTINUOUS</td><td>mrna</td><td>mRNA data</td></tr>
<tr><td>DISCRETE</td><td>mrna_outliers</td><td>mRNA data of outliers</td></tr>
<tr><td>Z-SCORE</td><td>mrna_zbynorm</td><td>?</td></tr>
<tr><td>CONTINUOUS</td><td>rna_seq_mrna_capture</td><td>data from Roche mRNA Capture Kit</td></tr>
<tr><td>Z-SCORE</td><td>rna_seq_mrna_capture_Zscores</td><td>data from Roche mRNA Capture Kit</td></tr>
</table>


### Example

An example metadata, e.g. meta_expression.txt file would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: MRNA_EXPRESSION
datatype: CONTINUOUS
stable_id: rna_seq_mrna
show_profile_in_analysis_tab: false
profile_name: mRNA expression 
profile_description: Expression levels 
data_filename: data_expression.txt
```

### Data file

For each gene (row) in the data file, the following columns are required in the order specified:

One or both of:
- ***Hugo_Symbol***: A [HUGO](https://www.genenames.org/) gene symbol.
- ***Entrez_Gene_Id***: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier.

And:
- An additional column for each sample in the dataset using the sample id as the column header.

For each gene-sample combination, a value is specified:
- A real number for each sample id (column) in the dataset, representing the expression value for the gene in the respective sample.
- or `NA` for when the expression value for the gene in the respective sample could not (or was not) be measured (or detected).

### z-score instructions

For mRNA expression data, we typically expect the relative expression of an individual gene and tumor to the gene's expression distribution in a reference population. That reference population is either all tumors that are diploid for the gene in question, or, when available, normal adjacent tissue. The returned value indicates the number of standard deviations away from the mean of expression in the reference population (Z-score). This measure is useful to determine whether a gene is up- or down-regulated relative to the normal samples or all other tumor samples.  **Note, the importer tool can create normalized (z-score) expression data on your behalf.  Please visit the [Z-Score normalization script](Z-Score-normalization-script.md) wiki page for more information.
A corresponding z-score metadata file would be something like**:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: MRNA_EXPRESSION
datatype: Z-SCORE
stable_id: rna_seq_mrna_median_Zscores
show_profile_in_analysis_tab: true
profile_name: mRNA expression z-scores
profile_description: Expression levels z-scores
data_filename: data_expression_zscores.txt
```

### Examples of data files: 

An example data file which includes the required column header and *leaves out Hugo_Symbol* (recommended) would look like:

```
Entrez_Gene_Id<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
116983<TAB>-0.005<TAB>-0.550<TAB>...
375790<TAB>0.142<TAB>0.091<TAB>...
...
...
```

An example data file which includes both Hugo_Symbo and Entrez_Gene_Id would look like (supported, but not recommended as it increases the chances of errors regarding ambiguous gene symbols):
```
Hugo_Symbol<TAB>Entrez_Gene_Id<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
ACAP3<TAB>116983<TAB>-0.005<TAB>-0.550<TAB>...
AGRN<TAB>375790<TAB>0.142<TAB>0.091<TAB>...
...
...
```

An example data file with only Hugo_Symbol column (supported, but not recommended as it increases the chances of errors regarding ambiguous gene symbols):
```
Hugo_Symbol<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
ACAP3<TAB>-0.005<TAB>-0.550<TAB>...
AGRN<TAB>0.142<TAB>0.091<TAB>...
...
...
```


## Mutation Data
When loading mutation data, the `_sequenced` case list is required. See the [case list section](#case-lists).

### Meta file
The mutation metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: MUTATION_EXTENDED
3. **datatype**: MAF
4. **stable_id**: mutations
5. **show_profile_in_analysis_tab**: true
6. **profile_name**: A name for the mutation data, e.g., "Mutations".
7. **profile_description**: A description of the mutation data, e.g., "Mutation data from whole exome sequencing.".
8. **data_filename**: your data file
9. **gene_panel (optional)**: gene panel stable id. See [Gene panels for mutation data](#gene-panels-for-mutation-data).
10. **swissprot_identifier (optional)**: `accession` or `name`, indicating the type of identifier in the `SWISSPROT` column
11. **variant_classification_filter (optional)**: List of `Variant_Classifications` values to be filtered out.
12. **namespaces (optional)**: Comma-delimited list of `namespaces` to import. 

### Gene panels for mutation data
Using the `gene_panel` property it is possible to annotate **all samples in the MAF file** as being profiled on the **same** specified gene panel. 

Please use the [Gene Panel Matrix file](#gene-panel-matrix-file) when:
- Data contains samples that are profiled but no mutations are called. Also please add these to the `_sequenced` case list. 
- Multiple gene panels are used to profile the samples in the MAF file.

### Variant classification filter
The `variant_classification_filter` field can be used to filter out specific mutations. This field should contain a comma separated list of `Variant_Classification` values. By default, cBioPortal filters out `Silent, Intron, IGR, 3'UTR, 5'UTR, 3'Flank and 5'Flank`, except for the promoter mutations of the TERT gene. For no filtering, include this field in the metadata file, but leave it empty. For cBioPortal default filtering, do not include this field in the metadata file.
Allowed values to filter out (mainly from [Mutation Annotation Format page](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/)): `Frame_Shift_Del, Frame_Shift_Ins, In_Frame_Del, In_Frame_Ins, Missense_Mutation, Nonsense_Mutation, Silent, Splice_Site, Translation_Start_Site, Nonstop_Mutation, 3'UTR, 3'Flank, 5'UTR, 5'Flank, IGR, Intron, RNA, Targeted_Region, De_novo_Start_InFrame, De_novo_Start_OutOfFrame, Splice_Region and Unknown`

### Tumor seq allele ambiguity
Bugs may exist in MAF data that make it ambiguous as to whether `Tumor_Seq_Allele1` or `Tumor_Seq_Allele2` should be seen as the variant allele to be used when a new mutation record is created and imported in cBioPortal. In such cases, preference is given to the tumor seq allele value that matches a valid nucleotide pattern `^[ATGC]*$` versus a null or empty value, or "-".
For example, given `Reference_Allele` = "G", `Tumor_Seq_Allele` = "-", and `Tumor_Seq_Allele2` = "A", preference will be given to `Tumor_Seq_Allele2`. Using this same example with `Tumor_Seq_Allele1` = "T", preference will be given to `Tumor_Seq_Allele1` if it does not match `Reference_Allele`, which in this case it does not.

When curating MAF data, it is best practice to leave `Tumor_Seq_Allele1` empty if this information is not provided in your data source to avoid this ambiguity.

### Namespaces
The `namespaces` field can be used to specify additional MAF columns for import. This field should contain a comma separated list of namespaces. Namespaces can be identified as prefixes to an arbitrary set of additional MAF columns (separated with a period e.g `ASCN.total_copy_number`, `ASCN.minor_copy_number`). All columns with a prefix matching a namespace specified in the metafile will be imported; columns with an unspecified namespace will be ignored. If no additional columns beyond the required set need to be imported, the field should be left blank. 

### Example
An example metadata file would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: MUTATION_EXTENDED
datatype: MAF
stable_id: mutations
show_profile_in_analysis_tab: true
profile_description: Mutation data from whole exome sequencing.
profile_name: Mutations
data_filename: data_mutations.txt
namespaces: ASCN
```

### Data file
The cBioPortal mutation data file extends the [Mutation Annotation Format](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/) (MAF) created as part of The Cancer Genome Atlas (TCGA) project, by adding *extra annotations* to each mutation record. This section describes:
1. How to create the cBioPortal mutation data file with a minimal MAF file using the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline).
2. The description of the cBioPortal mutation data file. You can also get the cBioPortal mutation data file from vcf using: [vcf2maf](https://github.com/mskcc/vcf2maf).

### Create the cBioPortal mutation data file with Genome Nexus with a minimal MAF file
#### Minimal MAF file format
A minimal mutation annotations file can contain just the five genomic change columns plus one sample identifier column. From this minimal MAF, it is possible to create the cBioPortal mutation data file by running it through the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline).
1. **Chromosome (Required)**:  A chromosome number, e.g., "7".
2. **Start_Position (Required)**: Start position of event.
3. **End_Position (Required)**: End position of event.
4. **Reference_Allele (Required)**: The plus strand reference allele at this position.
5. **Tumor_Seq_Allele2 (Required)**: Primary data genotype.
6. **Tumor_Sample_Barcode (Required)**: This is the sample ID. Either a TCGA barcode (patient identifier will be extracted), or for non-TCGA data, a literal SAMPLE_ID as listed in the clinical data file.

In addition to the above columns, it is recommended to have the read counts to calculate variant allele frequencies:

7. **t_alt_count (Optional, but recommended)**: Variant allele count (tumor). 
8. **t_ref_count (Optional, but recommended)**: Reference allele count (tumor).

The following extra annotation columns are important for making sure mutation specific UI functionality works well in the portal:

9. **Protein_position (Optional)**: (annotation column) Required to initialize the 3D viewer in [mutations view](https://www.cbioportal.org/index.do?cancer_study_list=brca_tcga_pub&cancer_study_id=brca_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_pub_gistic&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_pub_mrna_median_Zscores&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=brca_tcga_pub_complete&case_ids=&patient_case_select=sample&gene_set_choice=prostate-cancer%3A-ar-signaling-%2810-genes%29&gene_list=TP53&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit#mutation_details)
10. **SWISSPROT (Optional)**: (annotation column) UniProtKB/SWISS-PROT name (formerly called ID) or accession code depending on the value of the `swissprot_identifier` metadatum, e.g. O11H1_HUMAN or Q8NG94. Is not required, but not having it may result in inconsistent PDB structure matching in [mutations view](https://www.cbioportal.org/index.do?cancer_study_list=brca_tcga_pub&cancer_study_id=brca_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_pub_gistic&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_pub_mrna_median_Zscores&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=brca_tcga_pub_complete&case_ids=&patient_case_select=sample&gene_set_choice=prostate-cancer%3A-ar-signaling-%2810-genes%29&gene_list=TP53&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit#mutation_details).

### Creating the cBioPortal mutation data file
Once you have a minimal MAF you can run it through the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline).
This tool runs annotates variants against the [Genome Nexus Server](https://genomenexus.org), which in turn leverages Ensembl Variant Effect Predictor (VEP) and selects a single effect per variant. Protein identifiers will be mapped to UniProt canonical isoforms (see also [this mapping file](https://github.com/genome-nexus/genome-nexus-importer/blob/master/data/grch37_ensembl92/export/ensembl_biomart_canonical_transcripts_per_hgnc.txt)).

### cBioPortal mutation data file format
The cBioPortal mutation data file format recognized by the portal has:
* 32 columns from the [TCGA MAF format](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/).
* 1 column with the amino acid change.
* 4 columns with information on reference and variant allele counts in tumor and normal samples. 

1. **Hugo_Symbol (Required)**: A [HUGO](https://www.genenames.org/) gene symbol.
2. **Entrez_Gene_Id (Optional, but recommended)**: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier.
3. **Center (Optional)**: The sequencing center.
4. **NCBI_Build (Required)<sup>1</sup>**: The Genome Reference Consortium Build is used by a variant calling software. It must be "GRCh37" or "GRCh38" for a human, and "GRCm38" for a mouse.
5. **Chromosome (Required)**: A chromosome number, e.g., "7".
6. **Start_Position (Optional, but recommended for additional features such as [Cancer Hotspots annotations](https://www.cancerhotspots.org/))**: Start position of event.
7. **End_Position (Optional, but recommended for additional features such as [Cancer Hotspots annotations](https://www.cancerhotspots.org/))**: End position of event.
8. **Strand (Optional)**: We assume that the mutation is reported for the + strand.
9. **Variant_Classification (Required)**: Translational effect of variant allele, e.g. Missense_Mutation, Silent, etc.
10. **Variant_Type <sup>1</sup>(Optional)**: Variant Type, e.g. SNP, DNP, etc.
11. **Reference_Allele (Required)**: The plus strand reference allele at this position.
12. **Tumor_Seq_Allele1 (Optional)**: Primary data genotype.
13. **Tumor_Seq_Allele2 (Required)**: Primary data genotype.
14. **dbSNP_RS<sup>1</sup> (Optional)**: Latest dbSNP rs ID.
15. **dbSNP_Val_Status<sup>1</sup> (Optional)**: dbSNP validation status.
16. **Tumor_Sample_Barcode (Required)**: This is the sample ID. Either a TCGA barcode (patient identifier will be extracted), or for non-TCGA data, a literal SAMPLE_ID as listed in the clinical data file.
17. **Matched_Norm_Sample_Barcode<sup>1</sup> (Optional)**: The sample ID for the matched normal sample.
18. **Match_Norm_Seq_Allele1 (Optional)**: Primary data.
19. **Match_Norm_Seq_Allele2 (Optional)**: Primary data.
20. **Tumor_Validation_Allele1 (Optional)**: Secondary data from orthogonal technology.
21. **Tumor_Validation_Allele2 (Optional)**: Secondary data from orthogonal technology.
22. **Match_Norm_Validation_Allele1<sup>1</sup> (Optional)**: Secondary data from orthogonal technology.
23. **Match_Norm_Validation_Allele2<sup>1</sup> (Optional)**: Secondary data from orthogonal technology.
24. **Verification_Status<sup>1</sup> (Optional)**: Second pass results from independent attempt using same methods as primary data source. "Verified", "Unknown" or "NA".
25. **Validation_Status (Optional)**: Second pass results from orthogonal technology. "Valid", "Invalid", "Untested", "Inconclusive", "Redacted", "Unknown" or "NA".
26. **Mutation_Status (Optional)**: "Somatic" or "Germline" are supported by the UI in Mutations tab. "None", "LOH" and "Wildtype" will not be loaded. Other values will be displayed as text.
27. **Sequencing_Phase<sup>1</sup> (Optional)**: Indicates current sequencing phase.
28. **Sequence_Source<sup>1</sup> (Optional)**: Molecular assay type used to produce the analytes used for sequencing.
29. **Validation_Method<sup>1</sup> (Optional)**: The assay platforms used for the validation call.
30. **Score<sup>1</sup> (Optional)**: Not used.
31. **BAM_File<sup>1</sup> (Optional)**: Not used.
32. **Sequencer<sup>1</sup> (Optional)**: Instrument used to produce primary data.
33. **HGVSp_Short (Required)**: Amino Acid Change, e.g. p.V600E.
34. **t_alt_count (Optional)**: Variant allele count (tumor). 
35. **t_ref_count (Optional)**: Reference allele count (tumor).
36. **n_alt_count (Optional)**: Variant allele count (normal).
37. **n_ref_count (Optional)**: Reference allele count (normal).

<sup>**1**</sup> These columns are currently not shown in the Mutation tab and Patient view.

### Custom driver annotations
It is possible to manually add columns for defining custom driver annotations. These annotations can be used to complement or replace default driver annotation resources OncoKB and HotSpots.

38. **cbp_driver (Optional)**: "Putative_Driver", "Putative_Passenger", "Unknown", "NA" or "" (empty value). This field must be present if the cbp_driver_annotation is also present in the MAF file. 
39. **cbp_driver_annotation (Optional)**: Description field for the cbp_driver value (limited to 80 characters). This field must be present if the cbp_driver is also present in the MAF file. This field is free text. Example values for this field are: "Pathogenic" or "VUS".
40. **cbp_driver_tiers (Optional)**: Free label/category that marks the mutation as a putative driver such as "Driver", "Highly actionable", "Potential drug target". . This field must be present if the cbp_driver_tiers_annotation is also present in the MAF file. In the OncoPrint view's Mutation Color dropdown menu, these tiers are ordered alphabetically. This field is free text and limited to 20 characters. For mutations without a custom annotation, leave the field blank or type "NA".
41. **cbp_driver_tiers_annotation (Optional)**: Description field for the cbp_driver_tiers value (limited to 80 characters). This field must be present if the cbp_driver_tiers is also present in the MAF file. This field can not be present when the cbp_driver_tiers field is not present. 

The `cbp_driver` column flags the mutation as either driver or passenger. In cBioPortal, passenger mutations are also known as variants of unknown significance (VUS). The `cbp_driver_tiers` column assigns an annotation tier to the mutation, such as "Driver", "Highly actionable" or "Potential drug target". When a tier is selected, mutations with that annotation are highlighted as driver. Both types of custom annotations contain a second column with the suffix `_annotation`, to add a description. This is displayed in the tooltip that appears when hovering over the sample's custom annotation icon in the OncoPrint view.

You can learn more about configuring these annotations in the [application.properties documentation](/deployment/customization/application.properties-Reference.md#custom-annotation-of-driver-and-passenger-mutations). When properly configured, the customized annotations appear in the "Mutation Color" menu of the OncoPrint view: \
![schreenshot mutation color menu](/images/screenshot-mutation-color-menu.png) 

### Adding your own mutation annotation columns
Additional mutation annotation columns can be added to the cBioPortal mutation data file. In this way, the portal will
parse and store your own MAF fields in the database. For example, mutation data that you find on cBioPortal.org comes
from MAF files that have been further enriched with information
from [mutationassessor.org](https://mutationassessor.org/), which leads to a "Mutation Assessor" column in
the [mutation table](https://www.cbioportal.org/index.do?cancer_study_list=acc_tcga&cancer_study_id=acc_tcga&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=acc_tcga_mutations&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=acc_tcga_sequenced&case_ids=&patient_case_select=sample&gene_set_choice=user-defined-list&gene_list=ZFPM1&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit).

See [Custom namespace columns](#custom-namespace-columns) for more information on adding custom columns to data files.

### Allele specific copy number (ASCN) annotations
Allele specific copy number (ASCN) annotation is also supported and may be added using namespaces, described [here](#adding-mutation-annotation-columns-through-namespaces). If ASCN data is present in the cBioPortal mutation data file, the deployed cBioPortal instance will display additional columns in the mutation table showing ASCN data.

**The ASCN columns below are optional by default. If `ascn` is a defined namespace in `meta_mutations_extended.txt`, then these columns are ALL required.**

42. **ASCN.ASCN_METHOD (Optional)**: Method used to obtain ASCN data e.g "FACETS".
43. **ASCN.CCF_EXPECTED_COPIES (Optional)**: Cancer-cell fraction if mutation exists on major allele. 
44. **ASCN.CCF_EXPECTED_COPIES_UPPER (Optional)**: Upper error for CCF estimate.
45. **ASCN.EXPECTED_ALT_COPIES (Optional)**: Estimated number of copies harboring mutant allele.
46. **ASCN.CLONAL (Optional)**: "Clonal", "Subclonal", or "Indeterminate". 
47. **ASCN.TOTAL_COPY_NUMBER (Optional)**: Total copy number of the gene.
48. **ASCN.MINOR_COPY_NUMBER (Optional)**: Copy number of the minor allele.
49. **ASCN.ASCN_INTEGER_COPY_NUMER (Optional)**: Absolute integer copy-number estimate.

### Example cBioPortal mutation data file
An example cBioPortal mutation data file can be found in the cBioPortal test study [study_es_0](https://github.com/cBioPortal/cbioportal/blob/master/core/src/test/scripts/test_data/study_es_0/data_mutations_extended.maf).

### Filtered mutations
A special case for **Entrez_Gene_Id=0** and **Hugo_Symbol=Unknown**: when this combination is given, the record is parsed in the same way as **Variant_Classification=IGR** and therefore filtered out.


## Methylation Data
The Portal expects a single value for each gene in each sample, usually a beta-value from the Infinium methylation array platform.

### Meta file

The methylation metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: METHYLATION
3. **datatype**: CONTINUOUS
4. **stable_id**: "methylation_hm27" or "methylation_hm450" (depending on platform).
5. **show_profile_in_analysis_tab**: false
6. **profile_name**: A name for the methylation data, e.g., "Methlytation (HM27)".
7. **profile_description**: A description of the methlytation data, e.g., "Methylation beta-values (HM27 platform). For genes with multiple methylation probes, the probe least correlated with expression is selected.".
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id


### Example

An example metadata file would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: METHYLATION
datatype: CONTINUOUS
stable_id: methylation_hm27
show_profile_in_analysis_tab: false
profile_name: Methylation (HM27)
profile_description: Methylation beta-values (HM27 platform). For genes with multiple methylation probes, the probe least correlated with expression is selected.
data_filename: data_methylation_hm27.txt
```

### Data file

The methylation data file follows the same format as expression data files. See [Expression Data](#expression-data) for a description of the expression data file format. The Portal expects a single value for each gene in each sample, usually a beta-value from the Infinium methylation array platform.


## Protein level Data

Protein expression measured by reverse-phase protein array or mass spectrometry. Antibody-sample pairs, with a real number representing the protein level for that sample.

### Meta file

The protein level metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: PROTEIN_LEVEL
3. **datatype**: LOG2-VALUE or Z-SCORE
4. **stable_id**: rppa, rppa_Zscores, protein_quantification or protein_quantification_zscores 
5. **show_profile_in_analysis_tab**: false (**true** for Z-SCORE datatype)
6. **profile_name**: A name for the RPPA data, e.g., "RPPA data".
7. **profile_description**: A description of the RPPA data, e.g., "RPPA levels.".
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id

An example metadata file would be:
```
cancer_study_identifier: brca_tcga
genetic_alteration_type: PROTEIN_LEVEL
datatype: LOG2-VALUE
stable_id: rppa
show_profile_in_analysis_tab: false
profile_description: Protein expression measured by reverse-phase protein array
profile_name: Protein expression (RPPA)
data_filename: data_rppa.txt
```

**NB:** You also need a Z-SCORE file if you want protein levels to be available in query UI and in [Oncoprint visualization](https://www.cbioportal.org/index.do?Ôú´cancer_study_list=brca_tcga_pub&cancer_study_id=brca_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_pub_gistic&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_pub_mrna_median_Zscores&Z_SCORE_THRESHOLD=2.0&genetic_profile_ids_PROFILE_PROTEIN_EXPRESSION=brca_tcga_pub_rppa_Zscores&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=brca_tcga_pub_complete&case_ids=&patient_case_select=sample&gene_set_choice=prostate-cancer%3A-ar-signaling-%2810-genes%29&gene_list=TP53+SOX9+RAN+TNK2+EP300+PXN+NCOA2+AR+NRIP1+NCOR1+NCOR2&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit). E.g.:

```
cancer_study_identifier: brca_tcga
genetic_alteration_type: PROTEIN_LEVEL
datatype: Z-SCORE
data_filename: data_rppa_zscores.txt
stable_id: rppa_Zscores
show_profile_in_analysis_tab: true
profile_description: Protein expression Z-scores (RPPA)
profile_name: Protein expression Z-scores (RPPA)
```

### Data file

A protein level data file is a two dimensional matrix with a RPPA antibody per row and a sample per column. For each antibody-sample pair, a real number represents the protein level for that sample.  The antibody information can contain one or more HUGO gene symbols and/or entrez gene identifiers, separated by a space, and an antibody ID pair separated by the "|" symbol.

### Example 

An example data file which includes the required column header would look like:

```
Composite.Element.REF<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
BRAF|B-Raf-M-NA<TAB>1.09506676325<TAB>0.5843256495...
MAPK1 MAPK3|MAPK_PT202_Y204<TAB>1.70444582025<TAB>1.0982864685...
AKT1 AKT2 10000|AKT<TAB>0.17071492725<TAB>0.264067254391
...
```


## Structural Variant Data

The cBioPortal can load all kinds of structural variant data but at the moment only a subset of them, fusions, are displayed.

### Meta file
The structural variant metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: STRUCTURAL_VARIANT
3. **datatype**: SV
4. **stable_id**: structural_variants
5. **show_profile_in_analysis_tab**: true.
6. **profile_name**: A name for the fusion data, e.g., "Structural Variants".
7. **profile_description**: A description of the structural variant data.
8. **data_filename**: your datafile (e.g. data_sv.txt)
9. **gene_panel (Optional)**:  gene panel stable id

An example metadata file would be:

```
cancer_study_identifier: msk_impact_2017
genetic_alteration_type: STRUCTURAL_VARIANT
datatype: SV
stable_id: structural_variants
show_profile_in_analysis_tab: true
profile_name: Structural variants
profile_description: Structural Variant Data for mskimpact2017
data_filename: data_sv.txt
```

### Data file

A structural variant data file is a tab-delimited file with one structural variant per row.  For each structural variant (row) in the data file, the following tab-delimited values are required:

| Field                        | Example Values         | Allowed  Values       | Comments |
| -- | -- | -- | -- |
| Sample_Id (Required)         | Sample_1      |  | As defined in the clinical sample file. |
| SV_Status (Required)         | SOMATIC | SOMATIC or GERMLINE |  |
| Site1_Hugo_Symbol            | TMPRSS2 |   | [HUGO](http://www.genenames.org/) gene symbol of gene 1. One might call this the left site(3‚Äô) as well. (strongly recommended field) |
| Site1_Ensembl_Transcript_Id  | ENST00000398585  |  | Ensembl transcript ID of gene 1. |
| Site1_Entrez_Gene_Id         | 7113 |  | [Entrez Gene](http://www.ncbi.nlm.nih.gov/gene) identifier of gene 1. (strongly recommended field) |
| Site1_Region_Number          | 2 |  | Number of Site 1 region e.g. exon 2.(strongly recommended field) |
| Site1_Region                 | Exon | 5_Prime_UTR,3_Prime_UTR,Promoter,Exon,Intron | We advise using one of these {5_Prime_UTR,3_Prime_UTR,Promoter,Exon,Intron},but it is a free text. (strongly recommended field) |
| Site1_Chromosome             | 21 |  | Chromosome of Gene 1.(strongly recommended field) |
| Site1_Contig                 | q22.3 |  | The contig of Site 1.(strongly recommended field) |
| Site1_Position               | 42874110 |  | Genomic position of breakpoint of Gene 1.(strongly recommended field) |
| Site1_Description            | Intron of TMPRSS2(-): 511bp before exon 3  |  | Description of this event at site 2. This could be the location of the 2nd breakpoint in case of a fusion event. |
| Site2_Hugo_Symbol            | ERG |  | [HUGO](http://www.genenames.org/) gene symbol of gene 2.One might call this the right site(5‚Äô) as well. |
| Site2_Ensembl_Transcript_Id  | ENST00000288319 |   | Ensembl transcript ID of gene 2.|
| Site2_Entrez_Gene_Id         | 2078 |  | [Entrez Gene](http://www.ncbi.nlm.nih.gov/gene) identifier of gene 2. |
| Site2_Region_Number          | 4 |  | Number of Site 2 region e.g. exon 4.|
| Site2_Region                 | EXON | 5_PRIME_UTR,3_PRIME_UTR,PROMOTER,EXON,INTRON | We advise using one of these {5_PRIME_UTR,3_PRIME_UTR,PROMOTER,EXON,INTRON},but it is a free text. |
| Site2_Chromosome             | 21 |  | Chromosome of Gene 2.|
| Site2_Contig                 | q22.2 |  | The contig of Site 2.|
| Site2_Position               | 39842043 |  | Genomic position of breakpoint of Gene 2. |
| Site2_Description            | Intron of ERG(-): 6Kb after exon 1 |  | Description of this event at site 1. This could be the location of 1st breakpoint in case of a fusion event |
| Site2_Effect_On_Frame        | Out-of-frame | In_frame, Out-of-frame,Frameshift  | The effect on frame reading in gene 2. Frame_Shift or InFrame,but it is a free text. |
| NCBI_Build                   | GRCh37 | GRCh37,GRCh38 | The NCBI assembly. Only one assembly per study can be used, see[ study metadata](https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#meta-file). |
| Class                        | Deletion | Deletion,Duplication,Insertion,Inversion,Translocation. | We advise using one of these terms [DELETION, DUPLICATION, INSERTION, INVERSION or TRANSLOCATION], but it is free text. |
| Tumor_Split_Read_Count       | 4 |  | The number of split reads of the tumor tissue that support the call.[Tumor Split Read Count is the same as ‚ÄúJunction Reads‚Äù.] |
| Tumor_Paired_End_Read_Count  | 55 |  | The number of paired-end reads of the tumor tissue that support the call. [Tumor Paired End Read Count is the same as ‚ÄúSpanning Fragments‚Äù.] |
| Event_Info                   | Protein fusion: out of frame (TMPRSS2-ERG) | Antisense fusion, Deletion within transcript: mid-exon, Duplication of 1 exon: in frame | Description of the event. For a fusion event, fill in Fusion. It is a free text. |
| Connection_Type              | 3to5 | 3to5 or 5to3 or 5to5 or 3to3 | Which direction the connection is made (3' to 5', 5' to 3', etc) |
| Breakpoint_Type              | PRECISE  | PRECISE/IMPRECISE  | PRECISE or IMPRECISE which explain the resolution. Fill in PRECISE if the breakpoint resolution is known down to the exact base pair. |
| Annotation                   | TMPRSS2 (NM_001135099) - ERG (NM_001243428) fusion (TMPRSS2 exons 1-2 fused with ERG exons 4-11):(c.126+879:TMRPSS2_c.40-63033:ERGdel) | Free Text                                                    | Free text description of the gene or transcript rearrangement. |
| DNA_Support                  | Yes  | Yes or No  | Fusion detected from DNA sequence data, "Yes" or "No".|
| RNA_Support                  |   | Yes or No | Fusion detected from DNA sequence data, "Yes" or "No".|
| SV_Length                    | 3032067 |  | Length of the structural variant in number of bases.|
| Normal_Read_Count            | 93891 |  | The total number of reads of the normal tissue.|
| Tumor_Read_Count             | 45556 |  | The total number of reads of the tumor tissue. |
| Normal_Variant_Count         | 0 |  | The number of reads of the normal tissue that have the variant/allele. |
| Tumor_Variant_Count          | 39 |  | The number of reads of the tumor tissue that have the variant/allele. |
| Normal_Paired_End_Read_Count | 41 |  | The number of paired-end reads of the normal tissue that support the call.|
| Normal_Split_Read_Count      | 63 |  | The number of split reads of the normal tissue that support the call. |
| Comments                     |  |  | Any comments/free text.|

For an example see [datahub](https://github.com/cBioPortal/datahub/blob/master/public/msk_impact_2017/data_sv.txt). For an example see [datahub](https://github.com/cBioPortal/datahub/blob/master/public/msk_impact_2017/data_sv.txt). At a minimum `Sample_Id`, either `Site1_Hugo_Symbol`/ `Site1_Entrez_Gene_Id` or  `Site2_Hugo_Symbol`/ `Site2_Entrez_Gene_Id` and `SV_Status` are required. For the stuctural variant tab visualization (still in development) one needs to provide those field as well as `Site1_Ensembl_Transcript_Id`, `Site2_Ensembl_Transcript_Id`, `Site1_Region` and `Site2_Region`. Some of the other columns are shown at several other pages on the website. The `Class`, `Annotation` and `Event_Info` columns are shown prominently on several locations.
**Note**: We strongly recommend all the data providers to submit genomic locations  in addition to required fields for future visualization features. 

### Adding your own structural variant columns
Additional mutation annotation columns can be added to the structural variant data file. In this way, the portal will
parse and store your own structural variant fields in the database.

See [Custom namespace columns](#custom-namespace-columns) for more information on adding custom columns to data files.

## Fusion Data
**‚ö†Ô∏è DEPRECATED Use the: [SV format](#structural-variant-data) instead**

### Meta file
The fusion metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: FUSION
3. **datatype**: FUSION
4. **stable_id**: fusion
5. **show_profile_in_analysis_tab**: true.
6. **profile_name**: A name for the fusion data, e.g., "Fusions".
7. **profile_description**: A description of the fusion data.
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id

### Example
An example metadata file would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: FUSION
datatype: FUSION
stable_id: fusion
show_profile_in_analysis_tab: true
profile_name: Fusions
profile_description: Fusion data.
data_filename: data_fusions.txt
```

### Data file
A fusion data file is a two dimensional matrix with one gene per row.  For each gene (row) in the data file, the following tab-delimited values are required in the order specified:

1. **Hugo_Symbol**: A [HUGO](https://www.genenames.org/) gene symbol.
2. **Entrez_Gene_Id**: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier.
3. **Center**: The sequencing center.
4. **Tumor_Sample_Barcode**: This is the sample ID.
5. **Fusion**: A description of the fusion, e.g., "TMPRSS2-ERG fusion".
6. **DNA_support**: Fusion detected from DNA sequence data, "yes" or "no".
7. **RNA_support**: Fusion detected from RNA sequence data, "yes" or "no".
8. **Method**: Fusion detected algorithm/tool.
9. **Frame**: "in-frame" or "frameshift".
10. **Fusion_Status (OPTIONAL)**: An assessment of the mutation type (i.e., "SOMATIC", "GERMLINE", "UNKNOWN", or empty)

**Note:** If a fusion event includes a gene, e.g., Hugo_Symbol or Entrez_Gene_Id, that is not profiled, the event will be filter out during import into the database.

An example data file which includes the required column header would look like:
```
Hugo_Symbol<TAB>Entrez_Gene_Id<TAB>Center<TAB>Tumor_Sample_Barcode<TAB>Fusion<TAB>DNA_support<TAB>RNA_support<TAB>Method<TAB>Frame>
ALK<TAB>238<TAB>center.edu<TAB>SAMPLE_ID_1<TAB>Fusion<TAB>unknown<TAB>yes<TAB>unknown<TAB>in-frame
ALK<TAB>238<TAB>center.edu<TAB>SAMPLE_ID_2<TAB>Fusion<TAB>unknown<TAB>yes<TAB>unknown<TAB>in-frame
RET<TAB>5979<TAB>center.edu<TAB>SAMPLE_ID_3<TAB>Fusion<TAB>unknown<TAB>yes<TAB>unknown<TAB>in-frame
...
...
```

## Case Lists

Case lists are used to define sample lists that can be selected on the query page. Some case lists have specific functionality, but it's also possible to add custom case lists. The case list files should be placed in a sub-directory called `case_lists` which exists alongside all the other cancer study data. 
The case list file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **stable_id**: it must contain the cancer_study_identifier followed by an underscore. Typically, after this a relevant suffix, e.g., `_custom`, is added. There are some naming rules to follow if you want the case list to be selected automatically in the query UI base on the selected sample profiles. See subsection below.
3. **case_list_name**: A name for the patient list, e.g., "All Tumors".
4. **case_list_description**: A description of the patient list, e.g., "All tumor samples (825 samples).".
5. **case_list_ids**: A tab-delimited list of sample ids from the dataset.
6. **case_list_category**: Optional *alternative* way of linking your case list to a specific molecular profile. E.g. setting this to `all_cases_with_cna_data` will signal to the portal that this is the list of samples to be associated with CNA data in *some* of the analysis. 

### Example
An example case list file would be:
```
cancer_study_identifier: brca_tcga_pub
stable_id: brca_tcga_pub_custom
case_list_name: Custom subset of samples
case_list_description: Custom subset of samples (825 samples)
case_list_ids: SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>SAMPLE_ID_3<TAB>...
```

### Case list stable id suffixes
In order for sample counts to propagate to the data sets widget on the home page and the table on the [Data Sets](https://www.cbioportal.org/data_sets.jsp) page, the following case list suffixes need to be used in the stable_id property (e.g. `brca_tcga_pub_sequenced`). This is also needed for correct statistics in the Study view page when calculating the frequency of CNA and of mutations per gene in the respective summary tables.

* **Sequenced**: `_sequenced` . When only a mutation profile is selected on the query page, this is the default case list. Also used in the Study Summary to calculate the proportion of samples with mutations.
* **CNA**: `_cna`. When only a CNA profile is selected on the query page, this is the default case list. Also used in the Study Summary to calculate the proportion of samples with CNA.
* **Sequenced and CNA**: `_cnaseq`. When a mutation and CNA genetic profile are selected on the query page, this is the default case list.
* **mRNA (microarray)**: `_mrna`. When only a mRNA (microarray) profile is selected on the query page, this is the default case list.
* **mRNA (RNA-Seq)**: `_rna_seq_mrna`. When only a mRNA (RNA-Seq) profile is selected on the query page, this is the default case list.
* **mRNA (RNA-SeqV2)**: `_rna_seq_v2_mrna`. When only a mRNA (RNA-SeqV2) profile is selected on the query page, this is the default case list.
* **mRNA normal**: `_normal_mrna`. Used for the datasets page to calculate the number of normal samples.
* **mRNA normal**: `_microrna`. Used for the datasets page to calculate the number of microRNA samples.
* **Methylation (HM27)**: `_methylation_hm27`.
* **RPPA**: `_rppa`. When only a RPPA profile is selected on the query page, this is the default case list.
* **Sequenced, CNA and mRNA**: `_3way_complete` When a mutation, CNA and mRNA profile are selected on the query page, this is the default case list.
* **SV**: `_sv` . When a structural variant profile is selected on the query page, this is the default case list. Also used in the Study Summary to calculate the proportion of samples with fusions.
* **All**: `_all`. If you are not using *add_global_case_list* attribute in [Study metadata](#cancer-study), you need to add this case list.

### Required case lists
Some case lists are required:

- `_all`. This can be generated by the importer if you set the attribute `add_global_case_list` to `true` in the [Study metadata](#cancer-study).
- `_sequenced`. This case list is required when loading mutation data.
- `_cna`. This case list is required when loading discrete cna data.

### Case list categories
These are the valid case lists categories for `case_list_category: ` in the meta file.
- `all_cases_in_study`
- `all_cases_with_mutation_data`
- `all_cases_with_cna_data`
- `all_cases_with_log2_cna_data`
- `all_cases_with_methylation_data`
- `all_cases_with_mrna_array_data`
- `all_cases_with_mrna_rnaseq_data`
- `all_cases_with_rppa_data`
- `all_cases_with_microrna_data`
- `all_cases_with_mutation_and_cna_data`
- `all_cases_with_mutation_and_cna_and_mrna_data`
- `all_cases_with_gsva_data`
- `all_cases_with_sv_data`
- `other`

## Timeline Data
The timeline data is a representation of the various events that occur during the course of treatment for a patient from initial diagnosis. In cBioPortal timeline data is represented as one or more tracks in the patient view. Each main track is based on an event type, such as "Specimen", "Imaging", "Lab_test", etc.

**Attention**: some clinical attributes affect the timeline visualization. Please check the [Clinical Data](#clinical-data) section for more information.  

This type data is not yet being validated. It can, however, be uploaded.

### Meta file
Each event type requires its own meta file. A timeline meta file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: CLINICAL
3. **datatype**: TIMELINE
4. **data_filename**: your datafile

An example metadata file would be:
```
cancer_study_identifier: brca_tcga
genetic_alteration_type: CLINICAL
datatype: TIMELINE
data_filename: data_timeline_imaging.txt
```


### Data file
Each event type requires its own data file, which contains all the events that each patient undergoes. The data format used for timeline data is extremely flexible. There are four required columns:

1. **PATIENT_ID**: the patient ID from the dataset
2. **START_DATE**: the start point of any event, calculated in **_days_* from the date of diagnosis (which will act as point zero on the timeline scale)
3. **STOP_DATE**: The end date of the event is calculated in days from the date of diagnosis (which will act as point zero on the timeline scale). If the event occurs over time (e.g. a Treatment, ...) the STOP_DATE column should have values. If the event occurs at a time point (e.g. a Lab_test, Imaging, ...) the STOP_DATE is still mandatory, but the values should be blanks.
4. **EVENT_TYPE**: the category of the event. You are free to define any type of event here. For several event types cBioPortal has column naming suggestions and for several events there are column names which have special effects. See [event types](#event-types) for more information. 

And optional columns with special effects:

1. **SPECIMEN_REFERENCE_NUMBER**: SPECIMEN_REFERENCE_NUMBER is no longer supported. Please replace this column header with SAMPLE_ID.
2. **STYLE_SHAPE**: when this column has a valid value, this event will be rendered using that shape. The valid shapes are `circle`, `square`, `triangle`, `diamond`, `star`, and `camera`.
3. **STYLE_COLOR**: when this column has a hexadecimal color value (e.g. #ffffff), it will be used as the color for rendering this event.

To embed hyperlinks in custom columns:

Values in custom columns can include markdown for hyperlinks that will show up in event tooltips, allowing users to click through to external resources associated with events.

![link-in-timeline-data](images/timeline-link.png)

Example column:

SOME_HYPERLINK: ```[Link text Here](https://link-url-here.org)```

The external link can be opened in a new tab, instead of an IFRAME within the same window/tab. To do this, the string `:blank` is to be added as a suffix at the end of the URL.

HYPERLINK_OPEN_IN_NEWTAB: ```[Link text Here](https://link-url-here.org/path:blank)```

### Event Types
As previously mentioned, the EVENT_TYPE can be anything. However, several event types have columns with special effects. Furthermore, for some event types cBioPortal has column naming suggestions.

_**EVENT_TYPE: TREATMENT**_

Suggested columns

 * **TREATMENT_TYPE**: This can be either Medical Therapy or Radiation Therapy.
 * **SUBTYPE**: Depending upon the TREATMENT_TYPE, this can either be Chemotherapy, Hormone Therapy, Targeted Therapy  etc. (for Medical Therapies) or WPRT, IVRT etc. (for Radiation Therapies).
 * **AGENT**: for medical therapies, the agent is defined with number of cycles if applicable and for radiation therapy, the agent is defined as standard dose given to the patient during the course.
 * **AGENT_CLASS**: This allows you to classify your agents into useful groups.
 * Based on different cancer types you can add additional data here.

Special: When using the AGENT and SUBTYPE columns, each agent and subtype will be split into its own track.

_**EVENT_TYPE: LAB_TEST**_

Suggested columns
 * **TEST**: type of test performed
 * **RESULT**: corresponding value of the test
 * Based on different cancer types you can add additional data here.

Special: When using the TEST and RESULT columns, each test gets its own track. Any TEST that has only numerical RESULT values will be rendered as a line chart.

_**EVENT_TYPE: IMAGING**_

Suggested columns
 * **DIAGNOSTIC_TYPE**: This attribute will cover the different diagnostics tools used (for example: MRI, CT scan etc.)
 * **DIAGNOSTIC_TYPE_DETAILED**: Detailed description of the event type.
 * **RESULT**: Results of the diagnostic tests
 * **SOURCE**: Where was the Imaging done.
 * Based on different cancer types you can add additional data here.

Special: all dots in the IMAGING track are squares.

_**EVENT_TYPE: STATUS**_

Suggested columns

 * **STATUS**: If the EVENT_TYPE is status, data is entered under STATUS to define either the best response from the treatment or if there is a diagnosis of any stage progression etc.
 * **SOURCE**: Where the status was monitored.
 * Based on different cancer types you can add additional data here.

_**EVENT_TYPE: SPECIMEN**_

Suggested columns

 * **SPECIMEN_SITE**: This is the site from where the specimen was collected.
 * **SPECIMEN_TYPE**: This can either be tissue or blood.
 * **SOURCE**: Where was the specimen collection done.
 * Based on different cancer types you can add additional data here.


### Clinical Track Ordering
Clinical tracks are ordered as follows (if available):

1. Specimen
2. Surgery
3. Status
4. Diagnostics
5. Diagnostic
6. Imaging
7. Lab_test
8. Treatment
9. First custom event
10. etc.

### Example

An example timeline file for SPECIMEN would be:
```
PATIENT_ID<TAB>START_DATE<TAB>EVENT_TYPE<TAB>SAMPLE_ID<TAB>SPECIMEN_SITE<TAB>SPECIMEN_TYPE<TAB>SOURCE<TAB>MyCustomColumn
CACO2<TAB>0<TAB>SPECIMEN<TAB>CACO2_S1<TAB>liver<TAB>tissue<TAB>hospital<TAB>T1
CACO2<TAB>100<TAB>SPECIMEN<TAB>CACO2_S2<TAB>lung<TAB>tissue<TAB>hospital<TAB>T2
...
```
Assuming the sample identifiers were also defined in the clinical file, this will lead to a timeline track with numbered specimen samples.

An example timeline file for Lab_test would be:
```
PATIENT_ID<TAB>START_DATE<TAB>EVENT_TYPE<TAB>TEST<TAB>RESULT
CACO2<TAB>100<TAB>LAB_TEST<TAB>PSA<TAB>10
CACO2<TAB>250<TAB>LAB_TEST<TAB>PSA<TAB>100
...
```
This will lead to a timeline track for Lab_test with an additional subtrack specifically for PSA. PSA's events will be sized based on the result.


## GISTIC 2.0 Data
Running GISTIC 2.0 on e.g. [GenePattern](https://cloud.genepattern.org) not only provides the [Discrete Copy Number Data](#discrete-copy-number-data), but also provides an amp_genes and a del_genes file. These cannot be directly imported into cBioPortal, but first have to be converted to a different file format. An example can be found in the [ACC TCGA study](https://github.com/cBioPortal/datahub/blob/master/public/acc_tcga/data_gistic_genes_amp.txt) on cBioPortal Datahub.

After uploading a gistic_amp and/or gistic_del file, significantly recurrently copy-number altered genes will be labeled and available for query in the "CNA Genes" table on the study view, like in the [TCGA Legacy / Firehose data set for bladder cancer](https://www.cbioportal.org/study/summary?id=blca_tcga):

![recurrently-altered-genes-table](images/recurrently-altered-genes-table.png)

### Meta file
The Gistic metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: GISTIC_GENES_AMP or GISTIC_GENES_DEL
3. **datatype**: Q-VALUE
4. **reference_genome_id**: reference genome version. Supported values: "hg19"
5. **data_filename**: your datafile

An example metadata file would be:
```
cancer_study_identifier: brca_tcga
genetic_alteration_type: GISTIC_GENES_AMP
datatype: Q-VALUE
reference_genome_id: hg19
data_filename: data_gistic_genes_amp.txt
```

### Data file

The following fields from the generated Gistic file are used by the cBioPortal importer:

* **chromosome**: chromosome on which the region was found, without the `chr` prefix
* **peak_start**: start coordinate of the region of maximal amplification or deletion within the significant region
* **peak_end**: end coordinate of the region of maximal amplification or deletion within the significant region
* **genes_in_region**: comma-separated list of HUGO gene symbols in the 'wide peak' (allowing for single-sample errors in the peak boundaries)
* **amp**: 1 for amp, 0 for del
* **cytoband**: cytogenetic band specification of the region, including chromosome (Giemsa stain)
* **q_value**: the q-value of the peak region

### Example 
An example data file which includes the required column header would look like:

```
chromosome<TAB>peak_start<TAB>peak_end<TAB>genes_in_region<TAB>amp<TAB>cytoband<TAB>q_value<TAB>
1<TAB>150563314<TAB>150621176<TAB>SNORA40|ENSG00000253047.1,RN7SL600P,RN7SL473P,C1orf138,LINC00568,CTSS,ECM1,ENSA,MCL1,RPRD2,ADAMTSL4,GOLPH3L,TARS2,HORMAD1,MIR4257,<TAB>1<TAB>1q21.3<TAB>2.7818E-43<TAB>
1<TAB>85988564<TAB>85991712<TAB>DDAH1,<TAB>1<TAB>1p22.3<TAB>4.1251E-13<TAB>
...
```


## MutSig Data
MutSig stands for "Mutation Significance".  MutSig analyzes lists of mutations discovered in DNA sequencing, to identify genes that were mutated more often than expected by chance given background mutation processes. You can download mutsig from [broadinstitute](https://www.broadinstitute.org/cancer/cga/mutsig_download) (MutSigCV 1.4 is available) or run mutsig (MutSigCV 1.2 is available) using [GenePattern](https://cloud.genepattern.org).

**Note:** The tcga files that are uploaded to cBioPortal are generated using MutSig2.0. This version is not available outside broadinstitute. 

_**The MutSigCV 1.2 output is different from the MutSig2.0 header. TODO: test the 1.4 version. Requires > 10GB of memory**_

After uploading a MutSig file, significantly recurrently mutated genes will be labeled and available for query in the "Mutated Genes" table on the study view, like in the [TCGA Legacy / Firehose data set for bladder cancer](https://www.cbioportal.org/study/summary?id=blca_tcga):

![recurrently-mutated-genes-table](images/recurrently-mutated-genes-table.png) 

This type data is not yet being validated. It can, however, be uploaded.

### Meta file
The MutSig metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: MUTSIG
3. **datatype**: Q-VALUE
4. **data_filename**: your datafile

An example metadata file would be:
```
cancer_study_identifier: brca_tcga
genetic_alteration_type: MUTSIG
datatype: Q-VALUE
data_filename: data_mutsig.txt
```


### Data file

The following fields from a MutSig file are used by the cBioPortal importer:
* **rank**
* **gene**: this is the HUGO symbol
* **N (or Nnon)**: bases covered
* **n (or nnon)**: number of mutations
* **p**: result of testing the hypothesis that all of the observed mutations in this gene are a consequence of random background mutation processes, taking into account the list of bases that are successfully interrogated by sequencing (i.e., ‚Äúcovered‚Äù) and the list of observed somatic mutations, as well as the length and composition of the gene in addition to the background mutation rates in different sequence contexts (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3059829/)
* **q**: p value correct for multiple testing


### Example 

An example data file which includes the required column header would look like:

```
rank<TAB>gene<TAB>N<TAB>n<TAB>p<TAB>q
1<TAB>RUNX1<TAB>1051659<TAB>29<TAB>1.11E-16<TAB>1.88E-12
2<TAB>PIK3CA<TAB>3200341<TAB>351<TAB><1.00e-15<TAB><2.36e-12
...
```

## Gene Panel Data
Gene panel functionality can specify which genes are assayed on a panel and assign samples and genetic profiles (such as mutation data) to a panel.

To include gene panel data in your instance, the following data and/or configurations can be used:
1. **Gene panel file**: This file contains the genes on the gene panel. A panel can be used for multiple studies within the instance and should be loaded prior to loading a study with gene panel data. For information on the format and import process please visit: [Import-Gene-Panels](Import-Gene-Panels.md).
2. **Gene panel matrix file**: This file is used to specify which samples are sequenced on which gene panel in which genetic profile. This is recommended for mutation and structural variant data, because the MAF and structural variant formats are unable to include samples which are sequenced but contain no called mutations, and only a single gene panel can be defined in the meta file. For other genetic profiles, columns can be added to specify their gene panel, but a property can also be added to their respective meta file, because these data files contain all profiled samples. Although the gene panel matrix functionality overlaps with the case list functionality, a case list for mutations (`_sequenced`) and Structural variants (`_sv`) is also required.
3. **Gene panel property in meta file**: Adding the `gene_panel:` property to the meta file of data profile will assign all samples from that profile to the gene panel. In this case it is not necessary to include a column for this profile in the gene panel matrix file.

### Gene Panel Matrix file

### Columns and rows
The gene panel matrix file contains a list of samples in the first column, and an additional column for each profile in the study using the stable_id as the column header. These stable_id's should match the ones in their respective meta files, for example `mutations` for mutation data and `gistic` for discrete CNA data. Columns should be separated by tabs. Fusion events are saved in the mutation table in the cBioPortal database, so they should be included in the `mutations` column. As described above, genetic profiles other than mutation and fusion data profiles can use the `gene_panel:` meta property if all samples are profiled on the same gene panel.

### Values
For each sample-profile combination, a gene panel should be specified. Please make sure this gene panel is imported before loading the study data. When the sample is not profiled on a gene panel, or if the sample is not profiled at all, use `NA` as value. If the sample is profiled for mutations, make sure it is also in the `_sequenced` case list. 

### Example 
An example file would look like this:

| SAMPLE_ID   | mutations | gistic    |
| ----------- | --------- | --------- |
| SAMPLE_ID_1 | IMPACT410 | IMPACT410 |
| SAMPLE_ID_2 | IMPACT410 | IMPACT410 |
| SAMPLE_ID_3 | NA       | NA        |

### Meta file
The gene panel matrix file requires a meta file, which should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: GENE_PANEL_MATRIX
3. **datatype**: GENE_PANEL_MATRIX
4. **data_filename**: your datafile

Example:
```
cancer_study_identifier: msk_impact_2017
genetic_alteration_type: GENE_PANEL_MATRIX
datatype: GENE_PANEL_MATRIX
data_filename: data_gene_panel_matrix.txt
```

### Gene panel property in meta file
If all samples in a genetic profile have the same gene panel associated with them, an optional field can be specified in the meta data file of that datatype called **gene_panel:**. If this is present, all samples in this data file will be assigned to this gene panel for this specific profile.

## Gene Set Data
A description of importing gene sets (which are required before loading gene set study) can be found [here](Import-Gene-Sets.md). This page also contains a decription to import gene set hierarchy data, which is required to show a hierarchical tree on the query page to select gene sets.

cBioPortal supports GSVA scores and GSVA-like scores, such as ssGSEA. The [Gene Set Variation Analysis](https://www.bioconductor.org/packages/release/bioc/html/GSVA.html) method in R (GSVA, [H√§nzelmann, 2013](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-7)) can calculate several types of scores (specified with the `methods=` argument) and outputs a score between -1 and 1. The GSVA method also calculates a p-value per score using a bootstrapping method.

To import the GSVA(-like) data, a score and p-value data file are required. It is important that the dimensions of the score and p-value file are the same and that they contain the same gene sets and samples. Both data files require a meta file.

### GSVA score meta file
The meta file will be similar to meta files of other genetic profiles, such as mRNA expression. For both GSVA and GSVA-like scores, `GSVA-SCORE` is used as `datatype` and `gsva_scores` is used as `stable_id`. 

Required fields: 
```
cancer_study_identifier: Same value as specified in study meta file
genetic_alteration_type: GENESET_SCORE
datatype: GSVA-SCORE
stable_id: Any unique identifier within the study
source_stable_id: Stable id of the genetic profile (in this same study) that was used as the input source for calculating the GSVA scores. Typically this will be one of the mRNA expression genetic profiles. 
profile_name: A name describing the analysis.
profile_description: A description of the data processing done.
data_filename: <your GSVA score datafile>
show_profile_in_analysis_tab: true
geneset_def_version: Version of the gene set definition this calculation was based on. 
```

Example:
```
cancer_study_identifier: study_es_0
genetic_alteration_type: GENESET_SCORE
datatype: GSVA-SCORE
stable_id: gsva_scores
source_stable_id: rna_seq_mrna
profile_name: GSVA scores on oncogenic signatures gene sets
profile_description: GSVA scores on oncogenic signatures gene sets using mRNA expression data calculated with GSVA version x with parameters x and y.
data_filename: data_gsva_scores.txt
show_profile_in_analysis_tab: true
geneset_def_version: msigdb_7.5.1
```

### GSVA score data file
The data file will be a simple tab separated format, similar to the expression data file: each sample is a column, each gene set a row, each cell contains the GSVA score for that sample x gene set combination.

The first column is `geneset_id` and contains the names of the gene sets. Gene set names should be formatted in uppercase. The other columns are sample columns: An additional column for each sample in the dataset using the sample id as the column header.

The cells contain the GSVA(-like) score: which is real number, between -1.0 and 1.0, representing the score for the gene set in the respective sample, or NA when the score for the gene set in the respective sample could not be (or was not) calculated. Example with 2 gene sets and 3 samples: 

<table>
<thead><tr><th>geneset_id</th><th>TCGA-AO-A0J</th><th>TCGA-A2-A0Y</th><th>TCGA-A2-A0S</th></tr></thead>
<tr><td>GO_POTASSIUM_ION_TRANSPORT</td><td>-0.987</td><td>0.423</td><td>-0.879</td></tr>
<tr><td>GO_GLUCURONATE_METABOLIC_PROCES</td><td>0.546</td><td>0.654</td><td>0.123</td></tr>
<tr><td>..</td><td></td><td></td><td></td></tr>
</table>

### GSVA p-value meta file

For both GSVA and GSVA-like p-values, `P-VALUE` is used as `datatype` and `gsva_pvalues` is used as `stable_id`. 

Required fields: 
```
cancer_study_identifier: Same value as specified in study meta file
genetic_alteration_type: GENESET_SCORE
datatype: P-VALUE
stable_id: Any unique identifier within the study
source_stable_id: Stable id of the GSVA-SCORE genetic profile (see above).
profile_name: A name describing the analysis.
profile_description: A description of the data processing done.
data_filename: <your GSVA p-value datafile>
geneset_def_version: Version of the gene sets definition this calculation was based on. 
```

Example:
```
cancer_study_identifier: study_es_0
genetic_alteration_type: GENESET_SCORE
datatype: P-VALUE
stable_id: gsva_pvalues
source_stable_id: gsva_scores
profile_name: GSVA p-values for GSVA scores on oncogenic signatures gene sets
profile_description: GSVA p-values for GSVA scores on oncogenic signatures gene sets using mRNA expression data calculated with the bootstrapping method in GSVA version x with parameters x and y.
data_filename: data_gsva_pvalues.txt
geneset_def_version: msigdb_7.5.1
```

### GSVA p-value data file
The data file will be a simple tab separated format, similar to the score file: each sample is a column, each gene set a row, each cell contains the p-value for the score found for sample x gene set combination.

The first column is `geneset_id` and contains the names of the gene sets. Gene set names should be formatted in uppercase. The other columns are sample columns: An additional column for each sample in the dataset using the sample id as the column header.

The cells contain the p-value for the GSVA score: A real number, between 0.0 and 1.0, representing the p-value for the GSVA score calculated for the gene set in the respective sample, or NA when the score for the gene is also NA. Example with 2 gene sets and 3 samples: 

<table>
<thead><tr><th>geneset_id</th><th>TCGA-AO-A0J</th><th>TCGA-A2-A0Y</th><th>TCGA-A2-A0S</th></tr></thead>
<tr><td>GO_POTASSIUM_ION_TRANSPORT</td><td>0.0811</td><td>0.0431</td><td>0.0087</td></tr>
<tr><td>GO_GLUCURONATE_METABOLIC_PROCES</td><td>0.6621</td><td>0.0031</td><td>1.52e-9</td></tr>
<tr><td>..</td><td></td><td></td><td></td></tr>
</table>

## Study Tags file
YAML or JSON file which contains extra information about the cancer study. No compulsory fields are required for this file (free-form). To enable this feature, you need to add a line in the cancer study meta file with `tags_file:` followed the YAML/JSON file name. The information on the YAML or JSON file will be displayed in a table when mousing over a tag logo in the studies on the query page.

## Generic Assay
Generic Assay is a two dimensional matrix generalized to capture non-genetic measurements per sample. Instead of a gene per row and a sample per column, a Generic Assay file contains a generic entity per row and a sample per column. A generic entity is defined by the data curator and generally means something other than a gene. Some examples include, treatment response or mutational signatures. For each generic entity - sample pair, a (real number / text / binary value) represents a captured measurement.

### Generic Assay meta file
The generic assay metadata file should contain the following fields:
```
cancer_study_identifier: Same value as specified in meta file of the study
genetic_alteration_type: GENERIC_ASSAY
generic_assay_type: <GENERIC_ASSAY_TYPE>, e.g., "TREATMENT_RESPONSE" or "MUTATIONAL_SIGNATURE"
datatype: LIMIT-VALUE, CATEGORICAL or BINARY
stable_id: Any unique identifier using a combination of alphanumeric characters, _ and -
profile_name: A name describing the analysis.
profile_description: A description of the data processing done.
data_filename: <name of generic assay data file>
show_profile_in_analysis_tab: true (or false if you don't want to show this profile in any analysis tab)
pivot_threshold_value: A threshold value beyond which a generic assay data is considered effective
value_sort_order: A flag that determines whether samples with small generic assay data values are displayed first or last; can be 'ASC' for small first, 'DESC' for small last. 
patient_level: false (or true if your data is patient level data)
generic_entity_meta_properties: A comma separate list of generic entity properties, e.g., "NAME,DESCRIPTION,URL"
```

Example:
```
cancer_study_identifier: study_es_0
genetic_alteration_type: GENERIC_ASSAY
generic_assay_type: TREATMENT_RESPONSE
datatype: LIMIT-VALUE
stable_id: treatment_ic50
profile_name: IC50 values of compounds on cellular phenotype readout
profile_description: IC50 (compound concentration resulting in half maximal inhibition) of compounds on cellular phenotype readout of cultured mutant cell lines.
data_filename: data_treatment_ic50.txt
show_profile_in_analysis_tab: true
pivot_threshold_value: 0.1
value_sort_order: ASC
patient_level: false
generic_entity_meta_properties: NAME,DESCRIPTION,URL
```

### Note on `stable_id`
The `stable_id` for the generic assay datatype is a user defined field. The only requirement is that the `stable_id` is unique across all metafiles in the study.

### Note on `pivot_threshold_value`
The `pivot_threshold_value` is an arbitrary value that specifies a critical boundary that distinguish _important_ from
_unimportant_ values in a generic assay profile. This boundary will be used in different visualizations to highlight
_important_ observations. Whether smaller or larger values are considered to be _important_ can be controlled with the
`value_sort_order` field (see [below](#note-on-value_sort_order)).

In a heatmap for a generic assay profile _important_ values are shown in darker shades of blue, whereas _unimportant_
values are shown in darker shades of red. The values represented by the `pivot_threshold_value` is shown in white. When
defined, the `pivot_threshold_value` will always be included in the legend, even when all datapoints are all more extreme.
In the Waterfall plot in Plots tab of Results view `pivot_threshold_value` determines the boundary between up- and downward
deflections in the plot.

When no sensible idea exists for a boundary between _important_ and _unimportant_ observations the `pivot_threshold_value`
field should not be defined (excluded from meta file). In this case all values will be shown as shades of blue in the
heatmap and the waterfall plot will show up- and downward deflections around 0.

### Note on `value_sort_order`
The `value_sort_order` field can be used to indicate whether small or large value are considered to be more `important`.
When `value_sort_order` is `ASC` smaller values are considered to be more _important_. When `value_sort_order` is `DESC`
larger values are considered to be more _important_.

The default value for `value_sort_order` is `ASC`.

The `value_sort_order` is used by the Oncoprint when aggregating generic assay data in a tooltip that covers multiple samples
from a single patient. When `value_sort_order` is `ASC` the sample with the smallest response value will be shown for that
patient. When `value_sort_order` is `DESC` the sample with the largest value will be shown for that patient.

In the heatmap for a generic assay profile in results view data points with smaller values will show as darker blue
when the `value_sort_order` is `ASC`. When `value_sort_order` is `DESC` larger values are assigned a darker blue color.

The `value_sort_order` is used by the Waterfall plot for orientation of the x-axis so that _important_ observations are
shown at the left side of the plot. When `value_sort_order` is `ASC` the x-axis will be in ascending order with smaller
values to the left.  When `value_sort_order` is `DESC` the x-axis will be in descending order with larger values to the left.

### Note on `generic_entity_meta_properties`
All meta properties must be specified in the `generic_entity_meta_properties` field. Every meta property listed here 
must appear as a column header in the corresponding data file. It's highly recommend to add `NAME`, `DESCRIPTION` and an 
optional `URL` to get the best visualization on OncoPrint tab and Plots tab.

### Note on `patient_level`
Generic Assay data will be considered `sample_level` data if the `patient_level` property is missing or set to `false`. 
In addition, the patient or sample identifiers need to be included in the [Clinical Data](#clinical-data) file.

### Note on `Generic Assay` genetic_alteration_type and datatype
All generic assay data is registered to be of the type of `genetic_alteration_type` and data type can choose from `LIMIT-VALUE`, `CATEGORICAL` and `BINARY`. 
- ***LIMIT-VALUE***: This datatype is intended to be used for any numerical data set with similar structure (entities measured in samples). The `LIMIT-VALUE` is validated to contain any continuous number optionally prefixed with a '>' or '<' threshold symbol (e.g., '>8.00').
- ***CATEGORICAL (under development)***: This datatype is intended to be used for any categorical data set with similar structure (entities measured in samples). Any text is allowed in `CATEGORICAL`.
- ***BINARY (under development)***: This datatype is intended to be used for any binary data set with similar structure (entities measured in samples). The `BINARY` is validated to contain only reserved text (`true`, `false`, `yes`, `no`).

If the value for the generic entity in the respective sample could not (or was not) be measured (or detected), the value should be 'NA' or leave that cell blank.

### Generic Assay data file
The data file will be a simple tab separated format, similar to the expression data file: each sample is a column, each generic entity is a row, each cell contains values for that generic entity x sample combination.

For each generic entity (row) in the data file, the following columns are required in the order specified:

- ***ENTITY_STABLE_ID***: Any unique identifier using a combination of alphanumeric characters, _ and -.

And:
- An additional column for each `generic_entity_meta_properties` in the metafile, using the property name as the column header (e.g., 'NAME').
- An additional column for each sample in the dataset using the sample id as the column header.

Example with 3 generic entities and 3 samples:

<table>
<thead><tr><th>ENTITY_STABLE_ID</th><th>NAME</th><th>DESCRIPTION</th><th>URL</th><th>1321N1_CENTRAL_NERVOUS_SYSTEM</th><th>22RV1_PROSTATE</th><th>42MGBA_CENTRAL_NERVOUS_SYSTEM</th></tr></thead>
<tr><td>17-AAG</td><td>Tanespimycin</td><td>Hsp90 inhibitor</td><td>https://en.wikipedia.org/wiki/Tanespimycin</td><td>0.228</td><td>0.330</td><td>0.0530</td></tr>
<tr><td>AEW541</td><td>Larotrectinib</td><td>TrkA/B/C inhibitor</td><td>https://en.wikipedia.org/wiki/Larotrectinib</td><td>>8</td><td>2.33</td><td>2.68</td></tr>
<tr><td>AZD0530</td><td>Saracatinib</td><td>Src/Bcr-Abl inhibitor</td><td>https://en.wikipedia.org/wiki/Saracatinib</td><td>NA</td><td>>8</td><td>4.60</td></tr>
</table>

### Arm Level CNA Data
Arm-level copy-number data is a predefined subtype of Generic Assay Data.

Allowed values for Arm-level copy-number data are `Loss`, `Gain`, and `Unchanged`, use `NA` or leave the cell blank to indicate a missing value.

Please find example file format here: [Meta file example](https://github.com/cBioPortal/cbioportal-frontend/blob/master/end-to-end-test/local/studies/lgg_ucsf_2014_test_generic_assay/meta_armlevel_CNA.txt) and [Data file example](https://github.com/cBioPortal/cbioportal-frontend/blob/master/end-to-end-test/local/studies/lgg_ucsf_2014_test_generic_assay/data_armlevel_CNA.txt)

### Mutational Signature Data
Mutational Signature data is a predefined subtype of Generic Assay Data. Setting `generic_assay_type: MUTATIONAL_SIGNATURE` 
in the meta file will make cBioPortal interpret the data as Mutational Signature data.

#### Mutational Signature meta files
The mutational signature meta files follow the same convention as the [Generic Assay Meta file](#generic-assay-meta-file),
however there are some key differences:
- `genetic_assay_type` should be set to `MUTATIONAL_SIGNATURE`
- `datatype` should be set to `LIMIT_VALUE`
- `stable_id` values should end with: `_{filetype}_{identifier}`, where:
  - `filetype` is either `contribution`, `pvalue` or `counts`
  - `identifier` is consistent between files belonging to the same analysis
  - Multiple signatures can be added to a single study, as long as they have different identifiers in their stable id (e.g., `contribution_SBS` and `contribution_DBS`)
- In `generic_entity_meta_properties` the `NAME` value is required. The `DESCRIPTION` and `URL` values can be added 
  to display more information and link to external resources in the mutational signatures tab.

#### Mutational Signature data files
The mutational signature data files follow the same convention as the [Generic Assay Data file](#generic-assay-data-file).
Each collection of mutational signatures can consist of up to three different data files, each with an accompanying meta file.
- Signature _contribution_ file (**required**)
  - Data file containing the contribution of each signature-sample pair. Values are expected to be 0 ‚â• x ‚â• 1.
- Signature _pvalue_ file (optional)
  - Data file containing p-values for each signature-sample pair. Values below 0.05 will be shown as significant.
- Mutational _counts_ matrix file (optional)
  - Data file containing nucleotide changes of a sample. cBioPortal has specific visualization options for single-base substitutions (96 channels), double-base substitutions (72 channels) and insertion/deletions (83 channels), following the signatures defined by [COSMIC](https://cancer.sanger.ac.uk/signatures/). But other channels can also be used. Values are expected to be positive integers.

## Resource Data

The resource data is used to capture resource data in patients, samples and studies. The resources will be represented by URLs with meta data. The types of resources include:
- Files: pdf, txt, png, json, etc.
- Web links: non-file links e.g. URLs to other systems

the resource file is split into a resource definition file, sample resource file, patient resource file and study resource file. All data files are required to have a matching meta file.

### Meta files
The resource metadata files have to contain the following fields:

1. **cancer_study_identifier**: same value specified in meta_study.txt
2. **resource_type**: value from (DEFINITION / SAMPLE / PATIENT / STUDY)
3. **data_filename**: your datafile

### Examples
An example metadata file, e.g. named meta_resource_definition.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
resource_type: DEFINITION
data_filename: data_resource_definition.txt
```

An example metadata file, e.g. named meta_resource_sample.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
resource_type: SAMPLE
data_filename: data_resource_sample.txt
```

### Data files

### Resource Definition Data File
The resource definition file should follow this format, it has three **required** columns:
- **RESOURCE_ID (required)**: a unique resource ID. This field allows only numbers, letters, points, underscores and hyphens.
- **DISPLAY_NAME (required)**: a display name for resources.
- **RESOURCE_TYPE (required)**: resource type for resources, must be SAMPLE, PATIENT or STUDY.
- **DESCRIPTION (optional)**: a discription for resources.
- **OPEN_BY_DEFAULT (optional)**: define if the resource will be open by default (`true` / `false`), dafault is `false`.
- **PRIORITY (optional)**: if not given, will give a default value.

### Example *Resource Definition* data file
<table>
<thead><tr><th>RESOURCE_ID</th><th>DISPLAY_NAME</th><th>RESOURCE_TYPE</th><th>DESCRIPTION</th><th>OPEN_BY_DEFAULT</th><th>PRIORITY</th></tr></thead>
<tr><td>PATHOLOGY_SLIDE</td><td>Pathology Slide</td><td>SAMPLE</td><td>The pathology slide for the sample</td><td>TRUE</td><td>1</td></tr>
<tr><td>PATIENT_NOTES</td><td>Patient Notes</td><td>PATIENT</td><td>Notes about the patient</td><td>FALSE</td><td>2</td></tr>
<tr><td>STUDY_SPONSORS</td><td>Study Sponsors</td><td>STUDY</td><td>Sponsors of this study</td><td>TRUE</td><td>3</td></tr>
</table>

### Sample Resource Data File
The sample resource file should follow this format, it has four **required** columns:
- **PATIENT_ID (required)**: a unique patient ID. This field allows only numbers, letters, points, underscores and hyphens.
- **SAMPLE_ID (required)**: a unique sample ID. This field allows only numbers, letters, points, underscores and hyphens.
- **RESOURCE_ID (required)**: a unique resource ID which should also be included in the `Resource Definition data file`.
- **URL (required)**: url to the resources, start with `http` or `https`.

### Example *Sample Resource* data file
<table>
<thead><tr><th>PATIENT_ID</th><th>SAMPLE_ID</th><th>RESOURCE_ID</th><th>URL</th></tr></thead>
<tr><td>TCGA-A2-A04P</td><td>TCGA-A2-A04P-01</td><td>PATHOLOGY_SLIDE</td><td>https://url-to-slide-sample1</td></tr>
<tr><td>TCGA-A1-A0SK</td><td>TCGA-A1-A0SK-01</td><td>PATHOLOGY_SLIDE</td><td>https://url-to-slide-sample2</td></tr>
</table>

### Patient Resource Data File
The patient resource file should follow this format, it has three **required** columns:
- **PATIENT_ID (required)**: a unique patient ID. This field allows only numbers, letters, points, underscores and hyphens.
- **RESOURCE_ID (required)**: a unique resource ID which should also be included in the `Resource Definition data file`.
- **URL (required)**: url to the resources, start with `http` or `https`.

### Example *Patient Resource* data file
<table>
<thead><tr><th>PATIENT_ID</th><th>RESOURCE_ID</th><th>URL</th></tr></thead>
<tr><td>TCGA-A2-A04P</td><td>PATIENT_NOTES</td><td>https://url-to-slide-patient1</td></tr>
<tr><td>TCGA-A1-A0SK</td><td>PATIENT_NOTES</td><td>https://url-to-slide-patient2</td></tr>
</table>

### Study Resource Data File
The study resource file should follow this format, it has two **required** columns:
- **RESOURCE_ID (required)**: a unique resource ID which should also be included in the `Resource Definition data file`.
- **URL (required)**: url to the resources, start with `http` or `https`.

### Example *Study Resource* data file
<table>
<thead><tr><th>RESOURCE_ID</th><th>URL</th></tr></thead>
<tr><td>STUDY_SPONSORS</td><td>https://url-to-study-sponsors</td></tr>
</table>

## Custom namespace columns

### Adding annotation columns through namespaces
Custom columns can be added to the data files of mutations, structural variants and discrete copy number (long) data. 
The columns can be imported through the namespace mechanism into a database table column called `ANNOTATION_JSON`. Any columns starting with a prefix specified in the `namespaces` field in the metafile will be imported into the database. Namespace columns should be formatted as the namespace and namespace attribute seperated with a period (e.g `ASCN.total_copy_number` where `ASCN` is the namespace and `total_copy_number` is the attribute).

An example cBioPortal mutation data file with the following **additional** columns:
```
ASCN.total_copy_number    ASCN.clonal     MUTATION.name    MUTATION.type
```
imported with the following `namespaces` field in the metafile:
```
namespaces: ascn
```
will import the `ASCN.total_copy_number` and `ASCN.clonal` column into the database. `MUTATION.name` and `MUTATION.type` 
will be ignored because `mutation` is not specified in the `namespaces` field. 

## Representation of namespace columns by mutation API endpoints

Columns added through namespaces will be returned by relevant mutation, discrete copy number and structural variant API 
endpoints. Namespace data will be available in the `namespaceColumn` of respective JSON representations of mutation 
records. The `namespaceColumns` property will be a JSON object where namespace data is keyed by name of the namespace in 
lowercase. For instance, when namespace `ZYGOSITY` is defined in the meta file and the data file has column 
`ZYGOSITY.status` with value `Homozygous` for a mutation row, the API will return the following JSON record for this 
mutation (only relevant fields are shown):

```
{
    "namespaceColumns": {
        "ZYGOSITY": {
            "status": "Homozygous"
        }
    },
}
```

Note: ASCN namespace data is not exported via the `namespaceColumns` field.

## Representation of namespace columns in the cBioPortal frontend

Namespace columns will be added as columns to mutation, structural variant and copy number alteration tables in Patient 
View and Results View. The case of the namespace in the column header will be as specified in the mutations meta file 
and the column name will be capitalized. 

For instance, this metafile entry:

```shell
namespaces: Zygosity
```

and this column header:

```shell
ZYGOSITY.status
```

will show in the mutation table with column name:

```shell
Zygosity Status
```

Note: namespace columns are recognized by a case-insensitive match of the namespace reported in the meta file
and the first word in the column header.
# Hardware Requirements

Hardware requirements will vary depending on the volume of users you anticipate
will access your cBioPortal instance and the amount of data loaded in the
portal. We run [cbioportal.org](https://www.cbioportal.org) on an AWS r5.xlarge
instance with 32 GB and 4 vCPUs. The public database consumes ~50 GB of disk
space. The site is visited by several thousands of users a day. For on-premise
installation recommendations one can look at the AWS instance type specs:
 
| Platform | instance type | (v)CPUs | RAM(GB) | Storage (GB) |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| aws | r5.xlarge | 4 | 32 | 50 |
| on-premise | - | 4 | 32 | 50 |


The hardware requirements are much lower when one has only a few users a day.
Minimally, 2GB of RAM is needed to run a cBioPortal instance. If you do not
plan to import public studies, depending on the size of your private data, 10GB
of disk space may be sufficient.

Another possible consideration is caching. The portal can cache responses to
requests so that repeated database queries are avoided. On the public cBioPortal
deployment we enable this cache and allocate 1GB of additional RAM and 4GB of
additional disk space for caching. For directions on configuring caching, see
[Ehcache Settings](/deployment/customization/application.properties-Reference.md#cache-settings)
# Import Gene Panels

This page describes how to import a gene panel into the cBioPortal database.  It assumes the following requirements have been satisfied:

1. The cBioPortal software has been correctly [built from source](/deployment/deploy-without-docker/Build-from-Source.md).
2. The gene panel to import is in the proper file format. See [Gene Panel File format](#gene-panel-file-format) for more information.
3. The `PORTAL_HOME` environment variable has been properly defined.  See [Loading a Sample Study](/deployment/deploy-without-docker/Load-Sample-Cancer-Study.md#set-the-portal_home-environment-variable) for more information.

## Gene panel file format
The gene panel file follows the format of a meta file with the following fields:
1. **stable_id**: The name of the gene panel. This should be unique across all studies, as gene panels can be globally applied to any sample and any genetic profile.
2. **description**: A description of the gene panel.
3. **gene_list**: Tab separated genes, represented either by all gene symbols or all Entrez gene IDs.

An example gene panel file would be:
```
stable_id: IMPACT410
description: Targeted (410 cancer genes) sequencing of various tumor types via MSK-IMPACT on Illumina HiSeq sequencers.
gene_list: ABL1    ACVR1   AKT1    AKT3 ...
```

#### Import command

In this example, we are loading the example gene panels which resides in the sample dataset `study_es_0`.

```
cd <cbioportal_source_folder>/core/src/main/scripts
./importGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel1.txt
./importGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel2.txt
```

After loading gene panels into the database, please restart Tomcat or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#cache-settings) for more information) so that the
validator can retrieve gene panel information from the cBioPortal API. 

#### Update existing gene panel

If a gene panel exists in the database with the same name as the one being imported, and there exists cancer study data that refers to this gene panel, the ImportGenePanel command will abort.  In order to reimport the gene panel in this situation, run the UpdateGenePanel command.  

If the incoming gene panel is the same as the original gene panel, whether through importing or updating, then no changes shall be made to the gene panel.  If the incoming gene panel is empty, then the script will abort.  Genes in the incoming gene panel that were not in the original shall be added to the existing gene panel. Conversely, genes not in the incoming gene panel that were in the original shall be removed from the existing gene panel.  The UpdateGenePanel command will prompt twice to confirm changes made to the gene panel, such as genes to be added or removed.  

```
cd <cbioportal_source_folder>/core/src/main/scripts
./updateGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel1.txt
```
# Import Gene Sets in cBioPortal

Gene sets are collections of genes that are grouped together based on higher level function or system characteristics, such as being part of the same molecular process or found to be co-regulated for example. Assessing gene sets in cBioPortal is useful when the user wants to visualize the number of mutations in sets of genes, or wants to see if all genes in a set are up- or down-regulated. To visualize gene set variation in a sample, the user can calculate scores per gene set per sample using the Gene Set Variation Analysis (GSVA) algorithm ([H√§nzelmann, 2013](#references)).

Before loading a study with gene set data, gene set definitions have to be added to the database. These can be custom user-defined sets, or sets downloaded from external sources such as [MSigDB](#references). Additionally, a gene set hierarchy can be imported which is used on the cBioPortal Query page for selecting gene sets.

## Quick example
This example shows how the process of importing gene set data using test data.

1. Navigate to scripts folder:

```
cd <cbioportal_source_folder>/core/src/main/scripts
```

2. Import gene sets and supplementary data:
Note: This removes existing gene set, gene set hierarchy and gene set genetic profile data.

```
./importGenesetData.pl \
	--data ../../test/resources/genesets/study_es_0_genesets.gmt \
	--new-version msigdb_6.1 \
	--supp ../../test/resources/genesets/study_es_0_supp-genesets.txt
```

3. Import gene set hierarchy data:

```
./importGenesetHierarchy.pl \
	--data ../../test/resources/genesets/study_es_0_tree.yaml
```

4. Restart Tomcat if you have it running or call the `/api/cache` endpoint with a `DELETE` http-request
   (see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information).


5. Import study (replace argument after `-u` with local cBioPortal and `-html` with preferred location for html report):

```
./importer/metaImport.py \
	-s ../../test/scripts/test_data/study_es_0 \
	-u http://cbioportal:8080/cbioportal \
	-html report_study_es_0.html \
	-v \
	-o
```

## Requirements for gene sets in cBioPortal
Gene set functionality was added in cBioPortal 1.7.0. Please use this or a later version. In addition, the database has to be updated to version 2.3.0 or higher, depending on the cBioPortal version. This can be done by running the python wrapper `migrate_db.py` for `migration.sql`.

Updating the database is described [here](https://github.com/cBioPortal/cbioportal/blob/master/docs/Updating-your-cBioPortal-installation.md#running-the-migration-script).

## Import Gene Sets

### File formats
Once you have initialized MySQL with cBioPortal database, it is possible to import gene sets. The format of the gene set data file is [the Gene Matrix Transposed file format (.gmt)](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29). This format is also used by the [MSigDB](#references), which hosts several collections of gene sets on: [https://software.broadinstitute.org/gsea/msigdb/](https://software.broadinstitute.org/gsea/msigdb/)

Sample of .gmt file:

```
GLI1_UP.V1_DN<TAB>https://...<TAB>22818<TAB>143384
GLI1_UP.V1_UP<TAB>https://...<TAB>3489<TAB>3119
E2F1_UP.V1_DN<TAB>https://...<TAB>7041<TAB>6374<TAB>5460
```

GMT files contain a row for every gene set. The first column contains the EXTERNAL_ID or `stable id` (MsigDB calls this "standard name"), e.g. GO_POTASSIUM_ION_TRANSPORT, not longer than 100 characters. The second column contains the REF_LINK. This is an optional URL linking to external information about this gene set. Column 3 to N contain the Entrez gene IDs that belong to this gene set.

Additional information can be placed in a supplementary file. This file should be a .txt, containing columns for the `stable id`, the long name (max 100 characters) and description of the gene set (max 300 characters).

Sample of supplementary .txt file:

```
GLI1_UP.V1_DN<TAB>GLI1 upregulated v1 down genes<TAB>Genes down-regulated in RK3E cells (kidney epithelium) over-expressing GLI1 [GeneID=2735].
GLI1_UP.V1_UP<TAB>GLI1 upregulated v1 up genes<TAB>Genes up-regulated in RK3E cells (kidney epithelium) over-expressing GLI1 [GeneID=2735].
E2F1_UP.V1_DN<TAB>E2F1 upregulated v1 down genes<TAB>Identification of E2F1-regulated genes that modulate the transition from quiescence into DNA synthesis, or have roles in apoptosis, signal transduction, membrane biology, and transcription repression.

```
### Run the gene set importer
The importer for gene sets can be run with a perl wrapper, which is located at the following location and requires the following arguments:

```
cd <cbioportal_source_folder>/core/src/main/scripts
perl importGenesetData.pl

required:     --data <data_file.gmt>
              --new-version <Version> OR --update-info
optional:     --supp <supp_file.txt>
```
The `--new-version` argument with a `<Version>` parameter is used for loading new gene set definitions. It is not possible to add new gene sets or change the genes of current gene sets, without removing the old gene sets first. This is to prevent the user from having gene sets from different definitions and data from older definitions. The user can choose the name or number of the `<Version>` as he likes, e.g. `msigdb_6.1` or `Oncogenic_2017`. Running the script with `--new-version` **removes all previous gene sets, gene set hierarchy and gene set genetic profiles.** A prompt is given to make sure the user wants to do this. Note that it is possible enter the same version as the previous version, but previous data is removed nevertheless.

The `--update info` can be used only to update only the long name, description and reference URL.

## Import Gene Set hierarchy

After importing gene sets, you can import a gene set hierarchy that is used on the query page to select gene sets.

### File format
For gene set hierarchy files, we use the YAML format. This is common format to structure hierarchical data.

Sample of format (note this is mock data):

```
Custom:
  Gene sets:
    - BCAT.100_UP.V1_DN
Cancer Gene sets from Broad:
  Gene sets:
    - AKT_UP.V1_DN
    - CYCLIN_D1_KE_.V1_UP
  Broad Subcategory 1:
    Gene sets:
      - HINATA_NFKB_MATRIX
  Broad Subcategory 2:
    Gene sets:
      - GLI1_UP.V1_UP
      - GLI1_UP.V1_DN
      - CYCLIN_D1_KE_.V1_UP

```

To make your own hierarchy, make sure every branchname ends with `:`. Every branch can contain new branches (which can be considered subcategories) or gene sets (which are designated by the `Gene sets:` statement). The gene set names are the `stable ids` imported by `ImportGenesetData.java` and should start with `-`.

### Running the gene set hierarchy importer
```
cd <cbioportal_source_folder>/core/src/main/scripts
perl importGenesetHierarchy.pl

required:     --data <data_file.yaml>
```

## Import a study with gene set data

Gene set data can be added to a study folder and subsequently import the whole study with `metaImport.py`. cBioPortal supports GSVA Scores and p-values (from bootstrapping) calculated using Gene Set Variation Analysis (GSVA, [H√§nzelmann, 2013](#references)).
A description of GSVA study data can be found in the [cBioPortal File Formats documentation](File-Formats.md#gene-set-data).

## References
**GSVA: gene set variation analysis for microarray and RNA-Seq data**<br>
Sonja H√§nzelmann, Robert Castelo and Justin Guinney, *BMC Bioinformatics, 2013*<br>
https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-7 <br>
https://www.bioconductor.org/packages/release/bioc/html/GSVA.html

**Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles**<br>
Aravind Subramanian, Pablo Tamayo, Vamsi K. Mootha, Sayan Mukherjee, Benjamin L. Ebert, Michael A. Gillette, Amanda Paulovich, Scott L. Pomeroy, Todd R. Golub, Eric S. Lander, and Jill P. Mesirov, *PNAS, 2005*<br>
https://www.pnas.org/content/102/43/15545 <br>
https://software.broadinstitute.org/gsea/msigdb
# Import OncoKB annotations as custom driver annotations

The _Annotation Configuration_ menu in Study View and Group Comparison is available only when _custom driver annotations_
are present in the cBioPortal database for the genes in the study (or studies). In order to use OncoKB annotations to filter
mutations and discrete copy number alteration in Study View and Group Comparison, OncoKB annotations can be added to the 
respective data files of a study prior to import into the database. This page describes how to import OncoKB annotations 
as custom driver annotations. It assumes the following requirements have been satisfied:

1. The cBioPortal software has been correctly [built from source](/deployment/deploy-without-docker/Build-from-Source.md).
2. The user is able to successfully import a [study into the database](/Data-Loading.md).
3. The study subjected to OncoKB import is [confirmed to be valid](/Data-Loading.md#validating-your-study-data)

## Import of OncoKB annotations when loading a study

OncoKB annotations can be added automatically to the study files when the study is loaded into the database by adding the
`--import_oncokb` parameter to the [metaImport.py](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/importer/metaImport.py) script like so:

```
python metaImport.py --import_oncokb -u http://cbioportal:8080 -s study/lgg_ucsf_2014/
```

This will add OncoKB data to the mutation and discrete CNA files of a study, revalidate the results and load the study
into the database.

The addition of mutation and discrete CNA files is explained in detail below.

## Update of MAF file with OncoKB annotations

OncoKB annotations can be added to the MAF file by running [importOncokbMutation.py](https://github.com/cBioPortal/cbioportal/tree/master/core/src/main/scripts/importer/importOncokbMutation.py) like so:

```
python importOncokbMutation.py -s study/lgg_ucsf_2014/ -u https://cbioportal.org
```

Where `-s` is the path to the directory of the MAF file and `-u` is the URL to 
a cBioPortal instance (needed for resolution of gene identifiers).

[importOncokbMutation.py](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/importer/importOncokbMutation.py) will add OncoKB annotations as
[custom driver annotation columns](/File-Formats.md#custom-driver-annotations) in the MAF file. The unmodified MAF file
will be stored in the study directory with the _ONCOKB_IMPORT_BACKUP__ prefix.


## Update of Discrete Copy Number file with OncoKB annotations

OncoKB annotations can be added to the Discrete Copy Number data by running [importOncokbDiscreteCNA.py](https://github.com/cBioPortal/cbioportal/tree/master/core/src/main/scripts/importer/importOncokbDiscreteCNA.py) like so:

```
python importOncokbDiscreteCNA.py -s /tmp/study -u https://cbioportal.org
```

Where `-s` is the path to the directory of the iscrete Copy Number data file and `-u` is the URL to 
a cBioPortal instance (needed for resolution of gene identifiers).

[importOncokbDiscreteCNA.py](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/importer/importOncokbMutation.py) will create a [custom driver annotation file](File-Formats.md#custom-driver-annotations-file)
with name _data_cna_pd_annotation.txt_ in the study directory. It will add a field `pd_annotations_filename` field in the 
CNA meta file that references the newly created custom driver annotation file. The unmodified CNA meta file will be stored
in the study directory with the _ONCOKB_IMPORT_BACKUP__ prefix.
# Add a new reference genome to the database
This manual is intended to guide you to load new reference genome(s) to the database. 

### What is Reference Genome (also know as reference assembly)?
A digital nucleic acid sequence database assembled by scientists as a representative example of a species' set of genes. Reference genomes can be accessed online using 
dedicated browsers such as [UCSC Genome Browser](https://genome.ucsc.edu/cgi-bin/hgGateway). 

### Why does reference genomes matter?
Mutation and Segment plots on Patient View are using genomic coordinates. The **cytoband** column in the Mutated Gene table on Study view rely on the Genome Reference Consortium Build 
matching the version used to analyse the original data (before it was loaded into cBioPortal DB).

## How to load reference genome(s) to the database?

### 1. prepare a tab delimited txt file with the following fields:
1. species: the group of organisms e.g human
2. name: the name of reference genome as used by the UCSC browser eg. hg19
3. build_name: the version of Genome Reference Consortium Build published by NCBI e.g GRCh38
4. nonN_bases:  the total number of non-N bases in reference genome FASTA-formatted file
5. URL: the URL to download the reference genome
6. release_date: when reference genome released, in a format of yyyy-mm-dd. The release date normally is included in the README.txt file in the download directory.

here is a sample reference genome file:
```
#species	name	build_name	nonN_bases	URL	release_date
human	hg19	GRCh37	2897310462	https://hgdownload.cse.ucsc.edu/goldenPath/hg19	2009-02-01
human	hg38	GRCh38	3049315783	https://hgdownload.cse.ucsc.edu/goldenPath/hg38	2013-12-24
mouse	mm10	GRCm38	2652783500	https://hgdownload.cse.ucsc.edu/goldenPath/mm10	2011-12-01
```
### 2. Using import script to load reference genome(s) data to the database:
```
 cd <your_cbioportal_dir>/core/src/main/scripts
 export PORTAL_HOME=<your_cbioportal_dir>
./importReferenceGenome.pl --ref-genome <your_reference_genome_file>
```
### 3. output of the script runs successfully
```
$ ./importReferenceGenome.pl --ref-genome ~/myspace/cbioportal/core/src/test/resources/reference_genomes.txt
   Reading reference genome from:  /Users/kelsyzhu/myspace/cbioportal/core/src/test/resources/reference_genomes.txt
    --> total number of lines:  3
   
   Done. Restart tomcat to make sure the cache is replaced with the new data.
   
   Warnings / Errors:
   -------------------
   0.  New reference genome added; 3x
   Done.
   Total time:  3432 ms

```# Import Study Using Docker

:warning: Every time you add/remove/overwrite a study please restart tomcat (or the Docker container), or 
call the `/api/cache` endpoint with a `DELETE` http-request (see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information).

## Adding a Study

In a docker terminal type the following command:

```bash
docker exec -it ${CBIOPORTAL_CONTAINER_NAME} \
metaImport.py -u http://${CBIOPORTAL_CONTAINER_NAME}:8080/cbioportal -s ${PATH_TO_STUDIES}
```

Where:    
- **${CBIOPORTAL_CONTAINER_NAME}**: The name of your cbioportal container instance, _i.e **cbioportal**_.
- **${PATH_TO_STUDIES}**: The external path where cBioPortal studies are stored.

## Restarting Tomcat

In a docker terminal type the following command:

```bash
docker exec -it ${CBIOPORTAL_CONTAINER_NAME} \
sh '${CATALINA_HOME}'/bin/shutdown.sh \
'&&' sh '${CATALINA_HOME}'/bin/startup.sh \
```

Where:    
- **${CBIOPORTAL_CONTAINER_NAME}**: The name of your cbioportal container instance, _i.e **cbioportal**_.
Relevant classes:
* [MafUtil](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/maf/MafUtil.java)
* [ExtendedMutationUtil](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/util/ExtendedMutationUtil.java)
* [MafRecord](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/maf/MafRecord.java)

### [MafUtil](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/maf/MafUtil.java) - to - cBioPortal Database mapping.
|        MafUtil Field Names       |         MAF Column Name          |                    `MUTATION.<FIELD>`                    |                                 `MUTATION_EVENT.<FIELD>`                                 |                                                                                                                                                                                           NOTES                                                                                                                                                                                           |
| -------------------------------- | -------------------------------- | -------------------------------------------------------- | ---------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| HUGO_SYMBOL                      | Hugo_Symbol                      |                                                          |                                                                                          | Helps resolve CanonicalGene (default by Entrez Gene id)                                                                                                                                                                                                                                                                                                                                   |
| ENTREZ_GENE_ID                   | Entrez_Gene_Id                   | mutation.entrez_gene_id                                  | mutation_event.entrez_gene_id                                                            | References gene.entrez_gene_id                                                                                                                                                                                                                                                                                                                                                            |
| CENTER                           | Center                           | mutation.center                                          |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| NCBI_BUILD                       | NCBI_Build                       |                                                          | mutation_event.ncbi_build                                                                |                                                                                                                                                                                                                                                                                                                                                                                           |
| CHROMOSOME                       | Chromosome                       |                                                          | mutation_event.chr                                                                       |                                                                                                                                                                                                                                                                                                                                                                                           |
| START_POSITION                   | Start_Position                   |                                                          | mutation_event.start_position                                                            |                                                                                                                                                                                                                                                                                                                                                                                           |
| END_POSITION                     | End_Position                     |                                                          | mutation_event.end_position                                                              |                                                                                                                                                                                                                                                                                                                                                                                           |
| STRAND                           | Strand                           |                                                          | mutation_event.strand                                                                    |                                                                                                                                                                                                                                                                                                                                                                                           |
| VARIANT_CLASSIFICATION           | Variant_Classification           |                                                          | mutation_event.mutation_type **                                                          | see ExtendedMutationUtil.getMutationType()                                                                                                                                                                                                                                                                                                                                                |
| VARIANT_TYPE                     | Variant_Type                     |                                                          | mutation_event.variant_type                                                              |                                                                                                                                                                                                                                                                                                                                                                                           |
| REFERENCE_ALLELE                 | Reference_Allele                 |                                                          | mutation_event.reference_allele                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| TUMOR_SEQ_ALLELE_1               | Tumor_Seq_Allele1                | mutation.tumor_seq_allele1                               | mutation_event.tumor_seq_allele                                                          | see MafUtil.resolveTumorSeqAllele()                                                                                                                                                                                                                                                                                                                                                       |
| TUMOR_SEQ_ALLELE_2               | Tumor_Seq_Allele2                | mutation.tumor_seq_allele2                               | mutation_event.tumor_seq_allele                                                          | see MafUtil.resolveTumorSeqAllele()                                                                                                                                                                                                                                                                                                                                                       |
| DBSNP_RS                         | dbSNP_RS                         |                                                          | mutation_event.db_snp_rs                                                                 |                                                                                                                                                                                                                                                                                                                                                                                           |
| DBSNP_VAL_STATUS                 | dbSNP_Val_Status                 |                                                          | mutation_event.db_snp_val_status                                                         |                                                                                                                                                                                                                                                                                                                                                                                           |
| TUMOR_SAMPLE_BARCODE             | Tumor_Sample_Barcode             | mutation.sample_id                                       |                                                                                          | sample.stable_id / sample.internal_id                                                                                                                                                                                                                                                                                                                                                     |
| MATCHED_NORM_SAMPLE_BARCODE      | Matched_Norm_Sample_Barcode      | mutation.matched_norm_sample_barcode                     |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MATCH_NORM_SEQ_ALLELE1           | Match_Norm_Seq_Allele1           | mutation.match_norm_seq_allele1                          |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MATCH_NORM_SEQ_ALLELE2           | Match_Norm_Seq_Allele2           | mutation.match_norm_seq_allele2                          |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| TUMOR_VALIDATION_ALLELE1         | Tumor_Validation_Allele1         | mutation.tumor_validation_allele1                        |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| TUMOR_VALIDATION_ALLELE2         | Tumor_Validation_Allele2         | mutation.tumor_validation_allele2                        |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MATCH_NORM_VALIDATION_ALLELE1    | Match_Norm_Validation_Allele1    | mutation.match_norm_validation_allele1                   |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MATCH_NORM_VALIDATION_ALLELE2    | Match_Norm_Validation_Allele2    | mutation.match_norm_validation_allele2                   |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| VERIFICATION_STATUS              | Verification_Status              | mutation.verification_status                             |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| VALIDATION_STATUS                | Validation_Status                | mutation.validation_status                               |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MUTATION_STATUS                  | Mutation_Status                  | mutation.mutation_status                                 |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| SEQUENCING_PHASE                 | Sequencing_Phase                 | mutation.sequencing_phase                                |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| SEQUENCE_SOURCE                  | Sequence_Source                  | mutation.sequence_source                                 |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| VALIDATION_METHOD                | Validation_Method                | mutation.validation_method                               |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| SCORE                            | Score                            | mutation.score                                           |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| BAM_FILE                         | BAM_File                         | mutation.bam_file                                        |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| SEQUENCER                        | Sequencer                        | mutation.sequencer                                       |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| AMINO_ACID_CHANGE                | Amino_Acid_Change                | mutation.amino_acid_change                               | mutation_event.protein_change **                                                         | see ExtendedMutationUtil.getProteinChange()                                                                                                                                                                                                                                                                                                                                               |
| T_REF_COUNT                      | t_ref_count                      | mutation.tumor_ref_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorRefCount()                                                                                                                                                                                                                                                                                                                                               |
| T_ALT_COUNT                      | t_alt_count                      | mutation.tumor_alt_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                                                               |
| N_REF_COUNT                      | n_ref_count                      | mutation.normal_ref_count **                             |                                                                                          | see ExtendedMutationUtil.getNormalRefCount()                                                                                                                                                                                                                                                                                                                                              |
| N_ALT_COUNT                      | n_alt_count                      | mutation.normal_alt_count **                             |                                                                                          | see ExtendedMutationutil.getNormalAltCount()                                                                                                                                                                                                                                                                                                                                              |
| I_T_REF_COUNT                    | i_t_ref_count                    | mutation.tumor_ref_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorRefCount()                                                                                                                                                                                                                                                                                                                                               |
| AD_REF                           | AD_Ref                           | mutation.tumor_ref_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorRefCount()                                                                                                                                                                                                                                                                                                                                               |
| I_T_ALT_COUNT                    | i_t_alt_count                    | mutation.tumor_alt_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                                                               |
| AD_ALT                           | AD_Alt                           | mutation.tumor_alt_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                                                               |
| NORM_AD_REF                      | Norm_AD_Ref                      | mutation.normal_ref_count **                             |                                                                                          | see ExtendedMutationUtil.getNormalRefCount()                                                                                                                                                                                                                                                                                                                                              |
| NORM_AD_ALT                      | Norm_AD_Alt                      | mutation.normal_alt_count **                             |                                                                                          | see ExtendedMutationutil.getNormalAltCount()                                                                                                                                                                                                                                                                                                                                              |
| T_TOT_COV                        | TTotCov                          | mutation.tumor_ref_count / mutation.tumor_alt_count **   |                                                                                          | see ExtendedMutationUtil.getTumorRefCount() / ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                     |
| T_VAR_COV                        | TVarCov                          | mutation.tumor_ref_count / mutation.tumor_alt_count **   |                                                                                          | see ExtendedMutationUtil.getTumorRefCount() / ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                     |
| N_TOT_COV                        | NTotCov                          | mutation.normal_ref_count / mutation.normal_alt_count ** |                                                                                          | see ExtendedMutationUtil.getNormalRefCount() / ExtendedMutationUtil.getNormalAltCount()                                                                                                                                                                                                                                                                                                   |
| N_VAR_COV                        | NVarCov                          | mutation.normal_ref_count / mutation.normal_alt_count ** |                                                                                          | see ExtendedMutationUtil.getNormalRefCount() / ExtendedMutationUtil.getNormalAltCount()                                                                                                                                                                                                                                                                                                   |
| TUMOR_DEPTH                      | tumor_depth                      | mutation.tumor_ref_count / mutation.tumor_alt_count **   |                                                                                          | see ExtendedMutationUtil.getTumorRefCount() / ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                     |
| TUMOR_VAF                        | tumor_vaf                        | mutation.tumor_ref_count / mutation.tumor_alt_count **   |                                                                                          | see ExtendedMutationUtil.getTumorRefCount() / ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                     |
| NORMAL_DEPTH                     | normal_depth                     | mutation.normal_ref_count / mutation.normal_alt_count ** |                                                                                          | see ExtendedMutationUtil.getNormalRefCount() / ExtendedMutationUtil.getNormalAltCount()                                                                                                                                                                                                                                                                                                   |
| NORMAL_VAF                       | normal_vaf                       | mutation.normal_ref_count / mutation.normal_alt_count ** |                                                                                          | see ExtendedMutationUtil.getNormalRefCount() / ExtendedMutationUtil.getNormalAltCount()                                                                                                                                                                                                                                                                                                   |
| PROTEIN_CHANGE                   | HGVSp_Short                      |                                                          | mutation_event.protein_change **                                                         | see ExtendedMutationUtil.getProteinChange()                                                                                                                                                                                                                                                                                                                                               |
| CODONS                           | Codons                           |                                                          | mutation_event.oncotator_codon_change **                                                 | see ImportExtendedMutation (codonChange = record.getCodons()) / MAF field "Codons" always overrides MAF value ONCOTATOR_CODON_CHANGE // "ONCOTATOR_CODON_CHANGE" never used?                                                                                                                                                                                                              |
| SWISSPROT                        | SWISSPROT                        |                                                          | mutation_event.oncotator_uniprot_accession **                                            | see ImportExtendedMutation (uniprotAccession = record.getSwissprot()) / if current accession type is SwissProt then record.getSwissprot() = uniprotAccession OR DaoUniProtIdMapping.mapFromUniprotIdToAccession(uniprotName) where uniprotName = record.getSwissprot() / OVERRIDES MAF field "ONCOTATOR_UNIPROT_ACCESSION"                                                                |
| REFSEQ                           | RefSeq                           |                                                          | mutation_event.oncotator_refseq_mrna_id **                                               | see ImportExtendedMutation(refseqMrnaId = record.getRefSeq()) / Always overrides MAF field "ONCOTATOR_REFSEQ_MRNA_ID"                                                                                                                                                                                                                                                                     |
| PROTEIN_POSITION                 | Protein_position                 |                                                          | mutation_event.oncotator_protein_pos_start / mutation_event.oncotator_protein_pos_end ** | see ExtendedMutationUtil.getProteinPosStart() / resolved from record.getProteinPosition() or record.getProteinChange() based on what is avialable / MAF field "ONCOTATOR_PROTEIN_POS_START" and "ONCOTATOR_PROTEIN_POS_END" never used?                                                                                                                                                   |
| ONCOTATOR_VARIANT_CLASSIFICATION | ONCOTATOR_VARIANT_CLASSIFICATION |                                                          | mutation_event.mutation_type **                                                          | see ExtendedMutationUtil.getMutationType()                                                                                                                                                                                                                                                                                                                                                |
| ONCOTATOR_REFSEQ_MRNA_ID         | ONCOTATOR_REFSEQ_MRNA_ID         |                                                          | mutation_event.oncotator_refseq_mrna_id ** [maps to DB but MAF value never imported]     | see ImportExtendedMutation(refseqMrnaId = record.getRefSeq()) / Always overrides MAF field "ONCOTATOR_REFSEQ_MRNA_ID" / MafRecord (record.getOncotatorRefseqMrnaId()) never used                                                                                                                                                                                                          |
| ONCOTATOR_UNIPROT_ENTRY_NAME     | ONCOTATOR_UNIPROT_ENTRY_NAME     |                                                          | mutation_event.oncotator_uniprot_entry_name ** [maps to DB but MAF value never imported] | see ImportExtendedMutation (record.getOncotatorUniprotName() never used to set mutation_event.oncotator_uniprot_entry_name) / MafRecord (record.getOncotatorUniprotName()) never used                                                                                                                                                                                                     |
| ONCOTATOR_UNIPROT_ACCESSION      | ONCOTATOR_UNIPROT_ACCESSION      |                                                          | mutation_event.oncotator_uniprot_accession ** [maps to DB but MAF value never imported]  | see ImportExtendedMutation (uniprotAccession = record.getSwissprot()) / if current accession type is SwissProt then record.getSwissprot() = uniprotAccession OR DaoUniProtIdMapping.mapFromUniprotIdToAccession(uniprotName) where uniprotName = record.getSwissprot() / OVERRIDES MAF field "ONCOTATOR_UNIPROT_ACCESSION" / MafRecord (record.getOncotatorUniprotAccession()) never used |
| ONCOTATOR_CODON_CHANGE           | ONCOTATOR_CODON_CHANGE           |                                                          | mutation_event.oncotator_codon_change ** [maps to DB but MAF value never imported]       | see ImportExtendedMutation (codonChange = record.getCodons()) / MAF field "Codons" always overrides MAF value ONCOTATOR_CODON_CHANGE // "ONCOTATOR_CODON_CHANGE" never used? / MafRecord (record.getOncotatorCodonChange()) never used                                                                                                                                                    |
| ONCOTATOR_PROTEIN_POS_START      | ONCOTATOR_PROTEIN_POS_START      |                                                          | mutation_event.oncotator_protein_pos_start ** [maps to DB but MAF value never imported]  | see ExtendedMutationUtil.getProteinPosStart() / resolved from record.getProteinPosition() or record.getProteinChange() based on what is avialable / MAF field "ONCOTATOR_PROTEIN_POS_START" never used?                                                                                                                                                                                   |
| ONCOTATOR_PROTEIN_POS_END        | ONCOTATOR_PROTEIN_POS_END        |                                                          | mutation_event.oncotator_protein_pos_end ** [maps to DB but MAF value never imported]    | see ExtendedMutationUtil.getProteinPosStart() / resolved from record.getProteinPosition() or record.getProteinChange() based on what is avialable / MAF field "ONCOTATOR_PROTEIN_POS_END" never used?                                                                                                                                                                                     |
| DRIVER_FILTER                    | cbp_driver                       | mutation.driver_filter                                   |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| DRIVER_FILTER_ANNOTATION         | cbp_driver_annotation            | mutation.driver_filter_annotation                        |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| DRIVER_TIERS_FILTER              | cbp_driver_tiers                 | mutation.driver_tiers_filter                             |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| DRIVER_TIERS_FILTER_ANNOTATION   | cbp_driver_tiers_annotation      | mutation.driver_tiers_filter_annotation                  |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
:warning:  ***Warning: this way of loading data is deprecated. See [Data loading](Data-Loading.md) for the recommended method.***

The page describes how to import data into the cBioPortal using Python scripts found in our [scripts](https://github.com/cBioPortal/cbioportal/tree/stable/core/src/main/scripts) directory.  The follow assumptions have been made:

1. The cBioPortal software has been correctly [built from source](/deployment/deploy-without-docker/Build-from-Source.md).
2. The data to import is in the proper [File Format](/File-Formats.md).
3. The `PORTAL_HOME` environment variable has been properly defined.

## Import Cancer Type

The following command is used to import cancer type metadata into the cBioPortal database:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-cancer-type --meta-filename /path-to-cancer-type-file/cancer_type.txt
```

The [cancer_type.txt](File-Formats.md#cancer-study) file should conform to the file format describe on the File Formats wiki page.

## Import Cancer Study

The following command is used to import cancer study metadata into the cBioPortal database:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-study --meta-filename /path-to-meta_study-file/meta_study.txt
```

The [meta_study.txt](File-Formats.md#cancer-study) file should conform to the file format describe on the File Formats wiki page.

## Remove Cancer Study

The following command is used to remove a cancer study from the cBioPortal database:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command remove-study --meta-filename /path-to-meta_study-file/meta_study.txt
```

The [meta_study.txt](File-Formats.md#cancer-study) file should conform to the file format describe on the File Formats wiki page.

## Import Study Data

The following command is used to import all types of genomic and clinicla data described on our [File Formats](File-Formats.md) wiki page.  Below are examples of import copy number and mutation data:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-study-data --meta-filename /path-to-meta_CNA.txt/meta_CNA.txt --data-filename /path-to-data_CNA.txt/data_CNA.txt
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-study-data --meta-filename /path-to-meta_mutations_extended.txt/meta_mutations_extended.txt --data-filename /path-to-data_mutations_extended.txt/data_mutations_extended.txt
```

## Import Case Lists

The following command is used to import case lists into the cBioPortal database:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-case-list --meta-filename /path-to-case-list/case-list.txt
```

The [case-list.txt](File-Formats.md#case-lists) file should conform to the file format describe on the File Formats wiki page.

Note - the import-case-list command can take a path to a folder that contains a set of case-list files in addition to a single case-list file.
# Inter-Repository Dependencies

## cbioportal-core Repository

[cbioportal-core](https://github.com/cBioPortal/cbioportal-core) is dependent on
[cbioportal](https://github.com/cBioPortal/cbioportal).
This can be seen in its [pom.xml](https://github.com/cBioPortal/cbioportal-core/blob/main/pom.xml).
By removing the dependency and attempting to build cbioportal-core, errors are reported. This
document captures a summary of the reported errors. This analysis was done at the time of release
v6.0.0.

cbioportal-core now primarily provides import functionality, which is mainly located
in the packages:
- [org.mskcc.cbio.portal.dao](https://github.com/cBioPortal/cbioportal-core/tree/main/src/main/java/org/mskcc/cbio/portal/dao)
- [org.mskcc.cbio.portal.model](https://github.com/cBioPortal/cbioportal-core/tree/main/src/main/java/org/mskcc/cbio/portal/model)
- [org.mskcc.cbio.portal.scripts](https://github.com/cBioPortal/cbioportal-core/tree/main/src/main/java/org/mskcc/cbio/portal/scripts)

### Dependencies in cbioportal-core import functionality

Code in cbioportal which is referenced directly by code in cbioportal-core consists of three
enum class definitions.

- cbioportal class [ResourceType](https://github.com/cBioPortal/cbioportal/blob/master/src/main/java/org/cbioportal/model/ResourceType.java) is imported by cbioportal-core class [ResourceDefinition](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/model/ResourceDefinition.java). This provides enum constants {STUDY, SAMPLE, PATIENT}.
- cbioportal class [EntityType](https://github.com/cBioPortal/cbioportal/blob/master/src/main/java/org/cbioportal/model/EntityType.java) is imported by cbioportal-core class [ImportTabDelimData](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/scripts/ImportTabDelimData.java). This provides enum constants {GENE, GENESET, PHOSPHOPROTEIN, GENERIC_ASSAY}.
- cbioportal class [CNA](https://github.com/cBioPortal/cbioportal/blob/master/src/main/java/org/cbioportal/model/CNA.java) is imported by cbioportal-core class [CnaEvent](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/model/CnaEvent.java). This provides enum constants {AMP, GAIN, DIPLOID, HETLOSS, HOMDEL}.

The cbioportal-core import functionality code also relies on several packaged libraries in cbioportal.

cbioportal packaged library org.apache.commons.commons-collections4 (version 4.4) is used by these cbioportal-core classes:
- [DaoPatient](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/dao/DaoPatient.java)
- [ImportClinicalData](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/scripts/ImportClinicalData.java)

cbioportal packaged library org.slf4j.slf4j-api (latest version) is used by these cbioportal-core classes:
- [JdbcUtil](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/dao/JdbcUtil.java)
- [ProgressMonitor](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/ProgressMonitor.java)
- [Patient](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/model/Patient.java)
- [AccessControlImpl](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/internal/AccessControlImpl.java)

cbioportal packaged library org.springframework.security.spring-security.core (latest compatible version)
and other related such as org.springframework.security.spring-security.web are used in these cbioportal-core classes:
- [AccessControl](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/AccessControl.java)
- [DaoCancerStudy](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java) (transitive dependency through AccessControl)

### Other dependencies in cbioportal-core

There are other (non-import) cbioportal-core uses of libraries packaged in cbioportal.

cbioportal packaged libraries jakarta.servlet jakarta.servlet.http (ServletConfig, HttpServlet, ServletException, HttpServletRequest, HttpServletResponse} are used by these cbioportal-core classes:
- [GetSurvivalDataJSON](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/GetSurvivalDataJSON.java)
- [XDebug](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/XDebug.java)
- [CalcCoExp](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/CalcCoExp.java)
- [BioGeneServlet](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/BioGeneServlet.java)
- [IGVLinkingJSON](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/IGVLinkingJSON.java)

cbioportal packaged library org.json.simple / com.googlecode.json-simple.json-simple (version 1.1.1) is used by these cbioportal-core classes:
- [CalcCoExp](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/CalcCoExp.java)
- [GetSurvivalDataJSON](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/GetSurvivalDataJSON.java)
- [IGVLinkingJSON](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/IGVLinkingJSON.java)
- [GetClinicalData](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/web_api/GetClinicalData.java)

cbioportal packaged library org.apache.commonsorg.apache.commons.commons-math3 (version 3.6.1) is used by this cbioportal-core class:
- [CalcCoExp](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/CalcCoExp.java)

cbioportal packaged library org.slf4j.slf4j-api (latest version) is used by this cbioportal-core class:
- [AccessControlImpl](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/internal/AccessControlImpl.java) (transitive dependency through many servlets)

cbioportal packaged library org.springframework.security.spring-security.core (latest compatible version)
and other related such as org.springframework.security.spring-security.web are used in these cbioportal-core classes:
- [AccessControl](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/AccessControl.java) (transitive dependency through many servlets)

### Consequences
When any of these dependencies are updated in cbioportal (or in the case of libraries, updated in a non-backwards-compatible way), the cbioportal-core repository should be built and tested using the updated dependencies and proper behavior of the import functionality should be tested. cbioportal-core should also be brought up to date by packaging the current version of cbioportal into cbioportal-core (in pom.xml).
# Migration Guide

This page describes various changes deployers will need to make as they deploy newer versions of the portal.

## v5 -> v6

- Override Spring Application Properties:
  - The process of overriding properties has been updated for example previously users could do the following `java -Xms2g -Xmx4g -Dauthenticate=saml -jar webapp-runner.jar`.
  Now users must pass Spring Application properties after the `.jar` and replace `-Dauthenticate=false` with `--authenticate=false`. 
  For example `java -Xms2g -Xmx4g -jar cbioportal/target/cbioportal-exec.jar --spring.config.location=cbioportal/application.properties --authenticate=false`
- `portal.properties` migration needed:
  - `portal.properties` has been renamed to `application.properties`. This is the Spring Boot default name 
  - `authenticate` values of `googleplus`, `social_auth_google` and `social_auth_microsoft` have been replaced by `optional_oauth2`
    - If you used this property before without authorization (unlikely, only the public cBioPortal instance uses this), add the property `always_show_study_group=PUBLIC` and confirm  that all studies in your database you'd like to be be public have `GROUPS` values set to `PUBLIC`
- `Redis HTTP Session`
  - To disable redis session (Must be disabled if redis is not setup) `spring.autoconfigure.exclude=org.springframework.boot.autoconfigure.data.redis.RedisAutoConfiguration`
  - To enable `spring.data.redis.host=localhost` `spring.data.redis.port=6379`
- `Security` changes:
  - Properties used to configure Security have changed see [security.properties.EXAMPLE](deployment/customization/security.properties-Reference.md)
  - `SAML` changes:
    - need to generate new keys (`.jks` no longer works)
       - if you have an existing client, update saml keys with newly generate keys
    - if you have an existing client, update redirect_uri to be without port 443
    - You can now use a URL here, instead of metadata XML: `spring.security.saml2.relyingparty.registration.cbio-saml-idp.assertingparty.metadata-uri`

## v5.3 -> v5.4

- Remove `db.host` and `db.portal_db_name` and `db.use_ssl` properties from the _portal.properties_ file or JVM
  parameters. Update property `db.connection_string` to encode the hostname, port, database and other parameters
  according to [Database Settings](deployment/customization/portal.properties-Reference.md#Database-Settings) documentation and pass via
  _portal.properties_ file or as JVM parameter.

## v4 -> v5

- All fusion profiles are now required to be migrated to structural variant format. One can use
  this [migration tool](https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/fusion-to-sv-converter)
  to migrate the fusion files.
- All fusion files on [datahub](https://github.com/cBioPortal/datahub) were migrated to the structural variant format
  and their molecular profile ids were renamed from `{study_id}_fusion` to `{study_id}_structural_variants`. If you are
  using these datahub files one would need to re-import them.
- Study view user setting will be outdated after migration, please follow `Clear Study View User settings` section
  in [Session Service Management](Session-Service-Management.md#Clear-Study-View-User-settings)
- The new default set of transcripts for each gene has changed from `uniprot` to `mskcc`. See
  the [Mutation Data Annotation Section](./mutation-data-transcript-annotation.md) for more details. To keep the old set
  of default transcripts add the property `genomenexus.isoform_override_source=uniprot`
  to [the properties file](https://docs.cbioportal.org/deployment/customization/portal.properties-reference/#properties).

See the [v5.0.0 release notes](https://github.com/cBioPortal/cbioportal/releases/tag/v5.0.0) for more details.

## v3 -> v4

- Introduces `logback` package for logging. If you don't have any custom log4j.properties file, no changes are necessary
- Cleans up several old databases ([PR](https://github.com/cBioPortal/cbioportal/pull/9360)). In theory the migration
  should be seamless, since the docker container detects an old database version and migrates it automatically.

See the [v4.0.0 release notes](https://github.com/cBioPortal/cbioportal/releases/tag/v4.0.0) for more details.

## v2 -> v3

- Session service is now required to be set up. You can't run it without session service. The recommended way to run
  cBioPortal is to use the Docker Compose instructions.

## v1 -> v2

- Changes cBioPortal to a Single Page App (SPA) written in React, Mobx and bootstrap that uses a REST API. It shouldn't
  change anything for a deployer.
# MSK Maintenance (In Progress)
We provide our cBioPortal's maintenance page publicly in the hope that it may be useful to others

## Database Migration
In the database migration process, we are going to have two main steps: building importers and updating database scheme.

### Building Importers
- Codebases:
    - [cbioportal](https://github.com/cBioPortal/cbioportal)
    - [genome nexus annotation pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline)
    - [pipelines](https://github.com/knowledgesystems/pipelines)
- Building single importer:
    - Take the cbioportal hash and add it to genome nexus annotation pipeline POM -> push to fork
    - Take the cbioportal hash and new genome nexus annotation pipeline hash (from above) and propagate to pipelines POM; also update db version if db migration is part of change
    - Login to pipelines server
    - Sets general env variables needed for building
        - Run `sh /data/portal-cron/scripts/automation-environment.sh`
    - Specific to cbioportal build, looks for properties and xml files off of $PORTAL_HOME
        - Run `export PORTAL_HOME=/data/portal-cron/git-repos/cbioportal`
    - Make sure the local cbioportal codebase is on the correct git hash at `/data/portal-cron/git-repos/cbioportal`
    - Specify importer to be build
        - Copy importer properties to `/data/portal-cron/git-repos/cbioportal/src/main/resources` (e.g. triage `/data/portal-cron/git-repos/pipelines-configuration/properties/import-triage/*`)
    - Navigate to pipelines folder
        - `cd /data/portal-cron/git-repos/pipelines`
    - Build importer
        - mvn clean install -DskipTests
- Build multiple importers
    - Take the cbioportal hash and add it to genome nexus annotation pipeline POM -> push to fork
    - Take the cbioportal hash and new genome nexus annotation pipeline hash (from above) and propagate to pipelines POM; also update db version if db migration is part of change
    - Login to pipelines server
    - Sets general env variables needed for building
        - Run `sh /data/portal-cron/scripts/automation-environment.sh`
    - Specific to cbioportal build, looks for properties and xml files off of $PORTAL_HOME
        - Run `export PORTAL_HOME=/data/portal-cron/git-repos/cbioportal`
    - Make sure the local cbioportal codebase is on the correct git hash at `/data/portal-cron/git-repos/cbioportal`
    - Go to build importer jars folder
        - `cd /data/portal-cron/git-repo/pipelines-configuration/build-importer-jars`
    - (Optional) Remove existing jars
        - `rm *.jar`
    - Build all importers at once (build all importers except cmo-pipelines)
        - `sh buildproductionjars.sh -sd=true -sgp=true -b=importers`
        - Available parameters:
          - [--cbioportal-git-hash|-cgh=<cbioportal_commit_hash>]
          - [--skip-deployment|-sd=<true|false>]
          - [--skip-git-pull|-sgp=<true|false>]
          - [--build|-b=<build_specifier>], build_specifier should be one of the following
            - all (build for all artifacts)
            - importers (all importers except cmo-pipelines)
            - cmo-pipelines (cmo-pipelines artifacts only)
            - triage-cmo-importer
            - msk-dmp-importer
            - msk-cmo-importer
            - public-importer
            - genie-aws-importer
            - genie-archive-importer
            - hgnc-importer

### Updating Database Scheme
Database needs to be updated one by one, we have four main databases: triage, private, genie, and public. Take triage database as an example.
- Migrate one database (e.g. triage)
    - SSH into pipeline server
    - Checkout to the commit that contains the latest database scheme
    - Check if property sets up correctly to the right database (triage)
        - `vi /data/portal-cron/git-repos/cbioportal/src/main/resources/application.properties`
    - Move to directory
        - `cd /data/portal-cron/git-repos/cbioportal`
    - Run database migration using script:
        - `python3 core/src/main/scripts/migrate_db.py --properties-file src/main/resources/application.properties --sql db-scripts/src/main/resources/migration.sql`
    - Monitor the DB migration process and look for possible errors
    - Access database and verify the DB scheme is updated
This document describes how each mutation in cBioPortal gets annotated with a specific gene symbol + protein change. 

### Biological Background 

This section explains the concepts of protein isoforms and transcripts. 

#### What is an isoform? 

From a single gene (string of nucleotides) multiple protein sequences can be formed (string of amino acids). For 
example: parts of the gene that code for proteins (exons) can be included or excluded through a process known as [alternative splicing](https://en.wikipedia.org/wiki/Alternative_splicing). Each of the different resulting proteins is called an [isoform](https://en.wikipedia.org/wiki/Protein_isoform). A single mutation can impact the isoforms differently. E.g. in one isoform
it might change a P to a T, but in the other isoform that particular exon does not get included and it is therefore not 
changing the amino acid sequence at all. In cBioPortal for convenience sake we assign a single gene symbol + protein change to each mutation. For most cases this works well because there is only one protein isoform relevant in a clinical setting. There are of course exceptions and we are 
therefore working on improving this representation. An explanation of the relation between transcripts and protein 
isoforms can be found in the next section. 

#### What is a transcript? 

DNA is transcribed to a pre-mRNA transcript which includes intron and exon regions. Splicing and other processes then take place to form the resulting mature mRNA transcript that only contains exons, which subsequently can be translated to a protein sequence. An mRNA transcript can thus be associated with a specific protein isoform. The Ensembl database assigns ids for these transcript with names like `ENSTxxx`. You can see this on e.g. the Ensembl 
website for the [BRAF gene](https://grch37.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000157764;r=7:140719327-140924929#): 

<img width="995" alt="Screen Shot 2022-10-05 at 9 35 55 AM" src="https://user-images.githubusercontent.com/1334004/194073821-9a43cab2-3d31-40ab-b47e-517f1ce8bca3.png">

The transcript `ENST00000288602.6` is 2480 base pairs long (nucleotides ACGT) and the associated protein isoform is 766 
amino acids (V/P/etc). You can see we are showing that same transcript and protein isoform on [cBioPortal](https://bit.ly/3vaclXN): 

<img width="1052" alt="Screen Shot 2022-10-05 at 9 37 32 AM" src="https://user-images.githubusercontent.com/1334004/194074248-c3e205b4-c2e4-4e19-a541-17517dc07913.png"> 

For each gene name in cBioPortal a canonical/default transcript is assigned. These assignments are stored in [Genome 
Nexus](https://www.genomenexus.org/) and explained below. Although cBioPortal does not store changes to different 
transcripts/isoforms for each mutation in the database itself, it does allow viewing them on the [Mutations Tab](https://bit.ly/39hVtDd) by re-annotating the mutations on the fly through [Genome Nexus](https://www.genomenexus.org/) whenever a user clicks on the transcript dropdown. 

### Transcript Assignment 

The cBioPortal database stores one gene + protein change annotation for each mutation event in the database. To allow 
comparing mutation data across studies it is important to annotate the mutation data (be it in MAF or VCF format) in 
the same way, otherwise the gene + protein changes can mean entirely different things. For all public studies stored 
in [datahub](https://github.com/cBioPortal/datahub/tree/master/public) we leverage [Genome Nexus](https://www.genomenexus.org) to do so. 
Genome Nexus assigns one canonical Ensembl Transcript + gene name + protein change for 
each mutation. You can find the mapping of hugo symbol to transcript id [here](https://github.com/genome-nexus/genome-nexus-importer/blob/master/data/grch37_ensembl92/export/ensembl_biomart_canonical_transcripts_per_hgnc.txt). There are 
two sets of default transcripts: `uniprot` and `mskcc`. We recommend to use the `mskcc` set of transcripts when 
starting from scratch, since these are more up to date and correspond to transcripts that were chosen as relevant for 
clinical sequencing at MSKCC. The `uniprot` set of transcripts was constructed several years ago, but we are no longer 
certain about the logic on how to reconstruct them hence they are not being kept up to date. One can see the 
differences between the two in [this file](https://github.com/cBioPortal/cbioportal-frontend/files/9498680/genes_with_different_uniprot_mskcc_isoforms.txt). For the public cBioPortal (https:
//www.cbioportal.org) and [datahub](https://github.com/cBioPortal/datahub/tree/master/public) we are using `mskcc`, for the GENIE cBioPortal (https://genie.cbioportal.org) we still use `uniprot`. As of cBioPortal v5 the default is `mskcc` for local installations. Prior to v5 it was `uniprot`. We recommend that people upgrading to v5 consider migrating to `mskcc` as well (see [migration guide](https://docs.cbioportal.org/migration-guide/) and [the properties reference docs](https://docs.cbioportal.org/deployment/customization/application.properties-reference/#properties)).

#### How default transcript assignment affects the Mutations Tab 

The [Mutations Tab](https://bit.ly/39hVtDd) shows the full protein sequence. The one shown by default is the canonical 
transcript (`mskcc` or `uniprot` depending on configuration). The mutations are drawn on the lollipop based on the 
protein position found in the cBioPortal database. For the [public cBioPortal](https://cbioportal.org) all mutation 
data in MAF format are annotated using [Genome Nexus](https://www.genomenexus.org) to add the gene and protein change 
columns. This is then imported into the cBioPortal database. Whether you choose to use the set of `uniprot` or `mskcc` 
transcripts, make sure to indicate it in the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-
nexus-annotation-pipeline#maf-annotation)(`--isoform-override <mskcc or uniprot>`) when annotating as well as in [the properties file](https://docs.cbioportal.org/deployment/customization/application.properties-reference/#properties)
of cBioPortal. That way the [Mutations Tab](https://bit.ly/39hVtDd) will show the correct canonical 
transcript. Note that whenever somebody uses the dropdown on the Mutations Tab to change the displayed transcript, 
Genome Neuxs re-annotates all mutations on the fly. The browser sends over the genomic location (chrom,start,end,ref,
alt) to get the protein change information for each transcript. Since many of the annotations are for the canonical transcripts
only we are currently hiding annotations for non-canonical transcripts.

#### Plans for default transcripts 

We are planning to move to a single set of default transcripts over time. Prior to v5 `uniprot` was used for the public 
facing portals and local installations. Our plan is to use `mskcc` everywhere and eventually we will most likely move to [MANE](https://www.ensembl.org/info/genome/genebuild/mane.html). MANE is only 
available for grch38 and since most of our data is for grch37 this is currently not feasible. Whichever set of 
transcripts you choose to use, make sure to indicate so in the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline#maf-annotation) (`--isoform-override <mskcc or uniprot>`) and put the same 
set of transcripts in [the properties file](https://docs.cbioportal.org/deployment/customization/application.properties-reference/#properties) of cBioPortal, such that the [Mutations Tab](https://bit.ly/39hVtDd) will show the correct canonical transcript (currently defaults to `mskcc`). The re-annotation of mutations only happens once a user clicks to change the transcript, which is why it's important that the protein change in the database is for the specific transcript displayed first. 
# Genie News
## April 3, 2024
*   **Public Release 15.1 of AACR Project GENIE:**
    *  The public release 15.1 version of AACR GENIE has 61 samples retracted that were present in AACR GENIE 15.0-public. 
    *  More detailed information can be found in the [AACR GENIE release notes](https://www.synapse.org/#!Synapse:syn55234828) and the data releases page from [Sage Bionetworks](https://www.synapse.org/#!Synapse:syn55234548).
## January 9, 2024
*   **Public Release 15.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 15.0-public now contains over 198,000 sequenced samples from over 172,000 patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.synapse.org/#!Synapse:syn7222066/wiki/409237).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](https://www.synapse.org/#!Synapse:syn50898929).
## November 9, 2023
*   **Public Release 14.1 of AACR Project GENIE:**
    *  The public release 14.1 version of AACR GENIE has 10 samples retracted that were present in AACR GENIE 14.0-public. 
    *  More detailed information can be found in the [AACR GENIE release notes](https://www.synapse.org/#!Synapse:syn52919317) and the data releases page from [Sage Bionetworks](https://www.synapse.org/#!Synapse:syn7222066/wiki/410924)
## September 20, 2023
*   **Public Release 14.0 of AACR Project GENIE:**
     *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 14.0-public now contains 183,000 sequenced samples from nearly 160,000 patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets released to date.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.synapse.org/#!Synapse:syn7222066/wiki/409237).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](https://www.synapse.org/#!Synapse:syn50898929).
## May 1, 2023
*   **Public Release 13.1 of AACR Project GENIE:**
    *  The public release 13.1 version of AACR GENIE has 65 samples retracted that were present in AACR GENIE 13.0-public. 
    *  More detailed information can be found in the [AACR GENIE release notes](https://www.synapse.org/#!Synapse:syn51355999) and the data releases page from [Sage Bionetworks](https://www.synapse.org/#!Synapse:syn7222066/wiki/410924)
## January 9, 2023
*   **Public Release 13.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 13.0-public now contains more than 167,000 sequenced samples from over 148,000 patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets released to date.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.synapse.org/#!Synapse:syn21683345).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](https://www.synapse.org/#!Synapse:syn50898929).
## November 3, 2022
*   **Public Release BPC CRC 2.0-PUBLIC**
    *   The GENIE BPC CRC v2.0-public dataset contains 1,485 CRC patients from three institutions: MSKCC, DFCI, and VICC.‚ÄØ
    *   The complete, post-processed data are available on [Synapse](https://www.synapse.org/#!Synapse:syn30102262)
## July 6, 2022
*   **Public Release GENIE ERBB2 Cohort**
    * The study contains 315 samples from 135 patients from 6 institues.
## July 22, 2022
*   **Public Release 12.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 12.0-public, was released in July 2022.The registry now contains more than 154,000 sequenced samples from 137,000+ patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets released to date.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.aacr.org/wp-content/uploads/2022/07/GENIE_12.0-public_data_guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
## May 19, 2022
*   **Public Release BPC NSCLC 2.0-PUBLIC**
    *   The GENIE BPC NSCLC v2.0-public dataset contains 1,846 NSCLC patients from 4 institutions: MSKCC, DFCI, VICC and UHN.
    *   The complete, post-processed data are available on [Synapse](https://www.synapse.org/#!Synapse:syn27056172/wiki/616601)
## January 7, 2022
*   **Public Release 11.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 11.0-public, was released in January 2022. The registry now contains over 136,000 sequenced samples from over 121,000 patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets released to date.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.aacr.org/wp-content/uploads/2022/02/GENIE_data_guide_11.0-public-1.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
## June 22, 2021
*   **Public Release 10.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The tenth data set, GENIE 10.0-public, was released in June 2021. With the most recent data release, the registry now contains genomic information from 120953 samples.  
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
## February 8, 2021
*   **Public Release 9.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The ninth data set, GENIE 9.0-public, was released in February 2021. With the most recent data release, the registry now contains genomic information from more nearly 17,000 non-small cell lung carcinomas, and nearly 12,000 breast and more than 11,000 colorectal cancers.  
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
## July 7, 2020
*   **Public Release 8.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The eighth data set, GENIE 8.0-public, was released in July 2020. The combined data set now includes nearly 96,000 de-identified genomic records collected from patients who were treated at each of the consortium's 19 participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for over 80 major cancer types, including data from greater than 14,000 patients with lung cancer, nearly 12,000 patients with breast cancer, and nearly 9,500 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).
## January 29, 2020
*   **Public Release 7.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The seventh data set, GENIE 7.0-public, was released in January 2020. The combined data set now includes nearly 79720 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for over 80 major cancer types.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).
## July 13, 2019
*   **Public Release 6.1 of AACR Project GENIE:**
    *   The sixth data set, GENIE 6.0-public, was released in early July 2019.  A patch to GENIE 6.0-public, GENIE 6.1-pubic, was subsequently released on July 13, 2019.  The combined data set now includes nearly 70,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for nearly 80 major cancer types, including data from nearly 11,000 patients with lung cancer, greater than  9,700 patients with breast cancer, and nearly 7,000 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](https://www.aacr.org/wp-content/uploads/2020/02/20200127_GENIE_Data_Guide_7.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](https://www.aacr.org/professionals/research/aacr-project-genie/).
## July 08, 2019
*   **Public Release 6.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The sixth data set, GENIE 6.0-public, was released in July 2019 adding more than 11,000 records to the database. The combined data set now includes nearly 71,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for over 80 major cancer types, including data from greater than 11,000 patients with lung cancer, nearly 9,800 patients with breast cancer, and more than 7,000 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## January 11, 2019
*   **Public Release 5.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The forth data set, GENIE 4.0-public, was released in July 2018 adding more than 7,800 records to the database. The combined data set now includes more than 59,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date.  This data will be released to the public every six months.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## July 16, 2018
*   **Public Release 4.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The forth data set, GENIE 4.0-public, was released in July 2018 adding more than 7,800 records to the database. The combined data set now includes more than 47,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date.  This data will be released to the public every six months. The public release of the fifth data set, GENIE 5.0-public, will take place in January, 2019.
    *   The combined data set now includes data for over 80 major cancer types, including data from greater than 7,500 patients with lung cancer, nearly 5,500 patients with breast cancer, and more than 5,100 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## January 2018
*   **Added 7500 samples to the GENIE Public Cohort** The combined dataset now includes samples from over 60 major cancer types including:
    *   6,000 lung cancer samples.
    *   4,500 breast cancer samples.
    *   4,300 colorectal cancer samples.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## November 20, 2017
*   **Public Release 2.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The second dataset was released in November 2017, adding more than 13,000 records to the database.  The combined data set now includes over 32,000 de-identified genomic records collected from patients who were treated at each of the consortium‚Äôs participating institutions, making it among the largest fully public cancer genomic data sets released to date. These data will be continuously updated on a quarterly basis.
    *   The combined data set now includes data for 59 major cancer types, including data on nearly 5,000 patients with lung cancer, nearly 4,000 patients with breast cancer, and more than 3,500 patients with colorectal cancer. For more details about the data, and how to use it, consult the [data guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).
    *   In addition to accessing the data on the AACR Project GENIE cBioPortal website, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
    *   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## January 5, 2017
*   **Initial public release of AACR Project GENIE:**
    *   Somatic alterations in 18,980 tumor samples from 18,500 patients sequenced at eight different institutions.
    *   Data is available for download from [Sage Bionetworks](http://synapse.org/genie).

## AACR Project GENIE cBioPortal Terms of Use
*   I will not attempt to identify or contact individual participants from whom these data were collected by any means.
*   I will not redistribute the data without express written permission from the AACR Project GENIE Coordinating Center (info at aacrgenie dot org).
*   When publishing or presenting work using or referencing the AACR Project GENIE dataset please include the following attributions:
    *  Please cite: AACR Project GENIE Consortium. AACR Project GENIE: Powering Precision Medicine Through an International Consortium, Cancer Discov. 2017 Aug;7(8):818-831 and include the version of the dataset used.
    *  Include the following acknowledgement: The authors would like to acknowledge the American Association for Cancer Research and its financial and material support in the development of the AACR Project GENIE registry, as well as members of the consortium for their commitment to data sharing. Interpretations are the responsibility of study authors.
*   Should you decide at any point in the future to stop using the AACR Project GENIE cBioPortal, send request: genie-cbioportal-access at cbio dot mskcc dot org and we will remove your user id and any other information provided during new user registration from our systems.
*   For more information see [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf) on the AACR website.
## May 29, 2024
*   **Added data** consisting of 9,632 samples from 10 studies:
	* [Soft Tissue and Bone Sarcoma (MSK, Nat Commun 2022)](https://www.cbioportal.org/study/summary?id=sarcoma_msk_2022) *7494 samples*
    * [Chronic Lymphocytic Leukemia (Broad, Nature Genetics 2022)](https://www.cbioportal.org/study/summary?id=cll_broad_2022) *1154 samples*
    * [Mature B-Cell Neoplasms (Simon Fraser University, Blood 2023)](https://www.cbioportal.org/study/summary?id=mbn_sfu_2023) *297 samples*
    * [Anaplastic Thyroid Cancers (GATCI, Cell Reports 2024)](https://www.cbioportal.org/study/summary?id=thyroid_gatci_2024) *190 samples*
    * [Colorectal Cancer (CAS Shanghai, Cancer Cell 2020)](https://www.cbioportal.org/study/summary?id=coadread_cass_2020) *146 samples*
    * [Prostate Cancer MDA PCa PDX (MD Anderson, Clin Cancer Res 2024)](https://www.cbioportal.org/study/summary?id=prad_msk_mdanderson_2023) *88 samples*
    * [Ovarian Cancer - MSK SPECTRUM (MSK, Nature 2022)](https://www.cbioportal.org/study/summary?id=msk_spectrum_tme_2022) *82 samples*
    * [IDH-mutated Diffuse Glioma (MSK, Clin Cancer Res 2024)](https://www.cbioportal.org/study/summary?id=difg_msk_2023) *73 samples*
    * [Pre-cancer Colorectal Polyps (HTAN Vanderbilt, Cell 2021)](https://www.cbioportal.org/study/summary?id=crc_hta11_htan_2021) *61 samples*
    * [Prostate Cancer (MSK, Science 2022)](https://www.cbioportal.org/study/summary?id=prad_organoids_msk_2022) *47 samples*

*   **Data Improvement**

    * Pan-can studies Ancestry addition: Genetic Ancestry data is added to all 32 TCGA Pan-Can studies,in generic assay format. Data source: [GDC](https://gdc.cancer.gov/about-data/publications/CCG-AIM-2020). [Example: Ancestry-associated somatic genetic alterations](https://www.cbioportal.org/comparison/alterations?comparisonId=664275e383e9543d61907074&patientEnrichments=true&selectedEnrichmentEventTypes=%5B%22missense%22%2C%22missense_mutation%22%2C%22missense_variant%22%2C%22indel%22%2C%22nonframeshift%22%2C%22inframe%22%2C%22inframe_del%22%2C%22inframe_deletion%22%2C%22in_frame_del%22%2C%22in_frame_deletion%22%2C%22nonframeshift_deletion%22%2C%22inframe_ins%22%2C%22inframe_insertion%22%2C%22in_frame_ins%22%2C%22in_frame_insertion%22%2C%22nonframeshift_insertion%22%2C%22truncating%22%2C%22feature_truncation%22%2C%22nonsense_mutation%22%2C%22nonsense%22%2C%22stopgain_snv%22%2C%22stop_gained%22%2C%22frameshift%22%2C%22frameshift_variant%22%2C%22frame_shift_del%22%2C%22frameshift_deletion%22%2C%22frame_shift_ins%22%2C%22frameshift_insertion%22%2C%22translation_start_site%22%2C%22initiator_codon_variant%22%2C%22start_codon_snp%22%2C%22start_codon_del%22%2C%22de_novo_start_outofframe%22%2C%22nonstop_mutation%22%2C%22stop_lost%22%2C%22splice_site%22%2C%22splice%22%2C%22splicing%22%2C%22splice_site_snp%22%2C%22splice_site_del%22%2C%22splice_site_indel%22%2C%22splice_region_variant%22%2C%22splice_region%22%2C%22silent%22%2C%22synonymous_variant%22%2C%22targeted_region%22%2C%22other%22%2C%22AMP%22%2C%22HOMDEL%22%2C%22structural_variant%22%5D)
 <img width="1006" alt="Screen Shot 2024-05-20 at 10 52 11 AM" src="https://github.com/cBioPortal/datahub/assets/61431648/628b7eed-9d72-4605-b114-2e438dbd442a"> 

## May 7, 2024
* **New Feature**: Gene-specific charts in Study View now support mutation data. Two chart types are available: sample mutational status and mutation types.

  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/d94c596a-1103-4aba-995a-ed1709236eee)
  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/5fd2fdd0-07a5-434a-ad42-3f72b7c67bc5)
  
* **New Feature**: Categorical data loaded into the generic assay format is now available in group comparison. [Example: Arm-level CNA compared between astrocytoma vs oligodendroglioma in Brain Lower Grade Glioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/comparison/generic_assay_armlevel_cna?comparisonId=616ff06a2e3200072db98b99)


  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/9f8c99a2-eef8-4b5a-ae33-4a2b295e0514)

* **Enhancement**: OncoPrint now allows you to remove a gene track.

  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/23cd422b-6a4e-4eff-9c04-ac88220591d2)

## Apr 2, 2024
* **Introducing the cBioPortal Newsletter!** Stay updated with the latest developments, insights, and community highlights of cBioPortal. Subscribe via [LinkedIn](https://www.linkedin.com/pulse/april-update-hello-cbioportal-community-cbioportal-22vle/) or [google groups](https://groups.google.com/g/cbioportal-news). We'll be sharing valuable updates every few months.

## March 27, 2024
* **New Feature**: The Plots tab is now available in the Study View. After exploring a cohort and applying any filters of interest in the Study View Summary page, you can now click over to the new Plots tab to explore the cohort by plotting any two attributes against each other. [Example: Mutation count vs subtype in Uterine Corpus Endometrial Carcinoma (TCGA, Nature 2013)](https://www.cbioportal.org/study/plots?id=ucec_tcga_pub&plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22SUBTYPE%22%7D&plots_vert_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22MUTATION_COUNT%22%2C%22logScale%22%3A%22false%22%7D&plots_coloring_selection=%7B%7D)

  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/cf74a4ff-57c6-4d71-86cb-4f7440fc3db6)

* **Enhancement**: The custom selection feature in Study View, which filters to a user-defined list of samples/patients, no longer requires the inclusion of the study ID with each sample ID. If viewing multiple studies in study view and filtering to a sample ID which exists in more than one study, all samples will be displayed by default or the study ID can be included for additional specificity.

  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/2143593d-5fd3-4b1d-b8a4-73a68e027e27)
  
## Mar 5, 2024
*   **Local Installations Feature**: Make representation of custom driver annotation configurable:
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/f67fd82f-1789-4427-9ca1-f17fc70c1a09)

## Feb 7, 2024
* **New feature**: The _Datasets Page_ now lists what studies have samples profiled for Structural Variants:
    
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/0e4e73e4-d9d8-417d-912c-710163480905)

* **New Major Release**: **v6.0.0** includes a major repackaging of the backend software to improve the development experience. The backend has been upgraded to use Java Spring Boot v3.1.4. The CORE and MAF modules have been moved to new repositories. All other modules have been compacted into a single source repository. We have also updated the JVM to 21 and many libraries have been updated to address security and performance issues. See more information in the [v6.0.0 release notes](https://github.com/cBioPortal/cbioportal/releases/tag/v6.0.0).

* **Local Installations Feature**: When mutational signature data is loaded, show COSMIC reference signatures on the _Patient View_:
    
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/10f73baa-d28c-4bb7-ad62-4751bd7f61cd)

## Dec 29, 2023
*   **Added data** consisting of 5,120 samples from 13 studies:
	* [Endometrial Cancer (MSK, Cancer Discovery 2023)](https://www.cbioportal.org/study/summary?id=ucec_ancestry_cds_msk_2023) *1882 samples*
    * [Esophagogastric Cancer (MSK, J Natl Cancer Inst 2023)](https://www.cbioportal.org/study/summary?id=egc_msk_2023) *902 samples*
    * [Diffuse Glioma (GLASS Consortium, 2022)](https://www.cbioportal.org/study/summary?id=difg_glass) *693 samples*
    * [Bladder Cancer (MSK, Clin Cancer Res 2023)](https://www.cbioportal.org/study/summary?id=bladder_msk_2023) *526 samples*
    * [Non-Small Cell Lung Cancer Brain Metastasis (MSK, Nat Commun 2023)](https://www.cbioportal.org/study/summary?id=bm_nsclc_mskcc_2023) *322 samples*
    * [MSK Make-an-IMPACT Rare Cancers (MSK, Clin Cancer Res 2023)](https://www.cbioportal.org/study/summary?id=makeanimpact_ccr_2023) *184 samples*
    * [Cervical Cancer (MSK, Clin Cancer Res 2023)](https://www.cbioportal.org/study/summary?id=cervix_msk_2023) *177 samples*
    * [Mature T and NK Neoplasms (MSK, Blood Adv 2023)](https://www.cbioportal.org/study/summary?id=mtnn_msk_2022) *132 samples*
    * [Hepatocellular Carcinoma (MSK, JCO Precis Oncol 2023)](https://www.cbioportal.org/study/summary?id=hcc_jcopo_msk_2023) *90 samples*
    * [Sarcoma (MSK, J Pathol 2023)](https://www.cbioportal.org/study/summary?id=sarcoma_msk_2023) *82 samples*
    * [Esophagogastric Cancer (MSK, Clin Cancer Res 2023)](https://www.cbioportal.org/study/summary?id=egc_trap_ccr_msk_2023) *64 samples*
    * [Rhabdomyosarcomas (MSK, NPJ Precis Oncol 2023)](https://www.cbioportal.org/study/summary?id=soft_tissue_msk_2023) *42 samples*
    * [Pediatric Rhabdomyosarcomas (MSK, JCO Precis Oncol 2023)](https://www.cbioportal.org/study/summary?id=rms_msk_2023) *24 samples*
    
*   **Gene Tables Update:** Updated tables of genes (main and alias), based on [Oct 1, 2023 HGNC release](http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/). See seedDB release notes [here](https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database-schema-2131) for details.

## Dec 5, 2023
* **New feature**: Filter by data types on the homepage:
    
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/cf6d5486-bfd7-4f69-add0-442dbb4d354a)
* **New feature**: Change colors of tracks in Oncoprint:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/301b0efb-3063-44db-bc4c-86973e7c566a" width="700" />

* **New feature**: Show alteration frequencies per group in OncoPrint. [Example: RTK-RAS alterations in Smokers vs Never Smokers Lung Cancer Cases in MSK-IMPACT Clinical Sequencing Cohort](https://bit.ly/3PwKsDJ)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/f2a9576e-c06d-447a-8eac-f0292bd1f935" width="700" />

## Oct 17, 2023
* **New feature**: Survival charts with landmark events and hazard ratios. [Example: TP53 in Lung Cancer Cases in MSK-IMPACT Clinical Sequencing Cohort](https://bit.ly/48QktxS)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/dc97c3ae-ebd7-4cd3-bb76-165517c9d2ba" width="700" />

## Oct 3, 2023
* **New feature**: New Structural Variants Tab on _Results View_. [Example: TMPRSS2 Structural Variants in MSK-IMPACT Clinical Sequencing Cohort](https://bit.ly/49FDhRk)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/141e2817-3300-4af2-a845-005403d75e74" width="700" />

## Sep 5, 2023
* **New feature**:  Add gene-specific CNA charts to show all levels of copy number alterations (including gain and hetloss) on _Study View_:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/873401ad-cbcf-4b07-9b80-8487af5abb18" width="700" />

## Aug 30, 2023
* **Enhancement**: Exclude a patient from your selection directly from the _Patient View_:
    
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/501fe998-8408-48e3-a183-e27ab3ff3f28)

## Aug 21, 2023
*   **Added data** consisting of 4,488 samples from 7 studies:
	* [Lung Adenocarcinoma Met Organotropism (MSK, Cancer Cell 2023)](https://www.cbioportal.org/study/summary?id=luad_mskcc_2023_met_organotropism) *2653 samples*
    * [Acute Myeloid Leukemia (OHSU, Cancer Cell 2022)](https://www.cbioportal.org/study/summary?id=aml_ohsu_2022) *942 samples*
    * [Colon Cancer (Sidra-LUMC AC-ICAM, Nat Med 2023)](https://www.cbioportal.org/study/summary?id=coad_silu_2022) *348 samples*
    * [Pediatric Neuroblastoma (MSK, Nat Genet 2023)](https://www.cbioportal.org/study/summary?id=nbl_msk_2023) *223 samples*    
    * [Colorectal Adenocarcinoma (MSK, Nat Commun 2022)](https://www.cbioportal.org/study/summary?id=bowel_colitis_msk_2022) *180 samples*
    * [Bladder Cancer (Columbia University/MSK, Cell 2018)](https://www.cbioportal.org/study/summary?id=bladder_columbia_msk_2018) *130 samples*
    * [Myoepithelial Carcinomas of Soft Tissue (WCM, CSH Molecular Case Studies 2022)](https://www.cbioportal.org/study/summary?id=stmyec_wcm_2022) *12 samples*


*   **Gene Tables Update:** Updated tables of genes (main and alias), based on [Apr 1, 2023 HGNC release](http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/). See seedDB release notes [here](https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database) for details.

## Aug 1, 2023
*   **Enhancement**: One-sided Fisher's exact tests were changed to be two-sided. The affected pages are:
	*    Results View Page - Mutual Exclusivity Tab
    *    Results View Page - Comparison Tab - Genomic Alterations Tab
    *    Comparison Page - Genomic Alterations Tab
    *    Comparison Page - Mutations Tab
    
    Please note that the Mutations tab on the Comparison page is a recent feature and was introduced with the two-sided Fisher's exact test already implemented.

    Several users pointed out that using a one-sided test was incorrect for these comparisons. Please see discussions [here](https://github.com/cBioPortal/cbioportal/issues/9943) for more information.

## Jul 18, 2023
* **New Feature**: Add mutations table and two-sided exact p-value to comparison. [Example: AR mutations in Primary vs Metastatic Prostate Cancer samples in MSK-IMPACT Clinical Sequencing Cohort](https://www.cbioportal.org/comparison/mutations?sessionId=5cf89323e4b0ab413787436c&groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D&selectedGene=AR)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/da0cd1ef-b9b4-49f0-a9d3-42cff0204dc2" width="700" />


## May 2, 2023
*   **New Feature**: The mutations tab now shows variant annotations from the [repository of Variant with Unexpected Effects (reVUE)](https://www.cancerrevue.org/).
    
    <img src="https://user-images.githubusercontent.com/16869603/234271379-6a0876ea-8477-426e-a0e4-5e0b0b66aef9.gif" width="700" />

## Apr 25, 2023
*   **Local Installations Feature**: When custom driver annotations for structural variants are loaded, one can now filter by them in both the Oncoprint and the _Study View_. 

## Apr 11, 2023
*   **New Feature**: Disable autocommit and manually commit filters in study view. Manually commit filters can improve cBioPortal performance when query large dataset.

    <img src="https://user-images.githubusercontent.com/3323006/221857599-1eb6e9a5-55c7-4fb5-bc14-283699be0990.png" width="700" />
    <img src="https://user-images.githubusercontent.com/3323006/221858244-bf359850-79b5-4f9c-8008-6b5271721854.png" width="700" />

## Apr 5, 2023
*   **Added data** consisting of 2,472 samples from 5 studies:
    * [Bladder Cancer (MSK, Cell Reports 2022)](https://www.cbioportal.org/study/summary?id=paired_bladder_2022) *1659 samples*
    * [Gastrointestinal Stromal Tumor (MSK, NPJ Precis Oncol 2023)](https://www.cbioportal.org/study/summary?id=gist_msk_2022) *499 samples*
    * [Appendiceal Cancer (MSK, J Clin Oncol 2022)](https://www.cbioportal.org/study/summary?id=appendiceal_msk_2022) *273 samples*
    * [Colorectal Cancer (MSK, Cancer Discovery 2022)](https://www.cbioportal.org/study/summary?id=coadread_mskresistance_2022) *22 samples*
    * [Nerve Sheath Tumors (Johns Hopkins, Sci Data 2020)](https://www.cbioportal.org/study/summary?id=nst_nfosi_ntap) *19 samples* [**First GRCh38 Study**]

*   **Data Improvement**
   	* Added TERT promoter mutation status to [Melanomas (TCGA, Cell 2015)](https://www.cbioportal.org/study/summary?id=skcm_tcga_pub_2015), [Papillary Thyroid Carcinoma (TCGA, Cell 2014)](https://www.cbioportal.org/study/summary?id=thca_tcga_pub) TCGA studies.

## Apr 4, 2023
*   **New Feature**: Allow numeric data type for custom data charts.
    
    <img src="https://user-images.githubusercontent.com/33220871/222459360-9c251f7f-1005-442f-811d-dd1ee1103161.png" width="700" />
    <img src="https://user-images.githubusercontent.com/33220871/222460439-ac7007c8-059a-4ecc-816c-90752adff444.png" width="700" />

    This also allows to have numerical custom data after we query based on genes (custom data 2 in the image):
    
    <img src="https://user-images.githubusercontent.com/33220871/223729410-03a094f4-7671-453f-b8b8-f7bc10b61c7c.png" width="700" />
    <img src="https://user-images.githubusercontent.com/33220871/223729522-fd9e19e2-6218-40d4-b9fb-f253c4faa858.png" width="700" />

## Jan 10, 2023
*   **New Feature**: New Pathways tab on the Group Comparison view. [Example: Primary vs Metastasis samples in MSK-IMPACT Clinical Sequencing Cohort](https://www.cbioportal.org/comparison/pathways?comparisonId=61791fa92e3200072db9b64e&groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D)
    
    <img src="https://user-images.githubusercontent.com/15748980/235750129-661682f7-6593-4b7e-8873-ebba70eeae32.png" width="700" />

## Dec 13, 2022
*   **New Feature**: New Mutations tab on the Group Comparison view. [Example: Primary vs Metastasis samples in MSK-IMPACT Clinical Sequencing Cohort](https://www.cbioportal.org/comparison/mutations?comparisonId=61791fa92e3200072db9b64e&groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D)
    
    <img src="https://user-images.githubusercontent.com/15748980/235748831-78652ccd-b3ab-496f-9626-74ceb8c7230d.png" width="700" />

## Oct 12, 2022
*   **Added data** consisting of 1,459 samples from 10 studies:
	* [Hepatocellular Carcinoma (MERiC/Basel, Nat Commun. 2022)](https://www.cbioportal.org/study/summary?id=hcc_meric_2021) *122 samples*
    * [Prostate Cancer Brain Metastases (Bern, Nat Commun. 2022)](https://www.cbioportal.org/study/summary?id=prostate_pcbm_swiss_2019) *168 samples*
    * [Pan-Cancer MSK-IMPACT MET Validation Cohort (MSK 2022)](https://www.cbioportal.org/study/summary?id=mixed_impact_subset_2022) *69 samples*
    * [Endometrial Carcinoma cfDNA (MSK, Clin Cancer Res 2022)](https://www.cbioportal.org/study/summary?id=ucec_ccr_cfdna_msk_2022) *44 samples*
    * [Endometrial Carcinoma MSI (MSK, Clin Cancer Res 2022)](https://www.cbioportal.org/study/summary?id=ucec_ccr_msk_2022) *181 samples*
    * [Gallbladder Cancer (MSK, Clin Cancer Res, 2022)](https://www.cbioportal.org/study/summary?id=gbc_mskcc_2022) *244 samples*
    * [Meningioma (University of Toronto, Nature 2021)](https://www.cbioportal.org/study/summary?id=mng_utoronto_2021) *121 samples*
	* [Mixed Tumors: Selpercatinib RET Trial (MSK, Nat Commun. 2022)](https://www.cbioportal.org/study/summary?id=mixed_selpercatinib_2020) *188 samples*
	* [Low-Grade Serous Ovarian Cancer (MSK, Clin Cancer Res 2022)](https://www.cbioportal.org/study/summary?id=lgsoc_mapk_msk_2022) *119 samples*
	* [Urothelial Carcinoma (BCAN/HCRN 2022)](http://www.cbioportal.org/study/summary?id=blca_bcan_hcrn_2022) *203 samples*

## Sep 6, 2022
*   **Enhancement**: Oncoprint can now save clinical tracks after login:
    
    <img src="https://user-images.githubusercontent.com/3323006/188828765-ded10947-7343-49dd-b605-50b58b67a388.png" width="700" />

## Aug 11, 2022
*   **New Major Release**: **v5.0.0** release drops support for fusions in the mutation data format. Going forward fusions can only be imported in the Structural Variant (SV) format. This is mainly a refactoring effort to simplify the codebase and pave the way for the development of novel structural variant visualizations in the future. For cBioPortal instance maintainer, please reference our [Migration Guide](https://docs.cbioportal.org/migration-guide/#v4---v5) for instruction.
	
## Jul 26, 2022
*   **Added data** consisting of 6,631 samples from 7 studies:
	* [Metastatic Biliary Tract Cancers (SUMMIT - Neratinib Basket Trial, 2022)](https://www.cbioportal.org/study/summary?id=biliary_tract_summit_2022) *36 samples*
    * [Rectal Cancer (MSK, Nature Medicine 2022)](https://www.cbioportal.org/study/summary?id=rectal_msk_2022) *801 samples*
    * [Lung Adenocarcinoma (MSK Mind,Nature Cancer 2022)](https://www.cbioportal.org/study/summary?id=lung_msk_mind_2020) *247 samples*
    * [Myelodysplastic Syndromes (MDS IWG, IPSSM, NEJM Evidence 2022)](https://www.cbioportal.org/study/summary?id=mds_iwg_2022) *3,323 samples*
    * [Esophagogastric Cancer (MSK, Clin Cancer Res 2022)](https://www.cbioportal.org/study/summary?id=egc_msk_tp53_ccr_2022) *237 samples*
    * [Pan-cancer Analysis of Advanced and Metastatic Tumors (BCGSC, Nature Cancer 2020)](https://www.cbioportal.org/study/summary?id=pog570_bcgsc_2020) *570 samples*
    * [Prostate Adenocarcinoma (MSK, Clin Cancer Res. 2022)](https://www.cbioportal.org/study/summary?id=prad_pik3r1_msk_2021) *1,417 samples*

## Jun 7, 2022
* **New Feature**: Add heatmap to plot options on _Comparison Page_. [Example: Primary vs Metastatic Prostate Cancer in MSK-IMPACT (2017)](https://www.cbioportal.org/comparison/clinical?sessionId=5cf89323e4b0ab413787436c)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/1a048863-7db6-43c6-8250-c31039a2d67f" width="700" />

## May 31, 2022
* **New Feature**: Added Quartiles, Median split and Generate bins options for bar charts on the study view page, where Generate bins allows user to define bin size and min value
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/290e6d45-1c10-4a51-a823-24db27063486" width="700" />


## May 12, 2022
* **New Feature**: Show cohort alteration frequencies in pathways from [NDEx](https://www.ndexbio.org/iquery/) on the _Results View_. [Example: Glioblastoma signaling pathways in MSK-IMPACT (2017) cohort](https://bit.ly/3sE4UqD)

    <img src="https://user-images.githubusercontent.com/1334004/168684594-fafd1fd6-b2a9-45d7-ad3a-d5611f3f469c.png" width="700" />

## May 24, 2022
* **New Feature**: Add Help buttons on various pages and tabs, including the homepage

## May 10, 2022
* **New Feature**: Use IGV for the genomic overview on the _Patient View_. [Example: Endometrial cancer patient in TCGA](https://www.cbioportal.org/patient?studyId=ucec_tcga_pub&caseId=TCGA-BK-A0CC)

   <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/6fa6ab4b-4c09-4bd5-a7c2-3be2f418d83a" width="700" /> 

## May 5, 2022
* **New Feature**: View mutations and copy number changes in the [Integrative Genomics Viewer (IGV)](https://igv.org/) on the _Patient View_. [Example: Endometrial cancer patient in TCGA](https://www.cbioportal.org/patient?studyId=ucec_tcga_pub&caseId=TCGA-BK-A0CC)

    <img src="https://user-images.githubusercontent.com/1334004/168593020-253a4bfe-3922-4bf9-8dcc-89e298ba87c7.png" width="700" />
* **New Feature**: Add charts that plot categorical vs continuous data on the _Study View_. [Example: MSK-IMPACT (2017) cohort](https://www.cbioportal.org/study/summary?id=msk_impact_2017)

    <img src="https://user-images.githubusercontent.com/1334004/168595352-6416db7e-b6b5-4ef8-95e7-eaf67c841d94.png" width="700" />
* **New Feature**: Several single cell data integrations are now available for the [CPTAC glioblastoma study](https://www.cbioportal.org/study/summary?id=gbm_cptac_2021), allowing one to:
    *  Compare genomic alterations and cell type fractions in oncoprints on the _Results View_ ([Example](https://bit.ly/37QiC2x))
    *  Explore the single cell data further in [Vitessce](http://vitessce.io/) on the _Patient View_ ([Example](https://www.cbioportal.org/patient/customTab1?studyId=gbm_cptac_2021&caseId=C3N-02783))
    *  Create cohorts and groups based on cell type fractions on the _Study View_ ([Example](https://tinyurl.com/singlecellbioportal))
    *  Compare differences in cell type fractions between groups on the _Comparison Page_ ([Example](https://www.cbioportal.org/comparison/generic_assay_single_cell_type_fractions?comparisonId=6282d44d0934121b56df47fb&unselectedGroups=%5B%220.05-0.06%22%2C%220.03-0.05%22%5D))

    <img src="https://user-images.githubusercontent.com/1334004/168699333-a50ef433-5efe-45cc-a477-a3a99ed419dc.png" width="700" />

## Apr 20, 2022
*   **Added data** consisting of 2,557 samples from 5 studies:
    * [Breast Cancer (HTAN OHSU, 2022)](https://www.cbioportal.org/study/summary?id=brca_hta9_htan_2022) *5 samples*
    * [Colorectal Cancer (MSK, 2022)](https://www.cbioportal.org/study/summary?id=crc_dd_2022) *47 samples*
    * [Pediatric Pancan Tumors (MSK, 2022)](https://www.cbioportal.org/study/summary?id=mixed_kunga_msk_2022) *135 samples*
    * [Sarcoma (MSK, 2022)](https://www.cbioportal.org/study/summary?id=sarcoma_mskcc_2022) *2,138 samples*
    * [Lung Cancer in Never Smokers (NCI, Nature Genetics 2021)](https://www.cbioportal.org/study/summary?id=lung_nci_2022) *232 samples*
*   **Gene Tables Update** Updated tables of genes (main and alias), based on [Jan 1, 2022 HGNC release](http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/). See seedDB release note [here](https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database) for details.
*   **Data Improvement**
    * Pan-can studies timeline addition: `TREATMENT`, `OTHER MALIGNANCY FORM`, `SAMPLE ACQUISITION`, `STATUS` are added to all 32 TCGA Pan-Can studies. Details for data source and transformation process can be found [here](https://github.com/cBioPortal/datahub/pull/1597) or in the [README.md](https://github.com/cBioPortal/datahub/blob/master/public/acc_tcga_pan_can_atlas_2018/README.md) files included in each study folder on datahub. [Example: patient view of TCGA-A2-A04P in Breast Invasive Carcinoma Tumor Type](https://www.cbioportal.org/patient?studyId=brca_tcga_pan_can_atlas_2018&caseId=TCGA-A2-A04P)<img width="1204" alt="Screen Shot 2022-04-25 at 11 40 51 AM" src="https://user-images.githubusercontent.com/5973438/165153293-3d400e03-48d5-4919-ae32-7f16fa7d2f4f.png">

    * Pan-can studies methylation addition: methylation profile (27k and 450k merged) are added to all 32 TCGA Pan-Can studies, in generic assay format. Data source: [GDC](https://gdc.cancer.gov/node/977). Example: search by gene or probe from dropdown, to add a chart in study view, a track in Oncoprint (single study query only), or plots in plots tab. <img width="860" alt="Screen Shot 2022-03-22 at 9 52 20 PM" src="https://user-images.githubusercontent.com/5973438/159626976-72601dc9-e5be-4b50-bae9-dedb5be6e7fc.png"><img width="340" alt="Screen Shot 2022-03-22 at 9 52 27 PM" src="https://user-images.githubusercontent.com/5973438/159626985-2423c556-2180-4a00-b9b8-69b4d2e361ba.png"><img width="430" alt="Screen Shot 2022-03-22 at 9 53 38 PM" src="https://user-images.githubusercontent.com/5973438/159626987-e6b51bef-814e-4a76-826a-211fcdd5e9a1.png">

    * Single cell (type fraction and phases) data (in generic assay format) is added to [Glioblastoma (CPTAC, Cell 2021)](https://www.cbioportal.org/study/summary?id=gbm_cptac_2021)
*   **New Feature** For the new HTAN OHSU study there is now also an integration with [Harvard LSP](https://labsyspharm.org/)'s [Minerva](https://www.cycif.org/software/minerva) for exploring multiplex imaging:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/aa0f0c99-4b9c-481d-9f7f-891ea03fdd34" width="700" />


## Mar 1, 2022
* **New Documentation**:
    * Add a [User Guide](https://docs.cbioportal.org/user-guide/)
    * Add [Study View and Query How-to videos](https://docs.cbioportal.org/user-guide/overview/#how-to-videos)

## Feb 8, 2022
* **New Feature**: Create X vs Y violin plots in _Study View_ using any categorical and numerical clinical data:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/94d7dd21-4415-405d-8e9b-46e1de3dc095" width="700" />

## Jan 19, 2022
* **New Feature**: Numerical filters on the _Study View_ are now editable:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/00926fec-f32b-48c7-a85f-fced7c29070f" width="700" />
* **New Feature**: In the annotation column choose between showing a single icon OncoKB icon or multiple (one for therapeutic, diagnostic and prognostic):
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/87c5f7d6-10d7-4631-8b37-fa8cef4cabe2" width="700" />

## Jan 10, 2022
* **New Documentation for Local cBioPortal Installations**:
    * Explain how to use [Genome Nexus to annotate MAF files](https://docs.cbioportal.org/file-formats/#create-the-cbioportal-mutation-data-file-with-genome-nexus-with-a-minimal-maf-file)

## Jan 4, 2022
*   **Added data** consisting of 27,447 samples from 10 studies:
    * [Endometrial Carcinoma (CPTAC, Cell 2020)](https://www.cbioportal.org/study/summary?id=ucec_cptac_2020) *95 samples*
    * [Pancreatic Ductal Adenocarcinoma (CPTAC, Cell 2021)](https://www.cbioportal.org/study/summary?id=paad_cptac_2021) *140 samples*
    * [Lung Squamous Cell Carcinoma (CPTAC, Cell 2021)](https://www.cbioportal.org/study/summary?id=lusc_cptac_2021) *108 samples*
    * [Lung Adenocarcinoma (CPTAC, Cell 2020)](https://www.cbioportal.org/study/summary?id=luad_cptac_2020) *110 samples*
    * [Glioblastoma (CPTAC, Cell 2021)](https://www.cbioportal.org/study/summary?id=gbm_cptac_2021) *99 samples*
    * [Breast Cancer (CPTAC, Cell 2020)](https://www.cbioportal.org/study/summary?id=brca_cptac_2020) *122 samples*
    * [Pediatric Brain Cancer (CPTAC/CHOP, Cell 2020)](https://www.cbioportal.org/study/summary?id=brain_cptac_2020) *218 samples*
    * [Metastatic Prostate Cancer (Provisional, June 2021)](https://www.cbioportal.org/study/summary?id=mpcproject_broad_2021) *123 samples*
    * [MSK MetTropism (MSK, Cell 2021)](https://www.cbioportal.org/study/summary?id=msk_met_2021) *25,775 samples*
    * [Cancer Therapy and Clonal Hematopoiesis (MSK, 2021)](https://www.cbioportal.org/study/summary?id=msk_ch_ped_2021) *657 samples*


*   **Added TMB (nonsynonymous) scores** for all studies. [Example: new TMB field for study gbm_cptac_2021](https://www.cbioportal.org/study/clinicalData?id=gbm_cptac_2021)   
    (Details for the calculation can be found [HERE](https://github.com/cBioPortal/datahub-study-curation-tools/blob/master/tmb/calculate_tmb/Readme.md))   
    <img width="361" alt="Screen Shot 2022-01-04 at 11 39 20 PM" src="https://user-images.githubusercontent.com/5973438/148133673-a23506e9-7957-4581-99f0-820ca9548672.png">

## Nov 12, 2021
*   **Added data** consisting of 3,680 samples from 6 studies:
    * [Breast Cancer MAPK (MSKCC, Nat Commun 2021)](https://www.cbioportal.org/study/summary?id=brca_mapk_hp_msk_2021) *145 samples*
    * [Colorectal Cancer (MSK, 2020)](https://www.cbioportal.org/study/summary?id=crc_nigerian_2020) *64 samples*
    * [Breast Cancer (MSK, Clinical Cancer Res 2020)](https://www.cbioportal.org/study/summary?id=brca_pareja_msk_2020) *60 samples*
    * [High-Grade Serous Ovarian Cancer (MSK, 2021)](https://www.cbioportal.org/study/summary?id=hgsoc_msk_2021) *45 samples*
    * [Diffuse Glioma (GLASS Consortium, Nature 2019)](https://www.cbioportal.org/study/summary?id=difg_glass_2019) *444 samples*
    * [Pan-cancer analysis of whole genomes (ICGC/TCGA, Nature 2020)](https://www.cbioportal.org/study/summary?id=pancan_pcawg_2020) *2,922 samples*

## Nov 3,2021
* **New Feature**: Add Uniprot topology as a new annotation track on the Mutations Tab of the Results View. [Example: EGFR in MSK-IMPACT (2017) cohort](https://www.cbioportal.org/results/mutations?cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&profileFilter=mutations%2Cfusion%2Ccna&case_set_id=msk_impact_2017_cnaseq&gene_list=EGFR&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/16869603/141861719-f91d787e-e8c3-4c27-8273-7642ec3fd5e1.png" width="700" />

## Oct 1, 2021
* **New Feature**: Arm level Copy Number events are now loaded into cBioPortal using the Categorial Generic Assay Data Type. They can be found in a tab under the Add Charts Button of the _Study View_ [Example: Arm Level Data in TCGA PanCancer Atlas](https://www.cbioportal.org/study/summary?id=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018)

     <img src="https://user-images.githubusercontent.com/1334004/139692484-8245d94f-1821-4476-974e-b6c1a6cd8d12.png" width="700" />

## Sep 22, 2021
*   **Added data** consisting of 14,844 samples from 7 studies:
    * [Colorectal Cancer (MSK, Gastroenterology 2020)](https://www.cbioportal.org/study/summary?id=crc_apc_impact_2020) *471 samples*
    * [Metastatic Breast Cancer (MSK, Cancer Discovery 2021)](https://www.cbioportal.org/study/summary?id=breast_ink4_msk_2021) *1,365 samples*
    * [Lung Adenocarcinoma (MSKCC, 2021)](https://www.cbioportal.org/study/summary?id=luad_mskimpact_2021) *186 samples*
    * [Race Differences in Prostate Cancer (MSK, 2021) ](https://www.cbioportal.org/study/summary?id=prad_msk_stopsack_2021) *2,069 samples*
    * [Medulloblastoma (DKFZ, Nature 2017)](https://www.cbioportal.org/study/summary?id=mbl_dkfz_2017) *491 samples*
    * [Thoracic Cancer (MSK, 2021)](https://www.cbioportal.org/study/summary?id=lung_pdx_msk_2021) *68 samples*
    * [China Pan-cancer (OrigiMed, 2020)](https://www.cbioportal.org/study/summary?id=pan_origimed_2020) *10,194 samples*

## Sep 21, 2021
* **Enhancement**: Dowloading the Lollipop plot on the Mutations Tab of the _Results View_ will now also include the annotation tracks:

    <img src="https://user-images.githubusercontent.com/1334004/139694195-51ad5580-efa5-4b25-9a10-2754726a246b.png" width="700" />

## Aug 17, 2021
* **New Feature**: The Mutations Tab of the _Results View_ can now show exon numbers as an annotation track [Example: MET Exon 14 Mutations in MSK-IMPACT (2017) cohort](https://www.cbioportal.org/results/mutations?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=mutations%2Cfusion%2Ccna&case_set_id=msk_impact_2017_cnaseq&gene_list=MET&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/1334004/139694614-62f62640-2fab-4117-ae09-db2281690f76.png" width="700" />

## Aug 10, 2021
* **New Feature**: Use the filtering capabilities in the Mutations Tab of the _Results View_ to create a custom cohort that one can open directly in the _Study View_ [Example: CTNNB1 in MSK-IMPACT (2017) cohort](https://bit.ly/3w6wu00)

    <img src="https://user-images.githubusercontent.com/1334004/139696976-6e145c05-10d3-44c9-8499-146ad6e13f2f.png" width="700" />

## Jul 27, 2021
* **New Feature**: Add a custom filter to any column of the Mutations Tab in the _Results View_ [Example: CTNNB1 in MSK-IMPACT (2017) cohort](https://bit.ly/3w6wu00)

    <img src="https://user-images.githubusercontent.com/1334004/139697555-34507fa8-2b57-4aa8-84ee-c477efffa864.png" width="700" />

* **New Feature**: Show detailed descriptions for each annotation source in the header of the the Mutations Table in both the _Results View_ and the _Patient View_ [Example link](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P04)

    <img src="https://user-images.githubusercontent.com/1334004/139699259-1be86599-c2e0-457c-b257-1b0d446facf4.png" width="350" />

## Jul 6, 2021
* **New Feature**: Add any clinical data as a column on the Mutations Tab in the _Results View_ [Example: EGFR in MSK-IMPACT (2017) cohort](https://bit.ly/3bxFlyp)

    <img src="https://user-images.githubusercontent.com/1334004/139700811-4b6ce432-4abf-4dc4-8371-634a14d2b603.png" width="700" />

## June 23, 2021
*   **Added data** consisting of 1,084 samples from 5 studies:
    * [Intrahepatic Cholangiocarcinoma (MSK, Hepatology 2021)](https://www.cbioportal.org/study/summary?id=ihch_msk_2021) *412 samples*
    * [Intrahepatic Cholangiocarcinoma (Mount Sinai 2015)](https://www.cbioportal.org/study/summary?id=ihch_ismms_2015) *8 samples*
    * [RAD51B Associated Mixed Cancers (Mandelker 2021](https://www.cbioportal.org/study/summary?id=mixed_msk_tcga_2021) *19 samples*
    * [Intrahepatic Cholangiocarcinoma (MSK, 2020)](https://www.cbioportal.org/study/summary?id=ihch_mskcc_2020) *219 samples*
    * [Lung Adenocarcinoma (NPJ Precision Oncology, MSK 2021)](https://www.cbioportal.org/study/summary?id=luad_msk_npjpo_2021) *426 samples*
*   **Added mass-spec proteome data** from CPTAC to [Breast Invasive Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018), [Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018) and [Colorectal Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=coadread_tcga_pan_can_atlas_2018).
*   **Added mass-spec phosphoproteome site level expression** from CPTAC to [Breast Invasive Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018) and [Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018).
*   **Updated gene tables** Updated tables of genes (main and alias), based on [HGNC](https://www.genenames.org/). See details [HERE](https://github.com/cBioPortal/datahub/tree/master/seedDB#seed-database-schema-2128) in section `Contents of seed database`. Sripts/resources/process used to construct new tables are described [HERE](https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/gene-table-update/build-input-for-importer).

## June 1, 2021
* **New Feature**: In certain studies where we have the data we show read counts for uncalled mutations on the _Patient View_ [Example: A patient in the Glioma (MSK, 2019) cohort](https://www.cbioportal.org/patient?studyId=glioma_msk_2018&caseId=IM-GBM-1)

    <img src="https://user-images.githubusercontent.com/1334004/139701937-91a40335-e95a-45f4-9800-356737a17fab.png" width="700" />

## May 10, 2021
*  **New Feature**: Pick color for User Defined Groups [Example: Color Bladder Cancer Group in MSK-IMPACT (2017) cohort](https://www.cbioportal.org/study/summary?id=msk_impact_2017), implemented by [The Hyve](https://www.thehyve.nl/services/cbioportal).

    <img src="https://user-images.githubusercontent.com/1334004/122818654-7b018280-d2a7-11eb-94c1-834b780618b9.png" width="700" />

## May 4, 2021
*   **New Feature**: Add more categories of mutations to the Mutations Tab on the _Results View_, including Driver/VUS, Splice and Structural Variants [Example: TP53 alterations in the MSK-IMPACT (2017) cohort](https://www.cbioportal.org/results/mutations?cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&profileFilter=mutations%2Cfusion%2Ccna&case_set_id=msk_impact_2017_cnaseq&gene_list=TP53&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/1334004/122817171-99ff1500-d2a5-11eb-8a58-374c9ea32013.png" width="700" />

## April 21, 2021
*   **Added data** consisting of 4074 samples from 9 studies:
    * [Metaplastic Breast Cancer (MSK, 2021)](https://www.cbioportal.org/study/summary?id=mbc_msk_2021) *19 samples*
    * [Lung Adenocarcinoma (MSKCC, 2020)](https://www.cbioportal.org/study/summary?id=luad_mskcc_2020) *604 samples*
    * [Cutaneous Squamous Cell Carcinoma (UCSF, 2021)](https://www.cbioportal.org/study/summary?id=cscc_ucsf_2021) *105 samples*
    * [MSK-IMPACT and MSK-ACCESS Mixed Cohort (MSK, 2021)](https://www.cbioportal.org/study/summary?id=msk_access_2021) *1446 samples*
    * [Melanoma (MSKCC, 2018)](https://www.cbioportal.org/study/summary?id=mel_mskimpact_2020) *720 samples*
    * [Cholangiocarcinoma (ICGC, Cancer Discov 2017)](https://www.cbioportal.org/study/summary?id=chol_icgc_2017) *489 samples*
    * [Esophageal/Stomach Cancer (MSK, 2020)](https://www.cbioportal.org/study/summary?id=egc_mskcc_2020) *487 samples*
    * [Retinoblastoma (MSK, Cancers 2021)](https://www.cbioportal.org/study/summary?id=rbl_mskcc_2020) *83 samples*
    * [Combined Hepatocellular and Intrahepatic Cholangiocarcinoma (Peking University, Cancer Cell 2019)](https://www.cbioportal.org/study/summary?id=hccihch_pku_2019) *121 samples*

## April 20, 2021
*   **New Feature**: Add driver annotations to download tab on _Results View_ [Example: RAS/RAF alterations in colorectal cancer](https://www.cbioportal.org/results/download?cancer_study_list=coadread_tcga_pub&cancer_study_id=coadread_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=coadread_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=coadread_tcga_pub_gistic&Z_SCORE_THRESHOLD=2.0&case_set_id=coadread_tcga_pub_nonhypermut&gene_list=KRAS%20NRAS%20BRAF&gene_set_choice=user-defined-list)

    <img src="https://user-images.githubusercontent.com/1334004/122816424-b2bafb00-d2a4-11eb-8a0f-5e878bb70ccd.png" width="700" />

## March 30, 2021
*   **Enhancement**: Add 95% Confidence Interval for Survival Plots [Example: Altered vs Unaltered EGFR in Lung Cancer](https://www.cbioportal.org/results/comparison?plots_horz_selection=%7B%7D&plots_vert_selection=%7B%7D&plots_coloring_selection=%7B%7D&tab_index=tab_visualize&Action=Submit&session_id=60d0d9fce4b015b63e9f1116&comparison_subtab=survival)

    <img src="https://user-images.githubusercontent.com/1334004/122810147-0295c400-d29d-11eb-9be7-fa14787d72b7.png" width="700" />

## March 11, 2021
*   **New Feature**: Combine different types of alterations in _Comparison View_ [Example: Deletions and Truncating events in primary vs metastases](https://www.cbioportal.org/comparison/alterations?comparisonId=5f73ac54e4b0242bd5d22ab0&selectedEnrichmentEventTypes=%5B%22truncating%22%2C%22feature_truncation%22%2C%22nonsense_mutation%22%2C%22nonsense%22%2C%22stopgain_snv%22%2C%22stop_gained%22%2C%22frameshift%22%2C%22frameshift_variant%22%2C%22frame_shift_del%22%2C%22frameshift_deletion%22%2C%22frame_shift_ins%22%2C%22frameshift_insertion%22%2C%22translation_start_site%22%2C%22initiator_codon_variant%22%2C%22start_codon_snp%22%2C%22start_codon_del%22%2C%22de_novo_start_outofframe%22%2C%22nonstop_mutation%22%2C%22stop_lost%22%2C%22splice_site%22%2C%22splice%22%2C%22splicing%22%2C%22splice_site_snp%22%2C%22splice_site_del%22%2C%22splice_site_indel%22%2C%22splice_region_variant%22%2C%22splice_region%22%2C%22HOMDEL%22%5D&groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D) or read more on [The Hyve's blog](https://www.thehyve.nl/articles/find-differentially-altered-genes-by-combining-alteration-types)

    <img src="https://user-images.githubusercontent.com/1334004/122808698-35d75380-d29b-11eb-8928-0fca5f6e3b34.png" width="700" />

*   **Enhancement**: Improve UI for OncoPrint, aggregating various data modalities in a single add track dropdown button [Example: Add clinical, heatmap and treatment response data into the OncoPrint](https://www.cbioportal.org/results/oncoprint?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ccle_broad_2019_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ccle_broad_2019_cna&cancer_study_list=ccle_broad_2019&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=0&case_set_id=ccle_broad_2019_cnaseq&gene_list=TP53&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/15748980/107722872-f32eba80-6cad-11eb-89d0-b6c5f300ab41.png" width="700" />

## February 16, 2021
*   **Enhancement**: Show only TCGA PanCancer Atlas Pathways in _Results_ and _Patient View_ to avoid showing many similar pathways [Example: Clinvar APC and CTNNB1 alterations in WNT pathway](https://www.cbioportal.org/results/pathways?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=mutations%2Cfusion%2Ccna&case_set_id=msk_impact_2017_cnaseq&gene_list=APC%2520CTNNB1&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/1334004/122806359-3a4e3d00-d298-11eb-89fc-9e107a4651b4.png" width="700" />

## January 28, 2021
*   **New Feature**: Show ClinVar Interpretation in Mutation tables [Example: Clinvar Interpretations in BRCA2](https://www.cbioportal.org/results/mutations?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=0&case_set_id=msk_impact_2017_cnaseq&gene_list=BRCA2&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/1334004/122805797-8a78cf80-d297-11eb-9454-fed37ed1ddb5.png" width="700" />


## January 12, 2021
*   **New Feature**: Add your own custom data for a sample or patient to use on the study or comparison view  [Example: Add custom data to three samples and do a comparison](https://www.cbioportal.org/study/summary?id=msk_impact_2017)

    <img src="https://user-images.githubusercontent.com/1334004/104651553-ce7efd00-5685-11eb-8017-c43f24421e3f.png" width="700">

*   **New Feature**: Show the mutations of a patient inside a pathway schematic using PathwayMapper  [Example: Notch signaling pathway in a prostate cancer patient](https://www.cbioportal.org/patient/pathways?studyId=msk_impact_2017&caseId=P-0000377)

    <img src="https://user-images.githubusercontent.com/1334004/104649378-9f1ac100-5682-11eb-91eb-4e0f024a4ded.png" width="700">

*   **New Feature**: Display and compare generic assays, such as microbiome and treatment response, on the study view [Example: Prasinovirus microbiome signatures in TCGA](https://www.cbioportal.org/study?id=6000b6aae4b015b63e9d4d44)
    
    <img src="https://user-images.githubusercontent.com/1334004/104651211-54e70f00-5685-11eb-94f1-f20930793c28.png" width="700">

*   **New Feature**: The Plots tab on _Results View_ now allows you to group alterations by Driver and VUS [Example: POLE driver mutations vs VUSs against mutation counts in TCGA Colorectal Adenocarcinoma](https://bit.ly/3ssY9Gu)
    <img src="https://user-images.githubusercontent.com/1334004/104768867-56bfd980-573c-11eb-8e03-5ad86d01f25a.png" width="700">


## December 31, 2020
*   **Added data** consisting of 430 samples from 5 studies:
    * [Juvenile Papillomatosis and Breast Cancer (MSK, 2020)](https://www.cbioportal.org/study/summary?id=brca_jup_msk_2020) *5 samples*
    * [Mixed cfDNA (MSKCC, 2020)](https://www.cbioportal.org/study/summary?id=mixed_cfdna_msk_2020) *229 samples*
    * [Metastatic Melanoma (DFCI, Nature Medicine 2019)](https://www.cbioportal.org/study/summary?id=mel_dfci_2019) *144 samples*
    * [Lung Cancer (SMC, Cancer Research 2016)](https://www.cbioportal.org/study/summary?id=lung_smc_2016) *22 samples*
    * [Upper Tract Urothelial Carcinoma (IGBMC, Genome Biology 2021)](https://www.cbioportal.org/study/summary?id=utuc_igbmc_2021) *30 samples*
*    **Added survival data** to [Breast Cancer (METABRIC, Nature 2012 & Nat Commun 2016)](https://www.cbioportal.org/study/summary?id=brca_metabric)

## November 3, 2020
*   **New Feature**: The [map](https://www.cbioportal.org/installations) of local installations of cBioPortal is available now. Please consider registering your instance [here](http://bit.ly/2XW3wNG ).
    ![image](https://user-images.githubusercontent.com/840895/99544685-96f41d00-2982-11eb-9de2-457a4456c294.png)

*   **Enhancement**: upgraded the _Genomic Evolution_ tab in _Patient View_ with timeline [Example](https://www.cbioportal.org/patient/genomicEvolution?caseId=P04&studyId=lgg_ucsf_2014)
    ![image](https://user-images.githubusercontent.com/840895/99543852-a3c44100-2981-11eb-8f53-14a0b51b767c.png)

## October 20, 2020
*   **Enhancement**: _Expression_ tab has now been merged into the _Plots_ tab
    ![image](https://user-images.githubusercontent.com/840895/99565969-4ee0f480-299a-11eb-8371-8cfd09c6e4ba.png)

## October 16, 2020
*   **Added data** consisting of 25,078 samples from 5 studies:
    * [Melanomas (TCGA, Cell 2015)](https://cbioportal.org/study/summary?id=skcm_tcga_pub_2015) *359 samples*
    * [Retinoblastoma cfDNA (MSKCC 2020)](https://www.cbioportal.org/study/summary?id=rbl_cfdna_msk_2020) *14 samples*
    * [The Angiosarcoma Project (Provisional, July 2020)](https://www.cbioportal.org/study/summary?id=angs_painter_2020) *83 samples*
    * [Bladder Cancer (MSK/TCGA, 2020)](https://www.cbioportal.org/study/summary?id=blca_msk_tcga_2020) *476 samples*
    * [Cancer Therapy and Clonal Hematopoiesis (MSK, 2020)](https://www.cbioportal.org/study/summary?id=msk_ch_2020) *24,146 samples*

*    **Added MSI data** (MSIsensor from [Mariamidze et al. 2018](https://www.sciencedirect.com/science/article/pii/S0092867418303131#bib66) and MANTIS scores from [Roychowdhury et al. 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5972025/)) for all 32 TCGA PanCan Atlas Cohorts.
     <img width="600" src="https://user-images.githubusercontent.com/5973438/95988728-52131e80-0df7-11eb-8e25-0617d4b9e8c4.png">

*    **Added new profile** ‚ÄúRNA-Seq V2 expression Z-scores relative to normal samples‚Äù for 16 TCGA PanCan Atlas Cohorts.
     The normals samples RNA-Seq V2 expression data were curated from [GDC](https://gdc.cancer.gov/about-data/publications/pancanatlas), and can be downloaded from our [Datahub](https://github.com/cBioPortal/datahub) or [Data Set](https://www.cbioportal.org/datasets) page. [Example: ERBB2 expression z-scores relative to normal expression](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&case_set_id=all&data_priority=0&gene_list=ERBB2&geneset_list=%20&plots_coloring_selection=%7B%22colorByCopyNumber%22%3A%22true%22%7D&plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_STUDY%22%7D&plots_vert_selection=%7B%22selectedGeneOption%22%3A2064%2C%22dataType%22%3A%22MRNA_EXPRESSION%22%2C%22selectedDataSourceOption%22%3A%22rna_seq_v2_mrna_median_all_sample_ref_normal_Zscores%22%2C%22logScale%22%3A%22true%22%7D&profileFilter=0&tab_index=tab_visualize)

![image](https://user-images.githubusercontent.com/840895/96191246-b32f1500-0f11-11eb-8727-bcfdd2d724a3.png)

## October 13, 2020
*   **Enhancement**: _Study View_ now allows comparing samples with mutations or copy number alterations in different genes
    ![image](https://user-images.githubusercontent.com/840895/99558433-ba729400-2991-11eb-9054-1d8751b4d84f.png)

*   **New Feature**: When treatment timeline is available (e.g. in [this study](https://www.cbioportal.org/study/summary?id=lgg_ucsf_2014)), _Study View_ now allows the selection and comparison of patients treated with specific drugs, or samples sequenced pre or post specific drug treatments
    ![image](https://user-images.githubusercontent.com/840895/99556958-248a3980-2990-11eb-97e3-d833518f8aef.png)

## September 30, 2020
*   **New Feature**: Microbiome signature data is available for comparison now. [Example: comparing colorectal subtypes for enriched microbiome signatures](https://www.cbioportal.org/comparison/generic_assay_microbiome_signature?sessionId=5dea8bbde4b05f1dfb8ace51)
    ![image](https://user-images.githubusercontent.com/840895/99566847-666cad00-299b-11eb-91d3-21b21f267733.png)

## September 22, 2020

*   **Enhancement**: The timeline feature in _Patient View_ has been refactored with an improved UI. [Example](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P04)
    ![image](https://user-images.githubusercontent.com/840895/94962574-ae577380-04c4-11eb-9ff7-1a79e242dd73.png)

*   **Enhancement**: Logrank p-values are now provided for all survival analysis (previously only availalbe when comparing two groups). [Example](https://www.cbioportal.org/comparison/survival?sessionId=5f77816de4b015b63e9c0f81)
    <img width="600" alt="" src="https://user-images.githubusercontent.com/840895/94962916-481f2080-04c5-11eb-811f-40d8fc4437cf.png">


## August 11, 2020
*   **New Feature**: microbiome data of TCGA samples from [Poore et al. 2020](https://www.nature.com/articles/s41586-020-2095-1) are now available for analysis in the *OncoPrint* and *Plots* tabs. [Example: Orthohepadnavirus across TCGA cancers](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&case_set_id=all&data_priority=0&gene_list=EGFR&geneset_list=%20&plots_coloring_selection=%7B%22selectedOption%22%3A%22-1_undefined%22%2C%22colorByMutationType%22%3A%22false%22%2C%22colorByCopyNumber%22%3A%22false%22%7D&plots_horz_selection=%7B%22selectedGeneOption%22%3A1956%2C%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_TYPE%22%2C%22mutationCountBy%22%3A%22MutationType%22%2C%22logScale%22%3A%22false%22%7D&plots_vert_selection=%7B%22selectedGeneOption%22%3A1956%2C%22selectedGenericAssayOption%22%3A%22Orthohepadnavirus%22%2C%22dataType%22%3A%22MICROBIOME_SIGNATURE%22%2C%22selectedDataSourceOption%22%3A%22microbiome_signature%22%2C%22mutationCountBy%22%3A%22MutationType%22%2C%22logScale%22%3A%22false%22%7D&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/90060684-1783ed00-dcb3-11ea-8368-e96dd69c552c.png)

*   **New Feature**: You can now compare *DNA Methylation* data between groups using the *Comparison* feature. [Example: Comparing DNA methylation levels between samples with high vs low BRCA1 expression](https://www.cbioportal.org/results/comparison?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_id=ov_tcga_pub&cancer_study_list=ov_tcga_pub&case_set_id=ov_tcga_pub_3way_complete&comparison_selectedGroups=%5B%22BRCA1%3A%20EXP%3E1%22%2C%22BRCA1%3A%20EXP%3C-1%22%5D&comparison_subtab=dna_methylation&data_priority=0&gene_list=BRCA1%253AEXP%253E1%2520%250ABRCA1%253A%2520EXP%253C-1&gene_set_choice=user-defined-list&geneset_list=%20&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=ov_tcga_pub_mrna_median_Zscores&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ov_tcga_pub_mutations&plots_coloring_selection=%7B%7D&plots_horz_selection=%7B%7D&plots_vert_selection=%7B%22selectedDataSourceOption%22%3A%22mrna_median_Zscores%22%7D&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/89957811-632c8d00-dc06-11ea-87dc-bdbb4fa23a0b.png)

*   **Added data** consisting of 513 samples from 3 studies:
    * [Breast Cancer (SMC 2018)](https://www.cbioportal.org/study/summary?id=brca_smc_2018) *187 samples*
    * [Germ Cell Tumors and Shared Leukemias (MSK     2020)](https://www.cbioportal.org/study/summary?id=gct_msk_2020) *21 samples*
    * [Lung Adenocarcinoma (OncoSG, Nat Genet 2020)](https://www.cbioportal.org/study/summary?id=luad_oncosg_2020) *305 samples*

*   **Added RPPA data** in addition to the microbiome data for [31 TCGA Pancan studies (except LAML)](https://www.cbioportal.org/study/summary?id=acc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018)
    <img width="400" alt="" src="https://user-images.githubusercontent.com/840895/90141147-ba3a7b00-dd48-11ea-964f-840dd591e535.png">

## July 21, 2020
*   **New Feature**: The *Mutations* tab now has the option to show mutation effects for different transcripts / isoforms. Note that some annotation features are only available for the canonical isoform. [example](https://www.cbioportal.org/results/mutations?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=msk_impact_2017&case_set_id=msk_impact_2017_cnaseq&data_priority=0&gene_list=FGFR2&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&mutations_transcript_id=ENST00000358487&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/88306535-d95d6400-ccd8-11ea-9e64-0c6600f65e50.png)

*   **Enhancement**: The *Plots* tab is now supported in multi-study queries. [example](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&case_set_id=all&data_priority=0&gene_list=TP53%2520PTEN%2520BRCA1%2520IGF2%2520EGFR&geneset_list=%20&plots_coloring_selection=%7B%22colorByCopyNumber%22%3A%22false%22%7D&plots_horz_selection=%7B%22selectedGeneOption%22%3A1956%7D&plots_vert_selection=%7B%22logScale%22%3A%22true%22%7D&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/88309239-07907300-ccdc-11ea-995b-7ddd658ec46e.png)

*   **New Feature**: You can now share custom groups in the *Study View* [example](https://www.cbioportal.org/study/summary?id=msk_impact_2017#sharedGroups=5d092670e4b0ab4137876374,5c99454ee4b0ab4137873dc5)

    <img width="333" alt="" src="https://user-images.githubusercontent.com/840895/88309724-b7fe7700-ccdc-11ea-969a-28ed8551ffd7.png">


## June 11, 2020
*   **Added data** consisting of 267 samples from 2 studies:
    * [Gastric Cancer (OncoSG, 2018)](https://www.cbioportal.org/study/summary?id=stad_oncosg_2018) *147 samples*
    * 120 ctDNA samples added to [Non-Small Cell Lung Cancer (TRACERx, NEJM & Nature 2017)](https://www.cbioportal.org/study/summary?id=nsclc_tracerx_2017) *447 samples*

## June 9, 2020

*   **Enhancement**: using [OQL](https://www.cbioportal.org/oql#oql-modifiers) to query for mutations based on a protein position range. [example](https://www.cbioportal.org/results/mutations?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=msk_impact_2017&case_set_id=msk_impact_2017_cnaseq&data_priority=0&gene_list=TP53%253AMUT_(95-288*)&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/84427197-83ce6b80-abf2-11ea-9d18-3a4f4524e545.png)

*   **New Feature**: you can now send the OncoPrint data to the [OncoPrinter tool](https://www.cbioportal.org/oncoprinter) for customization.
    ![image](https://user-images.githubusercontent.com/840895/84318326-3c85a380-ab3c-11ea-97c1-34343cb3e996.png)

*   **Enhancement**: Mutational spectrum data can be downloaded from OncoPrint
    ![image](https://user-images.githubusercontent.com/840895/84322695-68585780-ab43-11ea-9224-a965331e35fc.png)


## June 2, 2020

*   **Enhancement**: Pediatric cancer studies are now grouped and highlighted in the query page
    ![image](https://user-images.githubusercontent.com/840895/84318659-c897cb00-ab3c-11ea-8209-93c940321a0f.png)

## May 6, 2020
*   **Added data** consisting of 574 samples from 3 studies:
    * [Uterine Sarcoma/Mesenchymal (MSK, Clin Cancer Res 2020)](https://www.cbioportal.org/study/summary?id=usarc_msk_2020) *108 samples*
    * [Metastatic castration-sensitive prostate cancer (MSK, Clin Cancer Res 2020)](https://www.cbioportal.org/study/summary?id=prad_mcspc_mskcc_2020) *424 samples*
    *  [Glioblastoma (Columbia, Nat Med. 2019)](https://www.cbioportal.org/study/summary?id=gbm_columbia_2019) *42 samples*
*   **Updated one study:**
    *  Expression data was added to [The Metastatic Breast Cancer Project (Provisional, February 2020)](https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017).

## April 24, 2020
* **New Feature**: Add a new chart on the _Study View_ for selecting samples based on pre-defined case lists:

    <img width="333" alt="Screen Shot 2020-04-24 at 9 18 25 AM" src="https://user-images.githubusercontent.com/1334004/80216807-a1c51a00-860c-11ea-8ba5-3781bd603ee5.png">

## April 10, 2020
*   **New Feature**: Make cohorts on the _Study View_ using continuous molecular profiles of one or more gene(s), such as mRNA expression, methylation, RPPA and continuous CNA. [example](https://www.cbioportal.org/study/summary?id=brca_tcga#filterJson={"genomicDataFilters":[{"hugoGeneSymbol":"ERBB2","profileType":"mrna_median_Zscores","values":[{"start":1.5,"end":2},{"start":2,"end":2.5},{"start":2.5,"end":3},{"start":3,"end":3.5},{"start":3.5,"end":4},{"start":4}]}],"genomicProfiles":[["mrna_median_Zscores"]],"studyIds":["brca_tcga"]})

    <img src="https://user-images.githubusercontent.com/1334004/79270846-da2a6280-7e6c-11ea-8cb9-b40e7d201ea1.png" width=500 />

    Combine this with the group comparison feature to compare e.g. all quartiles of expression:

    <img src="https://user-images.githubusercontent.com/1334004/79270618-79028f00-7e6c-11ea-8a30-9eaedee948ca.png" width=400 />

*   **New Feature**: Annotate mutations using the _Mutation Mapper Tool_ on the GRCh38 reference genome:

    [![mutation_mapper_tool_grch38](https://user-images.githubusercontent.com/1334004/79233555-07f4b480-7e37-11ea-9d0a-0cafff434fa5.png)](https://www.cbioportal.org/mutation_mapper)

## April 3, 2020
*   **New Feature**: Extended the _Comparison_ tab to support the comparison of altered samples per gene or alteration. This [example query](https://www.cbioportal.org/results/comparison?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=msk_impact_2017&case_set_id=msk_impact_2017_Non-Small_Cell_Lung_Cancer&comparison_createdGroupsSessionId=5e8a2d4ae4b0ff7ef5fdd7d0&comparison_selectedGroups=%5B%22EGFR%3A%20AMP%20%26%20MUT%22%2C%22EGFR%3A%20MUT%20only%22%2C%22EGFR%3A%20AMP%20only%22%5D&comparison_subtab=survival&data_priority=0&gene_list=EGFR%253AAMP%250AEGFR%253AMUT_DRIVER&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&profileFilter=0&tab_index=tab_visualize) compares NSCLC patients with 1) both mutated and amplified EGFR, 2) mutated EGFR only, and 3) amplified EGFR only.

![image](https://user-images.githubusercontent.com/840895/78508085-3ea13f80-7752-11ea-8e87-e486061def12.png)

## March 27, 2020
*   **Enhancement**: User selections in the Plots tab are now saved in the URL. [example](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_id=coadread_tcga_pub&cancer_study_list=ov_tcga_pan_can_atlas_2018&case_set_id=ov_tcga_pan_can_atlas_2018_cnaseq&data_priority=0&gene_list=TP53&gene_set_choice=user-defined-list&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ov_tcga_pan_can_atlas_2018_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ov_tcga_pan_can_atlas_2018_mutations&plots_horz_selection=%7B%22dataType%22%3A%22MUTATION_EXTENDED%22%7D&profileFilter=0&tab_index=tab_visualize)

*   **New Feature**: Added table of data availability per profile in the _Study View_. [example](https://www.cbioportal.org/study/summary?id=ucec_tcga_pub)

<img src="https://user-images.githubusercontent.com/840895/78508182-e454ae80-7752-11ea-914b-bac17f725a9f.png" width="300">

## March 20, 2020
*   **Enhancement**: Extended _Survival Analysis_ to support more outcome measures. [example](https://www.cbioportal.org/results/comparison?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=luad_tcga_pan_can_atlas_2018&case_set_id=luad_tcga_pan_can_atlas_2018_cnaseq&comparison_selectedGroups=%5B%22Unaltered%20group%22%2C%22Altered%20group%22%5D&comparison_subtab=survival&data_priority=0&gene_list=KRAS%250ATP53&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=luad_tcga_pan_can_atlas_2018_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=luad_tcga_pan_can_atlas_2018_mutations&profileFilter=0&tab_index=tab_visualize)

![image](https://user-images.githubusercontent.com/840895/78508508-5c23d880-7755-11ea-9e16-36bff32678ad.png)

## March 18, 2020
*   **Added data** consisting of 1,393 samples from 3 studies:
    * [Breast Cancer (Alpelisib plus AI, Nature Cancer 2020)](https://www.cbioportal.org/study/summary?id=breast_alpelisib_2020) *141 samples*
    * [Glioma (MSKCC, Clin Cancer Res 2019)](https://www.cbioportal.org/study/summary?id=glioma_mskcc_2019) *1,004 samples*
    * [Mixed cfDNA (MSK, Nature Medicine 2019)](https://www.cbioportal.org/study/summary?id=cfdna_msk_2019) *248 samples*

## March 3, 2020
* **New Feature**: Added Pathways tab to the _Results View_ page, which visualizes the alteration frequencies of genes in pathways of interest. The pathways are pulled from https://www.pathwaymapper.org and shown in a read only view. One can edit these pathways in the PathwayMapper editor. For more information see the [tutorial](https://www.cbioportal.org/tutorials#pathways).

  [![pathwaymapper_screenshot](https://user-images.githubusercontent.com/1334004/76771001-133fbc00-6775-11ea-8b36-82edb1e7be09.png)](https://www.cbioportal.org/results/pathways?Action=Submit&Z_SCORE_THRESHOLD=1.0&cancer_study_id=gbm_tcga_pub&cancer_study_list=gbm_tcga_pub&case_set_id=gbm_tcga_pub_sequenced&gene_list=TP53%20MDM2%20MDM4&gene_set_choice=user-defined_list&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=gbm_tcga_pub_cna_rae&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=gbm_tcga_pub_mutations)

## February 12, 2020
*   **Added data** consisting of 1,605 samples from 3 studies:
    * [Tumors with TRK fusions (MSK, 2019)](https://www.cbioportal.org/study/summary?id=ntrk_msk_2019) *106 samples*
    * [Lymphoma Cell Lines (MSKCC, 2020)](https://www.cbioportal.org/study/summary?id=lymphoma_cellline_msk_2020) *34 samples*
    * [Prostate Adenocarcinoma (MSKCC, 2020)](https://www.cbioportal.org/study/summary?id=prad_cdk12_mskcc_2020) *1,465 samples*

## February 6, 2020
* **New Feature**: Extend the [recent group comparison feature](https://www.cbioportal.org/tutorials#group-comparison) by allowing comparisons inside the _Results View_ page. The new tab allows for quick comparison of altered vs unaltered cases by survival, clinical information, mutation, copy number events and mRNA expression:

  [![group_results640px](https://user-images.githubusercontent.com/1334004/74002155-2a041f00-493c-11ea-9867-14740202c368.gif)](https://www.cbioportal.org/results/comparison?Z_SCORE_THRESHOLD=2.0&cancer_study_id=coadread_tcga_pub&cancer_study_list=coadread_tcga_pub&case_set_id=coadread_tcga_pub_nonhypermut&gene_list=KRAS%20NRAS%20BRAF&gene_set_choice=user-defined-list&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=coadread_tcga_pub_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=coadread_tcga_pub_mutations)

* **Performance enhancement**: the _Study View_'s mutation table now loads faster for studies with multiple gene panels. For the [genie portal](https://genie.cbioportal.org), which has a study with many different gene panels this resulted in a speed-up from ~90-120 seconds to 5 seconds.
* Read more about the v3.2.2 release [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.2.2)

## January 30, 2020
* **Enhancement**: Show HGVSg in mutations table and linkout to [Genome Nexus](https://www.genomenexus.org):

  ![hgvsg genome nexus](https://user-images.githubusercontent.com/1334004/73494837-4d026200-4383-11ea-8968-f9a8a6e00675.png)


* **Enhancement**: Add a pencil button near gene list in results page which opens interface for quickly modifying the oql of the query:

  ![edit query pencil](https://user-images.githubusercontent.com/1334004/73494684-f72dba00-4382-11ea-8bd0-001fdf0cb681.png)


* See more updates [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.2.1)

## January 29, 2020
*   **Added data** consisting of 197 samples from 2 studies:
    * [Bladder/Urinary Tract Cancer (MSK, 2019)](https://www.cbioportal.org/study/summary?id=utuc_pdx_msk_2019) *78 samples*
    * [Upper Tract Urothelial Carcinoma (MSK, 2019)](https://www.cbioportal.org/study/summary?id=utuc_msk_2019) *119 samples*

## December 19, 2019
* **Enhancement**: We restored support for submitting large queries from external applications using HTTP POST requests.  Accepted parameters are the same
  as appear in the url of a query submitted from the homepage.

* See more updates [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.8)

## December 12, 2019
* **Enhancement**: Several enhancements to the display of gene panels on the _Patient View_ page, by [The Hyve](https://www.thehyve.nl/), described in more detail [here](https://www.thehyve.nl/articles/gene-panels-cbioportal-patient-view)

  ![image](https://api.thehyve.nl/uploads/gene-panels-patient-view-cbioportal.png)
* **Enhancement**: Add Count Bubbles to Oncoprint Toolbar

  ![Screenshot from 2019-12-06 11-36-21](https://user-images.githubusercontent.com/20069833/70339336-aa16e700-181c-11ea-94ac-c4acff272e4f.png)
* See more updates [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.7)

## November 29, 2019
* **Enhancement**: Support group comparison for custom charts in _Study View_ page
* **Enhancement**: Performance improvement of Co-Expression analysis.
* **Enhancement**: Kaplan-Meier plots now supports custom time range.
* See more updates [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.6)

<img src="https://user-images.githubusercontent.com/840895/69887458-02963380-12b5-11ea-8151-4d531db0fcc5.png" width="500">

## November 22, 2019

* **New Feature**: Support for Treatment response data in the Oncoprint and Plots tab, including new *Waterfall* plot type. Read more in [The Hyve's blog post](https://www.thehyve.nl/articles/treatment-response-data-cbioportal)

![image](https://api.thehyve.nl/uploads/Waterfall-plots-cbioportal.png)

## November 15, 2019
* **Enhancement**: heatmap tracks in OncoPrint now has separate headers and sub-menus. [example](https://www.cbioportal.org/results/oncoprint?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=brca_tcga&case_set_id=brca_tcga_protein_quantification&clinicallist=PROFILED_IN_brca_tcga_mutations&data_priority=0&gene_list=ESR1%2520PGR%2520ERBB2&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_gistic&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_mrna_median_Zscores&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_mutations&genetic_profile_ids_PROFILE_PROTEIN_EXPRESSION=brca_tcga_rppa_Zscores&heatmap_track_groups=brca_tcga_rna_seq_v2_mrna_median_Zscores%2CESR1%2CPGR%2CERBB2%3Bbrca_tcga_protein_quantification_zscores%2CESR1%2CPGR%2CERBB2&tab_index=tab_visualize)

![image](https://user-images.githubusercontent.com/840895/69502724-7fee2c80-0ee0-11ea-81b7-2ada01fc99a2.png)

* **Enhancement**: global settings for query session

<img src="https://user-images.githubusercontent.com/840895/69502843-b6787700-0ee1-11ea-927b-049635a5a8ac.png" width="300">


## November 7, 2019

*   **Added data** consisting of 212 samples from 3 studies:
    * [Metastatic Melanoma (DFCI, Science 2015)](https://www.cbioportal.org/study/summary?id=skcm_dfci_2015) *110 samples*
    * [Melanoma (MSKCC, NEJM 2014)](https://www.cbioportal.org/study/summary?id=skcm_mskcc_2014) *64 samples*
    * [Metastatic Melanoma (UCLA, Cell 2016)](https://www.cbioportal.org/study/summary?id=mel_ucla_2016) *38 samples*

## October 30, 2019

*   **Added data** consisting of 178 samples from 2 studies:
    * [Intrahepatic Cholangiocarcinoma (Shanghai, Nat Commun 2014)](https://www.cbioportal.org/study/summary?id=ihch_smmu_2014) *103 samples*
    * [Non-Small Cell Lung Cancer (MSK, Cancer Cell 2018)](https://www.cbioportal.org/study/summary?id=nsclc_mskcc_2018) *75 samples*

## October 23, 2019
* **Enhancement**: Quick example links in Plots tab. [example](https://www.cbioportal.org/results/plots?session_id=5ba11ce6498eb8b3d567e7ed)

## October 14, 2019

* **New Feature**: Fusion Genes table in `Study View`. [example](https://www.cbioportal.org/study/summary?id=msk_impact_2017)

![image](https://user-images.githubusercontent.com/840895/68320157-6ef28e00-008d-11ea-8cb3-e4b54aa1d848.png)

## October 11, 2019
* **Enhancement**: The Download interface on the homepage has been removed. Enhanced download functionality is now available after querying on the results page.

  _Home page_:

  ![homepage download tab removed](https://user-images.githubusercontent.com/1334004/66735303-48627e00-ee66-11e9-863f-5ec713ea444a.png)

  _Results page_:

  ![results page download tab](https://user-images.githubusercontent.com/1334004/66736057-7ea0fd00-ee68-11e9-942b-e123b9659126.png)

  Note that as before one can always download the full raw data on the [Data Sets page](https://www.cbioportal.org/datasets) or from [Datahub](https://github.com/cBioPortal/datahub/).

## October 9, 2019
*  **Added data** consisting of  2725 samples from 4 studies:

    * [Cancer Cell Line Encyclopedia (Broad, 2019)](https://www.cbioportal.org/study/summary?id=ccle_broad_2019) *1739 samples*
    * [Chronic Lymphocytic Leukemia (Broad, Nature 2015)](https://www.cbioportal.org/study/summary?id=cll_broad_2015#summary) *537 samples*
    * [Rectal Cancer (MSK,Nature Medicine 2019)](https://www.cbioportal.org/study/summary?id=rectal_msk_2019) *339 samples*
    * [Colon Cancer (CPTAC-2 Prospective, Cell 2019)](https://www.cbioportal.org/study/summary?id=coad_cptac_2019) *110 samples*

*  **Updated**  [Esophageal Carcinoma (TCGA, Nature 2017)](https://www.cbioportal.org/study/summary?id=stes_tcga_pub#summary) with addition of CNA data for Esophageal Squamous Cell Carcinoma cases *90 samples*.

## September 18, 2019
* **New Feature**: The list and order of charts of a study will be automatically saved now as a user preference on the _study view_ page.

## September 6, 2019
*   **Added data** consisting of 1216 samples from 3 studies:
    * [Breast Cancer (MSKCC, 2019)](https://www.cbioportal.org/study/summary?id=brca_mskcc_2019) *70 samples*
    * [Brain Tumor PDXs (Mayo Clinic, 2019)](https://www.cbioportal.org/study/summary?id=gbm_mayo_pdx_sarkaria_2019)*97 samples*
    * [Adenoid Cystic Carcinoma Project (2019)](https://www.cbioportal.org/study/summary?id=acc_2019) *1049 samples*

## August 13, 2019
*   **Added data** consisting of 295 samples from 3 studies:
    *  [Pediatric Preclinical Testing Consortium (PPTC, 2019)](https://www.cbioportal.org/study/summary?id=pptc_2019) *261 samples*
    *  [Non-small cell lung cancer (MSK, Science 2015)](https://www.cbioportal.org/study/summary?id=nsclc_mskcc_2015) *16 samples*
    *  [Prostate Cancer (MSK, 2019)](https://www.cbioportal.org/study/summary?id=prad_msk_2019) *18 samples*

## July 26, 2019
*   **Added data** consisting of 35 samples from 1 study:
    *  [Clear Cell Renal Cell Carcinoma (DFCI, Science 2019)](https://www.cbioportal.org/study/summary?id=ccrcc_dfci_2019) *35 samples*
*   Added Hypoxia data for:
    *  [Breast Invasive Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018)
    *  [Colorectal Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=coadread_tcga_pan_can_atlas_2018)
    *  [Cervical Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=cesc_tcga_pan_can_atlas_2018)
    *  [Glioblastoma Multiforme (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=gbm_tcga_pan_can_atlas_2018)
    *  [Head and Neck Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=hnsc_tcga_pan_can_atlas_2018)
    *  [Kidney Renal Clear Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=kirc_tcga_pan_can_atlas_2018)
    *  [Kidney Renal Papillary Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=kirp_tcga_pan_can_atlas_2018)
    *  [Brain Lower Grade Glioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=lgg_tcga_pan_can_atlas_2018)
    *  [Liver Hepatocellular Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=lihc_tcga_pan_can_atlas_2018)
    *  [Lung Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=luad_tcga_pan_can_atlas_2018)
    *  [Lung Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=lusc_tcga_pan_can_atlas_2018)
    *  [Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018)
    *  [Pancreatic Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=paad_tcga_pan_can_atlas_2018)
    *  [Pheochromocytoma and Paraganglioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=pcpg_tcga_pan_can_atlas_2018)
    *  [Prostate Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=prad_tcga_pan_can_atlas_2018)
    *  [Skin Cutaneous Melanoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=skcm_tcga_pan_can_atlas_2018)
    *  [Thyroid Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=thca_tcga_pan_can_atlas_2018)
    *  [Uterine Corpus Endometrial Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=ucec_tcga_pan_can_atlas_2018)
## July 24, 2019
*   **Added data** consisting of 151 samples from 1 study:
    *  [Myeloproliferative Neoplasms (CIMR, NEJM 2013)](https://www.cbioportal.org/study/summary?id=mpn_cimr_2013) *151 samples*
## July 13, 2019
*   **Public Release 6.1 of AACR Project GENIE:**
    *   The sixth data set, GENIE 6.0-public, was released in early July 2019.  A patch to GENIE 6.0-public, GENIE 6.1-pubic, was subsequently released on July 13, 2019.  The combined data set now includes nearly 70,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for nearly 80 major cancer types, including data from nearly 11,000 patients with lung cancer, greater than  9,700 patients with breast cancer, and nearly 7,000 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](https://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](https://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](https://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](https://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).
## July 2, 2019
*   **Added data** consistng of 785 samples from 4 studies:
    *  [Non-Small Cell Lung Cancer (TRACERx, NEJM 2017)](https://www.cbioportal.org/study/summary?id=nsclc_tracerx_2017) *327 samples*
    *  [Acute myeloid leukemia or myelodysplastic syndromes (WashU, 2016)](https://www.cbioportal.org/study/summary?id=mnm_washu_2016) *136 samples*
    *  [Basal Cell Carcinoma (UNIGE, Nat Genet 2016)](https://www.cbioportal.org/study/summary?id=bcc_unige_2016) *293 samples*
    *  [Colon Adenocarcinoma (CaseCCC, PNAS 2015)](https://www.cbioportal.org/study/summary?id=coad_caseccc_2015) *29 samples*

## June 19, 2019
* **New Feature**: Show *Genome Aggregation Database (gnomAD)* population frequencies in the mutations table - see [example](https://bit.ly/2ISHgiu):

  ![gnomad feature news](https://user-images.githubusercontent.com/1334004/59794400-e07c9c00-92a6-11e9-97ea-a79bfc8f3885.gif)

## June 12, 2019
*   **Added data**  of 1350 samples from 3 studies:
    *  [Pheochromocytoma and Paraganglioma (TCGA, Cell 2017)](https://www.cbioportal.org/study?id=pcpg_tcga_pub) *178 samples*
    *  [Metastatic Solid Cancers (UMich, Nature 2017)](https://www.cbioportal.org/study?id=metastatic_solid_tumors_mich_2017) *500 samples*
    *   [Acute Myeloid Leukemia (OHSU, Nature 2018)](https://www.cbioportal.org/study/summary?id=aml_ohsu_2018) *672 samples*
*    Added survival data for TCGA PanCan Atlas Cohorts (>10,000 samples across 33 tumor types).
*    Added hypoxia data for [Bladder Urothelial Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=blca_tcga_pan_can_atlas_2018)
## June 7, 2019
* **New Group Comparison Feature**: Compare clinical and genomic features of user-defined groups of samples/patients. [View Tutorial](https://www.cbioportal.org/tutorials#group-comparison)

[![group-comparison](https://user-images.githubusercontent.com/840895/59052073-40f9eb00-885c-11e9-9ddb-6d036533e5f5.png)](https://www.cbioportal.org/comparison/mutations?sessionId=5cf92520e4b0ab4137874376)


## May 8, 2019
* **New Feature**: Show *Post Translational Modification (PTM)* information from [dbPTM](https://dbptm.mbc.nctu.edu.tw/) on the _Mutation Mapper_ - see [example](https://bit.ly/2VVrWdi):

  ![ptm feature_news](https://user-images.githubusercontent.com/1334004/57391288-34667200-718c-11e9-9fcc-b849542c74b3.gif)

## April 26, 2019
*   **Added data**  of 568 samples from 4 studies:
    *  [Adenoid Cystic Carcinoma (JHU, Cancer Prev Res 2016)](https://www.cbioportal.org/study?id=acyc_jhu_2016) *25 samples*
    *  [Histiocytosis Cobimetinib (MSK, Nature 2019)](https://www.cbioportal.org/study?id=histiocytosis_cobi_msk_2019) *52 samples*
    *  [Upper Tract Urothelial Carcinoma (Cornell/Baylor/MDACC, Nat Comm 2019)](https://www.cbioportal.org/study?id=utuc_cornell_baylor_mdacc_2019) *47 samples*
    *  [Metastatic Prostate Adenocarcinoma (SU2C/PCF Dream Team, PNAS 2019)](https://www.cbioportal.org/study?id=prad_su2c_2019) *444 samples*

## March 29, 2019
* **New Feature**: Use the new quick search tab on the homepage to more easily navigate to a study, gene or patient:

  ![quick_search_news](https://user-images.githubusercontent.com/1334004/55113078-8f4c7a00-50b4-11e9-9d95-e9a6e1dcda52.gif)

## March 15, 2019
*   **Added data**  of 338 samples from 4 studies:
    *  [Adenoid Cystic Carcinoma (MGH, Nat Gen  2016)](https://www.cbioportal.org/study?id=acyc_mgh_2016) *10 samples*
    *   [Gallbladder Cancer (MSK, Cancer 2018)](https://www.cbioportal.org/study?id=gbc_msk_2018) *103 samples*
    *   [The Metastatic Prostate Cancer Project (Provisional, December 2018)](https://www.cbioportal.org/study?id=prad_mpcproject_2018) *19 samples*
    *   [Adult Soft Tissue Sarcomas (TCGA, Cell 2017)](https://www.cbioportal.org/study?id=sarc_tcga_pub) *206 samples*

## February 22, 2019
* **Enhancement**: Exon number and HGVSc annotations are available in optional columns in the Mutations tab on the _Results_ page and in the _Patient View_.
* **New feature**: option to a show regression line in the scatter plot in the Plots tab on the _Results_ page

![image](https://user-images.githubusercontent.com/840895/53193265-99271d00-35de-11e9-8ffc-352afc2ba906.png)

## February 19, 2019
* **New feature**: *Copy-Number Segments tab* on the _Study View_ page using [igv.js v2](https://github.com/igvteam/igv.js) - see [example](https://www.cbioportal.org/study?id=ucec_tcga_pub&tab=cnSegments)
* Improved *Copy-Number Segments* tab on the _Results_ page
* **New feature**: *OncoKB* and *Cancer Hotspots tracks* in the Mutations tab on the _Results_ page

![image](https://user-images.githubusercontent.com/840895/53054860-d1f6b300-3473-11e9-8178-1df10eed36fc.png)


## January 24, 2019
*   **Added data**  of 2328 samples from 8 studies:
    *   [Uveal Melanoma (QIMR, Oncotarget 2016)](https://www.cbioportal.org/study?id=um_qimr_2016#summary) *28 samples*
    *   [Squamous Cell Carcinoma of the Vulva (CUK, Exp Mol Med 2018)](https://www.cbioportal.org/study?id=vsc_cuk_2018) *15 samples*
    *   [TMB and Immunotherapy (MSKCC, Nat Genet 2019)](https://www.cbioportal.org/study?id=tmb_mskcc_2018#summary) *1661 samples*
    *   [Glioma (MSK, 2018)](https://www.cbioportal.org/study?id=glioma_msk_2018) *91 samples*
    *   [Urothelial Carcinoma (Cornell/Trento, Nat Gen 2016)](https://www.cbioportal.org/study?id=blca_cornell_2016) *72 samples*
    *   [Hepatocellular Carcinoma (MSK, Clin Cancer Res 2018)](https://www.cbioportal.org/study?id=hcc_mskimpact_2018) *127 samples*
    *   [MSK Thoracic PDX (MSK, Provisional)](https://www.cbioportal.org/study?id=lung_msk_pdx) *139 samples*
    *   [Cholangiocarcinoma (MSK, Clin Cancer Res 2018)](https://www.cbioportal.org/study?id=chol_msk_2018) *195 samples*
*  **Updated data** for [The Metastatic Breast Cancer Project (Provisional, October 2018)](https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017) *237 samples*
## January 10, 2019
* cBioPortal now supports queries for driver mutations, fusions and copy number alterations as well as germline/somatic mutations using [Onco Query Language (OQL)](https://www.cbioportal.org/oql) -- see [example](https://www.cbioportal.org/results/oncoprint?session_id=5c23ea81e4b05228701f9d44)
* A new [tutorial](https://www.cbioportal.org/tutorials) explores OQL and provides examples of how OQL can be a powerful tool to refine queries.


## December 17, 2018
* The 10th phase of cBioPortal architectural upgrade is now complete: the _Study View_ has been moved to the new architecture with numerous improvements. This marks the completion of the cBioPortal architectural refactoring! üéâüéâüéâ

[![image](https://user-images.githubusercontent.com/840895/50120599-11aadf80-0224-11e9-9f9f-adae0ed42c72.png)](https://www.cbioportal.org/study?id=msk_impact_2017)

## October 29, 2018
* The ninth phase of the cBioPortal architectural upgrade is now complete: the _results page_ is now a single-page application with better performance.
* Supported plotting mutations by type in _Plots_ tab

![image](https://user-images.githubusercontent.com/840895/47598695-d996c700-d96d-11e8-9371-9e591cf25ea3.png)

## October 19, 2018

* Support selection of transcript of interest in the [MutationMapper tool](https://www.cbioportal.org/mutation_mapper.jsp) via [Genome Nexus](https://www.genomenexus.org).

![mutation_mapper_dropdown](https://user-images.githubusercontent.com/1334004/47240017-f99a1980-d3b4-11e8-99a4-e8c73b25ecec.png)

## October 17, 2018
*   **Added data**  of 3578 samples from 8 studies:
    *   [Rhabdoid Cancer (BCGSC, Cancer Cell 2016)](https://www.cbioportal.org/study?id=mrt_bcgsc_2016#summary) *40 samples*
    *   [Diffuse Large B-Cell Lymphoma (Duke, Cell 2017)](https://www.cbioportal.org/study?id=dlbcl_duke_2017#summary) *1001 samples*
    *   [Diffuse Large B cell Lymphoma (DFCI, Nat Med 2018)](https://www.cbioportal.org/study?id=dlbcl_dfci_2018#summary) *135 samples*
    *   [Breast Fibroepithelial Tumors (Duke-NUS, Nat Genet 2015)](https://www.cbioportal.org/study?id=bfn_duke_nus_2015#summary) *22 samples*
    *   [Uterine Clear Cell Carcinoma (NIH, Cancer 2017)](https://www.cbioportal.org/study?id=uccc_nih_2017#summary) *16 samples*
    *   [Endometrial Cancer (MSK, 2018)](https://www.cbioportal.org/study?id=ucec_msk_2018#summary) *197 samples*
    *   [Breast Cancer (MSK, Cancer Cell 2018)](https://www.cbioportal.org/study?id=breast_msk_2018#summary) *1918 samples*
    *   [MSS Mixed Solid Tumors (Van Allen, 2018)](https://www.cbioportal.org/study?id=mixed_allen_2018#summary) *249 samples*
*   Updated data for [The Angiosarcoma Project (Provisional, September 2018)](https://www.cbioportal.org/study?id=angs_project_painter_2018#summary) *48 samples*

## August 20, 2018
*   Now you can log in on the public cBioPortal with your Google account and save your virtual studies for quick analysis.

![image](https://user-images.githubusercontent.com/840895/44370286-902a8700-a4a7-11e8-9c8f-dfda87fbd66b.png)

## August 7, 2018
*   The eighth phase of the cBioPortal architectural upgrade is now complete: The _Plots_, _Expression_, _Network_, and _Bookmarks_ tabs, and therefore all analysis tabs in the _results page_, have been moved to the new architecture.
*   Updated the [MutationMapper tool](https://www.cbioportal.org/mutation_mapper.jsp), now connecting to [Genome Nexus](https://genomenexus.org/) for annotating mutations on the fly.
*   _Total Mutations_ and _Fraction Genome Altered_ are now available in _Plots_ tab for visualization and analysis.
*   Enhanced clinical attribute selector for OncoPrint, now showing sample counts per attribute.

![image](https://user-images.githubusercontent.com/840895/43609786-ec946d64-9672-11e8-8218-4e63bd5007e8.png)


## July 27, 2018
*   **Added data**  of 2787 samples from 10 studies:
    *   [Mixed Tumors (PIP-Seq 2017)](https://www.cbioportal.org/study?id=mixed_pipseq_2017#summary) *103 samples*
    *   [Nonmuscle Invasive Bladder Cancer (MSK Eur Urol 2017)](https://www.cbioportal.org/study?id=blca_nmibc_2017#summary) *105 samples*
    *   [Pediatric Neuroblastoma (TARGET, 2018)](https://www.cbioportal.org/study?id=nbl_target_2018_pub) *1089 samples*
    *   [Pediatric Pan-Cancer (DKFZ - German Cancer Consortium, 2017)](https://www.cbioportal.org/study?id=pediatric_dkfz_2017) *961 samples*
    *   [Skin Cutaneous Melanoma (Broad, Cancer Discov 2014)](https://www.cbioportal.org/study?id=skcm_broad_brafresist_2012) *78 samples*
    *   [Cutaneous Squamous Cell Carcinoma (MD Anderson, Clin Cancer Res 2014)](https://www.cbioportal.org/study?id=cscc_hgsc_bcm_2014#summary) *39 samples*
    *   [Diffuse Large B-cell Lymphoma (BCGSC, Blood 2013)](https://www.cbioportal.org/study?id=nhl_bcgsc_2013#summary) *53 samples*
    *   [Non-Hodgkin Lymphoma (BCGSC, Nature 2011)](https://www.cbioportal.org/study?id=nhl_bcgsc_2011#summary) *14 samples*
    *   [Chronic lymphocytic leukemia (ICGA, Nat 2011)](https://www.cbioportal.org/study?id=cllsll_icgc_2011#summary) *105 samples*
    *   [Neuroblastoma (Broad Institute 2013)](https://www.cbioportal.org/study?id=nbl_broad_2013#summary) *240 samples*

## June 20, 2018
*   The seventh phase of the cBioPortal architectural upgrade is now complete: The _Enrichments_ and _Co-Expression_ tabs have been moved to the new architecture.
*   Supported merged gene tracks in OncoPrint and [Onco Query Language](https://www.cbioportal.org/onco_query_lang_desc.jsp) -- see [example](https://www.cbioportal.org/index.do?session_id=5b2bedca498eb8b3d566ab60)

![image](https://user-images.githubusercontent.com/840895/41738366-def14850-755f-11e8-8a7c-deb789dc03da.png)

## May 10, 2018
*   Enhanced OncoPrint to show germline mutations -- see [example](https://www.cbioportal.org/index.do?session_id=5af5a304498eb8b3d56615b6)

![image](https://user-images.githubusercontent.com/840895/39926636-c3304d8c-54fd-11e8-9cb7-76430c8772d7.png)

## April 17, 2018
*   **Added data**  of 3732 samples from 4 TARGET studies:
    *   [Pediatric Acute Lymphoid Leukemia - Phase II (TARGET, 2018)](https://www.cbioportal.org/study?id=all_phase2_target_2018_pub#summary) *1978 samples*
    *   [Pediatric Acute Myeloid Leukemia (TARGET, 2018)](https://www.cbioportal.org/study?id=aml_target_2018_pub#summary) *1025 samples*
    *   [Pediatric Rhabdoid Tumor (TARGET, 2018)](https://www.cbioportal.org/study?id=rt_target_2018_pub#summary) *72 samples*
    *   [Pediatric Wilms' Tumor (TARGET, 2018)](https://www.cbioportal.org/study?id=wt_target_2018_pub#summary) *657 samples*
*   **Added data**  of 3416 samples from 10 published studies:
    *    [Prostate Adenocarcinoma (MSKCC/DFCI, Nature Genetics 2018)](https://www.cbioportal.org/study?id=prad_p1000#summary) *1013 samples*
    *    [Prostate Adenocarcinoma (EurUrol, 2017)](https://www.cbioportal.org/study?id=prad_eururol_2017#summary) *65 samples*
    *    [Non-Small Cell Lung Cancer (MSK, JCO 2018)](https://www.cbioportal.org/study?id=nsclc_pd1_msk_2018#summary) *240 samples*
    *    [Small-Cell Lung Cancer (Multi-Institute 2017)](https://www.cbioportal.org/study?id=sclc_cancercell_gardner_2017#summary) *20 samples*
    *    [The Angiosarcoma Project (Provisional, February 2018)](https://www.cbioportal.org/study?id=angs_project_painter_2018#summary) *14 samples*
    *    [Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2016)](https://www.cbioportal.org/study?id=all_stjude_2016#summary) *73 samples*
    *    [Updated Segment data and Allele Frequencies for The Metastatic Breast Cancer Project (Provisional, October 2017)](https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017#summary) *103 samples*
    *    [Colorectal Cancer (MSK, Cancer Cell 2018)](https://www.cbioportal.org/study?id=crc_msk_2018#summary) *1134 samples*
    *    [Metastatic Esophagogastric Cancer (MSK,Cancer Discovery 2017)](https://www.cbioportal.org/study?id=egc_msk_2017#summary) *341 samples*
    * [Bladder Cancer (TCGA, Cell 2017)](https://www.cbioportal.org/study?id=blca_tcga_pub_2017#summary) *413 samples*

## April 5, 2018
*   **Added data** from the TCGA PanCanAtlas project with >10,000 samples from 33 tumor types:
    *   [Adrenocortical Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=acc_tcga_pan_can_atlas_2018)
    *   [Bladder Urothelial Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=blca_tcga_pan_can_atlas_2018)
    *   [Breast Invasive Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=brca_tcga_pan_can_atlas_2018)
    *   [Cervical Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=cesc_tcga_pan_can_atlas_2018)
    *   [Colon Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=coad_tcga_pan_can_atlas_2018)
    *   [Cholangiocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=chol_tcga_pan_can_atlas_2018)
    *   [Diffuse Large B-Cell Lymphoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=dlbc_tcga_pan_can_atlas_2018)
    *   [Esophageal Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=esca_tcga_pan_can_atlas_2018)
    *   [Glioblastoma Multiforme (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=gbm_tcga_pan_can_atlas_2018)
    *   [Head and Neck Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=hnsc_tcga_pan_can_atlas_2018)
    *   [Kidney Chromophobe (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=kich_tcga_pan_can_atlas_2018)
    *   [Kidney Renal Clear Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=kirc_tcga_pan_can_atlas_2018)
    *   [Kidney Renal Papillary Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=kirp_tcga_pan_can_atlas_2018)
    *   [Acute Myeloid Leukemia (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=laml_tcga_pan_can_atlas_2018)
    *   [Brain Lower Grade Glioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=lgg_tcga_pan_can_atlas_2018)
    *   [Liver Hepatocellular Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=lihc_tcga_pan_can_atlas_2018)
    *   [Lung Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=luad_tcga_pan_can_atlas_2018)
    *   [Lung Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=lusc_tcga_pan_can_atlas_2018)
    *   [Mesothelioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=meso_tcga_pan_can_atlas_2018)
    *   [Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=ov_tcga_pan_can_atlas_2018)
    *   [Pancreatic Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=paad_tcga_pan_can_atlas_2018)
    *   [Pheochromocytoma and Paraganglioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=pcpg_tcga_pan_can_atlas_2018)
    *   [Prostate Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=prad_tcga_pan_can_atlas_2018)
    *   [Rectum Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=read_tcga_pan_can_atlas_2018)
    *   [Sarcoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=sarc_tcga_pan_can_atlas_2018)
    *   [Skin Cutaneous Melanoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=skcm_tcga_pan_can_atlas_2018)
    *   [Stomach Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=stad_tcga_pan_can_atlas_2018)
    *   [Testicular Germ Cell Tumors (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=tgct_tcga_pan_can_atlas_2018)
    *   [Thyroid Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=thca_tcga_pan_can_atlas_2018)
    *   [Thymoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=thym_tcga_pan_can_atlas_2018)
    *   [Uterine Corpus Endometrial Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=ucec_tcga_pan_can_atlas_2018)
    *   [Uterine Carcinosarcoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=ucs_tcga_pan_can_atlas_2018)
    *   [Uveal Melanoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=uvm_tcga_pan_can_atlas_2018)

## March 20, 2018
*   The sixth phase of the cBioPortal architectural upgrade is now complete: The _Download_ tab has been moved to the new architecture.
*   Data can now be downloaded in [tabular format](https://www.thehyve.nl/articles/download-data-from-cbioportal-oncoprint) from OncoPrint.
*   Added an option to download an SVG file on the _Cancer Type Summary_ tab.

## January 15, 2018
*   The fifth phase of the cBioPortal architectural upgrade is now complete: The OncoPrint and Survival tabs have been moved to the new architecture.

## November 20, 2017
*   You can now combine multiple studies and view them on the study summary page. [Example: liver cancer studies](https://www.cbioportal.org/study?id=liad_inserm_fr_2014%2Chcc_inserm_fr_2015%2Clihc_amc_prv%2Clihc_riken%2Clihc_tcga)
*   You can now bookmark or share your selected samples as virtual studies with the _share_ icon on the study summary page. [Example: a virtual study of breast tumors](https://www.cbioportal.org/study?id=5a12fd57498eb8b3d5605cd4)
*   Cross-study query reimplemented: Now you can view an OncoPrint of multiple studies. [Example: querying NSCLC tumors from 5 studies](https://www.cbioportal.org/index.do?session_id=5a135bee498eb8b3d5605d01)

![image](https://user-images.githubusercontent.com/840895/33045546-3c8a636c-ce1b-11e7-9f52-7060c89a8dfd.png)

## October 17, 2017
*   The fourth phase of the cBioPortal architectural upgrade is now complete: The _Mutual Exclusivity_ and _Cancer Type Summary_ tabs have been moved to the new architecture.
*   Updated protein structure alignment data in Mutations tab are now retrieved from Genome Nexus via the [G2S web service](https://g2s.genomenexus.org/).

## October 2, 2017
*   **Added data** of 1646 samples from 7 published studies:
    *   [NGS in Anaplastic Oligodendroglioma and Anaplastic Oligoastrocytomas tumors (MSK, Neuro Oncol 2017)](https://www.cbioportal.org/study?id=odg_msk_2017#summary)   *22 samples*
    *   [MSK-IMPACT Clinical Sequencing Cohort for Non-Small Cell Cancer (MSK, Cancer Discovery 2017)](https://www.cbioportal.org/study?id=lung_msk_2017) *915 samples*
    *   [Paired-exome sequencing of acral melanoma (TGEN, Genome Res 2017)](https://www.cbioportal.org/study?id=mel_tsam_liang_2017#summary) *38 samples*
    *   [MSK-IMPACT Clinical Sequencing Cohort in Prostate Cancer (MSK, JCO Precision Oncology 2017)](https://www.cbioportal.org/study?id=prad_mskcc_2017#summary) *504 samples*
    *   [Whole-exome sequences (WES) of pretreatment melanoma tumors (UCLA, Cell 2016)](https://www.cbioportal.org/study?id=skcm_ucla_2016#summary) *39 samples*
    *   [Next generation sequencing (NGS) of pre-treatment metastatic melanoma samples (MSK, JCO Precision Oncology 2017)](https://www.cbioportal.org/study?id=skcm_vanderbilt_mskcc_2015#summary) *66 samples*
    *   [Targeted gene sequencing in 62 high-grade primary Unclassified Renal Cell Carcinoma (MSK, Nature 2016)](https://www.cbioportal.org/study?id=urcc_mskcc_2016#summary) *62 samples*
*   **Updated data** for [MSK-IMPACT Clinical Sequencing Cohort (MSK, Nat Med 2017)](https://www.cbioportal.org/study?id=msk_impact_2017) with overall survival data.
## August 3, 2017
*   The third phase of the cBioPortal architectural upgrade is now complete: The _Mutations_ tab now has a fresh look and faster performance -- see [example](https://www.cbioportal.org/index.do?session_id=598386e5498e5df2e29376ab&show_samples=false&#mutation_details&)

![image](https://user-images.githubusercontent.com/840895/28942244-9ea43bee-7868-11e7-9b24-0aaf0f9c010d.png)

*   Variant interpretations from the [CIViC database](https://civic.genome.wustl.edu) are now integrated into the annotation columns on the Mutations tab and in the patient view pages
*   New summary graph for all cancer studies and samples on the front page


## June 26, 2017
*   The second phase of the cBioPortal architectural upgrade is now complete: The query interface now has a fresh look and faster performance.

![image](https://user-images.githubusercontent.com/840895/27399281-3ba5357a-568a-11e7-9730-d230285d7805.png)

## May 12, 2017
*   **Added data** of 12,211 samples from 11 published studies:
    *   [MSK-IMPACT Clinical Sequencing Cohort (MSK, Nat Med 2017)](https://www.cbioportal.org/study?id=msk_impact_2017)   *10,945 samples*
    *   [Whole-genome sequencing of pilocytic astrocytomasatic (DKFZ, Nat Genetics, 2013)](https://www.cbioportal.org/study?id=past_dkfz_heidelberg_2013) *96 samples*
    *   [Hepatocellular Carcinomas (INSERM, Nat Genet 2015)](https://www.cbioportal.org/study?id=hcc_inserm_fr_2015) *243 samples*
    *   [Cystic Tumor of the Pancreas (Johns Hopkins, PNAS 2011)](https://www.cbioportal.org/study?id=pact_jhu_2011) *32 samples*
    *   [Whole-Genome Sequencing of Pancreatic Neuroendocrine Tumors (ARC- Net, Nature, 2017)](https://www.cbioportal.org/study?id=panet_arcnet_2017) *98 samples*
    *   [Medulloblastoma (Sickkids, Nature 2016)](https://www.cbioportal.org/study?id=mbl_sickkids_2016) *46 samples*
    *   [Genetic Characterization of NSCLC young adult patients (University of Turin, Lung Cancer 2016)](https://www.cbioportal.org/study?id=nsclc_unito_2016) *41 samples*
    *   [Genomic Profile of Patients with Advanced Germ Cell Tumors (MSK, JCO 2016).](https://www.cbioportal.org/study?id=gct_msk_2016) *180 samples*
    *   [Ampullary Carcinoma (Baylor, Cell Reports 2016)](https://www.cbioportal.org/study?id=ampca_bcm_2016) *160 samples*
    *   [Mutational profiles of metastatic breast cancer (INSERM, 2016)](https://www.cbioportal.org/study?id=brca_igr_2015) *216 samples*
    *   [Prostate Adenocarcinoma (Fred Hutchinson CRC, Nat Med 2016)](https://www.cbioportal.org/study?id=prad_fhcrc) *154 samples*


## May 5, 2017
*   First phase of cBioPortal architectural upgrade complete: Patient view now has fresh look and faster performance. [example](https://bit.ly/2pNY961)

## March 28, 2017
*   **New features**:
    * Per-sample mutation spectra are now available in OncoPrints -- see [example](https://www.cbioportal.org/index.do?cancer_study_list=ucec_tcga_pub&cancer_study_id=ucec_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ucec_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ucec_tcga_pub_gistic&Z_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=ucec_tcga_pub_manuscript&case_ids=&patient_case_select=sample&gene_set_choice=user-defined-list&gene_list=POLE%0D%0AERBB2%0D%0AKRAS%0D%0ACTNNB1&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit&show_samples=false&clinicallist=SUBTYPE,NO_CONTEXT_MUTATION_SIGNATURE,%23%20mutations&)

    ![image](https://cloud.githubusercontent.com/assets/840895/24209131/d34e33b2-0efb-11e7-945a-fd2b3c66d195.png)

    * mRNA heat map clustering is now supported in OncoPrints
    * MDACC Next-Generation Clustered Heat Maps are now available in the patient view
    * cBioPortal web site style change


## Feburary 2, 2017
*   **New features**:
    * 3D hotspot mutation annotations are now available from 3dhotspots.org
*   **New data**:
    * CPTAC proteomics data have been integrated for TCGA breast, ovarian, and colorectal provisional studies


## December 23, 2016
*   **New features**:
    * Heat map visualization of gene expression data in the OncoPrint

    ![OncoPrint Heatmap](https://cloud.githubusercontent.com/assets/840895/21407479/81f8c008-c79e-11e6-9b82-3f4eb8f8e1ea.png)

    * Heat map visualization of gene expression data in the Study View page connecting to MDACC's TCGA Next-Generation Clustered Heat Map Compendium

## October 7, 2016
*   **New features**:
    * All data sets can now be downloaded as flat files from the new [Data Hub](https://github.com/cBioPortal/datahub)
    * Annotation of putative driver missense mutations in OncoPrints, based on [OncoKB](https://www.oncokb.org), mutation hotspots, and recurrence in cBioPortal and COSMIC

    ![OncoPrint-OncoKB](https://cloud.githubusercontent.com/assets/840895/19208804/094dd320-8ccd-11e6-8012-f30104b62ff4.png)

    * Copy number segments visualization directly in the browser in a new *CN Segments* tab via [IGV.js](https://igv.org/)

    ![image](https://cloud.githubusercontent.com/assets/840895/19200747/ad03c824-8c98-11e6-8051-91fc40e1dc56.png)

*   **Improvements**:
    * Improved cancer study view page (bug fixes and increased performance)

## July 24, 2016
*   **Added data** of 4,375 samples from 21 published studies:
    *   [Adenoid Cystic Carcinoma (MDA, Clin Cancer Res 2015)](https://www.cbioportal.org/study?id=acyc_mda_2015) *102 samples*
    *   [Adenoid Cystic Carcinoma (FMI, Am J Surg Pathl. 2014)](https://www.cbioportal.org/study?id=acyc_fmi_2014) *28 samples*
    *   [Adenoid Cystic Carcinoma (Sanger/MDA, JCI 2013)](https://www.cbioportal.org/study?id=acyc_sanger_2013) *24 samples*
    *   [Adenoid Cystic Carcinoma of the Breast (MSKCC, J Pathol. 2015)](https://www.cbioportal.org/study?id=acbc_mskcc_2015) *12 samples*
    *   [Bladder Cancer, Plasmacytoid Variant (MSKCC, Nat Genet 2016)](https://www.cbioportal.org/study?id=blca_plasmacytoid_mskcc_2016) *34 samples*
    *   [Breast Cancer (METABRIC, Nat Commun 2016)](https://www.cbioportal.org/study?id=brca_metabric) *1980 samples*
    *   [Chronic Lymphocytic Leukemia (Broad, Cell 2013)](https://www.cbioportal.org/study?id=lcll_broad_2013) *160 samples*
    *   [Chronic Lymphocytic Leukemia (IUOPA, Nature 2015)](https://www.cbioportal.org/study?id=cll_iuopa_2015) *506 samples*
    *   [Colorectal Adenocarcinoma (DFCI, Cell Reports 2016)](https://www.cbioportal.org/study?id=coadread_dfci_2016) *619 samples*
    *   [Cutaneous T Cell Lymphoma (Columbia U, Nat Genet 2015)](https://www.cbioportal.org/study?id=ctcl_columbia_2015) *42 samples*
    *   [Diffuse Large B-Cell Lymphoma (Broad, PNAS 2012)](https://www.cbioportal.org/study?id=dlbc_broad_2012) *58 samples*
    *   [Hepatocellular Adenoma (Inserm, Cancer Cell 2014)](https://www.cbioportal.org/study?id=liad_inserm_fr_2014) *46 samples*
    *   [Hypodiploid Acute Lymphoid Leukemia (St Jude, Nat Genet 2013)](https://www.cbioportal.org/study?id=all_stjude_2013) *44 samples*
    *   [Insulinoma (Shanghai, Nat Commun 2013)](https://www.cbioportal.org/study?id=panet_shanghai_2013) *10 samples*
    *   [Malignant Pleural Mesothelioma (NYU, Cancer Res 2015)](https://www.cbioportal.org/study?id=plmeso_nyu_2015) *22 samples*
    *   [Mantle Cell Lymphoma (IDIBIPS, PNAS 2013)](https://www.cbioportal.org/study?id=mcl_idibips_2013) *29 samples*
    *   [Myelodysplasia (Tokyo, Nature 2011)](https://www.cbioportal.org/study?id=mds_tokyo_2011) *29 samples*
    *   [Neuroblastoma (Broad, Nat Genet 2013)](https://www.cbioportal.org/study?id=nbl_ucologne_2015) *56 samples*
    *   [Oral Squamous Cell Carcinoma (MD Anderson, Cancer Discov 2013)](https://www.cbioportal.org/study?id=hnsc_mdanderson_2013) *40 samples*
    *   [Pancreatic Adenocarcinoma (QCMG, Nature 2016)](https://www.cbioportal.org/study?id=paad_qcmg_uq_2016) *383 samples*
    *   [Recurrent and Metastatic Head & Neck Cancer (JAMA Oncology, 2016)](https://www.cbioportal.org/study?id=hnc_mskcc_2016) *151 samples*
*   **New TCGA study**:
    *   [Pan-Lung Cancer (TCGA, Nat Genet 2016)](https://www.cbioportal.org/study?id=nsclc_tcga_broad_2016) *1144 samples*
* Updated **TCGA provisional studies**
    *   updated to the Firehose run of January 28, 2016
    *   RPPA data updated with the latest data from MD Anderson
    *   [OncoTree](http://oncotree.mskcc.org/) codes assigned per sample

## June 6, 2016
*   **New features**:
    * Annotation of mutation effect and drug sensitivity on the Mutations tab and the patient view pages (via [OncoKB](https://www.oncokb.org))  
      ![oncokb-screenshot](https://cloud.githubusercontent.com/assets/840895/15825344/54bbbc44-2bd1-11e6-8b65-e265e03f453a.png)
*   **Improvements**:
    * Improved OncoPrint visualization using WebGL: faster, more zooming flexibility, visualization of recurrent variants
    * Improved Network tab with SBGN view for a single interaction
    * Performance improvement of tables in the study view page
    * Mutation type summary on the Mutations tab

## March 31, 2016
*   **New features**:
    *   Visualization of "Enrichments Analysis" results via volcano plots
    *   Improved performance of the cross cancer expression view  by switching to Plot.ly graphs
    *   Improvements to the "Clinical Data" tab on the study view page
    *   More customization options for the cross-cancer histograms
    *   Performance improvements in the study view and query result tabs
*   **Added data** of 1235 samples from 3 published studies:
    *   [Merged Cohort of LGG and GBM (TCGA, 2016)](https://www.cbioportal.org/study?id=lgggbm_tcga_pub)
    *   [Lung Adenocarcinoma (MSKCC, 2015)](https://www.cbioportal.org/study?id=luad_mskcc_2015)
    *   [Poorly-Differentiated and Anaplastic Thyroid Cancers (MSKCC, JCI 2016)](https://www.cbioportal.org/study?id=thyroid_mskcc_2016)

## January 12, 2016
*   **New features**:
    *   Visualization of multiple samples in a patient
    *   Visualization of timeline data of a patient ([example](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P04))<br/>
        ![timeline-example](https://cloud.githubusercontent.com/assets/840895/12055606/cca26160-aefc-11e5-93f9-2ecfe7e95caf.png)
*   All **TCGA data** updated to the latest Firehose run of August 21, 2015
*   **New TCGA studies**:
    *   [Cholangiocarcinoma (TCGA, Provisional)](https://www.cbioportal.org/study?id=chol_tcga)
    *   [Mesothelioma (TCGA, Provisional)](https://www.cbioportal.org/study?id=meso_tcga)
    *   [Testicular Germ Cell Cancer (TCGA, Provisional)](https://www.cbioportal.org/study?id=tgct_tcga)
    *   [Thymoma (TCGA, Provisional)](https://www.cbioportal.org/study?id=thym_tcga)
*   **Added data** of 650 samples from 10 published studies:
    *   [Neuroblastoma (AMC Amsterdam, Nature 2012)](https://www.cbioportal.org/study?id=nbl_amc_2012)
    *   [Clear Cell Renal Cell Carcinoma (U Tokyo, Nat Genet 2013)](https://www.cbioportal.org/study?id=ccrcc_utokyo_2013)
    *   [Multiregion Sequencing of Clear Cell Renal Cell Carcinoma (IRC, Nat Genet 2014)](https://www.cbioportal.org/study?id=ccrcc_irc_2014)
    *   [Bladder Urothelial Carcinoma (Dana Farber & MSKCC, Cancer Discovery 2014)](https://www.cbioportal.org/study?id=blca_dfarber_mskcc_2014)
    *   [Low-Grade Gliomas (UCSF, Science 2014)	](https://www.cbioportal.org/study?id=lgg_ucsf_2014)
    *   [Esophageal Squamous Cell Carcinoma (UCLA, Nat Genet 2014)](https://www.cbioportal.org/study?id=escc_ucla_2014)
    *   [Acinar Cell Carcinoma of the Pancreas (Johns Hopkins, J Pathol 2014)](https://www.cbioportal.org/study?id=paac_jhu_2014)
    *   [Gastric Adenocarcinoma (TMUCIH, PNAS 2015)](https://www.cbioportal.org/study?id=egc_tmucih_2015)
    *   [Primary Central Nervous System Lymphoma (Mayo Clinic, Clin Cancer Res 2015)](https://www.cbioportal.org/study?id=pcnsl_mayo_2015)
    *   [Desmoplastic Melanoma (Broad Institute, Nat Genet 2015)](https://www.cbioportal.org/study?id=desm_broad_2015)
*   All mutation data mapped to [UniProt canonical isoforms](https://www.uniprot.org/help/canonical_and_isoforms)

## December 23, 2015
*   **New features**:
    *   Visualization of RNA-seq expression levels across TCGA studies (cross-cancer queries)<br/>
        ![cross cancer expression](https://cloud.githubusercontent.com/assets/840895/11821643/8c9a6de4-a338-11e5-83d1-a9e92cb6dfb6.png)
    *   Selection of genes in the study view to initiate queries<br/>
        ![query gene in study view](https://cloud.githubusercontent.com/assets/840895/11977489/5af87ca8-a951-11e5-87ad-5c31b0451cc1.png)
*   **Improvement**:
    *   3-D structures in the "Mutations" tab are now rendered by 3Dmol.js (previously JSmol)
    *   Improved performance by code optimization and compressing large data by gzip


## December 1, 2015
*   **New feature**: Annotated statistically recurrent hotspots, via new algorithm by [Chang et al. 2015](https://www.nature.com/nbt/journal/vaop/ncurrent/full/nbt.3391.html)</br>
    ![Annotate recurrent hotspots](https://cloud.githubusercontent.com/assets/840895/11794851/7729839e-a281-11e5-9413-12dc885b947d.png)


## November 9, 2015
*   **New features**:
    *   Links to MyCancerGenome.org for mutations<br/>
        ![Link to MyCancerGenome.org](https://cloud.githubusercontent.com/assets/11892704/11050295/9745f78c-8712-11e5-8690-a9a04c4455fb.png)
    *   Improved display of selection samples on the study view page
*   **Improvements**:
    *   "Enrichments" analysis is now run across all genes
    *   The "Network" tab is now using Cytoscape.js (Adobe Flash is no longer required)


## October 6, 2015

*   **New TCGA data**:
    *   [Breast Invasive Carcinoma (TCGA, Cell 2015)](https://www.cbioportal.org/study?id=brca_tcga_pub2015)
    *   [Prostate Adenocarcinoma (TCGA, in press)](https://www.cbioportal.org/study?id=prad_tcga_pub)
    *   [Uveal Melanoma (TCGA, Provisional)](https://www.cbioportal.org/study?id=uvm_tcga)
*   **Added data** of 763 samples from 12 published studies:
    *   [Small Cell Lung Cancer (U Cologne, Nature 2015)](https://www.cbioportal.org/study?id=sclc_ucologne_2015)
    *   [Uterine Carcinosarcoma (JHU, Nat Commun 2014)](https://www.cbioportal.org/study?id=ucs_jhu_2014)
    *   [Microdissected Pancreatic Cancer Whole Exome Sequencing (UTSW, Nat Commun 2015)](https://www.cbioportal.org/study?id=paad_utsw_2015)
    *   [Pancreatic Neuroendocrine Tumors (JHU, Science 2011)](https://www.cbioportal.org/https://www.cbioportal.org/study?id=panet_jhu_2011)
    *   [Renal Non-Clear Cell Carcinoma (Genentech, Nat Genet 2014)](https://www.cbioportal.org/study?id=nccrcc_genentech_2014)
    *   [Infant MLL-Rearranged Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2015)](https://www.cbioportal.org/study?id=all_stjude_2015)
    *   [Rhabdomyosarcoma (NIH, Cancer Discov 2014)](https://www.cbioportal.org/study?id=rms_nih_2014)
    *   [Thymic epithelial tumors (NCI, Nat Genet 2014)](https://www.cbioportal.org/study?id=tet_nci_2014)
    *   [Pediatric Ewing Sarcoma (DFCI, Cancer Discov 2014)](https://www.cbioportal.org/study?id=es_dfarber_broad_2014)
    *   [Ewing Sarcoma (Institut Cuire, Cancer Discov 2014)](https://www.cbioportal.org/study?id=es_iocurie_2014)
    *   [Cutaneous squamous cell carcinoma (DFCI, Clin Cancer Res 2015)](https://www.cbioportal.org/study?id=cscc_dfarber_2015)
    *   [Gallbladder Carcinoma (Shanghai, Nat Genet 2014)](https://www.cbioportal.org/study?id=gbc_shanghai_2014)

## August 21, 2015

*   All **TCGA data** updated to the Firehose run of April 16, 2015.
*   **New feature**: Enrichments Analysis finds alterations that are enriched in either altered or unaltered samples.
*   **Improvement**: improved OncoPrint with better performance.

## June 3, 2015

*   **Improvements**:
    *   Allowed downloading data in each chart/table in study summary page.
    *   Added log-rank test _p_-values to the survival plots in study summary page.
    *   Improved visualization of patient clinical data in patient-centric view.
    *   Added option to merge multiple samples for the same patient in OncoPrint.

## April 28, 2015

*   **New features**:
    *   Redesigned query interface to allow selecting multiple cancer studies
    *   Redesigned **Plots** tab

## January 20, 2015

*   All **TCGA data** updated to the Firehose run of October 17, 2014
*   **COSMIC data** updated to V71
*   **New features**:
    *   Query page: better search functions to find cancer studies
    *   OncoPrints now support color coding of different mutation types
    *   OncoPrints now support multiple clinical annotation tracks
    *   [**OncoPrinter tool**](https://www.cbioportal.org/oncoprinter.jsp) now supports mRNA expression changes  
        ![Oncoprint with multiple clinical tracks](https://user-images.githubusercontent.com/1334004/47188323-9f8e4b00-d305-11e8-8fe9-972104518545.png)

## January 6, 2015

*   **New feature**: You can now view **frequencies of mutations and copy-number alterations** in the study view. These tables are updated dynamically when selecting subsets of samples.  
    ![Alterations in heavily copy-number altered endometrial cancer cases](https://user-images.githubusercontent.com/1334004/47188333-9f8e4b00-d305-11e8-8c64-de5cbda0fa54.png)

## December 9, 2014

*   **New TCGA data**:
    *   Added complete and up-to-date **clinical data** for all **TCGA** provisional studies
    *   All TCGA data updated to the Firehose run of July 15, 2014
    *   New TCGA provisional studies: Esophageal cancer, Pheochromocytoma and Paraganglioma (PCPG)
    *   New published TCGA studies: [Thyroid Cancer](https://www.cbioportal.org/study?id=thca_tcga_pub) and [Kidney Chromophobe](https://www.cbioportal.org/study?id=kich_tcga_pub)
*   **Added data** of 172 samples from 4 published studies:
    *   [Cholangiocarcinoma (National University of Singapore, Nature Genetics 2012)](https://www.cbioportal.org/study?id=chol_nus_2012)
    *   [Cholangiocarcinoma (National Cancer Centre of Singapore, Nature Genetics 2013)](https://www.cbioportal.org/study?id=chol_nccs_2013)
    *   [Intrahepatic Cholangiocarcinoma (Johns Hopkins University, Nature Genetics 2013)](https://www.cbioportal.org/study?id=chol_jhu_2013)
    *   [Bladder Cancer (MSKCC, Eur Urol 2014)](https://www.cbioportal.org/study?id=blca_mskcc_solit_2014)
*   **New features**:
    *   Redesigned **Mutual Exclusivity** tab
    *   Added **correlation scores** for scatter plots on the Plots tab
    *   Download links to [**GenomeSpace**](https://www.genomespace.org/)

## October 24, 2014

*   Added data of 885 samples from 11 published studies:
    *   [Colorectal Adenocarcinoma Triplets (MSKCC, Genome Biology 2014)](https://www.cbioportal.org/study?id=coadread_mskcc)
    *   [Esophageal Squamous Cell Carcinoma (ICGC, Nature 2014)](https://www.cbioportal.org/study?id=escc_icgc)
    *   [Malignant Peripheral Nerve Sheath Tumor (MSKCC, Nature Genetics 2014)](https://www.cbioportal.org/study?id=mpnst_mskcc)
    *   [Melanoma (Broad/Dana Farber, Nature 2012)](https://www.cbioportal.org/study?id=skcm_broad_dfarber)
    *   [Nasopharyngeal Carcinoma (National University Singapore, Nature Genetics 2014)](https://www.cbioportal.org/study?id=npc_nusingapore)
    *   [Prostate Adenocarcinoma CNA study (MSKCC, PNAS 2014)](https://www.cbioportal.org/study?id=prad_mskcc_2014)
    *   [Prostate Adenocarcinoma Organoids (MSKCC, Cell 2014)](https://www.cbioportal.org/study?id=prad_mskcc_cheny1_organoids_2014)
    *   [Stomach Adenocarcinoma (TCGA, Nature 2014)](https://www.cbioportal.org/study?id=stad_tcga_pub)
    *   [Stomach Adenocarcinoma (Pfizer and University of Hong Kong, Nature Genetics 2014)](https://www.cbioportal.org/study?id=stad_pfizer_uhongkong)
    *   [Stomach Adenocarcinoma (University of Hong Kong, Nature Genetics 2011)](https://www.cbioportal.org/study?id=stad_uhongkong)
    *   [Stomach Adenocarcinoma (University of Tokyo, Nature Genetics 2014)](https://www.cbioportal.org/study?id=stad_utokyo)

## August 8, 2014

*   Released two new tools
    *   [Oncoprinter](https://www.cbioportal.org/oncoprinter.jsp) lets you create Oncoprints from your own, custom data
    *   [MutationMapper](https://www.cbioportal.org/mutation_mapper.jsp) draws mutation diagrams (lollipop plots) from your custom data

## May 21, 2014

*   All TCGA data updated to the Firehose run of April 16, 2014

## May 12, 2014

*   Improved study summary page including survival analysis based on clinical attributes  
    e.g. [TCGA Endometrial Cancer cohort](https://www.cbioportal.org/study?id=ucec_tcga_pub)  
    ![Study view](https://user-images.githubusercontent.com/1334004/47188334-a026e180-d305-11e8-964c-6c417bb06ca2.png)

## March 27, 2014

*   New features:
    *   Visualizing of mutations mapped on 3D structures (individual or multiple mutations, directly in the browser)
    *   Gene expression correlation analysis (find all genes with expression correlation to your query genes)
    *   The Patient-Centric View now displays mutation frequencies across all cohorts in cBioPortal for each mutation
    *   The Mutation Details Tab and the Patient-Centric View now display the copy-number status of each mutation  
        ![3D viewer & Co-expression](https://user-images.githubusercontent.com/1334004/47188325-9f8e4b00-d305-11e8-96db-67936cd89f8f.png)

## March 18, 2014

*   All TCGA data updated to the Firehose run of January 15, 2014
*   Updated to the latest COSMIC data (v68)
*   Added two new provisional TCGA studies:
    *   Adrenocortical Carcinoma
    *   Uterine Carcinosarcoma
*   Added mutation data of 898 samples from 11 published studies:
    *   Hepatocellular Carcinoma (RIKEN, Nature Genetics 2012)
    *   Hepatocellular Carcinoma (AMC, Hepatology in press)
    *   Medulloblastoma (Broad, Nature 2012)
    *   Medulloblastoma (ICGC, Nature 2012)
    *   Medulloblastoma (PCGP, Nature 2012)
    *   Multiple Myeloma (Broad, Cancer Cell 2014)
    *   Pancreatic Adenocarcinoma (ICGC, Nature 2012)
    *   Small Cell Carcinoma of the Ovary (MSKCC, Nature Genetics in press)
    *   Small Cell Lung Cancer (CLCGP, Nature Genetics 2012)
    *   Small Cell Lung Cancer (Johns Hopkins, Nature Genetics 2012)
    *   NCI-60 Cell Lines (NCI, Cancer Res. 2012)

## December 9, 2013

*   Added mutation data of 99 bladder cancer samples (BGI, Nature Genetics 2013)

## December 6, 2013

*   Data sets matching four recently submitted or published TCGA studies are now available
    *   Glioblastoma (Cell 2013)
    *   Bladder carcinoma (Nature, in press)
    *   Head & neck squamous cell carcinoma (submitted)
    *   Lung adenocarcinoma (submitted)

## November 8, 2013

*   All TCGA data updated to the Firehose run of September 23, 2013.
*   Updated to the latest COSMIC data (v67).
*   Added mutation data of 792 samples from 9 published cancer studies:
    *   Esophageal Adenocarcinoma (Broad, Nature Genetics 2013)
    *   Head and Neck Squamous Cell Carcinoma (Broad, Science 2011)
    *   Head and Neck Squamous Cell Carcinoma (Johns Hopkins, Science 2011)
    *   Kidney Renal Clear Cell Carcinoma (BGI, Nature Genetics 2012)
    *   Prostate Adenocarcinoma, Metastatic (Michigan, Nature 2012)
    *   Prostate Adenocarcinoma (Broad/Cornell, Nature Genetics 2012)
    *   Prostate Adenocarcinoma (Broad/Cornell, Cell 2013)
    *   Skin Cutaneous Melanoma (Yale, Nature Genetics 2012)
    *   Skin Cutaneous Melanoma (Broad, Cell 2012)

## October 21, 2013

*   Improved interface for survival plots, including information on individual samples via mouse-over
*   New fusion glyph in OncoPrints    [![FGFR3 fusions in head and neck carcinoma](https://user-images.githubusercontent.com/1334004/47188320-9f8e4b00-d305-11e8-9d8c-09a7ec28d921.png)](https://www.cbioportal.org/index.do?cancer_study_id=hnsc_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=hnsc_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=hnsc_tcga_pub_gistic&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=hnsc_tcga_pub_sequenced&case_ids=&gene_set_choice=user-defined-list&gene_list=FGFR3%3A+AMP+MUT%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit)
*   Improved cross-cancer query: new alteration frequency histogram (example below - query gene: CDKN2A) and mutation diagram
    ![Cross Cancer Query](https://user-images.githubusercontent.com/1334004/47188319-9ef5b480-d305-11e8-9281-fb31993b7b42.png)

## September 9, 2013

*   Updated COSMIC data (v66 Release)
*   Improved / interactive visualization on the "Protein changes" tab
*   Enhanced mutation diagrams: color-coding by mutation time and syncing with table filters
*   Addition of DNA cytoband information in the patient view of copy-number changes
*   OncoPrints now allow the display of an optional track with clinical annotation (Endometrial cancer example below)
    ![Oncoprint with clinical track](https://user-images.githubusercontent.com/1334004/47188326-9f8e4b00-d305-11e8-9cf5-52e459d56310.png)

## July 25, 2013

*   Multi-gene correlation plots.
*   Variant allele frequency distribution plots for individual tumor samples.
*   Tissue images for TCGA samples in the patient view, via [Digital Slide Archive](https://cancer.digitalslidearchive.net/). [Example](https://www.cbioportal.org/patient?studyId=ucec_tcga&caseId=TCGA-BK-A0CC&tab=tissueImageTab).

## July 16, 2013

*   All TCGA data updated to the May Firehose run (May 23, 2013).
    *   TCGA Pancreatic Cancer study (provisional) added.

## July 4, 2013

*   Improved rendering of mutation diagrams, including ability to download in PDF format.
*   Improved home page: Searchable cancer study & gene set selectors, data sets selector.

## June 17, 2013

*   Improved interface for correlation plots, including information on individual samples via mouse-over.
*   Gene Details from Biogene are now available in the Network view.
*   Added mutation and copy number data from a new adenoid cystic carcinoma study: Ho et al., Nature Genetics 2013.
*   Added mutation data from 6 cancer studies.
    *   Breast Invasive Carcinoma (Shah et al., Nature 2012)
    *   Breast Invasive Carcinoma (Banerji et al., Nature 2012)
    *   Breast Invasive Carcinoma (Stephens et al., Nature 2012)
    *   Lung Adenocarcinoma (Imielinksi et al., Cell 2012)
    *   Lung Adenocarcinoma (Ding et al., Nature 2008)
    *   Colorectal Cancer (Seshagiri et al., Nature 2012)

## June 4, 2013

*   All TCGA data updated to the April Firehose run (April 21, 2012).

## May 14, 2013

*   Added a published TCGA study: Acute Myeloid Leukemia (TCGA, NEJM 2013).

## April 28, 2013

*   All TCGA data updated to the March Firehose run (March 26, 2012).
*   mRNA percentiles for altered genes shown in patient view.

## April 2, 2013

*   All TCGA data updated to the February Firehose run (February 22, 2012).

## March 28, 2013

*   All TCGA data updated to the January Firehose run (January 16, 2012).
*   Data from a new bladder cancer study from MSKCC has been added (97 samples, Iyer et al., JCO in press).

## February 16, 2013

*   The cBio Portal now contains mutation data from all provisional TCGA projects. Please adhere to [the TCGA publication guidelines](https://cancergenome.nih.gov/abouttcga/policies/publicationguidelines) when using these and any TCGA data in your publications.
*   All data updated to the October Firehose run (October 24, 2012).
*   **Sequencing read counts and frequencies** are now shown in the Mutation Details table when available.
*   Improved OncoPrints, resulting in performance improvements.

## November 21, 2012

*   Major new feature: Users can now visualize **genomic alterations and clinical data of individual tumors**, including:
    *   Summary of **mutations** and **copy-number alterations** of interest
    *   **Clinical trial** information
    *   TCGA **Pathology Reports**
*   New **cancer summary view** (Example [Endometrial Cancer](https://www.cbioportal.org/study?id=ucec_tcga))
*   **Updated drug data** from KEGG DRUG and NCI Cancer Drugs (aggregated by [PiHelper](https://bitbucket.org/armish/pihelper))

## October 22, 2012

*   All data updated to the **Broad Firehose** run from July 25, 2012.
*   **COSMIC data** added to Mutation Details (via Oncotator).
*   All predicted functional impact scores are updated to **Mutation Assessor 2.0**.
*   Users can now base queries on genes in recurrent regions of copy-number alteration (from **GISTIC** via Firehose).
*   The [Onco Query Language (OQL)](https://www.cbioportal.org/onco_query_lang_desc.jsp) now supports queries for specific mutations or mutation types.
*   Data sets added that match the data of all TCGA publications (GBM, ovarian, colorectal, and lung squamous).

## July 18, 2012

*   Mutation data for the TCGA lung squamous cell carcinoma and breast cancer projects (manuscripts in press at Nature).
*   All data updated to the **latest Broad Firehose run** (May 25, 2012).
*   **Drug information** added to the network view (via Drugbank).
*   **Improved cross-cancer** queries: Option to select data types, export of summary graphs.
*   Users can now base queries on frequently mutated genes (from **MutSig** via Firehose).

## May 16, 2012

*   All data updated to the **latest Broad Firehose run** (March 21, 2012).
*   Extended **cross-cancer** functionality, enabling users to query across all cancer studies in our database.
*   New **"build a case"** functionality, enabling users to generate custom case sets, based on one or more clinical attributes.
*   New OncoPrint features, including more **compact OncoPrints**, and support for **RPPA** visualization.

## February 27, 2012

*   All data updated to the **latest Broad Firehose run** (January 24, 2012).
*   Validated mutation data for colorectal cancer.
*   New feature: **Mutation Diagrams** that show mutations in the context of protein domains.
    ![TP53 Mutations in Ovarian Cancer](https://user-images.githubusercontent.com/1334004/47188347-a0bf7800-d305-11e8-8c92-7c8763711338.png)

## January 30, 2012

*   Updated data for several TCGA cancer studies.
*   Some small bug-fixes.

## December 22, 2011

*   **Fourteen new [TCGA](https://cancergenome.nih.gov/) cancer studies**: This includes complete data for TCGA Colorectal Carcinoma and provisional data for thirteen other cancer types in the TCGA production pipeline. Please note that data from these thirteen new cancer types are provisional, not final and do not yet include mutation data. As per NCI guidelines, preliminary mutation data cannot be redistributed until they have been validated.
    ![TCGA](https://cancergenome.nih.gov/PublishedContent/Images/SharedItems/Images/TCGA_54px-Logo.png)

*   **Four new data types**:
    *   Reverse-phase protein array (RPPA) data.
    *   microRNA expression and copy-number (including support for multiple loci)
    *   RNA-Seq based expression data.
    *   log2 copy-number data.
*   Updated TCGA GBM copy-number, expression, and methylation data.
*   New **gene symbol validation** service. You can now use gene aliases and/or Entrez Gene IDs within your gene sets.
*   **Links to IGV** for visualization of DNA copy-number changes.
*   Background information from the [Sanger Cancer Gene Census](https://www.sanger.ac.uk/genetics/CGP/Census/).
*   Two **new [Tutorials](https://www.cbioportal.org/tutorials)** to get you quickly started in using the portal.

## November 14, 2011

*   New and **improved mutation details**, with sorting and filtering capabilities.
*   In collaboration with Bilkent University, we have added a **new Network tab** to our results pages. The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from [Pathway Commons](https://www.pathwaycommons.org).
    ![GBM Network](https://user-images.githubusercontent.com/1334004/47188328-9f8e4b00-d305-11e8-87d7-6336285f7572.png)

## September 3, 2011

*   You can now query across different cancer studies (feature available directly from the home page).
*   Our [MATLAB CGDS Cancer Genomics Toolbox](https://www.cbioportal.org/cgds_r.jsp) is now available. The toolbox enables you to download data from the cBio Portal, and import it directly into MATLAB.
*   The code for the cBio Portal has now been fully open sourced, and made available at [Google Code](https://code.google.com/p/cbio-cancer-genomics-portal/). If you would like to join our open source efforts and make the portal even better, drop us an email.

## March 2, 2011

New plotting features and other improvements:

*   Correlation plots that show the relationship between different data types for individual genes.
*   Survival analysis - assess survival differences between altered and non-altered patient sets.
*   Updated [R Package](https://www.cbioportal.org/cgds_r.jsp) with support for correlation plots and general improvements for retrieving and accessing data in R data frames.
*   The [Web Interface](https://www.cbioportal.org/web_api.jsp) now supports basic clinical data, e.g. survival data.
*   [Networks](https://www.cbioportal.org/networks.jsp) for pathway analysis are now available for download.
    ![Survival Analysis](https://user-images.githubusercontent.com/1334004/47188329-9f8e4b00-d305-11e8-83e2-0ad816431888.png)

## December 15, 2010

Several new features, including:

*   Redesigned and streamlined user interface, based on user feedback and usability testing.
*   Advanced support for gene-specific alterations. For example, users can now view mutations within TP53, and ignore copy number alterations, or only view amplifications of EGFR, and ignore deletions.
*   Improved performance.
*   [Frequently Asked Questions](FAQ) document released.
*   Updated ~~Video Tutorial~~ (update: old link no longer functional. Now see: [YouTube](https://www.youtube.com/results?search_query=cbioportal)

## November 4, 2010

*   Enhanced [Oncoprints](https://www.cbioportal.org/faq.jsp#what-are-oncoprints), enabling users to quickly visualize genomic alterations across many cases. Oncoprints now also work in all major browsers, including Firefox, Chrome, Safari, and Internet Explorer.
*   Official release of our [Web Interface](https://www.cbioportal.org/web_api.jsp), enabling programmatic access to all data.
*   Official release of our [R Package](https://www.cbioportal.org/cgds_r.jsp), enabling programmatic access to all data from the R platform for statistical computing.
    ![OncoPrints](https://user-images.githubusercontent.com/1334004/47188322-9f8e4b00-d305-11e8-9fae-ca188de91267.png)
# Welcome to the documentation for cBioPortal!

We've organized this site by the different roles in the cBioPortal community.

<div class="homeList">
            
            <div>
                <a href="user-guide/">
                <img src="images/user-guide.png" width="125" height="125" alt="User guide  free icon" title="User guide free icon">
                </a>
                <a href="user-guide/">Portal User</a>
            </div>
        
            <div>
            <a href="web-api-and-clients/">
            <img src="images/api.png" width="125" height="125" alt="User guide  free icon" title="User guide free icon">
            </a> 
            <a href="web-api-and-clients/">API User</a>
            </div>
       
            <div>
            <a href="deployment/">
            <img src="images/deployment.png" width="125" height="125" alt="User guide  free icon" title="User guide free icon">
            </a>
            <a href="deployment/">Instance Maintainer</a>
            </div>
        
            <div>
            <a href="development/">
            <img src="images/develop.png" width="125" height="125" alt="User guide  free icon" title="User guide free icon">
            </a>
            <a href="development/">Developer</a>
            </div>


</div>


<div id="feedback">
<h3><strong>Feedback? Questions?</strong></h3>
<br />

[!button cBioPortal Google Group](https://groups.google.com/g/cbioportal)

<hr />

:::rfclist
We maintain an active [list of RFCs (Requests for Comments)](RFC-List.md) where we describe new features and solicit community feedback.
:::

</div>

<hr />

<h3>Quick Links</h3>

- [FAQ](user-guide/faq.md)
- [Tutorials](https://www.cbioportal.org/tutorials)
- [API documentation](web-API-and-Clients.md)


<script>
setTimeout(function(){
    document.querySelectorAll("#docs-sidebar-right .simplebar-content")[0].appendChild(document.getElementById("feedback"));
},500);
</script>



# List of Active RFCs

We maintain an active set of RFCs (Requests for Comments) where we spec out new features and solicit community feedback.

See this [shared google folder](https://drive.google.com/drive/folders/0B6z8AlO4d-qMT0Zpd0J4cW5LbHc) for the list of RFCs.


# For Developers Creating new RFCs

* Use the [RFC Template](https://docs.google.com/document/d/1nXi_to7U7qDR41hmrVtIPZBrRS1z4WI4OiT79s6Pc4I/edit)
* Create your RFC within this [shared google folder](https://drive.google.com/drive/folders/0B6z8AlO4d-qMT0Zpd0J4cW5LbHc), and pick a new incremental number.
* Add a Link to your RFC on this page.

# Session Service Management
## Clear Study View User settings
### **Using kubernetes:**
Go inside session service pod (replace `CBIOPORTAL_SESSION_SERVICE_MONGO_POD_ID` with your session service pod id):
```shell
kubectl exec  -it CBIOPORTAL_SESSION_SERVICE_MONGO_POD_ID  -- /bin/sh -c 'mongo admin -u root -p ${MONGODB_ROOT_PASSWORD}'
```
Choose database:
```shell
use session_service;
```
(Optional) Make sure there are matching records to remove (replace `PORTAL_SOURCE` with correct text, one can find it in `session.service.url`, e.g. `public_portal`) 
```shell
db.settings.find({ "source": "PORTAL_SOURCE", "data.page": "study_view" }).count();
```
Remove user settings in study view:
```shell
db.settings.remove({ "source": "PORTAL_SOURCE", "data.page": "study_view" });
```

### **Using docker-compose:**
Go inside session service container (replace `CBIOPORTAL_SESSION_SERVICE_MONGO_CONTAINER_ID` with your session service database container id):
```shell
docker exec  -it CBIOPORTAL_SESSION_SERVICE_MONGO_CONTAINER_ID  bash -c 'mongo'
```
Then follow the same procedure described in the `Using kubernetes` section.The cBioPortal is open source software, available on [GitHub](https://github.com/cbioportal).  If you are interested in joining our efforts, please check out our documentation on [how to contribute](https://github.com/cBioPortal/cbioportal/blob/master/CONTRIBUTING.md).

As an open source project, we are grateful for all the many great tools and services that keep our project humming.

We gratefully acknowledge a large collection of open source languages, tools and frameworks, including:

* [Java](https://www.java.com/), [Python](https://www.python.org/) and [Typescript](https://www.typescriptlang.org/)
* [React](https://reactjs.org/) and [MobX](https://mobx.js.org/)
* [Bootstrap](https://getbootstrap.com/) and [React-bootstrap](https://react-bootstrap.github.io/)
* [Yarn](https://yarnpkg.com/), [npm](https://www.npmjs.com/) and [webpack](https://webpack.js.org/)
* [Mocha](https://mochajs.org/), [Chai](https://www.chaijs.com/) and [Selenium](https://selenium.dev/)
* [D3](https://d3js.org/), [Victory.js](https://formidable.com/open-source/victory/) and [Chart JS](https://www.chartjs.org/)
* [Spring](https://spring.io/) and [Hibernate](https://hibernate.org/)
* [MySQL](https://www.mysql.com/) and [MongoDB](https://www.mongodb.com/)

We also gratefully acknowledge the following service providers:

* [GitHub](https://github.com) for source control management and community engagement.
* [Slack](https://slack.cbioportal.org/)  for coordinating work across a geographically dispersed team.
* [Travis CI](https://travis-ci.org/) and [Circle CI](https://circleci.com/) for continuous integration.
* [Netlify](https://www.netlify.com/), [Heroku](https://www.heroku.com/) and [Amazon Web Services](https://aws.amazon.com/) for cloud hosting services.# Study Curation Guide
This guide is to help data curators learn how to curate a study on their own computer

## Prerequisites
- To follow this guide the curator should have some familiarity with running commands on the command line. We will be using Docker. No Docker knowledge is required, one will obtain some basic understanding by following the guide.
- Learn how to setup cBioPortal locally [here](docker/README.md) first.

## Load an example study
After having followed the steps in the [Docker Deployment instructions](docker/README.md), you will end up with the study [Low-Grade Gliomas (UCSF, Science 2014)
](https://www.cbioportal.org/study/summary?id=lgg_ucsf_2014) loaded locally. Now let's try to import another study:

1. Choose another example study from the [datahub](https://github.com/cBioPortal/datahub/tree/master/public). Note the name of the folder which is identical to the study id defined in [meta_study.txt](https://github.com/cBioPortal/datahub/blob/master/public/lgg_ucsf_2014/meta_study.txt#L2).
2. From the root of the `cbioportal-docker-compose` folder run `DATAHUB_STUDIES=my_study_id ./study/init.sh`. Change `my_study_id` to the study you picked in 1. The study should now be downloaded in `./study/`.
3. Import the study by running `docker-compose exec cbioportal metaImport.py -u http://cbioportal:8080 -s study/my_study_id/ -o`. Again change `my_study_id` to the study you picked in 1. This should import the study.
4. Restart the cbioportal instance `docker-compose restart cbioportal` and see if the new study shows up on http://localhost:8080

## Curate a new study
The cBioPortal team has curated many published studies in formats suitable for import in cBioPortal. These can
be found on the [datahub](https://github.com/cbioportal/datahub) and can serve as an example of how our curation
processes works. You can find a step by step description of how to curate a new study [here](https://github.com/cBioPortal/datahub/blob/master/docs/curation-process.md).
# Table of contents

* [Home](README.md)
* [About Us](About-Us.md)
* [List of RFCs](RFC-List.md)
* [User Guide](user-guide/README.md)
  * [New Users](user-guide/new-users.md)
  * [Frequently Asked Questions](user-guide/faq.md)
  * [Tutorials](user-guide/overview.md)
  * [Page Specific Resources](user-guide/by-page.md)
  * [OQL](user-guide/oql.md)
  * [News](News.md)
  * [Genie News](News-Genie.md)
* [Web API and API Clients](web-API-and-Clients.md)
* [Deployment](deployment/README.md)
  * [Architecture overview](Architecture-Overview.md)
  * [Hardware Requirements](Hardware-Requirements.md)
  * [Deploy with Docker (Recommended)](deployment/docker/README.md)
    * [Import data with Docker](deployment/docker/import\_data.md)
    * [Example Commands](deployment/docker/example\_commands.md)
    * [Authenticating and Authorizing Users using Keycloak in Docker](deployment/docker/using-keycloak.md)
  * [Deploy without Docker](deployment/deploy-without-docker/README.md)
    * [Software Requirements](deployment/deploy-without-docker/Software-Requirements.md)
    * [Pre-Build Steps](deployment/deploy-without-docker/Pre-Build-Steps.md)
    * [Building from Source](deployment/deploy-without-docker/Build-from-Source.md)
    * [Importing the Seed Database](deployment/deploy-without-docker/Import-the-Seed-Database.md)
    * [Deploying the Web Application](deployment/deploy-without-docker/Deploying.md)
    * [Loading a Sample Study](deployment/deploy-without-docker/Load-Sample-Cancer-Study.md)
  * [Authorization and Authentication](deployment/authorization-and-authentication/README.md)
    * [User Authorization](deployment/authorization-and-authentication/User-Authorization.md)
    * [Authenticating Users via SAML](deployment/authorization-and-authentication/Authenticating-Users-via-SAML.md)
    * [Authenticating Users via LDAP](deployment/authorization-and-authentication/Authenticating-Users-via-LDAP.md)
    * [Authenticating and Authorizing Users via Keycloak](deployment/authorization-and-authentication/Authenticating-and-Authorizing-Users-via-keycloak.md)
    * [Authenticating Users via Tokens](deployment/authorization-and-authentication/Authenticating-Users-via-Tokens.md)
  * [Customization]()
    * [Customizing your cBioPortal Instance via application.properties](deployment/customization/Customizing-your-instance-of-cBioPortal.md)
    * [More application.properties Settings](deployment/customization/application.properties-Reference.md)
    * [Security Properties Settings](deployment/customization/security.properties-Reference.md)
    * [Configuring Caching Behavior](deployment/customization/Caching.md)
    * [How does the study view organize the charts](deployment/customization/Studyview.md)
    * [Utilize priority setting from database to visualize charts](deployment/customization/Studyview.md)
  * [Integration with Other Webservices]()
    * [OncoKB Data Access](deployment/integration-with-other-webservices/OncoKB-Data-Access.md)
  * [Data Loading Overview](Data-Loading.md)
    * [Downloads](Downloads.md)
    * [Using the Dataset Validator](Using-the-dataset-validator.md)
    * [Using the metaImport Script](Using-the-metaImport-script.md)
    * [File Formats](File-Formats.md)
      * [Z-Score Normalization](Z-Score-normalization-script.md)
    * [Study Curation Guide](Study-Curation-Guide.md)
    * [Maintaining Studies](Data-Loading-Maintaining-Studies.md)
    * [For Developers](Data-Loading-For-Developers.md)
    * [Tips and Best Practices](Data-Loading-Tips-and-Best-Practices.md)
    * [Mutation Data Transcript Annotation](mutation-data-transcript-annotation.md)
    * [Import OncoKB annotations as custom driver annotations](Import-OncoKB-annotations-as-Custom-Driver-Annotations.md)
    * [Importing Gene Sets for Gene Set Scoring Data](Import-Gene-Sets.md)
    * [Importing Gene Panels](Import-Gene-Panels.md)
    * [Study View Customization with Priority Data](deployment/customization/Studyview.md)
  * [cBioPortal Maintenance](deployment/cbioportal-maintenance/README.md)
    * [Updating your cBioPortal Database Scheme](Updating-your-cBioPortal-installation.md)
    * [Update genes and gene aliases](Updating-gene-and-gene_alias-tables.md)
    * [Migration Guide](Migration-Guide.md)
    * [MSK Maintenance](MSK-Maintenance.md)
* [Development](development/README.md)
  * [Feature Development Guide](development/feature-development-guide.md)
  * [cBioPortal Entity-relationship Diagram](development/cBioPortal-ER-Diagram.md)
  * [cBioPortal Database Versioning](development/Database-Versioning.md)
  * [Build cBioPortal with a different frontend version](development/Build-Different-Frontend.md)
  * [Session Service](development/session-service-working.md)
  * [Manual test cases](development/manual-test-cases.md)
  * [Release Procedure](development/Release-Procedure.md)
  * [Deployment Procedure](development/Deployment-Procedure.md)
  * [This documentation site](development/Documentation-site.md)

# Introduction

This page is about how to run and debug, in an IDE agnostic way, any test classes added to the backend (Java layer).

# Steps

## Set up test instance of MySQL server

### System installed (assumes Linux OS)

1. Start mysql:

```
sudo systemctl start mysql
```

2. Log into database:

```shell
sudo mysql -u root
```

or

```shell
mysql -u root -p
```

3. In the MySQL console run:

```shell
SET GLOBAL local_infile=1;
CREATE DATABASE cgds_test;
CREATE USER 'cbio_user'@'localhost' IDENTIFIED BY 'somepassword';
GRANT ALL ON cgds_test.* TO 'cbio_user'@'localhost';
FLUSH PRIVILEGES;
SET default_storage_engine=InnoDB;
SET SESSION sql_mode = 'ANSI_QUOTES';
```

4. (Optional) Load schema and seed:

When running tests through Maven this step can be skipped. In MySQL console run:

```shell
source <path-to-file>/cgds-test.sql
source <path-to-file>/seed_mini.sql
```

#### Note

_Cgds-test.sql_ should be generated with `/db-scripts/src/main/resources/gen-cgds-test-schema.sh`.

## Docker container

1. Create file `init.sql` with the following contents:

```shell
SET GLOBAL local_infile=1;
CREATE DATABASE cgds_test;
CREATE USER 'cbio_user'@'localhost' IDENTIFIED BY 'somepassword';
GRANT ALL ON cgds_test.* TO 'cbio_user'@'localhost'";
FLUSH PRIVILEGES;
SET default_storage_engine=InnoDB;
SET SESSION sql_mode = 'ANSI_QUOTES';
```

2. Run (making sure the file paths are correct):

```shell
 docker run -ti --rm \
  --name testdb \
  --env MYSQL_ROOT_PASSWORD=root \
  --env MYSQL_USER=cbio_user \
  --env MYSQL_PASSWORD=somepassword \
  --env MYSQL_DATABASE=cgds_test \
  -v <path-to-file>/init.sql:/docker-entrypoint-initdb.d/first_file.sql \
  -v <path-to-file>/cgds-test.sql:/docker-entrypoint-initdb.d/second_file.sql \
  -v <path-to-file>/seed_mini.sql:/docker-entrypoint-initdb.d/third_file.sql \
  -p 3306:3306 mysql:5.7
  --ssl=0
```

#### Notes

- When running tests through Maven, the line with volume mounts (starting with `-v`) to the schema (_cgds-test.sql_) and
  seed (_seed_mini.sql_) should be skipped.
- _Cgds-test.sql_ should be generated with `/db-scripts/src/main/resources/gen-cgds-test-schema.sh`.

## Running the tests with Maven

1. Make sure you have built the project once with e.g. `mvn clean install -DskipTests`
2. Run integration tests with

```
mvn integration-test -Ddb.test.username=cbio_user -Ddb.test.password=somepassword
```

This will create the tables in `cgds_test` and populate your schema with test data.

3. you can run all other tests with

```
mvn test -Ddb.test.username=cbio_user -Ddb.test.password=somepassword
```

4. you can run a specific test with

```
mvn test -pl core -Dtest=TestIntegrationTest -Ddb.test.username=cbio -Ddb.test.password= <your_db_password>
```

where `-pl` is the name of the module (in this example `core`) and `-Dtest` is the name of the Java test class (in this
example `TestIntegrationTest`).

5. you can debug a specific test with

```
mvn integration-test -Dmaven.surefire.debug -pl core test -Dtest=TestIntegrationTest -Ddb.test.username=cbio -Ddb.test.password=<your_db_password> 
```

where `-pl` is the name of the module (in this example `core`) and `-Dtest` is the name of the Java test class (in this
example `TestIntegrationTest`). This command will pause execution and wait for you to connect your IDE to the listening
port, reporting something like below:

```
[INFO] Surefire report directory: /cbioportal/core/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Listening for transport dt_socket at address: 5005
```

Once you connected your IDE (after setting a breakpoint in the code), the execution will continue in your debugger view.
# Uninstall Docker cBioPortal

## 1. Remove cBioPortal docker container

In a docker terminal type the following command:

```bash
docker rm -fv ${CONTAINER_NAME}
```

Where:    
- **${CONTAINER_NAME}**: The name of your container instance, _i.e **cbioportal**_.

## 2. Remove MySQL docker container

In a docker terminal type the following command:

```bash
docker rm -fv ${CONTAINER_NAME}
```

Where:    
- **${CONTAINER_NAME}**: The name of your container instance, _i.e **cbio-DB**_.

## 3. Remove the docker network

In a docker terminal type the following command:

```bash
docker network rm ${DOCKER_NETWORK_NAME}
```

Where:
- **${DOCKER_NETWORK_NAME}** is the name of the network that cBioPortal and the cBioPortal DB are accessible. _i.e **cbio-net**_# Updating the gene names and aliases tables
This manual is intended for users that have knowledge about the structure of the cBioPortal seed database.

When loading studies into cBioPortal it is possible for warnings to occur that are caused by an outdated seed database. Gene symbols can be deprecated or be assigned to a different Entrez Gene in a new release. Also Entrez Gene IDs can be added. This markdown explains how to update the seed database, in order to use the most recent Entrez Gene IDs.

The cBioPortal scripts package provides a method to update the `gene` and `gene_alias` tables.

## Prepare

### Human genes
Download [gene_info.txt](https://github.com/cBioPortal/datahub-study-curation-tools/blob/master/gene-table-update/build-input-for-importer/gene_info.txt)  
Generated based on latest HGNC release using script [HERE](https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/gene-table-update/build-input-for-importer) 

### Mouse genes
Download `Mus_musculus.gene_info.gz` from   
ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Mus_musculus.gene_info.gz
Unzip the downloaded file with the command `gunzip Mus_musculus.gene_info.gz`

## MySQL steps
Execute these steps in case you want to reset your database to the most recent genes list from NCBI.

1- Start a new MySQL database with the previous seed database, which can be found on cBioPortal Datahub for [human](https://github.com/cBioPortal/datahub/tree/master/seedDB) and [mouse](https://github.com/cBioPortal/datahub/tree/master/seedDB_mouse).

2- If DB engine supports foreign key (FK) constraints, e.g. InnoDB, drop constraints:
```sql
ALTER TABLE cosmic_mutation
  DROP FOREIGN KEY cosmic_mutation_ibfk_1;

ALTER TABLE uniprot_id_mapping
  DROP FOREIGN KEY uniprot_id_mapping_ibfk_1;
```

3- Empty tables `gene` and `gene_alias`
```sql
TRUNCATE TABLE gene_alias;
DELETE from genetic_entity;
DELETE from geneset_hierarchy_node;
ALTER TABLE `genetic_entity` AUTO_INCREMENT = 1;
ALTER TABLE `geneset_hierarchy_node` AUTO_INCREMENT = 1;
ALTER TABLE `geneset` AUTO_INCREMENT = 1;
```

4- Restart cBioPortal (restart webserver)  or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information)
to clean-up any cached gene lists.

5- To import gene data type the following commands when in the folder `<cbioportal_source_folder>/core/src/main/scripts`:
### Human genes
```
export PORTAL_HOME=<cbioportal_configuration_folder>
export JAVA_HOME=<jre_installation_folder>
./importGenes.pl --hgnc <gene_info.txt> --genome-build <GRCh37>
```
### Mouse genes
```
export PORTAL_HOME=<cbioportal_configuration_folder>
export JAVA_HOME=<jre_installation_folder>
./importGenes.pl --genes <Mus_musculus.gene_info.gz> --gtf <gencode.v25.annotation.gtf> --genome-build <GRCh37>
```
**IMPORTANT NOTE**: 
1. The **reference_genome** table needs to be populated before updating the **gene** table. Further details can be found in [this document](import-reference-genome.md). 
2. Use **--species** option when importing genes for a species other than human
3. Use the **gene** table if you query information such as hugo symbols, types of the gene 
4. Use **reference_genome_gene** table if you query information such as chromosome, cytoband, exonic length, or the start or end of the gene
5. Load genes only to the **reference_genome_gene** table without updating the **gene** table, please use the following command:
### Human genes
```
./importGenes.pl --gtf <gencode.v27.annotation.gtf> --genome-build <GRCh38>
```
### Mouse genes
```
./importGenes.pl --gtf <gencode.v27.annotation.gtf> --genome-build <GRCh38>
```
6- :warning: Check the `gene` and `gene_alias` tables to verify that they are filled correctly.
```sql
SELECT count(*) FROM cbioportal.gene;
SELECT count(*) FROM cbioportal.gene_alias;
```

7- Additionally, there are other tables you may want to update now (only in human).

* Updating the COSMIC coding mutations, can be downloaded from [here](https://cancer.sanger.ac.uk/cosmic/download) and require the script `importCosmicData.pl`

8- Clean-up old data:
```sql
SET SQL_SAFE_UPDATES = 0;
DELETE FROM cosmic_mutation where ENTREZ_GENE_ID not in (SELECT ENTREZ_GENE_ID from gene);
DELETE FROM sanger_cancer_census where ENTREZ_GENE_ID not in (SELECT ENTREZ_GENE_ID from gene);
DELETE FROM uniprot_id_mapping where ENTREZ_GENE_ID not in (SELECT ENTREZ_GENE_ID from gene);
DELETE FROM interaction where GENE_A not in (SELECT ENTREZ_GENE_ID from gene) or GENE_B not in (SELECT ENTREZ_GENE_ID from gene);
DELETE FROM drug_interaction where target not in (SELECT ENTREZ_GENE_ID from gene);
SET SQL_SAFE_UPDATES = 1;
commit;
```

9- If DB engine supports FK constraints, e.g. InnoDB, restore constraints:
```sql
ALTER TABLE cosmic_mutation
  ADD CONSTRAINT cosmic_mutation_ibfk_1 FOREIGN KEY (`ENTREZ_GENE_ID`) REFERENCES `gene` (`ENTREZ_GENE_ID`);

ALTER TABLE uniprot_id_mapping
  ADD CONSTRAINT uniprot_id_mapping_ibfk_1 FOREIGN KEY (`ENTREZ_GENE_ID`) REFERENCES `gene` (`ENTREZ_GENE_ID`);
```

10- You can import new gene sets using the gene set importer. These gene sets are currently only used for gene set scoring. See [Import-Gene-Sets.md](Import-Gene-Sets.md) and [File-Formats.md#gene-set-data](File-Formats.md#gene-set-data).

For example, run in folder `<cbioportal_source_folder>/core/src/main/scripts`:
```bash
./importGenesetData.pl --data ~/Desktop/msigdb.v6.1.entrez.gmt --new-version msigdb_6.1
```
Please make sure the version gene sets is the same as the version used to calculate gene set scores in your data.
# Updating your cBioPortal installation

As of release **1.1.0** cBioPortal has a Database schema update mechanism which comes into play whenever the new version of the portal code relies on specific DB schema changes to be applied. The portal will automatically check if the DB schema is according to what it expects. It does so by comparing the version number of the portal code with the version number of the DB schema. If they are equal, it assumes the DB schema has been upgraded. If not, it will require the administrator to run a migration script. Below are the steps to ensure your DB schema is updated correctly.

## First time

The first time you update from release **1.0.4** (or lower) to release **1.1.0** (or higher), you should get a an error banner page after restarting your webserver. The error should state something like: 

```
Current DB schema version: xxx 
DB schema version expected by Portal: yyy
```
where `xxx` and `yyy` will be different version numbers. 

If you get `DB version expected by Portal: 0` (i.e. you are building the new release from source), you need to  add a new property to your `application.properties` file which is needed for this check. 

#### Step1

In your `application.properties` file (e.g. `<your_cbioportal_dir>/src/main/resources/application.properties`) add the following property:
```
# this is the *expected* DB version (expected by the code). Don't set it manually, it is filled by maven:
db.version=${db.version}
```

#### Step2

Compile your code again. After restarting the webserver the page should now state something like: `DB version expected by Portal: 1.1.0` (or higher), while the DB version remains as `Current DB version: -1`. 

## Running the migration script

First, make sure you have the DB connection properties correctly set in your application.properties file (see [DB connection settings here](/deployment/customization/application.properties-Reference.md#database-settings)).

**Dependencies:** the migration script is a Python script that depends on the `mysqlclient` library. If necessary, you can install it with the following commands (example for Ubuntu):
```console
sudo apt-get install python3-dev default-libmysqlclient-dev
sudo python3 -m pip install mysqlclient
```

For macOS, try the following:

```
brew install mysql-connector-c
sudo python3 -m pip install mysqlclient
```
and see <https://github.com/PyMySQL/mysqlclient-python/blob/master/README.md#prerequisites>
if problems occur during installation.

To run the migration script first go to the scripts folder
`<your_cbioportal_dir>/core/src/main/scripts` 
and then run the following command:
```console
$ python migrate_db.py
```
This should result in the following output:
```console
WARNING: This script will alter your database! Be sure to back up your data before running.
Continue running DB migration? (y/n) y
Running statments for version: 1.0.0
	Executing statement: CREATE TABLE info (DB_SCHEMA_VERSION VARCHAR(8));
	Executing statement: INSERT INTO info VALUES ("1.0.0");
Running statments for version: 1.1.0
	Executing statement: CREATE TABLE sample_list LIKE patient_list;
	Executing statement: INSERT sample_list SELECT * FROM patient_list;
	Executing statement: CREATE TABLE sample_list_list LIKE patient_list_list;
	Executing statement: INSERT sample_list_list SELECT * FROM patient_list_list;
	Executing statement: ALTER TABLE sample_list_list CHANGE PATIENT_ID SAMPLE_ID INT(11);
	Executing statement: UPDATE info SET DB_SCHEMA_VERSION="1.1.0";
...
etc
```

**Final step:** Restart your webserver or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information).
## Resources by Page
### Study View
* Tutorial Slides: Single Study Exploration [Google slides](https://docs.google.com/presentation/d/1_OGK69lO4Z62WaxHHkNYmWvY0LQN2v0slfaLyY1_IQ0) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%201%20Single%20Study%20Exploration.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=760s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
* How-To Video: Comparing samples based on expression level of a gene [youtube.com](https://www.youtube.com/watch?v=HTiKUXk0j0s)
* Tutorial Slides: Virtual Studies [Google slides](https://docs.google.com/presentation/d/1rQE5rbFNdmup-rAtySHFxlLp3i4qa8SBA7MiQpMdn1I) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%204%20Virtual%20Studies.pdf)


### Group Comparison
* Tutorial Slides: Group Comparison [Google slides](https://docs.google.com/presentation/d/1P2boDph8IfpvjxoxDj_496CLHGtshzJnbbZhszPsmf4) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
* Webinar: Group Comparison [youtube.com](https://www.youtube.com/watch?v=Tx4HZCrIe5c) | [bilibili.com](https://www.bilibili.com/video/BV1VZ4y1W76p)
* How-To Video: Comparing samples based on expression level of a gene [youtube.com](https://www.youtube.com/watch?v=HTiKUXk0j0s)

### Running a Query / Results View
#### General
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=1462s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)


#### OQL
* Documentation [OQL](/user-guide/oql.md)
* Tutorial Slides: Onco Query Language (OQL) [Google slides](https://docs.google.com/presentation/d/1U39xgVujtBodwW20qIfcGu4E5n2zzaKkl2KmzzHqj4A) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
* Webinar: Expression Data Analysis [youtube.com](https://www.youtube.com/watch?v=YUxVv6pkxD4&t=1559s) | [bilibili.com](https://www.bilibili.com/video/BV1HK4y1t7dE)
* How-To Video: Using Onco Query Language (OQL) to query based on the expression level of genes [youtube.com](https://www.youtube.com/watch?v=kHlFXw2TMzc)


#### OncoPrint
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=1682s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Cancer Types Summary
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2024s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Mutual Exclusivity
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=1822s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Plots
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2060s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
#### Mutations
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2149s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
* Webinar: Mutation Details & Patient View [youtube.com](https://www.youtube.com/watch?v=uJsp9kd2jIk) | [bilibili.com](https://www.bilibili.com/video/BV1Qf4y1m7Lx)
* How-To Video: Filtering and adding clinical data to Mutations tab [youtube.com](https://www.youtube.com/watch?v=q9No2073c5o)

#### Co-expression
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2277s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Comparison/Survival
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2334s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
* Tutorial Slides: Group Comparison [Google slides](https://docs.google.com/presentation/d/1P2boDph8IfpvjxoxDj_496CLHGtshzJnbbZhszPsmf4) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
* Webinar: Group Comparison [youtube.com](https://www.youtube.com/watch?v=Tx4HZCrIe5c) | [bilibili.com](https://www.bilibili.com/video/BV1VZ4y1W76p)

#### CN Segments
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2449s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Pathways
* Tutorial Slides: Pathways [Google slides](https://docs.google.com/presentation/d/1O5WGucz0lrfdY25b5QS6zaID_26i434EYXBluqZfT2g) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%207%20Pathways.pdf)
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2512s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Downloads
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2543s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

### Patient View
* Tutorial Slides: Patient View [Google slides](https://docs.google.com/presentation/d/1Jr_2yEfgjKBn4DBiXRk4kmhIbtsRp6gd0iD3k1fIUUk) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%203%20Patient%20View.pdf)
* Tutorial Slides: Pathways [Google slides](https://docs.google.com/presentation/d/1O5WGucz0lrfdY25b5QS6zaID_26i434EYXBluqZfT2g) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%207%20Pathways.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=1022s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
* Webinar: Mutation Details & Patient View [youtube.com](https://www.youtube.com/watch?v=uJsp9kd2jIk) | [bilibili.com](https://www.bilibili.com/video/BV1Qf4y1m7Lx)
* How-To Video: Exploring the longitudinal evolution of individual patients [youtube.com](https://www.youtube.com/watch?v=Hbbs-tHh9LQ)
# cBioPortal FAQs


* [General Questions](#general-questions)
    * [What is the cBioPortal for Cancer Genomics?](/user-guide/faq.md#what-is-the-cbioportal-for-cancer-genomics)
    * [How do I get started?](/user-guide/faq.md#how-do-i-get-started)
    * [What data types are in the portal?](/user-guide/faq.md#what-data-types-are-in-the-portal)
    * [What does ___ stand for?](#what-does-___-stand-for)
    * [What is the process of data curation?](/user-guide/faq.md#what-is-the-process-of-data-curation)
    * [How do I get updates on new portal developments and new data sets?](/user-guide/faq.md#how-do-i-get-updates-on-new-portal-developments-and-new-data-sets)
    * [Does the portal work on all browsers and operating systems?](/user-guide/faq.md#does-the-portal-work-on-all-browsers-and-operating-systems)
    * [How do I cite the cBioPortal?](/user-guide/faq.md#how-do-i-cite-the-cbioportal)
    * [Can I use figures from the cBioPortal in my publications or presentations?](/user-guide/faq.md#can-i-use-figures-from-the-cbioportal-in-my-publications-or-presentations)
    * [Can I save or bookmark my results in cBioPortal?](/user-guide/faq.md#can-i-save-or-bookmark-my-results-in-cbioportal)
    * [How is the cBioPortal for Cancer Genomics different from the Genomic Data Commons (GDC)?](/user-guide/faq.md#how-is-the-cbioportal-for-cancer-genomics-different-from-the-genomic-data-commons-gdc)
    * [Does the cBioPortal provide a Web Service API? R interface? MATLAB interface?](/user-guide/faq.md#does-the-cbioportal-provide-a-web-service-api-r-interface-matlab-interface)
    * [Can I use cBioPortal with my own data?](/user-guide/faq.md#can-i-use-cbioportal-with-my-own-data)
    * [Can I create a local instance of cBioPortal to host my own data?](/user-guide/faq.md#can-i-create-a-local-instance-of-cbioportal-to-host-my-own-data)
    * [I'd like to contribute code to the cBioPortal. How do I get started?](/user-guide/faq.md#id-like-to-contribute-code-to-the-cbioportal-how-do-i-get-started)
    * [What is a combined Study?](/user-guide/faq.md#what-is-a-combined-study)
    * [How can I form a combined Study?](/user-guide/faq.md#how-can-i-form-a-combined-study)
    * [How can I create a subset or sub cohort of a study with specific samples or patients?](/user-guide/faq.md#how-can-i-create-a-subset-or-sub-cohort-of-a-study-with-specific-samples-or-patients)
    * [What is a Virtual Study?](/user-guide/faq.md#what-is-a-virtual-study)
    * [Is it necessary to log in to use virtual studies? If I do log in, what additional functionality do I gain?](/user-guide/faq.md#is-it-necessary-to-log-in-to-use-virtual-studies-if-i-do-log-in-what-additional-functionality-do-i-gain)
    * [What is Group Comparison?](/user-guide/faq.md#what-is-group-comparison)
    * [What are the statistical significance tests in Group Comparison?](/user-guide/faq.md#what-are-the-statistical-significance-tests-in-group-comparison)
* [Data Questions](/user-guide/faq.md#data-questions)
    * [General Data](/user-guide/faq.md#general-data)
        * [Does the portal contain cancer study X?](/user-guide/faq.md#does-the-portal-contain-cancer-study-x)
        * [Which resources are integrated for variant annotation?](/user-guide/faq.md#which-resources-are-integrated-for-variant-annotation)
        * [What version of the human reference genome is being used in cBioPortal?](/user-guide/faq.md#what-version-of-the-human-reference-genome-is-being-used-in-cbioportal)
        * [How does cBioPortal handle duplicate samples or sample IDs across different studies?](/user-guide/faq.md#how-does-cbioportal-handle-duplicate-samples-or-sample-ids-across-different-studies)
        * [Are there any normal tissue samples available through cBioPortal?](/user-guide/faq.md#are-there-any-normal-tissue-samples-available-through-cbioportal)
        * [How can I find which studies have mRNA expression data (or any other specific data type)?](/user-guide/faq.md#how-can-i-find-which-studies-have-mrna-expression-data-or-any-other-specific-data-type)
        * [Can I download all data at once?](/user-guide/faq.md#can-i-download-all-data-at-once)
        * [The data today is different than the last time i looked. What happened?](/user-guide/faq.md#the-data-today-is-different-than-the-last-time-i-looked-what-happened)
        * [How do I access data from AACR Project GENIE?](/user-guide/faq.md#how-do-i-access-data-from-aacr-project-genie)
    * [TCGA](/user-guide/faq.md#tcga)
        * [How does TCGA data in cBioPortal compare to TCGA data in Genome Data Commons?](/user-guide/faq.md#how-does-tcga-data-in-cbioportal-compare-to-tcga-data-in-genome-data-commons)
        * [What happened to TCGA Provisional datasets?](/user-guide/faq.md#what-happened-to-tcga-provisional-datasets)
        * [What are TCGA Firehose Legacy datasets and how do they compare to the publication-associated datasets and the PanCancer Atlas datasets?](/user-guide/faq.md#what-are-tcga-firehose-legacy-datasets-and-how-do-they-compare-to-the-publication-associated-datasets-and-the-pancancer-atlas-datasets)
        * [Where do the thresholded copy number call in TCGA Firehose Legacy data come from?](/user-guide/faq.md#where-do-the-thresholded-copy-number-call-in-tcga-firehose-legacy-data-come-from)
        * [Which studies have MutSig and GISTIC results? How do these results compare to the data in the TCGA publications?](/user-guide/faq.md#which-studies-have-mutsig-and-gistic-results-how-do-these-results-compare-to-the-data-in-the-tcga-publications)
        * [How can I download the PanCancer Atlas data?](/user-guide/faq.md#how-can-i-download-the-pancancer-atlas-data)
    * [DNA (Mutations, Copy Number &amp; Fusions)](/user-guide/faq.md#dna-mutations-copy-number--fusions)
        * [Does the cBioPortal contain synonymous mutation data?](/user-guide/faq.md#does-the-cbioportal-contain-synonymous-mutation-data)
        * [What processing or filtering is applied to generate the mutation data?](/user-guide/faq.md#what-processing-or-filtering-is-applied-to-generate-the-mutation-data)
        * [What transcripts are used for annotating mutations?](/user-guide/faq.md#what-transcripts-are-used-for-annotating-mutations)
        * [How are protein domains in the mutational lollipop diagrams specified?](/user-guide/faq.md#how-are-protein-domains-in-the-mutational-lollipop-diagrams-specified)
        * [What is the difference between a ‚Äúsplice site‚Äù mutation and a ‚Äúsplice region‚Äù mutation?](/user-guide/faq.md#what-is-the-difference-between-a-splice-site-mutation-and-a-splice-region-mutation)
        * [What do ‚ÄúAmplification‚Äù, ‚ÄúGain‚Äù, ‚ÄúDeep Deletion‚Äù, ‚ÄúShallow Deletion‚Äù and "-2", "-1", "0", "1", and "2" mean in the copy-number data?](/user-guide/faq.md#what-do-amplification-gain-deep-deletion-shallow-deletion-and--2--1-0-1-and-2-mean-in-the-copy-number-data)
        * [What is GISTIC? What is RAE?](/user-guide/faq.md#what-is-gistic-what-is-rae)
    * [RNA](/user-guide/faq.md#rna)
        * [Does the portal store raw or probe-level data?](/user-guide/faq.md#does-the-portal-store-raw-or-probe-level-data)
        * [What are mRNA and microRNA Z-Scores?](/user-guide/faq.md#what-are-mrna-and-microrna-z-scores)
        * [Is there any normal RNA-seq data in cBioPortal?](/user-guide/faq.md#is-there-any-normal-rna-seq-data-in-cbioportal)
        * [How is TCGA RNASeqV2 processed? What units are used?](/user-guide/faq.md#how-is-tcga-rnaseqv2-processed-what-units-are-used)
        * [Is there microRNA data?](/user-guide/faq.md#is-there-microrna-data)
        * [How can I query microRNAs in the portal?](/user-guide/faq.md#how-can-i-query-micrornas-in-the-portal)
    * [Protein](/user-guide/faq.md#protein)
        * [How can I query phosphoprotein levels in the portal?](/user-guide/faq.md#how-can-i-query-phosphoprotein-levels-in-the-portal)
        * [Why isn‚Äôt there protein data for my gene of interest?](/user-guide/faq.md#why-isnt-there-protein-data-for-my-gene-of-interest)
    * [DNA Methylation](/user-guide/faq.md#dna-methylation)
        * [Which methylation probe is used for genes with multiple probes?](/user-guide/faq.md#which-methylation-probe-is-used-for-genes-with-multiple-probes)
    * [Clinical Data](/user-guide/faq.md#clinical-data)
        * [What kind of clinical data is stored in the portal?](/user-guide/faq.md#what-kind-of-clinical-data-is-stored-in-the-portal)
        * [What is the meaning of OS_STATUS / OS_MONTHS, and PFS_STATUS / PFS_MONTHS?](/user-guide/faq.md#what-is-the-meaning-of-os_status--os_months-and-pfs_status--pfs_months)
* [Analysis Questions](/user-guide/faq.md#analysis-questions)
    * [How can I query/explore a select subset of samples?](/user-guide/faq.md#how-can-i-queryexplore-a-select-subset-of-samples)
    * [How can I compare two or more subsets of samples?](/user-guide/faq.md#how-can-i-compare-two-or-more-subsets-of-samples)
    * [Is it possible to determine if a particular mutation is heterozygous or homozygous in a sample? When a sample has 2 mutations in one gene, is it possible to determine whether the mutations are in cis or in trans with each other?](/user-guide/faq.md#is-it-possible-to-determine-if-a-particular-mutation-is-heterozygous-or-homozygous-in-a-sample-when-a-sample-has-2-mutations-in-one-gene-is-it-possible-to-determine-whether-the-mutations-are-in-cis-or-in-trans-with-each-other)
    * [How can I query over/under expression of a gene?](/user-guide/faq.md#how-can-i-query-overunder-expression-of-a-gene)
    * [How can I compare outcomes in patients with high vs low expression of a gene?](/user-guide/faq.md#how-can-i-compare-outcomes-in-patients-with-high-vs-low-expression-of-a-gene)
* [Results View](/user-guide/faq.md#results-view)
    * [OncoPrint](/user-guide/faq.md#oncoprint)
        * [What are OncoPrints?](/user-guide/faq.md#what-are-oncoprints)
        * [Can I change the order of genes in the OncoPrint?](/user-guide/faq.md#can-i-change-the-order-of-genes-in-the-oncoprint)
        * [Can I visualize my own data within an OncoPrint?](/user-guide/faq.md#can-i-visualize-my-own-data-within-an-oncoprint)
        * [Why are some samples ‚ÄúNot profiled‚Äù for certain genes?](/user-guide/faq.md#why-are-some-samples-not-profiled-for-certain-genes)
    * [Other pages](/user-guide/faq.md#other-pages)
        * [Does the Mutual Exclusivity tab calculate its statistics using all samples/alterations or only a specific subset?](/user-guide/faq.md#does-the-mutual-exclusivity-tab-calculate-its-statistics-using-all-samplesalterations-or-only-a-specific-subset)
        * [What are the values of the box and whiskers in a boxplot?](/user-guide/faq.md#what-are-the-values-of-the-box-and-whiskers-in-a-boxplot)
* [Study View](/user-guide/faq.md#study-view)
    * [How to use filter in the URL of Study View page?](/user-guide/faq.md#how-to-use-filter-in-the-url-of-study-view-page)
* [What if I have other questions or comments?](/user-guide/faq.md#what-if-i-have-other-questions-or-comments)

## General Questions
#### What is the cBioPortal for Cancer Genomics?
The cBioPortal for Cancer Genomics is an open-access, open-source resource for interactive exploration of multidimensional cancer genomics data sets. The goal of cBioPortal is to significantly lower the barriers between complex genomic data and cancer researchers by providing rapid, intuitive, and high-quality access to molecular profiles and clinical attributes from large-scale cancer genomics projects, and therefore to empower researchers to translate these rich data sets into biologic insights and clinical applications.
#### How do I get started?
Check out our [tutorial slides](https://www.cbioportal.org/tutorials) to get started or go through our [tutorial paper](https://www.ncbi.nlm.nih.gov/pubmed/23550210).
#### What data types are in the portal?
The portal supports and stores non-synonymous mutations, DNA copy-number data (putative, discrete values per gene, e.g. "deeply deleted" or "amplified", as well as log2 or linear copy number data), mRNA and microRNA expression data, protein-level and phosphoprotein level data (RPPA or mass spectrometry based), DNA methylation data, and de-identified clinical data. For a complete breakdown of available data types per cancer study go to the [Data Sets Page](https://www.cbioportal.org/datasets). Note that for many studies, only somatic mutation data and limited clinical data are available. For TCGA studies, the other data types are also available. Germline mutations are supported by cBioPortal, but are, with a few exceptions, not available in the public instance.

#### What does ___ stand for?
Here are the meanings of some of the abbreviations used by cBioPortal:
* VUS: variant of unknown significance
* CNA: copy number alteration
* AMP: amplification
* HOMDEL: deep deletion
* TMB: tumor mutational burden, calculated as mutations per megabase of sequenced DNA
* KM: Kaplan-Meier
* MSI: microsatellite instability
* OQL: Onco Query Language, used within cBioPortal to define the types of alterations included in a query. For more on OQL, review the [documentation](https://www.cbioportal.org/oql), [tutorial slides](https://www.cbioportal.org/tutorials#oql), and [videos](https://www.cbioportal.org/tutorials)

#### What is the process of data curation?
The TCGA firehose legacy datasets are imported directly from the original TCGA Data Coordinating Center via the [Broad Firehose](https://gdac.broadinstitute.org/).

We are also actively curating datasets from the literature. Studies from the literature were curated from the data published with the manuscripts. We sometimes reach out to the investigators to acquire additional data, such as clinical attributes. All mutation calls (in VCF or MAF format) are processed through an internal pipeline to annotate the variant effects in a consistent way across studies. Please [contact us](mailto:cbioportal@googlegroups.com) to suggest additional public datasets to curate or view the list of studies suggested for curation in our [Datahub on Github](https://github.com/cBioPortal/datahub/issues?q=is%3Aissue+is%3Aopen+label%3A%22new+public+study%22).
#### How do I get updates on new portal developments and new data sets?
Please subscribe to our low-volume [news mailing list](https://groups.google.com/group/cbioportal-news) or follow [@cbioportal on Twitter](https://twitter.com/cbioportal).
#### Does the portal work on all browsers and operating systems?
We support and test on the following web browsers: Safari, Google Chrome, Firefox and Edge. (As of release v3.5.4 we no longer support Internet Explorer 11). If you notice any incompatibilities, please let us know.
#### How do I cite the cBioPortal?
Please cite the following portal papers:
* Cerami et al. The cBio Cancer Genomics Portal: An Open Platform for Exploring Multidimensional Cancer Genomics Data. Cancer Discovery. May 2012 2; 401. [PubMed](https://www.ncbi.nlm.nih.gov/pubmed/22588877).
* Gao et al. Integrative analysis of complex cancer genomics and clinical profiles using the cBioPortal. Sci. Signal. 6, pl1 (2013). [PubMed](https://www.ncbi.nlm.nih.gov/pubmed/23550210).
* de Bruijn et al. Analysis and Visualization of Longitudinal Genomic and Clinical Data from the AACR Project GENIE Biopharma Collaborative in cBioPortal. Cancer Res (2023). [PubMed](https://pubmed.ncbi.nlm.nih.gov/37668528/).

Remember also to cite the source of the data if you are using a publicly available dataset.
#### Can I use figures from the cBioPortal in my publications or presentations?
Yes, you are free to use any of the figures from the portal in your publications or presentations (many are available in SVG or PDF format for easier scaling and editing). When you do, please cite Cerami et al., Cancer Discov. 2012, and Gao et al., Sci. Signal. 2013 (see the previous question for full citations).
#### Can I save or bookmark my results in cBioPortal?
You can bookmark your query results and share the URL with collaborators. We store all queries via Session IDs, and these are saved indefinitely. Use the bookmark tab to retrieve the full link, or generate a short link via the bit.ly link generator.
#### How is the cBioPortal for Cancer Genomics different from the Genomic Data Commons (GDC)?
The cBioPortal is an exploratory analysis tool for exploring large-scale cancer genomic data sets that hosts data from large consortium efforts, like [TCGA](https://cancergenome.nih.gov/) and [TARGET](https://ocg.cancer.gov/programs/target), as well as publications from individual labs. You can quickly view genomic alterations across a set of patients, across a set of cancer types, perform survival analysis and perform group comparisons. If you want to explore specific genes or a pathway of interest in one or more cancer types, the cBioPortal is probably where you want to start.

By contrast, the [Genomic Data Commons (GDC)](https://gdc.cancer.gov/) aims to be the definitive place for full-download and access to all data generated by TCGA and TARGET. If you want to download raw mRNA expression files or full segmented copy number files, the GDC is probably where you want to start.
#### Does the cBioPortal provide a Web Service API? R interface? MATLAB interface?
Yes, the cBioPortal provides a [Swagger API](https://www.cbioportal.org/api/swagger-ui.html), and [R/MATLAB interfaces](/web-API-and-Clients.md#r-client).
#### Can I use cBioPortal with my own data?
cBioPortal provides several options for analyzing your own data. Visit our [Visualize Your Data](https://www.cbioportal.org/visualize) page to generate an OncoPrint or Lollipop Plot with your own data. To utilize the entire suite of analysis and visualization tools, you can also install your own instance of cBioPortal (see next question).
#### Can I create a local instance of cBioPortal to host my own data?
Yes, the cBioPortal is open-source, and available on [GitHub](https://github.com/cBioPortal/cbioportal). Our [documentation](/) provides complete download and installation instructions.
#### I'd like to contribute code to the cBioPortal. How do I get started?
Great! We would love to have your contributions. To get started, head over to our GitHub repository and check out our page on [how to contribute](https://github.com/cBioPortal/cbioportal/blob/master/CONTRIBUTING.md).
#### What is a combined Study?
A combined study is a custom study comprised of samples from multiple studies. The combined study feature 
enables you to combine samples from multiple studies to form a bigger study. This cohort of samples can then be queried or explored just like a traditional study, and can be returned to at a later date or shared with a collaborator. 
#### How can I form a combined Study?
A combined or merged study is a custom study comprised of samples from multiple studies. In the homepage of cbioportal, studies can be selected using the checkbox located on the left of the study. Once the studies are selected, they can be combined and explored using the "Explore Selected Studies" button. Alternatively, after the studies are selected, you can run queries on the combined study using the "Query by Gene" button.
#### How can I create a subset or sub cohort of a study with specific samples or patients?
A subset or sub cohort of a study can be created by specifying individual patients or samples. After a study is selected, user can click on the "Custom selection" button to create a new filter by specifying the sampleID or patientID that the user is interested to explore. Another way is to filter a set of patients using the charts on the study view and then view the IDs of the patients and samples that were selected or unselected based on the current filter.
#### What is a Virtual Study?
A virtual study is a custom study comprised of samples from one or more existing studies. The virtual study feature allows you to define a custom cohort of samples that fit your specific genomic or clinical criteria of interest. These samples can be a subset of the data available in an existing study, or result from the combination of multiple existing studies. This cohort of samples can then be queried or explored just like a traditional study, and can be returned to at a later date or shared with a collaborator. For more information and examples, see our [tutorial on virtual studies](https://www.cbioportal.org/tutorials).
#### Is it necessary to log in to use virtual studies? If I do log in, what additional functionality do I gain?
No. A user that has not logged in can create virtual studies and run queries in those studies (by using the query box on the study summary page). Links to virtual studies are permanent, so you can save the link on your computer and come back to it anytime, or share it with others.

If you log in, you gain the ability to save your virtual study to the list of existing studies on the homepage. This makes a virtual study functionally the same as any other study: you can access your virtual studies in the query builder and you can combine an existing virtual study with any other study to create a new virtual study.
#### What is Group Comparison?
Group Comparison is a suite of analysis features which allows a user to compare clinical or genomic features of user-defined groups of samples. These groups can be defined based on any clinical or genomic features. For an overview, see [our tutorial on group comparison](https://www.cbioportal.org/tutorials).
#### What are the statistical significance tests in Group Comparison?
- Survival tab: Log-rank test
- Clinical tab:
    - Continuous data: Chi-squared test
    - Categorical data: Kruskal Wallis test
- Other tabs
    - 2 groups
        - Continuous data: one-sided t-test
        - Categorical data: two-sided Fisher's exact test
    - 3 or more groups
        - Continuous data: one-way ANOVA
        - Categorical data: Chi-squared test

## Data Questions
### General Data
#### Does the portal contain cancer study X?
Check out the [Data Sets Page](https://www.cbioportal.org/datasets) for the complete set of cancer studies currently stored in the portal. If you do not see your specific cancer study of interest, please [contact us](mailto:cbioportal@googlegroups.com), and we will let you know if it's in the queue.
#### Which resources are integrated for variant annotation?
cBioPortal supports the annotation of variants from several different databases. These databases provide information about the recurrence of, or prior knowledge about, specific amino acid changes. For each variant, the number of occurrences of mutations at the same amino acid position present in the COSMIC database are reported. Furthermore, variants are annotated as ‚Äúhotspots‚Äù if the amino acid positions were found to be recurrent linear hotspots, as defined by the Cancer Hotspots method ([cancerhotspots.org](https://www.cancerhotspots.org/)), or three-dimensional hotspots, as defined by 3D Hotspots ([3dhotspots.org](https://www.3dhotspots.org/)). Prior knowledge about variants, including clinical actionability information, is provided from three different sources: OncoKB ([www.oncokb.org](https://www.oncokb.org/)), CIViC ([civicdb.org](https://civicdb.org/)), as well as My Cancer Genome ([mycancergenome.org](https://www.mycancergenome.org/)). For OncoKB, exact levels of clinical actionability are displayed in cBioPortal, as defined by [the OncoKB paper](https://ascopubs.org/doi/full/10.1200/PO.17.00011).
#### What version of the human reference genome is being used in cBioPortal?
The [public cBioPortal](https://www.cbioportal.org) is currently using hg19/GRCh37.
#### How does cBioPortal handle duplicate samples or sample IDs across different studies?
The cBioPortal generally assumes that samples or patients that have the same ID are actually the same. This is important for cross-cancer queries, where each sample should only be counted once. If a sample is part of multiple cancer cohorts, its alterations are only counted once in the Mutations tab (it will be listed multiple times in the table, but is only counted once in the lollipop plot). However, other tabs (including OncoPrint and Cancer Types Summary) will count the sample twice - for this reason, we advise against querying multiple studies that contain the same samples (e.g., TCGA PanCancer Atlas and TCGA Firehose Legacy).
#### Are there any normal tissue samples available through cBioPortal?
No, we currently do not store any normal tissue data in our system.
#### How can I find which studies have mRNA expression data (or any other specific data type)?
Check out the [Data Sets Page](https://www.cbioportal.org/datasets) where you can view the complete set of cancer studies and sort by the number of samples with data available for any data type.
#### Can I download all data at once?
You can download all data for individual studies on the [Data Sets Page](https://www.cbioportal.org/datasets) or the study view page for the study of interest. You can also download all studies from our [Data Hub](https://github.com/cBioPortal/datahub).
#### The data today is different than the last time I looked. What happened?
We do occasionally update existing datasets to provide the most up-to-date, accurate and consistent data possible. The data you see today is likely an improved version of what you have seen previously. However, if you suspect that there is an error in the current version, please let us know at [cbioportal@googlegroups.com](mailto:cbioportal@googlegroups.com).

If you need to reference an old version of a dataset, you can find previous versions in our [Datahub repository](https://github.com/cBioPortal/datahub/tree/master/public).
#### How do I access data from AACR Project GENIE?
Data from AACR Project GENIE are provided in a [dedicated instance of cBioPortal](https://www.cbioportal.org/genie/). You can also download GENIE data from the [Synapse Platform](https://synapse.org/genie). Note that you will need to register before accessing the data. Additional information about AACR Project GENIE can be found on the [AACR website](https://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

### TCGA
#### How does TCGA data in cBioPortal compare to TCGA data in Genome Data Commons?
We do not currently load the mutation data from the GDC. Instead, we have the original mutation data generated by the individual TCGA sequencing centers. The source of the data is the Broad Firehose (or the publication pages for data that matches a specific manuscript). These data are usually a combination of two mutation callers, but they differ by center (typically a variant caller like MuTect plus an indel caller), and sequencing centers have modified their mutation calling pipelines over time.
#### What happened to TCGA Provisional datasets?
We renamed TCGA Provisional datasets to TCGA Firehose Legacy to better reflect that this data comes from a legacy processing pipeline. The exact same data is now available in TCGA Firehose Legacy studies.
#### What are TCGA Firehose Legacy datasets and how do they compare to the publication-associated datasets and the PanCancer Atlas datasets?
The Firehose Legacy dataset (formerly Provisional datasets) for each TCGA cancer type contains all data available from the Broad Firehose. The publication datasets reflect the data that were used for each of the publications. The samples in a published dataset are usually a subset of the firehose legacy dataset, since manuscripts were often written before TCGA completed their goal of sequencing 500 tumors.

There can be differences between firehose legacy and published data. For example, the mutation data in the publication usually underwent more QC, and false positives might have been removed or, in rare cases, false negatives added. RNA-Seq and copy-number values may also differ slightly, as different versions of analysis pipelines could have been used. Additionally, due to additional curation during the publication process, the clinical data for the publication may be of higher quality or may contain a few more data elements, sometimes derived from the genomic data (e.g., genomic subtypes).

The TCGA PanCancer Atlas datasets derive from an effort to unify TCGA data across all tumor types. Publications resulting from this effort can be found at the [TCGA PanCancer Atlas site](https://www.cell.com/pb-assets/consortium/pancanceratlas/pancani3/index.html). In the cBioPortal, data from the PanCancer Atlas is divided by tumor type, but these studies have uniform clinical elements, consistent processing and normalization of mutations, copy number, mRNA data and are ideally processed for comparative analyses.
#### Where do the thresholded copy number call in TCGA Firehose Legacy data come from?
Thresholded copy number calls in the TCGA Firehouse Legacy datasets are generated by the GISTIC 2.0 algorithm and obtained from the Broad Firehose.
#### Which studies have MutSig and GISTIC results? How do these results compare to the data in the TCGA publications?
MutSig and GISTIC results about the statistical significance of recurrence of mutations and copy-number alterations in specific genes are available for many TCGA studies. The MutSig and GISTIC results reported in cBioPortal are based on the same mutations and copy number data reported in each TCGA publication, or the Broad Firehose for the firehose legacy data sets. However, the publication may or may not have included the complete MutSig and GISTIC output, and therefore there may be some discrepancies between the publication and the data in cBioPortal.
#### How can I download the PanCancer Atlas data?
PanCancer Atlas data can be downloaded on a study-by-study basis from cBioPortal through the [Datasets page](https://www.cbioportal.org/datasets) or our [DataHub](https://github.com/cBioPortal/datahub/tree/master/public). To download all cancer types together, try the [Genomic Data Commons PanCancer Atlas page](https://gdc.cancer.gov/about-data/publications/pancanatlas).

### DNA (Mutations, Copy Number & Fusions)
#### Does the cBioPortal contain synonymous mutation data?
No, the cBioPortal does not currently support synonymous mutations. This may change in the future, but we have no plans yet to add this feature.
#### What processing or filtering is applied to generate the mutation data?
Within cBioPortal, we utilize the mutation calls as provided by each publication. We do not perform any additional filtering. The only processing we do is to standardize the annotation of the mutations using [Genome Nexus](https://genomenexus.org) (which utilizes [VEP](https://useast.ensembl.org/info/docs/tools/vep/index.html) with the [canonical MSKCC transcript](https://github.com/genome-nexus/genome-nexus-importer/blob/master/data/common_input/isoform_overrides_at_mskcc_grch37.txt)). Read more about the transcript assignments [here](https://docs.cbioportal.org/mutation-data-transcript-annotation/#transcript-assignment). For specifics of which tools were used to call mutations and filters that may have been applied, refer to the publication manuscript.
#### What transcripts are used for annotating mutations?
Prior to loading a study into cBioPortal, we run all mutation data through a standard pipeline (see above), which re-annotates all mutations to the [canonical MSKCC transcript](https://github.com/genome-nexus/genome-nexus-importer/blob/master/data/common_input/isoform_overrides_at_mskcc_grch37.txt). Read more about the transcript assignments [here](https://docs.cbioportal.org/mutation-data-transcript-annotation/#transcript-assignment).
#### How are protein domains in the mutational lollipop diagrams specified?
Protein domain definitions come from [PFAM](https://pfam.xfam.org/).
#### What is the difference between a ‚Äúsplice site‚Äù mutation and a ‚Äúsplice region‚Äù mutation?
A ‚Äúsplice site‚Äù mutation occurs in an intron, in a splice acceptor or donor site (2bp into an intron adjacent to the intron/exon junction), defined by [Sequence Ontology](https://www.sequenceontology.org/browser/current_svn/term/SO:0001629). ‚ÄúSplice region‚Äù mutations are mutations that occur near the intron/exon junction, defined by [Sequence Ontology](https://www.sequenceontology.org/browser/current_svn/term/SO:0001630). While synonymous mutations are generally excluded from cBioPortal, these ‚Äúsplice region‚Äù synonymous mutations are included due to their potential impact on splicing.
#### What do ‚ÄúAmplification‚Äù, ‚ÄúGain‚Äù, ‚ÄúDeep Deletion‚Äù, ‚ÄúShallow Deletion‚Äù and "-2", "-1", "0", "1", and "2" mean in the copy-number data?
These levels are derived from copy-number analysis algorithms like GISTIC or RAE, and indicate the copy-number level per gene:
* -2 or Deep Deletion indicates a deep loss, possibly a homozygous deletion
* -1 or Shallow Deletion indicates a shallow loss, possibley a heterozygous deletion
* 0 is diploid
* 1 or Gain indicates a low-level gain (a few additional copies, often broad)
* 2 or Amplification indicate a high-level amplification (more copies, often focal)

Note that these calls are putative. We consider the deep deletions and amplifications as biologically relevant for individual genes by default. Note that these calls are usually not manually reviewed, and due to differences in purity and ploidy between samples, there may be false positives and false negatives.
#### What is GISTIC? What is RAE?
Copy number data sets within the portal are often generated by the [GISTIC](https://www.ncbi.nlm.nih.gov/sites/entrez?term=18077431) or [RAE](https://www.ncbi.nlm.nih.gov/sites/entrez?term=18784837) algorithms. Both algorithms attempt to identify significantly altered regions of amplification or deletion across sets of patients. Both algorithms also generate putative gene/patient copy number specific calls, which are then input into the portal.

For TCGA studies, the table in allthresholded.bygenes.txt (which is the part of the GISTIC output that is used to determine the copy-number status of each gene in each sample in cBioPortal) is obtained by applying both low- and high-level thresholds to to the gene copy levels of all the samples. The entries with value +/- 2 exceed the high-level thresholds for amplifications/deep deletions, and those with +/- 1 exceed the low-level thresholds but not the high-level thresholds. The low-level thresholds are just the 'ampthresh' and 'delthresh' noise threshold input values to GISTIC (typically 0.1 or 0.3) and are the same for every thresholds.

By contrast, the high-level thresholds are calculated on a sample-by-sample basis and are based on the maximum (or minimum) median arm-level amplification (or deletion) copy number found in the sample. The idea, for deletions anyway, is that this level is a good approximation for hemizygous losses given the purity and ploidy of the sample. The actual cutoffs used for each sample can be found in a table in the output file sample_cutoffs.txt. All GISTIC output files for TCGA are available at: gdac.broadinstitute.org.

### RNA
#### Does the portal store raw or probe-level data?
No, the portal only contains gene-level data. Data for different isoforms of a given gene are merged. Raw and probe-level data for data sets are available via [NCBI GEO](https://www.ncbi.nlm.nih.gov/geo/), [dbGaP](https://www.ncbi.nlm.nih.gov/gap/) or through the [GDC](https://portal.gdc.cancer.gov/). See the cancer type description on the main query page or refer to the original publication for links to the raw data.
#### What are mRNA and microRNA Z-Scores?
For mRNA and microRNA expression data, we typically compute the relative expression of an individual gene in a tumor sample to the gene's expression distribution in a reference population of samples. That reference population is all profiled samples (by default for mRNA), or normal samples (when specified), or all samples that are diploid for the gene in question (discontinued).  The returned value indicates the number of standard deviations away from the mean of expression in the reference population (Z-score). The normalization method is described [here](https://github.com/cBioPortal/cbioportal/blob/master/docs/Z-Score-normalization-script.md). Please note that the expression results by querying a gene with the default setting (z-score threshold of 2) oftentimes are not meaningful. Since the z-scores were usually calculated compared to other tumor samples, high or low expression does not necessarily mean that the gene is expressed irregularly in tumors. The data is useful for correlation analysis, for example, pick a threshold based on overall expression (using Plots tab) and compare survival data between expression high and low groups.
#### Is there any normal RNA-seq data in cBioPortal?
We have RNASeqV2 mRNA expression data for normal samples of 16 TCGA PanCan Atlas Cohorts. The data was curated from [GDC](https://gdc.cancer.gov/about-data/publications/pancanatlas), and can be downloaded from our [Datahub](https://github.com/cBioPortal/datahub) or [Data Set](https://www.cbioportal.org/datasets) page. This data is not directly queriable in portal; they are only used as reference data for calculating the "relavtive to normal expression z-score" profile. [Example: ERBB2 expression z-scores relative to normal expression](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&case_set_id=all&data_priority=0&gene_list=ERBB2&geneset_list=%20&plots_coloring_selection=%7B%22colorByCopyNumber%22%3A%22true%22%7D&plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_STUDY%22%7D&plots_vert_selection=%7B%22selectedGeneOption%22%3A2064%2C%22dataType%22%3A%22MRNA_EXPRESSION%22%2C%22selectedDataSourceOption%22%3A%22rna_seq_v2_mrna_median_all_sample_ref_normal_Zscores%22%2C%22logScale%22%3A%22true%22%7D&profileFilter=0&tab_index=tab_visualize). 
#### How is TCGA RNASeqV2 processed? What units are used?
RNASeqV2 from TCGA is processed and normalized using [RSEM](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323). Specifically, the RNASeq V2 data in cBioPortal corresponds to the rsem.genes.normalized_results file from TCGA. A more detailed explanation of RSEM output can be found [here](https://www.biostars.org/p/106127/). cBioPortal then calculates z-scores as described above in [What are mRNA and microRNA Z-Scores?](/user-guide/faq.md#what-are-mrna-and-microrna-z-scores)
#### Is there microRNA data?
We have microRNA data for only a few studies and they are not up to date. To download more updated miRNA data, please go to either [Broad Firehose](https://gdac.broadinstitute.org/), or [GDC](https://portal.gdc.cancer.gov/).
#### How can I query microRNAs in the portal?
You can input either precursor or mature miRNA IDs. Since one precursor ID may correspond to multiple mature IDs and vise versa, the portal creates one internal ID for each pair of precursor ID and mature ID mapping. For example, an internal ID of MIR-29B-1/29B stands for precursor microRNA hsa-mir-29b-1 and mature microRNA hsa-miR-29b. After entering a precursor or mature ID, you will be asked to select one internal ID for query and that internal ID will also be displayed in the Oncoprint.

### Protein
#### How can I query phosphoprotein levels in the portal?
You need to input special IDs for each phosphoprotein/phopshosite such as AKT1_pS473 (which means AKT1 protein phosphorylated at serine residue at position 473). You could also input aliases such as phosphoAKT1 or phosphoprotein, and the portal will ask you to select the phosphoprotein/phosphosite of your interest. Note that phosphoprotein data is only available for select studies and for a limited number of proteins / phosphorylation sites.
#### Why isn‚Äôt there protein data for my gene of interest?
Most of the protein expression data in cBioPortal comes from assays like RPPA which only interrogate a subset of all proteins. TCGA ovarian, breast, and colorectal firehose legacy studies also have mass-spectrometry-based proteomics data from [CPTAC](https://proteomics.cancer.gov/programs/cptac) which cover more genes/proteins.
### DNA Methylation
#### Which methylation probe is used for genes with multiple probes?
For genes with multiple probes (usually from the Infinium arrays), we only include methylation data from the probe with the strongest negative correlation between the methylation signal and the gene's expression in the study (TCGA only).

### Clinical Data
#### What kind of clinical data is stored in the portal?
The portal currently stores de-identified clinical data, such as gender, age, tumor type, tumor grade, overall and disease-free survival data, when available. The available clinical data will differ from study to study.
#### What is the meaning of OS_STATUS / OS_MONTHS, and PFS_STATUS / PFS_MONTHS?
OS_STATUS means overall survival status ("0" -> "living" or "1" -> "deceased") and OS_MONTHS indicates the number of months from time of diagnosis to time of death or last follow up. PFS refers to ‚Äúprogression free survival‚Äù, indicating whether patient‚Äôs disease has recurred/progressed (PFS_STATUS), and at what time the disease recurred or the patient was last seen (PFS_MONTHS).

## Analysis Questions
#### How can I query/explore a select subset of samples?
cBioPortal allows you to run a query or explore study view using a user-specified list of samples/patients.

The first step is to define your sample set. There are two slightly different approaches you can take to defining your sample set, depending on whether you are selecting based on a positive criteria (samples with TP53 mutations) or a negative criteria (samples without a KRAS mutation).

Let‚Äôs take the positive criteria example first. Run a query for TP53 mutations using OQL (TP53: MUT) in your study of interest. Click over to the ‚ÄúDownload‚Äù tab. In the table at the top, find the row that starts with ‚ÄúSamples affected‚Äù, and either Copy or Download that list. This is your list of samples that have a TP53 mutation.

Now for the negative criteria example. This also begins by using OQL to run a query for KRAS mutations (KRAS: MUT) in your study of interest. Click over to the ‚ÄúDownload‚Äù tab. Look at the table at the top again, but this time find the row that starts with ‚ÄúSample matrix‚Äù. Copy or download this data and open it in Excel. You will see a two column table that indicates whether a given sample is altered or not, indicated by 0 or 1. Sort by the second column and then copy all the sample IDs from the first column that have a 0 in the second column. This is your list of samples that do not have a KRAS mutation.

With a sample list in hand, you can now either run a query in just the selected samples (select ‚ÄúUser-defined Case List‚Äù in the ‚ÄúSelect Patient/Case Set:‚Äù dropdown) or explore this set of patients in study view (click ‚ÄúSelect cases by IDs‚Äù and then create a Virtual Study restricted to just those samples).

For more information about OQL, see the [specification page](/user-guide/oql.md) or view the [tutorial slides](https://www.cbioportal.org/tutorials#oql). For more information about virtual studies, read [this FAQ](/user-guide/faq.md#what-is-a-virtual-study) or view the [tutorial slides](https://www.cbioportal.org/tutorials).
#### How can I compare two or more subsets of samples?
cBioPortal has a suite of analysis tools to enable comparisons between user-defined groups of samples/patients. For an overview of this functionality, see our [tutorial on group comparison](https://www.cbioportal.org/tutorials).
#### Is it possible to determine if a particular mutation is heterozygous or homozygous in a sample? When a sample has 2 mutations in one gene, is it possible to determine whether the mutations are in cis or in trans with each other?
There is currently no way to definitively determine whether a mutation is heterozygous/homozygous or in cis/trans with another mutation. However, you can try to infer the status of mutations by noting the copy number status of the gene and the variant allele frequency of the mutation(s) of interest relative to other mutations in the same sample. The cBioPortal patient/sample view can help you accomplish this.

Specifically in the case of TCGA samples with two mutations in the same gene, you can also obtain access to the aligned sequencing reads from the [GDC](https://portal.gdc.cancer.gov/) and check if the mutations are in cis or in trans (if the mutations are close enough to each other).
#### How can I query over/under expression of a gene?
cBioPortal supports Onco Query Language (OQL) which can be used to query over/under expression of a gene. When writing a query, select an mRNA expression profile. By default, samples with expression z-scores >2 or <-2 in any queried genes are considered altered. Alternate cut-offs can be defined using OQL, for example: "EGFR: EXP>2" will query for samples with an EGFR expression z-score >2. Review for the OQL [specification page](/user-guide/oql.md) or [tutorial slides](https://www.cbioportal.org/tutorials#oql) for more specifics and examples.
#### How can I compare outcomes in patients with high vs low expression of a gene?
To compare outcomes in patients with high vs low expression of a gene (excluding those patients with intermediate levels of expression), we will follow a 2 step process that builds on the approach described above in [How can I query/explore a select subset of samples?](/user-guide/faq.md#how-can-i-queryexplore-a-select-subset-of-samples), utilizing [OQL](/user-guide/oql.md) to first identify and then stratify that cases of interest.

First, identify the sample set using OQL. For example, to stratify patients based on expression of EGFR, add an mRNA profile to the query, and write "EGFR: EXP>2 EXP<-2" in the gene set box. After running the query, go to the Download tab and copy/download the ‚ÄúSamples affected‚Äù list.

Second, return to the homepage and paste the list of sample IDs from the previous step into the ‚ÄúUser-defined Case List‚Äù in the ‚ÄúSelect Patient/Case Set:‚Äù dropdown. This query will now only look at samples with high or low expression. To now stratify into high vs low for survival analysis, enter "EGFR: EXP>2" in the gene set box (don‚Äôt forget to select the same mRNA profile). Run the query and click over to the Survival tab. The ‚Äúcases with alteration‚Äù are patients with high expression of EGFR and the cases without alteration are those with low expression of EGFR.

We use 2 and -2 as example thresholds above, but it is also a good idea to look at the distribution of expression data and select a threshold based on that. Plots tab can be useful for analyzing the expression distribution.
## Results View
### OncoPrint
#### What are OncoPrints?
OncoPrints are compact means of visualizing distinct genomic alterations, including somatic mutations, copy number alterations, and mRNA expression changes across a set of cases. They are extremely useful for visualizing gene set and pathway alterations across a set of cases, and for visually identifying trends, such as trends in mutual exclusivity or co-occurrence between gene pairs within a gene set. Individual genes are represented as rows, and individual cases or patients are represented as columns.

![image](https://cloud.githubusercontent.com/assets/840895/24209131/d34e33b2-0efb-11e7-945a-fd2b3c66d195.png)

#### Can I change the order of genes in the OncoPrint?
By default, the order of genes in the OncoPrint will be the same as in your query. You can change the order by (a) clicking on the gene name and dragging it up/down or (b) clicking on the three vertical dots next to the gene name to move the gene up/down.
#### Can I visualize my own data within an OncoPrint?
Yes, check out the OncoPrinter tool on our [Visualize Your Data page](https://www.cbioportal.org/visualize).
#### Why are some samples ‚ÄúNot profiled‚Äù for certain genes?
Some studies include data from one or more targeted sequencing platforms which do not include all genes. For samples sequenced on these smaller panels, cBioPortal will indicate that a particular gene was not included on the sequencing panel used for that sample. Alteration frequency calculations for each gene also take this information into account. Hover over a sample in OncoPrint to see the gene panel name, and click on that gene panel name to view a list of the genes included on that panel.
### Other pages
#### Does the Mutual Exclusivity tab calculate its statistics using all samples/alterations or only a specific subset?
The calculations on the Mutual Exclusivity tab are performed using all samples included in the query. A sample is defined as altered or unaltered for each gene based on the [OQL](/user-guide/oql.md) utilized in the query - by default, this will be non-synonymous mutations, fusions, amplifications and deep deletions.
#### What are the values of the box and whiskers in a boxplot?
In boxplots on cBioPortal, the box is drawn from the 25th percentile (Q1) to the 75th percentile (Q3), with the horizontal line in between representing the median. Whiskers are drawn independently above and below the box, and will extend to the maximum or minimum data values, unless there are outlier values, in which case the whisker will extend to 1.5 * IQR (interquartile range = Q3-Q1). Outliers are defined as values that extend beyond 1.5 * IQR.

## Study View
### How to use filter in the URL of Study View page?
You can filter the study based on values of one attribute in the URL. For example, <https://www.cbioportal.org/study/summary?id=msk_impact_2017#filterJson={"clinicalDataFilters":[{"attributeId":"CANCER_TYPE","values":[{"value":"Melanoma"}]}]}>

filterJson is set in the url hash string. Here are the allowed parameters and format for it in filterJson:
```
{
	"caseLists": [
		["string"]
	],
	"clinicalDataFilters": [{
		"attributeId": "string",
		"values": [{
			"end": 0,
			"start": 0,
			"value": "string"
		}]
	}],
	"geneFilters": [{
		"geneQueries": [
			["string"]
		],
		"molecularProfileIds": ["string"]
	}],
	"genomicDataFilters": [{
		"hugoGeneSymbol": "string",
		"profileType": "string",
		"values": [{
			"end": 0,
			"start": 0,
			"value": "string"
		}]
	}],
	"genomicProfiles": [
		["string"]
	],
	"sampleIdentifiers": [{
		"sampleId": "string",
		"studyId": "string"
	}],
	"studyIds": ["string"]
}
```

## What if I have other questions or comments?
Please contact us at [cbioportal@googlegroups.com](mailto:cbioportal@googlegroups.com). Previous discussions about cBioPortal are available on the [user discussion mailing list](https://groups.google.com/group/cbioportal).
# New Users

Are you new to cBioPortal? Welcome! We have a few options to help you get started.

If you have an hour, we highly recommend watching the recording of our Introduction to cBioPortal webinar ([youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo) or [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)).

Don't have an hour? Review our tutorial slides for exploring a study ([Google slides](https://docs.google.com/presentation/d/1_OGK69lO4Z62WaxHHkNYmWvY0LQN2v0slfaLyY1_IQ0) or [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%201%20Single%20Study%20Exploration.pdf)) and running a query ([Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) or [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf))

Or, watch two of our short how-to videos which demonstrate how to explore a study ([youtube.com](https://www.youtube.com/watch?v=N8ffDgkqDWc)) and how to run a query ([youtube.com](https://www.youtube.com/watch?v=MH-kY5usA70)).
# Onco Query Language (OQL)

## Introduction to OQL
The Onco Query Language (OQL) is used to define which specific types of alterations are included in a query on the cBioPortal for Cancer Genomics. By default, querying for a gene includes mutations, fusions, amplifications and deep deletions. OQL can be used to specify specific mutations (e.g. BRAF V600E) or types of mutations (e.g. BRCA1 truncating mutations), lower level copy number alterations (e.g. CDKN2A shallow deletions), changes in mRNA or protein expression, and more.

OQL-specified alterations will be reflected on most tabs, including OncoPrint, but are not currently reflected on the Plots, Co-Expression or Expression tabs.

Note that OQL assumes any word that it doesn't recognize is a mutation code.

Additional explanation and examples using OQL are available in the [User Guide](https://docs.cbioportal.org/user-guide/by-page/#oql).

## OQL Keywords
Users can define specific subsets of genetic alterations for five data types:

Data Type | Keywords and Syntax | Default*
--------- | ------------------- | --------
Mutations | `MUT` All non-synonymous mutations <br> `MUT = <protein change>` Specific amino acid changes (e.g. `V600E` or `V600`) <br> `MUT = <mutation type>` Acceptable values are: `MISSENSE, NONSENSE, NONSTART, NONSTOP, FRAMESHIFT, INFRAME, SPLICE, TRUNC` | `MUT`
Fusions | `FUSION` All fusions (note that many studies lack fusion data) | `FUSION`
Copy Number Alterations | `AMP` Amplifications <br> `HOMDEL` Deep Deletions <br> `GAIN` Gains <br> `HETLOSS` Shallow Deletions <br> Comparison operators can also be used with `CNA` (e.g. `CNA >= GAIN` is the same as `AMP GAIN`) | `AMP` <br> `HOMDEL`
mRNA Expression | `EXP < -x` mRNA expression is less than `x` standard deviations (SD) below the mean <br> `EXP > x` mRNA expression is greater than `x` SD above the mean <br> The comparison operators `<=` and `>=` also work | `EXP >= 2` <br> `EXP <= -2`
Protein/phosphoprotein level | `PROT < -x` Protein expression is less than `x` standard deviations (SD) below the mean <br> `PROT > x` Protein expression is greater than `x` SD above the mean <br> The comparison operators `<=` and `>=` also work | `PROT >= 2` <br> `PROT <= -2`

\* These are the default OQL keywords used for each data type when a gene is queried without any explicit OQL.

### OQL modifiers
Mutations and copy number alterations can be further refined using modifiers:

Keyword | Applicable Data Type | Explanation
------- | -------------------- | -----------
`DRIVER` | Mutations <br> Fusions <br> Copy Number Alterations | Include only mutations, fusions and copy number alterations which are driver events, as defined in OncoPrint (default: OncoKB and CancerHotspots).
`GERMLINE` | Mutations | Include only mutations that are defined as germline events by the study.
`SOMATIC` | Mutations | Include all mutations that are not defined as germline.
`(a-b)` (protein position range) | Mutations | Include all mutations that overlap with the protein position range `a-b`, where `a` and `b` are integers. If you add a `*` (i.e. `(a-b*)`) then it will only include those mutations that are fully contained inside `a-b`. The open-ended ranges `(a-)` and `(-b)` are also allowed.

## Basic Usage
When querying a gene without providing any OQL specifications, cBioPortal will default to these OQL terms for a query with Mutation and Copy Number selected in the Genomic Profiles section:
`MUT FUSION AMP HOMDEL`

![image of basic query](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/basic_query.png)

You can see the OQL terms applied by hovering over the gene name in OncoPrint:

![image of basic query oncoprint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/basic_query_oncoprint.png)


If you select RNA and/or Protein in the "Genomic Profiles" section of the query, the default settings are:

RNA: `EXP >= 2 EXP <= -2`

Protein: `PROT >= 2 PROT <= -2`

![image of exp prot query oncoprint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/basic_query_oncoprint_exp_prot.png)

You must select the relevant Genomic Profile in order for OQL to query that data type. For example, you can't add `EXP > 2` to the query without also selecting an RNA profile.


Proper formatting for OQL is straightforward: gene name, followed by a colon, followed by any OQL keywords and ending in a semicolon, an end-of-line, or both.
```
GENE1: OQL KEYWORDS;
GENE2: OQL KEYWORDS
```
In general, any combination of OQL keywords and/or expressions can annotate any gene, and the order of the keywords is immaterial.

Below we will go into greater detail about each data type.

## Mutations
To view cases with specific mutations, provide the specific amino acid change of interest:
```
BRAF: MUT = V600E
```

You can also view all mutations at a particular position:
```
BRAF: MUT = V600
```

Or all mutations of a specific type:
```
TP53: MUT = <mutation type>
```

`<mutation type>` can be one or more of:
* `MISSENSE`
* `NONSENSE`
* `NONSTART`
* `NONSTOP`
* `FRAMESHIFT`
* `INFRAME`
* `SPLICE`
* `TRUNC`

For example, to view TP53 truncating mutations and in-frame insertions/deletions:
```
TP53: MUT = TRUNC INFRAME
```

OQL for mutations can also be written without `MUT =`. The following examples are identical:
```
BRAF: MUT = V600E
BRAF: V600E
TP53: MUT = TRUNC INFRAME
TP53: TRUNC INFRAME
```

OQL can also be used to exclude a specific protein change, position or type of mutation. For example, below are examples to query all EGFR mutations except T790M, all BRAF mutations except those at V600 and all TP53 mutations except missense:
```
EGFR: MUT != T790M
BRAF: MUT != V600
TP53: MUT != MISSENSE
```
Note that this will only work to exclude a single event. Because OQL uses 'OR' logic, excluding multiple mutations or excluding a mutation while including another mutation (e.g. `BRAF: MUT=V600 MUT!=V600E`) will result in querying all mutations.

## Copy Number Alterations
To view cases with specific copy number alterations, provide the appropriate keywords for the copy number alterations of interest. For example, to see amplifications:
```
CCNE1: AMP
```

Or amplified and gained cases:
```
CCNE1: CNA >= GAIN
```

Which can also be written as:
```
CCNE1: GAIN AMP
```

## Expression
High or low mRNA expression of a gene is determined by the number of standard deviations (SD) from the mean. For example, to see cases where mRNA for CCNE1 is greater than 3 SD above the mean:
```
CCNE1: EXP > 3
```

## Protein
High or low protein expression is similarly determined by the number of SD from the mean. For example, to see cases where protein expression is 2 SD above the mean:
```
EGFR: PROT > 2
```

Protein expression can also be queried at the phospho-protein level:
```
EGFR_PY992: PROT > 2
```

## Modifiers
Modifiers can be used on their own or in combination with other OQL terms for mutations, fusions and copy number alterations to further refine the query. Modifiers can be combined with other OQL terms using an underscore. The order in which terms are combined is immaterial.

### Driver
The `DRIVER` modifier applies to mutations, fusions and copy number alterations. The definition of what qualifies as a driver alteration comes from the "Mutation Color" menu in OncoPrint. By default, drivers are defined as mutations, fusions and copy number alterations in <a href="https://www.oncokb.org">OncoKB</a> or <a href="https://www.cancerhotspots.org">CancerHotspots</a>.

On its own, the `DRIVER` modifier includes driver mutations, fusions and copy number alterations:
```
EGFR: DRIVER
```

Or it can be used in combination with another OQL term. For example, to see only driver fusion events:
```
EGFR: FUSION_DRIVER
```

Or driver missense mutations:
```
EGFR: MUT = MISSENSE_DRIVER
```

When combining `DRIVER` with another OQL term, the order doesn't matter: `MUT_DRIVER` and `DRIVER_MUT` are equivalent. `DRIVER` can be combined with:
* `MUT`
* `MUT = <mutation type>` or `MUT = <protein change>`
* `FUSION`
* `CNA`
* `AMP` or `GAIN` or `HETLOSS` or `HOMDEL`
* `GERMLINE` or `SOMATIC` (see below)

### Germline/Somatic
The `GERMLINE` and `SOMATIC` modifiers only apply to mutations. A mutation can be explicitly defined as germline during the data curation process. Note that very few studies on the public cBioPortal contain germline data.

`GERMLINE` or `SOMATIC` can be combined with:
* `MUT`
* `MUT = <mutation type>` or `MUT = <protein change>`
* `DRIVER`

To see all germline BRCA1 mutations:
```
BRCA1: GERMLINE
```

Or to see specifically truncating germline mutations:
```
BRCA1: TRUNC_GERMLINE
BRCA1: GERMLINE_TRUNC
```
The order is immaterial; both options produce identical results.

Or to see somatic missense mutations:
```
BRCA1: MUT = MISSENSE_SOMATIC
``` 

`GERMLINE` or `SOMATIC` can also be combined with `DRIVER` and, optionally, a more specific mutation term (e.g. `NONSENSE`):
```
BRCA1: NONSENSE_GERMLINE_DRIVER
```

## The DATATYPES Command
To save copying and pasting, the `DATATYPES` command sets the genetic annotation for all subsequent genes. Thus,
```
DATATYPES: AMP GAIN HOMDEL EXP > 1.5 EXP < -1.5; CDKN2A MDM2 TP53
```

is equivalent to:
```
CDKN2A: AMP GAIN HOMDEL EXP > 1.5 EXP < -1.5
MDM2: AMP GAIN HOMDEL EXP > 1.5 EXP < -1.5
TP53: AMP GAIN HOMDEL EXP > 1.5 EXP < -1.5
```

## Merged Gene Tracks
OQL can be used to create a merged gene track in OncoPrint, in which alterations in multiple genes appear as a single track. This is done by enclosing a list of genes in square brackets. By default, the track will be labeled by the gene names, separated by '/'. To instead specify a label, type the desired label within double quotes at the beginning of the square brackets. For example:
```
["CDK INHIBITORS" CDKN2A CDKN2B]
[MDM2 MDM4]
```

The resulting merged gene track will be visible in OncoPrint and can be expanded to view the individual gene tracks. For example:

![Image of merged genes in OncoPrint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/merged_genes_oncoprint.png)

<https://www.cbioportal.org/results/oncoprint?session_id=5c1966e2e4b05228701f958e>

It is possible to include OQL for specific alterations in merged gene tracks, as well as querying a combination of single and merged gene tracks.

Note that merged gene tracks only appear in OncoPrint. All other pages show the individual genes.

## Example: RB Pathway Alterations

Provided below is one example of the power of using OQL. Additional examples are available in the [User Guide](https://docs.cbioportal.org/user-guide/by-page/#oql).

### Using the Defaults
Select Ovarian Serous Cystadenocarcinoma (TCGA, Nature 2011) with the following data types:
* Mutations
* Putative copy-number alterations (GISTIC)
* mRNA expression (mRNA expression Z-scores (all genes))

Input the following three genes in the RB pathway:
* CCNE1
* RB1
* CDKN2A

![image of rb query](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/rb_query.png)

Submit this query and note how many samples have alterations in multiple of these genes:

![image of rb oncoprint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/rb_query_oncoprint.png)

<https://www.cbioportal.org/results/oncoprint?session_id=5c1966cee4b05228701f958d>

### Greater Insight with OQL
Given what is known about the RB pathway, the events that are most likely selected for in the tumors are CCNE1 amplification, RB1 deletions or mutations, and loss of expression of CDKN2A. To investigate this hypothesis, we can use OQL to display only these events. Modify the query to reflect this:
```
CCNE1: AMP MUT
RB1: HOMDEL MUT
CDKN2A: HOMDEL EXP < -1
```

Examine the updated OncoPrint:

![image of modified rb oncoprint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/rb_query_oncoprint_modified.png)

<https://www.cbioportal.org/results/oncoprint?session_id=5c1966aee4b05228701f958c>

This shows that alterations in these genes are almost entirely mutually-exclusive -- no cases are altered in all three genes and only six are altered in two genes. This supports the theory that the tumor has selected for these events.

## Questions? Feedback?
Please share any questions or feedback on OQL with us: <https://groups.google.com/group/cbioportal>

Also note that additional explanation and examples using OQL are available in the [User Guide](https://docs.cbioportal.org/user-guide/by-page/#oql).## Tutorials

### Tutorial Slides
These tutorial slides contain annoted screenshots to walk you through using the cBioPortal site.
1. Single Study Exploration [Google slides](https://docs.google.com/presentation/d/1_OGK69lO4Z62WaxHHkNYmWvY0LQN2v0slfaLyY1_IQ0) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%201%20Single%20Study%20Exploration.pdf)
2. Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
3. Patient View [Google slides](https://docs.google.com/presentation/d/1Jr_2yEfgjKBn4DBiXRk4kmhIbtsRp6gd0iD3k1fIUUk) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%203%20Patient%20View.pdf)
4. Virtual Studies [Google slides](https://docs.google.com/presentation/d/1rQE5rbFNdmup-rAtySHFxlLp3i4qa8SBA7MiQpMdn1I) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%204%20Virtual%20Studies.pdf)
5. Onco Query Language (OQL) [Google slides](https://docs.google.com/presentation/d/1U39xgVujtBodwW20qIfcGu4E5n2zzaKkl2KmzzHqj4A) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
6. Group Comparison [Google slides](https://docs.google.com/presentation/d/1P2boDph8IfpvjxoxDj_496CLHGtshzJnbbZhszPsmf4) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
7. Pathways [Google slides](https://docs.google.com/presentation/d/1O5WGucz0lrfdY25b5QS6zaID_26i434EYXBluqZfT2g) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%207%20Pathways.pdf)

### Webinar Recordings
Recordings of live webinars from April & May 2020
1. Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%201%20Introduction%20to%20cBioPortal.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing)
2. Mutation Details & Patient View [youtube.com](https://www.youtube.com/watch?v=uJsp9kd2jIk) | [bilibili.com](https://www.bilibili.com/video/BV1Qf4y1m7Lx) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%202%20Mutation%20Details%20and%20Patient%20View.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing)
3. Expression Data Analysis [youtube.com](https://www.youtube.com/watch?v=YUxVv6pkxD4) | [bilibili.com](https://www.bilibili.com/video/BV1HK4y1t7dE) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%203%20Expression%20Data%20Analysis.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing)
4. Group Comparison [youtube.com](https://www.youtube.com/watch?v=Tx4HZCrIe5c) | [bilibili.com](https://www.bilibili.com/video/BV1VZ4y1W76p) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%204%20Group%20Comparison.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing)
5. API & R Client [youtube.com](https://www.youtube.com/watch?v=Nq12o2i0yaw) | [bilibili.com](https://www.bilibili.com/video/BV1jz4y197iU) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%205%20API%20and%20R%20Client.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing) | [Workshop code](https://github.com/cBioPortal/2020-cbioportal-r-workshop)

### How-To Videos
Short videos that show how to perform specific analyses or how to use specific pages.
1. Comparing samples based on expression level of a gene [youtube.com](https://www.youtube.com/watch?v=HTiKUXk0j0s)
2. Proteomic profiles in cBioPortal - An example based on cancer cell lines from the Cancer Cell Line Encyclopedia (CCLE) [youtube.com](https://www.youtube.com/watch?v=62qbjQOH9qc) 
3. Filtering and adding clinical data to Mutations tab [youtube.com](https://www.youtube.com/watch?v=q9No2073c5o)
4. Exploring the longitudinal evolution of individual patients [youtube.com](https://www.youtube.com/watch?v=Hbbs-tHh9LQ)
5. Using Onco Query Language (OQL) to query based on the expression level of genes [youtube.com](https://www.youtube.com/watch?v=kHlFXw2TMzc)
6. How to explore the data in a study [youtube.com](https://www.youtube.com/watch?v=N8ffDgkqDWc)
7. How to run a query for genes of interest [youtube.com](https://www.youtube.com/watch?v=MH-kY5usA70)
8. How to download data [youtube.com](https://www.youtube.com/watch?v=JId304dp3tc)
9. Navigating AACR GENIE - Biopharma Collaborative (BPC) dataset [youtube.com](https://www.youtube.com/playlist?list=PLE2tjlUyHL8Gn4ythnuvf2BxWJVpAxkMM)

### Documentation
* Frequently Asked Questions [FAQ](/user-guide/faq.md)
* Onco Query Language [OQL](/user-guide/oql.md)

### Publications
* de Bruijn et al. Cancer Research 2023 [PubMed](https://pubmed.ncbi.nlm.nih.gov/37668528/)
* Cerami et al. Cancer Discovery 2012 [PubMed](http://cancerdiscovery.aacrjournals.org/content/2/5/401.abstract)
* Gao et al. Science Signaling 2013 [PubMed](https://www.ncbi.nlm.nih.gov/pubmed/23550210)

### Tutorials by others
* cBioPortal Tutorial Series by Jackson Laboratory [youtube.com](https://www.youtube.com/watch?v=tWdrcZOiXc0&list=PLWNp6Z5dXDZ7yRyny8aypSwrArs6e0EJi)
* Using the Cancer Digital Slide Archive in cBioPortal by Nicole M. Rivera Acevedo [youtube.com (English)](https://www.youtube.com/watch?v=oP0GbWmfVWM) | [youtube.com (Spanish)](https://www.youtube.com/watch?v=QyOVCxjVIPU)
* Visualizing and Downloading RNASeq data from cBioPortal by Farhan Haq [youtube.com](https://www.youtube.com/watch?v=I3KiLCE35Mk&list=PLvK49Q0ARs91Ilonq_07n0q9Cd5pOsdJW)
# User Guide

The cBioPortal for Cancer Genomics is a resource for interactive exploration of multidimensional cancer genomics data sets. The goal of cBioPortal is to significantly lower the barriers between complex genomic data and cancer researchers by providing rapid, intuitive, and high-quality access to molecular profiles and clinical attributes from large-scale cancer genomics projects, and therefore to empower researchers to translate these rich data sets into biologic insights and clinical applications.

The following section guides users in performing queries and analysis on any instance of cBioportal.
# Using the dataset validator

To facilitate the loading of new studies into its database, cBioPortal [provides a set of staging files formats](/File-Formats.md) for the various data types. To validate your files you can use the dataset validator script. 

## Running the validator

To run the validator first go to the importer folder
`<cbioportal_source_folder>/core/src/main/scripts/importer`
and then run the following command:
```bash
./validateData.py --help
```
This will tell you the parameters you can use: 
```console
usage: validateData.py [-h] -s STUDY_DIRECTORY
                       [-u URL_SERVER | -p PORTAL_INFO_DIR | -n]
                       [-P PORTAL_PROPERTIES] [-html HTML_TABLE]
                       [-e ERROR_FILE] [-v] [-r] [-m]

cBioPortal study validator

optional arguments:
  -h, --help            show this help message and exit
  -s STUDY_DIRECTORY, --study_directory STUDY_DIRECTORY
                        path to directory.
  -u URL_SERVER, --url_server URL_SERVER
                        URL to cBioPortal server. You can set this if your URL
                        is not http://localhost:8080
  -p PORTAL_INFO_DIR, --portal_info_dir PORTAL_INFO_DIR
                        Path to a directory of cBioPortal info files to be
                        used instead of contacting a server
  -n, --no_portal_checks
                        Skip tests requiring information from the cBioPortal
                        installation
  -html HTML_TABLE, --html_table HTML_TABLE
                        path to html report output file
  -e ERROR_FILE, --error_file ERROR_FILE
                        File to which to write line numbers on which errors
                        were found, for scripts
  -v, --verbose         report status info messages in addition to errors and
                        warnings
  -r, --relaxed-clinical_definitions
                        Option to enable relaxed mode for validator when validating
                        clinical data without header definitions
  -m, --strict_maf_checks
                        Option to enable strict mode for validator when validating
                        mutation data
```

For more information on the `--portal_info_dir` option, see [Offline validation](#offline-validation) below. If your cBioPortal is not using `hg19`, 
you have to specify the `reference_genome` field in your `meta_study.txt`.

For more information, see [Validation of non-human data](#validation-of-non-human-data).

When running the validator with parameter `-r` the validator will run the validation of the clinical data it will ignore all failing checks
about values in the headers of the clinical data file.

When running the validator with parameter `-m` the validator will run the validation of the specific [MAF file checks](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/) for the mutation file in strict maf check mode. This means that
when the validator encounters these validation checks it will report them as an error instead of a warning.

### Example 1: test study_es_0
As an example, you can try the validator with one of the test studies found in  `<cbioportal_source_folder>/core/src/test/scripts/test_data`. Example, assuming port 8080 and using -v option to also see the progress:
```bash
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -u http://localhost:8080 -v
```
Results in:
```console
DEBUG: -: Requesting info from portal at 'http://localhost:8080'
DEBUG: -: Requesting cancer-types from portal at 'http://localhost:8080'
DEBUG: -: Requesting genes from portal at 'http://localhost:8080'
DEBUG: -: Requesting genesets from portal at 'http://localhost:8080'
DEBUG: -: Requesting genesets_version from portal at 'http://localhost:8080'
DEBUG: -: Requesting gene-panels from portal at 'http://localhost:8080'

DEBUG: meta_cancer_type.txt: Starting validation of meta file
INFO: meta_cancer_type.txt: Validation of meta file complete

DEBUG: meta_clinical_patients.txt: Starting validation of meta file
INFO: meta_clinical_patients.txt: Validation of meta file complete

DEBUG: meta_clinical_samples.txt: Starting validation of meta file
INFO: meta_clinical_samples.txt: Validation of meta file complete

DEBUG: meta_cna_discrete.txt: Starting validation of meta file
INFO: meta_cna_discrete.txt: Validation of meta file complete

DEBUG: meta_cna_hg19_seg.txt: Starting validation of meta file
INFO: meta_cna_hg19_seg.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_cna_log2.txt: Starting validation of meta file
INFO: meta_cna_log2.txt: Validation of meta file complete

DEBUG: meta_expression_median.txt: Starting validation of meta file
INFO: meta_expression_median.txt: Validation of meta file complete

DEBUG: meta_expression_median_Zscores.txt: Starting validation of meta file
INFO: meta_expression_median_Zscores.txt: Validation of meta file complete

DEBUG: meta_fusions.txt: Starting validation of meta file
INFO: meta_fusions.txt: Validation of meta file complete

DEBUG: meta_gene_panel_matrix.txt: Starting validation of meta file
INFO: meta_gene_panel_matrix.txt: Validation of meta file complete

DEBUG: meta_gistic_genes_amp.txt: Starting validation of meta file
INFO: meta_gistic_genes_amp.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_gsva_pvalues.txt: Starting validation of meta file
INFO: meta_gsva_pvalues.txt: Validation of meta file complete

DEBUG: meta_gsva_scores.txt: Starting validation of meta file
INFO: meta_gsva_scores.txt: Validation of meta file complete

DEBUG: meta_methylation_hm27.txt: Starting validation of meta file
INFO: meta_methylation_hm27.txt: Validation of meta file complete

DEBUG: meta_mutational_signature.txt: Starting validation of meta file
INFO: meta_mutational_signature.txt: Validation of meta file complete

DEBUG: meta_mutations_extended.txt: Starting validation of meta file
INFO: meta_mutations_extended.txt: Validation of meta file complete

DEBUG: meta_resource_definition.txt: Starting validation of meta file
INFO: meta_resource_definition.txt: Validation of meta file complete

DEBUG: meta_resource_patient.txt: Starting validation of meta file
INFO: meta_resource_patient.txt: Validation of meta file complete

DEBUG: meta_resource_sample.txt: Starting validation of meta file
INFO: meta_resource_sample.txt: Validation of meta file complete

DEBUG: meta_resource_study.txt: Starting validation of meta file
INFO: meta_resource_study.txt: Validation of meta file complete

DEBUG: meta_structural_variants.txt: Starting validation of meta file
INFO: meta_structural_variants.txt: Validation of meta file complete

DEBUG: meta_study.txt: Starting validation of meta file
INFO: meta_study.txt: Validation of meta file complete

DEBUG: -: Study Tag file found. It will be validated.

DEBUG: meta_treatment_ec50.txt: Starting validation of meta file
INFO: meta_treatment_ec50.txt: Validation of meta file complete

DEBUG: meta_treatment_ic50.txt: Starting validation of meta file
INFO: meta_treatment_ic50.txt: Validation of meta file complete

DEBUG: data_cancer_type.txt: Starting validation of file
INFO: data_cancer_type.txt: line 1: New disease type will be added to the portal; value encountered: 'brca-es0'
INFO: data_cancer_type.txt: Validation of file complete
INFO: data_cancer_type.txt: Read 1 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_samples.txt: Starting validation of file
INFO: data_clinical_samples.txt: Validation of file complete
INFO: data_clinical_samples.txt: Read 847 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_definition.txt: Starting validation of file
INFO: data_resource_definition.txt: Validation of file complete
INFO: data_resource_definition.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_sample.txt: Starting validation of file
INFO: data_resource_sample.txt: Validation of file complete
INFO: data_resource_sample.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: study_tags.yml: Starting validation of study tags file
INFO: study_tags.yml: Validation of study tags file complete.

DEBUG: -: Validating case lists

DEBUG: case_lists/cases_cnaseq.txt: Starting validation of meta file
INFO: case_lists/cases_cnaseq.txt: Validation of meta file complete

DEBUG: case_lists/cases_test.txt: Starting validation of meta file
INFO: case_lists/cases_test.txt: Validation of meta file complete

DEBUG: case_lists/cases_sequenced.txt: Starting validation of meta file
INFO: case_lists/cases_sequenced.txt: Validation of meta file complete

DEBUG: case_lists/cases_custom.txt: Starting validation of meta file
INFO: case_lists/cases_custom.txt: Validation of meta file complete

DEBUG: case_lists/cases_cna.txt: Starting validation of meta file
INFO: case_lists/cases_cna.txt: Validation of meta file complete

INFO: -: Validation of case list folder complete

DEBUG: data_gene_panel_matrix.txt: Starting validation of file
INFO: data_gene_panel_matrix.txt: line 1: This column can be replaced by a 'gene_panel' property in the respective meta file; value encountered: 'gistic'
INFO: data_gene_panel_matrix.txt: Validation of file complete
INFO: data_gene_panel_matrix.txt: Read 21 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_discrete.txt: Starting validation of file
INFO: data_cna_discrete.txt: Validation of file complete
INFO: data_cna_discrete.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_patients.txt: Starting validation of file
INFO: data_clinical_patients.txt: Validation of file complete
INFO: data_clinical_patients.txt: Read 845 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median.txt: Starting validation of file
INFO: data_expression_median.txt: Validation of file complete
INFO: data_expression_median.txt: Read 7 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median_Zscores.txt: Starting validation of file
INFO: data_expression_median_Zscores.txt: Validation of file complete
INFO: data_expression_median_Zscores.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_fusions.txt: Starting validation of file
INFO: data_fusions.txt: Validation of file complete
INFO: data_fusions.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutational_signature.txt: Starting validation of file
INFO: data_mutational_signature.txt: Validation of file complete
INFO: data_mutational_signature.txt: Read 62 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ec50.txt: Starting validation of file
INFO: data_treatment_ec50.txt: Validation of file complete
INFO: data_treatment_ec50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ic50.txt: Starting validation of file
INFO: data_treatment_ic50.txt: Validation of file complete
INFO: data_treatment_ic50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gistic_genes_amp.txt: Starting validation of file
INFO: data_gistic_genes_amp.txt: Validation of file complete
INFO: data_gistic_genes_amp.txt: Read 13 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_pvalues.txt: Starting validation of file
INFO: data_gsva_pvalues.txt: Validation of file complete
INFO: data_gsva_pvalues.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_scores.txt: Starting validation of file
INFO: data_gsva_scores.txt: Validation of file complete
INFO: data_gsva_scores.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_log2.txt: Starting validation of file
INFO: data_cna_log2.txt: Validation of file complete
INFO: data_cna_log2.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_methylation_hm27.txt: Starting validation of file
INFO: data_methylation_hm27.txt: Validation of file complete
INFO: data_methylation_hm27.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutations_extended.maf: Starting validation of file
INFO: data_mutations_extended.maf: lines [4, 5, 6, (3 more)]: column 164: Values contained in the column cbp_driver_tiers that will appear in the "Mutation Color" menu of the Oncoprint; values encountered: ['Class 2', 'Class 1', 'Class 4', '(1 more)']
INFO: data_mutations_extended.maf: lines [7, 9]: Line will not be loaded due to the variant classification filter. Filtered types: [Silent, Intron, 3'UTR, 3'Flank, 5'UTR, 5'Flank, IGR, RNA]; value encountered: 'Silent'
INFO: data_mutations_extended.maf: Validation of file complete
INFO: data_mutations_extended.maf: Read 35 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_patient.txt: Starting validation of file
INFO: data_resource_patient.txt: Validation of file complete
INFO: data_resource_patient.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_study.txt: Starting validation of file
INFO: data_resource_study.txt: Validation of file complete
INFO: data_resource_study.txt: Read 2 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_hg19.seg: Starting validation of file
INFO: data_cna_hg19.seg: Validation of file complete
INFO: data_cna_hg19.seg: Read 10 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_structural_variants.txt: Starting validation of file
INFO: data_structural_variants.txt: Validation of file complete
INFO: data_structural_variants.txt: Read 46 lines. Lines with warning: 0. Lines with error: 0

INFO: -: Validation complete
Validation of study succeeded.
```

When using the `-html` option, a report will be generated, which looks like this for the previous example:
![Screenshot of a successful validation report](images/scripts/report.png)

### Example 2: test study_es_1
More test studies for trying the validator (`study_es_1` and `study_es_3`) are available in  `<cbioportal_source_folder>/core/src/test/scripts/test_data`. Example, assuming port 8080 and using -v option:
```bash
./validateData.py -s ../../../test/scripts/test_data/study_es_1/ -u http://localhost:8080 -v
```
Results in:
```console
DEBUG: -: Requesting info from portal at 'http://localhost:8081'
DEBUG: -: Requesting cancer-types from portal at 'http://localhost:8081'
DEBUG: -: Requesting genes from portal at 'http://localhost:8081'
DEBUG: -: Requesting genesets from portal at 'http://localhost:8081'
DEBUG: -: Requesting genesets_version from portal at 'http://localhost:8081'
DEBUG: -: Requesting gene-panels from portal at 'http://localhost:8081'

DEBUG: meta_expression_median.txt: Starting validation of meta file
ERROR: meta_expression_median.txt: Invalid stable id for genetic_alteration_type 'MRNA_EXPRESSION', data_type 'Z-SCORE'; expected one of [mrna_U133_Zscores, rna_seq_mrna_median_Zscores, mrna_median_Zscores, rna_seq_v2_mrna_median_Zscores, rna_seq_v2_mrna_median_normals_Zscores, mirna_median_Zscores, mrna_merged_median_Zscores, mrna_zbynorm, mrna_seq_tpm_Zscores, mrna_seq_cpm_Zscores, rna_seq_mrna_capture_Zscores, mrna_seq_fpkm_capture_Zscores, mrna_seq_fpkm_polya_Zscores, mrna_U133_all_sample_Zscores, mrna_all_sample_Zscores, rna_seq_mrna_median_all_sample_Zscores, mrna_median_all_sample_Zscores, rna_seq_v2_mrna_median_all_sample_Zscores, rna_seq_v2_mrna_median_all_sample_ref_normal_Zscores, mrna_seq_cpm_all_sample_Zscores, mrna_seq_tpm_all_sample_Zscores, rna_seq_mrna_capture_all_sample_Zscores, mrna_seq_fpkm_capture_all_sample_Zscores, mrna_seq_fpkm_polya_all_sample_Zscores]; value encountered: 'mrna'

DEBUG: meta_samples.txt: Starting validation of meta file
WARNING: meta_samples.txt: Unrecognized field in meta file; values encountered: ['show_profile_in_analysis_tab', 'profile_description', 'profile_name']
INFO: meta_samples.txt: Validation of meta file complete

DEBUG: meta_study.txt: Starting validation of meta file
INFO: meta_study.txt: Validation of meta file complete
INFO: meta_study.txt: No reference genome specified -- using default (hg19)

DEBUG: meta_treatment_ec50.txt: Starting validation of meta file
INFO: meta_treatment_ec50.txt: Validation of meta file complete

DEBUG: meta_treatment_ic50.txt: Starting validation of meta file
INFO: meta_treatment_ic50.txt: Validation of meta file complete

DEBUG: data_samples.txt: Starting validation of file
INFO: data_samples.txt: Validation of file complete
INFO: data_samples.txt: Read 831 lines. Lines with warning: 0. Lines with error: 0

DEBUG: -: Validating case lists

DEBUG: case_lists/cases_all.txt: Starting validation of meta file
INFO: case_lists/cases_all.txt: Validation of meta file complete
ERROR: case_lists/cases_all.txt: Sample ID not defined in clinical file; value encountered: 'INVALID-A2-A0T2-01'

INFO: -: Validation of case list folder complete

DEBUG: data_treatment_ec50.txt: Starting validation of file
ERROR: data_treatment_ec50.txt: line 2: column 1: Do not use space in the stable id; value encountered: '17 AAG'
ERROR: data_treatment_ec50.txt: line 7: column 5: Blank cell found in column; value encountered: ''' (in column 'TCGA-A1-A0SB-01')'
INFO: data_treatment_ec50.txt: Validation of file complete
INFO: data_treatment_ec50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 2

DEBUG: data_treatment_ic50.txt: Starting validation of file
ERROR: data_treatment_ic50.txt: line 7: column 5: Blank cell found in column; value encountered: ''' (in column 'TCGA-A1-A0SB-01')'
INFO: data_treatment_ic50.txt: Validation of file complete
INFO: data_treatment_ic50.txt: Read 10 lines. Lines with warning: 0. Lines with error: 1

INFO: -: Validation complete
Validation of study failed.
```

And respective HTML report:
![Screenshot of an unsuccessful validation report](images/scripts/report1.png)

## Offline validation ##
The validation script can be used offline, without connecting to a cBioPortal server. The tests that depend on information specific to the portal (which clinical attributes and cancer types have been previously defined, and which Entrez gene identifiers and corresponding symbols are supported), will instead be read from a folder with .json files generated from the portal.

### Example 3: validation with a portal info folder ###
To run the validator with a folder of portal information files, add the `-p/--portal_info_dir` option to the command line, followed by the path to the folder:
```bash
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -p ../../../test/scripts/test_data/api_json_system_tests/ -v
```
```console
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/cancer-types.json
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/genes.json
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/genesets.json
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/genesets_version.json
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/gene-panels.json

DEBUG: meta_cancer_type.txt: Starting validation of meta file
INFO: meta_cancer_type.txt: Validation of meta file complete

DEBUG: meta_clinical_patients.txt: Starting validation of meta file
INFO: meta_clinical_patients.txt: Validation of meta file complete

DEBUG: meta_clinical_samples.txt: Starting validation of meta file
INFO: meta_clinical_samples.txt: Validation of meta file complete

DEBUG: meta_cna_discrete.txt: Starting validation of meta file
INFO: meta_cna_discrete.txt: Validation of meta file complete

DEBUG: meta_cna_hg19_seg.txt: Starting validation of meta file
INFO: meta_cna_hg19_seg.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_cna_log2.txt: Starting validation of meta file
INFO: meta_cna_log2.txt: Validation of meta file complete

DEBUG: meta_expression_median.txt: Starting validation of meta file
INFO: meta_expression_median.txt: Validation of meta file complete

DEBUG: meta_expression_median_Zscores.txt: Starting validation of meta file
INFO: meta_expression_median_Zscores.txt: Validation of meta file complete

DEBUG: meta_fusions.txt: Starting validation of meta file
INFO: meta_fusions.txt: Validation of meta file complete

DEBUG: meta_gene_panel_matrix.txt: Starting validation of meta file
INFO: meta_gene_panel_matrix.txt: Validation of meta file complete

DEBUG: meta_gistic_genes_amp.txt: Starting validation of meta file
INFO: meta_gistic_genes_amp.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_gsva_pvalues.txt: Starting validation of meta file
INFO: meta_gsva_pvalues.txt: Validation of meta file complete

DEBUG: meta_gsva_scores.txt: Starting validation of meta file
INFO: meta_gsva_scores.txt: Validation of meta file complete

DEBUG: meta_methylation_hm27.txt: Starting validation of meta file
INFO: meta_methylation_hm27.txt: Validation of meta file complete

DEBUG: meta_mutational_signature.txt: Starting validation of meta file
INFO: meta_mutational_signature.txt: Validation of meta file complete

DEBUG: meta_mutations_extended.txt: Starting validation of meta file
INFO: meta_mutations_extended.txt: Validation of meta file complete

DEBUG: meta_resource_definition.txt: Starting validation of meta file
INFO: meta_resource_definition.txt: Validation of meta file complete

DEBUG: meta_resource_patient.txt: Starting validation of meta file
INFO: meta_resource_patient.txt: Validation of meta file complete

DEBUG: meta_resource_sample.txt: Starting validation of meta file
INFO: meta_resource_sample.txt: Validation of meta file complete

DEBUG: meta_resource_study.txt: Starting validation of meta file
INFO: meta_resource_study.txt: Validation of meta file complete

DEBUG: meta_structural_variants.txt: Starting validation of meta file
INFO: meta_structural_variants.txt: Validation of meta file complete

DEBUG: meta_study.txt: Starting validation of meta file
INFO: meta_study.txt: Validation of meta file complete

DEBUG: -: Study Tag file found. It will be validated.

DEBUG: meta_treatment_ec50.txt: Starting validation of meta file
INFO: meta_treatment_ec50.txt: Validation of meta file complete

DEBUG: meta_treatment_ic50.txt: Starting validation of meta file
INFO: meta_treatment_ic50.txt: Validation of meta file complete

DEBUG: data_cancer_type.txt: Starting validation of file
INFO: data_cancer_type.txt: Validation of file complete
INFO: data_cancer_type.txt: Read 1 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_samples.txt: Starting validation of file
INFO: data_clinical_samples.txt: Validation of file complete
INFO: data_clinical_samples.txt: Read 847 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_definition.txt: Starting validation of file
INFO: data_resource_definition.txt: Validation of file complete
INFO: data_resource_definition.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_sample.txt: Starting validation of file
INFO: data_resource_sample.txt: Validation of file complete
INFO: data_resource_sample.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: study_tags.yml: Starting validation of study tags file
INFO: study_tags.yml: Validation of study tags file complete.

DEBUG: -: Validating case lists

DEBUG: case_lists/cases_cnaseq.txt: Starting validation of meta file
INFO: case_lists/cases_cnaseq.txt: Validation of meta file complete

DEBUG: case_lists/cases_test.txt: Starting validation of meta file
INFO: case_lists/cases_test.txt: Validation of meta file complete

DEBUG: case_lists/cases_sequenced.txt: Starting validation of meta file
INFO: case_lists/cases_sequenced.txt: Validation of meta file complete

DEBUG: case_lists/cases_custom.txt: Starting validation of meta file
INFO: case_lists/cases_custom.txt: Validation of meta file complete

DEBUG: case_lists/cases_cna.txt: Starting validation of meta file
INFO: case_lists/cases_cna.txt: Validation of meta file complete

INFO: -: Validation of case list folder complete

DEBUG: data_gene_panel_matrix.txt: Starting validation of file
INFO: data_gene_panel_matrix.txt: line 1: This column can be replaced by a 'gene_panel' property in the respective meta file; value encountered: 'gistic'
INFO: data_gene_panel_matrix.txt: Validation of file complete
INFO: data_gene_panel_matrix.txt: Read 21 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_discrete.txt: Starting validation of file
INFO: data_cna_discrete.txt: Validation of file complete
INFO: data_cna_discrete.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_patients.txt: Starting validation of file
INFO: data_clinical_patients.txt: Validation of file complete
INFO: data_clinical_patients.txt: Read 845 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median.txt: Starting validation of file
INFO: data_expression_median.txt: Validation of file complete
INFO: data_expression_median.txt: Read 7 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median_Zscores.txt: Starting validation of file
INFO: data_expression_median_Zscores.txt: Validation of file complete
INFO: data_expression_median_Zscores.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_fusions.txt: Starting validation of file
INFO: data_fusions.txt: Validation of file complete
INFO: data_fusions.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutational_signature.txt: Starting validation of file
INFO: data_mutational_signature.txt: Validation of file complete
INFO: data_mutational_signature.txt: Read 62 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ec50.txt: Starting validation of file
INFO: data_treatment_ec50.txt: Validation of file complete
INFO: data_treatment_ec50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ic50.txt: Starting validation of file
INFO: data_treatment_ic50.txt: Validation of file complete
INFO: data_treatment_ic50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gistic_genes_amp.txt: Starting validation of file
INFO: data_gistic_genes_amp.txt: Validation of file complete
INFO: data_gistic_genes_amp.txt: Read 13 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_pvalues.txt: Starting validation of file
INFO: data_gsva_pvalues.txt: Validation of file complete
INFO: data_gsva_pvalues.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_scores.txt: Starting validation of file
INFO: data_gsva_scores.txt: Validation of file complete
INFO: data_gsva_scores.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_log2.txt: Starting validation of file
INFO: data_cna_log2.txt: Validation of file complete
INFO: data_cna_log2.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_methylation_hm27.txt: Starting validation of file
INFO: data_methylation_hm27.txt: Validation of file complete
INFO: data_methylation_hm27.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutations_extended.maf: Starting validation of file
INFO: data_mutations_extended.maf: lines [4, 5, 6, (3 more)]: column 164: Values contained in the column cbp_driver_tiers that will appear in the "Mutation Color" menu of the Oncoprint; values encountered: ['Class 2', 'Class 1', 'Class 4', '(1 more)']
INFO: data_mutations_extended.maf: lines [7, 9]: Line will not be loaded due to the variant classification filter. Filtered types: [Silent, Intron, 3'UTR, 3'Flank, 5'UTR, 5'Flank, IGR, RNA]; value encountered: 'Silent'
INFO: data_mutations_extended.maf: Validation of file complete
INFO: data_mutations_extended.maf: Read 35 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_patient.txt: Starting validation of file
INFO: data_resource_patient.txt: Validation of file complete
INFO: data_resource_patient.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_study.txt: Starting validation of file
INFO: data_resource_study.txt: Validation of file complete
INFO: data_resource_study.txt: Read 2 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_hg19.seg: Starting validation of file
INFO: data_cna_hg19.seg: Validation of file complete
INFO: data_cna_hg19.seg: Read 10 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_structural_variants.txt: Starting validation of file
INFO: data_structural_variants.txt: Validation of file complete
INFO: data_structural_variants.txt: Read 46 lines. Lines with warning: 0. Lines with error: 0

INFO: -: Validation complete
Validation of study succeeded.
```

### Example 4: generating the portal info folder ###
The portal information files can be generated on the server, using the dumpPortalInfo script. Go to `<cbioportal_source_folder>/core/src/main/scripts`, make sure the environment variables `$JAVA_HOME` and `$PORTAL_HOME` are set, and run dumpPortalInfo.pl with the name of the directory you want to create:
```bash
export JAVA_HOME='/usr/lib/jvm/default-java'
export PORTAL_HOME=<cbioportal_configuration_folder>
./dumpPortalInfo.pl /home/johndoe/my_portal_info_folder/
```

### Example 5: validating without portal-specific information ###
Alternatively, you can run the validation script with the `-n/--no_portal_checks` flag to entirely skip checks relating to installation-specific metadata. Be warned that files succeeding this validation may still fail to load (correctly).

```bash
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -n -v
```
```console
WARNING: -: Skipping validations relating to cancer types defined in the portal
WARNING: -: Skipping validations relating to gene identifiers and aliases defined in the portal
WARNING: -: Skipping validations relating to gene set identifiers
WARNING: -: Skipping validations relating to gene panel identifiers

DEBUG: meta_cancer_type.txt: Starting validation of meta file
INFO: meta_cancer_type.txt: Validation of meta file complete

DEBUG: meta_clinical_patients.txt: Starting validation of meta file
INFO: meta_clinical_patients.txt: Validation of meta file complete

DEBUG: meta_clinical_samples.txt: Starting validation of meta file
INFO: meta_clinical_samples.txt: Validation of meta file complete

DEBUG: meta_cna_discrete.txt: Starting validation of meta file
INFO: meta_cna_discrete.txt: Validation of meta file complete

DEBUG: meta_cna_hg19_seg.txt: Starting validation of meta file
INFO: meta_cna_hg19_seg.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_cna_log2.txt: Starting validation of meta file
INFO: meta_cna_log2.txt: Validation of meta file complete

DEBUG: meta_expression_median.txt: Starting validation of meta file
INFO: meta_expression_median.txt: Validation of meta file complete

DEBUG: meta_expression_median_Zscores.txt: Starting validation of meta file
INFO: meta_expression_median_Zscores.txt: Validation of meta file complete

DEBUG: meta_fusions.txt: Starting validation of meta file
INFO: meta_fusions.txt: Validation of meta file complete

DEBUG: meta_gene_panel_matrix.txt: Starting validation of meta file
INFO: meta_gene_panel_matrix.txt: Validation of meta file complete

DEBUG: meta_gistic_genes_amp.txt: Starting validation of meta file
INFO: meta_gistic_genes_amp.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_gsva_pvalues.txt: Starting validation of meta file
INFO: meta_gsva_pvalues.txt: Validation of meta file complete

DEBUG: meta_gsva_scores.txt: Starting validation of meta file
INFO: meta_gsva_scores.txt: Validation of meta file complete

DEBUG: meta_methylation_hm27.txt: Starting validation of meta file
INFO: meta_methylation_hm27.txt: Validation of meta file complete

DEBUG: meta_mutational_signature.txt: Starting validation of meta file
INFO: meta_mutational_signature.txt: Validation of meta file complete

DEBUG: meta_mutations_extended.txt: Starting validation of meta file
INFO: meta_mutations_extended.txt: Validation of meta file complete

DEBUG: meta_resource_definition.txt: Starting validation of meta file
INFO: meta_resource_definition.txt: Validation of meta file complete

DEBUG: meta_resource_patient.txt: Starting validation of meta file
INFO: meta_resource_patient.txt: Validation of meta file complete

DEBUG: meta_resource_sample.txt: Starting validation of meta file
INFO: meta_resource_sample.txt: Validation of meta file complete

DEBUG: meta_resource_study.txt: Starting validation of meta file
INFO: meta_resource_study.txt: Validation of meta file complete

DEBUG: meta_structural_variants.txt: Starting validation of meta file
INFO: meta_structural_variants.txt: Validation of meta file complete

DEBUG: meta_study.txt: Starting validation of meta file
INFO: meta_study.txt: Validation of meta file complete

DEBUG: -: Study Tag file found. It will be validated.

DEBUG: meta_treatment_ec50.txt: Starting validation of meta file
INFO: meta_treatment_ec50.txt: Validation of meta file complete

DEBUG: meta_treatment_ic50.txt: Starting validation of meta file
INFO: meta_treatment_ic50.txt: Validation of meta file complete

DEBUG: data_cancer_type.txt: Starting validation of file
INFO: data_cancer_type.txt: Validation of file complete
INFO: data_cancer_type.txt: Read 1 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_samples.txt: Starting validation of file
INFO: data_clinical_samples.txt: Validation of file complete
INFO: data_clinical_samples.txt: Read 847 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_definition.txt: Starting validation of file
INFO: data_resource_definition.txt: Validation of file complete
INFO: data_resource_definition.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_sample.txt: Starting validation of file
INFO: data_resource_sample.txt: Validation of file complete
INFO: data_resource_sample.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: study_tags.yml: Starting validation of study tags file
INFO: study_tags.yml: Validation of study tags file complete.

DEBUG: -: Validating case lists

DEBUG: case_lists/cases_cnaseq.txt: Starting validation of meta file
INFO: case_lists/cases_cnaseq.txt: Validation of meta file complete

DEBUG: case_lists/cases_test.txt: Starting validation of meta file
INFO: case_lists/cases_test.txt: Validation of meta file complete

DEBUG: case_lists/cases_sequenced.txt: Starting validation of meta file
INFO: case_lists/cases_sequenced.txt: Validation of meta file complete

DEBUG: case_lists/cases_custom.txt: Starting validation of meta file
INFO: case_lists/cases_custom.txt: Validation of meta file complete

DEBUG: case_lists/cases_cna.txt: Starting validation of meta file
INFO: case_lists/cases_cna.txt: Validation of meta file complete

INFO: -: Validation of case list folder complete

DEBUG: data_gene_panel_matrix.txt: Starting validation of file
INFO: data_gene_panel_matrix.txt: line 1: This column can be replaced by a 'gene_panel' property in the respective meta file; value encountered: 'gistic'
INFO: data_gene_panel_matrix.txt: Validation of file complete
INFO: data_gene_panel_matrix.txt: Read 21 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_discrete.txt: Starting validation of file
INFO: data_cna_discrete.txt: Validation of file complete
INFO: data_cna_discrete.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_patients.txt: Starting validation of file
INFO: data_clinical_patients.txt: Validation of file complete
INFO: data_clinical_patients.txt: Read 845 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median.txt: Starting validation of file
INFO: data_expression_median.txt: Validation of file complete
INFO: data_expression_median.txt: Read 7 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median_Zscores.txt: Starting validation of file
INFO: data_expression_median_Zscores.txt: Validation of file complete
INFO: data_expression_median_Zscores.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_fusions.txt: Starting validation of file
INFO: data_fusions.txt: Validation of file complete
INFO: data_fusions.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutational_signature.txt: Starting validation of file
INFO: data_mutational_signature.txt: Validation of file complete
INFO: data_mutational_signature.txt: Read 62 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ec50.txt: Starting validation of file
INFO: data_treatment_ec50.txt: Validation of file complete
INFO: data_treatment_ec50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ic50.txt: Starting validation of file
INFO: data_treatment_ic50.txt: Validation of file complete
INFO: data_treatment_ic50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gistic_genes_amp.txt: Starting validation of file
INFO: data_gistic_genes_amp.txt: Validation of file complete
INFO: data_gistic_genes_amp.txt: Read 13 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_pvalues.txt: Starting validation of file
INFO: data_gsva_pvalues.txt: Validation of file complete
INFO: data_gsva_pvalues.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_scores.txt: Starting validation of file
INFO: data_gsva_scores.txt: Validation of file complete
INFO: data_gsva_scores.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_log2.txt: Starting validation of file
INFO: data_cna_log2.txt: Validation of file complete
INFO: data_cna_log2.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_methylation_hm27.txt: Starting validation of file
INFO: data_methylation_hm27.txt: Validation of file complete
INFO: data_methylation_hm27.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutations_extended.maf: Starting validation of file
INFO: data_mutations_extended.maf: lines [4, 5, 6, (3 more)]: column 164: Values contained in the column cbp_driver_tiers that will appear in the "Mutation Color" menu of the Oncoprint; values encountered: ['Class 2', 'Class 1', 'Class 4', '(1 more)']
INFO: data_mutations_extended.maf: lines [7, 9]: Line will not be loaded due to the variant classification filter. Filtered types: [Silent, Intron, 3'UTR, 3'Flank, 5'UTR, 5'Flank, IGR, RNA]; value encountered: 'Silent'
INFO: data_mutations_extended.maf: Validation of file complete
INFO: data_mutations_extended.maf: Read 35 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_patient.txt: Starting validation of file
INFO: data_resource_patient.txt: Validation of file complete
INFO: data_resource_patient.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_study.txt: Starting validation of file
INFO: data_resource_study.txt: Validation of file complete
INFO: data_resource_study.txt: Read 2 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_hg19.seg: Starting validation of file
INFO: data_cna_hg19.seg: Validation of file complete
INFO: data_cna_hg19.seg: Read 10 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_structural_variants.txt: Starting validation of file
INFO: data_structural_variants.txt: Validation of file complete
INFO: data_structural_variants.txt: Read 46 lines. Lines with warning: 0. Lines with error: 0

INFO: -: Validation complete
Validation of study succeeded with warnings.
```

## Validation of non-human data ##
When importing a study with a reference genome other than hg19/GRCh37, this should be specified in the `meta_study.txt` file, next to the `reference_genome` field. Supported values are **hg19**, **hg38** and **mm10**.

cBioPortal is gradually introducing support for mouse. If you want to load mouse studies and you have to [set up your database for mouse](/deployment/deploy-without-docker/Import-the-Seed-Database.md#download-the-cbioportal-seed-database).

As an example, the command for the mouse example using the three parameters is given:
```
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -P ../../../../../src/main/resources/application.properties -u http://localhost:8080 -v
```

## Running the validator for multiple studies
The importer folder `<cbioportal_source_folder>/core/src/main/scripts/importer` also contains a script for running the validator for multiple studies:
```bash
./validateStudies.py --help
```
The following parameters can be used:
```console
usage: validateStudies.py [-h] [-d ROOT_DIRECTORY] [-l LIST_OF_STUDIES]
                          [-html HTML_FOLDER]
                          [-u URL_SERVER | -p PORTAL_INFO_DIR | -n]
                          [-P PORTAL_PROPERTIES] [-m]

Wrapper where cBioPortal study validator is run for multiple studies

optional arguments:
  -h, --help            show this help message and exit
  -d ROOT_DIRECTORY, --root-directory ROOT_DIRECTORY
                        Path to directory with all studies that should be
                        validated
  -l LIST_OF_STUDIES, --list-of-studies LIST_OF_STUDIES
                        List with paths of studies which should be validated
  -html HTML_FOLDER, --html-folder HTML_FOLDER
                        Path to folder for output HTML reports
  -u URL_SERVER, --url_server URL_SERVER
                        URL to cBioPortal server. You can set this if your URL
                        is not http://localhost:8080
  -p PORTAL_INFO_DIR, --portal_info_dir PORTAL_INFO_DIR
                        Path to a directory of cBioPortal info files to be
                        used instead of contacting a server
  -n, --no_portal_checks
                        Skip tests requiring information from the cBioPortal
                        installation 
  -m, --strict_maf_checks
                        Option to enable strict mode for validator when
                        validating mutation data
```

Parameters `--url_server`, `--portal_info_dir`, `--no_portal_checks` and `--portal_properties` are equal to the parameters with the same name in `validateData.py`. The script will save a log file with validation output (`log-validate-studies.txt`) and output the validation status from the input studies:

```console
=== Validating study ../../../test/scripts/test_data/study_es_0
Result: VALID (WITH WARNINGS)

=== Validating study ../../../test/scripts/test_data/study_es_1
Result: INVALID

=== Validating study ../../../test/scripts/test_data/study_es_invalid
directory cannot be found: ../../../test/scripts/test_data/study_es_invalid
Result: INVALID (PROBLEMS OCCURRED)

```

### Example 1: Root directory parameter
Validation can be run for all studies in a certain directory by using the `--root-directory` parameter. The script will append each folder in the root directory to the study list to validate:
```bash
./validateStudies.py -d ../../../test/scripts/test_data/
```

### Example 2: List of studies parameter
Validation can also be run for specific studies by using the `--list-of-studies` parameter. The paths to the different studies can be defined and seperated by a comma:
```bash
./validateStudies.py -l ../../../test/scripts/test_data/study_es_0,../../../test/scripts/test_data/study_es_1
```

### Example 3: Combination root directory and list of studies parameter
Validation can also be run on specific studies in a certain directory by combining the `--root-directory` and `--list-of-studies` parameter:
```bash
./validateStudies.py -d ../../../test/scripts/test_data/ -l study_es_0,study_es_1
```

### Example 4: HTML folder parameter
When HTML validation reports are desired, an output folder for these HTML files can be specified. This folder does not have to exist, the script can create the folder. The HTML validation reports will get the following name: `<study_name>-validation.html`. To create HTML validation reports for each study the `--html-folder` parameter needs to be defined:
```bash
./validateStudies.py -d ../../../test/scripts/test_data/ -l study_es_0,study_es_1 -html ../../../test/scripts/test_data/validation-reports
```
## Importing Data into cBioPortal
The metaImport script should be used to automate the process of validating and loading datasets. It also has some nice features like an extra option to only load datasets that completely pass validation (i.e. with no errors, while warnings can be explicitly allowed by the user). 

#### Running the metaImport Script
To run the metaImport script first go to the importer folder
`<your_cbioportal_dir>/core/src/main/scripts/importer` 
and then run the following command:
```
./metaImport.py -h
```
This will tell you the parameters you can use:
```
$./metaImport.py -h
usage: metaImport.py [-h] -s STUDY_DIRECTORY
                     [-u URL_SERVER | -p PORTAL_INFO_DIR | -n]
                     [-jar JAR_PATH] [-html HTML_TABLE]
                     [-v] [-o] [-r] [-m]

cBioPortal meta Importer

optional arguments:
  -h, --help            show this help message and exit
  -s STUDY_DIRECTORY, --study_directory STUDY_DIRECTORY
                        path to directory.
  -u URL_SERVER, --url_server URL_SERVER
                        URL to cBioPortal server. You can set this if your URL
                        is not http://localhost/cbioportal
  -p PORTAL_INFO_DIR, --portal_info_dir PORTAL_INFO_DIR
                        Path to a directory of cBioPortal info files to be
                        used instead of contacting the web API
  -n, --no_portal_checks
                        Skip tests requiring information from the cBioPortal
                        installation
  -jar JAR_PATH, --jar_path JAR_PATH
                        Path to scripts JAR file (default: locate it relative
                        to the import script)
  -html HTML_TABLE, --html_table HTML_TABLE
                        path to html report
  -v, --verbose         report status info messages while validating
  -o, --override_warning
                        override warnings and continue importing
  -r, --relaxed_clinical_definitions
                        Option to enable relaxed mode for validator when
                        validating clinical data without header definitions
  -m, --strict_maf_checks
                        Option to enable strict mode for validator when
                        validating mutation data
```

#### Example of Importing a study
Export `PORTAL_HOME` as explained [here](/deployment/deploy-without-docker/Load-Sample-Cancer-Study.md), e.g.

```
export PORTAL_HOME=<cbioportal_configuration_folder>
```

and then run (this simple command only works if your cBioPortal is running at http://localhost/cbioportal - if this is not the case, follow the advanced example):

```
./metaImport.py -s ../../../test/scripts/test_data/study_es_0/
```

#### Advanced Example
This example imports the study to the localhost, creates an html report and shows status messages.
```
./metaImport.py -s ../../../test/scripts/test_data/study_es_0/ -u http://localhost:8080 -html myReport.html -v
```

By adding `-o`, warnings will be overridden and import will start after validation.

## Development / debugging mode
For developers and specific testing purposes, an extra script, cbioportalImporter.py, is available which imports data regardless of validation results. Check [this](Data-Loading-For-Developers.md) page for more information on how to use it.
# API and API Clients

cBioPortal provides a REST API for programmatic access to the data. The visualizations one can see on the website leverage the same API. By connecting to the API directly, anyone can build their own visalizations/reports.

Please see the full reference documentation for the API [here](https://www.cbioportal.org/api/swagger-ui/index.html).

## API Clients

The cBioPortal REST API is described using Swagger/OpenAPI, which allows one to generate a client in most programming languages. One can use the command line tool `curl` for dowloading data on the command line or use another language such as `Python` or `R` to make visualizations. We list some common examples below, but if your language is not listed, there is likely a client generator available elsewhere (see e.g. https://swagger.io/tools/swagger-codegen/). Do reach out if you'd like us to add a language.

### R clients

There are multiple ways to access the API using R. Below are two recommended R packages to access cBioPortal data.

#### cBioPortalData (recommended)

cBioPortalData aims to import all cBioPortal datasets as MultiAssayExperiment objects in Bioconductor. Some of its key features:

1. The MultiAssayExperiment class explicitly links all assays to the patient clinical/pathological data
2. The MultiAssayExperiment class provides a flexible API including harmonized subsetting and reshaping to convenient wide and long formats.
3. It provides complete datasets, not just for subsets of genes
4. It provides automatic local caching, thanks to BiocFileCache.

For a comprehensive user guide to `cBioportalData` see: https://waldronlab.io/cBioPortalData/articles/cBioPortalData.html

See also the workshop materials from our webinar which include an intro to `cBioPortalData`: https://github.com/cBioPortal/2020-cbioportal-r-workshop.

Note that one can point to private authenticated instances like this:

```
cBioPortal(
    hostname = "genie.cbioportal.org",
    token = "~/Downloads/cbioportal_data_access_token.txt"
)
```

#### cbioportalR (recommended)

cbioportalR offers easy-to-use functions that allow users to browse and pull data from public or institutional cBioPortal sites without knowledge of web service or Bioconductor infrastructures. The package is tidyverse-compatible. Key package features include:

1. Comprehensive documentation aimed at helping clinical researchers understand the underlying structure of cBioPortal data
2. [Tutorials]([https://www.karissawhiting.com/cbioportalR/articles/overview-of-workflow.html]) for quick API authentication and set up
3. Functions to pull complete clinical and genomic data by study ID, molecular profile ID, sample list IDs or individual sample ID (e.g. `get_genetics_by_study()`, `get_genetics_by_sample()`)
4. Functions to navigate and identify patient IDs, sample IDs or study IDs as needed, or infer necessary ID information for queries when not supplied by user.
5. Helper functions to pull information on gene panels (`get_gene_panel()`), or lookup entrez ID (`get_entrez_id()`), Hugo Symbol (`get_hugo_symbol()`) or common gene aliases (`get_alias()`) of genes
6. Capability to query multiple sample IDs from different studies concurrently

For a detailed tutorial on `cbioportalR`, see the package website: https://www.karissawhiting.com/cbioportalR/articles/overview-of-workflow.html

#### rapiclient

Although we recommend [cBioPortalData](/#cbioportaldata-recommended) or [cbioportalR](/#cbioportalR-recommended) for most use cases, it is possible to connect to the API directly using [rapiclient](https://github.com/bergant/rapiclient):

```
library(rapiclient)
client <- get_api(url = "https://www.cbioportal.org/api/v2/api-docs")
```

#### CGDSR (will be deprecated)

The CGDS-R package connects an older version of our web API (`webservice.do`). Althought we will continue to keep `webservice.do` running for a while, we can't guarantee the same level of quality as our new API (`cbioportal.org/api`) provides. Therefore we recommend that you use `cBioPortalData` instead.

### Python client

There are multiple ways to access the API using Python. One can use the `bravado` package to access the API directly, or use the `cbio_py` client, which provides a simple wrapper for the API and returns data in a format that is easy to work with.

#### bravado

Generate a client in Python using [bravado](https://github.com/Yelp/bravado) like this:

```python
from bravado.client import SwaggerClient
cbioportal = SwaggerClient.from_url('https://www.cbioportal.org/api/v2/api-docs',
                                    config={"validate_requests":False,"validate_responses":False,"validate_swagger_spec": False})
```

This allows you to access all API endpoints:

```python
>>> dir(cbioportal)
['Cancer Types',
 'Clinical Attributes',
 'Clinical Data',
 'Clinical Events',
 'Copy Number Segments',
 'Discrete Copy Number Alterations',
 'Gene Panels',
 'Genes',
 'Molecular Data',
 'Molecular Profiles',
 'Mutations',
 'Patients',
 'Sample Lists',
 'Samples',
 'Studies']
```

For easy tab completion you can add lower cases and underscores:

```python
for a in dir(cbioportal):
    cbioportal.__setattr__(a.replace(' ', '_').lower(), cbioportal.__getattr__(a))
```

This example gets you all mutation data for the MSK-IMPACT 2017 study:

```python
muts = cbioportal.mutations.getMutationsInMolecularProfileBySampleListIdUsingGET(
    molecularProfileId="msk_impact_2017_mutations", # {study_id}_mutations gives default mutations profile for study 
    sampleListId="msk_impact_2017_all", # {study_id}_all includes all samples
    projection="DETAILED" # include gene info
).result()
```

For a portal that requires authentication one can use (see [Data Access Using Tokens](/deployment/authorization-and-authentication/Authenticating-Users-via-Tokens.md)):

```
from bravado.client import SwaggerClient
from bravado.requests_client import RequestsClient

http_client = RequestsClient()
http_client.set_api_key(
    'genie.cbioportal.org', 'Bearer <TOKEN>',
    param_name='Authorization', param_in='header'
)

cbioportal = SwaggerClient.from_url('https://genie.cbioportal.org/api/v2/api-docs',
                                    http_client=http_client,
                                    config={"validate_requests":False,
                                            "validate_responses":False,
                                            "validate_swagger_spec": False}
)
```

A Jupyter notebook with more examples can be found [here](https://github.com/mskcc/cbsp-hackathon/blob/master/0-introduction/cbsp\_hackathon.ipynb).

#### cbio\_py

See the `cbio_py` documentation: https://pypi.org/project/cbio-py/.
## Introduction
For some data types, when uploading to cBioPortal, it is currently necessary to also provide a z-score transformed version of your input file. The z-score data is essential for the oncoprint functionality. The oncoprint shows high or low mRNA expression of the genes, based on the threshold the user sets when selecting the genomic profile. 

:warning: 
Please keep in mind that the z-scores are calculated using only patient data. Hence, 'mRNA High' in this case implies higher expression than the average patient. Also, the source data on which the z-score data is based does not necessarily follow the normal distribution. If your data does not follow the normal distribution, the z-score threshold is less reliable and will result in more false positives or false negatives. You can consider log transforming your value before calculating z-scores to improve this. However, even the logged values may not follow the normal distribution, especially if the data is bimodal.

## The cBioPortal Z-Score calculation method

cBioPortal currently generates two z-score profiles using two different base populations:
- Distribution based on **diploid** samples only: The expression distribution for unaltered copies of the gene is estimated by calculating the mean and variance of the expression values for samples in which the gene is diploid (i.e. value is "0" as reported by [discrete CNA data](File-Formats.md#discrete-copy-number-data)). We call this the unaltered distribution. If the gene has no diploid samples, then its normalized expression is reported as NA.

- Distribution based on **all** samples: The expression distribution of the gene is estimated by calculating the mean and variance of all samples with expression values. If the gene has samples whose expression values are all zeros or non-numeric, then its normalized expression is reported as NA.

Otherwise for every sample, the gene's normalized expression for both the profiles is reported as

```
(r - mu)/sigma
```
where `r` is the raw expression value, and `mu` and `sigma` are the mean and standard deviation of the base population, respectively.

### How to proceed
cBioPortal expects z-score normalization to take place per gene. You can calculate z-scores with your own preferred method, or use one of the cBioPortal provided approaches:
- [convertExpressionZscores.pl](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/convertExpressionZscores.pl) applies Method 1 (diploid samples as base population)
- [NormalizeExpressionLevels_allsampleref.py](https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/zscores/zscores_relative_allsamples) applies Method 2 (all samples as base population)

Examples of the calculation and running the programs are below.

## convertExpressionZscores method
Given expression and Copy Number Variation data for a set of samples (patients), generate normalized expression values.

### Parameters

`<copy_number_file> <expression_file> <output_file> <normal_sample_suffix> <[min_number_of_diploids]>`

- `<copy_number_file>` : the [discrete copy number (CNA) file](File-Formats.md#discrete-copy-number-data) 
- `<expression_file>` : the [expression (exp) data file](File-Formats.md#expression-data). 
- `<output_file>` : the output file to be generated
- `<normal_sample_suffix>` : use this to identify which of your samples are "normal" samples (if any). E.g. normal TCGA samples have a suffix "-11". Set it to some dummy value, e.g. "NONE", if you have no normal samples in your data.

### Algorithm
Input: [discrete copy number (CNA)](File-Formats.md#discrete-copy-number-data) and [expression (exp)](File-Formats.md#expression-data) files

```
    for each gene{
       identify diploid cases in CNA
       obtain mean and sd of exp of diploid cases
       for each case{
          zScore <- (value - mean)/sd
       }
    }
```
### Example Calculation
<table>
<tr><td>Hugo_Symbol</td><td>Entrez_Gene_Id</td><td>A1-A0SD-01</td><td>A1-A0SE-01</td><td>A1-A0SH-01</td><td>A1-A0SJ-01</td><td>A1-A0SK-01</td><td>A1-A0SM-01</td><td>A1-A0SO-01</td><td>A1-A0SP-01</td><td>A2-A04N-01</td><td>A2-A04P-01</td><td>A2-A04Q-01</td><td>A2-A04R-01</td><td>A2-A04T-01</td><td>A2-A04U-01</td><td>A2-A04V-01</td><td>A2-A04W-01</td><td>A2-A04X-01</td><td>A2-A04Y-01</td></tr>
<tr><td>RPS11 Expr</td><td>6205</td><td>0.765</td><td>0.716</td><td>0.417125</td><td>0.115</td><td>0.492875</td><td>-0.525</td><td>-0.169</td><td>0.396</td><td>0.50475</td><td>0.400875</td><td>0.393125</td><td>0.9165</td><td>0.627125</td><td>0.337125</td><td>0.705</td><td>0.16425</td><td>0.325</td><td>0.11175</td></tr>
<tr><td>RPS11 CNA</td><td>6205</td><td>0</td><td>0</td><td>0</td><td>1</td><td>1</td><td>0</td><td>-1</td><td>0</td><td>0</td><td>2</td><td>0</td><td>0</td><td>1</td><td>-1</td><td>0</td><td>0</td><td>-1</td><td>0</td></tr>
</table>

Calculate mean and stdev where CNA is 0 (=diploid):

<table>
<tr><td>Diploid avg</td><td>Diploid std</td></tr>
<tr><td>---</td><td>---</td></tr>
<tr><td>0.414954545454545</td><td>0.399504498851105</td></tr>
</table>

Calculate the z-scores:
<table>
<tr><td>Hugo_Symbol</td><td>Entrez_Gene_Id</td><td>A1-A0SD-01</td><td>A1-A0SE-01</td><td>A1-A0SH-01</td><td>A1-A0SJ-01</td><td>A1-A0SK-01</td><td>A1-A0SM-01</td><td>A1-A0SO-01</td><td>A1-A0SP-01</td><td>A2-A04N-01</td><td>A2-A04P-01</td><td>A2-A04Q-01</td><td>A2-A04R-01</td><td>A2-A04T-01</td><td>A2-A04U-01</td><td>A2-A04V-01</td><td>A2-A04W-01</td><td>A2-A04X-01</td><td>A2-A04Y-01</td></tr>
<tr><td>RPS11 Expr Output</td><td>6205</td><td>0.8762</td><td>0.7535</td><td>0.0054</td><td>-0.7508</td><td>0.1950</td><td>-2.3528</td><td>-1.4617</td><td>-0.0474</td><td>0.2248</td><td>-0.0352</td><td>-0.0546</td><td>1.2554</td><td>0.5311</td><td>-0.1948</td><td>0.7260</td><td>-0.6275</td><td>-0.2252</td><td>-0.7590</td></tr>
</table>

Note: this implies that your full dataset does not have average=0, std=1

### Running the script


To run the script type the following commands when in the folder `<cbioportal_source_folder>/core/src/main/scripts`:

```
export PORTAL_HOME=<cbioportal_configuration_folder>
```

and then 

```
./convertExpressionZscores.pl <copy_number_file> <expression_file> <output_file> <normal_sample_suffix> <[min_number_of_diploids]>
```

#### Example:

```
./convertExpressionZscores.pl  ../../test/scripts/test_data/study_es_0/data_CNA.txt ../../test/scripts/test_data/study_es_0/data_expression_median.txt data_expression_ZSCORES.txt NONE
```
## NormalizeExpressionLevels_allsampleref method
Given the expression data for a set of samples, generate normalized expression values with the reference population of all samples independent of sample diploid status.

### Parameters
`<expression_file> <output_file> <[log_transform]> <[exclude_zero_negative_values]>`

- `<expression_file>` : the [expression (exp) data file](File-Formats.md#expression-data).
- `<output_file>` : the output file to be generated.
- `<log_transform>` : Use this to log transform the data before calculating z-scores (optional).
- `<exclude_zero_negative_values>` : Use this to exclude zero's or negative counts from the reference population when normalizing the data (optional). 

### Algorithm
Input [expression data file](File-Formats.md#expression-data)
```
for each gene{
  log-transform the raw data, if -l is passed
  identify the base population (if -e is passed, n = # of samples where expression value is not zero or negative)
  compute mean and standard deviation for samples in the base poplulation 
  for each sample in the input dataset{
    Z-Score <- (value - mean)/sd when standard deviation != 0
    Z-Score <- NA when standard deviation = 0
  }
}
```
### Log-transforming the data
Using the `-l` option above calculates log base 2 of the expression values.

Here's how we handle the Negative values when log transforming:
```
Replace the negative values to 0 and add a pseudo-count (+1) to all observed counts prior to applying log transform.
example, if raw value is -1, the log transform would be log(0+1)
         if the value is 0, the log transform would be log(0+1)
         if the value is 1, the log transform would be log(1+1)
```

### Example Calculation:

Log transform and calculate the z-scores (without -e option):
<table>
<tr><td>Hugo_Symbol</td><td>Entrez_Gene_Id</td><td>A1-A0SD-01</td><td>A1-A0SE-01</td><td>A1-A0SH-01</td><td>A1-A0SJ-01</td><td>A1-A0SK-01</td><td>A1-A0SM-01</td><td>A1-A0SO-01</td><td>A1-A0SP-01</td><td>A2-A04N-01</td><td>A2-A04P-01</td><td>A2-A04Q-01</td><td>A2-A04R-01</td><td>A2-A04T-01</td><td>A2-A04U-01</td><td>A2-A04V-01</td><td>A2-A04W-01</td><td>A2-A04X-01</td><td>A2-A04Y-01</td></tr>
<tr><td>RPS11 Expr</td><td>6205</td><td>0.765</td><td>0.716</td><td>0.417125</td><td>0.115</td><td>0.492875</td><td>-0.525</td><td>-0.169</td><td>0.396</td><td>0.50475</td><td>0.400875</td><td>0.393125</td><td>0.9165</td><td>0.627125</td><td>0.337125</td><td>0.705</td><td>0.16425</td><td>0.325</td><td>0.11175</td></tr>
<tr><td>Output</td><td>6205</td><td>1.2879</td><td>1.1378</td><td>0.1177</td><td>-1.1605</td><td>0.3953</td><td>-1.7408</td><td>-1.7408</td><td>0.0376</td><td>0.4375</td><td>0.0562</td><td>0.0266</td><td>1.7268</td><td>0.8543</td><td>-0.1921</td><td>1.1035</td><td>-0.9301</td><td>-0.2407</td><td>-1.1761</td></tr>
  </table>

### Running the script
To run the script clone the datahub-study-curation-tools from [here](https://github.com/cBioPortal/datahub-study-curation-tools) and type the commands when in the folder `zscores/zscores_relative_allsamples`:

```
python NormalizeExpressionLevels_allsampleref.py -i <expression_file> -o <output_file> [-l] [-e]
```

#### Example:

```
python NormalizeExpressionLevels_allsampleref.py -i path/to/your/study/data_expression_median.txt -o path/to/your/study/data_mRNA_median_Zscores.txt -l
```
cBioPortal, 2015# cBioPortal

The cBioPortal for Cancer Genomics provides visualization, analysis, and download of large-scale cancer genomics data sets. For a short intro on cBioPortal, see [these introductory slides](https://docs.google.com/presentation/d/1hm0G77UklZnpQfFvywBfW2ZIsy8deKi5r1RfJarOPLg/edit?usp=sharing).

If you would like to know how to setup a private instance of the portal and/or get set up for developing, see the [documentation](https://docs.cbioportal.org). For details on contributing code changes via pull requests, see our [Contributing document](CONTRIBUTING.md).

If you are interested in coordinating the development of new features, please contact cbioportal@cbioportal.org or reach out on https://slack.cbioportal.org.

## üìò Documentation
See [https://docs.cbioportal.org](https://docs.cbioportal.org)

## ü§ù License
See [LICENSE](./LICENSE)

## üíª Run Backend
cBioPortal consists of several components, please read the [Architecture docs](https://docs.cbioportal.org/architecture-overview/) to figure out what repo would be relevant to edit. If you e.g. only want to make frontend changes, one can directly edit [the frontend repo](https://github.com/cbioportal/cbioportal-frontend) instead. Read the instructions in that repo for more info on how to do frontend development. This repo only contains the backend part. Before editing the backend, it's good to read the [backend code organization](docs/Backend-Code-Organization.md).

### Local Development
#### What MySQL database to use
We recommend to set up a MySQL database using [Docker Compose](https://github.com/cBioPortal/cbioportal-docker-compose). It's useful to know how to do this as it allows you to import any dataset of your choice. For debugging production issues, we also have a database available with all the data on https://cbioportal.org that one can connect to directly. Please reach out on slack to get the credentials.

#### Command Line
If you want to run the cBioPortal web app from the command line please follow these instructions. First, we want to make sure that all ports are open for the services set up through [docker compose](https://github.com/cBioPortal/cbioportal-docker-compose) (i.e. not just accessible to other containers within the same Docker Compose file). To do so, in the [docker compose repo](https://github.com/cBioPortal/cbioportal-docker-compose) run:

```
docker compose -f docker-compose.yml -f open-ports.yml up
```
This should open the ports. Now we are ready to run the cBioPortal web app locally. You can compile the backend code with:

```

java -Xms2g -Xmx4g \
     -Dauthenticate=noauthsessionservice \
     -Dsession.service.url=http://localhost:5000/api/sessions/my_portal/ \
     -Dsession.service.origin='*' \
     -Dspring.datasource.username=cbio_user \
     -Dspring.datasource.password=somepassword \
     -Dspring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver \
     -Dspring.jpa.database-platform=org.hibernate.dialect.MySQL5InnoDBDialect \
     -Dspring.datasource.url='jdbc:mysql://cbio_user:somepassword@localhost:3306/cbioportal?useSSL=false&allowPublicKeyRetrieval=true' \
     -Dshow.civic=true \
     -Dskin.footer='' \
     -Dapp.name='my-portal' \
     -Ddbconnector=dbcp \
     -cp "$PWD:$PWD/BOOT-INF/lib/*" \
     org.cbioportal.PortalApplication
```

The app should now show up at http://localhost:8080.

#### Deploy your development image inside Docker Compose
Another option is to deploy your development image directly into the [docker-compose](https://github.com/cBioPortal/cbioportal-docker-compose/blob/5da068f0eb9b4f42db52ab5e91321b26a1826d7a/docker-compose.yml#L6) file. First build the image like this

```
docker build -t cbioportal/cbioportal:my-dev-cbioportal-image -f docker/web-and-data/Dockerfile .
```

Then change the [env file](https://github.com/cBioPortal/cbioportal-docker-compose/blob/master/.env) to use `cbioportal/cbioportal:my-dev-cbioportal-image`.

### Local Development


Note: internally we have a dev database available with the public data set that one can connect to directly. Please reach out on slack to get the credentials. It is usually best to use a small test dataset, but if a copy of the production database is necessary for e.g. fixing a bug specific to production data that can be useful.

### üïµÔ∏è‚Äç‚ôÄÔ∏è Debugging

If you want to attach a debugger you can change the `docker-compose.yml` file to include the parameters: `-Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=5005` (make sure to expose the debug port by adding `5005:5005` in the ports section of the cbioportal container). If you are running the java app outside of docker you can add the same parameters to the java command line arguments instead.

You can then use a JAVA IDE to connect to that port. E.g. in [VSCode](https://code.visualstudio.com/), one would add the following configuration to `launch.json` to connect:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "java",
            "name": "Debug (Attach)",
            "request": "attach",
            "hostName": "localhost",
            "port": 5005,
            "projectName": "cbioportal"
        }
    ]
}
```

## üå≥ Branch Information

| | main branch | upcoming release branch | later release candidate branch |
| --- | --- | --- | --- |
| Branch name | [`master`](https://github.com/cBioPortal/cbioportal/tree/master) |  -- |  [`rc`](https://github.com/cBioPortal/cbioportal/tree/rc) |
| Description | All bug fixes and features not requiring database migrations go here. This code is either already in production or will be released this week | Next release that requires database migrations. Thorough manual product review often takes place for this branch before release | Later releases with features that require database migrations. This is useful to allow merging in new features without affecting the upcoming release. Could be seen as a development branch, but note that only high quality pull requests are merged. That is the feature should be pretty much ready for release after merge. |
| Live instance | https://www.cbioportal.org / https://master.cbioportal.org | -- | https://rc.cbioportal.org |
| Live instance version | https://www.cbioportal.org/api/info / https://master.cbioportal.org/api/info | -- | https://rc.cbioportal.org/api/info |
| Docker Image | cbioportal/cbioportal:master | --| cbioportal/cbioportal:rc |
| Kubernetes Config | [production](https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/cbioportal/cbioportal_spring_boot.yaml) / [master](https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/cbioportal/cbioportal_backend_master.yaml) | -- | [rc](https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/cbioportal/cbioportal_backend_rc.yaml) |
| Status | [![master build status](https://github.com/cbioportal/cbioportal/workflows/Core%20tests/badge.svg)](https://github.com/cBioPortal/cbioportal/actions/workflows/core-test.yml?query=branch%3Amaster) [![master build status](https://github.com/cbioportal/cbioportal/workflows/Integration%20tests/badge.svg)](https://github.com/cBioPortal/cbioportal/actions/workflows/integration-test.yml?query=branch%3Amaster) [![master build status](https://github.com/cbioportal/cbioportal/workflows/Docker%20Image%20CI/badge.svg)](https://github.com/cBioPortal/cbioportal/actions/workflows/dockerimage.yml?query=branch%3Amaster) [![master build status](https://github.com/cbioportal/cbioportal/workflows/Python%20validator/badge.svg)](https://github.com/cBioPortal/cbioportal/actions/workflows/validate-data.yml?query=branch%3Amaster) [![CircleCI](https://circleci.com/gh/cBioPortal/cbioportal/tree/master.svg?style=svg)](https://app.circleci.com/pipelines/github/cBioPortal/cbioportal?branch=master&filter=all) | -- | -- |

## üöÄ Releases
Release Notes on GitHub:

https://github.com/cBioPortal/cbioportal/releases

See also the cBioPortal News section for user focused release information:

https://www.cbioportal.org/news

Docker Images are available for each tag and branch:

https://hub.docker.com/repository/docker/cbioportal/cbioportal/tags

## üëâ Other Repos
Read the [Architecture docs](https://docs.cbioportal.org/2.1-deployment/architecture-overview) to see how these relate:

- https://github.com/cBioPortal/cbioportal-frontend
- https://github.com/cbioportal/session-service
- https://github.com/cBioPortal/datahub/
# Security Policy

## Supported Versions

We release security releases for the following versions:

| Version | Supported          |
| ------- | ------------------ |
| 6.x   | :white_check_mark: |
| <= 5.x   | :x:  |

## Reporting a Vulnerability
If there are any vulnerabilities, don't hesitate to report them.

Use the private contact address cbioportal@cbioportal.org to report the fix. 

Describe the vulnerability.

If you have a fix, that is most welcome -- please attach or summarize it in your message!

We will evaluate the vulnerability and, if necessary, release a fix or mitigating steps to address it. We will contact you to let you know the outcome, and will credit you in the report.

Please do not disclose the vulnerability publicly until a fix is released!

Once we have either a) published a fix, or b) declined to address the vulnerability for whatever reason, you are free to publicly disclose it.
# cBioPortal Community Code of Conduct

cBioPortal follows the [CNCF Code of Conduct](https://github.com/cncf/foundation/blob/master/code-of-conduct.md).

Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting <conduct@cbioportal.org>.
# How to contribute

Thank you for your interest in contributing to cBioPortal! This document provides a brief set of guidelines for contributing.

# Who are you?

We are curious to learn more about you! We would love to help you get started! The contributors in our community all have different backgrounds. For instance some people have:

1. Engineering experience but no to little knowledge of cancer genomics
2. Knowledge about cancer genomics but no to little engineering experience
3. No engineering nor cancer genomics experience but an eagerness to contribute

if you feel like you don't fall into any of these categories, please reach out so you can help us update the above list üôÇ! Note that there are many contributions that can be made to an open source commmunity without coding a single line of code. You can reach us through our [public slack channel](https://slack.cbioportal.org).

# Join the Slack!

Come and chat with us at https://slack.cbioportal.org üëã

# Making a code contribution

The cBioPortal currently uses a "fork and pull" model for collaborative software development.

From the [GitHub Help Page of Using Pull Requests](https://help.github.com/articles/using-pull-requests/):

"The fork & pull model lets anyone fork an existing repository and push changes to their personal fork without requiring access be granted to the source repository. The changes must then be pulled into the source repository by the project maintainer. This model reduces the amount of friction for new contributors and is popular with open source projects because it allows people to work independently without upfront coordination."

## Getting Started

### Programming
 * Make sure you have a [GitHub account](https://github.com/signup/free).
 * Familiarize yourself with the [project documentation](https://docs.cbioportal.org), including the [Feature Development Guide](https://docs.cbioportal.org/4.-development/feature-development-guide), the [Architecture docs](https://docs.cbioportal.org/2.1-deployment/architecture-overview), the [backend code organization](docs/Backend-Code-Organization.md) and [backend development guidelines](docs/Backend-Development-Guidelines.md).
 * Find a [good first issue](https://github.com/cBioPortal/cbioportal/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22good+first+issue%22) to start with
 * Check if the issue will require frontend or backend changes. If it is for the frontend look at how to set up the [frontend repo](https://github.com/cbioPortal/cbioportal-frontend/) instead
 * Fork the cbioportal or cbioportal-frontend project on GitHub depending on what your working on.  For general instructions on forking a GitHub project, see [Forking a Repo](https://help.github.com/articles/fork-a-repo/) and [Syncing a fork](https://help.github.com/articles/syncing-a-fork/).
 * Reach out on slack or our [Google user group](https://groups.google.com/g/cbioportal) if you run into any issues

### Documentation
If you'd like to improve our documentation, have a look at https://docs.cbioportal.org. At the bottom of each page is an "edit page" button to make changes. You can use the GitHub UI to edit the pages and submit them.

## Contributing Code Changes via a Pull Request

Once you have forked the repo, you need to create your code contributions within a new branch of your forked repo.  For general background on creating and managing branches within GitHub, see:  [Git Branching and Merging](https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging).

* To begin, create a topic branch from where you want to base your work.
 * For any change that requires database migrations, this will be the **rc branch**. For all other changes, this will be the **master branch**. For additional details, see [Branches within cBioPortal](#branches-within-cbioportal) below.

You usually create a branch like so:

```
git checkout master
git checkout -b [name_of_your_new_branch]
```

You then usually commit code changes, and push your branch back to GitHub like so:

```git push origin [name_of_your_new_branch]```

A few tips:

* Make commits in logical/cohesive units.
* Make sure you have added the necessary tests for your changes.
* Run _all_ tests to assure nothing else was accidentally broken in the java part (data loading and front-end parts are tested by other scripts in github actions). This is done by running:  ```mvn integration-test```.

When you are ready to submit your pull-request:

* Push your branch to your GitHub project.
* Open the pull request to the branch you've based your work on

For more details on submitting a pull-request, please see:  [GitHub Guide to Collaborating with issues and pull requests](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests).

## Branches within cBioPortal

To figure out where your first pull request might go, it helps to have an understanding of cBioPortal's branching model. The cBioPortal currently maintains three branches in both the [frontend](https://github.com/cbioportal/cbioportal-frontend) and [backend repo](https://github.com/cbioportal/cbioportal):

 * **master**:  this reflects what will be released with our next weekly release (https://github.com/cBioPortal/cbioportal/releases). For the [frontend repo](https://github.com/cbioportal/cbioportal-frontend) this branch is automatically deployed to production. On the backend it is deployed at least once a week. New features, bug fixes and documentation updates can go here. Only if the feature requires a database migration it should go to **rc**.
  * **rc**:  release candidate branch, this branch contains new changes that require a database migration. It is deployed to https://rc.cbioportal.org. Once it's ready for more thorough product review a new branch is created with the name **release-x.y.z**. This way people can still continue to submit new changes to **rc**, while more thorough testing takes place of the **release-x.y.z** branch.
  *  **release-x.y.z**: this branch contains changes that require a database migration. It will be merged to master after thorough product review on https://beta.cbioportal.org.
  
We try to continuously merge new changes from `master` to `release-x.y.z`, and subsequently from the `release-x.y.z` branch to `rc` such that everybody is working on the latest code. Keep in mind though that occasionally there are conflicts that need to be resolved before we can merge. If you're working on e.g. the rc branch, you can check whether all the changes in master are in rc like this: https://github.com/cBioPortal/cbioportal/compare/rc...master. If a particular change you are waiting for is not there, one can help creating a pull request that merges these changes in. Try e.g. (if origin points to the cbioportal repo): `git fetch origin && git checkout origin/rc -b merge-master-to-rc && git merge origin/master`. Then resolve conflicts and push the branch `merge-master-to-rc` to your repo and open the PR.

### Getting your changes reviewed

Once you've submitted your pull request, you want
other members of the development community to review
whether integrating your change will cause problems
for any users or the maintainability of the software.

If you have an idea who might be able to spot such issues
in the parts of the code and functionality affected by your changes,
notify them by requesting a review using the **Reviewers** menu
to the right of the summary you just wrote
and/or `@`-mentioning them in a comment. Or reaching out them on [slack](https://slack.cbioportal.org).

Reviewers may request you to rephrase or adjust things
before they allow the changes to be integrated.
If they do, commit the amendments as new, separate changes,
to allow the reviewers to see what changed since they last read your code.
Do not overwrite previously-reviewed commits with
ones that include additional changes (by `--amend`ing or squashing)
until the reviewers approve.
Reviewers may request you to squash such amendment commits afterwards,
or offer to push rewritten versions of your commits themselves.

## Pull Request Reviewers Guide
If someone requests your review on a pull request,
read the title and description and assign any other collaborators
who would want to know about the proposed change.

Decide whether you think that your input is needed,
and that the PR should wait for your further review before being merged.
If not, un-assign yourself as a reviewer and leave a comment.

Here we describe the guidelines for the reviewer. Always follow the checks in
general, then follow the other checks that apply:

### General
- Double check all the things in the **Checks** section of the Pull Request.
  Remind the submitter if any of them are not fulfilled
- Are the test cases spanning a decent amount of scenarios? It is the
  submitters as well as the reviewers responsibility to not let any errors
  sneak into the portal.

Bug fixes:

- Should the bug that causes the issue be added as a test case?

New features:

- If this is a new feature make sure the proposed changes are in line with the
  current planning of cBioPortal e.g. is the right API used, is this in line
  with current refactoring efforts.

### Backend
New features:

- Is the new persistence stack used?

### Frontend
New features:

- What APIs are used to get the data? Is the REST API used?
- Should this be a separate library in a separate repo or should it be part of cBioPortal?
- Are dependencies properly listed? Ideally in a package.json
- How is the package included in cBioPortal?

### Devops
New features:

- Does the configuration style follow the config guidelines? That is compile.
- Runtime (Spring) goes in `application.properties`. Default values should be in `GlobalProperties.java`.
- Non-stable configuration should be done through war overlays.
- Is the configuration tested as part of the CI tests? It's not a necessity but be
  aware that untested configuration will be tough to maintain.
- Is there documentation on the proposed changes?

## Additional Resources

* [cBioPortal Issue Tracker](https://github.com/cBioPortal/cbioportal/issues)
* [General GitHub documentation](http://help.github.com/)
* [GitHub Pull Request documentation](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests)
# Tools for development

In this folder is some additional configuration that can be useful for local development. None of this should be deployed directly to production

# Set up keycloak for cBioPortal >v6

Requirements:
- System runs docker (including docker compose)

  1. Run from the root of the repository:

```
cd dev
docker compose up -d 
```

2. (Option 1) Apply SAML2 configuration to _security.properties_ in cBioPortal:

```properties
authenticate=saml
spring.security.saml2.relyingparty.registration.keycloak.assertingparty.metadata-uri=http://localhost:8084/realms/cbio/protocol/saml/descriptor
spring.security.saml2.relyingparty.registration.keycloak.assertingparty.entity-id=http://localhost:8084/realms/cbio
spring.security.saml2.relyingparty.registration.keycloak.entity-id=cbioportal
spring.security.saml2.relyingparty.registration.keycloak.signing.credentials[0].certificate-location=classpath:/dev/security/signing-cert.pem
spring.security.saml2.relyingparty.registration.keycloak.signing.credentials[0].private-key-location=classpath:/dev/security/signing-key.pem
```

3. (Option 2) Apply OIDC configuration to _security.properties_ in cBioPortal:

```properties
authenticate=oauth2
spring.security.oauth2.client.registration.keycloak.redirect-uri=http://localhost:8080/login/oauth2/code/keycloak
spring.security.oauth2.client.registration.keycloak.client-name=cbioportal_oauth2
spring.security.oauth2.client.registration.keycloak.client-id=cbioportal_oauth2
spring.security.oauth2.client.registration.keycloak.client-secret=client_secret
spring.security.oauth2.client.registration.keycloak.authorization-grant-type=authorization_code
spring.security.oauth2.client.registration.keycloak.client-authentication-method=client_secret_post
spring.security.oauth2.client.registration.keycloak.scope=openid,email,roles
spring.security.oauth2.client.provider.keycloak.issuer-uri=http://localhost:8084/realms/cbio
spring.security.oauth2.client.provider.keycloak.user-name-attribute=email
```

4. Set the following in _application.properties_:

```properties
persistence.cache_type=no-cache
session.service.url=http://localhost:5000/api/sessions/my_portal/

spring.datasource.url=jdbc:mysql://localhost:3306/cbioportal?useSSL=false&allowPublicKeyRetrieval=true
spring.datasource.username=cbio_user
spring.datasource.password=somepassword
spring.jpa.database-platform=org.hibernate.dialect.MySQL5InnoDBDialect
spring.datasource.driver-class-name=com.mysql.jdbc.Driver
```

4. Start cBioPortal application on port 8080. The login credentials are `testuser:P@assword1`.

‚ö†Ô∏è Warning: Do not use this directly for production use as it takes several shortcuts to get a quick keycloak instance up.
See [../docs/docker/README.md](../docs/docker/README.md)
# File Not Found

<script>
window.redirectMap = {
    "5.1-data-loading/data-loading/file-formats":"file-formats",
    "2.1-deployment/architecture-overview":"architecture-overview",
    "6.-web-api-and-clients/api-and-api-clients":"web-api-and-api-clients",
    "2.2-authorization-and-authentication/authenticating-users-via-tokens":"deployment/authorization-and-authentication/authenticating-users-via-tokens",
    "2.4-integration-with-other-webservices/oncokb-data-access":"deployment/integration-with-other-webservices/oncokb-data-access",
    "4.-development/release-procedure#a-note-on-versioning":"release-procedure",
    "2.1.2-deploy-without-docker/deploying":"deployment/deploy-without-docker/deploying",
    "2.2-authorization-and-authentication/authenticating-users-via-tokens":"deployment/authorization-and-authentication/authenticating-users-via-tokens",
    "1.-general/news":"news",
    "1.-general/faq":"user-guide/faq",
    "1.-general/about-us":"about-us",
    "1.-general/rfc-list":"rfc-list",
    "2.1-deployment/architecture-overview":"architecture-overview",
    "2.1-deployment/hardware-requirements":"hardware-requirements",
    "2.1.1-deploy-with-docker-recommended/docker":"deployment/docker",
    "2.1.1-deploy-with-docker-recommended/import_data":"deployment/docker/import_data",
    "2.1.1-deploy-with-docker-recommended/example_commands":"deployment/docker/example_commands",
    "2.1.1-deploy-with-docker-recommended/using-keycloak":"deployment/docker/using-keycloak",
    "2.1.2-deploy-without-docker/software-requirements":"deployment/deploy-without-docker/software-requirements",
    "2.1.2-deploy-without-docker/pre-build-steps":"deployment/deploy-without-docker/pre-build-steps",
    "2.1.2-deploy-without-docker/build-from-source":"deployment/deploy-without-docker/build-from-source",
    "2.1.2-deploy-without-docker/import-the-seed-database":"deployment/deploy-without-docker/import-the-seed-database",
    "2.1.2-deploy-without-docker/deploying":"deployment/deploy-without-docker/deploying",
    "2.1.2-deploy-without-docker/load-sample-cancer-study":"deployment/deploy-without-docker/load-sample-cancer-study",
    "2.2-authorization-and-authentication/user-authorization":"deployment/authorization-and-authentication/user-authorization",
    "2.2-authorization-and-authentication/authenticating-users-via-saml":"deployment/authorization-and-authentication/authenticating-users-via-saml",
    "2.2-authorization-and-authentication/authenticating-users-via-ldap":"deployment/authorization-and-authentication/authenticating-users-via-ldap",
    "2.2-authorization-and-authentication/authenticating-and-authorizing-users-via-keycloak":"deployment/authorization-and-authentication/authenticating-and-authorizing-users-via-keycloak",
    "2.2-authorization-and-authentication/authenticating-users-via-tokens":"deployment/authorization-and-authentication/authenticating-users-via-tokens",
    "2.3-customization/customizing-your-instance-of-cbioportal":"deployment/customization/customizing-your-instance-of-cbioportal",
    "2.3-customization/portal.properties-reference":"deployment/customization/application.properties-reference/",
    "deployment/customization/portal.properties-reference":"deployment/customization/application.properties-reference/",
    "2.3-customization/caching":"deployment/customization/caching",
    "2.3-customization/layout":"deployment/customization/studyview/#how-does-the-study-view-organize-the-charts",
    "2.3-customization/priority":"deployment/customization/studyview/#study-view-customization-with-priority-data",
    "2.4-integration-with-other-webservices/oncokb-data-access":"deployment/integration-with-other-webservices/oncokb-data-access",
    "3.-cbioportal-maintenance/updating-your-cbioportal-installation":"updating-your-cbioportal-installation",
    "3.-cbioportal-maintenance/updating-gene-and-gene_alias-tables":"updating-gene-and-gene_alias-tables",
    "4.-development/feature-development-guide":"feature-development-guide",
    "4.-development/cbioportal-er-diagram":"cbioportal-er-diagram",
    "4.-development/build-different-frontend":"build-different-frontend",
    "4.-development/session-service-working":"ession-service-working",
    "4.-development/manual-test-cases":"manual-test-cases",
    "4.-development/release-procedure":"release-procedure",
    "4.-development/deployment-procedure":"deployment-procedure",
    "5.1-data-loading/data-loading":"data-loading",
    "5.2-datasets/downloads":"downloads",
    "6.-web-api-and-clients/api-and-api-clients":"web-api-and-clients",
    "7.-user-guide/user-guide":"user-guide",
    "5.1-data-loading/data-loading/using-the-dataset-validator":"using-the-dataset-validator",
    "5.1-data-loading/data-loading/using-the-metaimport-script":"using-the-metaimport-script",
    "5.1-data-loading/data-loading/file-formats/z-score-normalization-script":"z-score-normalization-script",
    "5.1-data-loading/data-loading/data-loading-maintaining-studies":"data-loading-maintaining-studies",
    "5.1-data-loading/data-loading/data-loading-for-developers":"data-loading-for-developers",
    "5.1-data-loading/data-loading/data-loading-tips-and-best-practices":"data-loading-tips-and-best-practices",
    "5.1-data-loading/data-loading/import-gene-sets":"import-gene-sets",
    "5.1-data-loading/data-loading/import-gene-panels":"import-gene-panels",
};

let newLocation = null;
Object.keys(redirectMap).forEach(function(path){
    if ((new RegExp(path + "$").test(window.location.pathname))) {
        let root = document.querySelector("[name=turbo-root]")?.content || "";
        root = root.replace("/","");
        newLocation = `${window.location.protocol}//${window.location.hostname}${window.location.port ? ":" : ""}${window.location.port}/${root}/${window.redirectMap[path]}${window.location.hash}`; 
    } 

});

if (newLocation) {
    window.location.href = newLocation;
} else {
    document.getElementById("file-not-found").innerHTML = ` 
    <h1>Document not found</h1>
    <p>We recently changed our link structure and you may have a defunct url.  Please search for your topic above.</p> 
    `;
    document.getElementById("file-not-found").style.display = "block";
}
</script>




<h2 id=""></h2>




The cBioPortal for Cancer Genomics was originally developed at [Memorial Sloan Kettering Cancer Center](https://www.mskcc.org/) (MSK). The [public cBioPortal site](https://www.cbioportal.org) is hosted by the [Center for Molecular Oncology](https://www.mskcc.org/research/molecular-oncology) at MSK. The cBioPortal software is now available under an open source license via [GitHub](https://github.com/cBioPortal/). The software is now developed and maintained by a multi-institutional team, consisting of MSK, the Dana Farber Cancer Institute, Princess Margaret Cancer Centre in Toronto, Children's Hospital of Philadelphia, Caris Life Sciences, [The Hyve](https://thehyve.nl) and [SE4BIO](https://www.se4.bio/) in the Netherlands, and Bilkent University in Ankara, Turkey.

## Memorial Sloan Kettering Cancer Center
* Benjamin Gross
* S Onur Sumer
* Hongxin Zhang
* Ritika Kundra
* Ino de Bruijn
* Robert Sheridan
* Angelica Ochoa
* Aaron Lisman
* Manda Wilson
* Avery Wang
* Ramyasree Madupuri
* Gaofei Zhao
* Xiang Li
* Bryan Lai
* Calla Chennault
* Anusha Satravada
* Rima AlHamad
* Nikolaus Schultz

## Dana-Farber Cancer Institute
* Ethan Cerami
* Tali Mazor
* Jeremy Easton-Marks
* Zhaoyuan (Ryan) Fu
* Augustin Luna
* James Lindsay
* Chris Sander

## Princess Margaret Cancer Centre, Toronto
* Prasanna Jagannathan
* Trevor Pugh

## Children's Hospital of Philadelphia
* Charles Haynes
* David Higgins
* Allison Heath
* John Maris
* Adam Resnick
* Miguel Brown

## Caris Life Sciences
* Jianjiong Gao
* Priti Kumari
* Karthik Kalletla

## The Hyve
* Oleguer Plantalech
* Pim van Nierop
* Sander Rodenburg
* Bas Leenknegt
* Elena G Lara
* Jessica Singh
* Matthijs Pon
* Tim Kuijpers
* Mirella Kalafati
* Sjoerd van Hagen

## SE4BIO
* Pieter Lukasse
* Ruslan Forostianov

## Bilkent University
* Ugur Dogrusoz
* Yusuf Ziya Ozgul

## Alumni
* Adam Abeshouse
* B Arman Aksoy
* M Salih Altun
* Istemi Bahceci
* Caitlin Byrne
* Hsiao-Wei Chen
* Ersin Ciftci
* Fred Criscuolo
* Leonard Dervishi
* Gideon Dresdner
* Andy Dufilie
* Catherine Del Vecchio Fitz
* Olivier Elemento
* Ziya Erkoc
* Arthur Goldberg
* Zachary Heins
* Michael Heuer
* Anders Jacobsen
* Peter Kok
* Erik Larsson
* Dong Li
* Divya Madala
* Tamba Monrose
* Irina Pulyakhina
* Pichai Raman
* M Furkan Sahin
* Kaan Sancak
* Alexandros Sigaras
* Luke Sikina
* Jing Su
* Yichao Sun
* Sander Tan
* Jiaojiao Wang
* Stuart Watt
* James Xu
* Dionne Zaal
* Kelsey Zhu
* Fedde Schaeffer
* Riza Nugraha
* Ngoc Nguyen
* Diana Baiceanu
* Paul van Dijk
* Kees van Bochove

## Funding for the cBioPortal for Cancer Genomics is or has been provided by:

### Current:
* NCI, through [ITCR grant NCI-U24CA274633](https://itcr.nci.nih.gov/funded-project/cbioportal-cancer-genomics) and [HTAN grant NCI-U24CA233243](https://humantumoratlas.org/)
* Marie-Jos√© and Henry R. Kravis Center for Molecular Oncology at MSK
* Dana Farber Cancer Institute
* American Association for Cancer Research through AACR Project GENIE
* Prostate Cancer Foundation
* The Cholangiocarcinoma Foundation
* Robertson Foundation

### Past:
* NCI, through [ITCR grant NCI-U24CA220457](https://itcr.nci.nih.gov/funded-project/cbioportal-cancer-genomics)
* Stand Up 2 Cancer
* The Ben & Catherine Ivy Foundation
* NCI, as a [TCGA Genome Data Analysis Center (GDAC)](https://tcga.cancer.gov/wwd/program/research_network/gdac.asp) (NCI-U24CA143840)
* NCRR, as the [National Resource for Network Biology (NRNB)](https://nrnb.org/) Research Resource (RR 031228-02)
* Starr Cancer Consortium
* Breast Cancer Research Foundation
* Adenoid Cystic Carcinoma Research Foundation
* POETIC Consortium
* Parker Institute for Cancer Immunotherapy
# Architecture Overview
cBioPortal consists of the following components:

- [backend](https://github.com/cBioPortal/cbioportal)
    - MySQL database
    - REST API written in Java Spring
    - Redis cache for storing frequently used queries (optional)
- [validator](https://github.com/cBioPortal/cbioportal/tree/master/core/src/main/scripts/importer)
  checks file formats before importing data into the database
- [frontend](https://github.com/cBioPortal/cbioportal-frontend)
  built with React, Mobx and Bootstrap
- [session service](https://github.com/cBioPortal/session-service) for storing
  user saved data such as virtual studies and groups
     - REST API written in Java Spring enabling retrieval and writing to the database
     - MongoDB database
- cBioPortal also uses the APIs from various [external services](#external-services) to provide more information about a variant

## Backend

The [backend](https://github.com/cBioPortal/cbioportal) is written in Java and
connects to a MySQL database to serve a REST API following the OpenAPI
specification (https://www.cbioportal.org/api/). Note that the repo where this
lives in (https://github.com/cBioPortal/cbioportal) also contains Java classes
to import data as well as the validator. The backend can be configured to
connect to a Redis cache to store database query results for improved performance.

The backend is organized as a multi-module Maven project.
See [cBioPortal backend code organization](Backend-Code-Organization.md).

## Validator
The
[validator](https://github.com/cBioPortal/cbioportal/tree/master/core/src/main/scripts/importer)
checks file formats before importing data into the database. There is a wrapper
script `metaImport.py` that validates the data and subsequently calls the
relevant Java classes to import the data.

## Session Service

The [session service](https://github.com/cBioPortal/session-service) is used
for storing user saved data such as virtual studies and groups. See the
[tutorials](https://www.cbioportal.org/tutorials) section to read more about
these features. Session service is a Java app that serves a REST API backed by
a Mongo database. The session service is served as a proxy through the
cBioPortal backend REST API. The backend is therefore the only component that
needs to be able to connect to it. The frontend does not connect to it
directly. 

## Frontend
The frontend is a single page app built with React, Mobx and Bootstrap. The
data gets pulled from the backend REST API. The frontend is by default included
with the backend so no extra setup is required. 

## External Services
cBioPortal uses the APIs from several external services to provide more
information about a variant:

- [OncoKB](#oncokb)
- [CIVIC](#civic)
- [Genome Nexus](#genome-nexus)
- [G2S](#g2s)

For privacy concerns see the section: [A note on privacy](#a-note-on-privacy).

### OncoKB
[OncoKB](https://www.oncokb.org) is a precision oncology knowledge base that
contains information about the effects and treatment implications of specific
cancer gene alterations. See the section [OncoKB Data Access](/deployment/integration-with-other-webservices/OncoKB-Data-Access.md)
for how to configure external OncoKB service.

### CIVIC
[CIVIC](https://civicdb.org) is a community-edited forum for discussion and
interpretation of peer-reviewed publications pertaining to the clinical
relevance of variants (or biomarker alterations) in cancer. For information on
how to deploy this service yourself see:
https://github.com/griffithlab/civic-server. It is also possible to disable
showing CIVIC in cBioPortal by setting `show.civic=false` in the
`application.properties` (See [application.properties reference](/deployment/customization/application.properties-Reference.md#civic-integration)).

### Genome Nexus
[Genome Nexus](https://www.genomenexus.org) is a comprehensive one-stop
resource for fast, automated and high-throughput annotation and interpretation
of genetic variants in cancer. For information on how to deploy this service
yourself see: https://github.com/genome-nexus/genome-nexus. For more
information on the various annotation sources and versions provided by Genome
Nexus see: https://docs.genomenexus.org/annotation-sources.

### G2S
[G2S (Genome to Structure)](https://g2s.genomenexus.org) maps genomic variants
to 3D structures. cBioPortal uses it on the mutations tab to show the variants
on a 3D structure. For information on how to deploy this service yourself see:
https://github.com/genome-nexus/g2s.

### A note on privacy

cBioPortal calls these services with variant information from the cBioPortal
database. It however does not send over information that links a variant to a
particular sample or patient. If this is a concern for your use case we recommmend
to deploy your own versions of these services. See the sections above to
linkouts for instructions on how to do this.
# cBioPortal Backend Code Organization

## Maven Project

The backend code is structured as a [Maven](https://maven.apache.org/index.html)
project. Included in the repository are [pom.xml files](https://maven.apache.org/pom.html#What_is_the_POM)
which define the code components and modules which make up the backend. The pom.xml
files also list dependencies on external open source libraries, and dependencies between cBioPortal Maven modules.
They also give instructions for how to build and package code components into a
deployable format.

### Refactoring

The code base is undergoing a long term refactoring effort which is nearing completion.

#### Web service refactoring

The original data distribution service (still in operation) was a custom servlet called
[webservice.do](https://github.com/cBioPortal/cbioportal/blob/v3.1.5/core/src/main/java/org/mskcc/cbio/portal/servlet/WebService.java)
This servlet takes a "cmd" argument which selects the category of data requested.
More details can be seen on the cBioPortal [webAPI](https://www.cbioportal.org/webAPI) tab.

An example request:
`curl -X GET "https://www.cbioportal.org/webservice.do?cmd=getGeneticProfiles&cancer_study_id=brca_tcga"`

The new data distribution approach follows a [RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer) approach,
mapping various URL paths to data services, allowing the specification of database
entities within the path as helpful.

An example request:
`curl -X GET "https://www.cbioportal.org/api/studies/brca_tcga/molecular-profiles"`

#### Presentation / frontend refactoring

The original presentation code was written as a set of java service pages ([JSP](https://en.wikipedia.org/wiki/JavaServer_Pages)),
connected to servlets which populated needed data for page rendering into request attributes
within the backend server. You can see examples of the setting of these request attributes in
[this example servlet](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/servlet/QueryBuilder.java)

This has largely been superceded and expanded on by a new [React](https://reactjs.org/)
javascript application located in a separate
[GitHub repository](https://github.com/cBioPortal/cbioportal-frontend), which maintains a
browser data store populated from the new RESTful Web API. Although the new frontend
project is not dependent on the original custom servlets for data delivery, we may continue
to use the servlet / JSP delivery path to embed configuration data in the initial HTML page
in order to allow us to optimize loading times. Some of this logic has already been
migrated out of the core module and moved to the portal module.

### Current Maven Modules

#### The core module

The [core](https://github.com/cBioPortal/cbioportal/blob/master/core) module contains the oldest code in the code base. Much of
the backend web data services logic has been re-implemented and expanded in
other backend maven modules, such as [model](https://github.com/cBioPortal/cbioportal/blob/master/model),
[persistence-mybatis](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-mybatis),
[service](https://github.com/cBioPortal/cbioportal/blob/master/service), and [web](https://github.com/cBioPortal/cbioportal/blob/master/web). We plan to discontinue the legacy
webservice.do data servlet and fully transition to the new data services API in
the near future. The core module will then be purged of much of the legacy
code. Any surviving functionality (such as the handling of global configuration
properties) may be relocated into new modules. It is therefore important that
no new code features be introduced which are dependent on the core module
functionality.

#### The central stack modules : web, service, persistence

Three maven modules make up the central stack of the new web API implementation.

* [web](https://github.com/cBioPortal/cbioportal/blob/master/web) : define Web API request handlers, map endpoints, select services
* [service](https://github.com/cBioPortal/cbioportal/blob/master/service) : call persistence module or utils, apply business logic
* [persistence](https://github.com/cBioPortal/cbioportal/blob/master/persistence) : retrieve data from database. Two submodules:
  * [persistence-api](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api) : declaration of repository classes / functions, caching markup
  * [persistence-mybatis](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-mybatis) : implementation of persistence-api using [MyBatis](https://mybatis.org/mybatis-3/) mappers

#### Other modules

* [model](https://github.com/cBioPortal/cbioportal/blob/master/model) : data model POJO classes (used throughout the stack)
* [security](https://github.com/cBioPortal/cbioportal/blob/master/security) : user authentication methods and authorization, request filters 
* [business](https://github.com/cBioPortal/cbioportal/blob/master/business) : a legacy refactor of webservice.do, has active dataSource definition
* [portal](https://github.com/cBioPortal/cbioportal/blob/master/portal) : web application packaging and launch
* [scripts](https://github.com/cBioPortal/cbioportal/blob/master/scripts) : data import tool packaging (from core module scripts package)
* [db_scripts](https://github.com/cBioPortal/cbioportal/blob/master/db_scripts) : installation and migration scripts for the database

#### External code modules

* [cbioportal-frontend](https://github.com/cBioPortal/cbioportal-frontend) : a React application using MobX and TypeScript
* [session-service](https://github.com/cBioPortal/session-service) : an external session key/query specifier storage system

**cbioportal-frontend** is packaged in the web application as a default frontend
implementation, but the source of the frontend code can also be directed to an external
source host and be deployed independently of the backend web application. See
[details](Deployment-Procedure.md)

**session-service** is imported into the web module in order to set up a proxy service
which receives and forwards requests for saved cBioPortal sessions to a separate system
providing this storage (using a document based database) The code is needed for handling
modeled types such as VirtualStudy and Session

### cBioPortal Module Dependencies

![Module Dependencies](images/maven-module-dependencies.png)

_A module is directly dependent on all modules which touch it from below (touching only at a corner
does not count). Dependencies are transitive : web is dependent on service, service
is dependent on persistence-mybatis, therefore web is also dependent on
persistence-mybatis. During packaging, all dependencies are made available to the
dependent package for use. Avoid creating new dependencies between modules.
**cyclical dependencies will cause errors during build**_

Module | Dependencies
------ | ------------
portal | core, _cbioportal-frontend_
scripts | core
core | web 
web | business, service, _session-service_
security-spring | service
service | persistence-mybatis
persistence-mybatis | persistence-api, db-scripts
persistence-api | model
business | model
db-scripts |
model |
# cBioPortal Development Guidelines

1. Avoid introducing new functionality in the core module or creating new dependencies on the existing core module code.
1. One exception to the previous guideline is that new data import functions should be added to the core module scripts package.
1. Do not expand the business module.
1. Place all new web API controllers in the web module.
1. Make abstract data-driven web controllers part of the default/public interface (tag them with @PublicApi).
1. Make Special purpose or visualization-driven web controllers part of the internal interface (tag them with @InternalApi).
1. Do not include business logic in the web module handler functions. Limit processing to argument examination and service module method selection.
1. Test new data-driven web controllers for proper behavior on a portal deployment which requires user authentication and authorities.
1. Do not call persistence module functions directly from the web module. Create pass-through service layer functions instead.
1. Locate database query code in the persistence modules, and follow existing patterns.
1. Consider the tradeoffs between using database query constructs to accomplish business logic requirements versus writing service layer java code.
# Create and Publish Virtual Study

A [Virtual Study](./user-guide/faq.md#what-is-a-virtual-study) defines a subset or a combination of samples from one or more studies in the system.

*Note*: To publish or un-publish a virtual study, your cBioPortal instance must be configured with  `session.endpoint.publisher-api-key` in the `application.properties`.

## Create Virtual Study

To create a virtual study in cBioPortal, follow these steps:

1. Define the desired filters on the study or multiple studies summary page.
2. Click the button with the bookmark icon (ÔÄÆ) in the top right corner of the screen.
3. Provide a title and description, then click the Save button. You will see a link that looks like:

```
https://<cbioportal_host>/study?id=<virtual_study_id>
```

4. Save the virtual study link or ID if you want to publish it.

If you are logged in, this virtual study will appear in the `My Virtual Studies` section on the landing page.
You can always find the ID of the virtual study from the URL of the page that opens after clicking on it.

## Publish Virtual Study

To publish a virtual study, you need to supply the publisher API key in the `X-PUBLISHER-API-KEY` header.

Here is a curl command to publish a virtual study:
```shell
curl \
  -X POST \
  -H 'X-PUBLISHER-API-KEY: <session.endpoint.publisher-api-key>' \
  -v 'http://<cbioportal_host>/api/public_virtual_studies/<virtual_study_id>'
```
The published virtual study will appear under the `Public Virtual Studies` section (next to the `My Virtual Studies` section) on the landing page for all users of cBioPortal.

While publishing, you can specify the PubMed ID (`pmid`) and `typeOfCancerId` of the virtual study using the following command:
```shell
curl \
  -X POST \
  -H 'X-PUBLISHER-API-KEY: <session.endpoint.publisher-api-key>' \
  -v 'http://<cbioportal_host>/api/public_virtual_studies/<virtual_study_id>?pmid=<pmid>&typeOfCancerId=<code>'
```

The type of cancer code should match the known types of cancers in the cBioPortal database.
If the type of cancer is specified, the published virtual study will appear under the respective cancer section on the landing page.
Specifying the `pmid` enables a link to the PubMed page of the study.

## Un-publish Virtual Study

To un-publish a virtual study, you need to supply the publisher API key in the `X-PUBLISHER-API-KEY` header.
After un-publishing, virtual study will no longer be displayed in the `Public Virtual Studies` section on the landing page.
However, it reappears in the `My Virtual Studies` section for the owner.

Here is the command to un-publish a virtual study:
```shell
curl \
  -X DELETE \
  -H 'X-PUBLISHER-API-KEY: <session.endpoint.publisher-api-key>' \
  -v 'http://<cbioportal_host>/api/public_virtual_studies/<virtual_study_id>'
```# Importing single data files for development
In some cases, for example during development, it may be useful to import a single data file into an existing study. To import one data file at a time, you can use the following command. Note that this process will not validate the data.

This can be done by running `cbioportalImporter.py` from `<cbioportal_source_folder>/core/src/main/scripts/importer/`.

## Requirements

This script requires `$PORTAL_HOME` to point to the folder containing your
cBioPortal configuration. This can be done with:
```
export PORTAL_HOME=<cbioportal_configuration_folder>
```

## Workflow

First, if your cancer type does not yet exist, you need to create it:
```
./cbioportalImporter.py -c import-cancer-type -data <path to study directory>/<name of data file>
```

Next, create the study using
```
./cbioportalImporter.py -c import-study -meta <path to study directory>/<name of meta file>
```
The meta file has to contain the study information.

Now you can import your data file(s): 
```
./cbioportalImporter.py -c import-study-data -meta <path to study directory>/<name of meta file> -data <path to study directory>/<name of data file> 
```
:warning: 
Your first data file should always be the clinical data!

Finally, after you've imported all data, import your case lists:
```
./cbioportalImporter.py -c import-case-list -meta <path to study directory>/<path to case lists>
```

#### Example:
```
./cbioportalImporter.py -c import-cancer-type -data /data/brca_small/data_cancer_type.txt
./cbioportalImporter.py -c import-study -meta /data/brca_small/meta_study.txt
./cbioportalImporter.py -c import-study-data -meta /data/brca_small/meta_clinical.txt -data /data/brca_small/data_clinical.txt
./cbioportalImporter.py -c import-study-data -meta /data/brca_small/meta_expression.txt -data /data/brca_small/data_expression.txt
./cbioportalImporter.py -c import-case-list -meta /data/brca_small/case_lists
```
# Data loading: Importing without validation and deleting studies
For data curators and developers cbioportalImporter.py is available. This script can import data regardless of validation results. If data format is incorrect, the importer may stop with an error or crash, or leave the database in an inconsistent state.

This script can also be used to delete studies.

- [Requirements](#requirements)
- [Importing a study without validation](#importing-a-study-without-validation)
- [Deleting a study](#deleting-a-study)

## Requirements
This script requires `$PORTAL_HOME` to point to the folder containing your
cBioPortal configuration. This can be done with:
```
export PORTAL_HOME=<cbioportal_configuration_folder>
```

The script itself can be found in `<cbioportal_source_folder>/core/src/main/scripts/importer`.

## Importing a study without validation 
To import a study without validation, run: 
```
./cbioportalImporter.py -s <path to study directory>
```

For example:
```
./cbioportalImporter.py -s ../../../test/scripts/test_data/study_es_0/
```

## Deleting a study
To remove a study, run: 
```
./cbioportalImporter.py -c remove-study -meta <path to study directory>/meta_study.txt
```
The `meta_study.txt` file should contain the study ID in `cancer_study_identifier: ` of the study you would like to remove.

For example:
```
./cbioportalImporter.py -c remove-study -meta /data/brca_small/meta_study.txt
```

If you have the Cancer Study Id of the study, or studies you want to remove, you can also use:
```
./cbioportalImporter.py -c remove-study -id study1_id
```
Where `study1_id` is the Cancer Study Id of the study you would like to remove.

You can also remove multiple studies at once by passing the Cancer Study Ids separated by commas:
```
./cbioportalImporter.py -c remove-study -id study1_id,study2_id,study3_id
```
Where `study1_id`, `study2_id` and `study3_id` are the Cancer Study IDs of the studies you would like to remove.
Here we describe some Tips and Best Practices.

## Running GISTIC 2.0
To generate discrete copy number data file you may need to run GISTIC 2.0. GISTIC 2.0 can be [installed](https://www.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=216&p=t) or run online using the GISTIC 2.0 module on [GenePattern](https://cloud.genepattern.org). 
Running GISTIC 2.0 requires two input files: 

1. A segmentation file, which contains the segmented data 
2. A marker file, which identifies the marker names and positions of the markers in the original dataset (before segmentation). 

In some cases the marker file may not be available. Your can create one as follows:
Using your segmentation file, create a line for each start and end position. E.g. if your seg file contains

<table>
<tr>
<td>Sample</td><td>Chrom</td><td>Start</td><td>Stop</td><td>#Mark</td><td>Seg.CN</td>
</tr>
<tr>
<td>S1</td><td>1</td><td>61735</td><td>77473</td><td>10</td><td>-0.1234 </td>
</tr>
</table>

In your markerfile this becomes

<table>
<tr>
<td>Marker Name</td><td>Chrom</td><td>Pos</td>
</tr>
<tr>
<td>S1</td><td>1</td><td>61735</td>
</tr>
<tr>
<td>S1</td><td>1</td><td>77473</td>
</tr>
</table>


## Effect of cBioPortal instance on validation
When validating data, you can decide against which server to validate your data with the -u flag. The selected server can have a significant effect on the validation results in the following ways:

1. Genes may or may not be available on a specific server
2. Clinical data and its description may vary per server
3. ...

It is advised to use the server where you plan on upload your data as validation server. 
# Data Loading
## Introduction
This page is the starting point for data loading. The [General Overview](#general-overview) section below contains all the required steps to get you started. 

## General Overview
Getting your study data into cBioPortal requires four steps:

1. [Setting up the validator](#setting-up-the-validator)
2. [Preparing your study data](#preparing-study-data)
3. [Validating your study data](#validating-your-study-data)
4. [Loading your study data](#loading-data)

## Setting up the validator
### Installation

If you have a git clone of cBioPortal, the relevant scripts can be found in the folder: `<your_cbioportal_dir>/core/src/main/scripts/importer`

### Dependencies
The scripts run in Python 3.4 or newer, and they require the modules `requests` and `pyyaml`.
You can use this command to install those modules:
```console
$ sudo python3 -m pip install requests pyyaml
```

If you want the scripts to be able to generate html reports (recommended way for reading the validation errors, if any), then you will also need to install `Jinja2`. You can use this command:
```console
$ sudo python3 -m pip install Jinja2
```

## Preparing Study Data 
A study to be loaded in cBioPortal can basically consist of a directory where all the data files are located. 
Each *data* file needs a *meta* file that refers to it and both files need to comply to the format required for the specific data type. The format and fields expected for each file are documented in the [File Formats page](File-Formats.md). Below is an example of the files in such a directory.

```
dir
|-meta_study.txt
|-meta_cancer_type.txt -> cancer_type.txt
|-meta_clinical.txt -> data_clinical.txt
|-meta_[expression|mutations|CNA|etc] -> data_[expression|mutations|CNA|etc]
```
#### Rules
There are just a few rules to follow:
- meta_study, meta_clinical and respective clinical data file are the only **mandatory** files.
- cancer type files can be mandatory if the study is referring to a cancer type that does not yet exist in the DB.
- meta files can be named anything, as long as it starts or ends with name 'meta'. E.g. meta_test, meta.test, test.meta are all fine; metal_test and metastudy are wrong.
- data files can be named anything and are referenced by a property `data_filename` set in the meta file. 

## Validating your study data
Once all files are in place and follow the proper format, you can [validate your files using the dataset validator script](/Using-the-dataset-validator.md). 

The validation can be run standalone, but it is also integrated into the [metaImport script](/Using-the-metaImport-script.md), which validates the data and then loads it if validation succeeds. 

## Loading Data
To load the data into cBioPortal, the [metaImport script](/Using-the-metaImport-script.md) has to be used. This script first validates the data and, if validation succeeds, loads the data. 

## Removing a Study
To remove a study, the [cbioportalImporter script](/Data-Loading-Maintaining-Studies.md#deleting-a-study) can be used.

## Example studies
Examples for the different types of data are available on the [File Formats](/File-Formats.md) page. The Provisional TCGA studies, downloadable from the [Data Sets section](https://www.cbioportal.org/datasets) are complete studies that can be used as reference when creating data files.

## Public Virtual Studies
If your new study data is a subset or a combination of existing studies in the system, consider using [Public Virtual Studies](./Create-And-Publish-Virtual-Study.md) instead of duplicating data. # Authenticating and Authorizing Users via Keycloak

**‚ö†Ô∏è This documentation is for keycloak <v20, see related [ticket](https://github.com/cBioPortal/cbioportal/issues/10360) ‚ö†Ô∏è**

## Introduction

The cBioPortal includes support for Keycloak authentication. Keycloak can function as an Identity Provider (IDP) for cBioPortal.
This document explains why you might find Keycloak authentication useful for storing your user login information outside
the cBioPortal database. It also shows you how to configure Keycloak to communicate with your instance of cBioPortal using
SAML (Security Assertion Markup Language).

Please note that configuring your local instance to use Keycloak authentication requires a Keycloak server to be set up. For details on how to set up a Keycloak server, please read online document at <https://www.keycloak.org/guides#server>.

This document focuses mainly on the steps to configure Keycloak for **authenticating** cBioPortal users.

To skip to the authorization section see: [authorization with Keycloak](#authorization-with-keycloak). Or continue reading to learn how to integrate Keycloak with cBioPortal.

### Why Keycloak?

Keycloak is an [open source identity and access management solution](https://www.keycloak.org/docs/latest/server_admin/index.html). It has a built-in RDBM system to store login information. It can help build a security layer on top of the cBioPortal web application.

Keycloak boils down to three simple terms:
* **realm**: A realm secures and manages security metadata for a set of users, application, and registered auth clients.
* **client**: Clients are entities that can request authentication of a user within a realm.
* **role**: Roles identify a type or category of user. Keycloak often assigns access and permissions to specific roles rather than individual users for a fine-grained access control.

Keycloak offers three types of roles:
* Realm-level roles are in global namespace shared by all clients.
* Client roles have basically a namespace dedicated to a client.
* A composite role is a role that has one or more additional roles associated with it.


### How does Keycloak SAML Authentication work?

Keycloak supports both OpenID-Connect and SAML authentication. When you use SAML authentication, the Keycloak server exchanges XML documents with a web application. XML signatures and encryption are then used to verify requests from the application.

## Configure Keycloak to authenticate your cbioportal instance
1. Log in to your Keycloak Identity Provider, e.g. <http://localhost:8080/auth>, as an admin user. :warning: when setting this up on something else than localhost (e.g. production), you will need to use/enable https on your Keycloak server. For simplicity, the rest of the documentation below continues on http://localhost.
2. Hover over the top-left‚Äìcorner drop down menu (titled ‚Äò**Master**‚Äô) to create a new realm.
![](/images/previews/add-realm.png)
Please note if you are logged in the master realm, this drop-down menu lists all the realms created. The last entry of this drop-down menu is always **Add Realm**. Click this to add a realm. Then type '_cbioportal_' in the name field and click the **Create** button.
3. To create a SAML client, go to the **Clients** item in the left menu. On this page, click the **Create** button on the right. This will bring you to the **Add Client** page.
    * Enter a **Client ID** for the client, e.g. '_cbioportal_', this will be the expected `issuer` value in SAML requests sent by the application.
    * Select _saml_ in the **Client Protocol** drop down box.
    * Enter `http://localhost:8081/saml` in the **Client SAML Endpoint** textbox, this is the URL that the Keycloak server will send SAML requests and responses to. Then click the **Save** button; this will take you to the client page below.

![](/images/previews/edit-client.png)

4. Choose _email_ as your **Name ID Format**.
5. Next, enter a pattern for **Valid Redirect URIs** that Keycloak can use upon a successful authentication, e.g. `http://localhost:8081/*`. :information_source: notice that you can add multiple URLs in this field. You could use this in some cases to support
the URLs with and without port (e.g. if tomcat is running on port `80` and you want to allow both `http://localhost:80/*` and `http://localhost/*` as redirect URLs).
6. Set **Force POST Binding** and **Front Channel Logout** to _OFF_ and **Force Name ID Format** to _ON_.
7. Expand the subsection **Fine Grain SAML Endpoint Configuration** and set **Logout Service POST Binding URL** to `http://localhost:8081/saml/logout`.
8. Leave everything else as it is and click **Save**.

### Map SAML Assertion Attributes
To specify attributes included in the SAML assertion, simply click on the **Mappers** tab, and add mappers using the **Add Builtin** and **Create** buttons.
Make sure you add at least:
- the built-in User Property mapper named _X500 email_ and
- a _Role list_-type attribute (keep the default word _Role_ as its **Role attribute name**).
- a _User Property_-type attribute with the name _username_. Use _username_ for the attributes **Property**, **Name**, and **SAML Attribute Name**.

![](/images/previews/add-mappers.png)

Edit the email attribute to use the word _email_ as the **SAML Attribute Name**.
![](/images/previews/edit-email-attribute.png)

Finally, head to the **Scope** tab for the client and switch off
**Full Scope Allowed**, to ensure that only those roles relevant to a
particular cBioPortal instance are listed in assertions sent to the
instance, and not any other roles tracked in Keycloak.

### Export configuration for cBioPortal

There are two known ways to download the keycloak configuration (aka IDP SSO Descriptor) file for cBioPortal.

#### I. Download via link

The file can be fetched by the following url:

```
http(s)://{KEYCLOAK-URL}/auth/realms/{REALM-NAME}/protocol/saml/descriptor
```

For example:

```
curl -o client-tailored-saml-idp-metadata.xml "http://localhost:8081/auth/realms/cbioportal/protocol/saml/descriptor"
```

**Note:** if you use https protocol with self-signed protocol you need to add `--insecure` option to the above curl command.

#### II. Export via GUI (legacy)
1. Next, navigate to the **Installation** tab for the same client.
2. Select _SAML Metadata IDPSSODescriptor_ as the Format Option and click the **Download** button.
‚ö†Ô∏è This GUI option has been removed from the newer versions of Keycloak.
![](/images/previews/download-IDPSSODescriptor-file.png)

After you've downloaded the XML file with one of the above ways, move it to `portal/src/main/resources/` if you're compiling cBioPortal yourself or if you're using the Docker container, mount the file in the `/cbioportal-webapp` folder with `-v /path/to/client-tailored-saml-idp-metadata.xml:/cbioportal-webapp/WEB-INF/classes/client-tailored-saml-idp-metadata.xml`.

## Create a signing key for cBioPortal

Use the Java '`keytool`' command to generate keystore, as described
[here](Authenticating-Users-via-SAML.md#creating-a-keystore)
on the page about SAML in cBioPortal:

```
keytool -genkey -alias secure-key -keyalg RSA -keystore samlKeystore.jks
```

**Important:** The validity of this keystore is **90 days**. You can change the default
value by adding the `-validity` parameter and the number of days (e.g. `-validity 200`
for 200 days). If the keystore expires, then **'invalid requester'** errors are thrown.

Install the generated JKS file to `portal/src/main/resources/` if you're compiling cBioPortal yourself or if you're using the Docker container, mount the file in the `/cbioportal-webapp` folder with `-v /path/to/samlKeystore.jks:/cbioportal-webapp/WEB-INF/classes/samlKeystore.jks`.

Import the key's certificate into Keycloak, so that Keycloak knows that it can trust the holder of this
key. To do that, head to the **SAML Keys** tab in the keycloak admin screen about the `cbioportal` client and:
1. Click the **Import** button.
2. Select the _JKS_ archive format.
3. Specify the key alias _secure-key_.
4. Type the store password _apollo1_ (not the private key password, as Keycloak only
needs to know the certificate)
5. Select the file you just installed.

**Important:** Keycloak may not give an indication of successful
completion, but when navigating to the **SAML Keys** tab again you
should now see the certificate and no private key.

## Modifying configuration

1. Within the application.properties file , make sure that this line is present:
```
    app.name=cbioportal
```

2. Then, modify the properties under the comment `# authentication`. In particular, see the options listed in the example below:

```properties
    filter_groups_by_appname=false
    saml.sp.metadata.entityid=cbioportal
    saml.sp.metadata.wantassertionsigned=true
    saml.idp.metadata.location=classpath:/client-tailored-saml-idp-metadata.xml
    saml.idp.metadata.entityid=http://localhost:8080/auth/realms/cbioportal
    saml.keystore.location=classpath:/samlKeystore.jks
    saml.keystore.password=apollo1
    saml.keystore.private-key.key=secure-key
    saml.keystore.private-key.password=apollo2
    saml.keystore.default-key=secure-key
    saml.idp.comm.binding.settings=defaultBinding
    saml.idp.comm.binding.type=
    saml.idp.metadata.attribute.email=email
    saml.idp.metadata.attribute.userName=username
    saml.idp.metadata.attribute.role=Role
    saml.custom.userservice.class=org.cbioportal.security.spring.authentication.keycloak.SAMLUserDetailsServiceImpl
    # global logout (as opposed to local logout):
    saml.logout.local=false
    saml.logout.url=/
```

## Obtain user identities

### Optional: create users in Keycloak

To create a user, click on **Users** in the left menu bar. This menu
option brings you to the user list page. On the right side of the
empty user list, you should see an **Add User** button. Click that to
start creating your new user.

![](/images/previews/add-user.png)

### Optional: integrate company-wide authentication services

Keycloak can read credentials from existing user databases, for
instance over LDAP. This is referred to as _user federation_. Keycloak
can also allow authentication by an external login form altogether
using a protocol such as SAML, it calls this _identity brokering_. In
either case, Keycloak acts as a proxy between your user directory and
cBioPortal, deciding which authorities to grant when telling
cBioPortal that the user has authenticated.

Please refer to the Keycloak documentation on
[user federation](https://www.keycloak.org/docs/latest/server_admin/index.html#_user-storage-federation)
and
[identity brokering](https://www.keycloak.org/docs/latest/server_admin/index.html#_identity_broker)
for more information on how to integrate Keycloak with your local LDAP
or SAML service.

#### Federate LDAP/AD user directories

Some notes on user federation using LDAP/Active Directory:

By specifying the **Vendor** of your LDAP server, Keycloak will choose
sensible defaults for the required objectClasses and attributes of
your user entries. Apart from the Dedicated Name of the tree in which
to search for users and the DN and password that Keycloak should use
to bind to the server, make sure to specify the following **Custom
User LDAP Filter** to ensure that only user entries that have an email
address are considered:

```
(mail=*)
```

When using LDAP to load users from your institute's user directory,
you will most likely want Keycloak to refrain from trying to
synchronise changes in user details back to the central directory. You
should set **Edit Mode** to `READ_ONLY`, as the alternative `UNSYNCED`
would mean that users can be changed in the Keycloak database once
imported from LDAP, and start diverging. Also disable **Sync
Registrations** unless you want Keycloak to add new users to the LDAP
store.

Do turn on **Import Users** to make Keycloak remember users after the
first login, if you want to be able to assign non-default roles. If
the LDAP tree holding your users is large and you do not want to
import all users into Keycloak, make sure to disable **Periodic Full
Sync** and **Periodic Changed Users Sync**.

## Authorization with Keycloak

### Create roles to authorize cBioPortal users

The roles you assign to users will be used to tell cBioPortal which
studies a user is allowed to see.

To create a role, head to the **Roles** tab that is displayed along
the top while configuring the `cbioportal` client ‚Äì this tab is _not_
the link of the same name in the left sidebar.  Click the **Add
Role** button. Enter a name (e.g.  `brca_tcga_pub`) and description
for the role and hit the **Save** button.

![](/images/previews/add-role-for-study.png)

**Note:** if `filter_groups_by_appname` is set to `false` as specified above, the `Role Name` has to match with an id of the study you would give access to by assigning this role. Otherwise, if `filter_groups_by_appname` is set to `true` (**DEFAULT**), you have to add the application name (`app.name`) followed by the colon as a prefix to the study id. e.g. `cbioportal:brca_tcga_pub`

#### Groups

Keycloak allows you to create Groups for easy mapping of multiple
studies to multiple users. One can, for example, make a Keycloak group
with name `PUBLIC_STUDIES` and add all the individual Keycloak roles
corresponding to public studies to this group. It is also possible to
configure a group to be "default" in Keycloak, meaning new users are
automatically added to this group when logging in for the first time.

Alternatively, the Keycloak roles can correspond to the **groups** specified
in the [metadata files of studies](/File-Formats.md#cancer-study) instead
of corresponding to individual **study identifiers**. Although this will
result in less roles that need to be added and maintained in Keycloak,
it does result in group configuration being spread over both Keycloak
and meta study files.

### Assign roles to users

Next, assign roles to users. Head to **Users** in the left sidebar,
find a user (users from external providers should be known to Keycloak
after they have logged in for the first time), click the ID and
open the **Role Mappings** tab for that user. Select the
_cbioportal_ client in the dropdown under **Client Roles**, and use
the **Available Roles** selection and its **Add selected** button to
assign client roles to this user.

![](/images/previews/assign-role-to-user.png)

To automatically assign roles to all users when Keycloak first sees
them, the **Roles** pane accessed from the left sidebar has a
**Default Roles** tab. The interface for assigning roles here is much
the same as the one for assigning roles to individual users.

### Doing a Test Run

Rebuild the WAR file and follow the [Deployment with authentication
steps](/deployment/deploy-without-docker/Deploying.md#required-login) using `authenticate=saml`.

Then, go to:  [http://localhost:8081/](http://localhost:8081/).

If all goes well, the following should happen:

* You will be redirected to the Keycloak Login Page.
* After authenticating, you will be redirected back to your local instance of cBioPortal.

If this does not happen, see the Troubleshooting Tips  below.

### Add client for OAuth2 token-based data access

With cBioPortal instances that require user authentication the API can be queried when including a data access token in the request header (see [Authenticating Users via Tokens](/deployment/authorization-and-authentication/Authenticating-Users-via-Tokens.md)). KeyCloak can be configured as an OAuth2 authentication provider that distributes data access tokens to users and validates these tokens when used while querying the API. This feature is enabled by creating a `cbioportal_api` OpenID Connect client that has access to the user roles defined in the `cbioportal` SAML client.

The step below were verified to work with Keycloak versions 4.8.3.Final and 8.0.2.

1. Create a client with name `cbioportal_api`. Set _Client Protocol_ to `openid-connect`.

![](/images/previews/oauth2_client_1.png)

2. On the configuration page of `cbioportal_api` client apply the following settings:

#### Settings tab

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| Access Type      | confidential |  |
| Standard Flow Enabled      | ON      |   (default value) |
| Enabled      | ON      |   (default value) |
| Consent Required      | OFF        |   (default value) |
| Client Protocol      | openid-connect   |   (default value) |
| Access Type | confidential      |     (default value) |
| Standard Flow Enabled | ON      |     (default value) |
| Implicit Flow Enabled | OFF      |     (default value) |
| Direct Access Grants Enabled | OFF      |    |
| Service Accounts Enabled | OFF      |  (default value)   |
| Authorization Enabled | OFF      |  (default value)  |
| Valid Redirect URIs | _url_/api/data-access-token/oauth2  |  _url_ refers to base url of cBioPortal instance |

![](/images/previews/oauth2_client_3.png)

#### Credentials tab

Select `Client Id and Secret`. Take notice of the value of _Secret_ the secret field. This secret should be added to `application.properties` file of the cBioPortal backend.

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| Client Authenticator     | Client Id and Secret |   (default value) |

![](/images/previews/oauth2_client_6.png)

#### Client Scopes tab

Keep only scopes `roles` and `offline_access` (remove all others).

![](/images/previews/oauth2_client_2.png)

#### Mapper tab

Create a new _Audience_ mapper with name `cbioportal_api_audience`. This value will be used by the cBioPortal backend during validation of access tokens.

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| Name       | cbioportal_api_audience |  |
| Mapper Type     | Audience      |   |
| Included Client Audience      | cbioportal_api      |    |
| Add to ID token      | OFF        |  (default value)  |
| Add to access token  | ON      |   (default value)  |

![](/images/previews/oauth2_client_4.png)

#### Scope tab

Enable _Full Scope_. This setting will include the user roles defined in the `cbioportal` SAML client in access tokens distributed by the `cbioportal_api` client.

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| Full Scope Allowed       | ON | (default value) |

![](/images/previews/oauth2_client_5.png)

3. Add these parameters to `application.properties` of the cBioPortal backend.

| parameter        | value  | comment  |
| ------------- |:-------------:| -----:|
| dat.method       | oauth2 |  |
| dat.oauth2.clientId       | cbioportal_api |  |
| dat.oauth2.clientSecret    | ?      | see _Secret_ field in the _Credentials_ tab  |
| dat.oauth2.accessTokenUri     | _keycloak_url_/auth/realms/cbioportal/protocol/openid-connect/token      |   _keycloak_url_ refers to URL of the KeyCloak server from perspective of the cBioPortal instance |
| dat.oauth2.jwkUrl  | _keycloak_url_/auth/realms/cbioportal/protocol/openid-connect/certs      |   _keycloak_url_ refers to URL of the KeyCloak server from perspective of the cBioPortal instance |
| dat.oauth2.issuer      | _keycloak_url_/auth/realms/cbioportal        |  _keycloak_url_ refers to URL of the KeyCloak server from perspective of the browser |
| dat.oauth2.userAuthorizationUri  | _keycloak_url_/auth/realms/cbioportal/protocol/openid-connect/auth      |  _keycloak_url_ refers to URL of the KeyCloak server from perspective of the browser |
| dat.oauth2.redirectUri  | _cbioportal_url_/api/data-access-token/oauth2 | _cbioportal_url_ is url up to _/api_ path |
| dat.oauth2.jwtRolesPath  | '::'-separated path to array with user roles in JWT token returned by Keycloak | example: _resource_access::cbioportal::roles_ |

More information on configuration of the cBioPortal backend can be found in [Authenticating Users via Tokens](/deployment/authorization-and-authentication/Authenticating-Users-via-Tokens.md).

### Troubleshooting

#### Logging

Getting this to work requires many steps, and can be a bit tricky.  If you get stuck or get an obscure error message, your best bet is to turn on all DEBUG logging.
This can be done via `src/main/resources/logback.xml`. See [logback.DEBUG.EXAMPLE.xml](./logback.DEBUG.EXAMPLE.xml) file for an example of how to configure debug levels for cbioportal.

Then, rebuild the WAR, redeploy, and try to authenticate again.  Your log file will then include hundreds of SAML-specific messages, even the full XML of each SAML message, and this should help you debug the error.

If you're using the Docker container, mount the file instead with `-v ./logback.xml:/cbioportal-webapp/WEB-INF/classes/logback.xml`.

#### Determining jwtRolesPath for OAuth2 Token
By default user-roles are extracted from path `resource_access::cbioportal::roles` in the JWT json. Changes to the configuration of roles at the realm and client level in Keycloak instance can alter this path and must be set acordingly with the `dat.oauth2.jwtRolesPath` property in the `application.properties` file. 

To check the the roles path, go into the `Client Scopes` tab inside KeyCloak. Enter the `Evaluate` section, select a test user, and click `Evaluate`. In the section below, select the `Generated Access Token` tab to examine the JWT structure. 

![](/images/previews/oauth2_client_7.png)

A sample JWT might look like this:
```
{
  "exp": 1234567891,
  "iat": 1234567892,
  "jti": "transient-id",
  "iss": "issuer",
  "sub": "subject",
  "typ": "Bearer",
  "session_state": "sessionstate",
  "acr": "1",
  "realm_access": {
    "roles": [
      "all"
    ]
  },
  "scope": "openid"
}
```
The `jwtRolesPath` in this case would be `realm_access::roles`. Double check this against the `jwtRolesPath` value set in `application.properties`.

To connect cBioPortal to an external user database such as Active Directory will require the installation of [Keycloak](http://www.keycloak.org).  Please read the Wiki page on [Authenticating and Authorizing Users via Keycloak](Authenticating-and-Authorizing-Users-via-keycloak.md) for information on how to connect the cBioPortal with Keycloak.  You can also read how to connected Keycloak to Active Directory via LDAP on the [User Storage Federation](https://www.keycloak.org/docs/latest/server_admin/#_user-storage-federation) webpage of the Keycloak website.  
# Introduction

The cBioPortal includes support for SAML (Security Assertion Markup Language).  This document explains why you might find SAML useful, and how to configure SAML within your own instance of cBioPortal.

Please note that configuring your local instance to support SAML requires many steps.  This includes configuration changes and a small amount of debugging.  If you follow the steps below, you should be up and running relatively quickly, but be forewarned that you may have do a few trial runs to get everything working.

In the documentation below, we also provide details on how to perform SAML authentication via a commercial company:  [OneLogin](https://www.onelogin.com/).  OneLogin provides a free tier for testing out SAML authentication, and is one of the easier options to get a complete SAML workflow set-up.  Once you have OneLogin working, you should then have enough information to transition to your final authentication service.

## What is SAML?

SAML is an open standard that enables one to more easily add an authentication service on top of any existing web application.  For the full definition, see the [SAML Wikipedia entry](https://en.wikipedia.org/wiki/Security_Assertion_Markup_Language).

In its simplest terms, SAML boils down to four terms:

* **identity provider**:  this is a web-based service that stores user names and passwords, and provides a login form for users to authenticate.  Ideally, it also provides easy methods to add / edit / delete users, and also provides methods for users to reset their password.  In the documentation below, OneLogin.com serves as the identity provider.

* **service provider**:  any web site or web application that provides a service, but should only be available to authenticated and authorized users.  In the documentation below, the cBioPortal is the service provider.

* **authentication**:  a means of verifying that a user is who they purport to be.  Authentication is performed by the identify provider, by extracting the user name and password provided in a login form, and matching this with information stored in a database. When authentication is enabled, multiple cancer studies can be stored within a single instance of cBioPortal while providing fine-grained control over which users can access which studies.  Authorization is implemented within the core cBioPortal code, and *not* the identify provider.

## Why is SAML Relevant to cBioPortal?

The cBioPortal code has no means of storing user name and passwords and no means of directly authenticating users.  If you want to restrict access to your instance of cBioPortal, you therefore have to consider an external authentication service.  SAML is one means of doing so, and your larger institution may already provide SAML support.  For example, at Sloan Kettering and Dana-Farber, users of the internal cBioPortal instances login with their regular credentials via SAML.  This greatly simplifies user management.

# Setting up an Identity Provider

As noted above, we provide details on how to perform SAML authentication via a commercial company:  [OneLogin](https://www.onelogin.com/).  
If you already have an IDP set up, you can skip this part and go to [Configuring SAML within cBioPortal](#configuring-saml-within-cbioportal).

OneLogin provides a free tier for testing out SAML authentication, and is one of the easier options to get a complete SAML workflow set-up.  Once you have OneLogin working, you should then have enough information to transition to your final authentication service.  As you follow the steps below, the following link may be helpful: [How to Use the OneLogin SAML Test Connector](https://support.onelogin.com/hc/en-us/articles/202673944-How-to-Use-the-OneLogin-SAML-Test-Connector).

To get started:

* [Register a new OneLogin.com Account](https://www.onelogin.com/signup?ref=floating-sidebar)

## Setting up a SAML Test Connector

* [Login to OneLogin.com](https://app.onelogin.com/login).
* Under Apps, Select Add Apps.
* Search for SAML.
* Select the option labeled: OneLogin SAML Test (IdP w/attr).
* "SAVE" the app, then select the Configuration Tab.

* Under the Configuration Tab for OneLogin SAML Test (IdP w/attr), paste the following fields (this is assuming you are testing everything via localhost).

    * Audience: cbioportal
    * Recipient: http://localhost:8080/saml/SSO
    * ACS (Consumer) URL Validator*:  ^http:\/\/localhost\:8080\/cbioportal\/saml\/SSO$
    * ACS (Consumer) URL*:  http://localhost:8080/saml/SSO

![](/images/previews/onelogin-config.png)


* Add at least the parameters: 
    * Email (Attribute)
    * Email (SAML NameID)

![](/images/previews/onelogin-config-parameters.png)

* Find your user in the "Users" menu

![](/images/previews/onelogin-users-search.png)

* Link the SAML app to your user (click "New app" on the **+** icon found on the top right of the "Applications" table to do this - see screenshot below): 

![](/images/previews/onelogin-add-app.png)

* Configure these **email** parameters in the Users menu:

![](/images/previews/onelogin-users.png)
 


## Downloading the SAML Test Connector Meta Data

* Go to the SSO Tab within OneLogin SAML Test (IdP), find the field labeled:  Issuer URL.  Copy this URL and download it's contents.  This is an XML file that describes the identity provider.

![](https://s3.amazonaws.com/f.cl.ly/items/1R0v0L3P1U2E23202V1d/Image%202015-05-24%20at%2010.07.56%20PM.png)

then, move this XML file to:

    portal/src/main/resources/

You should now be all set with OneLogin.com.  Next, you need to configure your instance of cBioPortal.

# Configuring SAML within cBioPortal

## Creating a KeyStore

In order to use SAML, you must create a [Java Keystore](https://docs.oracle.com/javase/7/docs/api/java/security/KeyStore.html).  

This can be done via the Java `keytool` command, which is bundled with Java.

Type the following:

    keytool -genkey -alias secure-key -keyalg RSA -keystore samlKeystore.jks

This will create a Java keystore for a key called:  `secure-key` and place the keystore in a file named `samlKeystore.jks`.  You will be prompted for:

* keystore password (required, for example:  apollo1)
* your name, organization and location (optional)
* key password for `secure-key` (required, for example apollo2)

When you are done, copy `samlKeystore.jsk` to the correct location:
    
    mv samlKeystore.jks portal/src/main/resources/

If you need to export the public certificate associated within your keystore, run:

    keytool -export -keystore samlKeystore.jks -alias secure-key -file cBioPortal.cer

##### HTTPS and Tomcat

:warning: If you already have an official (non-self-signed) SSL certificate, and need to get your site 
running on HTTPS directly from Tomcat, then you need to import your certificate into the keystore instead. 
See [this Tomcat documentation page](https://tomcat.apache.org/tomcat-8.0-doc/ssl-howto.html) for more details.

:warning: An extra warning for when configuring HTTPS for Tomcat: use the same password for 
both keystore and secure-key. This seems to be an extra restriction by Tomcat.


## Modifying configuration

Within application.properties, make sure that:

    app.name=cbioportal

Then, modify the section labeled `authentication`. See SAML parameters shown in example below:

    saml.sp.metadata.entityid=cbioportal
    saml.sp.metadata.wantassertionsigned=true
    saml.idp.metadata.location=classpath:/onelogin_metadata_620035.xml
    saml.idp.metadata.entityid=https://app.onelogin.com/saml/metadata/620035
    saml.keystore.location=classpath:/samlKeystore.jks
    saml.keystore.password=apollo1
    saml.keystore.private-key.key=secure-key
    saml.keystore.private-key.password=apollo2
    saml.keystore.default-key=secure-key
    saml.idp.comm.binding.settings=defaultBinding
    saml.idp.comm.binding.type=
    saml.idp.metadata.attribute.email=User.email
    saml.idp.metadata.attribute.userName=User.name
    saml.custom.userservice.class=org.cbioportal.security.spring.authentication.saml.SAMLUserDetailsServiceImpl
    # global logout (as opposed to local logout):
    saml.logout.local=false
    saml.logout.url=/

Please note that you will have to modify all the above to match your own settings. `saml.idp.comm.binding.type` can be left empty if `saml.idp.comm.binding.settings=defaultBinding`. The `saml.logout.*` settings above reflect the settings of an IDP that supports Single Logout (hopefully the default in most cases - more details in section below).

In the case that you are running cBioPortal behind a reverse proxy that handles the SSL certificates (such as nginx or traefik), you will have to also specify `saml.sp.metadata.entitybaseurl`. This should point to `https://host.example.come:443`. This setting is required such that cBioPortal uses the Spring SAML library appropriately for creating redirects back into cBioPortal.

In addition there is a known bug where redirect from the cBioPortal instance always goes over http instead of https (https://github.com/cBioPortal/cbioportal/issues/6342). To get around this issue you can pass the full URL including https to the `webapp-runnner.jar` command with e.g. `--proxy-base-url https://mycbioportalinstance.org`.

### Custom scenarios

:information_source: Some settings may need to be adjusted to non-default values, depending on your IDP. For example, if your 
IDP required HTTP-GET requests instead of HTTP-POST, you need to set these properties as such:
 
    saml.idp.comm.binding.settings=specificBinding
    saml.idp.comm.binding.type=bindings:HTTP-Redirect

If you need a very different parsing of the SAML tokens than what is done at `org.cbioportal.security.spring.authentication.saml.SAMLUserDetailsServiceImpl`, you can point the `saml.custom.userservice.class` to your own implementation: 

    saml.custom.userservice.class=<your_package.your_class_name>

:warning: The properties `saml.idp.metadata.attribute.email`, and `saml.idp.metadata.attribute.userName` can also vary per IDP. It is important to set these correctly since these are a required field by the cBioPortal SAML parser (that is, if `org.cbioportal.security.spring.authentication.saml.SAMLUserDetailsServiceImpl` is chosen for property `saml.custom.userservice.class`). 

:warning: Some IDPs like to provide their own logout page (e.g. when they don't support the custom SAML Single Logout protocol). For this you can adjust the  
`saml.logout.url` property to a custom URL provided by the IDP. Also set the `saml.logout.local=true` property in this case to indicate that global logout (or Single Logout) is not supported by IDP:

    # local logout followed by a redirect to a global logout page:
    saml.logout.local=true
    saml.logout.url=<idp specific logout URL, e.g. https://idp.logoutpage.com >
    

:warning: Some IDPs (e.g. Azure Active Directory) cache user data for more than 2 hours causing cbioportal to complain that the authentication statement is too old to be used. You can fix this problem by setting `forceAuthN` to true. Below is an example how you can do this with the properties. You can choose any binding type you like. `bindings:HTTP-Redirect` is given just as an example.

```
    saml.idp.comm.binding.settings=specificBinding
    saml.idp.comm.binding.type=bindings:HTTP-Redirect
    saml.idp.comm.binding.force-auth-n=true
```

## More customizations

If your IDP does not have the flexibility of sending the specific credential fields expected by our 
default "user details parsers" implementation (i.e. `security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/saml/SAMLUserDetailsServiceImpl.java` 
expects field `mail` to be present in the SAML credential), then please let us know via a [new 
issue at our issue tracking system](https://github.com/cBioPortal/cbioportal/issues/new), so we can 
evaluate whether this is a scenario we would like to support in the default code. You can also consider 
adding your own version of the `SAMLUserDetailsService` class. 

## Authorizing Users

Next, please read the Wiki page on [User Authorization](User-Authorization.md), and add user rights for a single user.


## Configuring the Login.html Page (not applicable to most external IDPs)

The login page is configurable via the `application.properties` properties `skin.authorization_message` and `skin.login.saml.registration_htm`. 
For example in `skin.authorization_message` you can be set to something like this:

```
skin.authorization_message= Welcome to this portal. Access to this portal is available to authorized test users at YOUR ORG.  [<a href="https://thehyve.nl/">Request Access</a>].
```

and `skin.login.saml.registration_htm` can be set to:

```
skin.login.saml.registration_htm=Sign in via XXXX
```

You can also set a standard text in `skin.login.contact_html` that will appear in case of problems: 

```
skin.login.contact_html=If you think you have received this message in error, please contact us at <a style="color:#FF0000" href="mailto:cbioportal-access@your.org">cbioportal-access@your.org</a>
```


## Doing a Test Run

You are now ready to go.

Rebuild the WAR file and follow the [Deployment with authentication
steps](/deployment/deploy-without-docker/Deploying.md#required-login) using `authenticate=saml`.

Then, go to:  [http://localhost:8080/](http://localhost:8080/).

If all goes well, the following should happen:

* You will be redirected to the OneLogin Login Page.
* After authenticating, you will be redirected back to your local instance of cBioPortal.

If this does not happen, see the Troubleshooting Tips  below.

## Troubleshooting Tips 

### Logging

Getting this to work requires many steps, and can be a bit tricky.  If you get stuck or get an obscure error message,
your best bet is to turn on all DEBUG logging.  This can be done via `src/main/resources/logback.xml`.  For example:

```
<root level="debug">
    <appender-ref ref="STDOUT" />
    <appender-ref ref="FILE" />
</root>

<logger name="org.mskcc" level="debug">
    <appender-ref ref="STDOUT" />
    <appender-ref ref="FILE" />
</logger>

<logger name="org.cbioportal.security" level="debug">
    <appender-ref ref="STDOUT" />
    <appender-ref ref="FILE" />
</logger>
```

Then, rebuild the WAR, redeploy, and try to authenticate again.  Your log file will then include hundreds of SAML-specific messages, even the full XML of each SAML message, and this should help you debug the error.

### Seeing the SAML messages

Another tool we can use to troubleshoot is SAML tracer (https://addons.mozilla.org/en-US/firefox/addon/saml-tracer/ ). You can add this to Firefox and it will give you an extra menu item in "Tools". Go through the loging steps and you will see the SAML messages that are sent by the IDP. 

### Obtaining the Service Provider Meta Data File

By default, the portal will automatically generate a Service Provider (SP) Meta Data File.  You may need to provide this file to your Identity Provider (IP).

You can access the Service Provider Meta Data File via a URL such as:

[http://localhost:8080/saml/metadata](http://localhost:8080/saml/metadata)

# Authenticating Users via Tokens

The cBioPortal provides access to data within its database through the [cBioPortal web service](https://www.cbioportal.org/api/swagger-ui.html). While most users interact with this web service indirectly while visiting cBioPortal web pages, it is possible to access this web service directly. Direct access to the web service is appealing for those users with an interest in mining or analyzing cBioPortal data beyond the confines of a web browser, such as from ‚ÄòR‚Äô or Python clients.

If your instance of the cBioPortal will not require [user authorization](User-Authorization.md) (login), direct access to the web service is available without any additional server configuration or client side prerequisites. However, if your instance of the cBioPortal will require login, additional data access token server configuration is necessary before direct access to the web service is available to your users. Additionally, prior to web service access, clients will need to obtain a data access token and present this token with each web service request.

This document contains information for both cBioPortal administrators and users. For administrators, this document describes how to configure an instance of the cBioPortal with data access token support. For users, this document provides details about how to obtain a data access token and present it with a web service request (see [Using Data Access Tokens](Authenticating-Users-via-Tokens.md#using-data-access-tokens)).

## Configuring Data Access Tokens within cBioPortal

### Implementations

The cBioPortal codebase comes with three different data access token implementations, each of which come with strengths and weaknesses.

The first implementation based on [JSON Web Token](https://jwt.io/introduction/) (JWT) technology, is appealing because the data access token contains all the information required to validate the token during a web service request. When a data access token is generated by the cBioPortal, its expiration date is inserted into the token. The data access token is then signed and encrypted before presenting it to the user. When the data access token is presented in a subsequent web service request, the cBioPortal is able to inspect the token and determine its validity before satisfying the request. The downside to this implementation is that since the data access token is self-validating, no information about the token is stored in the database. This means there is no way to revoke a data access token in the event that it has been compromised or limit the number of data access tokens per user.

The second implementation based on [Universally unique identifier](https://en.wikipedia.org/wiki/Universally\_unique\_identifier) (UUID) technology, allows for greater control over the data access token at the cost of additional database support. Unlike JWT based data access tokens which are self-validating, a UUID based token is a string representation of a unique number. When a data access token is generated by the cBioPortal, a record is inserted into the `data_access_tokens` database table:

```
mysql> describe data_access_tokens;
+------------+--------------+------+-----+---------+-------+
| Field      | Type         | Null | Key | Default | Extra |
+------------+--------------+------+-----+---------+-------+
| TOKEN      | varchar(50)  | NO   | PRI | NULL    |       |
| USERNAME   | varchar(255) | NO   | MUL | NULL    |       |
| EXPIRATION | datetime     | NO   |     | NULL    |       |
| CREATION   | datetime     | YES  |     | NULL    |       |
+------------+--------------+------+-----+---------+-------+
4 rows in set (0.00 sec)
```

In addition to storing the token (the unique number), its creation and expiration dates along with the associated username is also stored prior to presenting it to the user. When the data access token is presented in a subsequent web service request, the cBioPortal compares the token presented in the request against this table for validity. If the token exists in the table and has not yet expired, the request will be satisfied. The creation date is used by the cBioPortal when a user requests a new data access token. Based on the server configuration, if the number of outstanding data access tokens together with this new one exceeds the number of allowable data access tokens per user, the cBioPortal may refuse to present a new one or it may invalidate the oldest one before presenting the new one to the user.

A third implementation makes use of an external OAuth2-authentication provider and can integrate with single-sign on (SSO) solutions. The OAuth2 provider provides users with an **offline token** that does not allow for direct access to protected cBioPortal resources. Instead, the offline token is passed to the OAuth2 authorization server that returns an **access token**, a short-lived permission to access cBioPortal resources. The access token contains up-to-date user permissions (the studies that the user is permitted to view). This implementation ensures that any changes of user permissions registered at the authentication provider are effective immediately and do not rely on the expiration date of the token of the user. At any moment OAuth2 offline tokens can be revoked by the Authentication provider on a per-user basis.

A step-by-step guide to configure KeyCloak to provide OAuth2 client functionality can be found in [Authenticating and Authorizing Users via keycloak](Authenticating-and-Authorizing-Users-via-keycloak.md#add-client-for-oauth2-token-based-data-access).

### Modifying Configuration

The following properties must be present in application.properties in order to allow direct access to the cBioPortal web service when login is required:

**Property**: dat.method (required)

* **Description**: The desired data access token implementation.
* **Permissible Values**: jwt, uuid, oauth2, none
* **Default Value**: none

**Property**: dat.filter\_user\_role (optional)

* **Description**: This property determines users access in token generation. If present, this role will be checked in user roles before generating a token.
* **Permissible Values**: A string value.
* **Default Value**: none

**Property**: dat.unauth\_users (optional, not used for dat.method = oauth2)

* **Description**: A list of users that should not be allowed to download a data access token.
* **Permissible Values**: A comma-delimited list of valid user names as found in the users table.
* **Default Value**: anonymousUser

**Property**: dat.ttl\_seconds (optional, not used for dat.method = oauth2)

* **Description**: The time in seconds between token creation and token expiration.
* **Permissible Values**: An integer value greater than zero.
* **Default Value**: 2592000 (30 days)

**Property**: dat.jwt.secret\_key (required only when dat.method = jwt)

* **Description**: The signature used to sign JWT based data access tokens during creation. This signature is used in subsequent web service requests to validate the request. This property is used only when dat.method = jwt. Keys used with the HS256 algorithm MUST have a size >= 256 bits. See https://tools.ietf.org/html/rfc7518#section-3.2 for more information.
* **Permissible Values**: A string value.
* **Default Value**: none

**Property**: dat.uuid.max\_number\_per\_user (required only when dat.method = uuid)

* **Description**: This property determines the number of data access tokens that can be allocated to a single user at any given time. If a user requests a data access token that puts them over this limit, the cBioPortal will present a new token to the user and it may invalidate the oldest one before presenting the new one to the user. This property is used only when dat.method = uuid.
* **Permissible Values**: An integer value greater than zero.
* **Default Value**: 1

**Property**: dat.oauth2.clientId (required only when dat.method = oauth2)

* **Description**: Identifier of the OAuth2 client of the authentication provider.
* **Permissible Values**: string
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.clientSecret (required only when dat.method = oauth2)

* **Description**: Secret of the OAuth2 client of the authentication provider used for back channel communication between cBioPortal and the provider. This value is specified by the authentication provider.
* **Permissible Values**: string
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.issuer (required only when dat.method = oauth2)

* **Description**: Name of the issuer of the tokens. It is used to assert the correct origin of the token. This value is specified by the authentication provider.
* **Permissible Values**: string
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.userAuthorizationUri (required only when dat.method = oauth2)

* **Description**: Url of the user login page of the OAuth2 authorization provider. According to OAuth2 specification this endpoint must end with `/auth`.
* **Permissible Values**: url
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.accessTokenUri (required only when dat.method = oauth2)

* **Description**: Url for retrieval of tokens from the OAuth2 authorization provider. According to OAuth2 specification this endpoint must end with `/token`.
* **Permissible Values**: url
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.jwkUrl (required only when dat.method = oauth2)

* **Description**: Url for retrieval of the public key used by the OAuth2 authorization provider to sign tokens. According to OAuth2 specification this endpoint must end with `/certs`.
* **Permissible Values**: url
* **Default Value**: This value is specified by the authentication provider.

**Property**: dat.oauth2.jwtRolesPath (required only when dat.method = oauth2)

* **Description**: `::`-separated path to array with user roles in JWT token returned by Keycloak.
* **Permissible Values**: `::`-separated strings
* **Default Value**: resource\_access::cbioportal::roles

## Using Data Access Tokens

To directly access the web service of a cBioPortal instance which requires user login, clients will need to obtain a data access token and present this token with each web service request. To obtain a data access token, login to the cBioPortal and then choose ‚ÄúData Access Token‚Äù from the ‚ÄúLogged in as‚Ä¶‚Äù dropdown menu located on the top-right section of the webpage (**note this menu option will not be present if the instance of the cBioPortal has not be configured to support data access tokens**):

![](/images/previews/dat-download-menu.png)

This will bring you to a section of the Web API page where you can download your token using the "Download Token" button.

![](/images/previews/download-dat-button.png)

The downloaded data access token file will contain the token, and, for `uuid` and `jwt` token methods, a creation and expiration date:

```
token: 63efa81c-2490-4e15-9d1c-fb6e8e50e35d
creation_date: 2018-11-28T20:23:55.241Z
expiration_date: 2019-05-27T20:23:55.241Z
```

To directly access the web service, include the token as an **Authorization** header to the web service request:

```
Authorization: Bearer 63efa81c-2490-4e15-9d1c-fb6e8e50e35d
```

Here is an example curl command showing a request to GET /api/studies with the proper **Authorization** header set:

```
curl -X GET "https://www.cbioportal.org/api/studies" -H "accept: application/json" -H "Authorization: Bearer 63efa81c-2490-4e15-9d1c-fb6e8e50e35d"
```

If a web service request is made without an **Authorization** header or if the token within the header is not valid, the web service will not satisfy the request and will return with an HTTP Status 401. Additionally, there are a handful of properties which determine the behavior of cBioPortal data access tokens (see [Configuring Data Access Tokens within cBioPortal](Authenticating-Users-via-Tokens.md#configuring-data-access-tokens-within-cbioportal)). For example, a data access token may have an expiration date of 30 days after its creation date or 300 days. Similarly, there may be a limit to the number of data access tokens a user may have at any one time. If this limit is reached a new data access token may or may not be downloadable. Please check with the local cBioPortal system administrator for more information about how data access token behavior has been configured.
# User Authorization

This step is only required if you intend on running an instance of the portal that supports user authorization.

Two tables need to be populated in order to support user authorization.

## Table:  users

This table contains all the users that have authorized access to the instance of the portal.  The table requires a user's email address, name, and integer flag indicating if the account is enabled.

```
mysql> describe users;
+---------+--------------+------+-----+---------+-------+
| Field   | Type         | Null | Key | Default | Extra |
+---------+--------------+------+-----+---------+-------+
| EMAIL   | varchar(128) | NO   | PRI | NULL    |       |
| NAME    | varchar(255) | NO   |     | NULL    |       |
| ENABLED | tinyint(1)   | NO   |     | NULL    |       |
+---------+--------------+------+-----+---------+-------+
3 rows in set (0.00 sec)
```

An example entry would be:

```
mysql> select * from users where email = "john.smith@gmail.com";
+--------------------------+----------------+---------+
| EMAIL                    | NAME           | ENABLED |
+--------------------------+----------------+---------+
| john.smith@gmail.com     | John Smith     |       1 | 
+--------------------------+----------------+---------+
1 row in set (0.00 sec)
```

Note, if the ENABLED value is set to 0, the user will be able to login to the portal, but will see no studies.

You need to add users via MySQL directly.  For example:

```
INSERT INTO cbioportal.users (EMAIL, NAME, ENABLED) 
    VALUES ('john.smith@gmail.com', 'John Smith', 1);
```

## Table:  authorities

This table contains the list of cancer studies that each user is authorized to view.  The table requires a user email address and an authority (e.g. cancer study) granted to the user.

```
mysql> describe authorities;
+-----------+--------------+------+-----+---------+-------+
| Field     | Type         | Null | Key | Default | Extra |
+-----------+--------------+------+-----+---------+-------+
| EMAIL     | varchar(128) | NO   |     | NULL    |       | 
| AUTHORITY | varchar(50)  | NO   |     | NULL    |       | 
+-----------+--------------+------+-----+---------+-------+
2 rows in set (0.00 sec)
```

Some example entries would be:

```
mysql> select * from authorities where email = "john.smith@gmail.com";
+--------------------------+---------------------------+
| EMAIL                    | AUTHORITY                 |
+--------------------------+---------------------------+
| john.smith@gmail.com     | cbioportal:CANCER_STUDY_1 | 
| john.smith@gmail.com     | cbioportal:CANCER_STUDY_2 | 
| john.smith@gmail.com     | cbioportal:CANCER_STUDY_3 | 
+--------------------------+---------------------------+
5 rows in set (0.00 sec)
```
The value in the EMAIL column should be the same email address contained in the USER table.  

The value in the AUTHORITY column is made of two parts:

* The first part is the name of your portal instance.  This name should also match the `app.name` property found in the `application.properties` file.  
* Following a colon delimiter, the second part is the [cancer_study_identifier](../../File-Formats.md#cancer-study) of the cancer study this user has rights to access. 

**If the user has rights to all available cancer studies, a single entry with the keyword `app.name:` + "ALL" is sufficient (so e.g. "cbioportal:ALL").**

You need to add users via MySQL directly.  For example:

```
mysql> INSERT INTO cbioportal.authorities (EMAIL, AUTHORITY) VALUES
    ('john.smith@gmail.com', 'cbioportal:CANCER_STUDY_1');
```
**Important Note:**  The cancer study identifier is *not* CASE sEnsitive. So it can be UPPER CASE, or just how it is stored in the `cancer_study` table.
Changes to these tables become effective the *next* time the user logs in. 

## Using groups

It is also possible to define groups and assign multiple studies and users to a group. You can add a group name to the `cancer_study` table `GROUPS` column. This same group name can be used in the `AUTHORITY` column of the `authorities` table mentioned above. 

### Example: 

We want to create the group "TEST_GROUP1" and assign two existing studies to it and give our user 'john.smith@gmail.com' access to this group of studies. Steps:

1- Find your studies in table `cancer_study`
```
mysql> select CANCER_STUDY_ID, CANCER_STUDY_IDENTIFIER,GROUPS from cancer_study where CANCER_STUDY_ID in (93,94);
+-----------------+-------------------------+-------------+
| CANCER_STUDY_ID | CANCER_STUDY_IDENTIFIER | GROUPS      |
+-----------------+-------------------------+-------------+
|              93 | acc_tcga                | GROUPB      |
|              94 | brca_tcga               |             |
+-----------------+-------------------------+-------------+
```
2- Update the `GROUPS` field, adding your "TEST_GROUP1" to it. :warning: This is a `;` separated column, so if you want a study to be part of multiple groups, separate them with `;`.  
```
mysql> update cancer_study set GROUPS='TEST_GROUP1' where CANCER_STUDY_ID = 94;
```

If `GROUPS` already has a value (like for study 93 in example above) then add ";TEST_GROUP1" to ensure existing groups are not ovewritten.
```
mysql> update cancer_study set GROUPS=concat(GROUPS,';TEST_GROUP1') where CANCER_STUDY_ID = 93;
```
3- Check the result:
```
mysql> select CANCER_STUDY_ID, CANCER_STUDY_IDENTIFIER,GROUPS from cancer_study where CANCER_STUDY_ID in (93,94);
+-----------------+-------------------------+--------------------+
| CANCER_STUDY_ID | CANCER_STUDY_IDENTIFIER | GROUPS             |
+-----------------+-------------------------+--------------------+
|              93 | acc_tcga                | GROUPB;TEST_GROUP1 |
|              94 | brca_tcga               | TEST_GROUP1        |
+-----------------+-------------------------+--------------------+
```
4- Add the group to user 'john.smith@gmail.com', using `app.name:` + "TEST_GROUP1" like so:
```
mysql> INSERT INTO cbioportal.authorities (EMAIL, AUTHORITY) VALUES
    ('john.smith@gmail.com', 'cbioportal:TEST_GROUP1');
```

After **next login**, the user 'john.smith@gmail.com' will have access to these two studies. 


## Configuring PUBLIC studies

To enable a set of public studies that should be visible to all users, without the need to configure this for each user in the `authorities` and `users` tables, you can set the property `always_show_study_group` in **application.properties** file. For example, you can set:

```
always_show_study_group=PUBLIC
```

This will enable the word "PUBLIC" to be used in the column `GROUPS` of the table `cancer_study` to indicate which studies should be always shown to *any authenticated user*, regardless of authorization configurations.

### Example:

To reuse the example table above, let's assume the property `always_show_study_group` is set as indicated above and the `cancer_study` table contents are set to the following: 

```
+-----------------+-------------------------+--------------------+
| CANCER_STUDY_ID | CANCER_STUDY_IDENTIFIER | GROUPS             |
+-----------------+-------------------------+--------------------+
|              93 | acc_tcga                | GROUPB;TEST_GROUP1 |
|              94 | brca_tcga               | TEST_GROUP1;PUBLIC |
+-----------------+-------------------------+--------------------+
```

In this case, the study `brca_tcga` will be visible to *any authenticated user* while the study `acc_tcga` will be visible only to users configured to be part of `GROUPB` or `TEST_GROUP1`
# More application.properties Settings

This page describes the main properties within application.properties.

## Database Settings

```
db.user=
db.password=
db.connection_string=
db.driver=[this is the name of your JDBC driver, e.g., com.mysql.jdbc.Driver]
```

The format of the `db.connection_string` is:
```
jdbc:mysql://<host>:<port>/<database name>?<parameter1>&<parameter2>&<parameter...>
```

For example:

```
jdbc:mysql://localhost:3306/cbiodb?zeroDateTimeBehavior=convertToNull&useSSL=false
```

:warning: The fields `db.host` and `db.portal_db_name` and `db.use_ssl` are deprecated. It is required to configure the database connection using
the `db.connection_string` instead.

`db.tomcat_resource_name` is required in order to work with the tomcat database connection pool and should have the default value jdbc/cbioportal in order to work correctly with the your WAR file.

```
db.tomcat_resource_name=jdbc/cbioportal
```

## cBioPortal Customization

### Hide tabs (pages)

Settings controlling which tabs (pages) to hide. Set them to `false` if you want to hide those tabs, otherwise set the properties to `true`.

```
skin.show_data_tab=
skin.show_web_api_tab=
skin.show_r_matlab_tab=
skin.show_tutorials_tab=
skin.show_faqs_tab=
skin.show_news_tab=
skin.show_tools_tab=
skin.show_about_tab=
```

**Note:** `skin.show_tools_tab` refers to the `Visualize Your Data` tab, while `skin.show_data_tab` refers to the `Data Sets` tab.

### Cross Cancer Study Query Default

The cross cancer study query default is a list of studies used when querying one or more genes and not specifying a specific study or list of studies. There are two ways in which the default cross cancer study list is used:

1. When using the linkout links without a study e.g. `/ln?q=TP53:MUT`. Those links are used mostly used to allow for easy linking to particular queries. One can't get those links using the cBioPortal user interface itself, they are only mentioned in the documentation of the Web API (https://www.cbioportal.org/webAPI).
2. In the quick search when querying for a gene. Quick search is disabled by default. It is a beta feature. See the [quick search documentation](application.properties-Reference.md#quick-search-beta).

The configuration is set with the following if you have session service enabled:

```
default_cross_cancer_study_session_id=
```

The title will be pulled from the virtual study. Make sure to create a `virtual_study` with studies that everybody has access to and don't use a `main_session` id.

If session service is disabled one can use the following instead:

```
# query this comma separated list of studies
default_cross_cancer_study_list=
default_cross_cancer_study_list_name=
```

### Quick Search (BETA)

![Quick search example](/images/previews/quick\_search\_example.png)

Enable or disable the quick search with the following:

```
# Enable/Disable quick search (default is false)
quick_search.enabled=true
```

The default studies queried when searching for a single gene is defined with the `default_cross_cancer_study_session_id` or `default_cross_cancer_study_list` properties as described in the [cross cancer study query default section](application.properties-Reference.md#cross-cancer-study-query-default).

### Hide sections in the right navigation bar

Settings controlling what to show in the right navigation bar. Set them to `false` if you want to hide those sections, otherwise set the properties to `true`.

```
#Cancer Studies section:
skin.right_nav.show_data_sets=

#Example Queries section:
skin.right_nav.show_examples=

#Testimonials section:
skin.right_nav.show_testimonials=

#What's New section
skin.right_nav.show_whats_new=

#Twitter feed in What's New section
skin.right_nav.show_twitter=
```

#Interactive tours section
skin.right_nav.show_web_tours=

### Control the content of specific sections

Setting controlling the blurb: you can add any HTML code here that you want to visualize. This will be shown between the cBioPortal menu and the Query selector in the main page.

```
skin.blurb=
```

Setting controlling the citation below the blurb: you can add any HTML code here that you want to visualize. If the field is left empty, this HTML code will be shown: `Please cite: <a href="https://cancerdiscovery.aacrjournals.org/content/2/5/401.abstract" target="_blank">Cerami et al., 2012</a> &amp; <a href="https://www.ncbi.nlm.nih.gov/pubmed/23550210" target="_blank">Gao et al., 2013</a>`).

```
skin.citation_rule_text=
```

Setting controlling the footer: you can add any HTML code here that you want to visualize. If the field is left empty, the default footer (from www.cbioportal.org) will be shown.

```
skin.footer=
```

Setting controlling listing the development channels in the site's footer section

```
skin.footer_show_dev=
```

Settings controlling the "What's New" blurb in the right navigation bar: you can add any HTML code here that you want to visualize. If the field is left empty, the Twitter timeline will be shown (as long as `skin.right_nav.show_whats_new` is `true`, otherwise this section will not be displayed).

```
skin.right_nav.whats_new_blurb=
```

Add custom logos to the left or right side of the header section. Place here the full name of the logo file (e.g. `logo.png`). This file should be present in `$PORTAL_HOME/portal/images/`.

```
skin.left_logo=
skin.right_logo=
```

Prevent users from saving data by removing all Download tabs and download and copy-to-clipboard controls.
```
skin.hide_download_controls=
```



### Control default setting for filtering of genes in mutation and CNA tables of patient view

Different samples of a patient may have been analyzed with different gene panels. In patient view mutations and discrete CNA's can be filtered based on whether the gene of respective mutations/CNA's was profiled in all samples of the patient (mutations profiled in `all samples`), or not (mutations profiled in `any sample`). Setting this field to `true` will make patient view select the `all samples` filter at startup. When set to false or left blank the patient view will default to the `any samples` filter setting.

```
skin.patientview.filter_genes_profiled_all_samples=
```

### Control default settings of the VAF line chart in the genomic evolution tab of patient view
If you want to enable log scale and sequential mode by default, set this property to `true`:
```
vaf.log_scale.default=true|false
vaf.sequential_mode.default=true|false
```

### Control unauthorized studies to be displayed on the home page

By default, on an authenticated portal the home page will only show studies for which the current user is authorized. By setting the _skin.home\_page.show\_unauthorized\_studies_ property to _true_ the home page will also show unauthorized studies. The unauthorized studies will appear greyed out and cannot be selected for downstream analysis in Results View or Study View.

```
skin.home_page.show_unauthorized_studies=
```

If _show\_unauthorized\_studies_ feature has been enabled, a message (template) can be defined by the property below that informs the user of insufficient permissions. This message will appear inside a tooltip when hovering the lock icon next to the study name on the Study Selection page. This message may contain placeholders for study-specific information derived from [study tags data](/File-Formats.md#study-tags-file). The information in the study tags JSON file can be accessed using Json Path placeholders. For example `{$.Owner.email}` points to member `{Owner: {email: "me@myself.org"}}`. For the studies that don't have this information available in the study tags, the default message "The study is unauthorized. You need to request access." will be displayed. In addition to the study tags information, the cancer study identifier can be included in the message using {$.studyId} placeholder (does not have to be present in study tags file). For example: skin.home\_page.unauthorized\_studies\_global\_message=You do not have access to this study. You can request access with {$.Owner.email} (please mention the '{$.studyId}' study identifier).

```
skin.home_page.unauthorized_studies_global_message=
```

### Show badge with reference genome
In instances with hg19 and hg38 studies you can show the reference genome in the home page next to the number of samples. This can be done setting this property to `true` (`false` by default):

```
skin.home_page.show_reference_genome=
```

### Control the appearance of the settings menu in study view and group comparison that controls custom annotation-based filtering

A settings menu that allows the user to filter alterations in study view and group comparison may be used when [custom driver annotations](/File-Formats.md#custom-driver-annotations) were loaded for the study or studies displayed in these sections. This menu will only appear, when setting the property _skin.show\_settings\_menu_ to _true_.

```
skin.show_settings_menu=
```

### Hide logout button

When the user is logged-in, a button with logout and data access token options is shown at the right side of the header section. This button can be hidden by setting the property \_skin.hide\_logout\_button to _true_ (default is _false_). Hiding the logout button will remove the ability to perform a manual logout and download data access tokens by the user.

```
skin.hide_logout_button=
```

### Namespace columns visible in Mutation Table by default

Namespace columns are custom columns in the MAF file that can be shown in Mutation Table components. By setting this property to _true_ these columns will be visible when the mutation table is created. By default, the namespace column is hidden and can be made visible using the column selection menu.

```
skin.mutation_table.namespace_column.show_by_default=
```

### Default visible columns on init in Mutation, Copy-Number and Structural Variant Tables

Define the columns that are going to be visible in the Mutation, Copy-Number and Structural Variant Tables on the Patient View and the Mutation Table in the Results View. The `skin.mutation_table.namespace_column.show_by_default` takes precedence over these settings for namespace columns.
```
skin.results_view.mutation_table.columns.show_on_init=
skin.patient_view.mutation_table.columns.show_on_init=
skin.patient_view.copy_number_table.columns.show_on_init=
skin.patient_view.structural_variant_table.columns.show_on_init=
```

### Default sort columns on Mutation, Copy-Number and Structural Variant Tables

Define the column that are going to sort be default in the Mutation, Copy-Number and Structural Variant Tables on the Patient View and the Mutation Table in the Results View.
Column name should be exactly the same as shown in tables.
```
skin.results_view.tables.default_sort_column=
skin.patient_view.tables.default_sort_column=
```

### Define custom sample type colors
Define the colors of custom sample types in the patient view using a json object with for each sample type a color:
```
skin.patient_view.custom_sample_type_colors_json=classpath:/skin-patient-view-custom-sample-type-colors.json
```
Example of json file contents:
```json
{
    "Primary": "green",
    "Biopsy 3": "#00c040ff"
}
```

### Choose the display name for authenticated users

By default the display name for authenticated users is email, but it can be changed for the user name:

```
skin.user_display_name=username
```

### Hide p- and q-values in survival types table

```
# Show/hide p- and q-values in survival types table (default is true)
survival.show_p_q_values_in_survival_type_table=
```

### Set the initial x-axis limit for the survival plot

By default, the initial x-axis limit for a survival plot is the time of the latest event in the data. With this setting, you can make the initial x-axis limit be smaller than that. ![Survival x-axis limit examples](/images/previews/survival\_x\_axis\_limit.png)

```
# Set initial x-axis limit for survival plot (by default, initial limit will be the latest event in the data)
survival.initial_x_axis_limit=
```

### Display installation map

This setting specifies the URL for the iframes of a given portal's installation map on the homepage and standalone page. By default, the installation map configuration is commented out with the iframes and standalone page hidden. The configuration below is used for https://www.cbioportal.org/.

```
installation_map_url=https://installationmap.netlify.app/
```

To set up an installation map instance, one may consult the source code for the installation map [here](https://github.com/cbioportal/installation-map).


### Show structural variants table on study view

The structural variants table widget on study view allows users to define cohorts based on gene-orientation specific structural variant data format (see [here](/File-Formats.md#structural-variant-data)).
The structural variants table widget supports cohort selection based on _gene1_, _gene2_ and _gene1/gene2_ orientation specific genomic events.
This property enables the structural variants table widget on Study View.

:warning: Although _gene1_ and _gene2_ specific queries may be used to investigate up- and downstream fusion partners, respectively, the validity of this
depends on supports for this interpretation in the underlying data.

```
skin.study_view.show_sv_table=true
```


## Ensembl transcript lookup URL

The Mutations tab contains various links, redirecting the user to external information resources regarding the displayed transcript. The Ensembl template URL can be customized by modifying the property:

```
ensembl.transcript_url=
```

The default setting is `https://ensembl.org/homo_sapiens/Transcript/Summary?t=<%= transcriptId %>`. The `<%= transcriptId %>` is substituted by the frontend code into respective transcript ID.

## Segment File URL

This is a root URL to where segment files can be found. This is used when you want to provide segment file viewing via external tools such as [IGV](https://www.broadinstitute.org/igv/).

```
segfile.url=
```

## Bitly API Username and Key

The following properties are used to provide shortened bookmarks to the cBioPortal:

```
bitly.user=
bitly.api_key=
```

To obtain a bitly username and key, first register at: https://bitly.com/

Then, go to: https://bitly.com/a/your\_api\_key

**Note:** If you are developing on a local machine, and using localhost, the bitly URL shortening service will not work. This is because bitly will not shorten URLs for localhost. Once you deploy to your final server, the issue should go away.

## Google Analytics

When the following property is defined, Google Analytics will track site usage.

:warning: In contrast to what this property's name suggests, this should be the **GA Universal Analytics ID** (UA-XXXXX-YY), not to be confused with the new [GA-4 ID, introduced in 2020](https://support.google.com/analytics/answer/10089681). A Universal Analytics GA property can be created at [Google Analytics](https://analytics.google.com/analytics/web/provision/), create a new GA property, and make sure to click **Show advanced options** and switch on **Create a Universal Analytics property**. After creating a Universal Analytics property, the tracking ID can be found at **Admin**, column **property**, under **Property Settings**

:warning: Adblockers may block site tracking by GA

```
google_analytics_profile_id
```

## Password Authentication

The portal supports password authentication via Google+. Before you start you need to setup a google account that will own the authentication API. Follow https://developers.google.com/identity/sign-in/web/devconsole-project to get clientID and secret. Fill it in application.properties:

```
googleplus.consumer.key=195047654890-499gl89hj65j8d2eorqe0jvjnfaxcln0.apps.googleusercontent.com
googleplus.consumer.secret=2jCfg4SPWdGfXF44WC588dK
```

(note: these are just examples, you need to get your own) You will also need to go to "Google+ API" and click Enable button. In case of problems make sure to enable DEBUG logging for org.springframework.social and org.springframework.security.web.authentication.

To activate password authentication follow the [Deployment with authentication steps](/deployment/deploy-without-docker/Deploying.md#required-login) and set `authenticate=googleplus`.

In addition, set this property in `application.properties`:

```
app.name=cbioportal
```

app.name should be set to the name of the portal instance referenced in the "AUTHORITY" column of the "AUTHORITIES" table. See the [User Authorization](/deployment/authorization-and-authentication/User-Authorization.md) for more information.

## OncoKB integration

OncoKB integration can be turned on or off with the following property (default: true):

```
show.oncokb=true|false
```

A private token is required to access the OncoKB Data (for details see the section [OncoKB Data Access](/deployment/integration-with-other-webservices/OncoKB-Data-Access.md)):

```
oncokb.token=
```

"cBioPortal>=" driver annotation sources in the settings menu of Results View can be hidden by turning off the following property (default: true):

```
show.cbioportal=true|false
```

"COSMIC>=" driver annotation sources in the settings menu of Results View can be hidden by turning off the following property (default: true):

```
show.cosmic=true|false
```

## CIViC integration

CIViC integration can be turned on or off with the following property (default: true):

```
show.civic=true|false
```

The CIViC API url is set to https://civic.genome.wustl.edu/api/ by default. It can be overridden using the following property:

```
civic.url=
```

## Genome Nexus Integration

Genome Nexus provides annotations of mutations in cBioPortal. The mutations tab relies heavily on the Genome Nexus API, therefore that tab won't work well without it. By default cBioPortal will use the public Genome Nexus API, such that no extra configuration is necessary.

### Genome Build

Genome Nexus supports both GRCh37 and GRCh38, but support for the latter is limited. Several annotation sources served by Genome Nexus might not have official GRCh38 support yet i.e. [OncoKB](https://www.oncokb.org), [CIViC](https://civicdb.org), [Cancer Hotspots](https://www.cancerhotspots.org), [My Cancer Genome](https://www.mycancergenome.org) and [3D structures](https://g2s.genomenexus.org). Although most of the time the canonical transcript for a gene will be the same between GRCh37 and GRCh38 there might be some that cause issues. In addition the complete integration of cBioPortal with Genome Nexus' GRCh38 is not complete yet. That is cBioPortal currently only connects to one Genome Nexus API by default (the GRCh37 one), so it's not possible to have multiple genome builds in one instance of cBioPortal and get the correct annotations from Genome Nexus for both. Currently only the [mutation mapper tool page](https://www.cbioportal.org/mutation\_mapper) is able to handle both.

### Properties

By default the Genome Nexus API url is set to https://v1.genomenexus.org/, which uses GRCh37. It can be overridden using the following property:

```
genomenexus.url=
```

Genome Nexus provides a set of mappings from Hugo genes names to Ensembl transcript IDs. There are two mappings: `mskcc` and `uniprot`. 
You can read more about the difference between those in the [Mutation Data Annotation Section](./../../mutation-data-transcript-annotation.md). The default is currently `uniprot`, but we recommend new installers to use `mskcc` and people with older installations to consider migrating. The property
can be changed with:

```
genomenexus.isoform_override_source=mskcc
```

The mutation mapper tool page can annotate GRCh38 coordinates. By default it uses https://grch38.genomenexus.org. It can be overridden by setting:

```
genomenexus.url.grch38=
```

The GRCh38 annotation in mutation mapper can be hidden by setting `show.mutation_mappert_tool.grch38=false`, by default it's set to `true`.

## MDACC Heatmap Integration

MDACC Heatmap integration (button in OncoPrint heatmap dropdown and tab on Study page can be turned on or off by setting the following property:

```
show.mdacc.heatmap=true
```

## OncoPrint

The default view in OncoPrint ("patient" or "sample") can be set with the following option. The default is "patient".

```
oncoprint.defaultview=sample
```

Configuration of tracks that will be visible by default in the oncoprint. It points to a JSON file on the classpath.

```
oncoprint.clinical_tracks.config_json=classpath:/oncoprint-default-tracks.json
```

## Custom annotation of driver and passenger mutations

cBioPortal supports 2 formats to add custom annotations for driver and passenger mutations.

1. **cbp\_driver**: This will define whether a mutation is a driver or not.
2. **cbp\_driver\_tiers**: This can be used to define multiple classes of driver mutations.

These data formats are described in the [cBioPortal MAF specifications](/File-Formats.md#extended-maf-format).

**Enabling custom annotations in the OncoPrint**

To enable functionality for one or both types of custom annotations, enter values for the following properties. These values will appear in the OncoPrint's "Mutation color" menu, Patient View's (mutation, CNA, SV) tables, Results View's mutation table, and Group Comparison View's mutation table.

```
oncoprint.custom_driver_annotation.binary.menu_label=Custom Driver
oncoprint.custom_driver_annotation.binary.menu_description=Custom driver tiers
oncoprint.custom_driver_annotation.tiers.menu_label=Custom Driver Tiers
oncoprint.custom_driver_annotation.tiers.menu_description=Custom driver tiers
```

**Automatic selection of OncoKB, hotspots and custom annotations**

OncoKB and Hotspots are by default automatically selected as annotation source, if `show.oncokb` and `show.hotspots` are set to `true`. To add automatic selection of custom driver or custom driver tiers annotations, set the respective property to `true`. Default is `false`.

```
oncoprint.custom_driver_annotation.binary.default=true|false
oncoprint.custom_driver_annotation.tiers.default=true|false
```

If you want to disable the automatic selection of OncoKB and hotspots as annotation source, set these properties to `false`:

```
oncoprint.oncokb.default=true|false
oncoprint.hotspots.default=true|false
```

If you want to enable oncoprint heatmap clustering by default, set this property to `true`:
```
oncoprint.clustered.default=true|false
```

**Automatic hiding of variants of unknown significance (VUS)**

By default, the selection box to hide VUS mutations is unchecked. If you want to automatically hide VUS, set this property to `true`. Default is `false`.

```
oncoprint.hide_vus.default=true|false
```

## Gene sets used for gene querying

To change the gene sets used for gene querying, create a JSON file and add gene sets, following the format specified in the examples below. Set the path to this file (e.g. `file:/cbioportal/custom_gene_sets.json`) in the following property and restart Tomcat to apply the update. The default gene sets will be replaced by the ones in `custom_gene_sets.json`.

```
querypage.setsofgenes.location=file:/<path>
```

### Example with gene names

In this example, two gene sets will appear in the query page, under the names "Prostate Cancer: AR Signaling" and "Prostate Cancer: AR and steroid synthesis enzymes".

```
[{
	"id": "Prostate Cancer: AR Signaling",
	"genes": ["SOX9", "RAN", "TNK2", "EP300", "PXN", "NCOA2", "AR", "NRIP1", "NCOR1", "NCOR2"]
}, {
	"id": "Prostate Cancer: AR and steroid synthesis enzymes",
	"genes": ["AKR1C3", "AR", "CYB5A", "CYP11A1", "CYP11B1", "CYP11B2", "CYP17A1", "CYP19A1", "CYP21A2", "HSD17B1", "HSD17B10", "HSD17B11", "HSD17B12", "HSD17B13", "HSD17B14", "HSD17B2", "HSD17B3", "HSD17B4", "HSD17B6", "HSD17B7", "HSD17B8", "HSD3B1", "HSD3B2", "HSD3B7", "RDH5", "SHBG", "SRD5A1", "SRD5A2", "SRD5A3", "STAR"]
}]
```

### Example with specific alterations

In this example, only one gene set will appear in the query page, under the name "Genes with alterations", which will add the different genetic alterations stated below in the query box.

```
[{
	"id": "Genes with alterations",
	"genes": ["TP53: MUT=R273C;", "KRAS: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;"]
}]
```

### Example with merged gene tracks

In this example, only one gene set will appear in the query page, under the name "BRCA genes test", containing the merged gene track called "BRCA genes".

```
[{
	"id": "BRCA genes test",
	"genes": ["[\\\"BRCA genes\\\" BRCA1: MUT=E1258D;", "BRCA2: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;]"]
}]
```

This gene set will add the following in the query box:

```
"BRCA genes" BRCA1: MUT=E1258D; BRCA2: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;
```

## Cache Settings

cBioPortal is supported on the backend with Ehcache or Redis. These caches are configurable from within application.properties through the following properties.

The cache type is set using `persistence.cache_type`. Valid values are `no-cache`, `redis` (redis), `ehache-heap` (ehcache heap-only), `ehache-disk` (ehcache disk-only), and `ehache-hybrid` (ehcache disk + heap). By default, `persistence.cache_type` is set to `no-cache` which disables the cache. When the cache is disabled, no responses will be stored in the cache.

:warning: the 'redis' caching option will likely cause a conflict when installing the portal in a Tomcat installation which uses redisson for session management. If you plan to deploy cbioportal to such a system, avoid the 'redis' caching option for `persistence.cache_type` and be sure to build cbioportal.war with the maven option `-Dexclude-redisson` (see [Building with Maven](/deployment/deploy-without-docker/Build-from-Source.md#building-with-maven)).

```
persistence.cache_type=[no-cache or ehache-heap or ehcache-disk or ehcache-hybrid or redis]
```

Logged metrics and additional information such as cache size and cached keys are available through an optional endpoint. The optional endpoint is turned off by default but can be turned on by setting `cache.statistics_endpoint_enabled` to true.

```
cache.statistics_endpoint_enabled=false[true or false]
```

The cache statistics endpoint is hidden on the api page; users must directly access the URL to view the response. The cache statistics endpoint can be accessed in the following ways.

For a list of all keys in the cache:

```
/api/[name of cache]/keysInCache
```

For a list of counts of keys in cache per repository class:

```
/api/[name of cache]/keyCountsPerClass
```

For general statistics about the cache such as memory usage (not currently implemented for Redis):

```
/api/cacheStatistics
```

**WARNING**: It must be noted that since cache statistics endpoint returns data on cache keys, the endpoint may expose otherwise hidden database query parameters such as sample identifiers, study names, etc. Generally, it is recommended that the endpoint only be turned on during cache-related development for testing. Deployers of a protected portal where users only have authorities to a subset of studies should carefully consider whether or not to turn on the cache statistics endpoint, as it does not filter the results.

For more information on how caching is implemented in cBioPortal refer to the [Caching](/deployment/customization/Caching.md) documentation.

### Redis

To cache with Redis set `persistence.cache_type` to `redis`.

To setup the Redis cache servers the following properties are required:

```
redis.name=unique_cbioportal_instance_name
redis.leader_address=redis://{host/servicename}:6379
redis.follower_address=redis://{host/servicename}:6379
redis.database=0
redis.password=password
```

If you are running one redis instance for multiple instances of cBioPortal, one can use the properties `redis.name` and `redis.database` to avoid clashes. If you are running only one instance of cBioPortal any value for name/database will do.

There are also some optional parameters:

`redis.clear_on_startup`: If `true`, the caches will clear on startup. This is important to do to avoid reading old study data from the cache. You may want to turn it off and clear redis yourself if you are running in a clustered environments, as you'll have frequent restarts that do not require you to clear the redis cache.\
`redis.ttl_mins`: The time to live of items in the general cache, in minutes. The default value is 10000, or just under 7 days.

For more information on Redis, refer to the official documentation [here](https://redis.io/documentation)

### Ehcache

To cache with Ehcache set `persistence.cache_type` to `ehache-heap` (ehcache heap-only), `ehache-disk` (ehcache disk-only), or `ehache-hybrid` (ehcache disk + heap).

Ehcache initializes caches using a template found in an Ehcache xml configuration file. When caching is enabled, set `ehcache.xml_configuration` to the name of the Ehcache xml configuration file. The default provided is `ehcache.xml`; to change the cache template, directly edit this file. Alternatively, you can create your own Ehcache xml configuration file, place it under `/persistence/persistence-api/src/main/resources/` and set `ehcache.xml_configuration` to `/[Ehcache xml configuration filename]`.

```
ehcache.xml_configuration=
```

If the cache is configured to use disk resources, users must make a directory available and set it with the `ehcache.persistence_path` property. Ehcache will create separate directories under the provided path for each cache defined in the ehcache.xml\_configuration file.

```
ehcache.persistence_path=[location on the disk filesystem where Ehcache can write the cache to /tmp/]
```

Cache size must be set for heap and/or disk depending on which are in use; Ehcache requires disk size to be greater than heap size in a hybrid configuration. Zero is not a supported size and will cause an exception. Units are in megabytes. Default values are provided. The general repository cache is specified to use 1024MB of heap and 4096MB of disk. The static repository cache is specified to use 30MB of heap and 32MB of disk. For installations with increased traffic or data, cache sizes can be increased to further improve performance.

```
ehcache.general_repository_cache.max_mega_bytes_heap=
ehcache.general_repository_cache.max_mega_bytes_local_disk=

ehcache.static_repository_cache_one.max_mega_bytes_heap=
ehcache.static_repository_cache_one.max_mega_bytes_local_disk=
```

For more information on Ehcache, refer to the official documentation [here](https://www.ehcache.org/documentation/3.7/index.html)

## Evict caches with the /api/cache endpoint

`DELETE` http requests to the `/api/cache` endpoint will flush the cBioPortal caches, and serves as an alternative to restarting the cBioPortal application.

By default the endpoint is disabled. The endpoint can be enabled by setting:

```
cache.endpoint.enabled=true
```

Access to the endpoint is not regulated by the configured user authorization mechanism. Instead, an API key should be passed with the `X-API-KEY` header. The accepted value for the API key can be configured by setting (for example):

```
cache.endpoint.api-key=7d70fecb-cda8-490f-9ea2-ef874b6512f4
```

## Delegate user-authorization cache to Spring-managed cache

For evaluation fo user permissions cBioPortal uses a user-authorization cache that is populated at startup. By setting the `cache.cache-map-utils.spring-managed` property to _true_ this cache will be managed by the Spring caching solution such as EHCache or Redis. For more extended information, see [here](Caching.md#user-authorization-cache)

```
cache.cache-map-utils.spring-managed=false
```

## Enable GSVA functionality

[GSVA functionality](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#gene-set-data) can be enabled by uncommenting this line (and making sure it is set to `true`):

```
skin.show_gsva=true
```

## Set default thresholds for geneset hierarchy

```
skin.geneset_hierarchy.default_gsva_score=0.3
skin.geneset_hierarchy.default_p_value=0.02
```

## Collapses the tree widget of the geneset hierarchy dialog on initialization

By default, the tree is expanded (property value is `false`).

```
skin.geneset_hierarchy.collapse_by_default = true
```

## Cross study expression and protein data


By default we hide expression data for multi-study queries as they are 
usually not normalized across studies. For the public cBioPortal for instance, only TCGA Pancan Atlas studies expression data has been normalized.  

If you know the expression data in your instance is comparable, or is comparable for a subset of
studies, you can configure a rule as follows. 

The value of this property can be boolean (true|false) or a javascript function
which executes at runtime and is passed the list of study objects being queried
by the user and evaluates whether expression data can be safely displayed.

```
// a function that accepts the users's selected studies and
// returns whether or not to allow expression data from the involved studies to be mixed
enable_cross_study_expression = (selectedStudies)=>{ [your logic] return true|false }
```

```
// boolean
enable_cross_study_expression = true|false
```

## Combined Study View Summary Limits
### Background
A limit is added to prevent poor performance of Study View when selecting too large sample numbers.
### Properties
* `studyview.max_samples_selected`: Limit is disabled when not set

### Behavior
When these limits are exceeded the "Explore Selected Studies" button will be disabled on the Study View Page.

## Request Body Compression

### Background

Some REST endpoints that the cBioPortal frontend uses have request bodies that scale as your dataset increases. In portals where users commonly query more than 100,000 samples, we found that some of these request bodies could get as large as 20 Mb. These large request bodies pose a significant problem for users with poor upload speeds - some users experienced upload times of more than five minutes for these requests. Request body compression is our temporary solution to this problem. When this feature is toggled on, we compress the request bodies of a few problematic endpoints.

### Properties

There are two `portal.property` values related to this feature:

* `enable_request_body_gzip_compression`: when `true`, the feature will be enabled.
* `request_gzip_body_size_bytes`: the maximum allowable unzipped request body in bytes. Defaults to 80000000 (80 Mb).

### Behavior

* This is a nonbreaking change. Any consumers of the cBioPortal API you have that send requests with uncompressed request bodies will continue to work, regardless of whether you turn this feature on or off.
* If you turn this feature on, the cBioPortal API will now be able to handle any request with a gzipped request body, provided:
  * It is a POST request.
  * It has a `Content-Encoding: gzip` header.

### Reasons to Enable This Feature

* You have studies with tens of thousands of samples.
* You have users with poor upload speeds (< 1mb up).

### Reasons to Disable This Feature

* It is harder to debug gzipped requests
  * Chrome's `copy request as CURL` will not work.
  * The compressed request body is not human-readable.
* It is a potential vector for denial of memory attacks.
  * Any request that has a body that takes significantly more space in memory than it does in the request body is potentially dangerous. We try to mitigate this by limiting the size of the unzipped request body via the `request_gzip_body_size_bytes` property, but at a fundamental level, this is still a concern.
  * Along these lines, if you do enable this feature, setting `request_gzip_body_size_bytes` to an arbitrarily large number would be unwise.
* This is not a cure-all for performance issues
  * Most requests the cBioPortal makes do not have large request bodies, so most requests will not be compressed, and will see no performance improvement.
  * Users with good upload speeds will see minimal performance improvements, as their upload speed is not a bottleneck.

# DataSets Tab (Study Download Links)
### Background
The DataSets tab has the ability to create a ``download`` button that allows users to quickly download "raw" public studies.
### Properties
* ``study_download_url`` : when set, the feature will be enabled
### Behavior
For private instances that want to replicate the``public-portal`` they first must set up their studies 
they want available for download in a similar format to what is described in the Example section below.
The studies are located on the ``public-portal`` at  `https://cbioportal-datahub.s3.amazonaws.com/`. 
Then there is a ``study_list.json`` defined that list the studies that can be downloaded.
The studies to be downloaded need to be compressed with the extension ``tar.gz``

### Example
* We have set ``study_download_url`` property to `https://cbioportal-datahub.s3.amazonaws.com/`
* ``study_list.json`` resides ``https://cbioportal-datahub.s3.amazonaws.com/study_list.json``
  * ``[ "acbc_mskcc_2015", "acc_2019"]`` Example of contents
* ``acbc_mskcc_2015.tar.gz`` resides `https://cbioportal-datahub.s3.amazonaws.com/acbc_mskcc_2015.tar.gz`

# Prioritized studies on study selector view 
By default, the studies loaded into a local cBioPortal instance are organized based on their cancer type (i.e. Breast >> Other).
```
priority_studies=
```
The value of this variable will create a custom category with studies on the top of the study selector view. The format for the string should be category1#study1a,study1b,study1c;category2#study2 (e.g., PanCancer Studies#msk_impact_2017), where the ``category`` can be any string and the ``study`` should be the study ID of the required uploaded study. 

## Study Tag functionality
Study Tags allow portal maintainers to define miscellaneous descriptive meta data to studies, which will be shown to users in tooltips and are also searchable. This feature
is on by default but can be disabled using the following property.
```
//boolean
enable_study_tags=true|false
```
# Backend Caching

cBioPortal provides the option of caching information on the backend to improve performance. Without caching, every time
a request is received by the backend, a query is sent to the database system for information, and the returned data is
processed to construct a response. This may lead to performance issues as the entire process can be rather costly,
especially for queries on larger studies. With caching turned on, query responses can be taken directly from the cache
if they have already been constructed. They would only be constructed for the initial query.

## Cache Configuration

The portal is configured to use Ehcache or Redis for backend caching. Ehcache supports a hybrid (disk + heap),
disk-only, and heap-only mode. Redis stores the cache in memory and periodically writes the updated data to disk. Cache
configuration is specified inside `application.properties`(more
information [here](/deployment/customization/application.properties-Reference.md#cache-settings)).

## Creating additional caches

The default configuration initializes two separate caches; however, you may wish to introduce new caches for different
datatypes. Please see the [Redis](#redis) and [Ehcache](#ehcache) sections to see how to set up a new cache in whichever
system you are using.

### Redis

Cache initialization is handled inside
the [CustomRedisCachingProvider](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api/src/main/java/org/cbioportal/persistence/util/CustomRedisCachingProvider.java)
. To create additional caches (e.g creating a cache specifically for clinical data), new code must be added to
the `CustomRedisCachingProvider`.

Within the `CustomRedisCachingProvider`, create your new cache using the `CacheManager`. The appName must be prepended
to your cache name.

```
manager.createCache(appName + "ClinicalDataCache", config);
```

You also need to create a new cache resolver
in [RedisConfig.java](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api/src/main/java/org/cbioportal/persistence/config/RedisConfig.java):

```
@Bean
public CacheResolver generalRepositoryCacheResolver() {
    return new NamedCacheResolver(cacheManager(), ${redis.name:cbioportal} + "GeneralRepositoryCache");
}
```

The `@Cacheable` annotation must also be added (or adjusted) to function declarations to indicate which functions are to
be cached. Those might look like this example:

``` 
@Cacheable(cacheResolver = "clinicalDataCacheResolver", condition = "@cacheEnabledConfig.getEnabled()")
public String getDataFromClinicalDataRepository(String param) {}
```

For more information on linking caches to functions, refer to the
documentation [here](https://spring.io/guides/gs/caching/).

### Ehcache

Within the `CustomEhcachingProvider`, initialize a new `ResourcePoolsBuilder` for the new cache and set the resources
accordingly.

```
ResourcePoolsBuilder clinicalDataCacheResourcePoolsBuilder = ResourcePoolsBuilder.newResourcePoolsBuilder();
clinicalDataCacheResourcePoolsBuilder = clinicalDataCacheResourcePoolsBuilder.heap(clinicalDataCacheHeapSize, MemoryUnit.MB);
clinicalDataCacheResourcePoolsBuilder = clinicalDataCacheResourcePoolsBuilder.disk(clinicalDataCacheDiskSize, MemoryUnit.MB);
```

After initialzing the `ResourcePoolsBuilder`, create a CacheConfiguration for the new cache using the
new `ResourcePoolsBuilder` just created.

```
CacheConfiguration<Object, Object> clinicalDataCacheConfiguration = xmlConfiguration.newCacheConfigurationBuilderFromTemplate("RepositoryCacheTemplate", 
Object.class, Object.class, clinicalDataCacheResourcePoolsBuilder)
.withSizeOfMaxObjectGraph(Long.MAX_VALUE)
.withSizeOfMaxObjectSize(Long.MAX_VALUE, MemoryUnit.B)
.build();
```

Finally, add the new `CacheConfiguration` to the map of managed caches with a name for the cache.

```
caches.put("ClinicalDataCache", ClinicalDataCacheConfiguration);
```

You also need to create a new cache resolver
in [EhCacheConfig.java](https://github.com/cBioPortal/cbioportal/blob/master/persistence-api/src/main/java/org/cbioportal/persistence/config/EhCacheConfig.java):

```
@Bean
public NamedCacheResolver generalRepositoryCacheResolver() {
    return new NamedCacheResolver(cacheManager(), "GeneralRepositoryCache");
}
```

The `@Cacheable` annotation must also be added (or adjusted) to function declarations to indicate which functions are to
be cached. Those might look like this example:

``` 
@Cacheable(cacheResolver = "clinicalDataCacheResolver", condition = "@cacheEnabledConfig.getEnabled()")
public String getDataFromClinicalDataRepository(String param) {}
```

Additionally, new properties for setting cache sizes should be added to `application.properties` and loaded into
the [CustomEhcachingProvider](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api/src/main/java/org/cbioportal/persistence/util/CustomEhcachingProvider.java)
. Alternatively, values may be hardcoded directly
inside [CustomEhcachingProvider](https://github.com/cBioPortal/cbioportal/blob/master/persistence/persistence-api/src/main/java/org/cbioportal/persistence/util/CustomEhcachingProvider.java)
.

For more information on cache templates and the Ehcache xml configuration file, refer to the
documentation [here](https://www.ehcache.org/documentation/3.7/xml.html).

For more information on linking caches to functions, refer to the
documentation [here](https://spring.io/guides/gs/caching/).


## User-authorization cache

In addition to the above-mentioned Spring-managed caches, cBioPortal maintains a separate cache that holds references to
sample lists, molecular profiles and cancer studies. This user-authorization cache is used to establish
whether a user has access to the data of a particular sample list or molecular profile based on study-level permissions.

By default, the user-authorization cache is implemented as a HashMap that is populated when cBioPortal is started. This
implementation allows for very fast response times of user-permission evaluation.

The user-authorization cache can be delegated to the Spring-managed caching solution by setting the [cache.cache-map-utils.spring-managed](application.properties-Reference.md#cache-settings)
to _true_. Depending on the implementation, this may add a delay to any data request that is caused by the additional consultation
of the external cache. This configuration should only be used where a central caching solution is required or no
instance/container-specific local caches are allowed. For example, cache eviction via the `api/cache` endpoint in a Kubernetes
deployment of cBioPortal where multiple pods/containers that represent a single cBioPortal instance is possible with a
Spring-managed user-authorization cache because a call to this endpoint in a single pod/container invalidates Redis caches
for the entire deployment thereby preventing inconsistent state of user-authorization caches between pods.


## Cache eviction

When the database is updated (e.g new studies loaded, existing study updated, new gene panel imported) the caches of a
cBioPortal instance should be updated. One way is to restart the cBioPortal spring application. When using Redis, this
will work only when `redis.clear_on_startup` is not set to _false_.

Alternatively, caches can be cleared (a.k.a. _evicted_) by calling the _/api/cache_ endpoint. Advantage of the cache
eviction end point is that user-sessions remain undisturbed since the portal instance is not restarted. By default the
cache eviction enpoint is disabled and can be enabled by setting `cache.endpoint.enabled` to _true_. The endpoint is
secured by a secret API key that can be customized with the `cache.endpoint.api-key` property. Caches are evicted by
making a DELETE request to the endoint while passing the API key in the `X-API-KEY` header. When using _curl_ use the
following command (replace the API key for the value configured in _application.properties_):

```
curl -X DELETE http://my-portal-url.org/api/cache  -H 'X-API-KEY: fd15f1ae-66f2-4b8a-8d54-fb899b03557e'
```

### Cache eviction after cancer study updates

When a study is added, deleted or updated, a more selective cache eviction strategy is possible, where only affected
cached data is evicted. This more selective cache eviction is triggered by calling the `/api/cache/{studyId}` endpoint
where _{studyId}_ is the _cancer_study_identifier_ stated in the meta-study.txt file. When using _curl_ use the
following command after update of a study with study identifier _my_cancer_study_ (replace the API key for the value
configured in _application.properties_):

```
curl -X DELETE http://my-portal-url.org/api/cache/my_cancer_study  -H 'X-API-KEY: fd15f1ae-66f2-4b8a-8d54-fb899b03557e'
```

:important: This endpoint can ony be used when adding/deleting/updating a study. When data related to gene panels or
gene sets is updated, all caches should be evicted with a call to `/api/cache`.

#### How does study-specific cache eviction work?

##### Structure of cache keys

The caching solutions integrated with cBioPortal (EHCache and Redis) store data as key-value pairs. Each key represents
a method call signature that contains the Java class name, the method name and a serialized representation of all method
arguments. For instance this is the key for a call to the CancerTypMyBatisRepository.getAllCancerTypes() method with
arguments _[ SUMMARY, 10000000, 0, null, ASC]_:

```
CancerTypeMyBatisRepository_getAllCancerTypes_"SUMMARY"_10000000_0_null_"ASC"
```

Cached data that relates to a specific study can be recognized by the occurrence of the cancer study identifier anywhere
in the method arguments. The study identifier can occur in the method arguments because it is passed as argument itself,
like here for a study with identifier _study_es_0_:

```
SignificantCopyNumberRegionMyBatisRepository_getSignificantCopyNumberRegions_"study_es_0"_"SUMMARY"_null_null_null_null
```

Alternatively the study identifier is present as the prefix of referenced study entities. For example, this is the
request for all molecular profiles:

```
MolecularProfileMyBatisRepository_getMolecularProfiles_["study_es_0_mrna_median_Zscores","study_es_0_log2CNA","study_es_0_fusion","study_es_0_treatment_ec50","study_es_0_mutational_signature","study_es_0_gsva_pvalues","study_es_0_mutations","study_es_0_methylation_hm27","study_es_0_gistic","study_es_0_generic_assay_patient_test","study_es_0_treatment_ic50","study_es_0_mrna","study_es_0_gsva_scores"]_"SUMMARY""
```

#### Cache eviction rules

When a study is added, deleted or updated, all caches are evicted where the respective key meets any of these
requirements:

1. The key contains the cancer study identifier of the study that is added, deleted or updated.
2. The key does not contain the cancer study identifier of any study present in the database.

The rationale behind rule 1. is that when a key references data for the affected study it points to potentially outdated
data and its associated cache should be evicted. The rationale behind rule 2. is that any key that does not reference
data for any study potentially points to data derived from all studies in the database, including the affected study,
and its associated cache should be evicted. Because not every key without study identifiers necessarily points to study
related data, this rule is overly broad. At the moment of this writing we were unable to implement reliable methods that
would further specify such keys. This might be a start-off point for future optimizations.


# Customizing your instance of cBioPortal

cBioPortal contains a number of properties that allow you to customize your cBioPortal instance. 

This page focuses on the skin properties, which allow you to
customize the web page cosmetics, such as custom images,
texts, which tabs are visible, etc.
Nearly all the skins properties have defaults, which can be overwritten by
changing the `application.properties` file located in `$PORTAL_HOME`--see
[the deployment manual](/deployment/deploy-without-docker/Deploying.md). If your cBioPortal instance
does not yet contain a `application.properties` file, you can copy
`src/main/resources/application.properties.EXAMPLE` and edit it.

Below you can find the complete list of all the available skin properties.

# cBioSkins Properties

**Please note!**  If you are reading this on the GitHub web site, you may need to scroll right to see all the columns in the table below.

<table>
   <tbody>
		<tr>
         <th>Name</th>
         <th>Description</th>
         <th>Default</th>
         <th>Type Allowed</th>
		</tr>
		<tr>
			<td>skin.authorization_message</td>
			<td>sets the authorization message that appears after an unsuccessful login</td>
			<td>Access to this portal is only available to authorized users at Memorial Sloan Kettering Cancer Center...</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.blurb</td>
			<td>sets the main text on the index page</td>
			<td>Text for public cBioPortal.org.</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.example_study_queries</td>
			<td>\n-separated list of study query suggestions displayed when clicking the ‚Äòselect a cancer study' search box</td>
			<td>tcga\ntcga -provisional\ntcga -moratorium\ntcga OR icgc\n-"cell line"\nprostate mskcc\nesophageal OR stomach\nserous\nbreast</td>
			<td>text separated by the escape sequence \n</td>
		</tr>
		<tr>
			<td>skin.data_sets_footer</td>
			<td>sets the text that is shown below the number of samples, after clicking on the "DATA SETS" tab in the header.</td>
			<td>The portal currently contains data from the following cancer genomics studies.  The table below lists the number of available samples per data type and tumor.</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.data_sets_header</td>
			<td>sets the text that is shown below the Data Sets title, after clicking on the "DATA SETS" tab in the header. By adding &lt;NUM_CANCER_STUDIES&gt; to the text, the number of cancer studies is inserted</td>
			<td>Data sets of TCGA studies were downloaded from Broad Firehose (https://gdac.broadinstitute.org) and updated monthly. In some studies, data sets were from the TCGA working groups directly.</td>
			<td>Any HTML text</td>
		</tr>
      <tr>
         <td>skin.documentation.about</td>
         <td>set the about page which can be accessed via ABOUT in the header. Please see the cBioportal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>About-Us.md</td>
         <td>text</td>
      </tr>
      <tr>
         <td>skin.documentation.baseurl</td>
         <td>set the base url for the documentation. Please see the cBioPortal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>https://github.com/cBioPortal/cbioportal/wiki/</td>
         <td>text</td>
      </tr>
      <tr>
         <td>skin.documentation.faq</td>
         <td>set the faq page which can be accessed via FAQ in the header. Please see the cBioPortal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>user-guide/faq.md</td>
         <td>text</td>
      </tr>
      <tr>
         <td>skin.documentation.markdown</td>
         <td>set whether the documentation files are in markdown. Please see the cBioPortal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>true</td>
         <td>text</td>
      </tr>
      <tr>
         <td>skin.documentation.news</td>
         <td>set the news page which can be accessed via NEWS in the header. Please see the cBioPortal Documentation section for details on how to properly configure the documentation pages.</td>
         <td>News.md</td>
         <td>text</td>
      </tr>
		<tr>
			<td>skin.email_contact</td>
			<td>sets the contact email address for Questions and feedback, located in the footer.</td>
			<td>cbioportal at googlegroups dot com</td>
			<td>Specify as name at somewhere dot extension or as name@somewhere.extension</td>
		</tr>
      <tr>
         <td>skin.examples_right_column_html</td>
         <td>set the Examples Queries section in the right navigation bar</td>
         <td>&lt;ul&gt;&lt;li&gt;&lt;a href="patient?studyId=ucec_tcga_pub&caseId=TCGA-BK-A0CC"&gt;Patient view of an endometrial cancer case&lt;/a&gt;&lt;/li&gt;&lt;/ul&gt;</td>
         <td>Any HTML text</td>
      </tr>
		<tr>
			<td>skin.footer</td>
			<td>sets the footer text. The text will be appended to the already existing cBioPortal link</td>
			<td>cBioPortal public instance footer.</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.login.contact_html</td>
			<td>sets the feedback message which appears after a failed login attempt</td>
			<td>If you think you have received this message in error, please contact us at...</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.login.saml.registration_html</td>
			<td>sets the button text for saml login</td>
			<td>Sign in with MSK</td>
			<td>Any HTML text</td>
		</tr>
		<tr>
			<td>skin.left_logo</td>
			<td>sets the left logo in the header. Logo should be placed in the images directory or in a subdirectory of the images directory. If placed in a subfolder, the skin.left_logo value has to contain the folder, e.g. skin.right_logo = myFolder/myImage.jpg</td>
			<td>cBioPortal Logo</td>
			<td>text</td>
		</tr>
		<tr>
			<td>skin.right_logo</td>
			<td>sets the right institute logo in the header. Logo should be placed in the images directory or in a subdirectory of the images directory. If placed in a subfolder, the skin.right_logo value has to contain the folder, e.g. skin.right_logo = myFolder/myImage.jpg</td>
			<td>no logo</td>
			<td>text</td>
		</tr>
		<tr>
			<td>skin.right_nav.show_data_sets</td>
			<td>show the "Data Sets" section in the right navigation bar</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.right_nav.show_examples</td>
			<td>show the "Example Queries" section in the right navigation bar</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.right_nav.show_testimonials</td>
			<td>show the "What People are Saying" section in the right navigation bar</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.right_nav.whats_new_blurb</td>
			<td>set the content of the "What‚Äôs New" section in the right navigation bar</td>
			<td>Text for public cBioPortal.org.</td>
			<td>Any HTML text</td>
		</tr>
        <tr>
			<td>skin.right_nav.show_web_tours</td>
			<td>set the "Interactive Tours" section in the right navigation bar</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_about_tab</td>
			<td>show the "ABOUT" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_data_tab</td>
			<td>show the "DATA SETS" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_faqs_tab</td>
			<td>show the "FAQ" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_news_tab</td>
			<td>show the "NEWS" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_r_matlab_tab</td>
			<td>show the "R/MATLAB" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_tools_tab</td>
			<td>show the "TOOLS" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_tutorials_tab</td>
			<td>show the "TUTORIALS" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.show_web_api_tab</td>
			<td>show the "WEB API" tab in the header</td>
			<td>true</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.tag_line_image</td>
			<td>sets the tag line image in the header. The image should be placed in the images directory or in a subdirectory of the images directory. If placed in a subdirectory, the skin.tag_line_image value has to contain the folder, e.g. skin.tag_line_image = myFolder/myImage.jpg</td>
			<td>tag_line.png</td>
			<td>text</td>
		</tr>
		<tr>
			<td>skin.title</td>
			<td>sets the title of the site / name of the browser tab. E.g. when navigating to About Us with the default setting, the title becomes "cBioPortal for Cancer Genomics::About Us"</td>
			<td>cBioPortal for Cancer Genomics</td>
			<td>text</td>
		</tr>
		<tr>
			<td>skin.show_gsva</td>
			<td>enables gsva to query</td>
			<td>false</td>
			<td>true / false</td>
		</tr>
		<tr>
			<td>skin.geneset_hierarchy.default_p_value</td>
			<td>modifies default value for p_value threshold</td>
			<td>0.05</td>
			<td>0.05</td>
		</tr>
		<tr>
			<td>skin.geneset_hierarchy.default_gsva_score</td>
			<td>modifies default value for gsva_score threshold</td>
			<td>0.5</td>
			<td>0.5</td>
		</tr>
		<tr>
			<td>skin.geneset_hierarchy.collapse_by_default</td>
			<td>collapses the tree widget of the gsva hierarchy dialog on init</td>
			<td>false</td>
			<td>true / false</td>
		</tr>
		<tr>
            <td>skin.query.max_tree_depth</td>
            <td>sets the maximum number of subcategories shown in the query component hierarchy before each study. E.g. when set to 0, the hierarchy is flat, meaning only the study elements show in the component, without any tissue or cancer type sorting.</td>
            <td>3</td>
            <td>integer</td>
        </tr>
		<tr>
            <td>skin.patientview.filter_genes_profiled_all_samples</td>
            <td>sets default setting for the genes filter in patient view to only show mutations for genes that were profiled for mutations or CNA's in all samples of that patient. If unset, patient view will show mutations/CNA's for genes that were profiled in any sample.</td>
            <td>false</td>
            <td>true / false</td>
        </tr>
		<tr>
            <td>skin.hide_download_controls</td>
            <td>controls download options in UI. **true**: hides all download options. **false**:  shows all download options. **data**: disable data download but not figure download</td>
            <td>false</td>
            <td>true / false / data</td>
        </tr>
		<tr>
            <td>download_group</td>
            <td>controls download options in UI for each user. If present and user is authenticated, this value is checked against user roles.
If the download_group is present in user groups then download options are shown in UI, else it fallback to **skin.hide_download_controls**</td>
            <td></td>
            <td>text</td>
        </tr>
        <tr>
            <td>skin.show_settings_menu</td>
            <td>controls the appearance of the settings menu in study view and group comparison that controls annotation-based filtering.</td>
            <td>false</td>
            <td>true / false</td>
        </tr>
        <tr>
            <td>skin.hide_logout_button</td>
            <td>controls the appearance the logout button on a portal with user authentication.</td>
            <td>false</td>
            <td>true / false</td>
        </tr>
        <tr>
            <td>google_analytics_profile_id</td>
            <td>enables google analaytics tracking on your site</td>
            <td>disabled</td>
            <td>string</td>
        </tr>
   </tbody>
</table>

# Customization when using Docker
To add images or other files to the Docker image one can mount them inside the `cbioportal-webapp` folder. For example to use a custom logo for `skin.right_logo`. Mount the image inside the`cbioportal-webapp` folder with `-v /path/to/custom_logo.png:/cbioportal-webapp/images/custom_logo.png` and set `skin.right_logo=custom_logo.png` in your properties file.

# Creating your own local news/about or FAQ page

Creating you own local news/about or FAQ page involves three steps. For example, to create your own news page:

1. Copy `<cbioportal_source_folder>/portal/src/main/webapp/content/news.html` to `news_XXXX.html`
2. Modify `news_XXXX.html` as needed.
3. Edit the `application.properties` file and change the `skin.documentation.news` property, giving it the name of your news HTML component.

You may need to modify additional settings depending on whether you're e.g. using markdown pages. Please refer to the [cBioPortal Documentation](#cbioportal-documentation-settings) for more information.

# cBioPortal Documentation Settings
cBioPortal supports internal as well as external pages and supports html and markdown pages. 

## Using External Pages
cBioPortal can retrieve external pages and use them as news/faq/about page. For example, if your news page is stored on github wiki and reachable via https://github.com/cBioPortal/cbioportal/wiki/News.md use

```
skin.documentation.baseurl=https://github.com/cBioPortal/cbioportal/wiki/
skin.documentation.news=News.md
```

## Using Internal Pages
cBioPortal can use pages stored in (a subdirectory of) your content directory as news/faq/about page. For example, if your news page is called myNews.md and stored in content/myFolder, use 

```
skin.documentation.baseurl=/
skin.documentation.news=myFolder/myNews.md
```

## Markdown
By default, cBioPortal assumes all documentation is in markdown. If your documentation is not in markdown, please use

```
skin.documentation.markdown=false
```

If your documentation contains a relative link, cBioPortal will assume it uses the same format and, when clicked, will try to interpret the page as such. If your documentation contains an absolute link, cBioPortal will open it in a new window when clicked.
Please be aware that the links may be case-sensitive! E.g. https://github.com/cBioPortal/cbioportal/wiki/News.md is not the same as https://github.com/cBioPortal/cbioportal/wiki/news.md

This file has been depreciated and information about customizing cBioPortal can be found in [application.properties documentation](/deployment/customization/application.properties-Reference.md).# Application Properties

The following are the properties for configuring authentication and authorization in the application.

## Authentication Configuration

### General Authentication Settings

```properties
# authentication (available options: [false, oauth2, optional_oauth2, saml])
authenticate=false
```
### OAUTH2
#### NOTE for Custom Authorization (validate users via db) 
```properties
authenticate=oauth2
authorization=true
```
#### Google OAuth2 Client/Login Configuration

#### Example of utilizing google client for oAuth2 (Authentication)
```properties
spring.security.oauth2.client.registration.google.clientId=
spring.security.oauth2.client.registration.google.clientSecret=
spring.security.oauth2.client.provider.google.user-name-attribute=email
```

### Microsoft OAUTH2 Multi-tenant Client/Login Config
#### Example with Utilizing AzureAD for oAuth2
```properties
spring.security.oauth2.client.registration.azure.authorization-grant-type=authorization_code
spring.security.oauth2.client.registration.azure.client-id=<client-id>
spring.security.oauth2.client.registration.azure.clientSecret=<secret>
spring.security.oauth2.client.provider.azure.user-name-attribute=email
spring.security.oauth2.client.registration.azure.redirect-uri=http://localhost:8080/login/oauth2/code/azure
spring.security.oauth2.client.provider.azure.authorization-uri=https://login.microsoftonline.com/common/oauth2/v2.0/authorize
spring.security.oauth2.client.provider.azure.token-uri=https://login.microsoftonline.com/common/oauth2/v2.0/token
spring.security.oauth2.client.provider.azure.jwk-set-uri=https://login.microsoftonline.com/common/discovery/v2.0/keys
spring.security.oauth2.client.registration.azure.scope=openid,profile,email
```
#### Custom OAUTH2 Client Configuration

```properties
# For OIDC clients the issuer-uri is sufficient to autoconfigure the provider (via .well-known endpoint)
spring.security.oauth2.client.provider.cbio-idp.issuer-uri=http://localhost:8080/realms/cbioportal
spring.security.oauth2.client.provider.cbio-idp.user-name-attribute=email
# Required Scopes [openid, email, roles] 
spring.security.oauth2.client.registration.cbio-idp.scope=openid,email,roles
spring.security.oauth2.client.registration.cbio-idp.client-id=
spring.security.oauth2.client.registration.cbio-idp.client-secret=
```

##### Configuring the individual settings below is not recommended.
```properties
# spring.security.oauth2.client.provider.cbio-idp.authorization-uri=
# spring.security.oauth2.client.provider.cbio-idp.token-uri=
# TODO update docs, the user info endpoint must expose the roles !!
# spring.security.oauth2.client.provider.cbio-idp.user-info-uri=
# spring.security.oauth2.client.provider.cbio-idp.jwk-set-uri=
# spring.security.oauth2.client.provider.cbio-idp.logout-uri= # NOTE: this is not an official property.
# TODO Can be authorization_code, ...
#spring.security.oauth2.client.registration.cbio-idp.authorization-grant-type=
# TODO Can be client_secret_post, ...
#spring.security.oauth2.client.registration.cbio-idp.client-authentication-method=
#spring.security.oauth2.client.registration.cbio-idp.redirect-uri=<server-url>/login/oauth2/authorization/<client-registration-id (cbio-idp in this case)>
```

### SAML Configuration
#### Example to generate cert and key 
```shell
openssl req -newkey rsa:2048 -nodes -keyout local.key -x509 -days 365 -out local.crt
```

```properties

# For SAML 2.0
## SAML settings
# TODO add options for auto- and manual config to docs
# TODO add to docs: metadata-uri can be both URL or metadata xml file
spring.security.saml2.relyingparty.registration.cbio-saml-idp.assertingparty.metadata-uri=classpath:/client-tailored-saml-idp-metadata.xml
#spring.security.saml2.relyingparty.registration.cbio-saml-idp.assertingparty.metadata-uri=http://localhost:8080/realms/cbioportal/protocol/saml/descriptor
spring.security.saml2.relyingparty.registration.cbio-saml-idp.entity-id=cbioportal-saml
spring.security.saml2.relyingparty.registration.cbio-saml-idp.signing.credentials[0].certificate-location=classpath:/local.crt
spring.security.saml2.relyingparty.registration.cbio-saml-idp.signing.credentials[0].private-key-location=classpath:/local.key
spring.security.saml2.relyingparty.registration.cbio-saml-idp.singlelogout.binding=POST
# TODO add to docs (in minutes; default 1)
spring.security.oauth2.allowed-clock-skew=
```
### Data Access Token Settings

```properties

## data access token settings
## Resource Server issuer-uri for Data Access Token
#spring.security.oauth2.resourceserver.jwt.issuer-uri=http://localhost:8080/realms/cbioportal
# TODO: Currently Dat.method has only been tested with oauth2
dat.unauth_users=
dat.method=oauth2
dat.ttl_seconds=2592000
dat.uuid.max_number_per_user=1
dat.jwt.secret_key=
dat.filter_user_role=

# OAuth2 token data access settings (If using OAuth2 for Login can copy setting here)
## TODO: Reuse OAUTH2 Spring settings defined above
dat.oauth2.clientId=<client-id>
dat.oauth2.clientSecret=<client-secret>
dat.oauth2.issuer=<token-issuer>
dat.oauth2.accessTokenUri=<authorization-server-url>/.../token
dat.oauth2.userAuthorizationUri=<authorization-server-url>/.../auth
dat.oauth2.jwkUrl=<authorization-server-url>/.../certs
dat.oauth2.redirectUri=<cbioportal-url>/.../api/data-access-token/oauth2
spring.security.oauth2.resourceserver.jwt.issuer-uri=http://localhost:8081/realms/cbioportal

```
### Authorization Configuration/Study View Settings

```properties

## Authorization
## study view settings
## always show studies with this group
always_show_study_group=PUBLIC
## Should the permissions for groups and users be filtered by this instance's app.name?
## (true means the system only handles "CBIOPORTAL:someGroupPermission" groups, false means "someGroupPermission" works)
filter_groups_by_appname=false
# Can disable authorization
security.method_authorization_enabled=true
```

### CORS Configuration
To Enable CORS set the allowed-origins urls. (comma delimited list)
To enable all origins use *
```properties
security.cors.allowed-origins=*
##Or http://localhost:8080,http://localhost:8081
```# Study View Customization

## How does the study view organize the charts

Study view page is fully responsive, it will try to fit as many charts as possible based on the browser's width and height.

The layout of a chart is determined mainly based on priority. Higher priority will promote chart closer to the left-top.

In order to improve the layout, we added a layout algorithm layer. The study view page is using grid layout. All charts will be put into 2-dimensional systems. For example, pie chart, by default, takes 1 block and bar chart uses two blocks. All charts will be placed from left to right, top to bottom. In order to prevent misalignment, we promote small charts to fit into the space.

For logged-in(authenticated) users, charts layout is saved to users profile i.e, whenever user tries to re-visits the same url, previously saved layout will be loaded.
Please note for optimal performance, the portal will initially display 20 clinical charts by default. These charts are prioritized from high to low. Additional clinical charts are initially hidden but can be easily selected by users through the 'Add charts' dropdown menu.

## Study View Customization with Priority Data
Example of study view in public portal: https://www.cbioportal.org/study?id=acc_tcga,lgg_tcga#summary

## Priorities

The priority system is represented with a final score.
The higher the final (numeric) score, the higher priority assigned.

The default score is 1.

To promote certain chart in study view, please increase priority in the database to a certain number.
The higher the score, the higher priority it will be displayed in the study view.
If you want to hide chart, please set the priority to 0.
For combination chart, as long as one of the clinical attributes has been set to 0, it will be hidden.

To disable the chart, set the priority to -1.(Currently disables charts for single clinical attributes only)

Currently, we preassigned priority to few charts, but as long as you assign a priority in the database except than 1, these preassigned priorities will be overwritten.

| Chart name(clinical attribute ID)                          	| Frontend default priority 	| Additional Info                                   |
|------------------------------------------------------------	|---------------------------	|-------------------------------------------------	|
| CANCER_TYPE                                                	| 3000                      	|                                                 	|
| CANCER_TYPE_DETAILED                                       	| 2000                      	|                                                 	|
| Overall Survival Plot                                      	| 400                       	| This is combination of OS_MONTH and OS_STATUS   	|
| Disease Free Survival Plot                                 	| 300                       	| This is combination of DFS_MONTH and DFS_STATUS 	|
| Mutation Count vs. Fraction of Genome Altered Density Plot 	| 200                       	|                                                 	|
| Mutated Genes Table                                        	| 90                        	|                                                 	|
| CNA Genes Table                                            	| 80                        	|                                                 	|
| Cancer Studies                                             	| 70                        	|                                                 	|
| Number of Samples Per Patient                                 | 40                        	|                                                 	|
| Mutation Count Bar Chart                                   	| 30                        	|                                                 	|
| CNA Bar Chart                                              	| 20                        	|                                                 	|
| GENDER, SEX                                                	| 9                         	|                                                 	|
| AGE                                                        	| 8                         	|                                                 	|
# Building from Source

## Building with Maven

To compile the cBioPortal source code, move into the root directory and run the following maven command:

```
mvn -DskipTests clean install
```

Note: cBioPortal 6.X requires Java 21
# Deploying the Web Application

## Before running cbioportal backend

You will need to update the src/main/resources/applications.properties to include your DB connection information.

The configuration defined in `application.properties` can also be passed as command line arguments. The priority of property loading is as follows:

1. `-D` command line parameters overrides all
2. `src/main/resources/application.properties`
3. `application.properties` supplied at compile time
4. Defaults defined in code

Note that the `authenticate` property is currently required to be set as a command line argument, it won't work when set in `application.properties` (See issue [#6109](https://github.com/cBioPortal/cbioportal/issues/6109)).

Some scripts require a `${PORTAL_HOME}/application.properties` file, so it is best to define the properties there.

For more information about the `application.properties` file, see the [reference](/deployment/customization/Customizing-your-instance-of-cBioPortal.md) page.


## Run the cbioportal backend

To run the backend execute the following commabd

```
java -jar target/cbioportal-exec.jar
```

This runs the app in the foreground. If a port is already in use it will raise an error mentioning that. To change the port use the `--server.port` flag.



There are three main ways to run the portal: without authentication, with optional login, and with required login. All of them require the cBioPortal session service to be running.

### Without authentication

In this mode users are able to use the portal, but they won't be able to save their own virtual studies and groups.

```
java -jar target/cbioportal-exec.jar -Dauthenticate=false
```

### With authentication

To configure the authentication and authorization please consult the [Authorization](./../authorization-and-authentication/User-Authorization.md) and [Authentication](./../authorization-and-authentication/Authenticating-and-Authorizing-Users-via-keycloak.md) Sections.
# Importing the Seed Database

The next step is to populate your cBioPortal instance with all the required background data sets. This includes for example gene data, ID mappings, and network interactions. Rather than importing each of these data sets individually, we have provided a simple "seed" database that you can import directly.

## Download the cBioPortal Seed Database

A cBioPortal seed database for human can be found on [the datahub page](https://github.com/cBioPortal/datahub/blob/master/seedDB/README.md). If you are looking for mouse, check [this link](https://github.com/cBioPortal/datahub/blob/master/seedDB_mouse/README.md).

After download, the files can be unzipped by entering the following command:

    gunzip *.sql.gz

## Import the cBioPortal Seed Database

**Important:** Before importing, make sure that you have [followed the pre-build steps](Pre-Build-Steps.md) for creating the `cbioportal` database (see section "Create the cBioPortal MySQL Databases and User").

1. Import the database schema (/db-scripts/src/main/resources/cgds.sql):

    ```
    mysql --user=cbio_user --password=somepassword cbioportal < cgds.sql
    ```

    Note that this may currently fail when using the default character encoding on MySQL 8.0 (`utf8mb4`); this is why MySQL 5.7 (which uses `latin1`) is recommended.

2. Import the main part of the seed database:

    ```
    mysql --user=cbio_user --password=somepassword cbioportal < seed-cbioportal_RefGenome_vX.Y.Z.sql
    ```

    **Important:** Replace `seed-cbioportal_RefGenome_vX.Y.Z.sql` with the downloaded version of the seed database, such as `seed-cbioportal_hg19_v2.3.1.sql` or `seed-cbioportal_mm10_v2.3.1.sql`.

3. (Human only) Import the Protein Data Bank (PDB) part of the seed database. This will enable the visualization of PDB structures in the mutation tab. Loading this file takes more time than loading the previous files, and is optional for users that do not require PDB structures.

    ```
    mysql --user=cbio_user --password=somepassword cbioportal < seed-cbioportal_hg19_vX.Y.Z_only-pdb.sql
    ```
    **Important:** Replace `seed-cbioportal_hg19_vX.Y.Z_only-pdb.sql` with the downloaded version of the PDB database, such as `seed-cbioportal_hg19_v2.3.1_only-pdb.sql`.

4. (optional : support for microRNA genomic profiles) Import constructed gene table records for microRNA genomic profiles. Currently, cBioPortal supports the combined display of copy number alterations (generally reported for microRNA precursors) and expression (generally reported for microRNA mature forms) by adding gene table records which represent the combination of microRNA precursor and microRNA mature form. Appropriate aliases are added to the gene_alias table so that both the name of the precursor and the name of the mature form are recognized references to the combination.

This involves downloading the cBioPortal Core code located [here](https://github.com/cBioPortal/cbioportal-core). 

After the code has been successfully configured and built, you can import the needed microRNA records by running the following command from the cBioPortal core directory:

    java -cp scripts/target/scripts-*.jar org.mskcc.cbio.portal.scripts.ImportGeneData -microrna core/src/main/resources/micrornas.tsv
    

**Important:** Please be aware of the version of the seed database. In the [README on datahub](https://github.com/cbioportal/datahub/blob/master/seedDB/README.md), we stated which version of cBioPortal is compatible with the current seed database.

If the database is older than what cBioPortal is expecting, the system will ask you (during startup or data loading) to migrate the database to a newer version. The migration process is described [here](/Updating-your-cBioPortal-installation.md#running-the-migration-script).
# Loading a Sample Study

Once you have confirmed that the cBioPortal server is installed,
you are ready to import data. Importing a sample study is recommended
to verify that everything is working correctly.

cBioPortal Core has a [small dummy study, `study_es_0`](https://https://github.com/cBioPortal/cbioportal-core/tree/main/src/test/scripts/test_data/study_es_0), which contains all datatypes supported by cBioPortal. This document describes how to import the prerequisites for the sample study and how to import the study itself.


## Download and Build cBioPortal Core

```
	git clone https://github.com/cBioPortal/cbioportal-core.git
	cd cbioportal-core
	git checkout main
	mvn -DskipTests clean install
```



## Set the PORTAL_HOME environment variable

Most cBioPortal command-line tools, including the data loading pipeline,
expect the environment variable `$PORTAL_HOME` to point to a folder
containing the `application.properties` configuration file,
as explained during [the previous step](./Deploying.md).

Configure your shell to keep the variable set to the right folder.
On GNU/Linux and macOS this usually means appending a line
like the following to `.bash_profile` in your home directory:

```
export PORTAL_HOME=/Users/johndoe/cbioportal
```

## Import Gene Panel for Sample Study

The sample gene panel has to be imported before gene panel study data can be added to the database.

```
cd cbioportal-core/src/main/resources/scripts 
./importGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel1.txt
./importGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel2.txt
```

After loading gene panels into the database, please restart Tomcat or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information)
so that the validator can retrieve gene panel information from the cBioPortal API.

More details to load your own gene panel and gene set data can be found here: [Import Gene Panels](/Import-Gene-Panels.md).

## Validating the Sample Study

First it's useful to validate the study `study_es_0`, to check if the data is formatted correctly.

To do so, go to the importer folder: 

```
cd cbioprtal-core/src/main/scripts/importer
```

and then run the following command:

```
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -n
```

If all goes well, you should see the final output message:

```
Validation of study succeeded with warnings.
```

## Importing the Sample Study

To import the sample study:

```
cd cbioportal-core/src/main/scripts/importer
```

and then run the following command:

```
./metaImport.py -s ../../../test/scripts/test_data/study_es_0/ -n -o
```

You will see a series of output messages, hopefully ending with a status message like this:

```
Done.
Total time:  7742 ms
```

After loading the study data, please restart the app  or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information).
# Pre-Build Steps

## Get the Latest Code

Make sure that you have cloned the last code, and make sure you are on the `master` branch:
```
	git clone https://github.com/cBioPortal/cbioportal.git
	cd cbioportal
	git checkout master
```

## Prepare the log4j.properties File

This file configures logging for the portal.
An example file is available within GitHub:
```
    cd src/main/resources
    cp logback.xml.EXAMPLE logback.xml
```

If you don't create your own `logback.xml`, maven will copy the EXAMPLE file to that location when it builds.
If `logback.xml` already exists, it will just use that. This allows us to give you a working, versioned
log config, which you can then override easily.

To modify the logging during tests the same EXAMPLE file can be copied to the relevant test resources folder.

## Prepare the global configuration file

The portal is configured using a global configuration file, `application.properties`. An example file is available in the `src/main/resources` folder. Use it as a template to create your own:

```
cd src/main/resources
cp application.properties.EXAMPLE application.properties
```

For more information about the `application.properties` file, see the [reference](/deployment/customization/Customizing-your-instance-of-cBioPortal.md) page.

## Create the cBioPortal MySQL Databases and User

You must create a `cbioportal` database and a `cgds_test` database within MySQL, and a user account with rights to access both databases.  This is done via the `mysql` shell.
```
    > mysql -u root -p
    Enter password: ********

    Welcome to the MySQL monitor.  Commands end with ; or \g.
    Your MySQL connection id is 64
    Server version: 5.6.23 MySQL Community Server (GPL)

    Copyright (c) 2000, 2015, Oracle and/or its affiliates. All rights reserved.

    mysql> create database cbioportal;
    Query OK, 1 row affected (0.00 sec)

    mysql> create database cgds_test;
    Query OK, 1 row affected (0.00 sec)

    mysql> CREATE USER 'cbio'@'localhost' IDENTIFIED BY 'P@ssword1';
    Query OK, 0 rows affected (0.00 sec)

    mysql> GRANT ALL ON cbioportal.* TO 'cbio'@'localhost';
    Query OK, 0 rows affected (0.00 sec)

    mysql> GRANT ALL ON cgds_test.* TO 'cbio'@'localhost';
    Query OK, 0 rows affected (0.00 sec)

    mysql>  flush privileges;
    Query OK, 0 rows affected (0.00 sec)
```
# Software Requirements

This page describes various system software required to run the cBioPortal.

## MySQL

The cBioPortal software should run properly on MySQL version 5.7.x. Versions higher than 5.7.x can cause an issue
while loading the database schema. Minor versions lower than 5.7.x will cause issues with persistent cache invalidation.
The software can be found and downloaded from the [MySQL website](https://www.mysql.com/).

On Ubuntu:  ```sudo apt-get install mysql-server```

## MongoDB

The session service uses MongoDB 3.6.6

## Java

cBioPortal requires Java 12 and above.  The software can be found and download from the [Oracle](https://www.oracle.com/us/technologies/java/overview/index.html) website.

On Ubuntu:  ```sudo apt-get install default-jdk```

## Apache Maven

The cBioPortal source code is an [Apache Maven](https://maven.apache.org/) driven project.  The software needs to be downloaded and installed prior to building the application from source code.  It can be found on the [Apache Maven](https://maven.apache.org/download.cgi) website. We are currently using version 3.9.5.

On Ubuntu:  ```sudo apt-get install maven```

## Git

You will need a git client to download the cBioPortal source code.

On Ubuntu:  ```sudo apt-get install git```
### Importing gene panel ###

Use this command to import a gene panel. Specify the gene panel file by replacing 
`<path_to_genepanel_file>` with the absolute path to the gene panel file. Another option is to add the 
gene panel files in `./study` which is mounted inside the container on `/study/.

```shell
docker-compose run \
    -v <path_to_genepanel_file>:/gene_panels/gene_panel.txt:ro \
    cbioportal \
    bash -c 'cd /cbioportal/core/src/main/scripts/ && ./importGenePanel.pl --data /gene_panels/gene_panel.txt'
```

### Importing data ###

Use this command to validate a dataset. Add the study to the `./study` folder. The
command will connect to the web API of the container `cbioportal-container`, and import 
the study in its associated database. Make sure to replace `<path_to_report_folder>` with 
the absolute path were the html report of the validation will be saved.

```shell
docker-compose run \
    -v "<path_to_report_folder>:/report" \
    cbioportal \
    metaImport.py -u http://cbioportal:8080 -s /study/name_of_study --html=/report/report.html
```
:warning: after importing a study, remember to restart `cbioportal-container`
to see the study on the home page. Run `docker-compose restart cbioportal`.

#### Using cached portal side-data ####

In some setups the data validation step may not have direct access to the web API, for instance when the web API is only accessible to authenticated browser sessions. You can use this command to generate a cached folder of files that the validation script can use instead. Make sure to replace `<path_to_portalinfo>` with the absolute path where the cached folder is going to be generated.

```shell
docker-compose run \
    -v "<path_to_portalinfo>/portalinfo:/portalinfo" \
    -w /cbioportal/core/src/main/scripts \
    cbioportal \
    ./dumpPortalInfo.pl /portalinfo
```

Then, grant the validation/loading command access to this folder and tell the script it to use it instead of the API:

```shell
docker-compose run \
    -v "<path_to_report_folder>:/report" \
    -v "<path_to_portalinfo>/portalinfo:/portalinfo:ro" \
    cbioportal \
    metaImport.py -p /portalinfo -s /study --html=/report/report.html
```

### Inspecting or adjusting the database ###

```shell
docker-compose run cbioportal_database \
    sh -c 'mysql -hcbioportal_database -u"$MYSQL_USER" -p"$MYSQL_PASSWORD" "$MYSQL_DATABASE"'
```

### Deleting a study ###

To remove a study, run:

```shell
docker-compose run \
    cbioportal \
    cbioportalImporter.py -c remove-study -id study_id
```

Where `study_id` is the `cancer_study_identifier` of the study you would like to remove.
# Import data with Docker

### Import data instructions

This is an example to import a sample study: `study_es_0`. When trying to import other studies, please follow the same routine:

1. import gene panels (if applicable, studies without gene panels are assumed to be whole exome/genome)
2. import study data

#### Step 1 - Import gene panels

To import gene panels for your study, please reference the example commands in [this file](example\_commands.md#importing-gene-panel)

These are the commands for importing `study_es_0` gene panels (`data_gene_panel_testpanel1` and `data_gene_panel_testpanel2`):

```shell
docker-compose run \
    cbioportal \
    bash -c 'cd /cbioportal/core/src/main/scripts/ && ./importGenePanel.pl --data /cbioportal/core/src/test/scripts/test_data/study_es_0/data_gene_panel_testpanel1.txt'
```

```shell
docker-compose run \
    cbioportal \
    bash -c 'cd /cbioportal/core/src/main/scripts/ && ./importGenePanel.pl --data /cbioportal/core/src/test/scripts/test_data/study_es_0/data_gene_panel_testpanel2.txt'
```

#### Step 2 - Import data

To import data for your study, please reference the example commands in [this file](example\_commands.md#importing-data)

Command for importing `study_es_0` data:

```shell
docker-compose run cbioportal metaImport.py -u http://cbioportal:8080 -s /cbioportal/core/src/test/scripts/test_data/study_es_0 -o
```

:warning: after importing a study, remember to restart `cbioportal` to see the study on the home page. Run `docker-compose restart cbioportal`.

You have now imported the test study `study_es_0`. Note that this study is included inside the cbioportal container. The process for adding a study that is outside of the container is similar. Just make sure to add the data files in the `./study` folder. This folder is mounted as `/study/` inside of the container.

### Frequently Asked Questions

#### Gene panel ID is not in database

If you see an error like this when you importing the data:\
`ERROR: data_gene_panel_matrix.txt: lines [2, 3, 4, (10 more)]: Gene panel ID is not in database. Please import this gene panel before loading study data.; values encountered: ['TESTPANEL1', 'TESTPANEL2']`

please follow the first step to import gene panels (e.g. import `data_gene_panel_testpanel1` and `data_gene_panel_testpanel2` for `study_es_0`), then try to import the data again.

#### Error occurred during validation step

Please make sure the seed database was correctly imported.

#### Study imported correctly, but got error when trying to query something

Remember to restart the `cbioportal` after data imported.

```shell
docker-compose restart cbioportal
```

#### Import GRCh38 data

If you are importing GRCh38 data, please remember to set the `reference_genome: hg38` field in the `meta_study.txt` file. See also [cancer study metadata](/File-Formats.md).
#### Notes for non-Linux systems

##### Docker for Mac/Windows (newer versions)
Make sure to assign enough memory to Docker when using Docker for Windows (Windows 10 Pro 64-bit) or Docker for Mac (Mac OS X Yosemite 10.10.3 or above). In Mac OS X this can be set when clicked on the Docker icon -> Preferences... -> Adjust the Memory slider. By default it's set to 2 GB, which is too low and causes problems when loading multiple studies.

##### Docker-machine (older versions)
Because the Docker Engine daemon uses Linux-specific kernel features, you can‚Äôt run Docker Engine natively in Windows or Mac OS X. In versions of these systems that do not support the newer lightweight virtualisation technologies mentioned above, you must instead use the Docker Machine command, `docker-machine`. This creates and attaches to a small Linux VM on your machine, which hosts Docker Engine.

The Docker Quickstart Terminal in the Docker Toolbox will automatically create a default VM for you (`docker-machine create`), boot it up (`docker-machine start`) and set environment variables in the running shell to transparently forward the docker commands to the VM (`eval $(docker-machine env)`). Do note however, that forwarded ports in the docker commands will pertain to the VM and not your Windows/OS X system. The local cBioPortal and MySQL servers will not be available on `localhost` or `127.0.0.1`, but on the address printed by the command `docker-machine ip`, unless you configure VirtualBox to further forward the port to the host system.
# Deploy with Docker

## Prerequisites

Docker provides a way to run applications securely isolated in a container, packaged with all its dependencies and libraries.
To learn more on Docker, kindly refer here: [Docker overview](https://docs.docker.com/engine/docker-overview/).

Make sure that you have the latest version of Docker installed on your machine. [Get latest version](https://www.docker.com/products/overview#/install_the_platform)

[Notes for non-Linux systems](notes-for-non-linux.md)

## Usage instructions
In this example we use [Docker Compose](https://docs.docker.com/compose/) to spin up all the different required containers/services for cBioPortal.

### Quick Start

```
git clone https://github.com/cBioPortal/cbioportal-docker-compose.git
cd cbioportal-docker-compose
./init.sh
docker compose up
```

You should now be able to see the cBioPortal website at http://localhost:8080

Import studies with:

```
docker compose run cbioportal metaImport.py -u http://cbioportal:8080 -s study/lgg_ucsf_2014/ -o
```

Clear persistent data volumes with:

```
docker compose down -v
```

### Comprehensive Start

#### Step 1 - Run Docker Compose

Download the git repo that has the Docker compose file and go to the root of that folder:

```
git clone https://github.com/cBioPortal/cbioportal-docker-compose.git
cd cbioportal-docker-compose
```

Then download all necessary files (seed data, example config and example study from datahub) with the init script:
```
./init.sh
```

Then run:

```
docker compose up
```

This will start all four containers (services) defined [here](https://github.com/cBioPortal/cbioportal-docker-compose/blob/master/docker-compose.yml). That is:

- the mysql database, which holds most of the cBioPortal data
- the cBioPortal Java web app, this serves the React frontend as well as the REST API
- the session service Java web app. This service has a REST API and stores session information (e.g. what genes are being queried) and user specific data (e.g. saved cohorts) in a separate mongo database
- the mongo database that persists the data for the session service

It will take a few minutes the first time to import the seed database and perform migrations if necessary. Each container outputs logs to the terminal. For each log you'll see the name of the container that outputs it (e.g. `cbioportal_container` or `cbioportal_session_database_container`). If all is well you won't see any significant errors (maybe some warnings, that's fine to ignore). If all went well you should be able to visit the cBioPortal homepage on http://localhost:8080. You'll notice that no studies are shown on the homepage yet:

<img width="1414" alt="Screen Shot 2022-01-24 at 2 10 10 PM" src="https://user-images.githubusercontent.com/1334004/150848276-dec9551f-6b90-470f-bb59-e7754829fc83.png">


Go to the next step to see how to import studies.

##### Notes on detached mode

If you prefer to run the services in detached mode (i.e. not logging everything to your terminal), you can run

```
docker compose up -d
```

In this mode, you'll have to check the logs of each container manually using e.g.:

```
docker logs -f cbioportal_container
```

You can list all containers running on your system with

```
docker ps -a
```

#### Step 2 - Import Studies
To import studies you can run:

```
docker compose run cbioportal metaImport.py -u http://cbioportal:8080 -s study/lgg_ucsf_2014/ -o
```
This will import the [lgg_ucsf_2014 study](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014) into your local database. It will take a few minutes to import. After importing, restart the cbioportal web container:

```
docker compose restart cbioportal
```

or 

All public studies can be downloaded from https://www.cbioportal.org/datasets, or https://github.com/cBioPortal/datahub/. You can add any of them to the `./study` folder and import them. There's also a script (`./study/init.sh`) to download multiple studies. You can set `DATAHUB_STUDIES` to any public study id (e.g. `lgg_ucsf_2014`) and run `./init.sh`.

##### Notes on restarting
To avoid having to restart one can alternatively hit an API endpoint. To do so, call the `/api/cache` endpoint with a `DELETE` http-request (see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information):

```
curl -x DELETE -H "X-API-KEY: my-secret-api-key-value" http://localhost:8080/api/cache
```

The value of the API key is configured in the _application.properties_ file. You can visit http://localhost:8080 again and you should be able to see the new study.

#### Step 3 - Customize your application.properties file ###

The properties file can be found in `./config/application.properties`. Which was set up when running `init.sh`.

This properties file allows you to customize your instance of cBioPortal with e.g. custom logos, or point the cBioPortal container to e.g. use an external mysql database. See the [properties](/deployment/customization/Customizing-your-instance-of-cBioPortal.md) documentation for a comprehensive overview.

If you would like to enable OncoKB see [OncoKB data access](/deployment/integration-with-other-webservices/OncoKB-Data-Access.md) for 
how to obtain a data access token. After obtaining a valid token use:

#### Step 4 - Customize cBioPortal setup
To read more about the various ways to use authentication and parameters for running the cBioPortal web app see the relevant [backend deployment documentation](/deployment/customization/Customizing-your-instance-of-cBioPortal.md).

On server systems that can easily spare 4 GiB or more of memory, set the `-Xms`
and `-Xmx` options to the same number. This should increase performance of
certain memory-intensive web services such as computing the data for the
co-expression tab. If you are using MacOS or Windows, make sure to take a look
at [these notes](notes-for-non-linux.md) to allocate more memory for the
virtual machine in which all Docker processes are running.

## More commands ##
For documentation on how to import a study, see [this tutorial](import_data.md)
For more uses of the cBioPortal image, see [this file](example_commands.md)

To Dockerize a Keycloak authentication service alongside cBioPortal,
see [this file](using-keycloak.md).

## Uninstalling cBioPortal ##
```
docker compose down -v --rmi all
```
# Authenticating and Authorizing Users using Keycloak in Docker

This guide describes a way to Dockerise Keycloak along with cBioPortal, for authentication.

First, create an isolated network in which the Keycloak and MySQL servers can talk to one another.

```shell
docker network create kcnet
```

Run a MySQL database in which Keycloak can store its data. This database server will not be addressable from outside the Docker network. Replace `<path_to_database>` with the absolute path where the folder `kcdb-files` will be placed. This folder is used by the database to store its files.

```shell
docker run -d --restart=always \
    --name=kcdb \
    --net=kcnet \
    -v "<path_to_database>/kcdb-files:/var/lib/mysql" \
    -e MYSQL_DATABASE=keycloak \
    -e MYSQL_USER=keycloak \
    -e MYSQL_PASSWORD=password \
    -e MYSQL_ROOT_PASSWORD=root_password \
    mysql:5.7
```

Then run the actual Keycloak server, using [this image](https://hub.docker.com/r/jboss/keycloak/) available from Docker Hub. This will by default connect to the database using the (non-root) credentials in the example above. The server will be accessible to the outside world on port 8180, so make sure to choose a strong administrator password.

The command below uses the default values for `MYSQL_DATABASE`, `MYSQL_USER` and `MYSQL_PASSWORD` (listed in the command above). If you wish to change these credentials, specify them in the command below. For instance, if `MYSQL_USER` in the database container is `user`, you need to add `-e MYSQL_USER=user`.

```
docker run -d --restart=always \
    --name=cbiokc \
    --net=kcnet \
    -p 8180:8080 \
    -e DB_VENDOR=mysql \
    -e DB_ADDR=kcdb \
    -e KEYCLOAK_USER=admin \
    -e "KEYCLOAK_PASSWORD=<admin_password_here>" \
    jboss/keycloak:4.8.3.Final
```

Finally, configure Keycloak and cBioPortal as explained in the [Keycloak documentation](./../authorization-and-authentication/Authenticating-and-Authorizing-Users-via-keycloak.md). Remember to specify port 8180 for the Keycloak server, wherever the guide says 8080.

After configuring Keycloak, set up cBioPortal containers [as specified in the documentation](./). Make sure to update the `-Dauthenticate` in the [docker-compose file](https://github.com/cBioPortal/cbioportal-docker-compose/blob/5da068f0eb9b4f42db52ab5e91321b26a1826d7a/docker-compose.yml#L20) to `-Dauthenticate=saml`.
# Introduction

Using OncoKB does not require a data access token. But the instance you are connecting to only includes biological information by default. 
If you want to include tumor type summary, therapeutic levels and more, please consider obtaining a license from OncoKB.

# How to obtain an OncoKB license
1. Please review OncoKB [terms](https://www.oncokb.org/terms) 
2. Please request for [data access](https://www.oncokb.org/dataAccess)
3. You can find your token in your [Account Settings](https://www.oncokb.org/account/settings) after login.

# Set up cBioPortal to include full OncoKB content
Following properties can be edited in the `application.properties` file or set in system variables if you are using docker. 
- `show.oncokb` should be set to `true`
- `oncokb.token` should be set to a valid OncoKB access token value
- `oncokb.public_api.url` should be set to `https://www.oncokb.org/api/v1`

Thank you for supporting future OncoKB development.

# Include MSI-H and TMB-H annotation
If you want to include the MSI-H and TMB-H annotation on patient view, please follow the instruction to [import required clinical data](/File-Formats.md#clinical-data).  
For MSI-H, a clinical attribute MSI_TYPE with value Instable is required.  
For TMB-H, a clinical attribute TMB_SCORE with value >=10 is required.  

# Disable OncoKB Service

Please set `show.oncokb` to `false` in `application.properties` or in system variables if you are using docker.
# Integration with Other Webservices

# Deployment

Private instances of cBioPortal are maintained by institutions and companies [around the world](https://www.cbioportal.org/installations).

An instance can be deployed using Docker **(recommended)** or by building and deploying from source. The source code of cBioPortal is [available](https://github.com/cBioPortal/cbioportal) on GitHub under the terms of [Affero GPL V3](https://www.gnu.org/licenses/agpl-3.0.en.html).

This section contains instructions for both of these paths.

Please note that installing a local version requires system administration skills; for example, installing and configuring Tomcat and MySQL. With limited resources, we cannot provide technical support on system administration.
# Build with different frontend versions

Maven will build cBioPortal with a cBioPortal-frontend version and git repository url as determined by respectively the _frontend.version_ and _frontend.groupId_ parameters in the root POM.xml.

```
  <properties>
    <frontend.version>v2.1.0_/frontend.version>
    <frontend.groupId>com.github.cbioportal</frontend.groupId>
    ...
```
To build cBioPortal with a different frontend version different values for  _frontend.version_ and _frontend.groupId_ parameters can be specified as part of the maven install command. For example:

```
mvn clean -DskipTests install -Dfrontend.version=93d9cbcbf007ff620ab51ef5af5927a0eb1ebed4 -Dfrontend.groupId=com.github.thehyve
```

Remarks:
- The _frontend.version_ parameter allows release tags (e.g. 'v2.1.0') and commit sha-hashes (e.g., '93d9cbcb').
- The _frontend.groupId_ is a reversed, dot-separated derivative of the git url. Git repository location _github.com/cbioportal_ is represented by the _com.github.cbioportal_ groupId.# cBioPortal ER Diagram
[cBioPortal ER Diagram - PDF Version](https://github.com/cBioPortal/cbioportal/blob/master/db-scripts/src/main/resources/cbioportal-er-diagram.pdf)

![cBioPortal ER Diagram](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/db-scripts/src/main/resources/cbioportal-er-diagram.png)
# cBioPortal Database Versioning

We follow the following logic when deciding how/when to increment the version of cBioPortal database. It's a complete modification of semantic versioning (MAJOR.MINOR.PATCH) more suitable for our purposes:

**MAJOR** : A non-backward compatible significant change in the database. Which requires the maintainer to reload and re-import all studies in the database entirely.

**MINOR** : Including deleting, renaming tables or columns, or changing constraints

Delete a table

```SQL
DROP TABLE t ;
```

Drop column c from the table

```SQL
ALTER TABLE t DROP COLUMN c ;
```

Add a constraint

```SQL
ALTER TABLE t ADD constraint;
```

Drop a constraint

```SQL
ALTER TABLE t DROP constraint;
```

Rename a table from t1 to t2

```SQL
ALTER TABLE t1 RENAME TO t2;
```

Rename column c1 to c2

```SQL
ALTER TABLE t1 RENAME c1 TO c2 ;
```

**PATCH** : Changes that don't change existing database schemes but add new tables or columns, manipulating data.

Create a new table

```SQL
CREATE TABLE t (
     id INT PRIMARY KEY,
);
```

Add a new column to the table

```SQL
ALTER TABLE t ADD column;
```

Remove all data in a table

```SQL
TRUNCATE TABLE t;
```

Delete data from a table:

```SQL
DELETE from TABLE t;
```

Update data in a table:

```SQL
UPDATE TABLE t SET ...;
```

Insert multiple rows into a table

```SQL
INSERT INTO t(column_list)
VALUES (value_list), 
       (value_list), ‚Ä¶;
```
# Deployment Procedure
This describes our internal deployment procedure. Shared publicly, in case it
may be of use. Instructions on how to deploy cBioPortal can be found elsewhere,
see e.g. [Deploying the web application](/deployment/deploy-without-docker/Deploying.md) and [Deploy using
Docker](/deployment/docker/).

We deploy the master branch of backend and the master branch of frontend to
production. The public portal (https://www.cbioportal.org) runs on AWS EKS. The configuration 
can be found in the knowledgesystems repo:

https://github.com/knowledgesystems/knowledgesystems-k8s-deployment 

Other internal MSK portals run on AWS EKS as well.

The frontend and backend can be upgraded independently. We have the following
events that can require a new deployment:

- New **frontend** commit in master
- New **backend** commit in master

## New frontend commit in master
Currently we are auto-deploying the frontend master branch to netlify:
https://frontend.cbioportal.org. So any change should be automatically
built and deployed to the relevant portals if the frontend configuration has
been set up properly. Do note that the current build time for the frontend
project is ~15 minutes or so. To see what frontend commit is deployed, check
`window.FRONTEND_COMMIT` in the console of the browser.

### Public Portal Frontend URL
The public portal is on AWS and running inside a Kubernetes cluster.
The URL that it gets the frontend version from is here:

https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/search?q=-Dfrontend.url&unscoped_q=-Dfrontend.url

This should be a URL pointing to netlify.

### Internal Portal Frontend URL
For the internally runnning portals the frontend.url is defined in the
application.properties file in the mercurial portal-configuration repo. If set up
correctly, this should point to a file on both dashi and dashi2 that in turn
points to a netlify frontend URL. The reason we have a separate file with the
URL in it is that it allows us to update the frontend URL without redeploying
the backend.

## New backend commit in master
A new backend commit usually also means a new frontend change is necessary. For
this reason the following sections assume that's the case.

### Public Portal Backend Upgrade
Once the backend repo has been tagged on github, a docker image gets build on Docker Hub automatically. It can take ~5 min before the image is available. You can check here what the status of the builds is: https://github.com/cBioPortal/cbioportal/actions?query=workflow%3A%22Docker+Image+CI%22.


After that, if you have access to the kubernetes cluster you can change the image in the configuration of the kubernetes cluster:

https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/public-eks/cbioportal-prod/cbioportal_spring_boot.yaml

point this line, to the new tag on docker hub e.g.:

```
image: cbioportal/cbioportal:6.0.2-web-shenandoah
```

Make sure it is an image with the postfix `-web-shenandoah`. This is the image that only has the web part of cBioPortal and uses the shenandoah garbage collector.

Also remove the `-Dfrontend.url` parameter such that the frontend version inside the war will be used:

```
"-Dfrontend.url=https://frontend.cbioportal.org"
```

Then running this command applies the changes to the cluster:

```
kubectl apply -f public-eks/cbioportal-prod/cbioportal_spring_boot.yaml
```

You can keep track of what's happening by looking at the pods:

```
kubectl get po
```

If you have the watch command installed you can also use that to see the output
of this every 2s:

```
watch kubectl get po
```

Another thing to look at is the events:

```
kubectl get events --sort-by='{.lastTimestamp}'
```

If there are any issues, point the image back to what it was, set
`-Dfrontend.url` and run `kubectl apply -f filename` again.

If everything went ok, you can re-enable auto deployment on netlify, set
`-Dfrontend.url` in the kubernetes file and run `kubectl apply -f filename`
again.

Make sure to commit your changes to the knowledgesystems-k8s-deployment repo
and push them to the main repo, so that other people making changes to the
kubernetes config will be using the latest version. 


## Upgrading Related Backend Components
Backend upgrades involving the database schema, DAO classes, etc. require updates to databases and importers. CBioPortal has multiple MySQL databases (all using AWS RDS) backing different portals. Similarly, there are multiple importers responsible for loading portal-specific data. Every database must be manually migrated on an individual basis; all importers/data fetchers can be updated simultaneously through an existing deployment script.

Before upgrading, make sure to turn off import jobs in the crontab and alert the backend pipelines team (Avery, Angelica, Rob, Manda). 

To access the crontab, log in to pipelines, log in as cbioportal_importer: `sudo su - cbioportal_importer`, and run `crontab -e`. Comment out any lines that run import jobs, save, and exit. Make sure to uncomment these lines once the upgrade (database and importers) is complete. Lines that need to be commented out will be under the `Import Jobs` section, shown [here](https://github.com/knowledgesystems/cmo-pipelines/blob/942de83c0f9a731e301151d10dad73744cd9c9a0/import-scripts/mycrontab#L4).

## Updating Databases
First, make sure there is a backup of the database being migrated. 
If there is not a weekly dump, backup the database being migrated using mysqldump. This process may take awhile depending on the size of the database. 

```
mysqldump -u <user> -h <host> -p <database name> | gzip > <database_name>_`date +%Y%m%d_%H%M`.sql.gz 
```
    
The second step is to migrate the database. Make sure that the migration script is the same version as the deployed cBioPortal website. It is recommended to first test the migration script manually line-by-line in a copy of the existing database. This will catch any data-related bugs that might not be captured by the python migration script. After testing is successful, migrate the production databases following these steps [here](/Updating-your-cBioPortal-installation.md#running-the-migration-script). 

These are all cBioPortal databases and their locations:

| Website  | Database | Location |
| ------------- | ------------- | ------------- |
| cbioportal.mskcc.org  | cgds_gdac  | pipelines |
| cbioportal.org  | cgds_public  | AWS |
| genie.cbioportal.org | cgds_genie | AWS | 
| triage.cbioportal.org | cgds_triage | pipelines |

To obtain information such as usernames, passwords, hostnames - ask Avery, Angelica, Rob, Manda, and Ino. 

## Updating Importers/Data Fetchers
Importers (code found [here](https://github.com/knowledgesystems/pipelines)) and data fetchers (code found [here](https://github.com/knowledgesystems/cmo-pipelines)) use code from the cBioPortal codebase. The cbioportal dependency is packaged with the genome-nexus-annotation-pipeline and specified in the pipelines importer pom.

The following steps are used during releases/updates to build new importers with the **most-up-to-date** cBioPortal and genome-nexus-annotation-pipeline code. All steps should be performed on the pipelines machine. 

1. Set the jitpack hash [here](https://github.com/genome-nexus/genome-nexus-annotation-pipeline/blob/9510299395986653d9e9b672a38d472e52e7625b/pom.xml#L71) in the genome-nexus-annotation-pipeline codebase to the most recent cbioportal/cbioportal commit hash in master.

2. Merge this change into genome-nexus-annotation-pipeline/master.

3. Set the commit hash [here](https://github.com/knowledgesystems/pipelines/blob/f6c52bbda86b3929222d42c9bc84581fd6333fb4/pom.xml#L76) in the pipelines codebase to the most most recent genome-nexus/genome-nexus-annotation-pipeline commit hash **(after merge specfied in step 2)**. Also ensure the db version in the pom [here](https://github.com/knowledgesystems/pipelines/blob/f6c52bbda86b3929222d42c9bc84581fd6333fb4/pom.xml#L76) matches the db schema version in the cbioportal codebase. 

4. Merge this change into pipelines/master.

5. Set the commit hash [here](https://github.com/knowledgesystems/cmo-pipelines/blob/e740c9fa3d409ab75988e7a157682733e261fca5/cvr/pom.xml#L70) in the cmo-pipelines codebase to the most recent genome-nexus/genome-nexus-annotation-pipeline commit hash **(after merge specified in step 2)**

6. Merge this change into cmo-pipelines/master

7. Run the deployment wrapper script. See details [here](/Deployment-Procedure.md#deployment-script).  

8. Verify new importers/data fetchers have been placed in `/data/portal-cron/lib` by checking timestamps.
```
ls -tlra /data/portal-cron/lib
```

## Deployment Script
The wrapper script is found on pipelines here:
`/data/portal-cron/git-repos/pipelines-configuration/build-importer-jars/buildproductionjars.sh`.

Run `git pull` to pull in any updates to the build script.

The wrapper script takes two arguments:
1. --cbioportal-git-hash (required): Set to the cBioPortal commit hash being used in the pipelines build (hash specified in **step 1** of [updating importers](#updating-importersdata-fetchers)). This must match because the build copies out resource files (e.g application-context-business.xml) from the cbioportal codebase. 
2. --skip-deployment (optional): Set to true to skip auto-deployment to `/data/portal-cron/lib`. Built jars will be found in `/data/portal-cron/git-repos/pipelines-configuration/build-importer-jars/` and can be be manually moved.

The wrapper script will automatically backup the importers/data-fetchers to `/data/portal-cron/lib/backup`.
# Documentation site

This documentation site is created using [https://retype.com/](https://retype.com/), a static site generator based on markdown.

Visit their site for installation instructions and a guide
on how
- Develop, build and test the site locally
- Markdown syntax
- Deploy to Github pages using Github actions.

### Navigation
Please note that the navigation for the site is defined in docs/SUMMARY.md.
This is a deprecated form of configuration which is not documented in Retype.
# Feature Development Guide

This is a guide for developers that are implementing a new feature.

## Before Implementation

As a first step it is important to determine how complex the proposed feature is. Incremental improvements on existing features are often easier to accomplish and require input from fewer people. Most minor changes can be submitted as a Pull Request. If the proposed feature would require one or more days of work it makes sense to connect on [slack](https://slack.cbioportal.org) to discuss the idea. For more complex new features that require weeks of work or more, it is best to get input from several people in the cBioPortal community, including people with a deep understanding of the cBioPortal product and its users as well as the engineers that write the software. In that case we often start out with a Request For Comments document that describes the feature in more detail, see [our list of RFCs ](../RFC-List.md)for some examples. The community can then help guide the feature development in the right direction.

During this process you will most likely receive some pointers which part of the stack you will be editing (see [Architecture Overview](../Architecture-Overview.md)). This will be helpful when actually starting your implementation and figuring out how to set up your development environment. For many features it is not necessary to understand all parts of the stack, so seeking out advice on this is highly recommended.

Before you start implementing a more complex feature, ideally many of these things are clear:

* Who can you contact for help?
* Who will be helping to review the code?
* What part of the stack will you work on?

Gold stars if you already start thinking about:

* How do we release the feature incrementally?
* When is feature development done?

See more thoughts about these topics further below

## Starting Implementation

Once you are ready to start implementing, the first thing is to set up the development environment. We strive to make this as easy as possible, but it can often still take some time so definitely reach out if you run into issues. If you haven't submitted a Pull Request to cBioPortal before, it might make sense to look at some [good first issues](https://github.com/cBioPortal/cbioportal/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) before starting with your feature. This will help to get some familiarity with the process of proposing a change, getting it reviewed, making edits and getting it to production. Don't pick anything to complicated for a first issue, it could be as simple as fixing some typos in the `README`.

## During Implementation

The most important part during implementation is communication. Continue getting feedback as your implementation evolves. One of the best ways to do this is to fully integrate into the development team while you work on a feature. Anybody is welcome to join our weekly planning meeting (Tuesdays 11AM-12PM Eastern Time) and our daily standups 2.30-3PM Eastern Time. Please reach out on [slack](https://slack.cbioportal.org) to get an invite. If these times are not ideal or you're working on the feature more sporadically then it's totally fine to skip them.

## Plan to release to production early

Don't wait until the feature is fully finished to get feedback from the product team and the engineering team. Think about ways we can release a portion of the feature to production without finishing the entire thing. We have found feature flags to work well here. Instead of using long running feature branches we try to add a configuration property that allows us to turn the feature on or off. That way portions of the code can be released to production early on. We want to avoid working on some piece of code for more than a week or so without being able to release it. For instance: if one is trying to add some new tab on the Patient View Page, one could start with adding the on/off configuration switch for this tab. That could be released to production relatively quickly.

## Regression Testing

Make sure to think about ways to incorporate testing for your feature. We have an extensive suite of unit, integration and end to end tests including automated browser testing that mimics user interaction. Adding some regression tests will make sure the feature won't break with new versions of cBioPortal.

## When is feature development done?

There are many stages in feature development:

* Design
* Implementation
* Review
* Production Deployment
* Production Usage Monitoring

The process is hardly ever a linear line and it can move back and forth between any stage. This is expected and one of the reasons why time estimations are notoriously hard. An additional note is that feature development usually isn't done at the moment it gets merged to the main branch and deployed, but rather only after a few weeks of using it in production and not identifying any new issues.
# Manual test cases

Linked below is a Google Doc with some manual test cases you might want to look at when developing, and deploying new changes. Note that this list is a work in progress and subject to change.

<https://docs.google.com/document/d/1IniMNrrSEGsPRtkaU3Vbcm6sMgcOhytBurIRg0SqXgA/edit?usp=sharing>
# Development

cBioPortal is an open source project hosted on [Github](https://www.github.com/cbioportal/cbioportal).  The core team will
consider pull requests from any source.

The following section describes how to set up a development environment and start contributing.

Note that cBioPortal is split into a backend project (api) and a frontend project which are kept in separate repositories.

- [Frontend](http://www.github.com/cbioportal/cbioportal-frontend)
  (Typescript, React, Mobx)
- [Backend](http://www.github.com/cbioportal/cbioportal)
  (Java, Spring, MySQL, MongoDB)

If you are interested in coordinating the development of new features, please contact cbioportal@cbioportal.org or reach out on https://slack.cbioportal.org.
# Release Procedure

We have release procedures for the following scenarios:

1. [cBioPortal community release of code already in production](Release-Procedure.md#cbioportal-community-release-of-code-already-in-production)
2. [Release with database migration](Release-Procedure.md#release-with-database-migration)

## cBioPortal community release of code already in production

We often run code in production that is not ready yet for use by the wider cBioPortal community. The frontend gets deployed to production after every merge to master. The backend gets updated every Tuesday before our community call (occasionally more frequently too if issues are identified). We tag whatever is in the master and frontend repo and put a "Pre-Release" indication on it. After a month of stable usage in production, one of the tags will get the "latest" indication on the GitHub. These are the steps we follow to create a release:

1. Create a new frontend tag. The releases can be found here: https://github.com/cBioPortal/cbioportal-frontend/releases. A draft of the release notes are automatically generated by https://github.com/marketplace/actions/release-drafter. If there are pull requests in the `Changes` section i.e. they have not been labeled with one of the labels defined [here](https://github.com/cBioPortal/cbioportal-frontend/blob/master/.github/release-drafter.yml). Try to label them and trigger a rerun by committing something to the master branch. Alternatively you can manually put them in a particular section. Note that our goal is to have automated release notes, so it would be great if you could send a PR to update the [release-drafter.yml](https://github.com/cBioPortal/cbioportal-frontend/blob/master/.github/release-drafter.yml) in case you find certain PRs don't fit in a particular section or a section should be altered. Look at other release notes for inspiration: https://github.com/cBioPortal/cbioportal-frontend/releases. You can save your work as a draft if necessary.
2.  Once the frontend code is tagged, create a pull request to the backend repo where the frontend version is incremented in `portal/pom.xml`:

    ```
       <groupId>com.github.cbioportal</groupId>
       <artifactId>cbioportal-frontend</artifactId>
       <version>CHANGE_THIS_TO_THE_FRONTEND_TAG</version>
    ```
3.  Once that PR is merged, one can create a tag for the backend repo with the same tag as the frontend repo. You should see a draft from release drafter similar to the frontend in the backend repo: https://github.com/cBioPortal/cbioportal/releases. The idea is to create one set of release notes in the backend repo that is a combination of the frontend and backend notes. To make the hyperlinks from the frontend repo work in the backend repo you can copy the frontend release notes raw markdown and run the following one liner to convert the links:

    ```
    pbpaste | sed 's|(#\(....\)|([cbioportal-frontend#\1](https://github.com/cBioPortal/cbioportal-frontend/pull/\1)|g'
    ```

    then put them in the right sections following same style as other releases: https://github.com/cBioPortal/cbioportal-frontend/releases.
4. Create a news item with a link to your carefully crafted release notes. Highlight a few major changes that could be interesting to users of cBioPortal ideally with a screenshot similar to: https://github.com/cBioPortal/cbioportal/pull/6914/files?short\_path=6f95322#diff-6f953229832059bab3fe229d4af08b52 (in the files changed section, you can click on view rich diff to see the converted markdown).

## Backend Hotfixes
Occasionally there are bugs identified in the backend running in production. Creating an entire new release can be cumbersome so in this case, we create a new release branch instead. E.g. if the current release is `6.0.2`, the new branch would be `release-6.0.3`. Any fixes are then merged to this branch and if it looks good, they are deployed to production. On the following Tuesday whatever's in this `release-x` branch get merged to master and included in that Tuesday's release. 

## Release with database migration

For releases with database migrations, we increase the MINOR number in MAJOR.MINOR.PATCH. For those releases we have a separate branch (see https://github.com/cBioPortal/cbioportal/blob/master/CONTRIBUTING.md#branches-within-cbioportal), which needs to be merged to master on both backend and frontend:

1. Make sure no auto deployment is running for frontend from netlify
2. Merge frontend release-x.y.z branch to frontend master
3. Follow same procedure as for [a PATCH release](Release-Procedure.md#cbioportal-community-release-of-code-already-in-production), but instead of having a separate PR to update the frontend (step 2) one can add it to the already existing backend branch release-x.y.z and open the PR from there to backend's master. This is merely for convenience to avoid having to create another branch just to update the frontend version.

## A note on versioning

### From pre-release to official release

On the [GitHub Release Page](https://github.com/cBioPortal/cbioportal/releases) you will see that some releases have the pre-release indication whereas others do not. In general we make a new pre-release release every week. We test it out in production (https://cbioportal.org) for one month and if no new critical issues are identified we make it an official release. Occasionally we make a new official release in less than a month's time if we identify a critical issue in the previous release.

### cBioPortal Software Version Numbers

We follow the following logic when deciding how/when to increment the version of cBioPortal. It's a complete modification of semantic versioning (MAJOR.MINOR.PATCH) more suitable for our purposes:

MAJOR : A big change in how cBioPortal works. We changed the major version from 1 to 2 when we completely moved from using JSPs to a Single Page App written in React calling a REST service. Another example: we changed from 2 to 3 when we made session-service a requirement.

MINOR : Changes that require a database migration, an upgrade to another cBioPortal component (e.g. [session-service](session-service-working.md)) or anything that could require additional effort for a deployer of cBioPortal (e.g. see [transcript change](https://github.com/cBioPortal/cbioportal-frontend/pull/4350)).

PATCH : Changes that don't require database migrations. Could be new features as well as bug fixes to either frontend, backend or both.

See also: https://github.com/cBioPortal/cbioportal/releases

### cBioPortal Database Version Numbers

cBioPortal database version numbers are different version numbers than the software version numbers, it's only updated when developers make database scheme changes, please see [cBioPortal Database Versioning](Database-Versioning.md)

## User announcements

The following is a provisional system of alerting user to new features and announcements. It would probably be better for these messages to be configurable on an instance level by installers. For now, the following suffices.

Top banner: see sample configuration in src/shared/components/userMessager/UserMessage.tsx

For a beacon and associated dialog message, use this as a model, where child of InfoBeacon component is any component that will be shown when InfoBeacon is moused over and conditions are met

```
<If condition={!isWebdriver()}>
  <InfoBeacon
      top={-15}
      right={45}
      interaction={'mouseover'}
      color={'green'}
      id={'groupComparison1'}
 >
 <WrappedTour/>
 </InfoBeacon>
</If>
```
# Session Service

## 1. Steps to add new Session type through session-service

1. Define new type at [SessionType.java](https://github.com/cBioPortal/session-service/blob/master/src/main/java/org/cbioportal/session\_service/domain/SessionType.java).
2. Add necessary tests if required.
3. Update [session-service](https://github.com/cBioPortal/session-service#valid-type) documentation.

## 2. cBioPortal Backend

### 2.1 configuring session-service

Here are the properties that needs to be set

```
WARNING: do not use session service with -Dauthenticate=false
 either use authentication or change to -Dauthenticate=noauthsessionservice
session.service.url=
#if basic authentication is enabled on session service one should set:
session.service.user=
session.service.password=
```

#### session.service.url format

```
session.service.url: http://[host]:[port]/[session_service_app]/api/sessions/[portal_instance]/
```

#### example

```
session.service.url: http://localhost:8080/session_service/api/sessions/public_portal/
```

### 2.2 Updates to code

1. Update session-service dependency version in [pom.xml](https://github.com/cBioPortal/cbioportal/blob/master/pom.xml).
2. Add/Update api's in [SessionServiceController.java](https://github.com/cBioPortal/cbioportal/blob/master/web/src/main/java/org/cbioportal/web/SessionServiceController.java) to support new session type.
3. Sometimes we might need to defined model for session. Check how session class is extended in [VirtualStudy](https://github.com/cBioPortal/cbioportal/blob/master/web/src/main/java/org/cbioportal/web/parameter/VirtualStudy.java) and how it is consumed in [SessionServiceController.java](https://github.com/cBioPortal/cbioportal/blob/master/web/src/main/java/org/cbioportal/web/SessionServiceController.java).

## 3. cBioPortal Frontend

1. Client code for cbio session service is in https://github.com/cBioPortal/cbioportal-frontend/blob/master/src/shared/api/sessionServiceAPI.ts. Update it accordingly when the new api added in cBioPortal backend SessionServiceController.java

## 4. Local development

### 4.1 Expose database port

If [docker-compose.yml](https://github.com/cBioPortal/session-service/blob/master/docker-compose.yml) is used for running session service and if you like to access mongodb directly then uncomment this block of [code](https://github.com/cBioPortal/session-service/blob/master/docker-compose.yml#L21)

### 4.2 Test results with cURL in terminal

After you have local cBioPortal instance running, you can use cURL to test session-service endpoints.

**Example for GET methods:**

```
curl 'http://localhost:8080/api/session/custom_gene_list' \
  -H 'Cookie: COOCKIE_COPIED_FROM_BROWSER_REQUEST'
```

```
curl 'http://localhost:8080/api/session/custom_gene_list' \
  -H 'Cookie: _ga=GA1.1.984976088.1616537420; _ga_F1EX2C1M11=GS1.1.1616537420.1.1.1616538749.0; JSESSIONID=E7498ECA52F30588CAA1B34F420122DA'
```

**Example for POST methods:**

```
curl 'http://localhost:8080/api/session/custom_gene_list/save' \
  -H 'Content-Type: application/json' \
  -H 'Accept: */*' \
  -H 'Cookie: COOCKIE_COPIED_FROM_BROWSER_REQUEST' \
  --data-raw '{"name":"test","description":"test description","geneList":["GENE1","GENE2"]}'
```

```
curl 'http://localhost:8080/api/session/custom_gene_list/save' \
  -H 'Content-Type: application/json' \
  -H 'Accept: */*' \
  -H 'Cookie: _ga=GA1.1.984976088.1616537420; _ga_F1EX2C1M11=GS1.1.1616537420.1.1.1616538749.0; JSESSIONID=04457E173F769F7BCF9072CD5480308F' \
  --data-raw '{"name":"test","description":"test description","geneList":["GENE1","GENE2"]}'
```

Note on **Cookie**: You can log in to your account in your local cBioPortal instance, after login, copy cookies from requests in network tab. See [login configuration](/deployment/deploy-without-docker/Deploying.md#optional-login) for more information.
# Downloads

This page describes the various files available for download. The first section is targeted towards [users](Downloads.md#user-downloads) of cBioPortal. The second section towards [maintainers](Downloads.md#instance-maintainer-downloads) of cBioPortal instances.

### User Downloads

There are several ways in which one can download data from cBioPortal including manual and programmatic approaches. See options outlined below.

#### Datasets Page

A zip file for each study on [cbioportal.org](https://www.cbioportal.org) can be download from the [Datasets Page](https://www.cbioportal.org/datasets). One can also use the R client [cBioPortalData](/API-and-API-Clients.md#r-client) to programmatically download all of these files.

#### Datahub

The files for each study are also available from our [datahub repository](https://github.com/cBioPortal/datahub). This is basically the extracted version of the zip files in the [Datasets Page](https://www.cbioportal.org/datasets). Note that this is a git LFS repo so if you are familiar with git you might prefer using this option.

#### API and API Clients

Besides downloading all the study data one can also request slices of the data using the API. A slice of the data could e.g. be "give me all the mutation data for one patient" or "get me all EGFR mutations for a particular group of samples". There are API clients available in a variety of languages including bash, R and Python. See for more information the [API documentation](/API-and-API-Clients.md#api-and-api-clients).

### Instance Maintainer Downloads

As an instance maintainer of cBioPortal there are a variety of files that might be helpful. See below.

#### Study staging files

Staging files for the studies on cbioportal.org can be download from the [Datasets Page](https://www.cbioportal.org/datasets). These studies can be validated and loaded in a local cBioPortal instances using the [validator and importer](Data-Loading.md). Any issues with a downloaded study can be reported on [cBioPortal DataHub](https://github.com/cBioPortal/datahub/).

**Example studies**

TCGA Provisional studies often contain many different data types. These are excellent examples to use as reference when creating your own staging files. A detailed description on supported data types can be found in the [File Formats documentation](/File-Formats.md).

#### Complete cBioPortal database

A MySQL database dump of the complete cbioportal.org database can be found here: http://download.cbioportal.org/mysql-snapshots/public-portal-dump.latest.sql.gz

#### Seed Database

The seed database is a MySQL dump for seeding a new instance of the cBioPortal. Instructions for loading the seed database can be found [here](/deployment/deploy-without-docker/Import-the-Seed-Database.md). The seed database for human can be downloaded from [cBioPortal Datahub](https://github.com/cBioPortal/datahub/tree/master/seedDB). A mouse version can be found [here](https://github.com/cBioPortal/datahub/tree/master/seedDB\_mouse).
* [Introduction](#introduction)
* [Formats](#formats)
    * [Cancer Study](#cancer-study)
    * [Cancer Type](#cancer-type)
    * [Clinical Data](#clinical-data)
    * [Discrete Copy Number Data](#discrete-copy-number-data)
    * [Continuous Copy Number Data](#continuous-copy-number-data)
    * [Segmented Data](#segmented-data)
    * [Expression Data](#expression-data)
    * [Mutation Data](#mutation-data)
    * [Structural Variant Data](#structural-variant-data)
    * [Fusion Data](#fusion-data) (DEPRECATED)
    * [Methylation Data](#methylation-data)
    * [Protein level Data](#protein-level-data)
    * [Case Lists](#case-lists)
    * [Timeline Data](#timeline-data)
    * [GISTIC 2.0 Data](#gistic-20-data)
    * [Mutsig Data](#mutsig-data)
    * [Gene Panel Data](#gene-panel-data)
    * [Gene Set Data](#gene-set-data)
    * [Study Tags file](#study-tags-file)
    * [Generic Assay](#generic-assay)
        * [Arm Level CNA Data](#arm-level-cna-data)
        * [Mutational Signature Data](#mutational-signature-data)
    * [Resource Data](#resource-data)
    * [Custom namespace columns](#custom-namespace-columns)

# Introduction

This page describes the file formats that cancer study data should assume in order to be successfully imported into the database.  Unless otherwise noted, all data files are in tabular-TSV (tab separated value) format and have an associated metadata file which is in a multiline record format.  The metadata and data files should follow a [few rules documented at the Data Loading page](Data-Loading.md#preparing-study-data).

# Formats

## Cancer Study

As described in the [Data Loading tool](Data-Loading.md) page, the following file is needed to describe the cancer study:

### Meta file
This file contains metadata about the cancer study. The file contains the following fields:

1. **type_of_cancer**: The cancer type abbreviation, e.g., "brca".  This should be the same cancer type as specified in the `meta_cancer_type.txt` file, if available. The type can be "mixed" for studies with multiple cancer types.
2. **cancer_study_identifier**: A string used to uniquely identify this cancer study within the database, e.g., "brca_joneslab_2013".  
3. **name**: The name of the cancer study, e.g., "Breast Cancer (Jones Lab 2013)".
4. **description**: A description of the cancer study, e.g., "Comprehensive profiling of 103 breast cancer samples. Generated by the Jones Lab 2013". This description may contain one or more URLs to relevant information.
5. **citation (Optional)**: A relevant citation, e.g., "TCGA, Nature 2012".
6. **pmid (Optional)**: One or more relevant pubmed ids (comma separated without whitespace). If used, the field citation has to be filled, too.
7. **groups (Optional)**: When using an authenticating cBioPortal, lists the user-groups that are allowed access to this study. Multiple groups are separated with a semicolon ";". The study will be invisible to users not in _at least one_ of the listed groups, as if it wasn't loaded at all. e.g., "PUBLIC;GDAC;SU2C-PI3K". see [User-Authorization](/deployment/authorization-and-authentication/User-Authorization.md) for more information on groups
8. **add_global_case_list (Optional)**: set to 'true' if you would like the "All samples" case list to be generated automatically for you. See also [Case lists](#case-lists).
9. **tags_file (Optional)**: the file name containing custom study tags for the [study tags](#study-tags-file).
10. **reference_genome (Optional)**: the study reference genome (e.g. `hg19`, `hg38`). Without specifying this property, the study will be assigned to the reference genome specified in `application.properties` (property `ucsc.build`).

### Example  
An example meta_study.txt file would be:
```
type_of_cancer: brca
cancer_study_identifier: brca_joneslab_2013
name: Breast Cancer (Jones Lab 2013)
description: Comprehensive profiling of 103 breast cancer samples. Generated by the Jones Lab 2013.
add_global_case_list: true
```

## Cancer Type
If the type_of_cancer specified in the meta_study.txt does not yet exist in the `type_of_cancer` database table, a `meta_cancer_type.txt` file is also mandatory.

### Meta file
The file is comprised of the following fields:

1. **genetic_alteration_type**: CANCER_TYPE
2. **datatype**: CANCER_TYPE
3. **data_filename**: your datafile

### Example
An example meta_cancer_type.txt file would be:
```
genetic_alteration_type: CANCER_TYPE
datatype: CANCER_TYPE
data_filename: cancer_type.txt
```

### Data file
The file is comprised of the following columns in the order specified:

1. **type_of_cancer**: The cancer type abbreviation, e.g., "brca".
2. **name**: The name of the cancer type, e.g., "Breast Invasive Carcinoma".
3. **dedicated_color**: CSS color name of the color associated with this cancer study, e.g., "HotPink". See [this list](https://www.w3.org/TR/css3-color/#svg-color) for supported names, and follow the [awareness ribbons](https://en.wikipedia.org/wiki/List_of_awareness_ribbons) color schema. This color is associated with the cancer study on various web pages within the cBioPortal.
4. **parent_type_of_cancer**:  The `type_of_cancer` field of the cancer type of which this is a subtype, e.g., "Breast". :information_source: : you can set parent to `tissue`, which is the reserved word to place the given cancer type at "root" level in the "studies oncotree" that will be generated in the homepage (aka query page) of the portal. 

### Example
An example record would be:
```
brca<TAB>Breast Invasive Carcinoma<TAB>HotPink<TAB>Breast
```

## Clinical Data

The clinical data is used to capture both clinical attributes and the mapping between patient and sample ids. The software supports multiple samples per patient.  

As of March 2016, the clinical file is split into a patient clinical file and a sample clinical file. The *sample* file is required, whereas the *patient* file is optional. cBioPortal has specific functionality for a core set of patient and sample columns, but can also display custom columns (see section ["Custom columns in clinical data"](#custom-columns-in-clinical-data)).

### Meta files
The two clinical metadata files (or just one metadata file if you choose to leave the *patient* file out) have to contain the following fields:

1. **cancer_study_identifier**: same value specified in meta_study.txt
2. **genetic_alteration_type**: CLINICAL
3. **datatype**: PATIENT_ATTRIBUTES or SAMPLE_ATTRIBUTES
4. **data_filename**: your datafile

### Examples
An example metadata file, e.g. named meta_clinical_sample.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: CLINICAL
datatype: SAMPLE_ATTRIBUTES
data_filename: data_clinical_sample.txt
```

An example metadata file, e.g. named meta_clinical_patient.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: CLINICAL
datatype: PATIENT_ATTRIBUTES
data_filename: data_clinical_patient.txt
```

### Data files
For both patients and samples, the clinical data file is a two dimensional matrix with multiple clinical attributes. When the attributes are defined in the *patient* file they are considered to be patient attributes; when they are defined in the *sample* file they are considered to be sample attributes. 

The first four rows of the clinical data file contain tab-delimited metadata about the clinical attributes. These rows **have to start with a '#' symbol**. Each of these four rows contain different type of information regarding each of the attributes that are defined in the fifth row:

- Row 1: **The attribute Display Names**: The display name for each clinical attribute
- Row 2: **The attribute Descriptions**: Long(er) description of each clinical attribute
- Row 3: **The attribute Datatype**: The datatype of each clinical attribute (must be one of:  STRING, NUMBER, BOOLEAN)
- Row 4: **The attribute Priority**: A number which indicates the importance of each attribute.  In the future, higher priority attributes will appear in more prominent places than lower priority ones on relevant pages (such as the [Study View](https://www.cbioportal.org/study?id=brca_tcga)). A higher number indicates a higher priority.
    ```
    To promote certain chart in study view, please increase priority to a certain number. The higher the score, the higher priority it will be displayed in the study view.
    If you want to hide chart, please set the priority to 0. For combination chart, as long as one of the clinical attribute has been set to 0, it will be hidden.
    
    Currently, we preassigned priority to few charts, but as long as you assign a priority except than 1, these preassigned priorities will be overwritten.
    
    CANCER_TYPE: 3000, CANCER_TYPE_DETAILED: 2000,
    Overall survival plot: 400 (This is combination of OS_MONTH and OS_STATUS) 
    Disease Free Survival Plot: 300 (This is combination of DFS_MONTH and DFS_STATUS) 
    Mutation Count vs. CNA Scatter Plot: 200,
    Mutated Genes Table: 90, CNA Genes Table: 80, study_id: 70, # of Samples Per Patient: 40,
    With Mutation Data Pie Chart: 60, With CNA Data Pie Chart: 50, 
    Mutation Count Bar Chart: 30, CNA Bar Chart: 20,
    GENDER: 9, SEX: 9, AGE: 8
    ```
    
    Please note: 
    Priority is not the sole factor determining which chart will be displayed first.
    A layout algorithm in study view also makes a minor adjustment on the layout.
    The algorithm tries to fit all charts into a 2 by 2 matrix (Mutated Genes Table occupies 2 by 2 space).
    When a chart can not be fitted in the first matrix, the second matrixed will be generated. 
    And the second matrix will have lower priority than the first one. 
    If later chart can fit into the first matrix, then its priority will be promoted.
    
    Please see [here](/deployment/customization/Studyview.md) for more detailed information about how study view utilize priority and how the layout is calculated based on priority.
- Row 5: **The attribute name for the database**: This name should be in upper case.
- Row 6: This is the first row that contains actual data.


### Example clinical header
Below is an example of the first 4 rows with the respective metadata for the attributes defined in the 5th row. 
```
#Patient Identifier<TAB>Overall Survival Status<TAB>Overall Survival (Months)<TAB>Disease Free Status<TAB>Disease Free (Months)<TAB>...
#Patient identifier<TAB>Overall survival status<TAB>Overall survival in months since diagnosis<TAB>Disease free status<TAB>Disease free in months since treatment<TAB>...
#STRING<TAB>STRING<TAB>NUMBER<TAB>STRING<TAB>NUMBER<TAB>...
#1<TAB>1<TAB>1<TAB>1<TAB>1<TAB>
PATIENT_ID<TAB>OS_STATUS<TAB>OS_MONTHS<TAB>DFS_STATUS<TAB>DFS_MONTHS<TAB>...
....
data - see examples below
....
```

### Clinical patient columns
The file containing the patient attributes has one **required** column:
- **PATIENT_ID (required)**: a unique patient ID. This field allows only numbers, letters, points, underscores and hyphens.

The following columns are used by the study view as well as the patient view. In the [study view](https://www.cbioportal.org/study?id=brca_tcga) they are used to create the survival plots. In the patient view they are used to add information to the [header](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P05).

**Note on survival plots**: to generate the survival plots successfully, the columns are required to be in pairs, which means the file should have a pair of columns that have the same prefix but ending with `_STATUS` and `_MONTHS` individually. For example, `PFS_STATUS` and `PFS_MONTHS` are a valid pair of columns that can generate the survival plots.

**Note on survival status value**: the value of survival status must prefixed with `0:` or `1:`. Value with prefix `0:` means that no event (e.g. `LIVING`, `DiseaseFree`). Value with prefix `1:` means that an event occurred (e.g. `DECEASED`, `Recurred/Progressed`). 

- **OS_STATUS**:  Overall patient survival status
    - Possible values: 1:DECEASED, 0:LIVING
    - In the patient view, 0:LIVING creates a green label, 1:DECEASED a red label.
- **OS_MONTHS**:  Overall survival in months since initial diagnosis
- **DFS_STATUS**: Disease free status since initial treatment
    - Possible values: 0:DiseaseFree, 1:Recurred/Progressed
    - In the patient view, 0:DiseaseFree creates a green label, 1:Recurred/Progressed a red label.
- **DFS_MONTHS**: Disease free (months) since initial treatment

These columns, when provided, add additional information to the patient description in the header:
- **PATIENT_DISPLAY_NAME**: Patient display name (string)
- **GENDER** or **SEX**: Gender or sex of the patient (string)
- **AGE**: Age at which the condition or disease was first diagnosed, in years (number)
- **TUMOR_SITE**

Custom attributes:
- **Custom Clinical Attribute Headers**: Any other custom attribute can be added as well. See section ["Custom columns in clinical data"](#custom-columns-in-clinical-data).

### Example *patient* data file
```
#Patient Identifier<TAB>Overall Survival Status<TAB>Overall Survival (Months)<TAB>Disease Free Status<TAB>Disease Free (Months)<TAB>...
#Patient identifier<TAB>Overall survival status<TAB>Overall survival in months since diagnosis<TAB>Disease free status<TAB>Disease free in months since treatment<TAB>...
#STRING<TAB>STRING<TAB>NUMBER<TAB>STRING<TAB>NUMBER<TAB>...
#1<TAB>1<TAB>1<TAB>1<TAB>1<TAB>
PATIENT_ID<TAB>OS_STATUS<TAB>OS_MONTHS<TAB>DFS_STATUS<TAB>DFS_MONTHS<TAB>...
PATIENT_ID_1<TAB>1:DECEASED<TAB>17.97<TAB>1:Recurred/Progressed<TAB>30.98<TAB>...
PATIENT_ID_2<TAB>0:LIVING<TAB>63.01<TAB>0:DiseaseFree<TAB>63.01<TAB>...
...
```

### Clinical sample columns
The file containing the sample attributes has two **required** columns:
- **PATIENT_ID (required)**: A patient ID. This field can only contain numbers, letters, points, underscores and hyphens.
- **SAMPLE_ID (required)**: A sample ID. This field can only contain numbers, letters, points, underscores and hyphens.

By adding `PATIENT_ID` here, cBioPortal will map the given sample to this patient. This enables one to associate multiple samples to one patient. For example, a single patient may have had multiple biopsies, each of which has been genomically profiled. See [this example for a patient with multiple samples](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P04).

The following columns are required for the pan-cancer summary statistics tab ([example](https://www.cbioportal.org/index.do?cancer_study_id=msk_impact_2017&Z_SCORE_THRESHOLD=2&RPPA_SCORE_THRESHOLD=2&data_priority=0&case_set_id=msk_impact_2017_cnaseq&gene_list=BRAF&geneset_list=+&tab_index=tab_visualize&Action=Submit&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna)).
- **CANCER_TYPE**: Cancer Type
- **CANCER_TYPE_DETAILED**: Cancer Type Detailed, a sub-type of the specified CANCER_TYPE

The following columns affect the header of the patient view by adding text to the samples in the header:
- **SAMPLE_DISPLAY_NAME**: displayed in addition to the ID
- **SAMPLE_CLASS**
- **METASTATIC_SITE** or **PRIMARY_SITE**: Override TUMOR_SITE (patient level attribute) depending on sample type

The following columns additionally affect the [Timeline data](#timeline-data) visualization:
- **OTHER_SAMPLE_ID**: OTHER_SAMPLE_ID is no longer supported. Please replace this column header with SAMPLE_ID.   
- **SAMPLE_TYPE**, **TUMOR_TISSUE_SITE** or **TUMOR_TYPE**: gives sample icon in the timeline a color.
    - If set to `recurrence`, `recurred`, `progression` or `progressed`: orange
    - If set to `metastatic` or `metastasis`: red
    - If set to `primary` or otherwise: black

Custom attributes:
- **Custom Clinical Attribute Headers**: Any other custom attribute can be added as well. See section ["Custom columns in clinical data"](#custom-columns-in-clinical-data).

### Example sample data file
```
#Patient Identifier<TAB>Sample Identifier<TAB>Subtype<TAB>...
#Patient identifier<TAB>Sample Identifier<TAB>Subtype description<TAB>...
#STRING<TAB>STRING<TAB>STRING<TAB>...
#1<TAB>1<TAB>1<TAB>...
PATIENT_ID<TAB>SAMPLE_ID<TAB>SUBTYPE<TAB>...
PATIENT_ID_1<TAB>SAMPLE_ID_1<TAB>basal-like<TAB>...
PATIENT_ID_2<TAB>SAMPLE_ID_2<TAB>Her2 enriched<TAB>...
...
```

### Columns with specific functionality
These columns can be in either the patient or sample file.
- **CANCER_TYPE**: Overrides study wide cancer type
- **CANCER_TYPE_DETAILED**
- **KNOWN_MOLECULAR_CLASSIFIER**
- **GLEASON_SCORE**: Radical prostatectomy Gleason score for prostate cancer
- **HISTOLOGY**
- **TUMOR_STAGE_2009**
- **TUMOR_GRADE**
- **ETS_RAF_SPINK1_STATUS**
- **TMPRSS2_ERG_FUSION_STATUS**
- **ERG_FUSION_ACGH**
- **SERUM_PSA**
- **DRIVER_MUTATIONS**

### Custom columns in clinical data
cBioPortal supports custom columns with clinical data in either the patient or sample file. They should follow the previously described 5-row header format. Be sure to provide the correct `Datatype`, for optimal search, sorting, filtering (in [clinical data tab](https://www.cbioportal.org/study?id=brca_tcga#clinical)) and visualization.

The Clinical Data Dictionary from MSKCC is used to normalize clinical data, and should be followed to make the clinical data comparable between studies. This dictionary provides a definition whether an attribute should be defined on the patient or sample level, as well as provides a name, description and datatype. The data curator can choose to ignore these proposed definitions, but not following this dictionary might make comparing data between studies more difficult. It should however not break any cBioPortal functionality. See GET /api/ at [https://oncotree.mskcc.org/cdd/swagger-ui.html#/](https://oncotree.mskcc.org/cdd/swagger-ui.html#/) for the data dictionary of all known clinical attributes.

### Banned column names
`MUTATION_COUNT` and `FRACTION_GENOME_ALTERED` are auto populated clinical attributes, and should therefore not be present in clinical data files.

## Discrete Copy Number Data
The discrete copy number data file contain values that would be derived from copy-number analysis algorithms like [GISTIC 2.0](https://www.ncbi.nlm.nih.gov/sites/entrez?term=18077431) or [RAE](https://www.ncbi.nlm.nih.gov/sites/entrez?term=18784837). GISTIC 2.0 can be [installed](https://www.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=216&p=t) or run online using the GISTIC 2.0 module on [GenePattern](https://cloud.genepattern.org). For some help on using GISTIC 2.0, check the [Data Loading: Tips and Best Practices](Data-Loading-Tips-and-Best-Practices.md) page. When loading case list data, the `_cna` case list is required. See the [case list section](#case-lists).

### Wide vs Long format
For CNA data two formats are supported: the wide, and the long format:
- **Wide format**: a matrix, where each row is a gene, and each column is a sample
- **Long format**: not a matrix, each row is a gene-sample combination; this makes the file longer

### Wide format

#### Meta file 
The meta file is comprised of the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: COPY_NUMBER_ALTERATION
3. **datatype**: `DISCRETE`
4. **stable_id**: gistic, cna, cna_rae or cna_consensus
5. **show_profile_in_analysis_tab**: true
6. **profile_name**: A name for the discrete copy number data, e.g., "Putative copy-number alterations from GISTIC"
7. **profile_description**: A description of the copy number data, e.g., "Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification."
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id
10. **pd_annotations_filename (Optional)**: name of [custom driver annotations file](File-Formats.md#custom-driver-annotations-file)

##### Example
An example metadata file could be named meta_cna.txt and its contents could be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: COPY_NUMBER_ALTERATION
datatype: DISCRETE
stable_id: gistic
show_profile_in_analysis_tab: true
profile_name: Putative copy-number alterations from GISTIC
profile_description: Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification.
data_filename: data_cna.txt
pd_annotations_filename: data_cna_pd_annotations.txt
```

#### Data file
For each gene (row) in the data file, the following columns are required in the order specified:

One or both of:
- ***Hugo_Symbol***: A [HUGO](https://www.genenames.org/) gene symbol.
- ***Entrez_Gene_Id***: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier.

And:
- An additional column for each sample in the dataset using the sample id as the column header.

For each gene-sample combination, a copy number level is specified:
- "-2" is a deep loss, possibly a homozygous deletion
- "-1" is a single-copy loss (heterozygous deletion)
- "0" is diploid
- "1" indicates a low-level gain
- "2" is a high-level amplification.

##### Example
An example data file which includes the required column header would look like:
```
Hugo_Symbol<TAB>Entrez_Gene_Id<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
ACAP3<TAB>116983<TAB>0<TAB>-1<TAB>...
AGRN<TAB>375790<TAB>2<TAB>0<TAB>...
...
...
`````

### Long format

#### Meta file 
The meta file of **wide format** is comprised of the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: COPY_NUMBER_ALTERATION
3. **datatype**: `DISCRETE_LONG`
   Note: It will end up as datatype `DISCRETE` in the database, because the LONG data format is only relevant while importing. 
4. **stable_id**: gistic, cna, cna_rae or cna_consensus
5. **show_profile_in_analysis_tab**: true
6. **profile_name**: A name for the discrete copy number data, e.g., "Putative copy-number alterations from GISTIC"
7. **profile_description**: A description of the copy number data, e.g., "Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification."
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id
10. **namespaces (Optional)**: Comma-delimited list of `namespaces` to import. 

##### Example
An example metadata file could be named meta_cna.txt and its contents could be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: COPY_NUMBER_ALTERATION
datatype: DISCRETE_LONG
stable_id: gistic
show_profile_in_analysis_tab: true
profile_name: Putative copy-number alterations from GISTIC
profile_description: Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification.
data_filename: data_cna.txt
namespaces: MyNamespace,MyNamespace2
```

#### Data file
Each row contains a row-sample combination. Custom driver annotations are added as columns to the data file, just like custom namespace columns.

##### Example
An example data file which includes the required column header would look like:
```
Hugo_Symbol	Entrez_Gene_Id	Sample_Id	Value	cbp_driver	cbp_driver_annotation	cbp_driver_tiers	cbp_driver_tiers_annotation	MyNamespace.column1
ACAP3	116983	TCGA-A2-A04U-01	2	Putative_Passenger	Test passenger	Class 2	Class annotation	value1
...
```

#### Adding your own discrete copy number columns
Additional columns can be added to the discrete copy number **long** data file. In this way, the portal will parse and store your own CNA fields in the database.

See [Custom namespace columns](#custom-namespace-columns) for more information on adding custom columns to data files.

### Custom driver annotations file

Custom driver annotations can be defined for discrete copy number data. These annotations can be used to complement or replace default driver annotation resources OncoKB and HotSpots.
Custom driver annotations can be placed in a separate file that is referenced by the `pd_annotations_file` field of the meta file. The annotation file can hold the following columns:

1. **Hugo_Symbol (Optional)**: A [HUGO](https://www.genenames.org/) gene symbol. Required when column `Entrez_Gene_Id` is not present.
2. **Entrez_Gene_Id (Optional)**: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier. Required when column `Hugo_Symbol` is not present.
3. **SAMPLE_ID**: A sample ID. This field can only contain numbers, letters, points, underscores and hyphens.
4. **cbp_driver (Optional)**: "Putative_Driver", "Putative_Passenger", "Unknown", "NA" or "" (empty value). This field must be present if the cbp_driver_annotation is also present in the MAF file. 
5. **cbp_driver_annotation (Optional)**: Description field for the cbp_driver value (limited to 80 characters). This field must be present if the cbp_driver is also present in the MAF file. This field is free text. Example values for this field are: "Pathogenic" or "VUS".
6. **cbp_driver_tiers (Optional)**: Free label/category that marks the mutation as a putative driver such as "Driver", "Highly actionable", "Potential drug target". . This field must be present if the cbp_driver_tiers_annotation is also present in the MAF file. In the OncoPrint view's Mutation Color dropdown menu, these tiers are ordered alphabetically. This field is free text and limited to 20 characters. For mutations without a custom annotation, leave the field blank or type "NA".
7. **cbp_driver_tiers_annotation (Optional)**: Description field for the cbp_driver_tiers value (limited to 80 characters). This field must be present if the cbp_driver_tiers is also present in the MAF file. This field can not be present when the cbp_driver_tiers field is not present. 

All genes referenced in the custom driver annotation file must be present in the data file for discrete copy number alterations.

The `cbp_driver` column flags the mutation as either driver or passenger. In cBioPortal, passenger mutations are also known as variants of unknown significance (VUS). The `cbp_driver_tiers` column assigns an annotation tier to the mutation, such as "Driver", "Highly actionable" or "Potential drug target". When a tier is selected, mutations with that annotation are highlighted as driver. Both types of custom annotations contain a second column with the suffix `_annotation`, to add a description. This is displayed in the tooltip that appears when hovering over the sample's custom annotation icon in the OncoPrint view.

You can learn more about configuring these annotations in the [application.properties documentation](/deployment/customization/application.properties-Reference.md#custom-annotation-of-driver-and-passenger-mutations). When properly configured, the customized annotations appear in the "Mutation Color" menu of the OncoPrint view: \
![schreenshot mutation color menu](images/screenshot-mutation-color-menu.png) 

### Example
An example data file which includes the required column header would look like:
```
SAMPLE_ID<TAB>Hugo_Symbol<TAB>Entrez_Gene_Id<TAB>cbp_driver<TAB>cbp_driver_annotation<TAB>cbp_driver_tiers<TAB>cbp_driver_tiers_annotation<TAB>...
TCGA-BH-A0E6-01<TAB>GENEA<TAB>116983<TAB>Putative_Driver<TAB>see: PMID:12345678<TAB>Highly actionable<TAB>Per decision 01/01/2020<TAB>
TCGA-BH-A0E6-01<TAB>GENEB<TAB>375790<TAB>Putative_Passenger<TAB>see: PMID:12345678<TAB><TAB><TAB>
...
`````

### GISTIC 2.0 Format
GISTIC 2.0 outputs a tabular file similarly formatted to the cBioPortal format, called `<prefix>_all_thresholded.by_genes.txt`.
In this file the gene symbol is found in the `Gene Symbol` column, while Entrez gene IDs are in the `Gene ID` or 
`Locus ID` column. Please rename `Gene Symbol` to `Hugo_Symbol` and `Gene ID` or `Locus ID` to `Entrez_Gene_Id`. The 
`Cytoband` column can be kept in the table, but note that these values are ignored in cBioPortal. cBioPortal uses 
cytoband annotations from the `map_location` column in NCBI's `Homo_sapiens.gene_info.gz` when loading genes into 
the seed database.

## Continuous Copy Number Data

### Meta file
The continuous copy number metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: COPY_NUMBER_ALTERATION.
3. **datatype**: CONTINUOUS
4. **stable_id**: linear_CNA 
5. **show_profile_in_analysis_tab**: false.
6. **profile_name**: A name for the copy number data, e.g., "copy-number values".
7. **profile_description**: A description of the copy number data, e.g., "copy-number values for each gene (from Affymetrix SNP6).".
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id

cBioPortal also supports log2 copy number data. If your data is in log2, change the following fields:

3. **datatype**: LOG2-VALUE
4. **stable_id**: log2CNA 

### Example
An example metadata file, e.g. meta_log2_cna.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: COPY_NUMBER_ALTERATION
datatype: LOG2-VALUE
stable_id: log2CNA
show_profile_in_analysis_tab: false
profile_description: Log2 copy-number values for each gene (from Affymetrix SNP6).
profile_name: Log2 copy-number values
data_filename: data_log2_cna.txt
```

### Data file

The log2 copy number data file follows the same format as expression data files.  See [Expression Data](#expression-data) for a description of the expression data file format.

### GISTIC 2.0 Format
GISTIC 2.0 outputs a tabular file similarly formatted to the cBioPortal format, called `<prefix>_all_data_by_genes.txt`. 
In this file the gene symbol is found in the `Gene Symbol` column, while Entrez gene IDs are in the `Gene ID` or 
`Locus ID` column. Please rename `Gene Symbol` to `Hugo_Symbol` and `Gene ID` or `Locus ID` to `Entrez_Gene_Id`. The 
`Cytoband` column can be kept in the table, but note that these values are ignored in cBioPortal. cBioPortal uses 
cytoband annotations from the `map_location` column in NCBI's `Homo_sapiens.gene_info.gz` when loading genes into 
the seed database.


## Segmented Data

A SEG file (segmented data; .seg or .cbs) is a tab-delimited text file that lists loci and associated numeric values. The segmented data file format is the output of the Circular Binary Segmentation algorithm (Olshen et al., 2004). This Segment data enables the 'CNA' lane in the Genomic overview of the Patient view (as [can be seen in this example](https://www.cbioportal.org/patient?sampleId=TCGA-BH-A0E6-01&studyId=brca_tcga)). 

### Meta file
The segmented metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: COPY_NUMBER_ALTERATION
3. **datatype**: SEG
4. **reference_genome_id**: Reference genome version. Supported values: "hg19"
5. **description**: A description of the segmented data, e.g., "Segment data for the XYZ cancer study.".
6. **data_filename**: your datafile

### Example:
An example metadata file, e.g. meta_cna_hg19_seg.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: COPY_NUMBER_ALTERATION
datatype: SEG
reference_genome_id: hg19
description: Somatic CNA data (copy number ratio from tumor samples minus ratio from matched normals) from TCGA.
data_filename: data_cna_hg19.seg
```

### Data file

The first row contains column headings and each subsequent row contains a locus and an associated numeric value. See also the [Broad IGV page on this format](https://www.broadinstitute.org/software/igv/SEG).

### Example:

An example data file which includes the required column header would look like:
```
ID<TAB>chrom<TAB>loc.start<TAB>loc.end<TAB>num.mark<TAB>seg.mean
SAMPLE_ID_1<TAB>1<TAB>3208470<TAB>245880329<TAB>128923<TAB>0.0025
SAMPLE_ID_2<TAB>2<TAB>474222<TAB>5505492<TAB>2639<TAB>-0.0112
SAMPLE_ID_2<TAB>2<TAB>5506070<TAB>5506204<TAB>2<TAB>-1.5012
SAMPLE_ID_2<TAB>2<TAB>5512374<TAB>159004775<TAB>80678<TAB>-0.0013
...
...
```

## Expression Data

An expression data file is a two dimensional matrix with a gene per row and a sample per column.  For each gene-sample pair, a real number represents the gene expression in that sample.  

### Meta file

The expression metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: MRNA_EXPRESSION
3. **datatype**: CONTINUOUS, DISCRETE or Z-SCORE
4. **stable_id**: see table below.
5. **source_stable_id**: Required when both conditions are true: (1) `datatype` = `Z-SCORE` and (2) this study contains [GSVA data](#gene-set-data). Should contain `stable_id` of the expression file for which this [Z-SCORE file is the statistic](#z-score-instructions).
5. **show_profile_in_analysis_tab**: false (you can set to **true** if Z-SCORE to enable it in the oncoprint, for example).
6. **profile_name**: A name for the expression data, e.g., "mRNA expression (microarray)".
7. **profile_description**: A description of the expression data, e.g., "Expression levels (Agilent microarray).".
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id

### Supported stable_id values for MRNA_EXPRESSION
For historical reasons, cBioPortal expects the `stable_id` to be one of those listed in the following static set.
The stable_id for continuous RNA-seq data has two options: `rna_seq_mrna` or `rna_seq_v2_mrna`. These options were added to distinguish between two different TCGA pipelines, which perform different types of normalization (RPKM and RSEM). However, for custom datasets either one of these `stable_id` can be chosen.

<table>
<thead>
<tr><th>datatype</th><th>stable_id</th><th>description</th></tr>
</thead>
<tr><td>CONTINUOUS</td><td>mrna_U133</td><td>Affymetrix U133 Array</td></tr>
<tr><td>Z-SCORE</td><td>mrna_U133_Zscores</td><td>Affymetrix U133 Array</td></tr>
<tr><td>Z-SCORE</td><td>rna_seq_mrna_median_Zscores</td><td>RNA-seq data</td></tr>
<tr><td>Z-SCORE</td><td>mrna_median_Zscores</td><td>mRNA data</td></tr>
<tr><td>CONTINUOUS</td><td>rna_seq_mrna</td><td>RNA-seq data</td></tr>
<tr><td>CONTINUOUS</td><td>rna_seq_v2_mrna</td><td>RNA-seq data</td></tr>
<tr><td>Z-SCORE</td><td>rna_seq_v2_mrna_median_Zscores</td><td>RNA-seq data</td></tr>
<tr><td>CONTINUOUS</td><td>mirna</td><td>MicroRNA data</td></tr>
<tr><td>Z-SCORE</td><td>mirna_median_Zscores</td><td>MicroRNA data</td></tr>
<tr><td>Z-SCORE</td><td>mrna_merged_median_Zscores</td><td>?</td></tr>
<tr><td>CONTINUOUS</td><td>mrna</td><td>mRNA data</td></tr>
<tr><td>DISCRETE</td><td>mrna_outliers</td><td>mRNA data of outliers</td></tr>
<tr><td>Z-SCORE</td><td>mrna_zbynorm</td><td>?</td></tr>
<tr><td>CONTINUOUS</td><td>rna_seq_mrna_capture</td><td>data from Roche mRNA Capture Kit</td></tr>
<tr><td>Z-SCORE</td><td>rna_seq_mrna_capture_Zscores</td><td>data from Roche mRNA Capture Kit</td></tr>
</table>


### Example

An example metadata, e.g. meta_expression.txt file would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: MRNA_EXPRESSION
datatype: CONTINUOUS
stable_id: rna_seq_mrna
show_profile_in_analysis_tab: false
profile_name: mRNA expression 
profile_description: Expression levels 
data_filename: data_expression.txt
```

### Data file

For each gene (row) in the data file, the following columns are required in the order specified:

One or both of:
- ***Hugo_Symbol***: A [HUGO](https://www.genenames.org/) gene symbol.
- ***Entrez_Gene_Id***: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier.

And:
- An additional column for each sample in the dataset using the sample id as the column header.

For each gene-sample combination, a value is specified:
- A real number for each sample id (column) in the dataset, representing the expression value for the gene in the respective sample.
- or `NA` for when the expression value for the gene in the respective sample could not (or was not) be measured (or detected).

### z-score instructions

For mRNA expression data, we typically expect the relative expression of an individual gene and tumor to the gene's expression distribution in a reference population. That reference population is either all tumors that are diploid for the gene in question, or, when available, normal adjacent tissue. The returned value indicates the number of standard deviations away from the mean of expression in the reference population (Z-score). This measure is useful to determine whether a gene is up- or down-regulated relative to the normal samples or all other tumor samples.  **Note, the importer tool can create normalized (z-score) expression data on your behalf.  Please visit the [Z-Score normalization script](Z-Score-normalization-script.md) wiki page for more information.
A corresponding z-score metadata file would be something like**:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: MRNA_EXPRESSION
datatype: Z-SCORE
stable_id: rna_seq_mrna_median_Zscores
show_profile_in_analysis_tab: true
profile_name: mRNA expression z-scores
profile_description: Expression levels z-scores
data_filename: data_expression_zscores.txt
```

### Examples of data files: 

An example data file which includes the required column header and *leaves out Hugo_Symbol* (recommended) would look like:

```
Entrez_Gene_Id<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
116983<TAB>-0.005<TAB>-0.550<TAB>...
375790<TAB>0.142<TAB>0.091<TAB>...
...
...
```

An example data file which includes both Hugo_Symbo and Entrez_Gene_Id would look like (supported, but not recommended as it increases the chances of errors regarding ambiguous gene symbols):
```
Hugo_Symbol<TAB>Entrez_Gene_Id<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
ACAP3<TAB>116983<TAB>-0.005<TAB>-0.550<TAB>...
AGRN<TAB>375790<TAB>0.142<TAB>0.091<TAB>...
...
...
```

An example data file with only Hugo_Symbol column (supported, but not recommended as it increases the chances of errors regarding ambiguous gene symbols):
```
Hugo_Symbol<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
ACAP3<TAB>-0.005<TAB>-0.550<TAB>...
AGRN<TAB>0.142<TAB>0.091<TAB>...
...
...
```


## Mutation Data
When loading mutation data, the `_sequenced` case list is required. See the [case list section](#case-lists).

### Meta file
The mutation metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: MUTATION_EXTENDED
3. **datatype**: MAF
4. **stable_id**: mutations
5. **show_profile_in_analysis_tab**: true
6. **profile_name**: A name for the mutation data, e.g., "Mutations".
7. **profile_description**: A description of the mutation data, e.g., "Mutation data from whole exome sequencing.".
8. **data_filename**: your data file
9. **gene_panel (optional)**: gene panel stable id. See [Gene panels for mutation data](#gene-panels-for-mutation-data).
10. **swissprot_identifier (optional)**: `accession` or `name`, indicating the type of identifier in the `SWISSPROT` column
11. **variant_classification_filter (optional)**: List of `Variant_Classifications` values to be filtered out.
12. **namespaces (optional)**: Comma-delimited list of `namespaces` to import. 

### Gene panels for mutation data
Using the `gene_panel` property it is possible to annotate **all samples in the MAF file** as being profiled on the **same** specified gene panel. 

Please use the [Gene Panel Matrix file](#gene-panel-matrix-file) when:
- Data contains samples that are profiled but no mutations are called. Also please add these to the `_sequenced` case list. 
- Multiple gene panels are used to profile the samples in the MAF file.

### Variant classification filter
The `variant_classification_filter` field can be used to filter out specific mutations. This field should contain a comma separated list of `Variant_Classification` values. By default, cBioPortal filters out `Silent, Intron, IGR, 3'UTR, 5'UTR, 3'Flank and 5'Flank`, except for the promoter mutations of the TERT gene. For no filtering, include this field in the metadata file, but leave it empty. For cBioPortal default filtering, do not include this field in the metadata file.
Allowed values to filter out (mainly from [Mutation Annotation Format page](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/)): `Frame_Shift_Del, Frame_Shift_Ins, In_Frame_Del, In_Frame_Ins, Missense_Mutation, Nonsense_Mutation, Silent, Splice_Site, Translation_Start_Site, Nonstop_Mutation, 3'UTR, 3'Flank, 5'UTR, 5'Flank, IGR, Intron, RNA, Targeted_Region, De_novo_Start_InFrame, De_novo_Start_OutOfFrame, Splice_Region and Unknown`

### Tumor seq allele ambiguity
Bugs may exist in MAF data that make it ambiguous as to whether `Tumor_Seq_Allele1` or `Tumor_Seq_Allele2` should be seen as the variant allele to be used when a new mutation record is created and imported in cBioPortal. In such cases, preference is given to the tumor seq allele value that matches a valid nucleotide pattern `^[ATGC]*$` versus a null or empty value, or "-".
For example, given `Reference_Allele` = "G", `Tumor_Seq_Allele` = "-", and `Tumor_Seq_Allele2` = "A", preference will be given to `Tumor_Seq_Allele2`. Using this same example with `Tumor_Seq_Allele1` = "T", preference will be given to `Tumor_Seq_Allele1` if it does not match `Reference_Allele`, which in this case it does not.

When curating MAF data, it is best practice to leave `Tumor_Seq_Allele1` empty if this information is not provided in your data source to avoid this ambiguity.

### Namespaces
The `namespaces` field can be used to specify additional MAF columns for import. This field should contain a comma separated list of namespaces. Namespaces can be identified as prefixes to an arbitrary set of additional MAF columns (separated with a period e.g `ASCN.total_copy_number`, `ASCN.minor_copy_number`). All columns with a prefix matching a namespace specified in the metafile will be imported; columns with an unspecified namespace will be ignored. If no additional columns beyond the required set need to be imported, the field should be left blank. 

### Example
An example metadata file would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: MUTATION_EXTENDED
datatype: MAF
stable_id: mutations
show_profile_in_analysis_tab: true
profile_description: Mutation data from whole exome sequencing.
profile_name: Mutations
data_filename: data_mutations.txt
namespaces: ASCN
```

### Data file
The cBioPortal mutation data file extends the [Mutation Annotation Format](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/) (MAF) created as part of The Cancer Genome Atlas (TCGA) project, by adding *extra annotations* to each mutation record. This section describes:
1. How to create the cBioPortal mutation data file with a minimal MAF file using the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline).
2. The description of the cBioPortal mutation data file. You can also get the cBioPortal mutation data file from vcf using: [vcf2maf](https://github.com/mskcc/vcf2maf).

### Create the cBioPortal mutation data file with Genome Nexus with a minimal MAF file
#### Minimal MAF file format
A minimal mutation annotations file can contain just the five genomic change columns plus one sample identifier column. From this minimal MAF, it is possible to create the cBioPortal mutation data file by running it through the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline).
1. **Chromosome (Required)**:  A chromosome number, e.g., "7".
2. **Start_Position (Required)**: Start position of event.
3. **End_Position (Required)**: End position of event.
4. **Reference_Allele (Required)**: The plus strand reference allele at this position.
5. **Tumor_Seq_Allele2 (Required)**: Primary data genotype.
6. **Tumor_Sample_Barcode (Required)**: This is the sample ID. Either a TCGA barcode (patient identifier will be extracted), or for non-TCGA data, a literal SAMPLE_ID as listed in the clinical data file.

In addition to the above columns, it is recommended to have the read counts to calculate variant allele frequencies:

7. **t_alt_count (Optional, but recommended)**: Variant allele count (tumor). 
8. **t_ref_count (Optional, but recommended)**: Reference allele count (tumor).

The following extra annotation columns are important for making sure mutation specific UI functionality works well in the portal:

9. **Protein_position (Optional)**: (annotation column) Required to initialize the 3D viewer in [mutations view](https://www.cbioportal.org/index.do?cancer_study_list=brca_tcga_pub&cancer_study_id=brca_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_pub_gistic&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_pub_mrna_median_Zscores&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=brca_tcga_pub_complete&case_ids=&patient_case_select=sample&gene_set_choice=prostate-cancer%3A-ar-signaling-%2810-genes%29&gene_list=TP53&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit#mutation_details)
10. **SWISSPROT (Optional)**: (annotation column) UniProtKB/SWISS-PROT name (formerly called ID) or accession code depending on the value of the `swissprot_identifier` metadatum, e.g. O11H1_HUMAN or Q8NG94. Is not required, but not having it may result in inconsistent PDB structure matching in [mutations view](https://www.cbioportal.org/index.do?cancer_study_list=brca_tcga_pub&cancer_study_id=brca_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_pub_gistic&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_pub_mrna_median_Zscores&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=brca_tcga_pub_complete&case_ids=&patient_case_select=sample&gene_set_choice=prostate-cancer%3A-ar-signaling-%2810-genes%29&gene_list=TP53&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit#mutation_details).

### Creating the cBioPortal mutation data file
Once you have a minimal MAF you can run it through the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline).
This tool runs annotates variants against the [Genome Nexus Server](https://genomenexus.org), which in turn leverages Ensembl Variant Effect Predictor (VEP) and selects a single effect per variant. Protein identifiers will be mapped to UniProt canonical isoforms (see also [this mapping file](https://github.com/genome-nexus/genome-nexus-importer/blob/master/data/grch37_ensembl92/export/ensembl_biomart_canonical_transcripts_per_hgnc.txt)).

### cBioPortal mutation data file format
The cBioPortal mutation data file format recognized by the portal has:
* 32 columns from the [TCGA MAF format](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/).
* 1 column with the amino acid change.
* 4 columns with information on reference and variant allele counts in tumor and normal samples. 

1. **Hugo_Symbol (Required)**: A [HUGO](https://www.genenames.org/) gene symbol.
2. **Entrez_Gene_Id (Optional, but recommended)**: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier.
3. **Center (Optional)**: The sequencing center.
4. **NCBI_Build (Required)<sup>1</sup>**: The Genome Reference Consortium Build is used by a variant calling software. It must be "GRCh37" or "GRCh38" for a human, and "GRCm38" for a mouse.
5. **Chromosome (Required)**: A chromosome number, e.g., "7".
6. **Start_Position (Optional, but recommended for additional features such as [Cancer Hotspots annotations](https://www.cancerhotspots.org/))**: Start position of event.
7. **End_Position (Optional, but recommended for additional features such as [Cancer Hotspots annotations](https://www.cancerhotspots.org/))**: End position of event.
8. **Strand (Optional)**: We assume that the mutation is reported for the + strand.
9. **Variant_Classification (Required)**: Translational effect of variant allele, e.g. Missense_Mutation, Silent, etc.
10. **Variant_Type <sup>1</sup>(Optional)**: Variant Type, e.g. SNP, DNP, etc.
11. **Reference_Allele (Required)**: The plus strand reference allele at this position.
12. **Tumor_Seq_Allele1 (Optional)**: Primary data genotype.
13. **Tumor_Seq_Allele2 (Required)**: Primary data genotype.
14. **dbSNP_RS<sup>1</sup> (Optional)**: Latest dbSNP rs ID.
15. **dbSNP_Val_Status<sup>1</sup> (Optional)**: dbSNP validation status.
16. **Tumor_Sample_Barcode (Required)**: This is the sample ID. Either a TCGA barcode (patient identifier will be extracted), or for non-TCGA data, a literal SAMPLE_ID as listed in the clinical data file.
17. **Matched_Norm_Sample_Barcode<sup>1</sup> (Optional)**: The sample ID for the matched normal sample.
18. **Match_Norm_Seq_Allele1 (Optional)**: Primary data.
19. **Match_Norm_Seq_Allele2 (Optional)**: Primary data.
20. **Tumor_Validation_Allele1 (Optional)**: Secondary data from orthogonal technology.
21. **Tumor_Validation_Allele2 (Optional)**: Secondary data from orthogonal technology.
22. **Match_Norm_Validation_Allele1<sup>1</sup> (Optional)**: Secondary data from orthogonal technology.
23. **Match_Norm_Validation_Allele2<sup>1</sup> (Optional)**: Secondary data from orthogonal technology.
24. **Verification_Status<sup>1</sup> (Optional)**: Second pass results from independent attempt using same methods as primary data source. "Verified", "Unknown" or "NA".
25. **Validation_Status (Optional)**: Second pass results from orthogonal technology. "Valid", "Invalid", "Untested", "Inconclusive", "Redacted", "Unknown" or "NA".
26. **Mutation_Status (Optional)**: "Somatic" or "Germline" are supported by the UI in Mutations tab. "None", "LOH" and "Wildtype" will not be loaded. Other values will be displayed as text.
27. **Sequencing_Phase<sup>1</sup> (Optional)**: Indicates current sequencing phase.
28. **Sequence_Source<sup>1</sup> (Optional)**: Molecular assay type used to produce the analytes used for sequencing.
29. **Validation_Method<sup>1</sup> (Optional)**: The assay platforms used for the validation call.
30. **Score<sup>1</sup> (Optional)**: Not used.
31. **BAM_File<sup>1</sup> (Optional)**: Not used.
32. **Sequencer<sup>1</sup> (Optional)**: Instrument used to produce primary data.
33. **HGVSp_Short (Required)**: Amino Acid Change, e.g. p.V600E.
34. **t_alt_count (Optional)**: Variant allele count (tumor). 
35. **t_ref_count (Optional)**: Reference allele count (tumor).
36. **n_alt_count (Optional)**: Variant allele count (normal).
37. **n_ref_count (Optional)**: Reference allele count (normal).

<sup>**1**</sup> These columns are currently not shown in the Mutation tab and Patient view.

### Custom driver annotations
It is possible to manually add columns for defining custom driver annotations. These annotations can be used to complement or replace default driver annotation resources OncoKB and HotSpots.

38. **cbp_driver (Optional)**: "Putative_Driver", "Putative_Passenger", "Unknown", "NA" or "" (empty value). This field must be present if the cbp_driver_annotation is also present in the MAF file. 
39. **cbp_driver_annotation (Optional)**: Description field for the cbp_driver value (limited to 80 characters). This field must be present if the cbp_driver is also present in the MAF file. This field is free text. Example values for this field are: "Pathogenic" or "VUS".
40. **cbp_driver_tiers (Optional)**: Free label/category that marks the mutation as a putative driver such as "Driver", "Highly actionable", "Potential drug target". . This field must be present if the cbp_driver_tiers_annotation is also present in the MAF file. In the OncoPrint view's Mutation Color dropdown menu, these tiers are ordered alphabetically. This field is free text and limited to 20 characters. For mutations without a custom annotation, leave the field blank or type "NA".
41. **cbp_driver_tiers_annotation (Optional)**: Description field for the cbp_driver_tiers value (limited to 80 characters). This field must be present if the cbp_driver_tiers is also present in the MAF file. This field can not be present when the cbp_driver_tiers field is not present. 

The `cbp_driver` column flags the mutation as either driver or passenger. In cBioPortal, passenger mutations are also known as variants of unknown significance (VUS). The `cbp_driver_tiers` column assigns an annotation tier to the mutation, such as "Driver", "Highly actionable" or "Potential drug target". When a tier is selected, mutations with that annotation are highlighted as driver. Both types of custom annotations contain a second column with the suffix `_annotation`, to add a description. This is displayed in the tooltip that appears when hovering over the sample's custom annotation icon in the OncoPrint view.

You can learn more about configuring these annotations in the [application.properties documentation](/deployment/customization/application.properties-Reference.md#custom-annotation-of-driver-and-passenger-mutations). When properly configured, the customized annotations appear in the "Mutation Color" menu of the OncoPrint view: \
![schreenshot mutation color menu](/images/screenshot-mutation-color-menu.png) 

### Adding your own mutation annotation columns
Additional mutation annotation columns can be added to the cBioPortal mutation data file. In this way, the portal will
parse and store your own MAF fields in the database. For example, mutation data that you find on cBioPortal.org comes
from MAF files that have been further enriched with information
from [mutationassessor.org](https://mutationassessor.org/), which leads to a "Mutation Assessor" column in
the [mutation table](https://www.cbioportal.org/index.do?cancer_study_list=acc_tcga&cancer_study_id=acc_tcga&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=acc_tcga_mutations&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=acc_tcga_sequenced&case_ids=&patient_case_select=sample&gene_set_choice=user-defined-list&gene_list=ZFPM1&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit).

See [Custom namespace columns](#custom-namespace-columns) for more information on adding custom columns to data files.

### Allele specific copy number (ASCN) annotations
Allele specific copy number (ASCN) annotation is also supported and may be added using namespaces, described [here](#adding-mutation-annotation-columns-through-namespaces). If ASCN data is present in the cBioPortal mutation data file, the deployed cBioPortal instance will display additional columns in the mutation table showing ASCN data.

**The ASCN columns below are optional by default. If `ascn` is a defined namespace in `meta_mutations_extended.txt`, then these columns are ALL required.**

42. **ASCN.ASCN_METHOD (Optional)**: Method used to obtain ASCN data e.g "FACETS".
43. **ASCN.CCF_EXPECTED_COPIES (Optional)**: Cancer-cell fraction if mutation exists on major allele. 
44. **ASCN.CCF_EXPECTED_COPIES_UPPER (Optional)**: Upper error for CCF estimate.
45. **ASCN.EXPECTED_ALT_COPIES (Optional)**: Estimated number of copies harboring mutant allele.
46. **ASCN.CLONAL (Optional)**: "Clonal", "Subclonal", or "Indeterminate". 
47. **ASCN.TOTAL_COPY_NUMBER (Optional)**: Total copy number of the gene.
48. **ASCN.MINOR_COPY_NUMBER (Optional)**: Copy number of the minor allele.
49. **ASCN.ASCN_INTEGER_COPY_NUMER (Optional)**: Absolute integer copy-number estimate.

### Example cBioPortal mutation data file
An example cBioPortal mutation data file can be found in the cBioPortal test study [study_es_0](https://github.com/cBioPortal/cbioportal/blob/master/core/src/test/scripts/test_data/study_es_0/data_mutations_extended.maf).

### Filtered mutations
A special case for **Entrez_Gene_Id=0** and **Hugo_Symbol=Unknown**: when this combination is given, the record is parsed in the same way as **Variant_Classification=IGR** and therefore filtered out.


## Methylation Data
The Portal expects a single value for each gene in each sample, usually a beta-value from the Infinium methylation array platform.

### Meta file

The methylation metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: METHYLATION
3. **datatype**: CONTINUOUS
4. **stable_id**: "methylation_hm27" or "methylation_hm450" (depending on platform).
5. **show_profile_in_analysis_tab**: false
6. **profile_name**: A name for the methylation data, e.g., "Methlytation (HM27)".
7. **profile_description**: A description of the methlytation data, e.g., "Methylation beta-values (HM27 platform). For genes with multiple methylation probes, the probe least correlated with expression is selected.".
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id


### Example

An example metadata file would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: METHYLATION
datatype: CONTINUOUS
stable_id: methylation_hm27
show_profile_in_analysis_tab: false
profile_name: Methylation (HM27)
profile_description: Methylation beta-values (HM27 platform). For genes with multiple methylation probes, the probe least correlated with expression is selected.
data_filename: data_methylation_hm27.txt
```

### Data file

The methylation data file follows the same format as expression data files. See [Expression Data](#expression-data) for a description of the expression data file format. The Portal expects a single value for each gene in each sample, usually a beta-value from the Infinium methylation array platform.


## Protein level Data

Protein expression measured by reverse-phase protein array or mass spectrometry. Antibody-sample pairs, with a real number representing the protein level for that sample.

### Meta file

The protein level metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: PROTEIN_LEVEL
3. **datatype**: LOG2-VALUE or Z-SCORE
4. **stable_id**: rppa, rppa_Zscores, protein_quantification or protein_quantification_zscores 
5. **show_profile_in_analysis_tab**: false (**true** for Z-SCORE datatype)
6. **profile_name**: A name for the RPPA data, e.g., "RPPA data".
7. **profile_description**: A description of the RPPA data, e.g., "RPPA levels.".
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id

An example metadata file would be:
```
cancer_study_identifier: brca_tcga
genetic_alteration_type: PROTEIN_LEVEL
datatype: LOG2-VALUE
stable_id: rppa
show_profile_in_analysis_tab: false
profile_description: Protein expression measured by reverse-phase protein array
profile_name: Protein expression (RPPA)
data_filename: data_rppa.txt
```

**NB:** You also need a Z-SCORE file if you want protein levels to be available in query UI and in [Oncoprint visualization](https://www.cbioportal.org/index.do?Ôú´cancer_study_list=brca_tcga_pub&cancer_study_id=brca_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_pub_gistic&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_pub_mrna_median_Zscores&Z_SCORE_THRESHOLD=2.0&genetic_profile_ids_PROFILE_PROTEIN_EXPRESSION=brca_tcga_pub_rppa_Zscores&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=brca_tcga_pub_complete&case_ids=&patient_case_select=sample&gene_set_choice=prostate-cancer%3A-ar-signaling-%2810-genes%29&gene_list=TP53+SOX9+RAN+TNK2+EP300+PXN+NCOA2+AR+NRIP1+NCOR1+NCOR2&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit). E.g.:

```
cancer_study_identifier: brca_tcga
genetic_alteration_type: PROTEIN_LEVEL
datatype: Z-SCORE
data_filename: data_rppa_zscores.txt
stable_id: rppa_Zscores
show_profile_in_analysis_tab: true
profile_description: Protein expression Z-scores (RPPA)
profile_name: Protein expression Z-scores (RPPA)
```

### Data file

A protein level data file is a two dimensional matrix with a RPPA antibody per row and a sample per column. For each antibody-sample pair, a real number represents the protein level for that sample.  The antibody information can contain one or more HUGO gene symbols and/or entrez gene identifiers, separated by a space, and an antibody ID pair separated by the "|" symbol.

### Example 

An example data file which includes the required column header would look like:

```
Composite.Element.REF<TAB>SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>...
BRAF|B-Raf-M-NA<TAB>1.09506676325<TAB>0.5843256495...
MAPK1 MAPK3|MAPK_PT202_Y204<TAB>1.70444582025<TAB>1.0982864685...
AKT1 AKT2 10000|AKT<TAB>0.17071492725<TAB>0.264067254391
...
```


## Structural Variant Data

The cBioPortal can load all kinds of structural variant data but at the moment only a subset of them, fusions, are displayed.

### Meta file
The structural variant metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: STRUCTURAL_VARIANT
3. **datatype**: SV
4. **stable_id**: structural_variants
5. **show_profile_in_analysis_tab**: true.
6. **profile_name**: A name for the fusion data, e.g., "Structural Variants".
7. **profile_description**: A description of the structural variant data.
8. **data_filename**: your datafile (e.g. data_sv.txt)
9. **gene_panel (Optional)**:  gene panel stable id

An example metadata file would be:

```
cancer_study_identifier: msk_impact_2017
genetic_alteration_type: STRUCTURAL_VARIANT
datatype: SV
stable_id: structural_variants
show_profile_in_analysis_tab: true
profile_name: Structural variants
profile_description: Structural Variant Data for mskimpact2017
data_filename: data_sv.txt
```

### Data file

A structural variant data file is a tab-delimited file with one structural variant per row.  For each structural variant (row) in the data file, the following tab-delimited values are required:

| Field                        | Example Values         | Allowed  Values       | Comments |
| -- | -- | -- | -- |
| Sample_Id (Required)         | Sample_1      |  | As defined in the clinical sample file. |
| SV_Status (Required)         | SOMATIC | SOMATIC or GERMLINE |  |
| Site1_Hugo_Symbol            | TMPRSS2 |   | [HUGO](http://www.genenames.org/) gene symbol of gene 1. One might call this the left site(3‚Äô) as well. (strongly recommended field) |
| Site1_Ensembl_Transcript_Id  | ENST00000398585  |  | Ensembl transcript ID of gene 1. |
| Site1_Entrez_Gene_Id         | 7113 |  | [Entrez Gene](http://www.ncbi.nlm.nih.gov/gene) identifier of gene 1. (strongly recommended field) |
| Site1_Region_Number          | 2 |  | Number of Site 1 region e.g. exon 2.(strongly recommended field) |
| Site1_Region                 | Exon | 5_Prime_UTR,3_Prime_UTR,Promoter,Exon,Intron | We advise using one of these {5_Prime_UTR,3_Prime_UTR,Promoter,Exon,Intron},but it is a free text. (strongly recommended field) |
| Site1_Chromosome             | 21 |  | Chromosome of Gene 1.(strongly recommended field) |
| Site1_Contig                 | q22.3 |  | The contig of Site 1.(strongly recommended field) |
| Site1_Position               | 42874110 |  | Genomic position of breakpoint of Gene 1.(strongly recommended field) |
| Site1_Description            | Intron of TMPRSS2(-): 511bp before exon 3  |  | Description of this event at site 2. This could be the location of the 2nd breakpoint in case of a fusion event. |
| Site2_Hugo_Symbol            | ERG |  | [HUGO](http://www.genenames.org/) gene symbol of gene 2.One might call this the right site(5‚Äô) as well. |
| Site2_Ensembl_Transcript_Id  | ENST00000288319 |   | Ensembl transcript ID of gene 2.|
| Site2_Entrez_Gene_Id         | 2078 |  | [Entrez Gene](http://www.ncbi.nlm.nih.gov/gene) identifier of gene 2. |
| Site2_Region_Number          | 4 |  | Number of Site 2 region e.g. exon 4.|
| Site2_Region                 | EXON | 5_PRIME_UTR,3_PRIME_UTR,PROMOTER,EXON,INTRON | We advise using one of these {5_PRIME_UTR,3_PRIME_UTR,PROMOTER,EXON,INTRON},but it is a free text. |
| Site2_Chromosome             | 21 |  | Chromosome of Gene 2.|
| Site2_Contig                 | q22.2 |  | The contig of Site 2.|
| Site2_Position               | 39842043 |  | Genomic position of breakpoint of Gene 2. |
| Site2_Description            | Intron of ERG(-): 6Kb after exon 1 |  | Description of this event at site 1. This could be the location of 1st breakpoint in case of a fusion event |
| Site2_Effect_On_Frame        | Out-of-frame | In_frame, Out-of-frame,Frameshift  | The effect on frame reading in gene 2. Frame_Shift or InFrame,but it is a free text. |
| NCBI_Build                   | GRCh37 | GRCh37,GRCh38 | The NCBI assembly. Only one assembly per study can be used, see[ study metadata](https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#meta-file). |
| Class                        | Deletion | Deletion,Duplication,Insertion,Inversion,Translocation. | We advise using one of these terms [DELETION, DUPLICATION, INSERTION, INVERSION or TRANSLOCATION], but it is free text. |
| Tumor_Split_Read_Count       | 4 |  | The number of split reads of the tumor tissue that support the call.[Tumor Split Read Count is the same as ‚ÄúJunction Reads‚Äù.] |
| Tumor_Paired_End_Read_Count  | 55 |  | The number of paired-end reads of the tumor tissue that support the call. [Tumor Paired End Read Count is the same as ‚ÄúSpanning Fragments‚Äù.] |
| Event_Info                   | Protein fusion: out of frame (TMPRSS2-ERG) | Antisense fusion, Deletion within transcript: mid-exon, Duplication of 1 exon: in frame | Description of the event. For a fusion event, fill in Fusion. It is a free text. |
| Connection_Type              | 3to5 | 3to5 or 5to3 or 5to5 or 3to3 | Which direction the connection is made (3' to 5', 5' to 3', etc) |
| Breakpoint_Type              | PRECISE  | PRECISE/IMPRECISE  | PRECISE or IMPRECISE which explain the resolution. Fill in PRECISE if the breakpoint resolution is known down to the exact base pair. |
| Annotation                   | TMPRSS2 (NM_001135099) - ERG (NM_001243428) fusion (TMPRSS2 exons 1-2 fused with ERG exons 4-11):(c.126+879:TMRPSS2_c.40-63033:ERGdel) | Free Text                                                    | Free text description of the gene or transcript rearrangement. |
| DNA_Support                  | Yes  | Yes or No  | Fusion detected from DNA sequence data, "Yes" or "No".|
| RNA_Support                  |   | Yes or No | Fusion detected from DNA sequence data, "Yes" or "No".|
| SV_Length                    | 3032067 |  | Length of the structural variant in number of bases.|
| Normal_Read_Count            | 93891 |  | The total number of reads of the normal tissue.|
| Tumor_Read_Count             | 45556 |  | The total number of reads of the tumor tissue. |
| Normal_Variant_Count         | 0 |  | The number of reads of the normal tissue that have the variant/allele. |
| Tumor_Variant_Count          | 39 |  | The number of reads of the tumor tissue that have the variant/allele. |
| Normal_Paired_End_Read_Count | 41 |  | The number of paired-end reads of the normal tissue that support the call.|
| Normal_Split_Read_Count      | 63 |  | The number of split reads of the normal tissue that support the call. |
| Comments                     |  |  | Any comments/free text.|

For an example see [datahub](https://github.com/cBioPortal/datahub/blob/master/public/msk_impact_2017/data_sv.txt). For an example see [datahub](https://github.com/cBioPortal/datahub/blob/master/public/msk_impact_2017/data_sv.txt). At a minimum `Sample_Id`, either `Site1_Hugo_Symbol`/ `Site1_Entrez_Gene_Id` or  `Site2_Hugo_Symbol`/ `Site2_Entrez_Gene_Id` and `SV_Status` are required. For the stuctural variant tab visualization (still in development) one needs to provide those field as well as `Site1_Ensembl_Transcript_Id`, `Site2_Ensembl_Transcript_Id`, `Site1_Region` and `Site2_Region`. Some of the other columns are shown at several other pages on the website. The `Class`, `Annotation` and `Event_Info` columns are shown prominently on several locations.
**Note**: We strongly recommend all the data providers to submit genomic locations  in addition to required fields for future visualization features. 

### Adding your own structural variant columns
Additional mutation annotation columns can be added to the structural variant data file. In this way, the portal will
parse and store your own structural variant fields in the database.

See [Custom namespace columns](#custom-namespace-columns) for more information on adding custom columns to data files.

## Fusion Data
**‚ö†Ô∏è DEPRECATED Use the: [SV format](#structural-variant-data) instead**

### Meta file
The fusion metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: FUSION
3. **datatype**: FUSION
4. **stable_id**: fusion
5. **show_profile_in_analysis_tab**: true.
6. **profile_name**: A name for the fusion data, e.g., "Fusions".
7. **profile_description**: A description of the fusion data.
8. **data_filename**: your datafile
9. **gene_panel (Optional)**:  gene panel stable id

### Example
An example metadata file would be:
```
cancer_study_identifier: brca_tcga_pub
genetic_alteration_type: FUSION
datatype: FUSION
stable_id: fusion
show_profile_in_analysis_tab: true
profile_name: Fusions
profile_description: Fusion data.
data_filename: data_fusions.txt
```

### Data file
A fusion data file is a two dimensional matrix with one gene per row.  For each gene (row) in the data file, the following tab-delimited values are required in the order specified:

1. **Hugo_Symbol**: A [HUGO](https://www.genenames.org/) gene symbol.
2. **Entrez_Gene_Id**: A [Entrez Gene](https://www.ncbi.nlm.nih.gov/gene) identifier.
3. **Center**: The sequencing center.
4. **Tumor_Sample_Barcode**: This is the sample ID.
5. **Fusion**: A description of the fusion, e.g., "TMPRSS2-ERG fusion".
6. **DNA_support**: Fusion detected from DNA sequence data, "yes" or "no".
7. **RNA_support**: Fusion detected from RNA sequence data, "yes" or "no".
8. **Method**: Fusion detected algorithm/tool.
9. **Frame**: "in-frame" or "frameshift".
10. **Fusion_Status (OPTIONAL)**: An assessment of the mutation type (i.e., "SOMATIC", "GERMLINE", "UNKNOWN", or empty)

**Note:** If a fusion event includes a gene, e.g., Hugo_Symbol or Entrez_Gene_Id, that is not profiled, the event will be filter out during import into the database.

An example data file which includes the required column header would look like:
```
Hugo_Symbol<TAB>Entrez_Gene_Id<TAB>Center<TAB>Tumor_Sample_Barcode<TAB>Fusion<TAB>DNA_support<TAB>RNA_support<TAB>Method<TAB>Frame>
ALK<TAB>238<TAB>center.edu<TAB>SAMPLE_ID_1<TAB>Fusion<TAB>unknown<TAB>yes<TAB>unknown<TAB>in-frame
ALK<TAB>238<TAB>center.edu<TAB>SAMPLE_ID_2<TAB>Fusion<TAB>unknown<TAB>yes<TAB>unknown<TAB>in-frame
RET<TAB>5979<TAB>center.edu<TAB>SAMPLE_ID_3<TAB>Fusion<TAB>unknown<TAB>yes<TAB>unknown<TAB>in-frame
...
...
```

## Case Lists

Case lists are used to define sample lists that can be selected on the query page. Some case lists have specific functionality, but it's also possible to add custom case lists. The case list files should be placed in a sub-directory called `case_lists` which exists alongside all the other cancer study data. 
The case list file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **stable_id**: it must contain the cancer_study_identifier followed by an underscore. Typically, after this a relevant suffix, e.g., `_custom`, is added. There are some naming rules to follow if you want the case list to be selected automatically in the query UI base on the selected sample profiles. See subsection below.
3. **case_list_name**: A name for the patient list, e.g., "All Tumors".
4. **case_list_description**: A description of the patient list, e.g., "All tumor samples (825 samples).".
5. **case_list_ids**: A tab-delimited list of sample ids from the dataset.
6. **case_list_category**: Optional *alternative* way of linking your case list to a specific molecular profile. E.g. setting this to `all_cases_with_cna_data` will signal to the portal that this is the list of samples to be associated with CNA data in *some* of the analysis. 

### Example
An example case list file would be:
```
cancer_study_identifier: brca_tcga_pub
stable_id: brca_tcga_pub_custom
case_list_name: Custom subset of samples
case_list_description: Custom subset of samples (825 samples)
case_list_ids: SAMPLE_ID_1<TAB>SAMPLE_ID_2<TAB>SAMPLE_ID_3<TAB>...
```

### Case list stable id suffixes
In order for sample counts to propagate to the data sets widget on the home page and the table on the [Data Sets](https://www.cbioportal.org/data_sets.jsp) page, the following case list suffixes need to be used in the stable_id property (e.g. `brca_tcga_pub_sequenced`). This is also needed for correct statistics in the Study view page when calculating the frequency of CNA and of mutations per gene in the respective summary tables.

* **Sequenced**: `_sequenced` . When only a mutation profile is selected on the query page, this is the default case list. Also used in the Study Summary to calculate the proportion of samples with mutations.
* **CNA**: `_cna`. When only a CNA profile is selected on the query page, this is the default case list. Also used in the Study Summary to calculate the proportion of samples with CNA.
* **Sequenced and CNA**: `_cnaseq`. When a mutation and CNA genetic profile are selected on the query page, this is the default case list.
* **mRNA (microarray)**: `_mrna`. When only a mRNA (microarray) profile is selected on the query page, this is the default case list.
* **mRNA (RNA-Seq)**: `_rna_seq_mrna`. When only a mRNA (RNA-Seq) profile is selected on the query page, this is the default case list.
* **mRNA (RNA-SeqV2)**: `_rna_seq_v2_mrna`. When only a mRNA (RNA-SeqV2) profile is selected on the query page, this is the default case list.
* **mRNA normal**: `_normal_mrna`. Used for the datasets page to calculate the number of normal samples.
* **mRNA normal**: `_microrna`. Used for the datasets page to calculate the number of microRNA samples.
* **Methylation (HM27)**: `_methylation_hm27`.
* **RPPA**: `_rppa`. When only a RPPA profile is selected on the query page, this is the default case list.
* **Sequenced, CNA and mRNA**: `_3way_complete` When a mutation, CNA and mRNA profile are selected on the query page, this is the default case list.
* **SV**: `_sv` . When a structural variant profile is selected on the query page, this is the default case list. Also used in the Study Summary to calculate the proportion of samples with fusions.
* **All**: `_all`. If you are not using *add_global_case_list* attribute in [Study metadata](#cancer-study), you need to add this case list.

### Required case lists
Some case lists are required:

- `_all`. This can be generated by the importer if you set the attribute `add_global_case_list` to `true` in the [Study metadata](#cancer-study).
- `_sequenced`. This case list is required when loading mutation data.
- `_cna`. This case list is required when loading discrete cna data.

### Case list categories
These are the valid case lists categories for `case_list_category: ` in the meta file.
- `all_cases_in_study`
- `all_cases_with_mutation_data`
- `all_cases_with_cna_data`
- `all_cases_with_log2_cna_data`
- `all_cases_with_methylation_data`
- `all_cases_with_mrna_array_data`
- `all_cases_with_mrna_rnaseq_data`
- `all_cases_with_rppa_data`
- `all_cases_with_microrna_data`
- `all_cases_with_mutation_and_cna_data`
- `all_cases_with_mutation_and_cna_and_mrna_data`
- `all_cases_with_gsva_data`
- `all_cases_with_sv_data`
- `other`

## Timeline Data
The timeline data is a representation of the various events that occur during the course of treatment for a patient from initial diagnosis. In cBioPortal timeline data is represented as one or more tracks in the patient view. Each main track is based on an event type, such as "Specimen", "Imaging", "Lab_test", etc.

**Attention**: some clinical attributes affect the timeline visualization. Please check the [Clinical Data](#clinical-data) section for more information.  

This type data is not yet being validated. It can, however, be uploaded.

### Meta file
Each event type requires its own meta file. A timeline meta file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: CLINICAL
3. **datatype**: TIMELINE
4. **data_filename**: your datafile

An example metadata file would be:
```
cancer_study_identifier: brca_tcga
genetic_alteration_type: CLINICAL
datatype: TIMELINE
data_filename: data_timeline_imaging.txt
```


### Data file
Each event type requires its own data file, which contains all the events that each patient undergoes. The data format used for timeline data is extremely flexible. There are four required columns:

1. **PATIENT_ID**: the patient ID from the dataset
2. **START_DATE**: the start point of any event, calculated in **_days_* from the date of diagnosis (which will act as point zero on the timeline scale)
3. **STOP_DATE**: The end date of the event is calculated in days from the date of diagnosis (which will act as point zero on the timeline scale). If the event occurs over time (e.g. a Treatment, ...) the STOP_DATE column should have values. If the event occurs at a time point (e.g. a Lab_test, Imaging, ...) the STOP_DATE is still mandatory, but the values should be blanks.
4. **EVENT_TYPE**: the category of the event. You are free to define any type of event here. For several event types cBioPortal has column naming suggestions and for several events there are column names which have special effects. See [event types](#event-types) for more information. 

And optional columns with special effects:

1. **SPECIMEN_REFERENCE_NUMBER**: SPECIMEN_REFERENCE_NUMBER is no longer supported. Please replace this column header with SAMPLE_ID.
2. **STYLE_SHAPE**: when this column has a valid value, this event will be rendered using that shape. The valid shapes are `circle`, `square`, `triangle`, `diamond`, `star`, and `camera`.
3. **STYLE_COLOR**: when this column has a hexadecimal color value (e.g. #ffffff), it will be used as the color for rendering this event.

To embed hyperlinks in custom columns:

Values in custom columns can include markdown for hyperlinks that will show up in event tooltips, allowing users to click through to external resources associated with events.

![link-in-timeline-data](images/timeline-link.png)

Example column:

SOME_HYPERLINK: ```[Link text Here](https://link-url-here.org)```

The external link can be opened in a new tab, instead of an IFRAME within the same window/tab. To do this, the string `:blank` is to be added as a suffix at the end of the URL.

HYPERLINK_OPEN_IN_NEWTAB: ```[Link text Here](https://link-url-here.org/path:blank)```

### Event Types
As previously mentioned, the EVENT_TYPE can be anything. However, several event types have columns with special effects. Furthermore, for some event types cBioPortal has column naming suggestions.

_**EVENT_TYPE: TREATMENT**_

Suggested columns

 * **TREATMENT_TYPE**: This can be either Medical Therapy or Radiation Therapy.
 * **SUBTYPE**: Depending upon the TREATMENT_TYPE, this can either be Chemotherapy, Hormone Therapy, Targeted Therapy  etc. (for Medical Therapies) or WPRT, IVRT etc. (for Radiation Therapies).
 * **AGENT**: for medical therapies, the agent is defined with number of cycles if applicable and for radiation therapy, the agent is defined as standard dose given to the patient during the course.
 * **AGENT_CLASS**: This allows you to classify your agents into useful groups.
 * Based on different cancer types you can add additional data here.

Special: When using the AGENT and SUBTYPE columns, each agent and subtype will be split into its own track.

_**EVENT_TYPE: LAB_TEST**_

Suggested columns
 * **TEST**: type of test performed
 * **RESULT**: corresponding value of the test
 * Based on different cancer types you can add additional data here.

Special: When using the TEST and RESULT columns, each test gets its own track. Any TEST that has only numerical RESULT values will be rendered as a line chart.

_**EVENT_TYPE: IMAGING**_

Suggested columns
 * **DIAGNOSTIC_TYPE**: This attribute will cover the different diagnostics tools used (for example: MRI, CT scan etc.)
 * **DIAGNOSTIC_TYPE_DETAILED**: Detailed description of the event type.
 * **RESULT**: Results of the diagnostic tests
 * **SOURCE**: Where was the Imaging done.
 * Based on different cancer types you can add additional data here.

Special: all dots in the IMAGING track are squares.

_**EVENT_TYPE: STATUS**_

Suggested columns

 * **STATUS**: If the EVENT_TYPE is status, data is entered under STATUS to define either the best response from the treatment or if there is a diagnosis of any stage progression etc.
 * **SOURCE**: Where the status was monitored.
 * Based on different cancer types you can add additional data here.

_**EVENT_TYPE: SPECIMEN**_

Suggested columns

 * **SPECIMEN_SITE**: This is the site from where the specimen was collected.
 * **SPECIMEN_TYPE**: This can either be tissue or blood.
 * **SOURCE**: Where was the specimen collection done.
 * Based on different cancer types you can add additional data here.


### Clinical Track Ordering
Clinical tracks are ordered as follows (if available):

1. Specimen
2. Surgery
3. Status
4. Diagnostics
5. Diagnostic
6. Imaging
7. Lab_test
8. Treatment
9. First custom event
10. etc.

### Example

An example timeline file for SPECIMEN would be:
```
PATIENT_ID<TAB>START_DATE<TAB>EVENT_TYPE<TAB>SAMPLE_ID<TAB>SPECIMEN_SITE<TAB>SPECIMEN_TYPE<TAB>SOURCE<TAB>MyCustomColumn
CACO2<TAB>0<TAB>SPECIMEN<TAB>CACO2_S1<TAB>liver<TAB>tissue<TAB>hospital<TAB>T1
CACO2<TAB>100<TAB>SPECIMEN<TAB>CACO2_S2<TAB>lung<TAB>tissue<TAB>hospital<TAB>T2
...
```
Assuming the sample identifiers were also defined in the clinical file, this will lead to a timeline track with numbered specimen samples.

An example timeline file for Lab_test would be:
```
PATIENT_ID<TAB>START_DATE<TAB>EVENT_TYPE<TAB>TEST<TAB>RESULT
CACO2<TAB>100<TAB>LAB_TEST<TAB>PSA<TAB>10
CACO2<TAB>250<TAB>LAB_TEST<TAB>PSA<TAB>100
...
```
This will lead to a timeline track for Lab_test with an additional subtrack specifically for PSA. PSA's events will be sized based on the result.


## GISTIC 2.0 Data
Running GISTIC 2.0 on e.g. [GenePattern](https://cloud.genepattern.org) not only provides the [Discrete Copy Number Data](#discrete-copy-number-data), but also provides an amp_genes and a del_genes file. These cannot be directly imported into cBioPortal, but first have to be converted to a different file format. An example can be found in the [ACC TCGA study](https://github.com/cBioPortal/datahub/blob/master/public/acc_tcga/data_gistic_genes_amp.txt) on cBioPortal Datahub.

After uploading a gistic_amp and/or gistic_del file, significantly recurrently copy-number altered genes will be labeled and available for query in the "CNA Genes" table on the study view, like in the [TCGA Legacy / Firehose data set for bladder cancer](https://www.cbioportal.org/study/summary?id=blca_tcga):

![recurrently-altered-genes-table](images/recurrently-altered-genes-table.png)

### Meta file
The Gistic metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: GISTIC_GENES_AMP or GISTIC_GENES_DEL
3. **datatype**: Q-VALUE
4. **reference_genome_id**: reference genome version. Supported values: "hg19"
5. **data_filename**: your datafile

An example metadata file would be:
```
cancer_study_identifier: brca_tcga
genetic_alteration_type: GISTIC_GENES_AMP
datatype: Q-VALUE
reference_genome_id: hg19
data_filename: data_gistic_genes_amp.txt
```

### Data file

The following fields from the generated Gistic file are used by the cBioPortal importer:

* **chromosome**: chromosome on which the region was found, without the `chr` prefix
* **peak_start**: start coordinate of the region of maximal amplification or deletion within the significant region
* **peak_end**: end coordinate of the region of maximal amplification or deletion within the significant region
* **genes_in_region**: comma-separated list of HUGO gene symbols in the 'wide peak' (allowing for single-sample errors in the peak boundaries)
* **amp**: 1 for amp, 0 for del
* **cytoband**: cytogenetic band specification of the region, including chromosome (Giemsa stain)
* **q_value**: the q-value of the peak region

### Example 
An example data file which includes the required column header would look like:

```
chromosome<TAB>peak_start<TAB>peak_end<TAB>genes_in_region<TAB>amp<TAB>cytoband<TAB>q_value<TAB>
1<TAB>150563314<TAB>150621176<TAB>SNORA40|ENSG00000253047.1,RN7SL600P,RN7SL473P,C1orf138,LINC00568,CTSS,ECM1,ENSA,MCL1,RPRD2,ADAMTSL4,GOLPH3L,TARS2,HORMAD1,MIR4257,<TAB>1<TAB>1q21.3<TAB>2.7818E-43<TAB>
1<TAB>85988564<TAB>85991712<TAB>DDAH1,<TAB>1<TAB>1p22.3<TAB>4.1251E-13<TAB>
...
```


## MutSig Data
MutSig stands for "Mutation Significance".  MutSig analyzes lists of mutations discovered in DNA sequencing, to identify genes that were mutated more often than expected by chance given background mutation processes. You can download mutsig from [broadinstitute](https://www.broadinstitute.org/cancer/cga/mutsig_download) (MutSigCV 1.4 is available) or run mutsig (MutSigCV 1.2 is available) using [GenePattern](https://cloud.genepattern.org).

**Note:** The tcga files that are uploaded to cBioPortal are generated using MutSig2.0. This version is not available outside broadinstitute. 

_**The MutSigCV 1.2 output is different from the MutSig2.0 header. TODO: test the 1.4 version. Requires > 10GB of memory**_

After uploading a MutSig file, significantly recurrently mutated genes will be labeled and available for query in the "Mutated Genes" table on the study view, like in the [TCGA Legacy / Firehose data set for bladder cancer](https://www.cbioportal.org/study/summary?id=blca_tcga):

![recurrently-mutated-genes-table](images/recurrently-mutated-genes-table.png) 

This type data is not yet being validated. It can, however, be uploaded.

### Meta file
The MutSig metadata file should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: MUTSIG
3. **datatype**: Q-VALUE
4. **data_filename**: your datafile

An example metadata file would be:
```
cancer_study_identifier: brca_tcga
genetic_alteration_type: MUTSIG
datatype: Q-VALUE
data_filename: data_mutsig.txt
```


### Data file

The following fields from a MutSig file are used by the cBioPortal importer:
* **rank**
* **gene**: this is the HUGO symbol
* **N (or Nnon)**: bases covered
* **n (or nnon)**: number of mutations
* **p**: result of testing the hypothesis that all of the observed mutations in this gene are a consequence of random background mutation processes, taking into account the list of bases that are successfully interrogated by sequencing (i.e., ‚Äúcovered‚Äù) and the list of observed somatic mutations, as well as the length and composition of the gene in addition to the background mutation rates in different sequence contexts (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3059829/)
* **q**: p value correct for multiple testing


### Example 

An example data file which includes the required column header would look like:

```
rank<TAB>gene<TAB>N<TAB>n<TAB>p<TAB>q
1<TAB>RUNX1<TAB>1051659<TAB>29<TAB>1.11E-16<TAB>1.88E-12
2<TAB>PIK3CA<TAB>3200341<TAB>351<TAB><1.00e-15<TAB><2.36e-12
...
```

## Gene Panel Data
Gene panel functionality can specify which genes are assayed on a panel and assign samples and genetic profiles (such as mutation data) to a panel.

To include gene panel data in your instance, the following data and/or configurations can be used:
1. **Gene panel file**: This file contains the genes on the gene panel. A panel can be used for multiple studies within the instance and should be loaded prior to loading a study with gene panel data. For information on the format and import process please visit: [Import-Gene-Panels](Import-Gene-Panels.md).
2. **Gene panel matrix file**: This file is used to specify which samples are sequenced on which gene panel in which genetic profile. This is recommended for mutation and structural variant data, because the MAF and structural variant formats are unable to include samples which are sequenced but contain no called mutations, and only a single gene panel can be defined in the meta file. For other genetic profiles, columns can be added to specify their gene panel, but a property can also be added to their respective meta file, because these data files contain all profiled samples. Although the gene panel matrix functionality overlaps with the case list functionality, a case list for mutations (`_sequenced`) and Structural variants (`_sv`) is also required.
3. **Gene panel property in meta file**: Adding the `gene_panel:` property to the meta file of data profile will assign all samples from that profile to the gene panel. In this case it is not necessary to include a column for this profile in the gene panel matrix file.

### Gene Panel Matrix file

### Columns and rows
The gene panel matrix file contains a list of samples in the first column, and an additional column for each profile in the study using the stable_id as the column header. These stable_id's should match the ones in their respective meta files, for example `mutations` for mutation data and `gistic` for discrete CNA data. Columns should be separated by tabs. Fusion events are saved in the mutation table in the cBioPortal database, so they should be included in the `mutations` column. As described above, genetic profiles other than mutation and fusion data profiles can use the `gene_panel:` meta property if all samples are profiled on the same gene panel.

### Values
For each sample-profile combination, a gene panel should be specified. Please make sure this gene panel is imported before loading the study data. When the sample is not profiled on a gene panel, or if the sample is not profiled at all, use `NA` as value. If the sample is profiled for mutations, make sure it is also in the `_sequenced` case list. 

### Example 
An example file would look like this:

| SAMPLE_ID   | mutations | gistic    |
| ----------- | --------- | --------- |
| SAMPLE_ID_1 | IMPACT410 | IMPACT410 |
| SAMPLE_ID_2 | IMPACT410 | IMPACT410 |
| SAMPLE_ID_3 | NA       | NA        |

### Meta file
The gene panel matrix file requires a meta file, which should contain the following fields:

1. **cancer_study_identifier**: same value as specified in [study meta file](#cancer-study)
2. **genetic_alteration_type**: GENE_PANEL_MATRIX
3. **datatype**: GENE_PANEL_MATRIX
4. **data_filename**: your datafile

Example:
```
cancer_study_identifier: msk_impact_2017
genetic_alteration_type: GENE_PANEL_MATRIX
datatype: GENE_PANEL_MATRIX
data_filename: data_gene_panel_matrix.txt
```

### Gene panel property in meta file
If all samples in a genetic profile have the same gene panel associated with them, an optional field can be specified in the meta data file of that datatype called **gene_panel:**. If this is present, all samples in this data file will be assigned to this gene panel for this specific profile.

## Gene Set Data
A description of importing gene sets (which are required before loading gene set study) can be found [here](Import-Gene-Sets.md). This page also contains a decription to import gene set hierarchy data, which is required to show a hierarchical tree on the query page to select gene sets.

cBioPortal supports GSVA scores and GSVA-like scores, such as ssGSEA. The [Gene Set Variation Analysis](https://www.bioconductor.org/packages/release/bioc/html/GSVA.html) method in R (GSVA, [H√§nzelmann, 2013](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-7)) can calculate several types of scores (specified with the `methods=` argument) and outputs a score between -1 and 1. The GSVA method also calculates a p-value per score using a bootstrapping method.

To import the GSVA(-like) data, a score and p-value data file are required. It is important that the dimensions of the score and p-value file are the same and that they contain the same gene sets and samples. Both data files require a meta file.

### GSVA score meta file
The meta file will be similar to meta files of other genetic profiles, such as mRNA expression. For both GSVA and GSVA-like scores, `GSVA-SCORE` is used as `datatype` and `gsva_scores` is used as `stable_id`. 

Required fields: 
```
cancer_study_identifier: Same value as specified in study meta file
genetic_alteration_type: GENESET_SCORE
datatype: GSVA-SCORE
stable_id: Any unique identifier within the study
source_stable_id: Stable id of the genetic profile (in this same study) that was used as the input source for calculating the GSVA scores. Typically this will be one of the mRNA expression genetic profiles. 
profile_name: A name describing the analysis.
profile_description: A description of the data processing done.
data_filename: <your GSVA score datafile>
show_profile_in_analysis_tab: true
geneset_def_version: Version of the gene set definition this calculation was based on. 
```

Example:
```
cancer_study_identifier: study_es_0
genetic_alteration_type: GENESET_SCORE
datatype: GSVA-SCORE
stable_id: gsva_scores
source_stable_id: rna_seq_mrna
profile_name: GSVA scores on oncogenic signatures gene sets
profile_description: GSVA scores on oncogenic signatures gene sets using mRNA expression data calculated with GSVA version x with parameters x and y.
data_filename: data_gsva_scores.txt
show_profile_in_analysis_tab: true
geneset_def_version: msigdb_7.5.1
```

### GSVA score data file
The data file will be a simple tab separated format, similar to the expression data file: each sample is a column, each gene set a row, each cell contains the GSVA score for that sample x gene set combination.

The first column is `geneset_id` and contains the names of the gene sets. Gene set names should be formatted in uppercase. The other columns are sample columns: An additional column for each sample in the dataset using the sample id as the column header.

The cells contain the GSVA(-like) score: which is real number, between -1.0 and 1.0, representing the score for the gene set in the respective sample, or NA when the score for the gene set in the respective sample could not be (or was not) calculated. Example with 2 gene sets and 3 samples: 

<table>
<thead><tr><th>geneset_id</th><th>TCGA-AO-A0J</th><th>TCGA-A2-A0Y</th><th>TCGA-A2-A0S</th></tr></thead>
<tr><td>GO_POTASSIUM_ION_TRANSPORT</td><td>-0.987</td><td>0.423</td><td>-0.879</td></tr>
<tr><td>GO_GLUCURONATE_METABOLIC_PROCES</td><td>0.546</td><td>0.654</td><td>0.123</td></tr>
<tr><td>..</td><td></td><td></td><td></td></tr>
</table>

### GSVA p-value meta file

For both GSVA and GSVA-like p-values, `P-VALUE` is used as `datatype` and `gsva_pvalues` is used as `stable_id`. 

Required fields: 
```
cancer_study_identifier: Same value as specified in study meta file
genetic_alteration_type: GENESET_SCORE
datatype: P-VALUE
stable_id: Any unique identifier within the study
source_stable_id: Stable id of the GSVA-SCORE genetic profile (see above).
profile_name: A name describing the analysis.
profile_description: A description of the data processing done.
data_filename: <your GSVA p-value datafile>
geneset_def_version: Version of the gene sets definition this calculation was based on. 
```

Example:
```
cancer_study_identifier: study_es_0
genetic_alteration_type: GENESET_SCORE
datatype: P-VALUE
stable_id: gsva_pvalues
source_stable_id: gsva_scores
profile_name: GSVA p-values for GSVA scores on oncogenic signatures gene sets
profile_description: GSVA p-values for GSVA scores on oncogenic signatures gene sets using mRNA expression data calculated with the bootstrapping method in GSVA version x with parameters x and y.
data_filename: data_gsva_pvalues.txt
geneset_def_version: msigdb_7.5.1
```

### GSVA p-value data file
The data file will be a simple tab separated format, similar to the score file: each sample is a column, each gene set a row, each cell contains the p-value for the score found for sample x gene set combination.

The first column is `geneset_id` and contains the names of the gene sets. Gene set names should be formatted in uppercase. The other columns are sample columns: An additional column for each sample in the dataset using the sample id as the column header.

The cells contain the p-value for the GSVA score: A real number, between 0.0 and 1.0, representing the p-value for the GSVA score calculated for the gene set in the respective sample, or NA when the score for the gene is also NA. Example with 2 gene sets and 3 samples: 

<table>
<thead><tr><th>geneset_id</th><th>TCGA-AO-A0J</th><th>TCGA-A2-A0Y</th><th>TCGA-A2-A0S</th></tr></thead>
<tr><td>GO_POTASSIUM_ION_TRANSPORT</td><td>0.0811</td><td>0.0431</td><td>0.0087</td></tr>
<tr><td>GO_GLUCURONATE_METABOLIC_PROCES</td><td>0.6621</td><td>0.0031</td><td>1.52e-9</td></tr>
<tr><td>..</td><td></td><td></td><td></td></tr>
</table>

## Study Tags file
YAML or JSON file which contains extra information about the cancer study. No compulsory fields are required for this file (free-form). To enable this feature, you need to add a line in the cancer study meta file with `tags_file:` followed the YAML/JSON file name. The information on the YAML or JSON file will be displayed in a table when mousing over a tag logo in the studies on the query page.

## Generic Assay
Generic Assay is a two dimensional matrix generalized to capture non-genetic measurements per sample. Instead of a gene per row and a sample per column, a Generic Assay file contains a generic entity per row and a sample per column. A generic entity is defined by the data curator and generally means something other than a gene. Some examples include, treatment response or mutational signatures. For each generic entity - sample pair, a (real number / text / binary value) represents a captured measurement.

### Generic Assay meta file
The generic assay metadata file should contain the following fields:
```
cancer_study_identifier: Same value as specified in meta file of the study
genetic_alteration_type: GENERIC_ASSAY
generic_assay_type: <GENERIC_ASSAY_TYPE>, e.g., "TREATMENT_RESPONSE" or "MUTATIONAL_SIGNATURE"
datatype: LIMIT-VALUE, CATEGORICAL or BINARY
stable_id: Any unique identifier using a combination of alphanumeric characters, _ and -
profile_name: A name describing the analysis.
profile_description: A description of the data processing done.
data_filename: <name of generic assay data file>
show_profile_in_analysis_tab: true (or false if you don't want to show this profile in any analysis tab)
pivot_threshold_value: A threshold value beyond which a generic assay data is considered effective
value_sort_order: A flag that determines whether samples with small generic assay data values are displayed first or last; can be 'ASC' for small first, 'DESC' for small last. 
patient_level: false (or true if your data is patient level data)
generic_entity_meta_properties: A comma separate list of generic entity properties, e.g., "NAME,DESCRIPTION,URL"
```

Example:
```
cancer_study_identifier: study_es_0
genetic_alteration_type: GENERIC_ASSAY
generic_assay_type: TREATMENT_RESPONSE
datatype: LIMIT-VALUE
stable_id: treatment_ic50
profile_name: IC50 values of compounds on cellular phenotype readout
profile_description: IC50 (compound concentration resulting in half maximal inhibition) of compounds on cellular phenotype readout of cultured mutant cell lines.
data_filename: data_treatment_ic50.txt
show_profile_in_analysis_tab: true
pivot_threshold_value: 0.1
value_sort_order: ASC
patient_level: false
generic_entity_meta_properties: NAME,DESCRIPTION,URL
```

### Note on `stable_id`
The `stable_id` for the generic assay datatype is a user defined field. The only requirement is that the `stable_id` is unique across all metafiles in the study.

### Note on `pivot_threshold_value`
The `pivot_threshold_value` is an arbitrary value that specifies a critical boundary that distinguish _important_ from
_unimportant_ values in a generic assay profile. This boundary will be used in different visualizations to highlight
_important_ observations. Whether smaller or larger values are considered to be _important_ can be controlled with the
`value_sort_order` field (see [below](#note-on-value_sort_order)).

In a heatmap for a generic assay profile _important_ values are shown in darker shades of blue, whereas _unimportant_
values are shown in darker shades of red. The values represented by the `pivot_threshold_value` is shown in white. When
defined, the `pivot_threshold_value` will always be included in the legend, even when all datapoints are all more extreme.
In the Waterfall plot in Plots tab of Results view `pivot_threshold_value` determines the boundary between up- and downward
deflections in the plot.

When no sensible idea exists for a boundary between _important_ and _unimportant_ observations the `pivot_threshold_value`
field should not be defined (excluded from meta file). In this case all values will be shown as shades of blue in the
heatmap and the waterfall plot will show up- and downward deflections around 0.

### Note on `value_sort_order`
The `value_sort_order` field can be used to indicate whether small or large value are considered to be more `important`.
When `value_sort_order` is `ASC` smaller values are considered to be more _important_. When `value_sort_order` is `DESC`
larger values are considered to be more _important_.

The default value for `value_sort_order` is `ASC`.

The `value_sort_order` is used by the Oncoprint when aggregating generic assay data in a tooltip that covers multiple samples
from a single patient. When `value_sort_order` is `ASC` the sample with the smallest response value will be shown for that
patient. When `value_sort_order` is `DESC` the sample with the largest value will be shown for that patient.

In the heatmap for a generic assay profile in results view data points with smaller values will show as darker blue
when the `value_sort_order` is `ASC`. When `value_sort_order` is `DESC` larger values are assigned a darker blue color.

The `value_sort_order` is used by the Waterfall plot for orientation of the x-axis so that _important_ observations are
shown at the left side of the plot. When `value_sort_order` is `ASC` the x-axis will be in ascending order with smaller
values to the left.  When `value_sort_order` is `DESC` the x-axis will be in descending order with larger values to the left.

### Note on `generic_entity_meta_properties`
All meta properties must be specified in the `generic_entity_meta_properties` field. Every meta property listed here 
must appear as a column header in the corresponding data file. It's highly recommend to add `NAME`, `DESCRIPTION` and an 
optional `URL` to get the best visualization on OncoPrint tab and Plots tab.

### Note on `patient_level`
Generic Assay data will be considered `sample_level` data if the `patient_level` property is missing or set to `false`. 
In addition, the patient or sample identifiers need to be included in the [Clinical Data](#clinical-data) file.

### Note on `Generic Assay` genetic_alteration_type and datatype
All generic assay data is registered to be of the type of `genetic_alteration_type` and data type can choose from `LIMIT-VALUE`, `CATEGORICAL` and `BINARY`. 
- ***LIMIT-VALUE***: This datatype is intended to be used for any numerical data set with similar structure (entities measured in samples). The `LIMIT-VALUE` is validated to contain any continuous number optionally prefixed with a '>' or '<' threshold symbol (e.g., '>8.00').
- ***CATEGORICAL (under development)***: This datatype is intended to be used for any categorical data set with similar structure (entities measured in samples). Any text is allowed in `CATEGORICAL`.
- ***BINARY (under development)***: This datatype is intended to be used for any binary data set with similar structure (entities measured in samples). The `BINARY` is validated to contain only reserved text (`true`, `false`, `yes`, `no`).

If the value for the generic entity in the respective sample could not (or was not) be measured (or detected), the value should be 'NA' or leave that cell blank.

### Generic Assay data file
The data file will be a simple tab separated format, similar to the expression data file: each sample is a column, each generic entity is a row, each cell contains values for that generic entity x sample combination.

For each generic entity (row) in the data file, the following columns are required in the order specified:

- ***ENTITY_STABLE_ID***: Any unique identifier using a combination of alphanumeric characters, _ and -.

And:
- An additional column for each `generic_entity_meta_properties` in the metafile, using the property name as the column header (e.g., 'NAME').
- An additional column for each sample in the dataset using the sample id as the column header.

Example with 3 generic entities and 3 samples:

<table>
<thead><tr><th>ENTITY_STABLE_ID</th><th>NAME</th><th>DESCRIPTION</th><th>URL</th><th>1321N1_CENTRAL_NERVOUS_SYSTEM</th><th>22RV1_PROSTATE</th><th>42MGBA_CENTRAL_NERVOUS_SYSTEM</th></tr></thead>
<tr><td>17-AAG</td><td>Tanespimycin</td><td>Hsp90 inhibitor</td><td>https://en.wikipedia.org/wiki/Tanespimycin</td><td>0.228</td><td>0.330</td><td>0.0530</td></tr>
<tr><td>AEW541</td><td>Larotrectinib</td><td>TrkA/B/C inhibitor</td><td>https://en.wikipedia.org/wiki/Larotrectinib</td><td>>8</td><td>2.33</td><td>2.68</td></tr>
<tr><td>AZD0530</td><td>Saracatinib</td><td>Src/Bcr-Abl inhibitor</td><td>https://en.wikipedia.org/wiki/Saracatinib</td><td>NA</td><td>>8</td><td>4.60</td></tr>
</table>

### Arm Level CNA Data
Arm-level copy-number data is a predefined subtype of Generic Assay Data.

Allowed values for Arm-level copy-number data are `Loss`, `Gain`, and `Unchanged`, use `NA` or leave the cell blank to indicate a missing value.

Please find example file format here: [Meta file example](https://github.com/cBioPortal/cbioportal-frontend/blob/master/end-to-end-test/local/studies/lgg_ucsf_2014_test_generic_assay/meta_armlevel_CNA.txt) and [Data file example](https://github.com/cBioPortal/cbioportal-frontend/blob/master/end-to-end-test/local/studies/lgg_ucsf_2014_test_generic_assay/data_armlevel_CNA.txt)

### Mutational Signature Data
Mutational Signature data is a predefined subtype of Generic Assay Data. Setting `generic_assay_type: MUTATIONAL_SIGNATURE` 
in the meta file will make cBioPortal interpret the data as Mutational Signature data.

#### Mutational Signature meta files
The mutational signature meta files follow the same convention as the [Generic Assay Meta file](#generic-assay-meta-file),
however there are some key differences:
- `genetic_assay_type` should be set to `MUTATIONAL_SIGNATURE`
- `datatype` should be set to `LIMIT_VALUE`
- `stable_id` values should end with: `_{filetype}_{identifier}`, where:
  - `filetype` is either `contribution`, `pvalue` or `counts`
  - `identifier` is consistent between files belonging to the same analysis
  - Multiple signatures can be added to a single study, as long as they have different identifiers in their stable id (e.g., `contribution_SBS` and `contribution_DBS`)
- In `generic_entity_meta_properties` the `NAME` value is required. The `DESCRIPTION` and `URL` values can be added 
  to display more information and link to external resources in the mutational signatures tab.

#### Mutational Signature data files
The mutational signature data files follow the same convention as the [Generic Assay Data file](#generic-assay-data-file).
Each collection of mutational signatures can consist of up to three different data files, each with an accompanying meta file.
- Signature _contribution_ file (**required**)
  - Data file containing the contribution of each signature-sample pair. Values are expected to be 0 ‚â• x ‚â• 1.
- Signature _pvalue_ file (optional)
  - Data file containing p-values for each signature-sample pair. Values below 0.05 will be shown as significant.
- Mutational _counts_ matrix file (optional)
  - Data file containing nucleotide changes of a sample. cBioPortal has specific visualization options for single-base substitutions (96 channels), double-base substitutions (72 channels) and insertion/deletions (83 channels), following the signatures defined by [COSMIC](https://cancer.sanger.ac.uk/signatures/). But other channels can also be used. Values are expected to be positive integers.

## Resource Data

The resource data is used to capture resource data in patients, samples and studies. The resources will be represented by URLs with meta data. The types of resources include:
- Files: pdf, txt, png, json, etc.
- Web links: non-file links e.g. URLs to other systems

the resource file is split into a resource definition file, sample resource file, patient resource file and study resource file. All data files are required to have a matching meta file.

### Meta files
The resource metadata files have to contain the following fields:

1. **cancer_study_identifier**: same value specified in meta_study.txt
2. **resource_type**: value from (DEFINITION / SAMPLE / PATIENT / STUDY)
3. **data_filename**: your datafile

### Examples
An example metadata file, e.g. named meta_resource_definition.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
resource_type: DEFINITION
data_filename: data_resource_definition.txt
```

An example metadata file, e.g. named meta_resource_sample.txt, would be:
```
cancer_study_identifier: brca_tcga_pub
resource_type: SAMPLE
data_filename: data_resource_sample.txt
```

### Data files

### Resource Definition Data File
The resource definition file should follow this format, it has three **required** columns:
- **RESOURCE_ID (required)**: a unique resource ID. This field allows only numbers, letters, points, underscores and hyphens.
- **DISPLAY_NAME (required)**: a display name for resources.
- **RESOURCE_TYPE (required)**: resource type for resources, must be SAMPLE, PATIENT or STUDY.
- **DESCRIPTION (optional)**: a discription for resources.
- **OPEN_BY_DEFAULT (optional)**: define if the resource will be open by default (`true` / `false`), dafault is `false`.
- **PRIORITY (optional)**: if not given, will give a default value.

### Example *Resource Definition* data file
<table>
<thead><tr><th>RESOURCE_ID</th><th>DISPLAY_NAME</th><th>RESOURCE_TYPE</th><th>DESCRIPTION</th><th>OPEN_BY_DEFAULT</th><th>PRIORITY</th></tr></thead>
<tr><td>PATHOLOGY_SLIDE</td><td>Pathology Slide</td><td>SAMPLE</td><td>The pathology slide for the sample</td><td>TRUE</td><td>1</td></tr>
<tr><td>PATIENT_NOTES</td><td>Patient Notes</td><td>PATIENT</td><td>Notes about the patient</td><td>FALSE</td><td>2</td></tr>
<tr><td>STUDY_SPONSORS</td><td>Study Sponsors</td><td>STUDY</td><td>Sponsors of this study</td><td>TRUE</td><td>3</td></tr>
</table>

### Sample Resource Data File
The sample resource file should follow this format, it has four **required** columns:
- **PATIENT_ID (required)**: a unique patient ID. This field allows only numbers, letters, points, underscores and hyphens.
- **SAMPLE_ID (required)**: a unique sample ID. This field allows only numbers, letters, points, underscores and hyphens.
- **RESOURCE_ID (required)**: a unique resource ID which should also be included in the `Resource Definition data file`.
- **URL (required)**: url to the resources, start with `http` or `https`.

### Example *Sample Resource* data file
<table>
<thead><tr><th>PATIENT_ID</th><th>SAMPLE_ID</th><th>RESOURCE_ID</th><th>URL</th></tr></thead>
<tr><td>TCGA-A2-A04P</td><td>TCGA-A2-A04P-01</td><td>PATHOLOGY_SLIDE</td><td>https://url-to-slide-sample1</td></tr>
<tr><td>TCGA-A1-A0SK</td><td>TCGA-A1-A0SK-01</td><td>PATHOLOGY_SLIDE</td><td>https://url-to-slide-sample2</td></tr>
</table>

### Patient Resource Data File
The patient resource file should follow this format, it has three **required** columns:
- **PATIENT_ID (required)**: a unique patient ID. This field allows only numbers, letters, points, underscores and hyphens.
- **RESOURCE_ID (required)**: a unique resource ID which should also be included in the `Resource Definition data file`.
- **URL (required)**: url to the resources, start with `http` or `https`.

### Example *Patient Resource* data file
<table>
<thead><tr><th>PATIENT_ID</th><th>RESOURCE_ID</th><th>URL</th></tr></thead>
<tr><td>TCGA-A2-A04P</td><td>PATIENT_NOTES</td><td>https://url-to-slide-patient1</td></tr>
<tr><td>TCGA-A1-A0SK</td><td>PATIENT_NOTES</td><td>https://url-to-slide-patient2</td></tr>
</table>

### Study Resource Data File
The study resource file should follow this format, it has two **required** columns:
- **RESOURCE_ID (required)**: a unique resource ID which should also be included in the `Resource Definition data file`.
- **URL (required)**: url to the resources, start with `http` or `https`.

### Example *Study Resource* data file
<table>
<thead><tr><th>RESOURCE_ID</th><th>URL</th></tr></thead>
<tr><td>STUDY_SPONSORS</td><td>https://url-to-study-sponsors</td></tr>
</table>

## Custom namespace columns

### Adding annotation columns through namespaces
Custom columns can be added to the data files of mutations, structural variants and discrete copy number (long) data. 
The columns can be imported through the namespace mechanism into a database table column called `ANNOTATION_JSON`. Any columns starting with a prefix specified in the `namespaces` field in the metafile will be imported into the database. Namespace columns should be formatted as the namespace and namespace attribute seperated with a period (e.g `ASCN.total_copy_number` where `ASCN` is the namespace and `total_copy_number` is the attribute).

An example cBioPortal mutation data file with the following **additional** columns:
```
ASCN.total_copy_number    ASCN.clonal     MUTATION.name    MUTATION.type
```
imported with the following `namespaces` field in the metafile:
```
namespaces: ascn
```
will import the `ASCN.total_copy_number` and `ASCN.clonal` column into the database. `MUTATION.name` and `MUTATION.type` 
will be ignored because `mutation` is not specified in the `namespaces` field. 

## Representation of namespace columns by mutation API endpoints

Columns added through namespaces will be returned by relevant mutation, discrete copy number and structural variant API 
endpoints. Namespace data will be available in the `namespaceColumn` of respective JSON representations of mutation 
records. The `namespaceColumns` property will be a JSON object where namespace data is keyed by name of the namespace in 
lowercase. For instance, when namespace `ZYGOSITY` is defined in the meta file and the data file has column 
`ZYGOSITY.status` with value `Homozygous` for a mutation row, the API will return the following JSON record for this 
mutation (only relevant fields are shown):

```
{
    "namespaceColumns": {
        "ZYGOSITY": {
            "status": "Homozygous"
        }
    },
}
```

Note: ASCN namespace data is not exported via the `namespaceColumns` field.

## Representation of namespace columns in the cBioPortal frontend

Namespace columns will be added as columns to mutation, structural variant and copy number alteration tables in Patient 
View and Results View. The case of the namespace in the column header will be as specified in the mutations meta file 
and the column name will be capitalized. 

For instance, this metafile entry:

```shell
namespaces: Zygosity
```

and this column header:

```shell
ZYGOSITY.status
```

will show in the mutation table with column name:

```shell
Zygosity Status
```

Note: namespace columns are recognized by a case-insensitive match of the namespace reported in the meta file
and the first word in the column header.
# Hardware Requirements

Hardware requirements will vary depending on the volume of users you anticipate
will access your cBioPortal instance and the amount of data loaded in the
portal. We run [cbioportal.org](https://www.cbioportal.org) on an AWS r5.xlarge
instance with 32 GB and 4 vCPUs. The public database consumes ~50 GB of disk
space. The site is visited by several thousands of users a day. For on-premise
installation recommendations one can look at the AWS instance type specs:
 
| Platform | instance type | (v)CPUs | RAM(GB) | Storage (GB) |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| aws | r5.xlarge | 4 | 32 | 50 |
| on-premise | - | 4 | 32 | 50 |


The hardware requirements are much lower when one has only a few users a day.
Minimally, 2GB of RAM is needed to run a cBioPortal instance. If you do not
plan to import public studies, depending on the size of your private data, 10GB
of disk space may be sufficient.

Another possible consideration is caching. The portal can cache responses to
requests so that repeated database queries are avoided. On the public cBioPortal
deployment we enable this cache and allocate 1GB of additional RAM and 4GB of
additional disk space for caching. For directions on configuring caching, see
[Ehcache Settings](/deployment/customization/application.properties-Reference.md#cache-settings)
# Import Gene Panels

This page describes how to import a gene panel into the cBioPortal database.  It assumes the following requirements have been satisfied:

1. The cBioPortal software has been correctly [built from source](/deployment/deploy-without-docker/Build-from-Source.md).
2. The gene panel to import is in the proper file format. See [Gene Panel File format](#gene-panel-file-format) for more information.
3. The `PORTAL_HOME` environment variable has been properly defined.  See [Loading a Sample Study](/deployment/deploy-without-docker/Load-Sample-Cancer-Study.md#set-the-portal_home-environment-variable) for more information.

## Gene panel file format
The gene panel file follows the format of a meta file with the following fields:
1. **stable_id**: The name of the gene panel. This should be unique across all studies, as gene panels can be globally applied to any sample and any genetic profile.
2. **description**: A description of the gene panel.
3. **gene_list**: Tab separated genes, represented either by all gene symbols or all Entrez gene IDs.

An example gene panel file would be:
```
stable_id: IMPACT410
description: Targeted (410 cancer genes) sequencing of various tumor types via MSK-IMPACT on Illumina HiSeq sequencers.
gene_list: ABL1    ACVR1   AKT1    AKT3 ...
```

#### Import command

In this example, we are loading the example gene panels which resides in the sample dataset `study_es_0`.

```
cd <cbioportal_source_folder>/core/src/main/scripts
./importGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel1.txt
./importGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel2.txt
```

After loading gene panels into the database, please restart Tomcat or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#cache-settings) for more information) so that the
validator can retrieve gene panel information from the cBioPortal API. 

#### Update existing gene panel

If a gene panel exists in the database with the same name as the one being imported, and there exists cancer study data that refers to this gene panel, the ImportGenePanel command will abort.  In order to reimport the gene panel in this situation, run the UpdateGenePanel command.  

If the incoming gene panel is the same as the original gene panel, whether through importing or updating, then no changes shall be made to the gene panel.  If the incoming gene panel is empty, then the script will abort.  Genes in the incoming gene panel that were not in the original shall be added to the existing gene panel. Conversely, genes not in the incoming gene panel that were in the original shall be removed from the existing gene panel.  The UpdateGenePanel command will prompt twice to confirm changes made to the gene panel, such as genes to be added or removed.  

```
cd <cbioportal_source_folder>/core/src/main/scripts
./updateGenePanel.pl --data ../../test/scripts/test_data/study_es_0/data_gene_panel_testpanel1.txt
```
# Import Gene Sets in cBioPortal

Gene sets are collections of genes that are grouped together based on higher level function or system characteristics, such as being part of the same molecular process or found to be co-regulated for example. Assessing gene sets in cBioPortal is useful when the user wants to visualize the number of mutations in sets of genes, or wants to see if all genes in a set are up- or down-regulated. To visualize gene set variation in a sample, the user can calculate scores per gene set per sample using the Gene Set Variation Analysis (GSVA) algorithm ([H√§nzelmann, 2013](#references)).

Before loading a study with gene set data, gene set definitions have to be added to the database. These can be custom user-defined sets, or sets downloaded from external sources such as [MSigDB](#references). Additionally, a gene set hierarchy can be imported which is used on the cBioPortal Query page for selecting gene sets.

## Quick example
This example shows how the process of importing gene set data using test data.

1. Navigate to scripts folder:

```
cd <cbioportal_source_folder>/core/src/main/scripts
```

2. Import gene sets and supplementary data:
Note: This removes existing gene set, gene set hierarchy and gene set genetic profile data.

```
./importGenesetData.pl \
	--data ../../test/resources/genesets/study_es_0_genesets.gmt \
	--new-version msigdb_6.1 \
	--supp ../../test/resources/genesets/study_es_0_supp-genesets.txt
```

3. Import gene set hierarchy data:

```
./importGenesetHierarchy.pl \
	--data ../../test/resources/genesets/study_es_0_tree.yaml
```

4. Restart Tomcat if you have it running or call the `/api/cache` endpoint with a `DELETE` http-request
   (see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information).


5. Import study (replace argument after `-u` with local cBioPortal and `-html` with preferred location for html report):

```
./importer/metaImport.py \
	-s ../../test/scripts/test_data/study_es_0 \
	-u http://cbioportal:8080/cbioportal \
	-html report_study_es_0.html \
	-v \
	-o
```

## Requirements for gene sets in cBioPortal
Gene set functionality was added in cBioPortal 1.7.0. Please use this or a later version. In addition, the database has to be updated to version 2.3.0 or higher, depending on the cBioPortal version. This can be done by running the python wrapper `migrate_db.py` for `migration.sql`.

Updating the database is described [here](https://github.com/cBioPortal/cbioportal/blob/master/docs/Updating-your-cBioPortal-installation.md#running-the-migration-script).

## Import Gene Sets

### File formats
Once you have initialized MySQL with cBioPortal database, it is possible to import gene sets. The format of the gene set data file is [the Gene Matrix Transposed file format (.gmt)](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29). This format is also used by the [MSigDB](#references), which hosts several collections of gene sets on: [https://software.broadinstitute.org/gsea/msigdb/](https://software.broadinstitute.org/gsea/msigdb/)

Sample of .gmt file:

```
GLI1_UP.V1_DN<TAB>https://...<TAB>22818<TAB>143384
GLI1_UP.V1_UP<TAB>https://...<TAB>3489<TAB>3119
E2F1_UP.V1_DN<TAB>https://...<TAB>7041<TAB>6374<TAB>5460
```

GMT files contain a row for every gene set. The first column contains the EXTERNAL_ID or `stable id` (MsigDB calls this "standard name"), e.g. GO_POTASSIUM_ION_TRANSPORT, not longer than 100 characters. The second column contains the REF_LINK. This is an optional URL linking to external information about this gene set. Column 3 to N contain the Entrez gene IDs that belong to this gene set.

Additional information can be placed in a supplementary file. This file should be a .txt, containing columns for the `stable id`, the long name (max 100 characters) and description of the gene set (max 300 characters).

Sample of supplementary .txt file:

```
GLI1_UP.V1_DN<TAB>GLI1 upregulated v1 down genes<TAB>Genes down-regulated in RK3E cells (kidney epithelium) over-expressing GLI1 [GeneID=2735].
GLI1_UP.V1_UP<TAB>GLI1 upregulated v1 up genes<TAB>Genes up-regulated in RK3E cells (kidney epithelium) over-expressing GLI1 [GeneID=2735].
E2F1_UP.V1_DN<TAB>E2F1 upregulated v1 down genes<TAB>Identification of E2F1-regulated genes that modulate the transition from quiescence into DNA synthesis, or have roles in apoptosis, signal transduction, membrane biology, and transcription repression.

```
### Run the gene set importer
The importer for gene sets can be run with a perl wrapper, which is located at the following location and requires the following arguments:

```
cd <cbioportal_source_folder>/core/src/main/scripts
perl importGenesetData.pl

required:     --data <data_file.gmt>
              --new-version <Version> OR --update-info
optional:     --supp <supp_file.txt>
```
The `--new-version` argument with a `<Version>` parameter is used for loading new gene set definitions. It is not possible to add new gene sets or change the genes of current gene sets, without removing the old gene sets first. This is to prevent the user from having gene sets from different definitions and data from older definitions. The user can choose the name or number of the `<Version>` as he likes, e.g. `msigdb_6.1` or `Oncogenic_2017`. Running the script with `--new-version` **removes all previous gene sets, gene set hierarchy and gene set genetic profiles.** A prompt is given to make sure the user wants to do this. Note that it is possible enter the same version as the previous version, but previous data is removed nevertheless.

The `--update info` can be used only to update only the long name, description and reference URL.

## Import Gene Set hierarchy

After importing gene sets, you can import a gene set hierarchy that is used on the query page to select gene sets.

### File format
For gene set hierarchy files, we use the YAML format. This is common format to structure hierarchical data.

Sample of format (note this is mock data):

```
Custom:
  Gene sets:
    - BCAT.100_UP.V1_DN
Cancer Gene sets from Broad:
  Gene sets:
    - AKT_UP.V1_DN
    - CYCLIN_D1_KE_.V1_UP
  Broad Subcategory 1:
    Gene sets:
      - HINATA_NFKB_MATRIX
  Broad Subcategory 2:
    Gene sets:
      - GLI1_UP.V1_UP
      - GLI1_UP.V1_DN
      - CYCLIN_D1_KE_.V1_UP

```

To make your own hierarchy, make sure every branchname ends with `:`. Every branch can contain new branches (which can be considered subcategories) or gene sets (which are designated by the `Gene sets:` statement). The gene set names are the `stable ids` imported by `ImportGenesetData.java` and should start with `-`.

### Running the gene set hierarchy importer
```
cd <cbioportal_source_folder>/core/src/main/scripts
perl importGenesetHierarchy.pl

required:     --data <data_file.yaml>
```

## Import a study with gene set data

Gene set data can be added to a study folder and subsequently import the whole study with `metaImport.py`. cBioPortal supports GSVA Scores and p-values (from bootstrapping) calculated using Gene Set Variation Analysis (GSVA, [H√§nzelmann, 2013](#references)).
A description of GSVA study data can be found in the [cBioPortal File Formats documentation](File-Formats.md#gene-set-data).

## References
**GSVA: gene set variation analysis for microarray and RNA-Seq data**<br>
Sonja H√§nzelmann, Robert Castelo and Justin Guinney, *BMC Bioinformatics, 2013*<br>
https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-7 <br>
https://www.bioconductor.org/packages/release/bioc/html/GSVA.html

**Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles**<br>
Aravind Subramanian, Pablo Tamayo, Vamsi K. Mootha, Sayan Mukherjee, Benjamin L. Ebert, Michael A. Gillette, Amanda Paulovich, Scott L. Pomeroy, Todd R. Golub, Eric S. Lander, and Jill P. Mesirov, *PNAS, 2005*<br>
https://www.pnas.org/content/102/43/15545 <br>
https://software.broadinstitute.org/gsea/msigdb
# Import OncoKB annotations as custom driver annotations

The _Annotation Configuration_ menu in Study View and Group Comparison is available only when _custom driver annotations_
are present in the cBioPortal database for the genes in the study (or studies). In order to use OncoKB annotations to filter
mutations and discrete copy number alteration in Study View and Group Comparison, OncoKB annotations can be added to the 
respective data files of a study prior to import into the database. This page describes how to import OncoKB annotations 
as custom driver annotations. It assumes the following requirements have been satisfied:

1. The cBioPortal software has been correctly [built from source](/deployment/deploy-without-docker/Build-from-Source.md).
2. The user is able to successfully import a [study into the database](/Data-Loading.md).
3. The study subjected to OncoKB import is [confirmed to be valid](/Data-Loading.md#validating-your-study-data)

## Import of OncoKB annotations when loading a study

OncoKB annotations can be added automatically to the study files when the study is loaded into the database by adding the
`--import_oncokb` parameter to the [metaImport.py](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/importer/metaImport.py) script like so:

```
python metaImport.py --import_oncokb -u http://cbioportal:8080 -s study/lgg_ucsf_2014/
```

This will add OncoKB data to the mutation and discrete CNA files of a study, revalidate the results and load the study
into the database.

The addition of mutation and discrete CNA files is explained in detail below.

## Update of MAF file with OncoKB annotations

OncoKB annotations can be added to the MAF file by running [importOncokbMutation.py](https://github.com/cBioPortal/cbioportal/tree/master/core/src/main/scripts/importer/importOncokbMutation.py) like so:

```
python importOncokbMutation.py -s study/lgg_ucsf_2014/ -u https://cbioportal.org
```

Where `-s` is the path to the directory of the MAF file and `-u` is the URL to 
a cBioPortal instance (needed for resolution of gene identifiers).

[importOncokbMutation.py](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/importer/importOncokbMutation.py) will add OncoKB annotations as
[custom driver annotation columns](/File-Formats.md#custom-driver-annotations) in the MAF file. The unmodified MAF file
will be stored in the study directory with the _ONCOKB_IMPORT_BACKUP__ prefix.


## Update of Discrete Copy Number file with OncoKB annotations

OncoKB annotations can be added to the Discrete Copy Number data by running [importOncokbDiscreteCNA.py](https://github.com/cBioPortal/cbioportal/tree/master/core/src/main/scripts/importer/importOncokbDiscreteCNA.py) like so:

```
python importOncokbDiscreteCNA.py -s /tmp/study -u https://cbioportal.org
```

Where `-s` is the path to the directory of the iscrete Copy Number data file and `-u` is the URL to 
a cBioPortal instance (needed for resolution of gene identifiers).

[importOncokbDiscreteCNA.py](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/importer/importOncokbMutation.py) will create a [custom driver annotation file](File-Formats.md#custom-driver-annotations-file)
with name _data_cna_pd_annotation.txt_ in the study directory. It will add a field `pd_annotations_filename` field in the 
CNA meta file that references the newly created custom driver annotation file. The unmodified CNA meta file will be stored
in the study directory with the _ONCOKB_IMPORT_BACKUP__ prefix.
# Add a new reference genome to the database
This manual is intended to guide you to load new reference genome(s) to the database. 

### What is Reference Genome (also know as reference assembly)?
A digital nucleic acid sequence database assembled by scientists as a representative example of a species' set of genes. Reference genomes can be accessed online using 
dedicated browsers such as [UCSC Genome Browser](https://genome.ucsc.edu/cgi-bin/hgGateway). 

### Why does reference genomes matter?
Mutation and Segment plots on Patient View are using genomic coordinates. The **cytoband** column in the Mutated Gene table on Study view rely on the Genome Reference Consortium Build 
matching the version used to analyse the original data (before it was loaded into cBioPortal DB).

## How to load reference genome(s) to the database?

### 1. prepare a tab delimited txt file with the following fields:
1. species: the group of organisms e.g human
2. name: the name of reference genome as used by the UCSC browser eg. hg19
3. build_name: the version of Genome Reference Consortium Build published by NCBI e.g GRCh38
4. nonN_bases:  the total number of non-N bases in reference genome FASTA-formatted file
5. URL: the URL to download the reference genome
6. release_date: when reference genome released, in a format of yyyy-mm-dd. The release date normally is included in the README.txt file in the download directory.

here is a sample reference genome file:
```
#species	name	build_name	nonN_bases	URL	release_date
human	hg19	GRCh37	2897310462	https://hgdownload.cse.ucsc.edu/goldenPath/hg19	2009-02-01
human	hg38	GRCh38	3049315783	https://hgdownload.cse.ucsc.edu/goldenPath/hg38	2013-12-24
mouse	mm10	GRCm38	2652783500	https://hgdownload.cse.ucsc.edu/goldenPath/mm10	2011-12-01
```
### 2. Using import script to load reference genome(s) data to the database:
```
 cd <your_cbioportal_dir>/core/src/main/scripts
 export PORTAL_HOME=<your_cbioportal_dir>
./importReferenceGenome.pl --ref-genome <your_reference_genome_file>
```
### 3. output of the script runs successfully
```
$ ./importReferenceGenome.pl --ref-genome ~/myspace/cbioportal/core/src/test/resources/reference_genomes.txt
   Reading reference genome from:  /Users/kelsyzhu/myspace/cbioportal/core/src/test/resources/reference_genomes.txt
    --> total number of lines:  3
   
   Done. Restart tomcat to make sure the cache is replaced with the new data.
   
   Warnings / Errors:
   -------------------
   0.  New reference genome added; 3x
   Done.
   Total time:  3432 ms

```# Import Study Using Docker

:warning: Every time you add/remove/overwrite a study please restart tomcat (or the Docker container), or 
call the `/api/cache` endpoint with a `DELETE` http-request (see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information).

## Adding a Study

In a docker terminal type the following command:

```bash
docker exec -it ${CBIOPORTAL_CONTAINER_NAME} \
metaImport.py -u http://${CBIOPORTAL_CONTAINER_NAME}:8080/cbioportal -s ${PATH_TO_STUDIES}
```

Where:    
- **${CBIOPORTAL_CONTAINER_NAME}**: The name of your cbioportal container instance, _i.e **cbioportal**_.
- **${PATH_TO_STUDIES}**: The external path where cBioPortal studies are stored.

## Restarting Tomcat

In a docker terminal type the following command:

```bash
docker exec -it ${CBIOPORTAL_CONTAINER_NAME} \
sh '${CATALINA_HOME}'/bin/shutdown.sh \
'&&' sh '${CATALINA_HOME}'/bin/startup.sh \
```

Where:    
- **${CBIOPORTAL_CONTAINER_NAME}**: The name of your cbioportal container instance, _i.e **cbioportal**_.
Relevant classes:
* [MafUtil](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/maf/MafUtil.java)
* [ExtendedMutationUtil](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/util/ExtendedMutationUtil.java)
* [MafRecord](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/maf/MafRecord.java)

### [MafUtil](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/maf/MafUtil.java) - to - cBioPortal Database mapping.
|        MafUtil Field Names       |         MAF Column Name          |                    `MUTATION.<FIELD>`                    |                                 `MUTATION_EVENT.<FIELD>`                                 |                                                                                                                                                                                           NOTES                                                                                                                                                                                           |
| -------------------------------- | -------------------------------- | -------------------------------------------------------- | ---------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| HUGO_SYMBOL                      | Hugo_Symbol                      |                                                          |                                                                                          | Helps resolve CanonicalGene (default by Entrez Gene id)                                                                                                                                                                                                                                                                                                                                   |
| ENTREZ_GENE_ID                   | Entrez_Gene_Id                   | mutation.entrez_gene_id                                  | mutation_event.entrez_gene_id                                                            | References gene.entrez_gene_id                                                                                                                                                                                                                                                                                                                                                            |
| CENTER                           | Center                           | mutation.center                                          |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| NCBI_BUILD                       | NCBI_Build                       |                                                          | mutation_event.ncbi_build                                                                |                                                                                                                                                                                                                                                                                                                                                                                           |
| CHROMOSOME                       | Chromosome                       |                                                          | mutation_event.chr                                                                       |                                                                                                                                                                                                                                                                                                                                                                                           |
| START_POSITION                   | Start_Position                   |                                                          | mutation_event.start_position                                                            |                                                                                                                                                                                                                                                                                                                                                                                           |
| END_POSITION                     | End_Position                     |                                                          | mutation_event.end_position                                                              |                                                                                                                                                                                                                                                                                                                                                                                           |
| STRAND                           | Strand                           |                                                          | mutation_event.strand                                                                    |                                                                                                                                                                                                                                                                                                                                                                                           |
| VARIANT_CLASSIFICATION           | Variant_Classification           |                                                          | mutation_event.mutation_type **                                                          | see ExtendedMutationUtil.getMutationType()                                                                                                                                                                                                                                                                                                                                                |
| VARIANT_TYPE                     | Variant_Type                     |                                                          | mutation_event.variant_type                                                              |                                                                                                                                                                                                                                                                                                                                                                                           |
| REFERENCE_ALLELE                 | Reference_Allele                 |                                                          | mutation_event.reference_allele                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| TUMOR_SEQ_ALLELE_1               | Tumor_Seq_Allele1                | mutation.tumor_seq_allele1                               | mutation_event.tumor_seq_allele                                                          | see MafUtil.resolveTumorSeqAllele()                                                                                                                                                                                                                                                                                                                                                       |
| TUMOR_SEQ_ALLELE_2               | Tumor_Seq_Allele2                | mutation.tumor_seq_allele2                               | mutation_event.tumor_seq_allele                                                          | see MafUtil.resolveTumorSeqAllele()                                                                                                                                                                                                                                                                                                                                                       |
| DBSNP_RS                         | dbSNP_RS                         |                                                          | mutation_event.db_snp_rs                                                                 |                                                                                                                                                                                                                                                                                                                                                                                           |
| DBSNP_VAL_STATUS                 | dbSNP_Val_Status                 |                                                          | mutation_event.db_snp_val_status                                                         |                                                                                                                                                                                                                                                                                                                                                                                           |
| TUMOR_SAMPLE_BARCODE             | Tumor_Sample_Barcode             | mutation.sample_id                                       |                                                                                          | sample.stable_id / sample.internal_id                                                                                                                                                                                                                                                                                                                                                     |
| MATCHED_NORM_SAMPLE_BARCODE      | Matched_Norm_Sample_Barcode      | mutation.matched_norm_sample_barcode                     |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MATCH_NORM_SEQ_ALLELE1           | Match_Norm_Seq_Allele1           | mutation.match_norm_seq_allele1                          |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MATCH_NORM_SEQ_ALLELE2           | Match_Norm_Seq_Allele2           | mutation.match_norm_seq_allele2                          |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| TUMOR_VALIDATION_ALLELE1         | Tumor_Validation_Allele1         | mutation.tumor_validation_allele1                        |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| TUMOR_VALIDATION_ALLELE2         | Tumor_Validation_Allele2         | mutation.tumor_validation_allele2                        |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MATCH_NORM_VALIDATION_ALLELE1    | Match_Norm_Validation_Allele1    | mutation.match_norm_validation_allele1                   |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MATCH_NORM_VALIDATION_ALLELE2    | Match_Norm_Validation_Allele2    | mutation.match_norm_validation_allele2                   |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| VERIFICATION_STATUS              | Verification_Status              | mutation.verification_status                             |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| VALIDATION_STATUS                | Validation_Status                | mutation.validation_status                               |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| MUTATION_STATUS                  | Mutation_Status                  | mutation.mutation_status                                 |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| SEQUENCING_PHASE                 | Sequencing_Phase                 | mutation.sequencing_phase                                |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| SEQUENCE_SOURCE                  | Sequence_Source                  | mutation.sequence_source                                 |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| VALIDATION_METHOD                | Validation_Method                | mutation.validation_method                               |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| SCORE                            | Score                            | mutation.score                                           |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| BAM_FILE                         | BAM_File                         | mutation.bam_file                                        |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| SEQUENCER                        | Sequencer                        | mutation.sequencer                                       |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| AMINO_ACID_CHANGE                | Amino_Acid_Change                | mutation.amino_acid_change                               | mutation_event.protein_change **                                                         | see ExtendedMutationUtil.getProteinChange()                                                                                                                                                                                                                                                                                                                                               |
| T_REF_COUNT                      | t_ref_count                      | mutation.tumor_ref_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorRefCount()                                                                                                                                                                                                                                                                                                                                               |
| T_ALT_COUNT                      | t_alt_count                      | mutation.tumor_alt_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                                                               |
| N_REF_COUNT                      | n_ref_count                      | mutation.normal_ref_count **                             |                                                                                          | see ExtendedMutationUtil.getNormalRefCount()                                                                                                                                                                                                                                                                                                                                              |
| N_ALT_COUNT                      | n_alt_count                      | mutation.normal_alt_count **                             |                                                                                          | see ExtendedMutationutil.getNormalAltCount()                                                                                                                                                                                                                                                                                                                                              |
| I_T_REF_COUNT                    | i_t_ref_count                    | mutation.tumor_ref_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorRefCount()                                                                                                                                                                                                                                                                                                                                               |
| AD_REF                           | AD_Ref                           | mutation.tumor_ref_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorRefCount()                                                                                                                                                                                                                                                                                                                                               |
| I_T_ALT_COUNT                    | i_t_alt_count                    | mutation.tumor_alt_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                                                               |
| AD_ALT                           | AD_Alt                           | mutation.tumor_alt_count **                              |                                                                                          | see ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                                                               |
| NORM_AD_REF                      | Norm_AD_Ref                      | mutation.normal_ref_count **                             |                                                                                          | see ExtendedMutationUtil.getNormalRefCount()                                                                                                                                                                                                                                                                                                                                              |
| NORM_AD_ALT                      | Norm_AD_Alt                      | mutation.normal_alt_count **                             |                                                                                          | see ExtendedMutationutil.getNormalAltCount()                                                                                                                                                                                                                                                                                                                                              |
| T_TOT_COV                        | TTotCov                          | mutation.tumor_ref_count / mutation.tumor_alt_count **   |                                                                                          | see ExtendedMutationUtil.getTumorRefCount() / ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                     |
| T_VAR_COV                        | TVarCov                          | mutation.tumor_ref_count / mutation.tumor_alt_count **   |                                                                                          | see ExtendedMutationUtil.getTumorRefCount() / ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                     |
| N_TOT_COV                        | NTotCov                          | mutation.normal_ref_count / mutation.normal_alt_count ** |                                                                                          | see ExtendedMutationUtil.getNormalRefCount() / ExtendedMutationUtil.getNormalAltCount()                                                                                                                                                                                                                                                                                                   |
| N_VAR_COV                        | NVarCov                          | mutation.normal_ref_count / mutation.normal_alt_count ** |                                                                                          | see ExtendedMutationUtil.getNormalRefCount() / ExtendedMutationUtil.getNormalAltCount()                                                                                                                                                                                                                                                                                                   |
| TUMOR_DEPTH                      | tumor_depth                      | mutation.tumor_ref_count / mutation.tumor_alt_count **   |                                                                                          | see ExtendedMutationUtil.getTumorRefCount() / ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                     |
| TUMOR_VAF                        | tumor_vaf                        | mutation.tumor_ref_count / mutation.tumor_alt_count **   |                                                                                          | see ExtendedMutationUtil.getTumorRefCount() / ExtendedMutationUtil.getTumorAltCount()                                                                                                                                                                                                                                                                                                     |
| NORMAL_DEPTH                     | normal_depth                     | mutation.normal_ref_count / mutation.normal_alt_count ** |                                                                                          | see ExtendedMutationUtil.getNormalRefCount() / ExtendedMutationUtil.getNormalAltCount()                                                                                                                                                                                                                                                                                                   |
| NORMAL_VAF                       | normal_vaf                       | mutation.normal_ref_count / mutation.normal_alt_count ** |                                                                                          | see ExtendedMutationUtil.getNormalRefCount() / ExtendedMutationUtil.getNormalAltCount()                                                                                                                                                                                                                                                                                                   |
| PROTEIN_CHANGE                   | HGVSp_Short                      |                                                          | mutation_event.protein_change **                                                         | see ExtendedMutationUtil.getProteinChange()                                                                                                                                                                                                                                                                                                                                               |
| CODONS                           | Codons                           |                                                          | mutation_event.oncotator_codon_change **                                                 | see ImportExtendedMutation (codonChange = record.getCodons()) / MAF field "Codons" always overrides MAF value ONCOTATOR_CODON_CHANGE // "ONCOTATOR_CODON_CHANGE" never used?                                                                                                                                                                                                              |
| SWISSPROT                        | SWISSPROT                        |                                                          | mutation_event.oncotator_uniprot_accession **                                            | see ImportExtendedMutation (uniprotAccession = record.getSwissprot()) / if current accession type is SwissProt then record.getSwissprot() = uniprotAccession OR DaoUniProtIdMapping.mapFromUniprotIdToAccession(uniprotName) where uniprotName = record.getSwissprot() / OVERRIDES MAF field "ONCOTATOR_UNIPROT_ACCESSION"                                                                |
| REFSEQ                           | RefSeq                           |                                                          | mutation_event.oncotator_refseq_mrna_id **                                               | see ImportExtendedMutation(refseqMrnaId = record.getRefSeq()) / Always overrides MAF field "ONCOTATOR_REFSEQ_MRNA_ID"                                                                                                                                                                                                                                                                     |
| PROTEIN_POSITION                 | Protein_position                 |                                                          | mutation_event.oncotator_protein_pos_start / mutation_event.oncotator_protein_pos_end ** | see ExtendedMutationUtil.getProteinPosStart() / resolved from record.getProteinPosition() or record.getProteinChange() based on what is avialable / MAF field "ONCOTATOR_PROTEIN_POS_START" and "ONCOTATOR_PROTEIN_POS_END" never used?                                                                                                                                                   |
| ONCOTATOR_VARIANT_CLASSIFICATION | ONCOTATOR_VARIANT_CLASSIFICATION |                                                          | mutation_event.mutation_type **                                                          | see ExtendedMutationUtil.getMutationType()                                                                                                                                                                                                                                                                                                                                                |
| ONCOTATOR_REFSEQ_MRNA_ID         | ONCOTATOR_REFSEQ_MRNA_ID         |                                                          | mutation_event.oncotator_refseq_mrna_id ** [maps to DB but MAF value never imported]     | see ImportExtendedMutation(refseqMrnaId = record.getRefSeq()) / Always overrides MAF field "ONCOTATOR_REFSEQ_MRNA_ID" / MafRecord (record.getOncotatorRefseqMrnaId()) never used                                                                                                                                                                                                          |
| ONCOTATOR_UNIPROT_ENTRY_NAME     | ONCOTATOR_UNIPROT_ENTRY_NAME     |                                                          | mutation_event.oncotator_uniprot_entry_name ** [maps to DB but MAF value never imported] | see ImportExtendedMutation (record.getOncotatorUniprotName() never used to set mutation_event.oncotator_uniprot_entry_name) / MafRecord (record.getOncotatorUniprotName()) never used                                                                                                                                                                                                     |
| ONCOTATOR_UNIPROT_ACCESSION      | ONCOTATOR_UNIPROT_ACCESSION      |                                                          | mutation_event.oncotator_uniprot_accession ** [maps to DB but MAF value never imported]  | see ImportExtendedMutation (uniprotAccession = record.getSwissprot()) / if current accession type is SwissProt then record.getSwissprot() = uniprotAccession OR DaoUniProtIdMapping.mapFromUniprotIdToAccession(uniprotName) where uniprotName = record.getSwissprot() / OVERRIDES MAF field "ONCOTATOR_UNIPROT_ACCESSION" / MafRecord (record.getOncotatorUniprotAccession()) never used |
| ONCOTATOR_CODON_CHANGE           | ONCOTATOR_CODON_CHANGE           |                                                          | mutation_event.oncotator_codon_change ** [maps to DB but MAF value never imported]       | see ImportExtendedMutation (codonChange = record.getCodons()) / MAF field "Codons" always overrides MAF value ONCOTATOR_CODON_CHANGE // "ONCOTATOR_CODON_CHANGE" never used? / MafRecord (record.getOncotatorCodonChange()) never used                                                                                                                                                    |
| ONCOTATOR_PROTEIN_POS_START      | ONCOTATOR_PROTEIN_POS_START      |                                                          | mutation_event.oncotator_protein_pos_start ** [maps to DB but MAF value never imported]  | see ExtendedMutationUtil.getProteinPosStart() / resolved from record.getProteinPosition() or record.getProteinChange() based on what is avialable / MAF field "ONCOTATOR_PROTEIN_POS_START" never used?                                                                                                                                                                                   |
| ONCOTATOR_PROTEIN_POS_END        | ONCOTATOR_PROTEIN_POS_END        |                                                          | mutation_event.oncotator_protein_pos_end ** [maps to DB but MAF value never imported]    | see ExtendedMutationUtil.getProteinPosStart() / resolved from record.getProteinPosition() or record.getProteinChange() based on what is avialable / MAF field "ONCOTATOR_PROTEIN_POS_END" never used?                                                                                                                                                                                     |
| DRIVER_FILTER                    | cbp_driver                       | mutation.driver_filter                                   |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| DRIVER_FILTER_ANNOTATION         | cbp_driver_annotation            | mutation.driver_filter_annotation                        |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| DRIVER_TIERS_FILTER              | cbp_driver_tiers                 | mutation.driver_tiers_filter                             |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
| DRIVER_TIERS_FILTER_ANNOTATION   | cbp_driver_tiers_annotation      | mutation.driver_tiers_filter_annotation                  |                                                                                          |                                                                                                                                                                                                                                                                                                                                                                                           |
:warning:  ***Warning: this way of loading data is deprecated. See [Data loading](Data-Loading.md) for the recommended method.***

The page describes how to import data into the cBioPortal using Python scripts found in our [scripts](https://github.com/cBioPortal/cbioportal/tree/stable/core/src/main/scripts) directory.  The follow assumptions have been made:

1. The cBioPortal software has been correctly [built from source](/deployment/deploy-without-docker/Build-from-Source.md).
2. The data to import is in the proper [File Format](/File-Formats.md).
3. The `PORTAL_HOME` environment variable has been properly defined.

## Import Cancer Type

The following command is used to import cancer type metadata into the cBioPortal database:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-cancer-type --meta-filename /path-to-cancer-type-file/cancer_type.txt
```

The [cancer_type.txt](File-Formats.md#cancer-study) file should conform to the file format describe on the File Formats wiki page.

## Import Cancer Study

The following command is used to import cancer study metadata into the cBioPortal database:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-study --meta-filename /path-to-meta_study-file/meta_study.txt
```

The [meta_study.txt](File-Formats.md#cancer-study) file should conform to the file format describe on the File Formats wiki page.

## Remove Cancer Study

The following command is used to remove a cancer study from the cBioPortal database:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command remove-study --meta-filename /path-to-meta_study-file/meta_study.txt
```

The [meta_study.txt](File-Formats.md#cancer-study) file should conform to the file format describe on the File Formats wiki page.

## Import Study Data

The following command is used to import all types of genomic and clinicla data described on our [File Formats](File-Formats.md) wiki page.  Below are examples of import copy number and mutation data:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-study-data --meta-filename /path-to-meta_CNA.txt/meta_CNA.txt --data-filename /path-to-data_CNA.txt/data_CNA.txt
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-study-data --meta-filename /path-to-meta_mutations_extended.txt/meta_mutations_extended.txt --data-filename /path-to-data_mutations_extended.txt/data_mutations_extended.txt
```

## Import Case Lists

The following command is used to import case lists into the cBioPortal database:

```
$PORTAL_HOME/core/src/main/scripts/cbioportalImporter.py --jvm-args "-Dspring.profiles.active=dbcp -cp $PORTAL_HOME/core/target/core-1.0-SNAPSHOT.jar" --command import-case-list --meta-filename /path-to-case-list/case-list.txt
```

The [case-list.txt](File-Formats.md#case-lists) file should conform to the file format describe on the File Formats wiki page.

Note - the import-case-list command can take a path to a folder that contains a set of case-list files in addition to a single case-list file.
# Inter-Repository Dependencies

## cbioportal-core Repository

[cbioportal-core](https://github.com/cBioPortal/cbioportal-core) is dependent on
[cbioportal](https://github.com/cBioPortal/cbioportal).
This can be seen in its [pom.xml](https://github.com/cBioPortal/cbioportal-core/blob/main/pom.xml).
By removing the dependency and attempting to build cbioportal-core, errors are reported. This
document captures a summary of the reported errors. This analysis was done at the time of release
v6.0.0.

cbioportal-core now primarily provides import functionality, which is mainly located
in the packages:
- [org.mskcc.cbio.portal.dao](https://github.com/cBioPortal/cbioportal-core/tree/main/src/main/java/org/mskcc/cbio/portal/dao)
- [org.mskcc.cbio.portal.model](https://github.com/cBioPortal/cbioportal-core/tree/main/src/main/java/org/mskcc/cbio/portal/model)
- [org.mskcc.cbio.portal.scripts](https://github.com/cBioPortal/cbioportal-core/tree/main/src/main/java/org/mskcc/cbio/portal/scripts)

### Dependencies in cbioportal-core import functionality

Code in cbioportal which is referenced directly by code in cbioportal-core consists of three
enum class definitions.

- cbioportal class [ResourceType](https://github.com/cBioPortal/cbioportal/blob/master/src/main/java/org/cbioportal/model/ResourceType.java) is imported by cbioportal-core class [ResourceDefinition](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/model/ResourceDefinition.java). This provides enum constants {STUDY, SAMPLE, PATIENT}.
- cbioportal class [EntityType](https://github.com/cBioPortal/cbioportal/blob/master/src/main/java/org/cbioportal/model/EntityType.java) is imported by cbioportal-core class [ImportTabDelimData](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/scripts/ImportTabDelimData.java). This provides enum constants {GENE, GENESET, PHOSPHOPROTEIN, GENERIC_ASSAY}.
- cbioportal class [CNA](https://github.com/cBioPortal/cbioportal/blob/master/src/main/java/org/cbioportal/model/CNA.java) is imported by cbioportal-core class [CnaEvent](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/model/CnaEvent.java). This provides enum constants {AMP, GAIN, DIPLOID, HETLOSS, HOMDEL}.

The cbioportal-core import functionality code also relies on several packaged libraries in cbioportal.

cbioportal packaged library org.apache.commons.commons-collections4 (version 4.4) is used by these cbioportal-core classes:
- [DaoPatient](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/dao/DaoPatient.java)
- [ImportClinicalData](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/scripts/ImportClinicalData.java)

cbioportal packaged library org.slf4j.slf4j-api (latest version) is used by these cbioportal-core classes:
- [JdbcUtil](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/dao/JdbcUtil.java)
- [ProgressMonitor](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/ProgressMonitor.java)
- [Patient](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/model/Patient.java)
- [AccessControlImpl](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/internal/AccessControlImpl.java)

cbioportal packaged library org.springframework.security.spring-security.core (latest compatible version)
and other related such as org.springframework.security.spring-security.web are used in these cbioportal-core classes:
- [AccessControl](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/AccessControl.java)
- [DaoCancerStudy](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java) (transitive dependency through AccessControl)

### Other dependencies in cbioportal-core

There are other (non-import) cbioportal-core uses of libraries packaged in cbioportal.

cbioportal packaged libraries jakarta.servlet jakarta.servlet.http (ServletConfig, HttpServlet, ServletException, HttpServletRequest, HttpServletResponse} are used by these cbioportal-core classes:
- [GetSurvivalDataJSON](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/GetSurvivalDataJSON.java)
- [XDebug](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/XDebug.java)
- [CalcCoExp](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/CalcCoExp.java)
- [BioGeneServlet](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/BioGeneServlet.java)
- [IGVLinkingJSON](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/IGVLinkingJSON.java)

cbioportal packaged library org.json.simple / com.googlecode.json-simple.json-simple (version 1.1.1) is used by these cbioportal-core classes:
- [CalcCoExp](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/CalcCoExp.java)
- [GetSurvivalDataJSON](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/GetSurvivalDataJSON.java)
- [IGVLinkingJSON](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/IGVLinkingJSON.java)
- [GetClinicalData](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/web_api/GetClinicalData.java)

cbioportal packaged library org.apache.commonsorg.apache.commons.commons-math3 (version 3.6.1) is used by this cbioportal-core class:
- [CalcCoExp](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/servlet/CalcCoExp.java)

cbioportal packaged library org.slf4j.slf4j-api (latest version) is used by this cbioportal-core class:
- [AccessControlImpl](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/internal/AccessControlImpl.java) (transitive dependency through many servlets)

cbioportal packaged library org.springframework.security.spring-security.core (latest compatible version)
and other related such as org.springframework.security.spring-security.web are used in these cbioportal-core classes:
- [AccessControl](https://github.com/cBioPortal/cbioportal-core/blob/main/src/main/java/org/mskcc/cbio/portal/util/AccessControl.java) (transitive dependency through many servlets)

### Consequences
When any of these dependencies are updated in cbioportal (or in the case of libraries, updated in a non-backwards-compatible way), the cbioportal-core repository should be built and tested using the updated dependencies and proper behavior of the import functionality should be tested. cbioportal-core should also be brought up to date by packaging the current version of cbioportal into cbioportal-core (in pom.xml).
# Migration Guide

This page describes various changes deployers will need to make as they deploy newer versions of the portal.

## v5 -> v6

- Override Spring Application Properties:
  - The process of overriding properties has been updated for example previously users could do the following `java -Xms2g -Xmx4g -Dauthenticate=saml -jar webapp-runner.jar`.
  Now users must pass Spring Application properties after the `.jar` and replace `-Dauthenticate=false` with `--authenticate=false`. 
  For example `java -Xms2g -Xmx4g -jar cbioportal/target/cbioportal-exec.jar --spring.config.location=cbioportal/application.properties --authenticate=false`
- `portal.properties` migration needed:
  - `portal.properties` has been renamed to `application.properties`. This is the Spring Boot default name 
  - `authenticate` values of `googleplus`, `social_auth_google` and `social_auth_microsoft` have been replaced by `optional_oauth2`
    - If you used this property before without authorization (unlikely, only the public cBioPortal instance uses this), add the property `always_show_study_group=PUBLIC` and confirm  that all studies in your database you'd like to be be public have `GROUPS` values set to `PUBLIC`
- `Redis HTTP Session`
  - To disable redis session (Must be disabled if redis is not setup) `spring.autoconfigure.exclude=org.springframework.boot.autoconfigure.data.redis.RedisAutoConfiguration`
  - To enable `spring.data.redis.host=localhost` `spring.data.redis.port=6379`
- `Security` changes:
  - Properties used to configure Security have changed see [security.properties.EXAMPLE](deployment/customization/security.properties-Reference.md)
  - `SAML` changes:
    - need to generate new keys (`.jks` no longer works)
       - if you have an existing client, update saml keys with newly generate keys
    - if you have an existing client, update redirect_uri to be without port 443
    - You can now use a URL here, instead of metadata XML: `spring.security.saml2.relyingparty.registration.cbio-saml-idp.assertingparty.metadata-uri`

## v5.3 -> v5.4

- Remove `db.host` and `db.portal_db_name` and `db.use_ssl` properties from the _portal.properties_ file or JVM
  parameters. Update property `db.connection_string` to encode the hostname, port, database and other parameters
  according to [Database Settings](deployment/customization/portal.properties-Reference.md#Database-Settings) documentation and pass via
  _portal.properties_ file or as JVM parameter.

## v4 -> v5

- All fusion profiles are now required to be migrated to structural variant format. One can use
  this [migration tool](https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/fusion-to-sv-converter)
  to migrate the fusion files.
- All fusion files on [datahub](https://github.com/cBioPortal/datahub) were migrated to the structural variant format
  and their molecular profile ids were renamed from `{study_id}_fusion` to `{study_id}_structural_variants`. If you are
  using these datahub files one would need to re-import them.
- Study view user setting will be outdated after migration, please follow `Clear Study View User settings` section
  in [Session Service Management](Session-Service-Management.md#Clear-Study-View-User-settings)
- The new default set of transcripts for each gene has changed from `uniprot` to `mskcc`. See
  the [Mutation Data Annotation Section](./mutation-data-transcript-annotation.md) for more details. To keep the old set
  of default transcripts add the property `genomenexus.isoform_override_source=uniprot`
  to [the properties file](https://docs.cbioportal.org/deployment/customization/portal.properties-reference/#properties).

See the [v5.0.0 release notes](https://github.com/cBioPortal/cbioportal/releases/tag/v5.0.0) for more details.

## v3 -> v4

- Introduces `logback` package for logging. If you don't have any custom log4j.properties file, no changes are necessary
- Cleans up several old databases ([PR](https://github.com/cBioPortal/cbioportal/pull/9360)). In theory the migration
  should be seamless, since the docker container detects an old database version and migrates it automatically.

See the [v4.0.0 release notes](https://github.com/cBioPortal/cbioportal/releases/tag/v4.0.0) for more details.

## v2 -> v3

- Session service is now required to be set up. You can't run it without session service. The recommended way to run
  cBioPortal is to use the Docker Compose instructions.

## v1 -> v2

- Changes cBioPortal to a Single Page App (SPA) written in React, Mobx and bootstrap that uses a REST API. It shouldn't
  change anything for a deployer.
# MSK Maintenance (In Progress)
We provide our cBioPortal's maintenance page publicly in the hope that it may be useful to others

## Database Migration
In the database migration process, we are going to have two main steps: building importers and updating database scheme.

### Building Importers
- Codebases:
    - [cbioportal](https://github.com/cBioPortal/cbioportal)
    - [genome nexus annotation pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline)
    - [pipelines](https://github.com/knowledgesystems/pipelines)
- Building single importer:
    - Take the cbioportal hash and add it to genome nexus annotation pipeline POM -> push to fork
    - Take the cbioportal hash and new genome nexus annotation pipeline hash (from above) and propagate to pipelines POM; also update db version if db migration is part of change
    - Login to pipelines server
    - Sets general env variables needed for building
        - Run `sh /data/portal-cron/scripts/automation-environment.sh`
    - Specific to cbioportal build, looks for properties and xml files off of $PORTAL_HOME
        - Run `export PORTAL_HOME=/data/portal-cron/git-repos/cbioportal`
    - Make sure the local cbioportal codebase is on the correct git hash at `/data/portal-cron/git-repos/cbioportal`
    - Specify importer to be build
        - Copy importer properties to `/data/portal-cron/git-repos/cbioportal/src/main/resources` (e.g. triage `/data/portal-cron/git-repos/pipelines-configuration/properties/import-triage/*`)
    - Navigate to pipelines folder
        - `cd /data/portal-cron/git-repos/pipelines`
    - Build importer
        - mvn clean install -DskipTests
- Build multiple importers
    - Take the cbioportal hash and add it to genome nexus annotation pipeline POM -> push to fork
    - Take the cbioportal hash and new genome nexus annotation pipeline hash (from above) and propagate to pipelines POM; also update db version if db migration is part of change
    - Login to pipelines server
    - Sets general env variables needed for building
        - Run `sh /data/portal-cron/scripts/automation-environment.sh`
    - Specific to cbioportal build, looks for properties and xml files off of $PORTAL_HOME
        - Run `export PORTAL_HOME=/data/portal-cron/git-repos/cbioportal`
    - Make sure the local cbioportal codebase is on the correct git hash at `/data/portal-cron/git-repos/cbioportal`
    - Go to build importer jars folder
        - `cd /data/portal-cron/git-repo/pipelines-configuration/build-importer-jars`
    - (Optional) Remove existing jars
        - `rm *.jar`
    - Build all importers at once (build all importers except cmo-pipelines)
        - `sh buildproductionjars.sh -sd=true -sgp=true -b=importers`
        - Available parameters:
          - [--cbioportal-git-hash|-cgh=<cbioportal_commit_hash>]
          - [--skip-deployment|-sd=<true|false>]
          - [--skip-git-pull|-sgp=<true|false>]
          - [--build|-b=<build_specifier>], build_specifier should be one of the following
            - all (build for all artifacts)
            - importers (all importers except cmo-pipelines)
            - cmo-pipelines (cmo-pipelines artifacts only)
            - triage-cmo-importer
            - msk-dmp-importer
            - msk-cmo-importer
            - public-importer
            - genie-aws-importer
            - genie-archive-importer
            - hgnc-importer

### Updating Database Scheme
Database needs to be updated one by one, we have four main databases: triage, private, genie, and public. Take triage database as an example.
- Migrate one database (e.g. triage)
    - SSH into pipeline server
    - Checkout to the commit that contains the latest database scheme
    - Check if property sets up correctly to the right database (triage)
        - `vi /data/portal-cron/git-repos/cbioportal/src/main/resources/application.properties`
    - Move to directory
        - `cd /data/portal-cron/git-repos/cbioportal`
    - Run database migration using script:
        - `python3 core/src/main/scripts/migrate_db.py --properties-file src/main/resources/application.properties --sql db-scripts/src/main/resources/migration.sql`
    - Monitor the DB migration process and look for possible errors
    - Access database and verify the DB scheme is updated
This document describes how each mutation in cBioPortal gets annotated with a specific gene symbol + protein change. 

### Biological Background 

This section explains the concepts of protein isoforms and transcripts. 

#### What is an isoform? 

From a single gene (string of nucleotides) multiple protein sequences can be formed (string of amino acids). For 
example: parts of the gene that code for proteins (exons) can be included or excluded through a process known as [alternative splicing](https://en.wikipedia.org/wiki/Alternative_splicing). Each of the different resulting proteins is called an [isoform](https://en.wikipedia.org/wiki/Protein_isoform). A single mutation can impact the isoforms differently. E.g. in one isoform
it might change a P to a T, but in the other isoform that particular exon does not get included and it is therefore not 
changing the amino acid sequence at all. In cBioPortal for convenience sake we assign a single gene symbol + protein change to each mutation. For most cases this works well because there is only one protein isoform relevant in a clinical setting. There are of course exceptions and we are 
therefore working on improving this representation. An explanation of the relation between transcripts and protein 
isoforms can be found in the next section. 

#### What is a transcript? 

DNA is transcribed to a pre-mRNA transcript which includes intron and exon regions. Splicing and other processes then take place to form the resulting mature mRNA transcript that only contains exons, which subsequently can be translated to a protein sequence. An mRNA transcript can thus be associated with a specific protein isoform. The Ensembl database assigns ids for these transcript with names like `ENSTxxx`. You can see this on e.g. the Ensembl 
website for the [BRAF gene](https://grch37.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000157764;r=7:140719327-140924929#): 

<img width="995" alt="Screen Shot 2022-10-05 at 9 35 55 AM" src="https://user-images.githubusercontent.com/1334004/194073821-9a43cab2-3d31-40ab-b47e-517f1ce8bca3.png">

The transcript `ENST00000288602.6` is 2480 base pairs long (nucleotides ACGT) and the associated protein isoform is 766 
amino acids (V/P/etc). You can see we are showing that same transcript and protein isoform on [cBioPortal](https://bit.ly/3vaclXN): 

<img width="1052" alt="Screen Shot 2022-10-05 at 9 37 32 AM" src="https://user-images.githubusercontent.com/1334004/194074248-c3e205b4-c2e4-4e19-a541-17517dc07913.png"> 

For each gene name in cBioPortal a canonical/default transcript is assigned. These assignments are stored in [Genome 
Nexus](https://www.genomenexus.org/) and explained below. Although cBioPortal does not store changes to different 
transcripts/isoforms for each mutation in the database itself, it does allow viewing them on the [Mutations Tab](https://bit.ly/39hVtDd) by re-annotating the mutations on the fly through [Genome Nexus](https://www.genomenexus.org/) whenever a user clicks on the transcript dropdown. 

### Transcript Assignment 

The cBioPortal database stores one gene + protein change annotation for each mutation event in the database. To allow 
comparing mutation data across studies it is important to annotate the mutation data (be it in MAF or VCF format) in 
the same way, otherwise the gene + protein changes can mean entirely different things. For all public studies stored 
in [datahub](https://github.com/cBioPortal/datahub/tree/master/public) we leverage [Genome Nexus](https://www.genomenexus.org) to do so. 
Genome Nexus assigns one canonical Ensembl Transcript + gene name + protein change for 
each mutation. You can find the mapping of hugo symbol to transcript id [here](https://github.com/genome-nexus/genome-nexus-importer/blob/master/data/grch37_ensembl92/export/ensembl_biomart_canonical_transcripts_per_hgnc.txt). There are 
two sets of default transcripts: `uniprot` and `mskcc`. We recommend to use the `mskcc` set of transcripts when 
starting from scratch, since these are more up to date and correspond to transcripts that were chosen as relevant for 
clinical sequencing at MSKCC. The `uniprot` set of transcripts was constructed several years ago, but we are no longer 
certain about the logic on how to reconstruct them hence they are not being kept up to date. One can see the 
differences between the two in [this file](https://github.com/cBioPortal/cbioportal-frontend/files/9498680/genes_with_different_uniprot_mskcc_isoforms.txt). For the public cBioPortal (https:
//www.cbioportal.org) and [datahub](https://github.com/cBioPortal/datahub/tree/master/public) we are using `mskcc`, for the GENIE cBioPortal (https://genie.cbioportal.org) we still use `uniprot`. As of cBioPortal v5 the default is `mskcc` for local installations. Prior to v5 it was `uniprot`. We recommend that people upgrading to v5 consider migrating to `mskcc` as well (see [migration guide](https://docs.cbioportal.org/migration-guide/) and [the properties reference docs](https://docs.cbioportal.org/deployment/customization/application.properties-reference/#properties)).

#### How default transcript assignment affects the Mutations Tab 

The [Mutations Tab](https://bit.ly/39hVtDd) shows the full protein sequence. The one shown by default is the canonical 
transcript (`mskcc` or `uniprot` depending on configuration). The mutations are drawn on the lollipop based on the 
protein position found in the cBioPortal database. For the [public cBioPortal](https://cbioportal.org) all mutation 
data in MAF format are annotated using [Genome Nexus](https://www.genomenexus.org) to add the gene and protein change 
columns. This is then imported into the cBioPortal database. Whether you choose to use the set of `uniprot` or `mskcc` 
transcripts, make sure to indicate it in the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-
nexus-annotation-pipeline#maf-annotation)(`--isoform-override <mskcc or uniprot>`) when annotating as well as in [the properties file](https://docs.cbioportal.org/deployment/customization/application.properties-reference/#properties)
of cBioPortal. That way the [Mutations Tab](https://bit.ly/39hVtDd) will show the correct canonical 
transcript. Note that whenever somebody uses the dropdown on the Mutations Tab to change the displayed transcript, 
Genome Neuxs re-annotates all mutations on the fly. The browser sends over the genomic location (chrom,start,end,ref,
alt) to get the protein change information for each transcript. Since many of the annotations are for the canonical transcripts
only we are currently hiding annotations for non-canonical transcripts.

#### Plans for default transcripts 

We are planning to move to a single set of default transcripts over time. Prior to v5 `uniprot` was used for the public 
facing portals and local installations. Our plan is to use `mskcc` everywhere and eventually we will most likely move to [MANE](https://www.ensembl.org/info/genome/genebuild/mane.html). MANE is only 
available for grch38 and since most of our data is for grch37 this is currently not feasible. Whichever set of 
transcripts you choose to use, make sure to indicate so in the [Genome Nexus Annotation Pipeline](https://github.com/genome-nexus/genome-nexus-annotation-pipeline#maf-annotation) (`--isoform-override <mskcc or uniprot>`) and put the same 
set of transcripts in [the properties file](https://docs.cbioportal.org/deployment/customization/application.properties-reference/#properties) of cBioPortal, such that the [Mutations Tab](https://bit.ly/39hVtDd) will show the correct canonical transcript (currently defaults to `mskcc`). The re-annotation of mutations only happens once a user clicks to change the transcript, which is why it's important that the protein change in the database is for the specific transcript displayed first. 
# Genie News
## April 3, 2024
*   **Public Release 15.1 of AACR Project GENIE:**
    *  The public release 15.1 version of AACR GENIE has 61 samples retracted that were present in AACR GENIE 15.0-public. 
    *  More detailed information can be found in the [AACR GENIE release notes](https://www.synapse.org/#!Synapse:syn55234828) and the data releases page from [Sage Bionetworks](https://www.synapse.org/#!Synapse:syn55234548).
## January 9, 2024
*   **Public Release 15.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 15.0-public now contains over 198,000 sequenced samples from over 172,000 patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.synapse.org/#!Synapse:syn7222066/wiki/409237).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](https://www.synapse.org/#!Synapse:syn50898929).
## November 9, 2023
*   **Public Release 14.1 of AACR Project GENIE:**
    *  The public release 14.1 version of AACR GENIE has 10 samples retracted that were present in AACR GENIE 14.0-public. 
    *  More detailed information can be found in the [AACR GENIE release notes](https://www.synapse.org/#!Synapse:syn52919317) and the data releases page from [Sage Bionetworks](https://www.synapse.org/#!Synapse:syn7222066/wiki/410924)
## September 20, 2023
*   **Public Release 14.0 of AACR Project GENIE:**
     *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 14.0-public now contains 183,000 sequenced samples from nearly 160,000 patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets released to date.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.synapse.org/#!Synapse:syn7222066/wiki/409237).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](https://www.synapse.org/#!Synapse:syn50898929).
## May 1, 2023
*   **Public Release 13.1 of AACR Project GENIE:**
    *  The public release 13.1 version of AACR GENIE has 65 samples retracted that were present in AACR GENIE 13.0-public. 
    *  More detailed information can be found in the [AACR GENIE release notes](https://www.synapse.org/#!Synapse:syn51355999) and the data releases page from [Sage Bionetworks](https://www.synapse.org/#!Synapse:syn7222066/wiki/410924)
## January 9, 2023
*   **Public Release 13.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 13.0-public now contains more than 167,000 sequenced samples from over 148,000 patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets released to date.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.synapse.org/#!Synapse:syn21683345).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](https://www.synapse.org/#!Synapse:syn50898929).
## November 3, 2022
*   **Public Release BPC CRC 2.0-PUBLIC**
    *   The GENIE BPC CRC v2.0-public dataset contains 1,485 CRC patients from three institutions: MSKCC, DFCI, and VICC.‚ÄØ
    *   The complete, post-processed data are available on [Synapse](https://www.synapse.org/#!Synapse:syn30102262)
## July 6, 2022
*   **Public Release GENIE ERBB2 Cohort**
    * The study contains 315 samples from 135 patients from 6 institues.
## July 22, 2022
*   **Public Release 12.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 12.0-public, was released in July 2022.The registry now contains more than 154,000 sequenced samples from 137,000+ patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets released to date.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.aacr.org/wp-content/uploads/2022/07/GENIE_12.0-public_data_guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
## May 19, 2022
*   **Public Release BPC NSCLC 2.0-PUBLIC**
    *   The GENIE BPC NSCLC v2.0-public dataset contains 1,846 NSCLC patients from 4 institutions: MSKCC, DFCI, VICC and UHN.
    *   The complete, post-processed data are available on [Synapse](https://www.synapse.org/#!Synapse:syn27056172/wiki/616601)
## January 7, 2022
*   **Public Release 11.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The current release, GENIE 11.0-public, was released in January 2022. The registry now contains over 136,000 sequenced samples from over 121,000 patients, making the AACR Project GENIE registry among the largest fully public cancer genomic data sets released to date.
    *   More detailed information can be found in the [AACR GENIE Data Guide](https://www.aacr.org/wp-content/uploads/2022/02/GENIE_data_guide_11.0-public-1.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
## June 22, 2021
*   **Public Release 10.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The tenth data set, GENIE 10.0-public, was released in June 2021. With the most recent data release, the registry now contains genomic information from 120953 samples.  
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
## February 8, 2021
*   **Public Release 9.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The ninth data set, GENIE 9.0-public, was released in February 2021. With the most recent data release, the registry now contains genomic information from more nearly 17,000 non-small cell lung carcinomas, and nearly 12,000 breast and more than 11,000 colorectal cancers.  
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
## July 7, 2020
*   **Public Release 8.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The eighth data set, GENIE 8.0-public, was released in July 2020. The combined data set now includes nearly 96,000 de-identified genomic records collected from patients who were treated at each of the consortium's 19 participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for over 80 major cancer types, including data from greater than 14,000 patients with lung cancer, nearly 12,000 patients with breast cancer, and nearly 9,500 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).
## January 29, 2020
*   **Public Release 7.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The seventh data set, GENIE 7.0-public, was released in January 2020. The combined data set now includes nearly 79720 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for over 80 major cancer types.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).
## July 13, 2019
*   **Public Release 6.1 of AACR Project GENIE:**
    *   The sixth data set, GENIE 6.0-public, was released in early July 2019.  A patch to GENIE 6.0-public, GENIE 6.1-pubic, was subsequently released on July 13, 2019.  The combined data set now includes nearly 70,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for nearly 80 major cancer types, including data from nearly 11,000 patients with lung cancer, greater than  9,700 patients with breast cancer, and nearly 7,000 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](https://www.aacr.org/wp-content/uploads/2020/02/20200127_GENIE_Data_Guide_7.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](https://www.aacr.org/professionals/research/aacr-project-genie/).
## July 08, 2019
*   **Public Release 6.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The sixth data set, GENIE 6.0-public, was released in July 2019 adding more than 11,000 records to the database. The combined data set now includes nearly 71,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for over 80 major cancer types, including data from greater than 11,000 patients with lung cancer, nearly 9,800 patients with breast cancer, and more than 7,000 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## January 11, 2019
*   **Public Release 5.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The forth data set, GENIE 4.0-public, was released in July 2018 adding more than 7,800 records to the database. The combined data set now includes more than 59,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date.  This data will be released to the public every six months.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## July 16, 2018
*   **Public Release 4.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The forth data set, GENIE 4.0-public, was released in July 2018 adding more than 7,800 records to the database. The combined data set now includes more than 47,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date.  This data will be released to the public every six months. The public release of the fifth data set, GENIE 5.0-public, will take place in January, 2019.
    *   The combined data set now includes data for over 80 major cancer types, including data from greater than 7,500 patients with lung cancer, nearly 5,500 patients with breast cancer, and more than 5,100 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## January 2018
*   **Added 7500 samples to the GENIE Public Cohort** The combined dataset now includes samples from over 60 major cancer types including:
    *   6,000 lung cancer samples.
    *   4,500 breast cancer samples.
    *   4,300 colorectal cancer samples.
*   More detailed information can be found in the [AACR GENIE Data Guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## November 20, 2017
*   **Public Release 2.0 of AACR Project GENIE:**
    *   The first set of cancer genomic data aggregated through AACR Project Genomics Evidence Neoplasia Information Exchange (GENIE) was released to the global community in January 2017. The second dataset was released in November 2017, adding more than 13,000 records to the database.  The combined data set now includes over 32,000 de-identified genomic records collected from patients who were treated at each of the consortium‚Äôs participating institutions, making it among the largest fully public cancer genomic data sets released to date. These data will be continuously updated on a quarterly basis.
    *   The combined data set now includes data for 59 major cancer types, including data on nearly 5,000 patients with lung cancer, nearly 4,000 patients with breast cancer, and more than 3,500 patients with colorectal cancer. For more details about the data, and how to use it, consult the [data guide](http://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).
    *   In addition to accessing the data on the AACR Project GENIE cBioPortal website, users can download the data directly from [Sage Bionetworks](http://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf).
    *   For frequently asked questions, visit the [AACR FAQ page](http://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

## January 5, 2017
*   **Initial public release of AACR Project GENIE:**
    *   Somatic alterations in 18,980 tumor samples from 18,500 patients sequenced at eight different institutions.
    *   Data is available for download from [Sage Bionetworks](http://synapse.org/genie).

## AACR Project GENIE cBioPortal Terms of Use
*   I will not attempt to identify or contact individual participants from whom these data were collected by any means.
*   I will not redistribute the data without express written permission from the AACR Project GENIE Coordinating Center (info at aacrgenie dot org).
*   When publishing or presenting work using or referencing the AACR Project GENIE dataset please include the following attributions:
    *  Please cite: AACR Project GENIE Consortium. AACR Project GENIE: Powering Precision Medicine Through an International Consortium, Cancer Discov. 2017 Aug;7(8):818-831 and include the version of the dataset used.
    *  Include the following acknowledgement: The authors would like to acknowledge the American Association for Cancer Research and its financial and material support in the development of the AACR Project GENIE registry, as well as members of the consortium for their commitment to data sharing. Interpretations are the responsibility of study authors.
*   Should you decide at any point in the future to stop using the AACR Project GENIE cBioPortal, send request: genie-cbioportal-access at cbio dot mskcc dot org and we will remove your user id and any other information provided during new user registration from our systems.
*   For more information see [terms of access](http://www.aacr.org/Documents/Terms%20of%20Access.pdf) on the AACR website.
## May 29, 2024
*   **Added data** consisting of 9,632 samples from 10 studies:
	* [Soft Tissue and Bone Sarcoma (MSK, Nat Commun 2022)](https://www.cbioportal.org/study/summary?id=sarcoma_msk_2022) *7494 samples*
    * [Chronic Lymphocytic Leukemia (Broad, Nature Genetics 2022)](https://www.cbioportal.org/study/summary?id=cll_broad_2022) *1154 samples*
    * [Mature B-Cell Neoplasms (Simon Fraser University, Blood 2023)](https://www.cbioportal.org/study/summary?id=mbn_sfu_2023) *297 samples*
    * [Anaplastic Thyroid Cancers (GATCI, Cell Reports 2024)](https://www.cbioportal.org/study/summary?id=thyroid_gatci_2024) *190 samples*
    * [Colorectal Cancer (CAS Shanghai, Cancer Cell 2020)](https://www.cbioportal.org/study/summary?id=coadread_cass_2020) *146 samples*
    * [Prostate Cancer MDA PCa PDX (MD Anderson, Clin Cancer Res 2024)](https://www.cbioportal.org/study/summary?id=prad_msk_mdanderson_2023) *88 samples*
    * [Ovarian Cancer - MSK SPECTRUM (MSK, Nature 2022)](https://www.cbioportal.org/study/summary?id=msk_spectrum_tme_2022) *82 samples*
    * [IDH-mutated Diffuse Glioma (MSK, Clin Cancer Res 2024)](https://www.cbioportal.org/study/summary?id=difg_msk_2023) *73 samples*
    * [Pre-cancer Colorectal Polyps (HTAN Vanderbilt, Cell 2021)](https://www.cbioportal.org/study/summary?id=crc_hta11_htan_2021) *61 samples*
    * [Prostate Cancer (MSK, Science 2022)](https://www.cbioportal.org/study/summary?id=prad_organoids_msk_2022) *47 samples*

*   **Data Improvement**

    * Pan-can studies Ancestry addition: Genetic Ancestry data is added to all 32 TCGA Pan-Can studies,in generic assay format. Data source: [GDC](https://gdc.cancer.gov/about-data/publications/CCG-AIM-2020). [Example: Ancestry-associated somatic genetic alterations](https://www.cbioportal.org/comparison/alterations?comparisonId=664275e383e9543d61907074&patientEnrichments=true&selectedEnrichmentEventTypes=%5B%22missense%22%2C%22missense_mutation%22%2C%22missense_variant%22%2C%22indel%22%2C%22nonframeshift%22%2C%22inframe%22%2C%22inframe_del%22%2C%22inframe_deletion%22%2C%22in_frame_del%22%2C%22in_frame_deletion%22%2C%22nonframeshift_deletion%22%2C%22inframe_ins%22%2C%22inframe_insertion%22%2C%22in_frame_ins%22%2C%22in_frame_insertion%22%2C%22nonframeshift_insertion%22%2C%22truncating%22%2C%22feature_truncation%22%2C%22nonsense_mutation%22%2C%22nonsense%22%2C%22stopgain_snv%22%2C%22stop_gained%22%2C%22frameshift%22%2C%22frameshift_variant%22%2C%22frame_shift_del%22%2C%22frameshift_deletion%22%2C%22frame_shift_ins%22%2C%22frameshift_insertion%22%2C%22translation_start_site%22%2C%22initiator_codon_variant%22%2C%22start_codon_snp%22%2C%22start_codon_del%22%2C%22de_novo_start_outofframe%22%2C%22nonstop_mutation%22%2C%22stop_lost%22%2C%22splice_site%22%2C%22splice%22%2C%22splicing%22%2C%22splice_site_snp%22%2C%22splice_site_del%22%2C%22splice_site_indel%22%2C%22splice_region_variant%22%2C%22splice_region%22%2C%22silent%22%2C%22synonymous_variant%22%2C%22targeted_region%22%2C%22other%22%2C%22AMP%22%2C%22HOMDEL%22%2C%22structural_variant%22%5D)
 <img width="1006" alt="Screen Shot 2024-05-20 at 10 52 11 AM" src="https://github.com/cBioPortal/datahub/assets/61431648/628b7eed-9d72-4605-b114-2e438dbd442a"> 

## May 7, 2024
* **New Feature**: Gene-specific charts in Study View now support mutation data. Two chart types are available: sample mutational status and mutation types.

  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/d94c596a-1103-4aba-995a-ed1709236eee)
  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/5fd2fdd0-07a5-434a-ad42-3f72b7c67bc5)
  
* **New Feature**: Categorical data loaded into the generic assay format is now available in group comparison. [Example: Arm-level CNA compared between astrocytoma vs oligodendroglioma in Brain Lower Grade Glioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/comparison/generic_assay_armlevel_cna?comparisonId=616ff06a2e3200072db98b99)


  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/9f8c99a2-eef8-4b5a-ae33-4a2b295e0514)

* **Enhancement**: OncoPrint now allows you to remove a gene track.

  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/23cd422b-6a4e-4eff-9c04-ac88220591d2)

## Apr 2, 2024
* **Introducing the cBioPortal Newsletter!** Stay updated with the latest developments, insights, and community highlights of cBioPortal. Subscribe via [LinkedIn](https://www.linkedin.com/pulse/april-update-hello-cbioportal-community-cbioportal-22vle/) or [google groups](https://groups.google.com/g/cbioportal-news). We'll be sharing valuable updates every few months.

## March 27, 2024
* **New Feature**: The Plots tab is now available in the Study View. After exploring a cohort and applying any filters of interest in the Study View Summary page, you can now click over to the new Plots tab to explore the cohort by plotting any two attributes against each other. [Example: Mutation count vs subtype in Uterine Corpus Endometrial Carcinoma (TCGA, Nature 2013)](https://www.cbioportal.org/study/plots?id=ucec_tcga_pub&plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22SUBTYPE%22%7D&plots_vert_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22MUTATION_COUNT%22%2C%22logScale%22%3A%22false%22%7D&plots_coloring_selection=%7B%7D)

  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/cf74a4ff-57c6-4d71-86cb-4f7440fc3db6)

* **Enhancement**: The custom selection feature in Study View, which filters to a user-defined list of samples/patients, no longer requires the inclusion of the study ID with each sample ID. If viewing multiple studies in study view and filtering to a sample ID which exists in more than one study, all samples will be displayed by default or the study ID can be included for additional specificity.

  ![image](https://github.com/cBioPortal/cbioportal/assets/10508276/2143593d-5fd3-4b1d-b8a4-73a68e027e27)
  
## Mar 5, 2024
*   **Local Installations Feature**: Make representation of custom driver annotation configurable:
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/f67fd82f-1789-4427-9ca1-f17fc70c1a09)

## Feb 7, 2024
* **New feature**: The _Datasets Page_ now lists what studies have samples profiled for Structural Variants:
    
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/0e4e73e4-d9d8-417d-912c-710163480905)

* **New Major Release**: **v6.0.0** includes a major repackaging of the backend software to improve the development experience. The backend has been upgraded to use Java Spring Boot v3.1.4. The CORE and MAF modules have been moved to new repositories. All other modules have been compacted into a single source repository. We have also updated the JVM to 21 and many libraries have been updated to address security and performance issues. See more information in the [v6.0.0 release notes](https://github.com/cBioPortal/cbioportal/releases/tag/v6.0.0).

* **Local Installations Feature**: When mutational signature data is loaded, show COSMIC reference signatures on the _Patient View_:
    
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/10f73baa-d28c-4bb7-ad62-4751bd7f61cd)

## Dec 29, 2023
*   **Added data** consisting of 5,120 samples from 13 studies:
	* [Endometrial Cancer (MSK, Cancer Discovery 2023)](https://www.cbioportal.org/study/summary?id=ucec_ancestry_cds_msk_2023) *1882 samples*
    * [Esophagogastric Cancer (MSK, J Natl Cancer Inst 2023)](https://www.cbioportal.org/study/summary?id=egc_msk_2023) *902 samples*
    * [Diffuse Glioma (GLASS Consortium, 2022)](https://www.cbioportal.org/study/summary?id=difg_glass) *693 samples*
    * [Bladder Cancer (MSK, Clin Cancer Res 2023)](https://www.cbioportal.org/study/summary?id=bladder_msk_2023) *526 samples*
    * [Non-Small Cell Lung Cancer Brain Metastasis (MSK, Nat Commun 2023)](https://www.cbioportal.org/study/summary?id=bm_nsclc_mskcc_2023) *322 samples*
    * [MSK Make-an-IMPACT Rare Cancers (MSK, Clin Cancer Res 2023)](https://www.cbioportal.org/study/summary?id=makeanimpact_ccr_2023) *184 samples*
    * [Cervical Cancer (MSK, Clin Cancer Res 2023)](https://www.cbioportal.org/study/summary?id=cervix_msk_2023) *177 samples*
    * [Mature T and NK Neoplasms (MSK, Blood Adv 2023)](https://www.cbioportal.org/study/summary?id=mtnn_msk_2022) *132 samples*
    * [Hepatocellular Carcinoma (MSK, JCO Precis Oncol 2023)](https://www.cbioportal.org/study/summary?id=hcc_jcopo_msk_2023) *90 samples*
    * [Sarcoma (MSK, J Pathol 2023)](https://www.cbioportal.org/study/summary?id=sarcoma_msk_2023) *82 samples*
    * [Esophagogastric Cancer (MSK, Clin Cancer Res 2023)](https://www.cbioportal.org/study/summary?id=egc_trap_ccr_msk_2023) *64 samples*
    * [Rhabdomyosarcomas (MSK, NPJ Precis Oncol 2023)](https://www.cbioportal.org/study/summary?id=soft_tissue_msk_2023) *42 samples*
    * [Pediatric Rhabdomyosarcomas (MSK, JCO Precis Oncol 2023)](https://www.cbioportal.org/study/summary?id=rms_msk_2023) *24 samples*
    
*   **Gene Tables Update:** Updated tables of genes (main and alias), based on [Oct 1, 2023 HGNC release](http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/). See seedDB release notes [here](https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database-schema-2131) for details.

## Dec 5, 2023
* **New feature**: Filter by data types on the homepage:
    
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/cf6d5486-bfd7-4f69-add0-442dbb4d354a)
* **New feature**: Change colors of tracks in Oncoprint:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/301b0efb-3063-44db-bc4c-86973e7c566a" width="700" />

* **New feature**: Show alteration frequencies per group in OncoPrint. [Example: RTK-RAS alterations in Smokers vs Never Smokers Lung Cancer Cases in MSK-IMPACT Clinical Sequencing Cohort](https://bit.ly/3PwKsDJ)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/f2a9576e-c06d-447a-8eac-f0292bd1f935" width="700" />

## Oct 17, 2023
* **New feature**: Survival charts with landmark events and hazard ratios. [Example: TP53 in Lung Cancer Cases in MSK-IMPACT Clinical Sequencing Cohort](https://bit.ly/48QktxS)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/dc97c3ae-ebd7-4cd3-bb76-165517c9d2ba" width="700" />

## Oct 3, 2023
* **New feature**: New Structural Variants Tab on _Results View_. [Example: TMPRSS2 Structural Variants in MSK-IMPACT Clinical Sequencing Cohort](https://bit.ly/49FDhRk)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/141e2817-3300-4af2-a845-005403d75e74" width="700" />

## Sep 5, 2023
* **New feature**:  Add gene-specific CNA charts to show all levels of copy number alterations (including gain and hetloss) on _Study View_:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/873401ad-cbcf-4b07-9b80-8487af5abb18" width="700" />

## Aug 30, 2023
* **Enhancement**: Exclude a patient from your selection directly from the _Patient View_:
    
    ![image](https://github.com/cBioPortal/cbioportal/assets/1334004/501fe998-8408-48e3-a183-e27ab3ff3f28)

## Aug 21, 2023
*   **Added data** consisting of 4,488 samples from 7 studies:
	* [Lung Adenocarcinoma Met Organotropism (MSK, Cancer Cell 2023)](https://www.cbioportal.org/study/summary?id=luad_mskcc_2023_met_organotropism) *2653 samples*
    * [Acute Myeloid Leukemia (OHSU, Cancer Cell 2022)](https://www.cbioportal.org/study/summary?id=aml_ohsu_2022) *942 samples*
    * [Colon Cancer (Sidra-LUMC AC-ICAM, Nat Med 2023)](https://www.cbioportal.org/study/summary?id=coad_silu_2022) *348 samples*
    * [Pediatric Neuroblastoma (MSK, Nat Genet 2023)](https://www.cbioportal.org/study/summary?id=nbl_msk_2023) *223 samples*    
    * [Colorectal Adenocarcinoma (MSK, Nat Commun 2022)](https://www.cbioportal.org/study/summary?id=bowel_colitis_msk_2022) *180 samples*
    * [Bladder Cancer (Columbia University/MSK, Cell 2018)](https://www.cbioportal.org/study/summary?id=bladder_columbia_msk_2018) *130 samples*
    * [Myoepithelial Carcinomas of Soft Tissue (WCM, CSH Molecular Case Studies 2022)](https://www.cbioportal.org/study/summary?id=stmyec_wcm_2022) *12 samples*


*   **Gene Tables Update:** Updated tables of genes (main and alias), based on [Apr 1, 2023 HGNC release](http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/). See seedDB release notes [here](https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database) for details.

## Aug 1, 2023
*   **Enhancement**: One-sided Fisher's exact tests were changed to be two-sided. The affected pages are:
	*    Results View Page - Mutual Exclusivity Tab
    *    Results View Page - Comparison Tab - Genomic Alterations Tab
    *    Comparison Page - Genomic Alterations Tab
    *    Comparison Page - Mutations Tab
    
    Please note that the Mutations tab on the Comparison page is a recent feature and was introduced with the two-sided Fisher's exact test already implemented.

    Several users pointed out that using a one-sided test was incorrect for these comparisons. Please see discussions [here](https://github.com/cBioPortal/cbioportal/issues/9943) for more information.

## Jul 18, 2023
* **New Feature**: Add mutations table and two-sided exact p-value to comparison. [Example: AR mutations in Primary vs Metastatic Prostate Cancer samples in MSK-IMPACT Clinical Sequencing Cohort](https://www.cbioportal.org/comparison/mutations?sessionId=5cf89323e4b0ab413787436c&groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D&selectedGene=AR)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/da0cd1ef-b9b4-49f0-a9d3-42cff0204dc2" width="700" />


## May 2, 2023
*   **New Feature**: The mutations tab now shows variant annotations from the [repository of Variant with Unexpected Effects (reVUE)](https://www.cancerrevue.org/).
    
    <img src="https://user-images.githubusercontent.com/16869603/234271379-6a0876ea-8477-426e-a0e4-5e0b0b66aef9.gif" width="700" />

## Apr 25, 2023
*   **Local Installations Feature**: When custom driver annotations for structural variants are loaded, one can now filter by them in both the Oncoprint and the _Study View_. 

## Apr 11, 2023
*   **New Feature**: Disable autocommit and manually commit filters in study view. Manually commit filters can improve cBioPortal performance when query large dataset.

    <img src="https://user-images.githubusercontent.com/3323006/221857599-1eb6e9a5-55c7-4fb5-bc14-283699be0990.png" width="700" />
    <img src="https://user-images.githubusercontent.com/3323006/221858244-bf359850-79b5-4f9c-8008-6b5271721854.png" width="700" />

## Apr 5, 2023
*   **Added data** consisting of 2,472 samples from 5 studies:
    * [Bladder Cancer (MSK, Cell Reports 2022)](https://www.cbioportal.org/study/summary?id=paired_bladder_2022) *1659 samples*
    * [Gastrointestinal Stromal Tumor (MSK, NPJ Precis Oncol 2023)](https://www.cbioportal.org/study/summary?id=gist_msk_2022) *499 samples*
    * [Appendiceal Cancer (MSK, J Clin Oncol 2022)](https://www.cbioportal.org/study/summary?id=appendiceal_msk_2022) *273 samples*
    * [Colorectal Cancer (MSK, Cancer Discovery 2022)](https://www.cbioportal.org/study/summary?id=coadread_mskresistance_2022) *22 samples*
    * [Nerve Sheath Tumors (Johns Hopkins, Sci Data 2020)](https://www.cbioportal.org/study/summary?id=nst_nfosi_ntap) *19 samples* [**First GRCh38 Study**]

*   **Data Improvement**
   	* Added TERT promoter mutation status to [Melanomas (TCGA, Cell 2015)](https://www.cbioportal.org/study/summary?id=skcm_tcga_pub_2015), [Papillary Thyroid Carcinoma (TCGA, Cell 2014)](https://www.cbioportal.org/study/summary?id=thca_tcga_pub) TCGA studies.

## Apr 4, 2023
*   **New Feature**: Allow numeric data type for custom data charts.
    
    <img src="https://user-images.githubusercontent.com/33220871/222459360-9c251f7f-1005-442f-811d-dd1ee1103161.png" width="700" />
    <img src="https://user-images.githubusercontent.com/33220871/222460439-ac7007c8-059a-4ecc-816c-90752adff444.png" width="700" />

    This also allows to have numerical custom data after we query based on genes (custom data 2 in the image):
    
    <img src="https://user-images.githubusercontent.com/33220871/223729410-03a094f4-7671-453f-b8b8-f7bc10b61c7c.png" width="700" />
    <img src="https://user-images.githubusercontent.com/33220871/223729522-fd9e19e2-6218-40d4-b9fb-f253c4faa858.png" width="700" />

## Jan 10, 2023
*   **New Feature**: New Pathways tab on the Group Comparison view. [Example: Primary vs Metastasis samples in MSK-IMPACT Clinical Sequencing Cohort](https://www.cbioportal.org/comparison/pathways?comparisonId=61791fa92e3200072db9b64e&groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D)
    
    <img src="https://user-images.githubusercontent.com/15748980/235750129-661682f7-6593-4b7e-8873-ebba70eeae32.png" width="700" />

## Dec 13, 2022
*   **New Feature**: New Mutations tab on the Group Comparison view. [Example: Primary vs Metastasis samples in MSK-IMPACT Clinical Sequencing Cohort](https://www.cbioportal.org/comparison/mutations?comparisonId=61791fa92e3200072db9b64e&groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D)
    
    <img src="https://user-images.githubusercontent.com/15748980/235748831-78652ccd-b3ab-496f-9626-74ceb8c7230d.png" width="700" />

## Oct 12, 2022
*   **Added data** consisting of 1,459 samples from 10 studies:
	* [Hepatocellular Carcinoma (MERiC/Basel, Nat Commun. 2022)](https://www.cbioportal.org/study/summary?id=hcc_meric_2021) *122 samples*
    * [Prostate Cancer Brain Metastases (Bern, Nat Commun. 2022)](https://www.cbioportal.org/study/summary?id=prostate_pcbm_swiss_2019) *168 samples*
    * [Pan-Cancer MSK-IMPACT MET Validation Cohort (MSK 2022)](https://www.cbioportal.org/study/summary?id=mixed_impact_subset_2022) *69 samples*
    * [Endometrial Carcinoma cfDNA (MSK, Clin Cancer Res 2022)](https://www.cbioportal.org/study/summary?id=ucec_ccr_cfdna_msk_2022) *44 samples*
    * [Endometrial Carcinoma MSI (MSK, Clin Cancer Res 2022)](https://www.cbioportal.org/study/summary?id=ucec_ccr_msk_2022) *181 samples*
    * [Gallbladder Cancer (MSK, Clin Cancer Res, 2022)](https://www.cbioportal.org/study/summary?id=gbc_mskcc_2022) *244 samples*
    * [Meningioma (University of Toronto, Nature 2021)](https://www.cbioportal.org/study/summary?id=mng_utoronto_2021) *121 samples*
	* [Mixed Tumors: Selpercatinib RET Trial (MSK, Nat Commun. 2022)](https://www.cbioportal.org/study/summary?id=mixed_selpercatinib_2020) *188 samples*
	* [Low-Grade Serous Ovarian Cancer (MSK, Clin Cancer Res 2022)](https://www.cbioportal.org/study/summary?id=lgsoc_mapk_msk_2022) *119 samples*
	* [Urothelial Carcinoma (BCAN/HCRN 2022)](http://www.cbioportal.org/study/summary?id=blca_bcan_hcrn_2022) *203 samples*

## Sep 6, 2022
*   **Enhancement**: Oncoprint can now save clinical tracks after login:
    
    <img src="https://user-images.githubusercontent.com/3323006/188828765-ded10947-7343-49dd-b605-50b58b67a388.png" width="700" />

## Aug 11, 2022
*   **New Major Release**: **v5.0.0** release drops support for fusions in the mutation data format. Going forward fusions can only be imported in the Structural Variant (SV) format. This is mainly a refactoring effort to simplify the codebase and pave the way for the development of novel structural variant visualizations in the future. For cBioPortal instance maintainer, please reference our [Migration Guide](https://docs.cbioportal.org/migration-guide/#v4---v5) for instruction.
	
## Jul 26, 2022
*   **Added data** consisting of 6,631 samples from 7 studies:
	* [Metastatic Biliary Tract Cancers (SUMMIT - Neratinib Basket Trial, 2022)](https://www.cbioportal.org/study/summary?id=biliary_tract_summit_2022) *36 samples*
    * [Rectal Cancer (MSK, Nature Medicine 2022)](https://www.cbioportal.org/study/summary?id=rectal_msk_2022) *801 samples*
    * [Lung Adenocarcinoma (MSK Mind,Nature Cancer 2022)](https://www.cbioportal.org/study/summary?id=lung_msk_mind_2020) *247 samples*
    * [Myelodysplastic Syndromes (MDS IWG, IPSSM, NEJM Evidence 2022)](https://www.cbioportal.org/study/summary?id=mds_iwg_2022) *3,323 samples*
    * [Esophagogastric Cancer (MSK, Clin Cancer Res 2022)](https://www.cbioportal.org/study/summary?id=egc_msk_tp53_ccr_2022) *237 samples*
    * [Pan-cancer Analysis of Advanced and Metastatic Tumors (BCGSC, Nature Cancer 2020)](https://www.cbioportal.org/study/summary?id=pog570_bcgsc_2020) *570 samples*
    * [Prostate Adenocarcinoma (MSK, Clin Cancer Res. 2022)](https://www.cbioportal.org/study/summary?id=prad_pik3r1_msk_2021) *1,417 samples*

## Jun 7, 2022
* **New Feature**: Add heatmap to plot options on _Comparison Page_. [Example: Primary vs Metastatic Prostate Cancer in MSK-IMPACT (2017)](https://www.cbioportal.org/comparison/clinical?sessionId=5cf89323e4b0ab413787436c)
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/1a048863-7db6-43c6-8250-c31039a2d67f" width="700" />

## May 31, 2022
* **New Feature**: Added Quartiles, Median split and Generate bins options for bar charts on the study view page, where Generate bins allows user to define bin size and min value
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/290e6d45-1c10-4a51-a823-24db27063486" width="700" />


## May 12, 2022
* **New Feature**: Show cohort alteration frequencies in pathways from [NDEx](https://www.ndexbio.org/iquery/) on the _Results View_. [Example: Glioblastoma signaling pathways in MSK-IMPACT (2017) cohort](https://bit.ly/3sE4UqD)

    <img src="https://user-images.githubusercontent.com/1334004/168684594-fafd1fd6-b2a9-45d7-ad3a-d5611f3f469c.png" width="700" />

## May 24, 2022
* **New Feature**: Add Help buttons on various pages and tabs, including the homepage

## May 10, 2022
* **New Feature**: Use IGV for the genomic overview on the _Patient View_. [Example: Endometrial cancer patient in TCGA](https://www.cbioportal.org/patient?studyId=ucec_tcga_pub&caseId=TCGA-BK-A0CC)

   <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/6fa6ab4b-4c09-4bd5-a7c2-3be2f418d83a" width="700" /> 

## May 5, 2022
* **New Feature**: View mutations and copy number changes in the [Integrative Genomics Viewer (IGV)](https://igv.org/) on the _Patient View_. [Example: Endometrial cancer patient in TCGA](https://www.cbioportal.org/patient?studyId=ucec_tcga_pub&caseId=TCGA-BK-A0CC)

    <img src="https://user-images.githubusercontent.com/1334004/168593020-253a4bfe-3922-4bf9-8dcc-89e298ba87c7.png" width="700" />
* **New Feature**: Add charts that plot categorical vs continuous data on the _Study View_. [Example: MSK-IMPACT (2017) cohort](https://www.cbioportal.org/study/summary?id=msk_impact_2017)

    <img src="https://user-images.githubusercontent.com/1334004/168595352-6416db7e-b6b5-4ef8-95e7-eaf67c841d94.png" width="700" />
* **New Feature**: Several single cell data integrations are now available for the [CPTAC glioblastoma study](https://www.cbioportal.org/study/summary?id=gbm_cptac_2021), allowing one to:
    *  Compare genomic alterations and cell type fractions in oncoprints on the _Results View_ ([Example](https://bit.ly/37QiC2x))
    *  Explore the single cell data further in [Vitessce](http://vitessce.io/) on the _Patient View_ ([Example](https://www.cbioportal.org/patient/customTab1?studyId=gbm_cptac_2021&caseId=C3N-02783))
    *  Create cohorts and groups based on cell type fractions on the _Study View_ ([Example](https://tinyurl.com/singlecellbioportal))
    *  Compare differences in cell type fractions between groups on the _Comparison Page_ ([Example](https://www.cbioportal.org/comparison/generic_assay_single_cell_type_fractions?comparisonId=6282d44d0934121b56df47fb&unselectedGroups=%5B%220.05-0.06%22%2C%220.03-0.05%22%5D))

    <img src="https://user-images.githubusercontent.com/1334004/168699333-a50ef433-5efe-45cc-a477-a3a99ed419dc.png" width="700" />

## Apr 20, 2022
*   **Added data** consisting of 2,557 samples from 5 studies:
    * [Breast Cancer (HTAN OHSU, 2022)](https://www.cbioportal.org/study/summary?id=brca_hta9_htan_2022) *5 samples*
    * [Colorectal Cancer (MSK, 2022)](https://www.cbioportal.org/study/summary?id=crc_dd_2022) *47 samples*
    * [Pediatric Pancan Tumors (MSK, 2022)](https://www.cbioportal.org/study/summary?id=mixed_kunga_msk_2022) *135 samples*
    * [Sarcoma (MSK, 2022)](https://www.cbioportal.org/study/summary?id=sarcoma_mskcc_2022) *2,138 samples*
    * [Lung Cancer in Never Smokers (NCI, Nature Genetics 2021)](https://www.cbioportal.org/study/summary?id=lung_nci_2022) *232 samples*
*   **Gene Tables Update** Updated tables of genes (main and alias), based on [Jan 1, 2022 HGNC release](http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/). See seedDB release note [here](https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database) for details.
*   **Data Improvement**
    * Pan-can studies timeline addition: `TREATMENT`, `OTHER MALIGNANCY FORM`, `SAMPLE ACQUISITION`, `STATUS` are added to all 32 TCGA Pan-Can studies. Details for data source and transformation process can be found [here](https://github.com/cBioPortal/datahub/pull/1597) or in the [README.md](https://github.com/cBioPortal/datahub/blob/master/public/acc_tcga_pan_can_atlas_2018/README.md) files included in each study folder on datahub. [Example: patient view of TCGA-A2-A04P in Breast Invasive Carcinoma Tumor Type](https://www.cbioportal.org/patient?studyId=brca_tcga_pan_can_atlas_2018&caseId=TCGA-A2-A04P)<img width="1204" alt="Screen Shot 2022-04-25 at 11 40 51 AM" src="https://user-images.githubusercontent.com/5973438/165153293-3d400e03-48d5-4919-ae32-7f16fa7d2f4f.png">

    * Pan-can studies methylation addition: methylation profile (27k and 450k merged) are added to all 32 TCGA Pan-Can studies, in generic assay format. Data source: [GDC](https://gdc.cancer.gov/node/977). Example: search by gene or probe from dropdown, to add a chart in study view, a track in Oncoprint (single study query only), or plots in plots tab. <img width="860" alt="Screen Shot 2022-03-22 at 9 52 20 PM" src="https://user-images.githubusercontent.com/5973438/159626976-72601dc9-e5be-4b50-bae9-dedb5be6e7fc.png"><img width="340" alt="Screen Shot 2022-03-22 at 9 52 27 PM" src="https://user-images.githubusercontent.com/5973438/159626985-2423c556-2180-4a00-b9b8-69b4d2e361ba.png"><img width="430" alt="Screen Shot 2022-03-22 at 9 53 38 PM" src="https://user-images.githubusercontent.com/5973438/159626987-e6b51bef-814e-4a76-826a-211fcdd5e9a1.png">

    * Single cell (type fraction and phases) data (in generic assay format) is added to [Glioblastoma (CPTAC, Cell 2021)](https://www.cbioportal.org/study/summary?id=gbm_cptac_2021)
*   **New Feature** For the new HTAN OHSU study there is now also an integration with [Harvard LSP](https://labsyspharm.org/)'s [Minerva](https://www.cycif.org/software/minerva) for exploring multiplex imaging:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/aa0f0c99-4b9c-481d-9f7f-891ea03fdd34" width="700" />


## Mar 1, 2022
* **New Documentation**:
    * Add a [User Guide](https://docs.cbioportal.org/user-guide/)
    * Add [Study View and Query How-to videos](https://docs.cbioportal.org/user-guide/overview/#how-to-videos)

## Feb 8, 2022
* **New Feature**: Create X vs Y violin plots in _Study View_ using any categorical and numerical clinical data:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/94d7dd21-4415-405d-8e9b-46e1de3dc095" width="700" />

## Jan 19, 2022
* **New Feature**: Numerical filters on the _Study View_ are now editable:
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/00926fec-f32b-48c7-a85f-fced7c29070f" width="700" />
* **New Feature**: In the annotation column choose between showing a single icon OncoKB icon or multiple (one for therapeutic, diagnostic and prognostic):
    
    <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/87c5f7d6-10d7-4631-8b37-fa8cef4cabe2" width="700" />

## Jan 10, 2022
* **New Documentation for Local cBioPortal Installations**:
    * Explain how to use [Genome Nexus to annotate MAF files](https://docs.cbioportal.org/file-formats/#create-the-cbioportal-mutation-data-file-with-genome-nexus-with-a-minimal-maf-file)

## Jan 4, 2022
*   **Added data** consisting of 27,447 samples from 10 studies:
    * [Endometrial Carcinoma (CPTAC, Cell 2020)](https://www.cbioportal.org/study/summary?id=ucec_cptac_2020) *95 samples*
    * [Pancreatic Ductal Adenocarcinoma (CPTAC, Cell 2021)](https://www.cbioportal.org/study/summary?id=paad_cptac_2021) *140 samples*
    * [Lung Squamous Cell Carcinoma (CPTAC, Cell 2021)](https://www.cbioportal.org/study/summary?id=lusc_cptac_2021) *108 samples*
    * [Lung Adenocarcinoma (CPTAC, Cell 2020)](https://www.cbioportal.org/study/summary?id=luad_cptac_2020) *110 samples*
    * [Glioblastoma (CPTAC, Cell 2021)](https://www.cbioportal.org/study/summary?id=gbm_cptac_2021) *99 samples*
    * [Breast Cancer (CPTAC, Cell 2020)](https://www.cbioportal.org/study/summary?id=brca_cptac_2020) *122 samples*
    * [Pediatric Brain Cancer (CPTAC/CHOP, Cell 2020)](https://www.cbioportal.org/study/summary?id=brain_cptac_2020) *218 samples*
    * [Metastatic Prostate Cancer (Provisional, June 2021)](https://www.cbioportal.org/study/summary?id=mpcproject_broad_2021) *123 samples*
    * [MSK MetTropism (MSK, Cell 2021)](https://www.cbioportal.org/study/summary?id=msk_met_2021) *25,775 samples*
    * [Cancer Therapy and Clonal Hematopoiesis (MSK, 2021)](https://www.cbioportal.org/study/summary?id=msk_ch_ped_2021) *657 samples*


*   **Added TMB (nonsynonymous) scores** for all studies. [Example: new TMB field for study gbm_cptac_2021](https://www.cbioportal.org/study/clinicalData?id=gbm_cptac_2021)   
    (Details for the calculation can be found [HERE](https://github.com/cBioPortal/datahub-study-curation-tools/blob/master/tmb/calculate_tmb/Readme.md))   
    <img width="361" alt="Screen Shot 2022-01-04 at 11 39 20 PM" src="https://user-images.githubusercontent.com/5973438/148133673-a23506e9-7957-4581-99f0-820ca9548672.png">

## Nov 12, 2021
*   **Added data** consisting of 3,680 samples from 6 studies:
    * [Breast Cancer MAPK (MSKCC, Nat Commun 2021)](https://www.cbioportal.org/study/summary?id=brca_mapk_hp_msk_2021) *145 samples*
    * [Colorectal Cancer (MSK, 2020)](https://www.cbioportal.org/study/summary?id=crc_nigerian_2020) *64 samples*
    * [Breast Cancer (MSK, Clinical Cancer Res 2020)](https://www.cbioportal.org/study/summary?id=brca_pareja_msk_2020) *60 samples*
    * [High-Grade Serous Ovarian Cancer (MSK, 2021)](https://www.cbioportal.org/study/summary?id=hgsoc_msk_2021) *45 samples*
    * [Diffuse Glioma (GLASS Consortium, Nature 2019)](https://www.cbioportal.org/study/summary?id=difg_glass_2019) *444 samples*
    * [Pan-cancer analysis of whole genomes (ICGC/TCGA, Nature 2020)](https://www.cbioportal.org/study/summary?id=pancan_pcawg_2020) *2,922 samples*

## Nov 3,2021
* **New Feature**: Add Uniprot topology as a new annotation track on the Mutations Tab of the Results View. [Example: EGFR in MSK-IMPACT (2017) cohort](https://www.cbioportal.org/results/mutations?cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&profileFilter=mutations%2Cfusion%2Ccna&case_set_id=msk_impact_2017_cnaseq&gene_list=EGFR&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/16869603/141861719-f91d787e-e8c3-4c27-8273-7642ec3fd5e1.png" width="700" />

## Oct 1, 2021
* **New Feature**: Arm level Copy Number events are now loaded into cBioPortal using the Categorial Generic Assay Data Type. They can be found in a tab under the Add Charts Button of the _Study View_ [Example: Arm Level Data in TCGA PanCancer Atlas](https://www.cbioportal.org/study/summary?id=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018)

     <img src="https://user-images.githubusercontent.com/1334004/139692484-8245d94f-1821-4476-974e-b6c1a6cd8d12.png" width="700" />

## Sep 22, 2021
*   **Added data** consisting of 14,844 samples from 7 studies:
    * [Colorectal Cancer (MSK, Gastroenterology 2020)](https://www.cbioportal.org/study/summary?id=crc_apc_impact_2020) *471 samples*
    * [Metastatic Breast Cancer (MSK, Cancer Discovery 2021)](https://www.cbioportal.org/study/summary?id=breast_ink4_msk_2021) *1,365 samples*
    * [Lung Adenocarcinoma (MSKCC, 2021)](https://www.cbioportal.org/study/summary?id=luad_mskimpact_2021) *186 samples*
    * [Race Differences in Prostate Cancer (MSK, 2021) ](https://www.cbioportal.org/study/summary?id=prad_msk_stopsack_2021) *2,069 samples*
    * [Medulloblastoma (DKFZ, Nature 2017)](https://www.cbioportal.org/study/summary?id=mbl_dkfz_2017) *491 samples*
    * [Thoracic Cancer (MSK, 2021)](https://www.cbioportal.org/study/summary?id=lung_pdx_msk_2021) *68 samples*
    * [China Pan-cancer (OrigiMed, 2020)](https://www.cbioportal.org/study/summary?id=pan_origimed_2020) *10,194 samples*

## Sep 21, 2021
* **Enhancement**: Dowloading the Lollipop plot on the Mutations Tab of the _Results View_ will now also include the annotation tracks:

    <img src="https://user-images.githubusercontent.com/1334004/139694195-51ad5580-efa5-4b25-9a10-2754726a246b.png" width="700" />

## Aug 17, 2021
* **New Feature**: The Mutations Tab of the _Results View_ can now show exon numbers as an annotation track [Example: MET Exon 14 Mutations in MSK-IMPACT (2017) cohort](https://www.cbioportal.org/results/mutations?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=mutations%2Cfusion%2Ccna&case_set_id=msk_impact_2017_cnaseq&gene_list=MET&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/1334004/139694614-62f62640-2fab-4117-ae09-db2281690f76.png" width="700" />

## Aug 10, 2021
* **New Feature**: Use the filtering capabilities in the Mutations Tab of the _Results View_ to create a custom cohort that one can open directly in the _Study View_ [Example: CTNNB1 in MSK-IMPACT (2017) cohort](https://bit.ly/3w6wu00)

    <img src="https://user-images.githubusercontent.com/1334004/139696976-6e145c05-10d3-44c9-8499-146ad6e13f2f.png" width="700" />

## Jul 27, 2021
* **New Feature**: Add a custom filter to any column of the Mutations Tab in the _Results View_ [Example: CTNNB1 in MSK-IMPACT (2017) cohort](https://bit.ly/3w6wu00)

    <img src="https://user-images.githubusercontent.com/1334004/139697555-34507fa8-2b57-4aa8-84ee-c477efffa864.png" width="700" />

* **New Feature**: Show detailed descriptions for each annotation source in the header of the the Mutations Table in both the _Results View_ and the _Patient View_ [Example link](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P04)

    <img src="https://user-images.githubusercontent.com/1334004/139699259-1be86599-c2e0-457c-b257-1b0d446facf4.png" width="350" />

## Jul 6, 2021
* **New Feature**: Add any clinical data as a column on the Mutations Tab in the _Results View_ [Example: EGFR in MSK-IMPACT (2017) cohort](https://bit.ly/3bxFlyp)

    <img src="https://user-images.githubusercontent.com/1334004/139700811-4b6ce432-4abf-4dc4-8371-634a14d2b603.png" width="700" />

## June 23, 2021
*   **Added data** consisting of 1,084 samples from 5 studies:
    * [Intrahepatic Cholangiocarcinoma (MSK, Hepatology 2021)](https://www.cbioportal.org/study/summary?id=ihch_msk_2021) *412 samples*
    * [Intrahepatic Cholangiocarcinoma (Mount Sinai 2015)](https://www.cbioportal.org/study/summary?id=ihch_ismms_2015) *8 samples*
    * [RAD51B Associated Mixed Cancers (Mandelker 2021](https://www.cbioportal.org/study/summary?id=mixed_msk_tcga_2021) *19 samples*
    * [Intrahepatic Cholangiocarcinoma (MSK, 2020)](https://www.cbioportal.org/study/summary?id=ihch_mskcc_2020) *219 samples*
    * [Lung Adenocarcinoma (NPJ Precision Oncology, MSK 2021)](https://www.cbioportal.org/study/summary?id=luad_msk_npjpo_2021) *426 samples*
*   **Added mass-spec proteome data** from CPTAC to [Breast Invasive Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018), [Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018) and [Colorectal Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=coadread_tcga_pan_can_atlas_2018).
*   **Added mass-spec phosphoproteome site level expression** from CPTAC to [Breast Invasive Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018) and [Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018).
*   **Updated gene tables** Updated tables of genes (main and alias), based on [HGNC](https://www.genenames.org/). See details [HERE](https://github.com/cBioPortal/datahub/tree/master/seedDB#seed-database-schema-2128) in section `Contents of seed database`. Sripts/resources/process used to construct new tables are described [HERE](https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/gene-table-update/build-input-for-importer).

## June 1, 2021
* **New Feature**: In certain studies where we have the data we show read counts for uncalled mutations on the _Patient View_ [Example: A patient in the Glioma (MSK, 2019) cohort](https://www.cbioportal.org/patient?studyId=glioma_msk_2018&caseId=IM-GBM-1)

    <img src="https://user-images.githubusercontent.com/1334004/139701937-91a40335-e95a-45f4-9800-356737a17fab.png" width="700" />

## May 10, 2021
*  **New Feature**: Pick color for User Defined Groups [Example: Color Bladder Cancer Group in MSK-IMPACT (2017) cohort](https://www.cbioportal.org/study/summary?id=msk_impact_2017), implemented by [The Hyve](https://www.thehyve.nl/services/cbioportal).

    <img src="https://user-images.githubusercontent.com/1334004/122818654-7b018280-d2a7-11eb-94c1-834b780618b9.png" width="700" />

## May 4, 2021
*   **New Feature**: Add more categories of mutations to the Mutations Tab on the _Results View_, including Driver/VUS, Splice and Structural Variants [Example: TP53 alterations in the MSK-IMPACT (2017) cohort](https://www.cbioportal.org/results/mutations?cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&profileFilter=mutations%2Cfusion%2Ccna&case_set_id=msk_impact_2017_cnaseq&gene_list=TP53&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/1334004/122817171-99ff1500-d2a5-11eb-8a58-374c9ea32013.png" width="700" />

## April 21, 2021
*   **Added data** consisting of 4074 samples from 9 studies:
    * [Metaplastic Breast Cancer (MSK, 2021)](https://www.cbioportal.org/study/summary?id=mbc_msk_2021) *19 samples*
    * [Lung Adenocarcinoma (MSKCC, 2020)](https://www.cbioportal.org/study/summary?id=luad_mskcc_2020) *604 samples*
    * [Cutaneous Squamous Cell Carcinoma (UCSF, 2021)](https://www.cbioportal.org/study/summary?id=cscc_ucsf_2021) *105 samples*
    * [MSK-IMPACT and MSK-ACCESS Mixed Cohort (MSK, 2021)](https://www.cbioportal.org/study/summary?id=msk_access_2021) *1446 samples*
    * [Melanoma (MSKCC, 2018)](https://www.cbioportal.org/study/summary?id=mel_mskimpact_2020) *720 samples*
    * [Cholangiocarcinoma (ICGC, Cancer Discov 2017)](https://www.cbioportal.org/study/summary?id=chol_icgc_2017) *489 samples*
    * [Esophageal/Stomach Cancer (MSK, 2020)](https://www.cbioportal.org/study/summary?id=egc_mskcc_2020) *487 samples*
    * [Retinoblastoma (MSK, Cancers 2021)](https://www.cbioportal.org/study/summary?id=rbl_mskcc_2020) *83 samples*
    * [Combined Hepatocellular and Intrahepatic Cholangiocarcinoma (Peking University, Cancer Cell 2019)](https://www.cbioportal.org/study/summary?id=hccihch_pku_2019) *121 samples*

## April 20, 2021
*   **New Feature**: Add driver annotations to download tab on _Results View_ [Example: RAS/RAF alterations in colorectal cancer](https://www.cbioportal.org/results/download?cancer_study_list=coadread_tcga_pub&cancer_study_id=coadread_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=coadread_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=coadread_tcga_pub_gistic&Z_SCORE_THRESHOLD=2.0&case_set_id=coadread_tcga_pub_nonhypermut&gene_list=KRAS%20NRAS%20BRAF&gene_set_choice=user-defined-list)

    <img src="https://user-images.githubusercontent.com/1334004/122816424-b2bafb00-d2a4-11eb-8a0f-5e878bb70ccd.png" width="700" />

## March 30, 2021
*   **Enhancement**: Add 95% Confidence Interval for Survival Plots [Example: Altered vs Unaltered EGFR in Lung Cancer](https://www.cbioportal.org/results/comparison?plots_horz_selection=%7B%7D&plots_vert_selection=%7B%7D&plots_coloring_selection=%7B%7D&tab_index=tab_visualize&Action=Submit&session_id=60d0d9fce4b015b63e9f1116&comparison_subtab=survival)

    <img src="https://user-images.githubusercontent.com/1334004/122810147-0295c400-d29d-11eb-9be7-fa14787d72b7.png" width="700" />

## March 11, 2021
*   **New Feature**: Combine different types of alterations in _Comparison View_ [Example: Deletions and Truncating events in primary vs metastases](https://www.cbioportal.org/comparison/alterations?comparisonId=5f73ac54e4b0242bd5d22ab0&selectedEnrichmentEventTypes=%5B%22truncating%22%2C%22feature_truncation%22%2C%22nonsense_mutation%22%2C%22nonsense%22%2C%22stopgain_snv%22%2C%22stop_gained%22%2C%22frameshift%22%2C%22frameshift_variant%22%2C%22frame_shift_del%22%2C%22frameshift_deletion%22%2C%22frame_shift_ins%22%2C%22frameshift_insertion%22%2C%22translation_start_site%22%2C%22initiator_codon_variant%22%2C%22start_codon_snp%22%2C%22start_codon_del%22%2C%22de_novo_start_outofframe%22%2C%22nonstop_mutation%22%2C%22stop_lost%22%2C%22splice_site%22%2C%22splice%22%2C%22splicing%22%2C%22splice_site_snp%22%2C%22splice_site_del%22%2C%22splice_site_indel%22%2C%22splice_region_variant%22%2C%22splice_region%22%2C%22HOMDEL%22%5D&groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D) or read more on [The Hyve's blog](https://www.thehyve.nl/articles/find-differentially-altered-genes-by-combining-alteration-types)

    <img src="https://user-images.githubusercontent.com/1334004/122808698-35d75380-d29b-11eb-8928-0fca5f6e3b34.png" width="700" />

*   **Enhancement**: Improve UI for OncoPrint, aggregating various data modalities in a single add track dropdown button [Example: Add clinical, heatmap and treatment response data into the OncoPrint](https://www.cbioportal.org/results/oncoprint?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ccle_broad_2019_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ccle_broad_2019_cna&cancer_study_list=ccle_broad_2019&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=0&case_set_id=ccle_broad_2019_cnaseq&gene_list=TP53&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/15748980/107722872-f32eba80-6cad-11eb-89d0-b6c5f300ab41.png" width="700" />

## February 16, 2021
*   **Enhancement**: Show only TCGA PanCancer Atlas Pathways in _Results_ and _Patient View_ to avoid showing many similar pathways [Example: Clinvar APC and CTNNB1 alterations in WNT pathway](https://www.cbioportal.org/results/pathways?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=mutations%2Cfusion%2Ccna&case_set_id=msk_impact_2017_cnaseq&gene_list=APC%2520CTNNB1&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/1334004/122806359-3a4e3d00-d298-11eb-89fc-9e107a4651b4.png" width="700" />

## January 28, 2021
*   **New Feature**: Show ClinVar Interpretation in Mutation tables [Example: Clinvar Interpretations in BRCA2](https://www.cbioportal.org/results/mutations?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&cancer_study_list=msk_impact_2017&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=0&case_set_id=msk_impact_2017_cnaseq&gene_list=BRCA2&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

    <img src="https://user-images.githubusercontent.com/1334004/122805797-8a78cf80-d297-11eb-9454-fed37ed1ddb5.png" width="700" />


## January 12, 2021
*   **New Feature**: Add your own custom data for a sample or patient to use on the study or comparison view  [Example: Add custom data to three samples and do a comparison](https://www.cbioportal.org/study/summary?id=msk_impact_2017)

    <img src="https://user-images.githubusercontent.com/1334004/104651553-ce7efd00-5685-11eb-8017-c43f24421e3f.png" width="700">

*   **New Feature**: Show the mutations of a patient inside a pathway schematic using PathwayMapper  [Example: Notch signaling pathway in a prostate cancer patient](https://www.cbioportal.org/patient/pathways?studyId=msk_impact_2017&caseId=P-0000377)

    <img src="https://user-images.githubusercontent.com/1334004/104649378-9f1ac100-5682-11eb-91eb-4e0f024a4ded.png" width="700">

*   **New Feature**: Display and compare generic assays, such as microbiome and treatment response, on the study view [Example: Prasinovirus microbiome signatures in TCGA](https://www.cbioportal.org/study?id=6000b6aae4b015b63e9d4d44)
    
    <img src="https://user-images.githubusercontent.com/1334004/104651211-54e70f00-5685-11eb-94f1-f20930793c28.png" width="700">

*   **New Feature**: The Plots tab on _Results View_ now allows you to group alterations by Driver and VUS [Example: POLE driver mutations vs VUSs against mutation counts in TCGA Colorectal Adenocarcinoma](https://bit.ly/3ssY9Gu)
    <img src="https://user-images.githubusercontent.com/1334004/104768867-56bfd980-573c-11eb-8e03-5ad86d01f25a.png" width="700">


## December 31, 2020
*   **Added data** consisting of 430 samples from 5 studies:
    * [Juvenile Papillomatosis and Breast Cancer (MSK, 2020)](https://www.cbioportal.org/study/summary?id=brca_jup_msk_2020) *5 samples*
    * [Mixed cfDNA (MSKCC, 2020)](https://www.cbioportal.org/study/summary?id=mixed_cfdna_msk_2020) *229 samples*
    * [Metastatic Melanoma (DFCI, Nature Medicine 2019)](https://www.cbioportal.org/study/summary?id=mel_dfci_2019) *144 samples*
    * [Lung Cancer (SMC, Cancer Research 2016)](https://www.cbioportal.org/study/summary?id=lung_smc_2016) *22 samples*
    * [Upper Tract Urothelial Carcinoma (IGBMC, Genome Biology 2021)](https://www.cbioportal.org/study/summary?id=utuc_igbmc_2021) *30 samples*
*    **Added survival data** to [Breast Cancer (METABRIC, Nature 2012 & Nat Commun 2016)](https://www.cbioportal.org/study/summary?id=brca_metabric)

## November 3, 2020
*   **New Feature**: The [map](https://www.cbioportal.org/installations) of local installations of cBioPortal is available now. Please consider registering your instance [here](http://bit.ly/2XW3wNG ).
    ![image](https://user-images.githubusercontent.com/840895/99544685-96f41d00-2982-11eb-9de2-457a4456c294.png)

*   **Enhancement**: upgraded the _Genomic Evolution_ tab in _Patient View_ with timeline [Example](https://www.cbioportal.org/patient/genomicEvolution?caseId=P04&studyId=lgg_ucsf_2014)
    ![image](https://user-images.githubusercontent.com/840895/99543852-a3c44100-2981-11eb-8f53-14a0b51b767c.png)

## October 20, 2020
*   **Enhancement**: _Expression_ tab has now been merged into the _Plots_ tab
    ![image](https://user-images.githubusercontent.com/840895/99565969-4ee0f480-299a-11eb-8371-8cfd09c6e4ba.png)

## October 16, 2020
*   **Added data** consisting of 25,078 samples from 5 studies:
    * [Melanomas (TCGA, Cell 2015)](https://cbioportal.org/study/summary?id=skcm_tcga_pub_2015) *359 samples*
    * [Retinoblastoma cfDNA (MSKCC 2020)](https://www.cbioportal.org/study/summary?id=rbl_cfdna_msk_2020) *14 samples*
    * [The Angiosarcoma Project (Provisional, July 2020)](https://www.cbioportal.org/study/summary?id=angs_painter_2020) *83 samples*
    * [Bladder Cancer (MSK/TCGA, 2020)](https://www.cbioportal.org/study/summary?id=blca_msk_tcga_2020) *476 samples*
    * [Cancer Therapy and Clonal Hematopoiesis (MSK, 2020)](https://www.cbioportal.org/study/summary?id=msk_ch_2020) *24,146 samples*

*    **Added MSI data** (MSIsensor from [Mariamidze et al. 2018](https://www.sciencedirect.com/science/article/pii/S0092867418303131#bib66) and MANTIS scores from [Roychowdhury et al. 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5972025/)) for all 32 TCGA PanCan Atlas Cohorts.
     <img width="600" src="https://user-images.githubusercontent.com/5973438/95988728-52131e80-0df7-11eb-8e25-0617d4b9e8c4.png">

*    **Added new profile** ‚ÄúRNA-Seq V2 expression Z-scores relative to normal samples‚Äù for 16 TCGA PanCan Atlas Cohorts.
     The normals samples RNA-Seq V2 expression data were curated from [GDC](https://gdc.cancer.gov/about-data/publications/pancanatlas), and can be downloaded from our [Datahub](https://github.com/cBioPortal/datahub) or [Data Set](https://www.cbioportal.org/datasets) page. [Example: ERBB2 expression z-scores relative to normal expression](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&case_set_id=all&data_priority=0&gene_list=ERBB2&geneset_list=%20&plots_coloring_selection=%7B%22colorByCopyNumber%22%3A%22true%22%7D&plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_STUDY%22%7D&plots_vert_selection=%7B%22selectedGeneOption%22%3A2064%2C%22dataType%22%3A%22MRNA_EXPRESSION%22%2C%22selectedDataSourceOption%22%3A%22rna_seq_v2_mrna_median_all_sample_ref_normal_Zscores%22%2C%22logScale%22%3A%22true%22%7D&profileFilter=0&tab_index=tab_visualize)

![image](https://user-images.githubusercontent.com/840895/96191246-b32f1500-0f11-11eb-8727-bcfdd2d724a3.png)

## October 13, 2020
*   **Enhancement**: _Study View_ now allows comparing samples with mutations or copy number alterations in different genes
    ![image](https://user-images.githubusercontent.com/840895/99558433-ba729400-2991-11eb-9054-1d8751b4d84f.png)

*   **New Feature**: When treatment timeline is available (e.g. in [this study](https://www.cbioportal.org/study/summary?id=lgg_ucsf_2014)), _Study View_ now allows the selection and comparison of patients treated with specific drugs, or samples sequenced pre or post specific drug treatments
    ![image](https://user-images.githubusercontent.com/840895/99556958-248a3980-2990-11eb-97e3-d833518f8aef.png)

## September 30, 2020
*   **New Feature**: Microbiome signature data is available for comparison now. [Example: comparing colorectal subtypes for enriched microbiome signatures](https://www.cbioportal.org/comparison/generic_assay_microbiome_signature?sessionId=5dea8bbde4b05f1dfb8ace51)
    ![image](https://user-images.githubusercontent.com/840895/99566847-666cad00-299b-11eb-91d3-21b21f267733.png)

## September 22, 2020

*   **Enhancement**: The timeline feature in _Patient View_ has been refactored with an improved UI. [Example](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P04)
    ![image](https://user-images.githubusercontent.com/840895/94962574-ae577380-04c4-11eb-9ff7-1a79e242dd73.png)

*   **Enhancement**: Logrank p-values are now provided for all survival analysis (previously only availalbe when comparing two groups). [Example](https://www.cbioportal.org/comparison/survival?sessionId=5f77816de4b015b63e9c0f81)
    <img width="600" alt="" src="https://user-images.githubusercontent.com/840895/94962916-481f2080-04c5-11eb-811f-40d8fc4437cf.png">


## August 11, 2020
*   **New Feature**: microbiome data of TCGA samples from [Poore et al. 2020](https://www.nature.com/articles/s41586-020-2095-1) are now available for analysis in the *OncoPrint* and *Plots* tabs. [Example: Orthohepadnavirus across TCGA cancers](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&case_set_id=all&data_priority=0&gene_list=EGFR&geneset_list=%20&plots_coloring_selection=%7B%22selectedOption%22%3A%22-1_undefined%22%2C%22colorByMutationType%22%3A%22false%22%2C%22colorByCopyNumber%22%3A%22false%22%7D&plots_horz_selection=%7B%22selectedGeneOption%22%3A1956%2C%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_TYPE%22%2C%22mutationCountBy%22%3A%22MutationType%22%2C%22logScale%22%3A%22false%22%7D&plots_vert_selection=%7B%22selectedGeneOption%22%3A1956%2C%22selectedGenericAssayOption%22%3A%22Orthohepadnavirus%22%2C%22dataType%22%3A%22MICROBIOME_SIGNATURE%22%2C%22selectedDataSourceOption%22%3A%22microbiome_signature%22%2C%22mutationCountBy%22%3A%22MutationType%22%2C%22logScale%22%3A%22false%22%7D&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/90060684-1783ed00-dcb3-11ea-8368-e96dd69c552c.png)

*   **New Feature**: You can now compare *DNA Methylation* data between groups using the *Comparison* feature. [Example: Comparing DNA methylation levels between samples with high vs low BRCA1 expression](https://www.cbioportal.org/results/comparison?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_id=ov_tcga_pub&cancer_study_list=ov_tcga_pub&case_set_id=ov_tcga_pub_3way_complete&comparison_selectedGroups=%5B%22BRCA1%3A%20EXP%3E1%22%2C%22BRCA1%3A%20EXP%3C-1%22%5D&comparison_subtab=dna_methylation&data_priority=0&gene_list=BRCA1%253AEXP%253E1%2520%250ABRCA1%253A%2520EXP%253C-1&gene_set_choice=user-defined-list&geneset_list=%20&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=ov_tcga_pub_mrna_median_Zscores&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ov_tcga_pub_mutations&plots_coloring_selection=%7B%7D&plots_horz_selection=%7B%7D&plots_vert_selection=%7B%22selectedDataSourceOption%22%3A%22mrna_median_Zscores%22%7D&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/89957811-632c8d00-dc06-11ea-87dc-bdbb4fa23a0b.png)

*   **Added data** consisting of 513 samples from 3 studies:
    * [Breast Cancer (SMC 2018)](https://www.cbioportal.org/study/summary?id=brca_smc_2018) *187 samples*
    * [Germ Cell Tumors and Shared Leukemias (MSK     2020)](https://www.cbioportal.org/study/summary?id=gct_msk_2020) *21 samples*
    * [Lung Adenocarcinoma (OncoSG, Nat Genet 2020)](https://www.cbioportal.org/study/summary?id=luad_oncosg_2020) *305 samples*

*   **Added RPPA data** in addition to the microbiome data for [31 TCGA Pancan studies (except LAML)](https://www.cbioportal.org/study/summary?id=acc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018)
    <img width="400" alt="" src="https://user-images.githubusercontent.com/840895/90141147-ba3a7b00-dd48-11ea-964f-840dd591e535.png">

## July 21, 2020
*   **New Feature**: The *Mutations* tab now has the option to show mutation effects for different transcripts / isoforms. Note that some annotation features are only available for the canonical isoform. [example](https://www.cbioportal.org/results/mutations?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=msk_impact_2017&case_set_id=msk_impact_2017_cnaseq&data_priority=0&gene_list=FGFR2&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&mutations_transcript_id=ENST00000358487&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/88306535-d95d6400-ccd8-11ea-9e64-0c6600f65e50.png)

*   **Enhancement**: The *Plots* tab is now supported in multi-study queries. [example](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&case_set_id=all&data_priority=0&gene_list=TP53%2520PTEN%2520BRCA1%2520IGF2%2520EGFR&geneset_list=%20&plots_coloring_selection=%7B%22colorByCopyNumber%22%3A%22false%22%7D&plots_horz_selection=%7B%22selectedGeneOption%22%3A1956%7D&plots_vert_selection=%7B%22logScale%22%3A%22true%22%7D&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/88309239-07907300-ccdc-11ea-995b-7ddd658ec46e.png)

*   **New Feature**: You can now share custom groups in the *Study View* [example](https://www.cbioportal.org/study/summary?id=msk_impact_2017#sharedGroups=5d092670e4b0ab4137876374,5c99454ee4b0ab4137873dc5)

    <img width="333" alt="" src="https://user-images.githubusercontent.com/840895/88309724-b7fe7700-ccdc-11ea-969a-28ed8551ffd7.png">


## June 11, 2020
*   **Added data** consisting of 267 samples from 2 studies:
    * [Gastric Cancer (OncoSG, 2018)](https://www.cbioportal.org/study/summary?id=stad_oncosg_2018) *147 samples*
    * 120 ctDNA samples added to [Non-Small Cell Lung Cancer (TRACERx, NEJM & Nature 2017)](https://www.cbioportal.org/study/summary?id=nsclc_tracerx_2017) *447 samples*

## June 9, 2020

*   **Enhancement**: using [OQL](https://www.cbioportal.org/oql#oql-modifiers) to query for mutations based on a protein position range. [example](https://www.cbioportal.org/results/mutations?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=msk_impact_2017&case_set_id=msk_impact_2017_cnaseq&data_priority=0&gene_list=TP53%253AMUT_(95-288*)&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&profileFilter=0&tab_index=tab_visualize)
    ![image](https://user-images.githubusercontent.com/840895/84427197-83ce6b80-abf2-11ea-9d18-3a4f4524e545.png)

*   **New Feature**: you can now send the OncoPrint data to the [OncoPrinter tool](https://www.cbioportal.org/oncoprinter) for customization.
    ![image](https://user-images.githubusercontent.com/840895/84318326-3c85a380-ab3c-11ea-97c1-34343cb3e996.png)

*   **Enhancement**: Mutational spectrum data can be downloaded from OncoPrint
    ![image](https://user-images.githubusercontent.com/840895/84322695-68585780-ab43-11ea-9224-a965331e35fc.png)


## June 2, 2020

*   **Enhancement**: Pediatric cancer studies are now grouped and highlighted in the query page
    ![image](https://user-images.githubusercontent.com/840895/84318659-c897cb00-ab3c-11ea-8209-93c940321a0f.png)

## May 6, 2020
*   **Added data** consisting of 574 samples from 3 studies:
    * [Uterine Sarcoma/Mesenchymal (MSK, Clin Cancer Res 2020)](https://www.cbioportal.org/study/summary?id=usarc_msk_2020) *108 samples*
    * [Metastatic castration-sensitive prostate cancer (MSK, Clin Cancer Res 2020)](https://www.cbioportal.org/study/summary?id=prad_mcspc_mskcc_2020) *424 samples*
    *  [Glioblastoma (Columbia, Nat Med. 2019)](https://www.cbioportal.org/study/summary?id=gbm_columbia_2019) *42 samples*
*   **Updated one study:**
    *  Expression data was added to [The Metastatic Breast Cancer Project (Provisional, February 2020)](https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017).

## April 24, 2020
* **New Feature**: Add a new chart on the _Study View_ for selecting samples based on pre-defined case lists:

    <img width="333" alt="Screen Shot 2020-04-24 at 9 18 25 AM" src="https://user-images.githubusercontent.com/1334004/80216807-a1c51a00-860c-11ea-8ba5-3781bd603ee5.png">

## April 10, 2020
*   **New Feature**: Make cohorts on the _Study View_ using continuous molecular profiles of one or more gene(s), such as mRNA expression, methylation, RPPA and continuous CNA. [example](https://www.cbioportal.org/study/summary?id=brca_tcga#filterJson={"genomicDataFilters":[{"hugoGeneSymbol":"ERBB2","profileType":"mrna_median_Zscores","values":[{"start":1.5,"end":2},{"start":2,"end":2.5},{"start":2.5,"end":3},{"start":3,"end":3.5},{"start":3.5,"end":4},{"start":4}]}],"genomicProfiles":[["mrna_median_Zscores"]],"studyIds":["brca_tcga"]})

    <img src="https://user-images.githubusercontent.com/1334004/79270846-da2a6280-7e6c-11ea-8cb9-b40e7d201ea1.png" width=500 />

    Combine this with the group comparison feature to compare e.g. all quartiles of expression:

    <img src="https://user-images.githubusercontent.com/1334004/79270618-79028f00-7e6c-11ea-8a30-9eaedee948ca.png" width=400 />

*   **New Feature**: Annotate mutations using the _Mutation Mapper Tool_ on the GRCh38 reference genome:

    [![mutation_mapper_tool_grch38](https://user-images.githubusercontent.com/1334004/79233555-07f4b480-7e37-11ea-9d0a-0cafff434fa5.png)](https://www.cbioportal.org/mutation_mapper)

## April 3, 2020
*   **New Feature**: Extended the _Comparison_ tab to support the comparison of altered samples per gene or alteration. This [example query](https://www.cbioportal.org/results/comparison?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=msk_impact_2017&case_set_id=msk_impact_2017_Non-Small_Cell_Lung_Cancer&comparison_createdGroupsSessionId=5e8a2d4ae4b0ff7ef5fdd7d0&comparison_selectedGroups=%5B%22EGFR%3A%20AMP%20%26%20MUT%22%2C%22EGFR%3A%20MUT%20only%22%2C%22EGFR%3A%20AMP%20only%22%5D&comparison_subtab=survival&data_priority=0&gene_list=EGFR%253AAMP%250AEGFR%253AMUT_DRIVER&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&profileFilter=0&tab_index=tab_visualize) compares NSCLC patients with 1) both mutated and amplified EGFR, 2) mutated EGFR only, and 3) amplified EGFR only.

![image](https://user-images.githubusercontent.com/840895/78508085-3ea13f80-7752-11ea-8e87-e486061def12.png)

## March 27, 2020
*   **Enhancement**: User selections in the Plots tab are now saved in the URL. [example](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_id=coadread_tcga_pub&cancer_study_list=ov_tcga_pan_can_atlas_2018&case_set_id=ov_tcga_pan_can_atlas_2018_cnaseq&data_priority=0&gene_list=TP53&gene_set_choice=user-defined-list&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ov_tcga_pan_can_atlas_2018_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ov_tcga_pan_can_atlas_2018_mutations&plots_horz_selection=%7B%22dataType%22%3A%22MUTATION_EXTENDED%22%7D&profileFilter=0&tab_index=tab_visualize)

*   **New Feature**: Added table of data availability per profile in the _Study View_. [example](https://www.cbioportal.org/study/summary?id=ucec_tcga_pub)

<img src="https://user-images.githubusercontent.com/840895/78508182-e454ae80-7752-11ea-914b-bac17f725a9f.png" width="300">

## March 20, 2020
*   **Enhancement**: Extended _Survival Analysis_ to support more outcome measures. [example](https://www.cbioportal.org/results/comparison?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=luad_tcga_pan_can_atlas_2018&case_set_id=luad_tcga_pan_can_atlas_2018_cnaseq&comparison_selectedGroups=%5B%22Unaltered%20group%22%2C%22Altered%20group%22%5D&comparison_subtab=survival&data_priority=0&gene_list=KRAS%250ATP53&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=luad_tcga_pan_can_atlas_2018_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=luad_tcga_pan_can_atlas_2018_mutations&profileFilter=0&tab_index=tab_visualize)

![image](https://user-images.githubusercontent.com/840895/78508508-5c23d880-7755-11ea-9e16-36bff32678ad.png)

## March 18, 2020
*   **Added data** consisting of 1,393 samples from 3 studies:
    * [Breast Cancer (Alpelisib plus AI, Nature Cancer 2020)](https://www.cbioportal.org/study/summary?id=breast_alpelisib_2020) *141 samples*
    * [Glioma (MSKCC, Clin Cancer Res 2019)](https://www.cbioportal.org/study/summary?id=glioma_mskcc_2019) *1,004 samples*
    * [Mixed cfDNA (MSK, Nature Medicine 2019)](https://www.cbioportal.org/study/summary?id=cfdna_msk_2019) *248 samples*

## March 3, 2020
* **New Feature**: Added Pathways tab to the _Results View_ page, which visualizes the alteration frequencies of genes in pathways of interest. The pathways are pulled from https://www.pathwaymapper.org and shown in a read only view. One can edit these pathways in the PathwayMapper editor. For more information see the [tutorial](https://www.cbioportal.org/tutorials#pathways).

  [![pathwaymapper_screenshot](https://user-images.githubusercontent.com/1334004/76771001-133fbc00-6775-11ea-8b36-82edb1e7be09.png)](https://www.cbioportal.org/results/pathways?Action=Submit&Z_SCORE_THRESHOLD=1.0&cancer_study_id=gbm_tcga_pub&cancer_study_list=gbm_tcga_pub&case_set_id=gbm_tcga_pub_sequenced&gene_list=TP53%20MDM2%20MDM4&gene_set_choice=user-defined_list&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=gbm_tcga_pub_cna_rae&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=gbm_tcga_pub_mutations)

## February 12, 2020
*   **Added data** consisting of 1,605 samples from 3 studies:
    * [Tumors with TRK fusions (MSK, 2019)](https://www.cbioportal.org/study/summary?id=ntrk_msk_2019) *106 samples*
    * [Lymphoma Cell Lines (MSKCC, 2020)](https://www.cbioportal.org/study/summary?id=lymphoma_cellline_msk_2020) *34 samples*
    * [Prostate Adenocarcinoma (MSKCC, 2020)](https://www.cbioportal.org/study/summary?id=prad_cdk12_mskcc_2020) *1,465 samples*

## February 6, 2020
* **New Feature**: Extend the [recent group comparison feature](https://www.cbioportal.org/tutorials#group-comparison) by allowing comparisons inside the _Results View_ page. The new tab allows for quick comparison of altered vs unaltered cases by survival, clinical information, mutation, copy number events and mRNA expression:

  [![group_results640px](https://user-images.githubusercontent.com/1334004/74002155-2a041f00-493c-11ea-9867-14740202c368.gif)](https://www.cbioportal.org/results/comparison?Z_SCORE_THRESHOLD=2.0&cancer_study_id=coadread_tcga_pub&cancer_study_list=coadread_tcga_pub&case_set_id=coadread_tcga_pub_nonhypermut&gene_list=KRAS%20NRAS%20BRAF&gene_set_choice=user-defined-list&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=coadread_tcga_pub_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=coadread_tcga_pub_mutations)

* **Performance enhancement**: the _Study View_'s mutation table now loads faster for studies with multiple gene panels. For the [genie portal](https://genie.cbioportal.org), which has a study with many different gene panels this resulted in a speed-up from ~90-120 seconds to 5 seconds.
* Read more about the v3.2.2 release [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.2.2)

## January 30, 2020
* **Enhancement**: Show HGVSg in mutations table and linkout to [Genome Nexus](https://www.genomenexus.org):

  ![hgvsg genome nexus](https://user-images.githubusercontent.com/1334004/73494837-4d026200-4383-11ea-8968-f9a8a6e00675.png)


* **Enhancement**: Add a pencil button near gene list in results page which opens interface for quickly modifying the oql of the query:

  ![edit query pencil](https://user-images.githubusercontent.com/1334004/73494684-f72dba00-4382-11ea-8bd0-001fdf0cb681.png)


* See more updates [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.2.1)

## January 29, 2020
*   **Added data** consisting of 197 samples from 2 studies:
    * [Bladder/Urinary Tract Cancer (MSK, 2019)](https://www.cbioportal.org/study/summary?id=utuc_pdx_msk_2019) *78 samples*
    * [Upper Tract Urothelial Carcinoma (MSK, 2019)](https://www.cbioportal.org/study/summary?id=utuc_msk_2019) *119 samples*

## December 19, 2019
* **Enhancement**: We restored support for submitting large queries from external applications using HTTP POST requests.  Accepted parameters are the same
  as appear in the url of a query submitted from the homepage.

* See more updates [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.8)

## December 12, 2019
* **Enhancement**: Several enhancements to the display of gene panels on the _Patient View_ page, by [The Hyve](https://www.thehyve.nl/), described in more detail [here](https://www.thehyve.nl/articles/gene-panels-cbioportal-patient-view)

  ![image](https://api.thehyve.nl/uploads/gene-panels-patient-view-cbioportal.png)
* **Enhancement**: Add Count Bubbles to Oncoprint Toolbar

  ![Screenshot from 2019-12-06 11-36-21](https://user-images.githubusercontent.com/20069833/70339336-aa16e700-181c-11ea-94ac-c4acff272e4f.png)
* See more updates [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.7)

## November 29, 2019
* **Enhancement**: Support group comparison for custom charts in _Study View_ page
* **Enhancement**: Performance improvement of Co-Expression analysis.
* **Enhancement**: Kaplan-Meier plots now supports custom time range.
* See more updates [here](https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.6)

<img src="https://user-images.githubusercontent.com/840895/69887458-02963380-12b5-11ea-8151-4d531db0fcc5.png" width="500">

## November 22, 2019

* **New Feature**: Support for Treatment response data in the Oncoprint and Plots tab, including new *Waterfall* plot type. Read more in [The Hyve's blog post](https://www.thehyve.nl/articles/treatment-response-data-cbioportal)

![image](https://api.thehyve.nl/uploads/Waterfall-plots-cbioportal.png)

## November 15, 2019
* **Enhancement**: heatmap tracks in OncoPrint now has separate headers and sub-menus. [example](https://www.cbioportal.org/results/oncoprint?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=brca_tcga&case_set_id=brca_tcga_protein_quantification&clinicallist=PROFILED_IN_brca_tcga_mutations&data_priority=0&gene_list=ESR1%2520PGR%2520ERBB2&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_gistic&genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_mrna_median_Zscores&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_mutations&genetic_profile_ids_PROFILE_PROTEIN_EXPRESSION=brca_tcga_rppa_Zscores&heatmap_track_groups=brca_tcga_rna_seq_v2_mrna_median_Zscores%2CESR1%2CPGR%2CERBB2%3Bbrca_tcga_protein_quantification_zscores%2CESR1%2CPGR%2CERBB2&tab_index=tab_visualize)

![image](https://user-images.githubusercontent.com/840895/69502724-7fee2c80-0ee0-11ea-81b7-2ada01fc99a2.png)

* **Enhancement**: global settings for query session

<img src="https://user-images.githubusercontent.com/840895/69502843-b6787700-0ee1-11ea-927b-049635a5a8ac.png" width="300">


## November 7, 2019

*   **Added data** consisting of 212 samples from 3 studies:
    * [Metastatic Melanoma (DFCI, Science 2015)](https://www.cbioportal.org/study/summary?id=skcm_dfci_2015) *110 samples*
    * [Melanoma (MSKCC, NEJM 2014)](https://www.cbioportal.org/study/summary?id=skcm_mskcc_2014) *64 samples*
    * [Metastatic Melanoma (UCLA, Cell 2016)](https://www.cbioportal.org/study/summary?id=mel_ucla_2016) *38 samples*

## October 30, 2019

*   **Added data** consisting of 178 samples from 2 studies:
    * [Intrahepatic Cholangiocarcinoma (Shanghai, Nat Commun 2014)](https://www.cbioportal.org/study/summary?id=ihch_smmu_2014) *103 samples*
    * [Non-Small Cell Lung Cancer (MSK, Cancer Cell 2018)](https://www.cbioportal.org/study/summary?id=nsclc_mskcc_2018) *75 samples*

## October 23, 2019
* **Enhancement**: Quick example links in Plots tab. [example](https://www.cbioportal.org/results/plots?session_id=5ba11ce6498eb8b3d567e7ed)

## October 14, 2019

* **New Feature**: Fusion Genes table in `Study View`. [example](https://www.cbioportal.org/study/summary?id=msk_impact_2017)

![image](https://user-images.githubusercontent.com/840895/68320157-6ef28e00-008d-11ea-8cb3-e4b54aa1d848.png)

## October 11, 2019
* **Enhancement**: The Download interface on the homepage has been removed. Enhanced download functionality is now available after querying on the results page.

  _Home page_:

  ![homepage download tab removed](https://user-images.githubusercontent.com/1334004/66735303-48627e00-ee66-11e9-863f-5ec713ea444a.png)

  _Results page_:

  ![results page download tab](https://user-images.githubusercontent.com/1334004/66736057-7ea0fd00-ee68-11e9-942b-e123b9659126.png)

  Note that as before one can always download the full raw data on the [Data Sets page](https://www.cbioportal.org/datasets) or from [Datahub](https://github.com/cBioPortal/datahub/).

## October 9, 2019
*  **Added data** consisting of  2725 samples from 4 studies:

    * [Cancer Cell Line Encyclopedia (Broad, 2019)](https://www.cbioportal.org/study/summary?id=ccle_broad_2019) *1739 samples*
    * [Chronic Lymphocytic Leukemia (Broad, Nature 2015)](https://www.cbioportal.org/study/summary?id=cll_broad_2015#summary) *537 samples*
    * [Rectal Cancer (MSK,Nature Medicine 2019)](https://www.cbioportal.org/study/summary?id=rectal_msk_2019) *339 samples*
    * [Colon Cancer (CPTAC-2 Prospective, Cell 2019)](https://www.cbioportal.org/study/summary?id=coad_cptac_2019) *110 samples*

*  **Updated**  [Esophageal Carcinoma (TCGA, Nature 2017)](https://www.cbioportal.org/study/summary?id=stes_tcga_pub#summary) with addition of CNA data for Esophageal Squamous Cell Carcinoma cases *90 samples*.

## September 18, 2019
* **New Feature**: The list and order of charts of a study will be automatically saved now as a user preference on the _study view_ page.

## September 6, 2019
*   **Added data** consisting of 1216 samples from 3 studies:
    * [Breast Cancer (MSKCC, 2019)](https://www.cbioportal.org/study/summary?id=brca_mskcc_2019) *70 samples*
    * [Brain Tumor PDXs (Mayo Clinic, 2019)](https://www.cbioportal.org/study/summary?id=gbm_mayo_pdx_sarkaria_2019)*97 samples*
    * [Adenoid Cystic Carcinoma Project (2019)](https://www.cbioportal.org/study/summary?id=acc_2019) *1049 samples*

## August 13, 2019
*   **Added data** consisting of 295 samples from 3 studies:
    *  [Pediatric Preclinical Testing Consortium (PPTC, 2019)](https://www.cbioportal.org/study/summary?id=pptc_2019) *261 samples*
    *  [Non-small cell lung cancer (MSK, Science 2015)](https://www.cbioportal.org/study/summary?id=nsclc_mskcc_2015) *16 samples*
    *  [Prostate Cancer (MSK, 2019)](https://www.cbioportal.org/study/summary?id=prad_msk_2019) *18 samples*

## July 26, 2019
*   **Added data** consisting of 35 samples from 1 study:
    *  [Clear Cell Renal Cell Carcinoma (DFCI, Science 2019)](https://www.cbioportal.org/study/summary?id=ccrcc_dfci_2019) *35 samples*
*   Added Hypoxia data for:
    *  [Breast Invasive Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018)
    *  [Colorectal Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=coadread_tcga_pan_can_atlas_2018)
    *  [Cervical Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=cesc_tcga_pan_can_atlas_2018)
    *  [Glioblastoma Multiforme (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=gbm_tcga_pan_can_atlas_2018)
    *  [Head and Neck Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=hnsc_tcga_pan_can_atlas_2018)
    *  [Kidney Renal Clear Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=kirc_tcga_pan_can_atlas_2018)
    *  [Kidney Renal Papillary Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=kirp_tcga_pan_can_atlas_2018)
    *  [Brain Lower Grade Glioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=lgg_tcga_pan_can_atlas_2018)
    *  [Liver Hepatocellular Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=lihc_tcga_pan_can_atlas_2018)
    *  [Lung Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=luad_tcga_pan_can_atlas_2018)
    *  [Lung Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=lusc_tcga_pan_can_atlas_2018)
    *  [Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018)
    *  [Pancreatic Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=paad_tcga_pan_can_atlas_2018)
    *  [Pheochromocytoma and Paraganglioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=pcpg_tcga_pan_can_atlas_2018)
    *  [Prostate Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=prad_tcga_pan_can_atlas_2018)
    *  [Skin Cutaneous Melanoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=skcm_tcga_pan_can_atlas_2018)
    *  [Thyroid Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=thca_tcga_pan_can_atlas_2018)
    *  [Uterine Corpus Endometrial Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=ucec_tcga_pan_can_atlas_2018)
## July 24, 2019
*   **Added data** consisting of 151 samples from 1 study:
    *  [Myeloproliferative Neoplasms (CIMR, NEJM 2013)](https://www.cbioportal.org/study/summary?id=mpn_cimr_2013) *151 samples*
## July 13, 2019
*   **Public Release 6.1 of AACR Project GENIE:**
    *   The sixth data set, GENIE 6.0-public, was released in early July 2019.  A patch to GENIE 6.0-public, GENIE 6.1-pubic, was subsequently released on July 13, 2019.  The combined data set now includes nearly 70,000 de-identified genomic records collected from patients who were treated at each of the consortium's participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for nearly 80 major cancer types, including data from nearly 11,000 patients with lung cancer, greater than  9,700 patients with breast cancer, and nearly 7,000 patients with colorectal cancer.
*   More detailed information can be found in the [AACR GENIE Data Guide](https://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf).  In addition to accessing the data via the cBioPortal, users can download the data directly from [Sage Bionetworks](https://synapse.org/genie). Users will need to create an account for either site and agree to the [terms of access](https://www.aacr.org/Documents/Terms%20of%20Access.pdf).
*   For frequently asked questions, visit the [AACR FAQ page](https://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).
## July 2, 2019
*   **Added data** consistng of 785 samples from 4 studies:
    *  [Non-Small Cell Lung Cancer (TRACERx, NEJM 2017)](https://www.cbioportal.org/study/summary?id=nsclc_tracerx_2017) *327 samples*
    *  [Acute myeloid leukemia or myelodysplastic syndromes (WashU, 2016)](https://www.cbioportal.org/study/summary?id=mnm_washu_2016) *136 samples*
    *  [Basal Cell Carcinoma (UNIGE, Nat Genet 2016)](https://www.cbioportal.org/study/summary?id=bcc_unige_2016) *293 samples*
    *  [Colon Adenocarcinoma (CaseCCC, PNAS 2015)](https://www.cbioportal.org/study/summary?id=coad_caseccc_2015) *29 samples*

## June 19, 2019
* **New Feature**: Show *Genome Aggregation Database (gnomAD)* population frequencies in the mutations table - see [example](https://bit.ly/2ISHgiu):

  ![gnomad feature news](https://user-images.githubusercontent.com/1334004/59794400-e07c9c00-92a6-11e9-97ea-a79bfc8f3885.gif)

## June 12, 2019
*   **Added data**  of 1350 samples from 3 studies:
    *  [Pheochromocytoma and Paraganglioma (TCGA, Cell 2017)](https://www.cbioportal.org/study?id=pcpg_tcga_pub) *178 samples*
    *  [Metastatic Solid Cancers (UMich, Nature 2017)](https://www.cbioportal.org/study?id=metastatic_solid_tumors_mich_2017) *500 samples*
    *   [Acute Myeloid Leukemia (OHSU, Nature 2018)](https://www.cbioportal.org/study/summary?id=aml_ohsu_2018) *672 samples*
*    Added survival data for TCGA PanCan Atlas Cohorts (>10,000 samples across 33 tumor types).
*    Added hypoxia data for [Bladder Urothelial Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study/summary?id=blca_tcga_pan_can_atlas_2018)
## June 7, 2019
* **New Group Comparison Feature**: Compare clinical and genomic features of user-defined groups of samples/patients. [View Tutorial](https://www.cbioportal.org/tutorials#group-comparison)

[![group-comparison](https://user-images.githubusercontent.com/840895/59052073-40f9eb00-885c-11e9-9ddb-6d036533e5f5.png)](https://www.cbioportal.org/comparison/mutations?sessionId=5cf92520e4b0ab4137874376)


## May 8, 2019
* **New Feature**: Show *Post Translational Modification (PTM)* information from [dbPTM](https://dbptm.mbc.nctu.edu.tw/) on the _Mutation Mapper_ - see [example](https://bit.ly/2VVrWdi):

  ![ptm feature_news](https://user-images.githubusercontent.com/1334004/57391288-34667200-718c-11e9-9fcc-b849542c74b3.gif)

## April 26, 2019
*   **Added data**  of 568 samples from 4 studies:
    *  [Adenoid Cystic Carcinoma (JHU, Cancer Prev Res 2016)](https://www.cbioportal.org/study?id=acyc_jhu_2016) *25 samples*
    *  [Histiocytosis Cobimetinib (MSK, Nature 2019)](https://www.cbioportal.org/study?id=histiocytosis_cobi_msk_2019) *52 samples*
    *  [Upper Tract Urothelial Carcinoma (Cornell/Baylor/MDACC, Nat Comm 2019)](https://www.cbioportal.org/study?id=utuc_cornell_baylor_mdacc_2019) *47 samples*
    *  [Metastatic Prostate Adenocarcinoma (SU2C/PCF Dream Team, PNAS 2019)](https://www.cbioportal.org/study?id=prad_su2c_2019) *444 samples*

## March 29, 2019
* **New Feature**: Use the new quick search tab on the homepage to more easily navigate to a study, gene or patient:

  ![quick_search_news](https://user-images.githubusercontent.com/1334004/55113078-8f4c7a00-50b4-11e9-9d95-e9a6e1dcda52.gif)

## March 15, 2019
*   **Added data**  of 338 samples from 4 studies:
    *  [Adenoid Cystic Carcinoma (MGH, Nat Gen  2016)](https://www.cbioportal.org/study?id=acyc_mgh_2016) *10 samples*
    *   [Gallbladder Cancer (MSK, Cancer 2018)](https://www.cbioportal.org/study?id=gbc_msk_2018) *103 samples*
    *   [The Metastatic Prostate Cancer Project (Provisional, December 2018)](https://www.cbioportal.org/study?id=prad_mpcproject_2018) *19 samples*
    *   [Adult Soft Tissue Sarcomas (TCGA, Cell 2017)](https://www.cbioportal.org/study?id=sarc_tcga_pub) *206 samples*

## February 22, 2019
* **Enhancement**: Exon number and HGVSc annotations are available in optional columns in the Mutations tab on the _Results_ page and in the _Patient View_.
* **New feature**: option to a show regression line in the scatter plot in the Plots tab on the _Results_ page

![image](https://user-images.githubusercontent.com/840895/53193265-99271d00-35de-11e9-8ffc-352afc2ba906.png)

## February 19, 2019
* **New feature**: *Copy-Number Segments tab* on the _Study View_ page using [igv.js v2](https://github.com/igvteam/igv.js) - see [example](https://www.cbioportal.org/study?id=ucec_tcga_pub&tab=cnSegments)
* Improved *Copy-Number Segments* tab on the _Results_ page
* **New feature**: *OncoKB* and *Cancer Hotspots tracks* in the Mutations tab on the _Results_ page

![image](https://user-images.githubusercontent.com/840895/53054860-d1f6b300-3473-11e9-8178-1df10eed36fc.png)


## January 24, 2019
*   **Added data**  of 2328 samples from 8 studies:
    *   [Uveal Melanoma (QIMR, Oncotarget 2016)](https://www.cbioportal.org/study?id=um_qimr_2016#summary) *28 samples*
    *   [Squamous Cell Carcinoma of the Vulva (CUK, Exp Mol Med 2018)](https://www.cbioportal.org/study?id=vsc_cuk_2018) *15 samples*
    *   [TMB and Immunotherapy (MSKCC, Nat Genet 2019)](https://www.cbioportal.org/study?id=tmb_mskcc_2018#summary) *1661 samples*
    *   [Glioma (MSK, 2018)](https://www.cbioportal.org/study?id=glioma_msk_2018) *91 samples*
    *   [Urothelial Carcinoma (Cornell/Trento, Nat Gen 2016)](https://www.cbioportal.org/study?id=blca_cornell_2016) *72 samples*
    *   [Hepatocellular Carcinoma (MSK, Clin Cancer Res 2018)](https://www.cbioportal.org/study?id=hcc_mskimpact_2018) *127 samples*
    *   [MSK Thoracic PDX (MSK, Provisional)](https://www.cbioportal.org/study?id=lung_msk_pdx) *139 samples*
    *   [Cholangiocarcinoma (MSK, Clin Cancer Res 2018)](https://www.cbioportal.org/study?id=chol_msk_2018) *195 samples*
*  **Updated data** for [The Metastatic Breast Cancer Project (Provisional, October 2018)](https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017) *237 samples*
## January 10, 2019
* cBioPortal now supports queries for driver mutations, fusions and copy number alterations as well as germline/somatic mutations using [Onco Query Language (OQL)](https://www.cbioportal.org/oql) -- see [example](https://www.cbioportal.org/results/oncoprint?session_id=5c23ea81e4b05228701f9d44)
* A new [tutorial](https://www.cbioportal.org/tutorials) explores OQL and provides examples of how OQL can be a powerful tool to refine queries.


## December 17, 2018
* The 10th phase of cBioPortal architectural upgrade is now complete: the _Study View_ has been moved to the new architecture with numerous improvements. This marks the completion of the cBioPortal architectural refactoring! üéâüéâüéâ

[![image](https://user-images.githubusercontent.com/840895/50120599-11aadf80-0224-11e9-9f9f-adae0ed42c72.png)](https://www.cbioportal.org/study?id=msk_impact_2017)

## October 29, 2018
* The ninth phase of the cBioPortal architectural upgrade is now complete: the _results page_ is now a single-page application with better performance.
* Supported plotting mutations by type in _Plots_ tab

![image](https://user-images.githubusercontent.com/840895/47598695-d996c700-d96d-11e8-9371-9e591cf25ea3.png)

## October 19, 2018

* Support selection of transcript of interest in the [MutationMapper tool](https://www.cbioportal.org/mutation_mapper.jsp) via [Genome Nexus](https://www.genomenexus.org).

![mutation_mapper_dropdown](https://user-images.githubusercontent.com/1334004/47240017-f99a1980-d3b4-11e8-99a4-e8c73b25ecec.png)

## October 17, 2018
*   **Added data**  of 3578 samples from 8 studies:
    *   [Rhabdoid Cancer (BCGSC, Cancer Cell 2016)](https://www.cbioportal.org/study?id=mrt_bcgsc_2016#summary) *40 samples*
    *   [Diffuse Large B-Cell Lymphoma (Duke, Cell 2017)](https://www.cbioportal.org/study?id=dlbcl_duke_2017#summary) *1001 samples*
    *   [Diffuse Large B cell Lymphoma (DFCI, Nat Med 2018)](https://www.cbioportal.org/study?id=dlbcl_dfci_2018#summary) *135 samples*
    *   [Breast Fibroepithelial Tumors (Duke-NUS, Nat Genet 2015)](https://www.cbioportal.org/study?id=bfn_duke_nus_2015#summary) *22 samples*
    *   [Uterine Clear Cell Carcinoma (NIH, Cancer 2017)](https://www.cbioportal.org/study?id=uccc_nih_2017#summary) *16 samples*
    *   [Endometrial Cancer (MSK, 2018)](https://www.cbioportal.org/study?id=ucec_msk_2018#summary) *197 samples*
    *   [Breast Cancer (MSK, Cancer Cell 2018)](https://www.cbioportal.org/study?id=breast_msk_2018#summary) *1918 samples*
    *   [MSS Mixed Solid Tumors (Van Allen, 2018)](https://www.cbioportal.org/study?id=mixed_allen_2018#summary) *249 samples*
*   Updated data for [The Angiosarcoma Project (Provisional, September 2018)](https://www.cbioportal.org/study?id=angs_project_painter_2018#summary) *48 samples*

## August 20, 2018
*   Now you can log in on the public cBioPortal with your Google account and save your virtual studies for quick analysis.

![image](https://user-images.githubusercontent.com/840895/44370286-902a8700-a4a7-11e8-9c8f-dfda87fbd66b.png)

## August 7, 2018
*   The eighth phase of the cBioPortal architectural upgrade is now complete: The _Plots_, _Expression_, _Network_, and _Bookmarks_ tabs, and therefore all analysis tabs in the _results page_, have been moved to the new architecture.
*   Updated the [MutationMapper tool](https://www.cbioportal.org/mutation_mapper.jsp), now connecting to [Genome Nexus](https://genomenexus.org/) for annotating mutations on the fly.
*   _Total Mutations_ and _Fraction Genome Altered_ are now available in _Plots_ tab for visualization and analysis.
*   Enhanced clinical attribute selector for OncoPrint, now showing sample counts per attribute.

![image](https://user-images.githubusercontent.com/840895/43609786-ec946d64-9672-11e8-8218-4e63bd5007e8.png)


## July 27, 2018
*   **Added data**  of 2787 samples from 10 studies:
    *   [Mixed Tumors (PIP-Seq 2017)](https://www.cbioportal.org/study?id=mixed_pipseq_2017#summary) *103 samples*
    *   [Nonmuscle Invasive Bladder Cancer (MSK Eur Urol 2017)](https://www.cbioportal.org/study?id=blca_nmibc_2017#summary) *105 samples*
    *   [Pediatric Neuroblastoma (TARGET, 2018)](https://www.cbioportal.org/study?id=nbl_target_2018_pub) *1089 samples*
    *   [Pediatric Pan-Cancer (DKFZ - German Cancer Consortium, 2017)](https://www.cbioportal.org/study?id=pediatric_dkfz_2017) *961 samples*
    *   [Skin Cutaneous Melanoma (Broad, Cancer Discov 2014)](https://www.cbioportal.org/study?id=skcm_broad_brafresist_2012) *78 samples*
    *   [Cutaneous Squamous Cell Carcinoma (MD Anderson, Clin Cancer Res 2014)](https://www.cbioportal.org/study?id=cscc_hgsc_bcm_2014#summary) *39 samples*
    *   [Diffuse Large B-cell Lymphoma (BCGSC, Blood 2013)](https://www.cbioportal.org/study?id=nhl_bcgsc_2013#summary) *53 samples*
    *   [Non-Hodgkin Lymphoma (BCGSC, Nature 2011)](https://www.cbioportal.org/study?id=nhl_bcgsc_2011#summary) *14 samples*
    *   [Chronic lymphocytic leukemia (ICGA, Nat 2011)](https://www.cbioportal.org/study?id=cllsll_icgc_2011#summary) *105 samples*
    *   [Neuroblastoma (Broad Institute 2013)](https://www.cbioportal.org/study?id=nbl_broad_2013#summary) *240 samples*

## June 20, 2018
*   The seventh phase of the cBioPortal architectural upgrade is now complete: The _Enrichments_ and _Co-Expression_ tabs have been moved to the new architecture.
*   Supported merged gene tracks in OncoPrint and [Onco Query Language](https://www.cbioportal.org/onco_query_lang_desc.jsp) -- see [example](https://www.cbioportal.org/index.do?session_id=5b2bedca498eb8b3d566ab60)

![image](https://user-images.githubusercontent.com/840895/41738366-def14850-755f-11e8-8a7c-deb789dc03da.png)

## May 10, 2018
*   Enhanced OncoPrint to show germline mutations -- see [example](https://www.cbioportal.org/index.do?session_id=5af5a304498eb8b3d56615b6)

![image](https://user-images.githubusercontent.com/840895/39926636-c3304d8c-54fd-11e8-9cb7-76430c8772d7.png)

## April 17, 2018
*   **Added data**  of 3732 samples from 4 TARGET studies:
    *   [Pediatric Acute Lymphoid Leukemia - Phase II (TARGET, 2018)](https://www.cbioportal.org/study?id=all_phase2_target_2018_pub#summary) *1978 samples*
    *   [Pediatric Acute Myeloid Leukemia (TARGET, 2018)](https://www.cbioportal.org/study?id=aml_target_2018_pub#summary) *1025 samples*
    *   [Pediatric Rhabdoid Tumor (TARGET, 2018)](https://www.cbioportal.org/study?id=rt_target_2018_pub#summary) *72 samples*
    *   [Pediatric Wilms' Tumor (TARGET, 2018)](https://www.cbioportal.org/study?id=wt_target_2018_pub#summary) *657 samples*
*   **Added data**  of 3416 samples from 10 published studies:
    *    [Prostate Adenocarcinoma (MSKCC/DFCI, Nature Genetics 2018)](https://www.cbioportal.org/study?id=prad_p1000#summary) *1013 samples*
    *    [Prostate Adenocarcinoma (EurUrol, 2017)](https://www.cbioportal.org/study?id=prad_eururol_2017#summary) *65 samples*
    *    [Non-Small Cell Lung Cancer (MSK, JCO 2018)](https://www.cbioportal.org/study?id=nsclc_pd1_msk_2018#summary) *240 samples*
    *    [Small-Cell Lung Cancer (Multi-Institute 2017)](https://www.cbioportal.org/study?id=sclc_cancercell_gardner_2017#summary) *20 samples*
    *    [The Angiosarcoma Project (Provisional, February 2018)](https://www.cbioportal.org/study?id=angs_project_painter_2018#summary) *14 samples*
    *    [Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2016)](https://www.cbioportal.org/study?id=all_stjude_2016#summary) *73 samples*
    *    [Updated Segment data and Allele Frequencies for The Metastatic Breast Cancer Project (Provisional, October 2017)](https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017#summary) *103 samples*
    *    [Colorectal Cancer (MSK, Cancer Cell 2018)](https://www.cbioportal.org/study?id=crc_msk_2018#summary) *1134 samples*
    *    [Metastatic Esophagogastric Cancer (MSK,Cancer Discovery 2017)](https://www.cbioportal.org/study?id=egc_msk_2017#summary) *341 samples*
    * [Bladder Cancer (TCGA, Cell 2017)](https://www.cbioportal.org/study?id=blca_tcga_pub_2017#summary) *413 samples*

## April 5, 2018
*   **Added data** from the TCGA PanCanAtlas project with >10,000 samples from 33 tumor types:
    *   [Adrenocortical Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=acc_tcga_pan_can_atlas_2018)
    *   [Bladder Urothelial Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=blca_tcga_pan_can_atlas_2018)
    *   [Breast Invasive Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=brca_tcga_pan_can_atlas_2018)
    *   [Cervical Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=cesc_tcga_pan_can_atlas_2018)
    *   [Colon Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=coad_tcga_pan_can_atlas_2018)
    *   [Cholangiocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=chol_tcga_pan_can_atlas_2018)
    *   [Diffuse Large B-Cell Lymphoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=dlbc_tcga_pan_can_atlas_2018)
    *   [Esophageal Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=esca_tcga_pan_can_atlas_2018)
    *   [Glioblastoma Multiforme (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=gbm_tcga_pan_can_atlas_2018)
    *   [Head and Neck Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=hnsc_tcga_pan_can_atlas_2018)
    *   [Kidney Chromophobe (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=kich_tcga_pan_can_atlas_2018)
    *   [Kidney Renal Clear Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=kirc_tcga_pan_can_atlas_2018)
    *   [Kidney Renal Papillary Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=kirp_tcga_pan_can_atlas_2018)
    *   [Acute Myeloid Leukemia (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=laml_tcga_pan_can_atlas_2018)
    *   [Brain Lower Grade Glioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=lgg_tcga_pan_can_atlas_2018)
    *   [Liver Hepatocellular Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=lihc_tcga_pan_can_atlas_2018)
    *   [Lung Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=luad_tcga_pan_can_atlas_2018)
    *   [Lung Squamous Cell Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=lusc_tcga_pan_can_atlas_2018)
    *   [Mesothelioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=meso_tcga_pan_can_atlas_2018)
    *   [Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=ov_tcga_pan_can_atlas_2018)
    *   [Pancreatic Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=paad_tcga_pan_can_atlas_2018)
    *   [Pheochromocytoma and Paraganglioma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=pcpg_tcga_pan_can_atlas_2018)
    *   [Prostate Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=prad_tcga_pan_can_atlas_2018)
    *   [Rectum Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=read_tcga_pan_can_atlas_2018)
    *   [Sarcoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=sarc_tcga_pan_can_atlas_2018)
    *   [Skin Cutaneous Melanoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=skcm_tcga_pan_can_atlas_2018)
    *   [Stomach Adenocarcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=stad_tcga_pan_can_atlas_2018)
    *   [Testicular Germ Cell Tumors (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=tgct_tcga_pan_can_atlas_2018)
    *   [Thyroid Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=thca_tcga_pan_can_atlas_2018)
    *   [Thymoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=thym_tcga_pan_can_atlas_2018)
    *   [Uterine Corpus Endometrial Carcinoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=ucec_tcga_pan_can_atlas_2018)
    *   [Uterine Carcinosarcoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=ucs_tcga_pan_can_atlas_2018)
    *   [Uveal Melanoma (TCGA, PanCancer Atlas)](https://www.cbioportal.org/study?id=uvm_tcga_pan_can_atlas_2018)

## March 20, 2018
*   The sixth phase of the cBioPortal architectural upgrade is now complete: The _Download_ tab has been moved to the new architecture.
*   Data can now be downloaded in [tabular format](https://www.thehyve.nl/articles/download-data-from-cbioportal-oncoprint) from OncoPrint.
*   Added an option to download an SVG file on the _Cancer Type Summary_ tab.

## January 15, 2018
*   The fifth phase of the cBioPortal architectural upgrade is now complete: The OncoPrint and Survival tabs have been moved to the new architecture.

## November 20, 2017
*   You can now combine multiple studies and view them on the study summary page. [Example: liver cancer studies](https://www.cbioportal.org/study?id=liad_inserm_fr_2014%2Chcc_inserm_fr_2015%2Clihc_amc_prv%2Clihc_riken%2Clihc_tcga)
*   You can now bookmark or share your selected samples as virtual studies with the _share_ icon on the study summary page. [Example: a virtual study of breast tumors](https://www.cbioportal.org/study?id=5a12fd57498eb8b3d5605cd4)
*   Cross-study query reimplemented: Now you can view an OncoPrint of multiple studies. [Example: querying NSCLC tumors from 5 studies](https://www.cbioportal.org/index.do?session_id=5a135bee498eb8b3d5605d01)

![image](https://user-images.githubusercontent.com/840895/33045546-3c8a636c-ce1b-11e7-9f52-7060c89a8dfd.png)

## October 17, 2017
*   The fourth phase of the cBioPortal architectural upgrade is now complete: The _Mutual Exclusivity_ and _Cancer Type Summary_ tabs have been moved to the new architecture.
*   Updated protein structure alignment data in Mutations tab are now retrieved from Genome Nexus via the [G2S web service](https://g2s.genomenexus.org/).

## October 2, 2017
*   **Added data** of 1646 samples from 7 published studies:
    *   [NGS in Anaplastic Oligodendroglioma and Anaplastic Oligoastrocytomas tumors (MSK, Neuro Oncol 2017)](https://www.cbioportal.org/study?id=odg_msk_2017#summary)   *22 samples*
    *   [MSK-IMPACT Clinical Sequencing Cohort for Non-Small Cell Cancer (MSK, Cancer Discovery 2017)](https://www.cbioportal.org/study?id=lung_msk_2017) *915 samples*
    *   [Paired-exome sequencing of acral melanoma (TGEN, Genome Res 2017)](https://www.cbioportal.org/study?id=mel_tsam_liang_2017#summary) *38 samples*
    *   [MSK-IMPACT Clinical Sequencing Cohort in Prostate Cancer (MSK, JCO Precision Oncology 2017)](https://www.cbioportal.org/study?id=prad_mskcc_2017#summary) *504 samples*
    *   [Whole-exome sequences (WES) of pretreatment melanoma tumors (UCLA, Cell 2016)](https://www.cbioportal.org/study?id=skcm_ucla_2016#summary) *39 samples*
    *   [Next generation sequencing (NGS) of pre-treatment metastatic melanoma samples (MSK, JCO Precision Oncology 2017)](https://www.cbioportal.org/study?id=skcm_vanderbilt_mskcc_2015#summary) *66 samples*
    *   [Targeted gene sequencing in 62 high-grade primary Unclassified Renal Cell Carcinoma (MSK, Nature 2016)](https://www.cbioportal.org/study?id=urcc_mskcc_2016#summary) *62 samples*
*   **Updated data** for [MSK-IMPACT Clinical Sequencing Cohort (MSK, Nat Med 2017)](https://www.cbioportal.org/study?id=msk_impact_2017) with overall survival data.
## August 3, 2017
*   The third phase of the cBioPortal architectural upgrade is now complete: The _Mutations_ tab now has a fresh look and faster performance -- see [example](https://www.cbioportal.org/index.do?session_id=598386e5498e5df2e29376ab&show_samples=false&#mutation_details&)

![image](https://user-images.githubusercontent.com/840895/28942244-9ea43bee-7868-11e7-9b24-0aaf0f9c010d.png)

*   Variant interpretations from the [CIViC database](https://civic.genome.wustl.edu) are now integrated into the annotation columns on the Mutations tab and in the patient view pages
*   New summary graph for all cancer studies and samples on the front page


## June 26, 2017
*   The second phase of the cBioPortal architectural upgrade is now complete: The query interface now has a fresh look and faster performance.

![image](https://user-images.githubusercontent.com/840895/27399281-3ba5357a-568a-11e7-9730-d230285d7805.png)

## May 12, 2017
*   **Added data** of 12,211 samples from 11 published studies:
    *   [MSK-IMPACT Clinical Sequencing Cohort (MSK, Nat Med 2017)](https://www.cbioportal.org/study?id=msk_impact_2017)   *10,945 samples*
    *   [Whole-genome sequencing of pilocytic astrocytomasatic (DKFZ, Nat Genetics, 2013)](https://www.cbioportal.org/study?id=past_dkfz_heidelberg_2013) *96 samples*
    *   [Hepatocellular Carcinomas (INSERM, Nat Genet 2015)](https://www.cbioportal.org/study?id=hcc_inserm_fr_2015) *243 samples*
    *   [Cystic Tumor of the Pancreas (Johns Hopkins, PNAS 2011)](https://www.cbioportal.org/study?id=pact_jhu_2011) *32 samples*
    *   [Whole-Genome Sequencing of Pancreatic Neuroendocrine Tumors (ARC- Net, Nature, 2017)](https://www.cbioportal.org/study?id=panet_arcnet_2017) *98 samples*
    *   [Medulloblastoma (Sickkids, Nature 2016)](https://www.cbioportal.org/study?id=mbl_sickkids_2016) *46 samples*
    *   [Genetic Characterization of NSCLC young adult patients (University of Turin, Lung Cancer 2016)](https://www.cbioportal.org/study?id=nsclc_unito_2016) *41 samples*
    *   [Genomic Profile of Patients with Advanced Germ Cell Tumors (MSK, JCO 2016).](https://www.cbioportal.org/study?id=gct_msk_2016) *180 samples*
    *   [Ampullary Carcinoma (Baylor, Cell Reports 2016)](https://www.cbioportal.org/study?id=ampca_bcm_2016) *160 samples*
    *   [Mutational profiles of metastatic breast cancer (INSERM, 2016)](https://www.cbioportal.org/study?id=brca_igr_2015) *216 samples*
    *   [Prostate Adenocarcinoma (Fred Hutchinson CRC, Nat Med 2016)](https://www.cbioportal.org/study?id=prad_fhcrc) *154 samples*


## May 5, 2017
*   First phase of cBioPortal architectural upgrade complete: Patient view now has fresh look and faster performance. [example](https://bit.ly/2pNY961)

## March 28, 2017
*   **New features**:
    * Per-sample mutation spectra are now available in OncoPrints -- see [example](https://www.cbioportal.org/index.do?cancer_study_list=ucec_tcga_pub&cancer_study_id=ucec_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ucec_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ucec_tcga_pub_gistic&Z_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=ucec_tcga_pub_manuscript&case_ids=&patient_case_select=sample&gene_set_choice=user-defined-list&gene_list=POLE%0D%0AERBB2%0D%0AKRAS%0D%0ACTNNB1&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit&show_samples=false&clinicallist=SUBTYPE,NO_CONTEXT_MUTATION_SIGNATURE,%23%20mutations&)

    ![image](https://cloud.githubusercontent.com/assets/840895/24209131/d34e33b2-0efb-11e7-945a-fd2b3c66d195.png)

    * mRNA heat map clustering is now supported in OncoPrints
    * MDACC Next-Generation Clustered Heat Maps are now available in the patient view
    * cBioPortal web site style change


## Feburary 2, 2017
*   **New features**:
    * 3D hotspot mutation annotations are now available from 3dhotspots.org
*   **New data**:
    * CPTAC proteomics data have been integrated for TCGA breast, ovarian, and colorectal provisional studies


## December 23, 2016
*   **New features**:
    * Heat map visualization of gene expression data in the OncoPrint

    ![OncoPrint Heatmap](https://cloud.githubusercontent.com/assets/840895/21407479/81f8c008-c79e-11e6-9b82-3f4eb8f8e1ea.png)

    * Heat map visualization of gene expression data in the Study View page connecting to MDACC's TCGA Next-Generation Clustered Heat Map Compendium

## October 7, 2016
*   **New features**:
    * All data sets can now be downloaded as flat files from the new [Data Hub](https://github.com/cBioPortal/datahub)
    * Annotation of putative driver missense mutations in OncoPrints, based on [OncoKB](https://www.oncokb.org), mutation hotspots, and recurrence in cBioPortal and COSMIC

    ![OncoPrint-OncoKB](https://cloud.githubusercontent.com/assets/840895/19208804/094dd320-8ccd-11e6-8012-f30104b62ff4.png)

    * Copy number segments visualization directly in the browser in a new *CN Segments* tab via [IGV.js](https://igv.org/)

    ![image](https://cloud.githubusercontent.com/assets/840895/19200747/ad03c824-8c98-11e6-8051-91fc40e1dc56.png)

*   **Improvements**:
    * Improved cancer study view page (bug fixes and increased performance)

## July 24, 2016
*   **Added data** of 4,375 samples from 21 published studies:
    *   [Adenoid Cystic Carcinoma (MDA, Clin Cancer Res 2015)](https://www.cbioportal.org/study?id=acyc_mda_2015) *102 samples*
    *   [Adenoid Cystic Carcinoma (FMI, Am J Surg Pathl. 2014)](https://www.cbioportal.org/study?id=acyc_fmi_2014) *28 samples*
    *   [Adenoid Cystic Carcinoma (Sanger/MDA, JCI 2013)](https://www.cbioportal.org/study?id=acyc_sanger_2013) *24 samples*
    *   [Adenoid Cystic Carcinoma of the Breast (MSKCC, J Pathol. 2015)](https://www.cbioportal.org/study?id=acbc_mskcc_2015) *12 samples*
    *   [Bladder Cancer, Plasmacytoid Variant (MSKCC, Nat Genet 2016)](https://www.cbioportal.org/study?id=blca_plasmacytoid_mskcc_2016) *34 samples*
    *   [Breast Cancer (METABRIC, Nat Commun 2016)](https://www.cbioportal.org/study?id=brca_metabric) *1980 samples*
    *   [Chronic Lymphocytic Leukemia (Broad, Cell 2013)](https://www.cbioportal.org/study?id=lcll_broad_2013) *160 samples*
    *   [Chronic Lymphocytic Leukemia (IUOPA, Nature 2015)](https://www.cbioportal.org/study?id=cll_iuopa_2015) *506 samples*
    *   [Colorectal Adenocarcinoma (DFCI, Cell Reports 2016)](https://www.cbioportal.org/study?id=coadread_dfci_2016) *619 samples*
    *   [Cutaneous T Cell Lymphoma (Columbia U, Nat Genet 2015)](https://www.cbioportal.org/study?id=ctcl_columbia_2015) *42 samples*
    *   [Diffuse Large B-Cell Lymphoma (Broad, PNAS 2012)](https://www.cbioportal.org/study?id=dlbc_broad_2012) *58 samples*
    *   [Hepatocellular Adenoma (Inserm, Cancer Cell 2014)](https://www.cbioportal.org/study?id=liad_inserm_fr_2014) *46 samples*
    *   [Hypodiploid Acute Lymphoid Leukemia (St Jude, Nat Genet 2013)](https://www.cbioportal.org/study?id=all_stjude_2013) *44 samples*
    *   [Insulinoma (Shanghai, Nat Commun 2013)](https://www.cbioportal.org/study?id=panet_shanghai_2013) *10 samples*
    *   [Malignant Pleural Mesothelioma (NYU, Cancer Res 2015)](https://www.cbioportal.org/study?id=plmeso_nyu_2015) *22 samples*
    *   [Mantle Cell Lymphoma (IDIBIPS, PNAS 2013)](https://www.cbioportal.org/study?id=mcl_idibips_2013) *29 samples*
    *   [Myelodysplasia (Tokyo, Nature 2011)](https://www.cbioportal.org/study?id=mds_tokyo_2011) *29 samples*
    *   [Neuroblastoma (Broad, Nat Genet 2013)](https://www.cbioportal.org/study?id=nbl_ucologne_2015) *56 samples*
    *   [Oral Squamous Cell Carcinoma (MD Anderson, Cancer Discov 2013)](https://www.cbioportal.org/study?id=hnsc_mdanderson_2013) *40 samples*
    *   [Pancreatic Adenocarcinoma (QCMG, Nature 2016)](https://www.cbioportal.org/study?id=paad_qcmg_uq_2016) *383 samples*
    *   [Recurrent and Metastatic Head & Neck Cancer (JAMA Oncology, 2016)](https://www.cbioportal.org/study?id=hnc_mskcc_2016) *151 samples*
*   **New TCGA study**:
    *   [Pan-Lung Cancer (TCGA, Nat Genet 2016)](https://www.cbioportal.org/study?id=nsclc_tcga_broad_2016) *1144 samples*
* Updated **TCGA provisional studies**
    *   updated to the Firehose run of January 28, 2016
    *   RPPA data updated with the latest data from MD Anderson
    *   [OncoTree](http://oncotree.mskcc.org/) codes assigned per sample

## June 6, 2016
*   **New features**:
    * Annotation of mutation effect and drug sensitivity on the Mutations tab and the patient view pages (via [OncoKB](https://www.oncokb.org))  
      ![oncokb-screenshot](https://cloud.githubusercontent.com/assets/840895/15825344/54bbbc44-2bd1-11e6-8b65-e265e03f453a.png)
*   **Improvements**:
    * Improved OncoPrint visualization using WebGL: faster, more zooming flexibility, visualization of recurrent variants
    * Improved Network tab with SBGN view for a single interaction
    * Performance improvement of tables in the study view page
    * Mutation type summary on the Mutations tab

## March 31, 2016
*   **New features**:
    *   Visualization of "Enrichments Analysis" results via volcano plots
    *   Improved performance of the cross cancer expression view  by switching to Plot.ly graphs
    *   Improvements to the "Clinical Data" tab on the study view page
    *   More customization options for the cross-cancer histograms
    *   Performance improvements in the study view and query result tabs
*   **Added data** of 1235 samples from 3 published studies:
    *   [Merged Cohort of LGG and GBM (TCGA, 2016)](https://www.cbioportal.org/study?id=lgggbm_tcga_pub)
    *   [Lung Adenocarcinoma (MSKCC, 2015)](https://www.cbioportal.org/study?id=luad_mskcc_2015)
    *   [Poorly-Differentiated and Anaplastic Thyroid Cancers (MSKCC, JCI 2016)](https://www.cbioportal.org/study?id=thyroid_mskcc_2016)

## January 12, 2016
*   **New features**:
    *   Visualization of multiple samples in a patient
    *   Visualization of timeline data of a patient ([example](https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&caseId=P04))<br/>
        ![timeline-example](https://cloud.githubusercontent.com/assets/840895/12055606/cca26160-aefc-11e5-93f9-2ecfe7e95caf.png)
*   All **TCGA data** updated to the latest Firehose run of August 21, 2015
*   **New TCGA studies**:
    *   [Cholangiocarcinoma (TCGA, Provisional)](https://www.cbioportal.org/study?id=chol_tcga)
    *   [Mesothelioma (TCGA, Provisional)](https://www.cbioportal.org/study?id=meso_tcga)
    *   [Testicular Germ Cell Cancer (TCGA, Provisional)](https://www.cbioportal.org/study?id=tgct_tcga)
    *   [Thymoma (TCGA, Provisional)](https://www.cbioportal.org/study?id=thym_tcga)
*   **Added data** of 650 samples from 10 published studies:
    *   [Neuroblastoma (AMC Amsterdam, Nature 2012)](https://www.cbioportal.org/study?id=nbl_amc_2012)
    *   [Clear Cell Renal Cell Carcinoma (U Tokyo, Nat Genet 2013)](https://www.cbioportal.org/study?id=ccrcc_utokyo_2013)
    *   [Multiregion Sequencing of Clear Cell Renal Cell Carcinoma (IRC, Nat Genet 2014)](https://www.cbioportal.org/study?id=ccrcc_irc_2014)
    *   [Bladder Urothelial Carcinoma (Dana Farber & MSKCC, Cancer Discovery 2014)](https://www.cbioportal.org/study?id=blca_dfarber_mskcc_2014)
    *   [Low-Grade Gliomas (UCSF, Science 2014)	](https://www.cbioportal.org/study?id=lgg_ucsf_2014)
    *   [Esophageal Squamous Cell Carcinoma (UCLA, Nat Genet 2014)](https://www.cbioportal.org/study?id=escc_ucla_2014)
    *   [Acinar Cell Carcinoma of the Pancreas (Johns Hopkins, J Pathol 2014)](https://www.cbioportal.org/study?id=paac_jhu_2014)
    *   [Gastric Adenocarcinoma (TMUCIH, PNAS 2015)](https://www.cbioportal.org/study?id=egc_tmucih_2015)
    *   [Primary Central Nervous System Lymphoma (Mayo Clinic, Clin Cancer Res 2015)](https://www.cbioportal.org/study?id=pcnsl_mayo_2015)
    *   [Desmoplastic Melanoma (Broad Institute, Nat Genet 2015)](https://www.cbioportal.org/study?id=desm_broad_2015)
*   All mutation data mapped to [UniProt canonical isoforms](https://www.uniprot.org/help/canonical_and_isoforms)

## December 23, 2015
*   **New features**:
    *   Visualization of RNA-seq expression levels across TCGA studies (cross-cancer queries)<br/>
        ![cross cancer expression](https://cloud.githubusercontent.com/assets/840895/11821643/8c9a6de4-a338-11e5-83d1-a9e92cb6dfb6.png)
    *   Selection of genes in the study view to initiate queries<br/>
        ![query gene in study view](https://cloud.githubusercontent.com/assets/840895/11977489/5af87ca8-a951-11e5-87ad-5c31b0451cc1.png)
*   **Improvement**:
    *   3-D structures in the "Mutations" tab are now rendered by 3Dmol.js (previously JSmol)
    *   Improved performance by code optimization and compressing large data by gzip


## December 1, 2015
*   **New feature**: Annotated statistically recurrent hotspots, via new algorithm by [Chang et al. 2015](https://www.nature.com/nbt/journal/vaop/ncurrent/full/nbt.3391.html)</br>
    ![Annotate recurrent hotspots](https://cloud.githubusercontent.com/assets/840895/11794851/7729839e-a281-11e5-9413-12dc885b947d.png)


## November 9, 2015
*   **New features**:
    *   Links to MyCancerGenome.org for mutations<br/>
        ![Link to MyCancerGenome.org](https://cloud.githubusercontent.com/assets/11892704/11050295/9745f78c-8712-11e5-8690-a9a04c4455fb.png)
    *   Improved display of selection samples on the study view page
*   **Improvements**:
    *   "Enrichments" analysis is now run across all genes
    *   The "Network" tab is now using Cytoscape.js (Adobe Flash is no longer required)


## October 6, 2015

*   **New TCGA data**:
    *   [Breast Invasive Carcinoma (TCGA, Cell 2015)](https://www.cbioportal.org/study?id=brca_tcga_pub2015)
    *   [Prostate Adenocarcinoma (TCGA, in press)](https://www.cbioportal.org/study?id=prad_tcga_pub)
    *   [Uveal Melanoma (TCGA, Provisional)](https://www.cbioportal.org/study?id=uvm_tcga)
*   **Added data** of 763 samples from 12 published studies:
    *   [Small Cell Lung Cancer (U Cologne, Nature 2015)](https://www.cbioportal.org/study?id=sclc_ucologne_2015)
    *   [Uterine Carcinosarcoma (JHU, Nat Commun 2014)](https://www.cbioportal.org/study?id=ucs_jhu_2014)
    *   [Microdissected Pancreatic Cancer Whole Exome Sequencing (UTSW, Nat Commun 2015)](https://www.cbioportal.org/study?id=paad_utsw_2015)
    *   [Pancreatic Neuroendocrine Tumors (JHU, Science 2011)](https://www.cbioportal.org/https://www.cbioportal.org/study?id=panet_jhu_2011)
    *   [Renal Non-Clear Cell Carcinoma (Genentech, Nat Genet 2014)](https://www.cbioportal.org/study?id=nccrcc_genentech_2014)
    *   [Infant MLL-Rearranged Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2015)](https://www.cbioportal.org/study?id=all_stjude_2015)
    *   [Rhabdomyosarcoma (NIH, Cancer Discov 2014)](https://www.cbioportal.org/study?id=rms_nih_2014)
    *   [Thymic epithelial tumors (NCI, Nat Genet 2014)](https://www.cbioportal.org/study?id=tet_nci_2014)
    *   [Pediatric Ewing Sarcoma (DFCI, Cancer Discov 2014)](https://www.cbioportal.org/study?id=es_dfarber_broad_2014)
    *   [Ewing Sarcoma (Institut Cuire, Cancer Discov 2014)](https://www.cbioportal.org/study?id=es_iocurie_2014)
    *   [Cutaneous squamous cell carcinoma (DFCI, Clin Cancer Res 2015)](https://www.cbioportal.org/study?id=cscc_dfarber_2015)
    *   [Gallbladder Carcinoma (Shanghai, Nat Genet 2014)](https://www.cbioportal.org/study?id=gbc_shanghai_2014)

## August 21, 2015

*   All **TCGA data** updated to the Firehose run of April 16, 2015.
*   **New feature**: Enrichments Analysis finds alterations that are enriched in either altered or unaltered samples.
*   **Improvement**: improved OncoPrint with better performance.

## June 3, 2015

*   **Improvements**:
    *   Allowed downloading data in each chart/table in study summary page.
    *   Added log-rank test _p_-values to the survival plots in study summary page.
    *   Improved visualization of patient clinical data in patient-centric view.
    *   Added option to merge multiple samples for the same patient in OncoPrint.

## April 28, 2015

*   **New features**:
    *   Redesigned query interface to allow selecting multiple cancer studies
    *   Redesigned **Plots** tab

## January 20, 2015

*   All **TCGA data** updated to the Firehose run of October 17, 2014
*   **COSMIC data** updated to V71
*   **New features**:
    *   Query page: better search functions to find cancer studies
    *   OncoPrints now support color coding of different mutation types
    *   OncoPrints now support multiple clinical annotation tracks
    *   [**OncoPrinter tool**](https://www.cbioportal.org/oncoprinter.jsp) now supports mRNA expression changes  
        ![Oncoprint with multiple clinical tracks](https://user-images.githubusercontent.com/1334004/47188323-9f8e4b00-d305-11e8-8fe9-972104518545.png)

## January 6, 2015

*   **New feature**: You can now view **frequencies of mutations and copy-number alterations** in the study view. These tables are updated dynamically when selecting subsets of samples.  
    ![Alterations in heavily copy-number altered endometrial cancer cases](https://user-images.githubusercontent.com/1334004/47188333-9f8e4b00-d305-11e8-8c64-de5cbda0fa54.png)

## December 9, 2014

*   **New TCGA data**:
    *   Added complete and up-to-date **clinical data** for all **TCGA** provisional studies
    *   All TCGA data updated to the Firehose run of July 15, 2014
    *   New TCGA provisional studies: Esophageal cancer, Pheochromocytoma and Paraganglioma (PCPG)
    *   New published TCGA studies: [Thyroid Cancer](https://www.cbioportal.org/study?id=thca_tcga_pub) and [Kidney Chromophobe](https://www.cbioportal.org/study?id=kich_tcga_pub)
*   **Added data** of 172 samples from 4 published studies:
    *   [Cholangiocarcinoma (National University of Singapore, Nature Genetics 2012)](https://www.cbioportal.org/study?id=chol_nus_2012)
    *   [Cholangiocarcinoma (National Cancer Centre of Singapore, Nature Genetics 2013)](https://www.cbioportal.org/study?id=chol_nccs_2013)
    *   [Intrahepatic Cholangiocarcinoma (Johns Hopkins University, Nature Genetics 2013)](https://www.cbioportal.org/study?id=chol_jhu_2013)
    *   [Bladder Cancer (MSKCC, Eur Urol 2014)](https://www.cbioportal.org/study?id=blca_mskcc_solit_2014)
*   **New features**:
    *   Redesigned **Mutual Exclusivity** tab
    *   Added **correlation scores** for scatter plots on the Plots tab
    *   Download links to [**GenomeSpace**](https://www.genomespace.org/)

## October 24, 2014

*   Added data of 885 samples from 11 published studies:
    *   [Colorectal Adenocarcinoma Triplets (MSKCC, Genome Biology 2014)](https://www.cbioportal.org/study?id=coadread_mskcc)
    *   [Esophageal Squamous Cell Carcinoma (ICGC, Nature 2014)](https://www.cbioportal.org/study?id=escc_icgc)
    *   [Malignant Peripheral Nerve Sheath Tumor (MSKCC, Nature Genetics 2014)](https://www.cbioportal.org/study?id=mpnst_mskcc)
    *   [Melanoma (Broad/Dana Farber, Nature 2012)](https://www.cbioportal.org/study?id=skcm_broad_dfarber)
    *   [Nasopharyngeal Carcinoma (National University Singapore, Nature Genetics 2014)](https://www.cbioportal.org/study?id=npc_nusingapore)
    *   [Prostate Adenocarcinoma CNA study (MSKCC, PNAS 2014)](https://www.cbioportal.org/study?id=prad_mskcc_2014)
    *   [Prostate Adenocarcinoma Organoids (MSKCC, Cell 2014)](https://www.cbioportal.org/study?id=prad_mskcc_cheny1_organoids_2014)
    *   [Stomach Adenocarcinoma (TCGA, Nature 2014)](https://www.cbioportal.org/study?id=stad_tcga_pub)
    *   [Stomach Adenocarcinoma (Pfizer and University of Hong Kong, Nature Genetics 2014)](https://www.cbioportal.org/study?id=stad_pfizer_uhongkong)
    *   [Stomach Adenocarcinoma (University of Hong Kong, Nature Genetics 2011)](https://www.cbioportal.org/study?id=stad_uhongkong)
    *   [Stomach Adenocarcinoma (University of Tokyo, Nature Genetics 2014)](https://www.cbioportal.org/study?id=stad_utokyo)

## August 8, 2014

*   Released two new tools
    *   [Oncoprinter](https://www.cbioportal.org/oncoprinter.jsp) lets you create Oncoprints from your own, custom data
    *   [MutationMapper](https://www.cbioportal.org/mutation_mapper.jsp) draws mutation diagrams (lollipop plots) from your custom data

## May 21, 2014

*   All TCGA data updated to the Firehose run of April 16, 2014

## May 12, 2014

*   Improved study summary page including survival analysis based on clinical attributes  
    e.g. [TCGA Endometrial Cancer cohort](https://www.cbioportal.org/study?id=ucec_tcga_pub)  
    ![Study view](https://user-images.githubusercontent.com/1334004/47188334-a026e180-d305-11e8-964c-6c417bb06ca2.png)

## March 27, 2014

*   New features:
    *   Visualizing of mutations mapped on 3D structures (individual or multiple mutations, directly in the browser)
    *   Gene expression correlation analysis (find all genes with expression correlation to your query genes)
    *   The Patient-Centric View now displays mutation frequencies across all cohorts in cBioPortal for each mutation
    *   The Mutation Details Tab and the Patient-Centric View now display the copy-number status of each mutation  
        ![3D viewer & Co-expression](https://user-images.githubusercontent.com/1334004/47188325-9f8e4b00-d305-11e8-96db-67936cd89f8f.png)

## March 18, 2014

*   All TCGA data updated to the Firehose run of January 15, 2014
*   Updated to the latest COSMIC data (v68)
*   Added two new provisional TCGA studies:
    *   Adrenocortical Carcinoma
    *   Uterine Carcinosarcoma
*   Added mutation data of 898 samples from 11 published studies:
    *   Hepatocellular Carcinoma (RIKEN, Nature Genetics 2012)
    *   Hepatocellular Carcinoma (AMC, Hepatology in press)
    *   Medulloblastoma (Broad, Nature 2012)
    *   Medulloblastoma (ICGC, Nature 2012)
    *   Medulloblastoma (PCGP, Nature 2012)
    *   Multiple Myeloma (Broad, Cancer Cell 2014)
    *   Pancreatic Adenocarcinoma (ICGC, Nature 2012)
    *   Small Cell Carcinoma of the Ovary (MSKCC, Nature Genetics in press)
    *   Small Cell Lung Cancer (CLCGP, Nature Genetics 2012)
    *   Small Cell Lung Cancer (Johns Hopkins, Nature Genetics 2012)
    *   NCI-60 Cell Lines (NCI, Cancer Res. 2012)

## December 9, 2013

*   Added mutation data of 99 bladder cancer samples (BGI, Nature Genetics 2013)

## December 6, 2013

*   Data sets matching four recently submitted or published TCGA studies are now available
    *   Glioblastoma (Cell 2013)
    *   Bladder carcinoma (Nature, in press)
    *   Head & neck squamous cell carcinoma (submitted)
    *   Lung adenocarcinoma (submitted)

## November 8, 2013

*   All TCGA data updated to the Firehose run of September 23, 2013.
*   Updated to the latest COSMIC data (v67).
*   Added mutation data of 792 samples from 9 published cancer studies:
    *   Esophageal Adenocarcinoma (Broad, Nature Genetics 2013)
    *   Head and Neck Squamous Cell Carcinoma (Broad, Science 2011)
    *   Head and Neck Squamous Cell Carcinoma (Johns Hopkins, Science 2011)
    *   Kidney Renal Clear Cell Carcinoma (BGI, Nature Genetics 2012)
    *   Prostate Adenocarcinoma, Metastatic (Michigan, Nature 2012)
    *   Prostate Adenocarcinoma (Broad/Cornell, Nature Genetics 2012)
    *   Prostate Adenocarcinoma (Broad/Cornell, Cell 2013)
    *   Skin Cutaneous Melanoma (Yale, Nature Genetics 2012)
    *   Skin Cutaneous Melanoma (Broad, Cell 2012)

## October 21, 2013

*   Improved interface for survival plots, including information on individual samples via mouse-over
*   New fusion glyph in OncoPrints    [![FGFR3 fusions in head and neck carcinoma](https://user-images.githubusercontent.com/1334004/47188320-9f8e4b00-d305-11e8-9d8c-09a7ec28d921.png)](https://www.cbioportal.org/index.do?cancer_study_id=hnsc_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=hnsc_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=hnsc_tcga_pub_gistic&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=hnsc_tcga_pub_sequenced&case_ids=&gene_set_choice=user-defined-list&gene_list=FGFR3%3A+AMP+MUT%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit)
*   Improved cross-cancer query: new alteration frequency histogram (example below - query gene: CDKN2A) and mutation diagram
    ![Cross Cancer Query](https://user-images.githubusercontent.com/1334004/47188319-9ef5b480-d305-11e8-9281-fb31993b7b42.png)

## September 9, 2013

*   Updated COSMIC data (v66 Release)
*   Improved / interactive visualization on the "Protein changes" tab
*   Enhanced mutation diagrams: color-coding by mutation time and syncing with table filters
*   Addition of DNA cytoband information in the patient view of copy-number changes
*   OncoPrints now allow the display of an optional track with clinical annotation (Endometrial cancer example below)
    ![Oncoprint with clinical track](https://user-images.githubusercontent.com/1334004/47188326-9f8e4b00-d305-11e8-9cf5-52e459d56310.png)

## July 25, 2013

*   Multi-gene correlation plots.
*   Variant allele frequency distribution plots for individual tumor samples.
*   Tissue images for TCGA samples in the patient view, via [Digital Slide Archive](https://cancer.digitalslidearchive.net/). [Example](https://www.cbioportal.org/patient?studyId=ucec_tcga&caseId=TCGA-BK-A0CC&tab=tissueImageTab).

## July 16, 2013

*   All TCGA data updated to the May Firehose run (May 23, 2013).
    *   TCGA Pancreatic Cancer study (provisional) added.

## July 4, 2013

*   Improved rendering of mutation diagrams, including ability to download in PDF format.
*   Improved home page: Searchable cancer study & gene set selectors, data sets selector.

## June 17, 2013

*   Improved interface for correlation plots, including information on individual samples via mouse-over.
*   Gene Details from Biogene are now available in the Network view.
*   Added mutation and copy number data from a new adenoid cystic carcinoma study: Ho et al., Nature Genetics 2013.
*   Added mutation data from 6 cancer studies.
    *   Breast Invasive Carcinoma (Shah et al., Nature 2012)
    *   Breast Invasive Carcinoma (Banerji et al., Nature 2012)
    *   Breast Invasive Carcinoma (Stephens et al., Nature 2012)
    *   Lung Adenocarcinoma (Imielinksi et al., Cell 2012)
    *   Lung Adenocarcinoma (Ding et al., Nature 2008)
    *   Colorectal Cancer (Seshagiri et al., Nature 2012)

## June 4, 2013

*   All TCGA data updated to the April Firehose run (April 21, 2012).

## May 14, 2013

*   Added a published TCGA study: Acute Myeloid Leukemia (TCGA, NEJM 2013).

## April 28, 2013

*   All TCGA data updated to the March Firehose run (March 26, 2012).
*   mRNA percentiles for altered genes shown in patient view.

## April 2, 2013

*   All TCGA data updated to the February Firehose run (February 22, 2012).

## March 28, 2013

*   All TCGA data updated to the January Firehose run (January 16, 2012).
*   Data from a new bladder cancer study from MSKCC has been added (97 samples, Iyer et al., JCO in press).

## February 16, 2013

*   The cBio Portal now contains mutation data from all provisional TCGA projects. Please adhere to [the TCGA publication guidelines](https://cancergenome.nih.gov/abouttcga/policies/publicationguidelines) when using these and any TCGA data in your publications.
*   All data updated to the October Firehose run (October 24, 2012).
*   **Sequencing read counts and frequencies** are now shown in the Mutation Details table when available.
*   Improved OncoPrints, resulting in performance improvements.

## November 21, 2012

*   Major new feature: Users can now visualize **genomic alterations and clinical data of individual tumors**, including:
    *   Summary of **mutations** and **copy-number alterations** of interest
    *   **Clinical trial** information
    *   TCGA **Pathology Reports**
*   New **cancer summary view** (Example [Endometrial Cancer](https://www.cbioportal.org/study?id=ucec_tcga))
*   **Updated drug data** from KEGG DRUG and NCI Cancer Drugs (aggregated by [PiHelper](https://bitbucket.org/armish/pihelper))

## October 22, 2012

*   All data updated to the **Broad Firehose** run from July 25, 2012.
*   **COSMIC data** added to Mutation Details (via Oncotator).
*   All predicted functional impact scores are updated to **Mutation Assessor 2.0**.
*   Users can now base queries on genes in recurrent regions of copy-number alteration (from **GISTIC** via Firehose).
*   The [Onco Query Language (OQL)](https://www.cbioportal.org/onco_query_lang_desc.jsp) now supports queries for specific mutations or mutation types.
*   Data sets added that match the data of all TCGA publications (GBM, ovarian, colorectal, and lung squamous).

## July 18, 2012

*   Mutation data for the TCGA lung squamous cell carcinoma and breast cancer projects (manuscripts in press at Nature).
*   All data updated to the **latest Broad Firehose run** (May 25, 2012).
*   **Drug information** added to the network view (via Drugbank).
*   **Improved cross-cancer** queries: Option to select data types, export of summary graphs.
*   Users can now base queries on frequently mutated genes (from **MutSig** via Firehose).

## May 16, 2012

*   All data updated to the **latest Broad Firehose run** (March 21, 2012).
*   Extended **cross-cancer** functionality, enabling users to query across all cancer studies in our database.
*   New **"build a case"** functionality, enabling users to generate custom case sets, based on one or more clinical attributes.
*   New OncoPrint features, including more **compact OncoPrints**, and support for **RPPA** visualization.

## February 27, 2012

*   All data updated to the **latest Broad Firehose run** (January 24, 2012).
*   Validated mutation data for colorectal cancer.
*   New feature: **Mutation Diagrams** that show mutations in the context of protein domains.
    ![TP53 Mutations in Ovarian Cancer](https://user-images.githubusercontent.com/1334004/47188347-a0bf7800-d305-11e8-8c92-7c8763711338.png)

## January 30, 2012

*   Updated data for several TCGA cancer studies.
*   Some small bug-fixes.

## December 22, 2011

*   **Fourteen new [TCGA](https://cancergenome.nih.gov/) cancer studies**: This includes complete data for TCGA Colorectal Carcinoma and provisional data for thirteen other cancer types in the TCGA production pipeline. Please note that data from these thirteen new cancer types are provisional, not final and do not yet include mutation data. As per NCI guidelines, preliminary mutation data cannot be redistributed until they have been validated.
    ![TCGA](https://cancergenome.nih.gov/PublishedContent/Images/SharedItems/Images/TCGA_54px-Logo.png)

*   **Four new data types**:
    *   Reverse-phase protein array (RPPA) data.
    *   microRNA expression and copy-number (including support for multiple loci)
    *   RNA-Seq based expression data.
    *   log2 copy-number data.
*   Updated TCGA GBM copy-number, expression, and methylation data.
*   New **gene symbol validation** service. You can now use gene aliases and/or Entrez Gene IDs within your gene sets.
*   **Links to IGV** for visualization of DNA copy-number changes.
*   Background information from the [Sanger Cancer Gene Census](https://www.sanger.ac.uk/genetics/CGP/Census/).
*   Two **new [Tutorials](https://www.cbioportal.org/tutorials)** to get you quickly started in using the portal.

## November 14, 2011

*   New and **improved mutation details**, with sorting and filtering capabilities.
*   In collaboration with Bilkent University, we have added a **new Network tab** to our results pages. The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from [Pathway Commons](https://www.pathwaycommons.org).
    ![GBM Network](https://user-images.githubusercontent.com/1334004/47188328-9f8e4b00-d305-11e8-87d7-6336285f7572.png)

## September 3, 2011

*   You can now query across different cancer studies (feature available directly from the home page).
*   Our [MATLAB CGDS Cancer Genomics Toolbox](https://www.cbioportal.org/cgds_r.jsp) is now available. The toolbox enables you to download data from the cBio Portal, and import it directly into MATLAB.
*   The code for the cBio Portal has now been fully open sourced, and made available at [Google Code](https://code.google.com/p/cbio-cancer-genomics-portal/). If you would like to join our open source efforts and make the portal even better, drop us an email.

## March 2, 2011

New plotting features and other improvements:

*   Correlation plots that show the relationship between different data types for individual genes.
*   Survival analysis - assess survival differences between altered and non-altered patient sets.
*   Updated [R Package](https://www.cbioportal.org/cgds_r.jsp) with support for correlation plots and general improvements for retrieving and accessing data in R data frames.
*   The [Web Interface](https://www.cbioportal.org/web_api.jsp) now supports basic clinical data, e.g. survival data.
*   [Networks](https://www.cbioportal.org/networks.jsp) for pathway analysis are now available for download.
    ![Survival Analysis](https://user-images.githubusercontent.com/1334004/47188329-9f8e4b00-d305-11e8-83e2-0ad816431888.png)

## December 15, 2010

Several new features, including:

*   Redesigned and streamlined user interface, based on user feedback and usability testing.
*   Advanced support for gene-specific alterations. For example, users can now view mutations within TP53, and ignore copy number alterations, or only view amplifications of EGFR, and ignore deletions.
*   Improved performance.
*   [Frequently Asked Questions](FAQ) document released.
*   Updated ~~Video Tutorial~~ (update: old link no longer functional. Now see: [YouTube](https://www.youtube.com/results?search_query=cbioportal)

## November 4, 2010

*   Enhanced [Oncoprints](https://www.cbioportal.org/faq.jsp#what-are-oncoprints), enabling users to quickly visualize genomic alterations across many cases. Oncoprints now also work in all major browsers, including Firefox, Chrome, Safari, and Internet Explorer.
*   Official release of our [Web Interface](https://www.cbioportal.org/web_api.jsp), enabling programmatic access to all data.
*   Official release of our [R Package](https://www.cbioportal.org/cgds_r.jsp), enabling programmatic access to all data from the R platform for statistical computing.
    ![OncoPrints](https://user-images.githubusercontent.com/1334004/47188322-9f8e4b00-d305-11e8-9fae-ca188de91267.png)
# Welcome to the documentation for cBioPortal!

We've organized this site by the different roles in the cBioPortal community.

<div class="homeList">
            
            <div>
                <a href="user-guide/">
                <img src="images/user-guide.png" width="125" height="125" alt="User guide  free icon" title="User guide free icon">
                </a>
                <a href="user-guide/">Portal User</a>
            </div>
        
            <div>
            <a href="web-api-and-clients/">
            <img src="images/api.png" width="125" height="125" alt="User guide  free icon" title="User guide free icon">
            </a> 
            <a href="web-api-and-clients/">API User</a>
            </div>
       
            <div>
            <a href="deployment/">
            <img src="images/deployment.png" width="125" height="125" alt="User guide  free icon" title="User guide free icon">
            </a>
            <a href="deployment/">Instance Maintainer</a>
            </div>
        
            <div>
            <a href="development/">
            <img src="images/develop.png" width="125" height="125" alt="User guide  free icon" title="User guide free icon">
            </a>
            <a href="development/">Developer</a>
            </div>


</div>


<div id="feedback">
<h3><strong>Feedback? Questions?</strong></h3>
<br />

[!button cBioPortal Google Group](https://groups.google.com/g/cbioportal)

<hr />

:::rfclist
We maintain an active [list of RFCs (Requests for Comments)](RFC-List.md) where we describe new features and solicit community feedback.
:::

</div>

<hr />

<h3>Quick Links</h3>

- [FAQ](user-guide/faq.md)
- [Tutorials](https://www.cbioportal.org/tutorials)
- [API documentation](web-API-and-Clients.md)


<script>
setTimeout(function(){
    document.querySelectorAll("#docs-sidebar-right .simplebar-content")[0].appendChild(document.getElementById("feedback"));
},500);
</script>



# List of Active RFCs

We maintain an active set of RFCs (Requests for Comments) where we spec out new features and solicit community feedback.

See this [shared google folder](https://drive.google.com/drive/folders/0B6z8AlO4d-qMT0Zpd0J4cW5LbHc) for the list of RFCs.


# For Developers Creating new RFCs

* Use the [RFC Template](https://docs.google.com/document/d/1nXi_to7U7qDR41hmrVtIPZBrRS1z4WI4OiT79s6Pc4I/edit)
* Create your RFC within this [shared google folder](https://drive.google.com/drive/folders/0B6z8AlO4d-qMT0Zpd0J4cW5LbHc), and pick a new incremental number.
* Add a Link to your RFC on this page.

# Session Service Management
## Clear Study View User settings
### **Using kubernetes:**
Go inside session service pod (replace `CBIOPORTAL_SESSION_SERVICE_MONGO_POD_ID` with your session service pod id):
```shell
kubectl exec  -it CBIOPORTAL_SESSION_SERVICE_MONGO_POD_ID  -- /bin/sh -c 'mongo admin -u root -p ${MONGODB_ROOT_PASSWORD}'
```
Choose database:
```shell
use session_service;
```
(Optional) Make sure there are matching records to remove (replace `PORTAL_SOURCE` with correct text, one can find it in `session.service.url`, e.g. `public_portal`) 
```shell
db.settings.find({ "source": "PORTAL_SOURCE", "data.page": "study_view" }).count();
```
Remove user settings in study view:
```shell
db.settings.remove({ "source": "PORTAL_SOURCE", "data.page": "study_view" });
```

### **Using docker-compose:**
Go inside session service container (replace `CBIOPORTAL_SESSION_SERVICE_MONGO_CONTAINER_ID` with your session service database container id):
```shell
docker exec  -it CBIOPORTAL_SESSION_SERVICE_MONGO_CONTAINER_ID  bash -c 'mongo'
```
Then follow the same procedure described in the `Using kubernetes` section.The cBioPortal is open source software, available on [GitHub](https://github.com/cbioportal).  If you are interested in joining our efforts, please check out our documentation on [how to contribute](https://github.com/cBioPortal/cbioportal/blob/master/CONTRIBUTING.md).

As an open source project, we are grateful for all the many great tools and services that keep our project humming.

We gratefully acknowledge a large collection of open source languages, tools and frameworks, including:

* [Java](https://www.java.com/), [Python](https://www.python.org/) and [Typescript](https://www.typescriptlang.org/)
* [React](https://reactjs.org/) and [MobX](https://mobx.js.org/)
* [Bootstrap](https://getbootstrap.com/) and [React-bootstrap](https://react-bootstrap.github.io/)
* [Yarn](https://yarnpkg.com/), [npm](https://www.npmjs.com/) and [webpack](https://webpack.js.org/)
* [Mocha](https://mochajs.org/), [Chai](https://www.chaijs.com/) and [Selenium](https://selenium.dev/)
* [D3](https://d3js.org/), [Victory.js](https://formidable.com/open-source/victory/) and [Chart JS](https://www.chartjs.org/)
* [Spring](https://spring.io/) and [Hibernate](https://hibernate.org/)
* [MySQL](https://www.mysql.com/) and [MongoDB](https://www.mongodb.com/)

We also gratefully acknowledge the following service providers:

* [GitHub](https://github.com) for source control management and community engagement.
* [Slack](https://slack.cbioportal.org/)  for coordinating work across a geographically dispersed team.
* [Travis CI](https://travis-ci.org/) and [Circle CI](https://circleci.com/) for continuous integration.
* [Netlify](https://www.netlify.com/), [Heroku](https://www.heroku.com/) and [Amazon Web Services](https://aws.amazon.com/) for cloud hosting services.# Study Curation Guide
This guide is to help data curators learn how to curate a study on their own computer

## Prerequisites
- To follow this guide the curator should have some familiarity with running commands on the command line. We will be using Docker. No Docker knowledge is required, one will obtain some basic understanding by following the guide.
- Learn how to setup cBioPortal locally [here](docker/README.md) first.

## Load an example study
After having followed the steps in the [Docker Deployment instructions](docker/README.md), you will end up with the study [Low-Grade Gliomas (UCSF, Science 2014)
](https://www.cbioportal.org/study/summary?id=lgg_ucsf_2014) loaded locally. Now let's try to import another study:

1. Choose another example study from the [datahub](https://github.com/cBioPortal/datahub/tree/master/public). Note the name of the folder which is identical to the study id defined in [meta_study.txt](https://github.com/cBioPortal/datahub/blob/master/public/lgg_ucsf_2014/meta_study.txt#L2).
2. From the root of the `cbioportal-docker-compose` folder run `DATAHUB_STUDIES=my_study_id ./study/init.sh`. Change `my_study_id` to the study you picked in 1. The study should now be downloaded in `./study/`.
3. Import the study by running `docker-compose exec cbioportal metaImport.py -u http://cbioportal:8080 -s study/my_study_id/ -o`. Again change `my_study_id` to the study you picked in 1. This should import the study.
4. Restart the cbioportal instance `docker-compose restart cbioportal` and see if the new study shows up on http://localhost:8080

## Curate a new study
The cBioPortal team has curated many published studies in formats suitable for import in cBioPortal. These can
be found on the [datahub](https://github.com/cbioportal/datahub) and can serve as an example of how our curation
processes works. You can find a step by step description of how to curate a new study [here](https://github.com/cBioPortal/datahub/blob/master/docs/curation-process.md).
# Table of contents

* [Home](README.md)
* [About Us](About-Us.md)
* [List of RFCs](RFC-List.md)
* [User Guide](user-guide/README.md)
  * [New Users](user-guide/new-users.md)
  * [Frequently Asked Questions](user-guide/faq.md)
  * [Tutorials](user-guide/overview.md)
  * [Page Specific Resources](user-guide/by-page.md)
  * [OQL](user-guide/oql.md)
  * [News](News.md)
  * [Genie News](News-Genie.md)
* [Web API and API Clients](web-API-and-Clients.md)
* [Deployment](deployment/README.md)
  * [Architecture overview](Architecture-Overview.md)
  * [Hardware Requirements](Hardware-Requirements.md)
  * [Deploy with Docker (Recommended)](deployment/docker/README.md)
    * [Import data with Docker](deployment/docker/import\_data.md)
    * [Example Commands](deployment/docker/example\_commands.md)
    * [Authenticating and Authorizing Users using Keycloak in Docker](deployment/docker/using-keycloak.md)
  * [Deploy without Docker](deployment/deploy-without-docker/README.md)
    * [Software Requirements](deployment/deploy-without-docker/Software-Requirements.md)
    * [Pre-Build Steps](deployment/deploy-without-docker/Pre-Build-Steps.md)
    * [Building from Source](deployment/deploy-without-docker/Build-from-Source.md)
    * [Importing the Seed Database](deployment/deploy-without-docker/Import-the-Seed-Database.md)
    * [Deploying the Web Application](deployment/deploy-without-docker/Deploying.md)
    * [Loading a Sample Study](deployment/deploy-without-docker/Load-Sample-Cancer-Study.md)
  * [Authorization and Authentication](deployment/authorization-and-authentication/README.md)
    * [User Authorization](deployment/authorization-and-authentication/User-Authorization.md)
    * [Authenticating Users via SAML](deployment/authorization-and-authentication/Authenticating-Users-via-SAML.md)
    * [Authenticating Users via LDAP](deployment/authorization-and-authentication/Authenticating-Users-via-LDAP.md)
    * [Authenticating and Authorizing Users via Keycloak](deployment/authorization-and-authentication/Authenticating-and-Authorizing-Users-via-keycloak.md)
    * [Authenticating Users via Tokens](deployment/authorization-and-authentication/Authenticating-Users-via-Tokens.md)
  * [Customization]()
    * [Customizing your cBioPortal Instance via application.properties](deployment/customization/Customizing-your-instance-of-cBioPortal.md)
    * [More application.properties Settings](deployment/customization/application.properties-Reference.md)
    * [Security Properties Settings](deployment/customization/security.properties-Reference.md)
    * [Configuring Caching Behavior](deployment/customization/Caching.md)
    * [How does the study view organize the charts](deployment/customization/Studyview.md)
    * [Utilize priority setting from database to visualize charts](deployment/customization/Studyview.md)
  * [Integration with Other Webservices]()
    * [OncoKB Data Access](deployment/integration-with-other-webservices/OncoKB-Data-Access.md)
  * [Data Loading Overview](Data-Loading.md)
    * [Downloads](Downloads.md)
    * [Using the Dataset Validator](Using-the-dataset-validator.md)
    * [Using the metaImport Script](Using-the-metaImport-script.md)
    * [File Formats](File-Formats.md)
      * [Z-Score Normalization](Z-Score-normalization-script.md)
    * [Study Curation Guide](Study-Curation-Guide.md)
    * [Maintaining Studies](Data-Loading-Maintaining-Studies.md)
    * [For Developers](Data-Loading-For-Developers.md)
    * [Tips and Best Practices](Data-Loading-Tips-and-Best-Practices.md)
    * [Mutation Data Transcript Annotation](mutation-data-transcript-annotation.md)
    * [Import OncoKB annotations as custom driver annotations](Import-OncoKB-annotations-as-Custom-Driver-Annotations.md)
    * [Importing Gene Sets for Gene Set Scoring Data](Import-Gene-Sets.md)
    * [Importing Gene Panels](Import-Gene-Panels.md)
    * [Study View Customization with Priority Data](deployment/customization/Studyview.md)
  * [cBioPortal Maintenance](deployment/cbioportal-maintenance/README.md)
    * [Updating your cBioPortal Database Scheme](Updating-your-cBioPortal-installation.md)
    * [Update genes and gene aliases](Updating-gene-and-gene_alias-tables.md)
    * [Migration Guide](Migration-Guide.md)
    * [MSK Maintenance](MSK-Maintenance.md)
* [Development](development/README.md)
  * [Feature Development Guide](development/feature-development-guide.md)
  * [cBioPortal Entity-relationship Diagram](development/cBioPortal-ER-Diagram.md)
  * [cBioPortal Database Versioning](development/Database-Versioning.md)
  * [Build cBioPortal with a different frontend version](development/Build-Different-Frontend.md)
  * [Session Service](development/session-service-working.md)
  * [Manual test cases](development/manual-test-cases.md)
  * [Release Procedure](development/Release-Procedure.md)
  * [Deployment Procedure](development/Deployment-Procedure.md)
  * [This documentation site](development/Documentation-site.md)

# Introduction

This page is about how to run and debug, in an IDE agnostic way, any test classes added to the backend (Java layer).

# Steps

## Set up test instance of MySQL server

### System installed (assumes Linux OS)

1. Start mysql:

```
sudo systemctl start mysql
```

2. Log into database:

```shell
sudo mysql -u root
```

or

```shell
mysql -u root -p
```

3. In the MySQL console run:

```shell
SET GLOBAL local_infile=1;
CREATE DATABASE cgds_test;
CREATE USER 'cbio_user'@'localhost' IDENTIFIED BY 'somepassword';
GRANT ALL ON cgds_test.* TO 'cbio_user'@'localhost';
FLUSH PRIVILEGES;
SET default_storage_engine=InnoDB;
SET SESSION sql_mode = 'ANSI_QUOTES';
```

4. (Optional) Load schema and seed:

When running tests through Maven this step can be skipped. In MySQL console run:

```shell
source <path-to-file>/cgds-test.sql
source <path-to-file>/seed_mini.sql
```

#### Note

_Cgds-test.sql_ should be generated with `/db-scripts/src/main/resources/gen-cgds-test-schema.sh`.

## Docker container

1. Create file `init.sql` with the following contents:

```shell
SET GLOBAL local_infile=1;
CREATE DATABASE cgds_test;
CREATE USER 'cbio_user'@'localhost' IDENTIFIED BY 'somepassword';
GRANT ALL ON cgds_test.* TO 'cbio_user'@'localhost'";
FLUSH PRIVILEGES;
SET default_storage_engine=InnoDB;
SET SESSION sql_mode = 'ANSI_QUOTES';
```

2. Run (making sure the file paths are correct):

```shell
 docker run -ti --rm \
  --name testdb \
  --env MYSQL_ROOT_PASSWORD=root \
  --env MYSQL_USER=cbio_user \
  --env MYSQL_PASSWORD=somepassword \
  --env MYSQL_DATABASE=cgds_test \
  -v <path-to-file>/init.sql:/docker-entrypoint-initdb.d/first_file.sql \
  -v <path-to-file>/cgds-test.sql:/docker-entrypoint-initdb.d/second_file.sql \
  -v <path-to-file>/seed_mini.sql:/docker-entrypoint-initdb.d/third_file.sql \
  -p 3306:3306 mysql:5.7
  --ssl=0
```

#### Notes

- When running tests through Maven, the line with volume mounts (starting with `-v`) to the schema (_cgds-test.sql_) and
  seed (_seed_mini.sql_) should be skipped.
- _Cgds-test.sql_ should be generated with `/db-scripts/src/main/resources/gen-cgds-test-schema.sh`.

## Running the tests with Maven

1. Make sure you have built the project once with e.g. `mvn clean install -DskipTests`
2. Run integration tests with

```
mvn integration-test -Ddb.test.username=cbio_user -Ddb.test.password=somepassword
```

This will create the tables in `cgds_test` and populate your schema with test data.

3. you can run all other tests with

```
mvn test -Ddb.test.username=cbio_user -Ddb.test.password=somepassword
```

4. you can run a specific test with

```
mvn test -pl core -Dtest=TestIntegrationTest -Ddb.test.username=cbio -Ddb.test.password= <your_db_password>
```

where `-pl` is the name of the module (in this example `core`) and `-Dtest` is the name of the Java test class (in this
example `TestIntegrationTest`).

5. you can debug a specific test with

```
mvn integration-test -Dmaven.surefire.debug -pl core test -Dtest=TestIntegrationTest -Ddb.test.username=cbio -Ddb.test.password=<your_db_password> 
```

where `-pl` is the name of the module (in this example `core`) and `-Dtest` is the name of the Java test class (in this
example `TestIntegrationTest`). This command will pause execution and wait for you to connect your IDE to the listening
port, reporting something like below:

```
[INFO] Surefire report directory: /cbioportal/core/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Listening for transport dt_socket at address: 5005
```

Once you connected your IDE (after setting a breakpoint in the code), the execution will continue in your debugger view.
# Uninstall Docker cBioPortal

## 1. Remove cBioPortal docker container

In a docker terminal type the following command:

```bash
docker rm -fv ${CONTAINER_NAME}
```

Where:    
- **${CONTAINER_NAME}**: The name of your container instance, _i.e **cbioportal**_.

## 2. Remove MySQL docker container

In a docker terminal type the following command:

```bash
docker rm -fv ${CONTAINER_NAME}
```

Where:    
- **${CONTAINER_NAME}**: The name of your container instance, _i.e **cbio-DB**_.

## 3. Remove the docker network

In a docker terminal type the following command:

```bash
docker network rm ${DOCKER_NETWORK_NAME}
```

Where:
- **${DOCKER_NETWORK_NAME}** is the name of the network that cBioPortal and the cBioPortal DB are accessible. _i.e **cbio-net**_# Updating the gene names and aliases tables
This manual is intended for users that have knowledge about the structure of the cBioPortal seed database.

When loading studies into cBioPortal it is possible for warnings to occur that are caused by an outdated seed database. Gene symbols can be deprecated or be assigned to a different Entrez Gene in a new release. Also Entrez Gene IDs can be added. This markdown explains how to update the seed database, in order to use the most recent Entrez Gene IDs.

The cBioPortal scripts package provides a method to update the `gene` and `gene_alias` tables.

## Prepare

### Human genes
Download [gene_info.txt](https://github.com/cBioPortal/datahub-study-curation-tools/blob/master/gene-table-update/build-input-for-importer/gene_info.txt)  
Generated based on latest HGNC release using script [HERE](https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/gene-table-update/build-input-for-importer) 

### Mouse genes
Download `Mus_musculus.gene_info.gz` from   
ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Mus_musculus.gene_info.gz
Unzip the downloaded file with the command `gunzip Mus_musculus.gene_info.gz`

## MySQL steps
Execute these steps in case you want to reset your database to the most recent genes list from NCBI.

1- Start a new MySQL database with the previous seed database, which can be found on cBioPortal Datahub for [human](https://github.com/cBioPortal/datahub/tree/master/seedDB) and [mouse](https://github.com/cBioPortal/datahub/tree/master/seedDB_mouse).

2- If DB engine supports foreign key (FK) constraints, e.g. InnoDB, drop constraints:
```sql
ALTER TABLE cosmic_mutation
  DROP FOREIGN KEY cosmic_mutation_ibfk_1;

ALTER TABLE uniprot_id_mapping
  DROP FOREIGN KEY uniprot_id_mapping_ibfk_1;
```

3- Empty tables `gene` and `gene_alias`
```sql
TRUNCATE TABLE gene_alias;
DELETE from genetic_entity;
DELETE from geneset_hierarchy_node;
ALTER TABLE `genetic_entity` AUTO_INCREMENT = 1;
ALTER TABLE `geneset_hierarchy_node` AUTO_INCREMENT = 1;
ALTER TABLE `geneset` AUTO_INCREMENT = 1;
```

4- Restart cBioPortal (restart webserver)  or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information)
to clean-up any cached gene lists.

5- To import gene data type the following commands when in the folder `<cbioportal_source_folder>/core/src/main/scripts`:
### Human genes
```
export PORTAL_HOME=<cbioportal_configuration_folder>
export JAVA_HOME=<jre_installation_folder>
./importGenes.pl --hgnc <gene_info.txt> --genome-build <GRCh37>
```
### Mouse genes
```
export PORTAL_HOME=<cbioportal_configuration_folder>
export JAVA_HOME=<jre_installation_folder>
./importGenes.pl --genes <Mus_musculus.gene_info.gz> --gtf <gencode.v25.annotation.gtf> --genome-build <GRCh37>
```
**IMPORTANT NOTE**: 
1. The **reference_genome** table needs to be populated before updating the **gene** table. Further details can be found in [this document](import-reference-genome.md). 
2. Use **--species** option when importing genes for a species other than human
3. Use the **gene** table if you query information such as hugo symbols, types of the gene 
4. Use **reference_genome_gene** table if you query information such as chromosome, cytoband, exonic length, or the start or end of the gene
5. Load genes only to the **reference_genome_gene** table without updating the **gene** table, please use the following command:
### Human genes
```
./importGenes.pl --gtf <gencode.v27.annotation.gtf> --genome-build <GRCh38>
```
### Mouse genes
```
./importGenes.pl --gtf <gencode.v27.annotation.gtf> --genome-build <GRCh38>
```
6- :warning: Check the `gene` and `gene_alias` tables to verify that they are filled correctly.
```sql
SELECT count(*) FROM cbioportal.gene;
SELECT count(*) FROM cbioportal.gene_alias;
```

7- Additionally, there are other tables you may want to update now (only in human).

* Updating the COSMIC coding mutations, can be downloaded from [here](https://cancer.sanger.ac.uk/cosmic/download) and require the script `importCosmicData.pl`

8- Clean-up old data:
```sql
SET SQL_SAFE_UPDATES = 0;
DELETE FROM cosmic_mutation where ENTREZ_GENE_ID not in (SELECT ENTREZ_GENE_ID from gene);
DELETE FROM sanger_cancer_census where ENTREZ_GENE_ID not in (SELECT ENTREZ_GENE_ID from gene);
DELETE FROM uniprot_id_mapping where ENTREZ_GENE_ID not in (SELECT ENTREZ_GENE_ID from gene);
DELETE FROM interaction where GENE_A not in (SELECT ENTREZ_GENE_ID from gene) or GENE_B not in (SELECT ENTREZ_GENE_ID from gene);
DELETE FROM drug_interaction where target not in (SELECT ENTREZ_GENE_ID from gene);
SET SQL_SAFE_UPDATES = 1;
commit;
```

9- If DB engine supports FK constraints, e.g. InnoDB, restore constraints:
```sql
ALTER TABLE cosmic_mutation
  ADD CONSTRAINT cosmic_mutation_ibfk_1 FOREIGN KEY (`ENTREZ_GENE_ID`) REFERENCES `gene` (`ENTREZ_GENE_ID`);

ALTER TABLE uniprot_id_mapping
  ADD CONSTRAINT uniprot_id_mapping_ibfk_1 FOREIGN KEY (`ENTREZ_GENE_ID`) REFERENCES `gene` (`ENTREZ_GENE_ID`);
```

10- You can import new gene sets using the gene set importer. These gene sets are currently only used for gene set scoring. See [Import-Gene-Sets.md](Import-Gene-Sets.md) and [File-Formats.md#gene-set-data](File-Formats.md#gene-set-data).

For example, run in folder `<cbioportal_source_folder>/core/src/main/scripts`:
```bash
./importGenesetData.pl --data ~/Desktop/msigdb.v6.1.entrez.gmt --new-version msigdb_6.1
```
Please make sure the version gene sets is the same as the version used to calculate gene set scores in your data.
# Updating your cBioPortal installation

As of release **1.1.0** cBioPortal has a Database schema update mechanism which comes into play whenever the new version of the portal code relies on specific DB schema changes to be applied. The portal will automatically check if the DB schema is according to what it expects. It does so by comparing the version number of the portal code with the version number of the DB schema. If they are equal, it assumes the DB schema has been upgraded. If not, it will require the administrator to run a migration script. Below are the steps to ensure your DB schema is updated correctly.

## First time

The first time you update from release **1.0.4** (or lower) to release **1.1.0** (or higher), you should get a an error banner page after restarting your webserver. The error should state something like: 

```
Current DB schema version: xxx 
DB schema version expected by Portal: yyy
```
where `xxx` and `yyy` will be different version numbers. 

If you get `DB version expected by Portal: 0` (i.e. you are building the new release from source), you need to  add a new property to your `application.properties` file which is needed for this check. 

#### Step1

In your `application.properties` file (e.g. `<your_cbioportal_dir>/src/main/resources/application.properties`) add the following property:
```
# this is the *expected* DB version (expected by the code). Don't set it manually, it is filled by maven:
db.version=${db.version}
```

#### Step2

Compile your code again. After restarting the webserver the page should now state something like: `DB version expected by Portal: 1.1.0` (or higher), while the DB version remains as `Current DB version: -1`. 

## Running the migration script

First, make sure you have the DB connection properties correctly set in your application.properties file (see [DB connection settings here](/deployment/customization/application.properties-Reference.md#database-settings)).

**Dependencies:** the migration script is a Python script that depends on the `mysqlclient` library. If necessary, you can install it with the following commands (example for Ubuntu):
```console
sudo apt-get install python3-dev default-libmysqlclient-dev
sudo python3 -m pip install mysqlclient
```

For macOS, try the following:

```
brew install mysql-connector-c
sudo python3 -m pip install mysqlclient
```
and see <https://github.com/PyMySQL/mysqlclient-python/blob/master/README.md#prerequisites>
if problems occur during installation.

To run the migration script first go to the scripts folder
`<your_cbioportal_dir>/core/src/main/scripts` 
and then run the following command:
```console
$ python migrate_db.py
```
This should result in the following output:
```console
WARNING: This script will alter your database! Be sure to back up your data before running.
Continue running DB migration? (y/n) y
Running statments for version: 1.0.0
	Executing statement: CREATE TABLE info (DB_SCHEMA_VERSION VARCHAR(8));
	Executing statement: INSERT INTO info VALUES ("1.0.0");
Running statments for version: 1.1.0
	Executing statement: CREATE TABLE sample_list LIKE patient_list;
	Executing statement: INSERT sample_list SELECT * FROM patient_list;
	Executing statement: CREATE TABLE sample_list_list LIKE patient_list_list;
	Executing statement: INSERT sample_list_list SELECT * FROM patient_list_list;
	Executing statement: ALTER TABLE sample_list_list CHANGE PATIENT_ID SAMPLE_ID INT(11);
	Executing statement: UPDATE info SET DB_SCHEMA_VERSION="1.1.0";
...
etc
```

**Final step:** Restart your webserver or call the `/api/cache` endpoint with a `DELETE` http-request
(see [here](/deployment/customization/application.properties-Reference.md#evict-caches-with-the-apicache-endpoint) for more information).
## Resources by Page
### Study View
* Tutorial Slides: Single Study Exploration [Google slides](https://docs.google.com/presentation/d/1_OGK69lO4Z62WaxHHkNYmWvY0LQN2v0slfaLyY1_IQ0) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%201%20Single%20Study%20Exploration.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=760s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
* How-To Video: Comparing samples based on expression level of a gene [youtube.com](https://www.youtube.com/watch?v=HTiKUXk0j0s)
* Tutorial Slides: Virtual Studies [Google slides](https://docs.google.com/presentation/d/1rQE5rbFNdmup-rAtySHFxlLp3i4qa8SBA7MiQpMdn1I) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%204%20Virtual%20Studies.pdf)


### Group Comparison
* Tutorial Slides: Group Comparison [Google slides](https://docs.google.com/presentation/d/1P2boDph8IfpvjxoxDj_496CLHGtshzJnbbZhszPsmf4) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
* Webinar: Group Comparison [youtube.com](https://www.youtube.com/watch?v=Tx4HZCrIe5c) | [bilibili.com](https://www.bilibili.com/video/BV1VZ4y1W76p)
* How-To Video: Comparing samples based on expression level of a gene [youtube.com](https://www.youtube.com/watch?v=HTiKUXk0j0s)

### Running a Query / Results View
#### General
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=1462s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)


#### OQL
* Documentation [OQL](/user-guide/oql.md)
* Tutorial Slides: Onco Query Language (OQL) [Google slides](https://docs.google.com/presentation/d/1U39xgVujtBodwW20qIfcGu4E5n2zzaKkl2KmzzHqj4A) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
* Webinar: Expression Data Analysis [youtube.com](https://www.youtube.com/watch?v=YUxVv6pkxD4&t=1559s) | [bilibili.com](https://www.bilibili.com/video/BV1HK4y1t7dE)
* How-To Video: Using Onco Query Language (OQL) to query based on the expression level of genes [youtube.com](https://www.youtube.com/watch?v=kHlFXw2TMzc)


#### OncoPrint
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=1682s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Cancer Types Summary
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2024s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Mutual Exclusivity
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=1822s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Plots
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2060s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
#### Mutations
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2149s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
* Webinar: Mutation Details & Patient View [youtube.com](https://www.youtube.com/watch?v=uJsp9kd2jIk) | [bilibili.com](https://www.bilibili.com/video/BV1Qf4y1m7Lx)
* How-To Video: Filtering and adding clinical data to Mutations tab [youtube.com](https://www.youtube.com/watch?v=q9No2073c5o)

#### Co-expression
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2277s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Comparison/Survival
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2334s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
* Tutorial Slides: Group Comparison [Google slides](https://docs.google.com/presentation/d/1P2boDph8IfpvjxoxDj_496CLHGtshzJnbbZhszPsmf4) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
* Webinar: Group Comparison [youtube.com](https://www.youtube.com/watch?v=Tx4HZCrIe5c) | [bilibili.com](https://www.bilibili.com/video/BV1VZ4y1W76p)

#### CN Segments
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2449s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Pathways
* Tutorial Slides: Pathways [Google slides](https://docs.google.com/presentation/d/1O5WGucz0lrfdY25b5QS6zaID_26i434EYXBluqZfT2g) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%207%20Pathways.pdf)
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2512s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

#### Downloads
* Tutorial Slides: Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=2543s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)

### Patient View
* Tutorial Slides: Patient View [Google slides](https://docs.google.com/presentation/d/1Jr_2yEfgjKBn4DBiXRk4kmhIbtsRp6gd0iD3k1fIUUk) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%203%20Patient%20View.pdf)
* Tutorial Slides: Pathways [Google slides](https://docs.google.com/presentation/d/1O5WGucz0lrfdY25b5QS6zaID_26i434EYXBluqZfT2g) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%207%20Pathways.pdf)
* Webinar: Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo&t=1022s) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)
* Webinar: Mutation Details & Patient View [youtube.com](https://www.youtube.com/watch?v=uJsp9kd2jIk) | [bilibili.com](https://www.bilibili.com/video/BV1Qf4y1m7Lx)
* How-To Video: Exploring the longitudinal evolution of individual patients [youtube.com](https://www.youtube.com/watch?v=Hbbs-tHh9LQ)
# cBioPortal FAQs


* [General Questions](#general-questions)
    * [What is the cBioPortal for Cancer Genomics?](/user-guide/faq.md#what-is-the-cbioportal-for-cancer-genomics)
    * [How do I get started?](/user-guide/faq.md#how-do-i-get-started)
    * [What data types are in the portal?](/user-guide/faq.md#what-data-types-are-in-the-portal)
    * [What does ___ stand for?](#what-does-___-stand-for)
    * [What is the process of data curation?](/user-guide/faq.md#what-is-the-process-of-data-curation)
    * [How do I get updates on new portal developments and new data sets?](/user-guide/faq.md#how-do-i-get-updates-on-new-portal-developments-and-new-data-sets)
    * [Does the portal work on all browsers and operating systems?](/user-guide/faq.md#does-the-portal-work-on-all-browsers-and-operating-systems)
    * [How do I cite the cBioPortal?](/user-guide/faq.md#how-do-i-cite-the-cbioportal)
    * [Can I use figures from the cBioPortal in my publications or presentations?](/user-guide/faq.md#can-i-use-figures-from-the-cbioportal-in-my-publications-or-presentations)
    * [Can I save or bookmark my results in cBioPortal?](/user-guide/faq.md#can-i-save-or-bookmark-my-results-in-cbioportal)
    * [How is the cBioPortal for Cancer Genomics different from the Genomic Data Commons (GDC)?](/user-guide/faq.md#how-is-the-cbioportal-for-cancer-genomics-different-from-the-genomic-data-commons-gdc)
    * [Does the cBioPortal provide a Web Service API? R interface? MATLAB interface?](/user-guide/faq.md#does-the-cbioportal-provide-a-web-service-api-r-interface-matlab-interface)
    * [Can I use cBioPortal with my own data?](/user-guide/faq.md#can-i-use-cbioportal-with-my-own-data)
    * [Can I create a local instance of cBioPortal to host my own data?](/user-guide/faq.md#can-i-create-a-local-instance-of-cbioportal-to-host-my-own-data)
    * [I'd like to contribute code to the cBioPortal. How do I get started?](/user-guide/faq.md#id-like-to-contribute-code-to-the-cbioportal-how-do-i-get-started)
    * [What is a combined Study?](/user-guide/faq.md#what-is-a-combined-study)
    * [How can I form a combined Study?](/user-guide/faq.md#how-can-i-form-a-combined-study)
    * [How can I create a subset or sub cohort of a study with specific samples or patients?](/user-guide/faq.md#how-can-i-create-a-subset-or-sub-cohort-of-a-study-with-specific-samples-or-patients)
    * [What is a Virtual Study?](/user-guide/faq.md#what-is-a-virtual-study)
    * [Is it necessary to log in to use virtual studies? If I do log in, what additional functionality do I gain?](/user-guide/faq.md#is-it-necessary-to-log-in-to-use-virtual-studies-if-i-do-log-in-what-additional-functionality-do-i-gain)
    * [What is Group Comparison?](/user-guide/faq.md#what-is-group-comparison)
    * [What are the statistical significance tests in Group Comparison?](/user-guide/faq.md#what-are-the-statistical-significance-tests-in-group-comparison)
* [Data Questions](/user-guide/faq.md#data-questions)
    * [General Data](/user-guide/faq.md#general-data)
        * [Does the portal contain cancer study X?](/user-guide/faq.md#does-the-portal-contain-cancer-study-x)
        * [Which resources are integrated for variant annotation?](/user-guide/faq.md#which-resources-are-integrated-for-variant-annotation)
        * [What version of the human reference genome is being used in cBioPortal?](/user-guide/faq.md#what-version-of-the-human-reference-genome-is-being-used-in-cbioportal)
        * [How does cBioPortal handle duplicate samples or sample IDs across different studies?](/user-guide/faq.md#how-does-cbioportal-handle-duplicate-samples-or-sample-ids-across-different-studies)
        * [Are there any normal tissue samples available through cBioPortal?](/user-guide/faq.md#are-there-any-normal-tissue-samples-available-through-cbioportal)
        * [How can I find which studies have mRNA expression data (or any other specific data type)?](/user-guide/faq.md#how-can-i-find-which-studies-have-mrna-expression-data-or-any-other-specific-data-type)
        * [Can I download all data at once?](/user-guide/faq.md#can-i-download-all-data-at-once)
        * [The data today is different than the last time i looked. What happened?](/user-guide/faq.md#the-data-today-is-different-than-the-last-time-i-looked-what-happened)
        * [How do I access data from AACR Project GENIE?](/user-guide/faq.md#how-do-i-access-data-from-aacr-project-genie)
    * [TCGA](/user-guide/faq.md#tcga)
        * [How does TCGA data in cBioPortal compare to TCGA data in Genome Data Commons?](/user-guide/faq.md#how-does-tcga-data-in-cbioportal-compare-to-tcga-data-in-genome-data-commons)
        * [What happened to TCGA Provisional datasets?](/user-guide/faq.md#what-happened-to-tcga-provisional-datasets)
        * [What are TCGA Firehose Legacy datasets and how do they compare to the publication-associated datasets and the PanCancer Atlas datasets?](/user-guide/faq.md#what-are-tcga-firehose-legacy-datasets-and-how-do-they-compare-to-the-publication-associated-datasets-and-the-pancancer-atlas-datasets)
        * [Where do the thresholded copy number call in TCGA Firehose Legacy data come from?](/user-guide/faq.md#where-do-the-thresholded-copy-number-call-in-tcga-firehose-legacy-data-come-from)
        * [Which studies have MutSig and GISTIC results? How do these results compare to the data in the TCGA publications?](/user-guide/faq.md#which-studies-have-mutsig-and-gistic-results-how-do-these-results-compare-to-the-data-in-the-tcga-publications)
        * [How can I download the PanCancer Atlas data?](/user-guide/faq.md#how-can-i-download-the-pancancer-atlas-data)
    * [DNA (Mutations, Copy Number &amp; Fusions)](/user-guide/faq.md#dna-mutations-copy-number--fusions)
        * [Does the cBioPortal contain synonymous mutation data?](/user-guide/faq.md#does-the-cbioportal-contain-synonymous-mutation-data)
        * [What processing or filtering is applied to generate the mutation data?](/user-guide/faq.md#what-processing-or-filtering-is-applied-to-generate-the-mutation-data)
        * [What transcripts are used for annotating mutations?](/user-guide/faq.md#what-transcripts-are-used-for-annotating-mutations)
        * [How are protein domains in the mutational lollipop diagrams specified?](/user-guide/faq.md#how-are-protein-domains-in-the-mutational-lollipop-diagrams-specified)
        * [What is the difference between a ‚Äúsplice site‚Äù mutation and a ‚Äúsplice region‚Äù mutation?](/user-guide/faq.md#what-is-the-difference-between-a-splice-site-mutation-and-a-splice-region-mutation)
        * [What do ‚ÄúAmplification‚Äù, ‚ÄúGain‚Äù, ‚ÄúDeep Deletion‚Äù, ‚ÄúShallow Deletion‚Äù and "-2", "-1", "0", "1", and "2" mean in the copy-number data?](/user-guide/faq.md#what-do-amplification-gain-deep-deletion-shallow-deletion-and--2--1-0-1-and-2-mean-in-the-copy-number-data)
        * [What is GISTIC? What is RAE?](/user-guide/faq.md#what-is-gistic-what-is-rae)
    * [RNA](/user-guide/faq.md#rna)
        * [Does the portal store raw or probe-level data?](/user-guide/faq.md#does-the-portal-store-raw-or-probe-level-data)
        * [What are mRNA and microRNA Z-Scores?](/user-guide/faq.md#what-are-mrna-and-microrna-z-scores)
        * [Is there any normal RNA-seq data in cBioPortal?](/user-guide/faq.md#is-there-any-normal-rna-seq-data-in-cbioportal)
        * [How is TCGA RNASeqV2 processed? What units are used?](/user-guide/faq.md#how-is-tcga-rnaseqv2-processed-what-units-are-used)
        * [Is there microRNA data?](/user-guide/faq.md#is-there-microrna-data)
        * [How can I query microRNAs in the portal?](/user-guide/faq.md#how-can-i-query-micrornas-in-the-portal)
    * [Protein](/user-guide/faq.md#protein)
        * [How can I query phosphoprotein levels in the portal?](/user-guide/faq.md#how-can-i-query-phosphoprotein-levels-in-the-portal)
        * [Why isn‚Äôt there protein data for my gene of interest?](/user-guide/faq.md#why-isnt-there-protein-data-for-my-gene-of-interest)
    * [DNA Methylation](/user-guide/faq.md#dna-methylation)
        * [Which methylation probe is used for genes with multiple probes?](/user-guide/faq.md#which-methylation-probe-is-used-for-genes-with-multiple-probes)
    * [Clinical Data](/user-guide/faq.md#clinical-data)
        * [What kind of clinical data is stored in the portal?](/user-guide/faq.md#what-kind-of-clinical-data-is-stored-in-the-portal)
        * [What is the meaning of OS_STATUS / OS_MONTHS, and PFS_STATUS / PFS_MONTHS?](/user-guide/faq.md#what-is-the-meaning-of-os_status--os_months-and-pfs_status--pfs_months)
* [Analysis Questions](/user-guide/faq.md#analysis-questions)
    * [How can I query/explore a select subset of samples?](/user-guide/faq.md#how-can-i-queryexplore-a-select-subset-of-samples)
    * [How can I compare two or more subsets of samples?](/user-guide/faq.md#how-can-i-compare-two-or-more-subsets-of-samples)
    * [Is it possible to determine if a particular mutation is heterozygous or homozygous in a sample? When a sample has 2 mutations in one gene, is it possible to determine whether the mutations are in cis or in trans with each other?](/user-guide/faq.md#is-it-possible-to-determine-if-a-particular-mutation-is-heterozygous-or-homozygous-in-a-sample-when-a-sample-has-2-mutations-in-one-gene-is-it-possible-to-determine-whether-the-mutations-are-in-cis-or-in-trans-with-each-other)
    * [How can I query over/under expression of a gene?](/user-guide/faq.md#how-can-i-query-overunder-expression-of-a-gene)
    * [How can I compare outcomes in patients with high vs low expression of a gene?](/user-guide/faq.md#how-can-i-compare-outcomes-in-patients-with-high-vs-low-expression-of-a-gene)
* [Results View](/user-guide/faq.md#results-view)
    * [OncoPrint](/user-guide/faq.md#oncoprint)
        * [What are OncoPrints?](/user-guide/faq.md#what-are-oncoprints)
        * [Can I change the order of genes in the OncoPrint?](/user-guide/faq.md#can-i-change-the-order-of-genes-in-the-oncoprint)
        * [Can I visualize my own data within an OncoPrint?](/user-guide/faq.md#can-i-visualize-my-own-data-within-an-oncoprint)
        * [Why are some samples ‚ÄúNot profiled‚Äù for certain genes?](/user-guide/faq.md#why-are-some-samples-not-profiled-for-certain-genes)
    * [Other pages](/user-guide/faq.md#other-pages)
        * [Does the Mutual Exclusivity tab calculate its statistics using all samples/alterations or only a specific subset?](/user-guide/faq.md#does-the-mutual-exclusivity-tab-calculate-its-statistics-using-all-samplesalterations-or-only-a-specific-subset)
        * [What are the values of the box and whiskers in a boxplot?](/user-guide/faq.md#what-are-the-values-of-the-box-and-whiskers-in-a-boxplot)
* [Study View](/user-guide/faq.md#study-view)
    * [How to use filter in the URL of Study View page?](/user-guide/faq.md#how-to-use-filter-in-the-url-of-study-view-page)
* [What if I have other questions or comments?](/user-guide/faq.md#what-if-i-have-other-questions-or-comments)

## General Questions
#### What is the cBioPortal for Cancer Genomics?
The cBioPortal for Cancer Genomics is an open-access, open-source resource for interactive exploration of multidimensional cancer genomics data sets. The goal of cBioPortal is to significantly lower the barriers between complex genomic data and cancer researchers by providing rapid, intuitive, and high-quality access to molecular profiles and clinical attributes from large-scale cancer genomics projects, and therefore to empower researchers to translate these rich data sets into biologic insights and clinical applications.
#### How do I get started?
Check out our [tutorial slides](https://www.cbioportal.org/tutorials) to get started or go through our [tutorial paper](https://www.ncbi.nlm.nih.gov/pubmed/23550210).
#### What data types are in the portal?
The portal supports and stores non-synonymous mutations, DNA copy-number data (putative, discrete values per gene, e.g. "deeply deleted" or "amplified", as well as log2 or linear copy number data), mRNA and microRNA expression data, protein-level and phosphoprotein level data (RPPA or mass spectrometry based), DNA methylation data, and de-identified clinical data. For a complete breakdown of available data types per cancer study go to the [Data Sets Page](https://www.cbioportal.org/datasets). Note that for many studies, only somatic mutation data and limited clinical data are available. For TCGA studies, the other data types are also available. Germline mutations are supported by cBioPortal, but are, with a few exceptions, not available in the public instance.

#### What does ___ stand for?
Here are the meanings of some of the abbreviations used by cBioPortal:
* VUS: variant of unknown significance
* CNA: copy number alteration
* AMP: amplification
* HOMDEL: deep deletion
* TMB: tumor mutational burden, calculated as mutations per megabase of sequenced DNA
* KM: Kaplan-Meier
* MSI: microsatellite instability
* OQL: Onco Query Language, used within cBioPortal to define the types of alterations included in a query. For more on OQL, review the [documentation](https://www.cbioportal.org/oql), [tutorial slides](https://www.cbioportal.org/tutorials#oql), and [videos](https://www.cbioportal.org/tutorials)

#### What is the process of data curation?
The TCGA firehose legacy datasets are imported directly from the original TCGA Data Coordinating Center via the [Broad Firehose](https://gdac.broadinstitute.org/).

We are also actively curating datasets from the literature. Studies from the literature were curated from the data published with the manuscripts. We sometimes reach out to the investigators to acquire additional data, such as clinical attributes. All mutation calls (in VCF or MAF format) are processed through an internal pipeline to annotate the variant effects in a consistent way across studies. Please [contact us](mailto:cbioportal@googlegroups.com) to suggest additional public datasets to curate or view the list of studies suggested for curation in our [Datahub on Github](https://github.com/cBioPortal/datahub/issues?q=is%3Aissue+is%3Aopen+label%3A%22new+public+study%22).
#### How do I get updates on new portal developments and new data sets?
Please subscribe to our low-volume [news mailing list](https://groups.google.com/group/cbioportal-news) or follow [@cbioportal on Twitter](https://twitter.com/cbioportal).
#### Does the portal work on all browsers and operating systems?
We support and test on the following web browsers: Safari, Google Chrome, Firefox and Edge. (As of release v3.5.4 we no longer support Internet Explorer 11). If you notice any incompatibilities, please let us know.
#### How do I cite the cBioPortal?
Please cite the following portal papers:
* Cerami et al. The cBio Cancer Genomics Portal: An Open Platform for Exploring Multidimensional Cancer Genomics Data. Cancer Discovery. May 2012 2; 401. [PubMed](https://www.ncbi.nlm.nih.gov/pubmed/22588877).
* Gao et al. Integrative analysis of complex cancer genomics and clinical profiles using the cBioPortal. Sci. Signal. 6, pl1 (2013). [PubMed](https://www.ncbi.nlm.nih.gov/pubmed/23550210).
* de Bruijn et al. Analysis and Visualization of Longitudinal Genomic and Clinical Data from the AACR Project GENIE Biopharma Collaborative in cBioPortal. Cancer Res (2023). [PubMed](https://pubmed.ncbi.nlm.nih.gov/37668528/).

Remember also to cite the source of the data if you are using a publicly available dataset.
#### Can I use figures from the cBioPortal in my publications or presentations?
Yes, you are free to use any of the figures from the portal in your publications or presentations (many are available in SVG or PDF format for easier scaling and editing). When you do, please cite Cerami et al., Cancer Discov. 2012, and Gao et al., Sci. Signal. 2013 (see the previous question for full citations).
#### Can I save or bookmark my results in cBioPortal?
You can bookmark your query results and share the URL with collaborators. We store all queries via Session IDs, and these are saved indefinitely. Use the bookmark tab to retrieve the full link, or generate a short link via the bit.ly link generator.
#### How is the cBioPortal for Cancer Genomics different from the Genomic Data Commons (GDC)?
The cBioPortal is an exploratory analysis tool for exploring large-scale cancer genomic data sets that hosts data from large consortium efforts, like [TCGA](https://cancergenome.nih.gov/) and [TARGET](https://ocg.cancer.gov/programs/target), as well as publications from individual labs. You can quickly view genomic alterations across a set of patients, across a set of cancer types, perform survival analysis and perform group comparisons. If you want to explore specific genes or a pathway of interest in one or more cancer types, the cBioPortal is probably where you want to start.

By contrast, the [Genomic Data Commons (GDC)](https://gdc.cancer.gov/) aims to be the definitive place for full-download and access to all data generated by TCGA and TARGET. If you want to download raw mRNA expression files or full segmented copy number files, the GDC is probably where you want to start.
#### Does the cBioPortal provide a Web Service API? R interface? MATLAB interface?
Yes, the cBioPortal provides a [Swagger API](https://www.cbioportal.org/api/swagger-ui.html), and [R/MATLAB interfaces](/web-API-and-Clients.md#r-client).
#### Can I use cBioPortal with my own data?
cBioPortal provides several options for analyzing your own data. Visit our [Visualize Your Data](https://www.cbioportal.org/visualize) page to generate an OncoPrint or Lollipop Plot with your own data. To utilize the entire suite of analysis and visualization tools, you can also install your own instance of cBioPortal (see next question).
#### Can I create a local instance of cBioPortal to host my own data?
Yes, the cBioPortal is open-source, and available on [GitHub](https://github.com/cBioPortal/cbioportal). Our [documentation](/) provides complete download and installation instructions.
#### I'd like to contribute code to the cBioPortal. How do I get started?
Great! We would love to have your contributions. To get started, head over to our GitHub repository and check out our page on [how to contribute](https://github.com/cBioPortal/cbioportal/blob/master/CONTRIBUTING.md).
#### What is a combined Study?
A combined study is a custom study comprised of samples from multiple studies. The combined study feature 
enables you to combine samples from multiple studies to form a bigger study. This cohort of samples can then be queried or explored just like a traditional study, and can be returned to at a later date or shared with a collaborator. 
#### How can I form a combined Study?
A combined or merged study is a custom study comprised of samples from multiple studies. In the homepage of cbioportal, studies can be selected using the checkbox located on the left of the study. Once the studies are selected, they can be combined and explored using the "Explore Selected Studies" button. Alternatively, after the studies are selected, you can run queries on the combined study using the "Query by Gene" button.
#### How can I create a subset or sub cohort of a study with specific samples or patients?
A subset or sub cohort of a study can be created by specifying individual patients or samples. After a study is selected, user can click on the "Custom selection" button to create a new filter by specifying the sampleID or patientID that the user is interested to explore. Another way is to filter a set of patients using the charts on the study view and then view the IDs of the patients and samples that were selected or unselected based on the current filter.
#### What is a Virtual Study?
A virtual study is a custom study comprised of samples from one or more existing studies. The virtual study feature allows you to define a custom cohort of samples that fit your specific genomic or clinical criteria of interest. These samples can be a subset of the data available in an existing study, or result from the combination of multiple existing studies. This cohort of samples can then be queried or explored just like a traditional study, and can be returned to at a later date or shared with a collaborator. For more information and examples, see our [tutorial on virtual studies](https://www.cbioportal.org/tutorials).
#### Is it necessary to log in to use virtual studies? If I do log in, what additional functionality do I gain?
No. A user that has not logged in can create virtual studies and run queries in those studies (by using the query box on the study summary page). Links to virtual studies are permanent, so you can save the link on your computer and come back to it anytime, or share it with others.

If you log in, you gain the ability to save your virtual study to the list of existing studies on the homepage. This makes a virtual study functionally the same as any other study: you can access your virtual studies in the query builder and you can combine an existing virtual study with any other study to create a new virtual study.
#### What is Group Comparison?
Group Comparison is a suite of analysis features which allows a user to compare clinical or genomic features of user-defined groups of samples. These groups can be defined based on any clinical or genomic features. For an overview, see [our tutorial on group comparison](https://www.cbioportal.org/tutorials).
#### What are the statistical significance tests in Group Comparison?
- Survival tab: Log-rank test
- Clinical tab:
    - Continuous data: Chi-squared test
    - Categorical data: Kruskal Wallis test
- Other tabs
    - 2 groups
        - Continuous data: one-sided t-test
        - Categorical data: two-sided Fisher's exact test
    - 3 or more groups
        - Continuous data: one-way ANOVA
        - Categorical data: Chi-squared test

## Data Questions
### General Data
#### Does the portal contain cancer study X?
Check out the [Data Sets Page](https://www.cbioportal.org/datasets) for the complete set of cancer studies currently stored in the portal. If you do not see your specific cancer study of interest, please [contact us](mailto:cbioportal@googlegroups.com), and we will let you know if it's in the queue.
#### Which resources are integrated for variant annotation?
cBioPortal supports the annotation of variants from several different databases. These databases provide information about the recurrence of, or prior knowledge about, specific amino acid changes. For each variant, the number of occurrences of mutations at the same amino acid position present in the COSMIC database are reported. Furthermore, variants are annotated as ‚Äúhotspots‚Äù if the amino acid positions were found to be recurrent linear hotspots, as defined by the Cancer Hotspots method ([cancerhotspots.org](https://www.cancerhotspots.org/)), or three-dimensional hotspots, as defined by 3D Hotspots ([3dhotspots.org](https://www.3dhotspots.org/)). Prior knowledge about variants, including clinical actionability information, is provided from three different sources: OncoKB ([www.oncokb.org](https://www.oncokb.org/)), CIViC ([civicdb.org](https://civicdb.org/)), as well as My Cancer Genome ([mycancergenome.org](https://www.mycancergenome.org/)). For OncoKB, exact levels of clinical actionability are displayed in cBioPortal, as defined by [the OncoKB paper](https://ascopubs.org/doi/full/10.1200/PO.17.00011).
#### What version of the human reference genome is being used in cBioPortal?
The [public cBioPortal](https://www.cbioportal.org) is currently using hg19/GRCh37.
#### How does cBioPortal handle duplicate samples or sample IDs across different studies?
The cBioPortal generally assumes that samples or patients that have the same ID are actually the same. This is important for cross-cancer queries, where each sample should only be counted once. If a sample is part of multiple cancer cohorts, its alterations are only counted once in the Mutations tab (it will be listed multiple times in the table, but is only counted once in the lollipop plot). However, other tabs (including OncoPrint and Cancer Types Summary) will count the sample twice - for this reason, we advise against querying multiple studies that contain the same samples (e.g., TCGA PanCancer Atlas and TCGA Firehose Legacy).
#### Are there any normal tissue samples available through cBioPortal?
No, we currently do not store any normal tissue data in our system.
#### How can I find which studies have mRNA expression data (or any other specific data type)?
Check out the [Data Sets Page](https://www.cbioportal.org/datasets) where you can view the complete set of cancer studies and sort by the number of samples with data available for any data type.
#### Can I download all data at once?
You can download all data for individual studies on the [Data Sets Page](https://www.cbioportal.org/datasets) or the study view page for the study of interest. You can also download all studies from our [Data Hub](https://github.com/cBioPortal/datahub).
#### The data today is different than the last time I looked. What happened?
We do occasionally update existing datasets to provide the most up-to-date, accurate and consistent data possible. The data you see today is likely an improved version of what you have seen previously. However, if you suspect that there is an error in the current version, please let us know at [cbioportal@googlegroups.com](mailto:cbioportal@googlegroups.com).

If you need to reference an old version of a dataset, you can find previous versions in our [Datahub repository](https://github.com/cBioPortal/datahub/tree/master/public).
#### How do I access data from AACR Project GENIE?
Data from AACR Project GENIE are provided in a [dedicated instance of cBioPortal](https://www.cbioportal.org/genie/). You can also download GENIE data from the [Synapse Platform](https://synapse.org/genie). Note that you will need to register before accessing the data. Additional information about AACR Project GENIE can be found on the [AACR website](https://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx).

### TCGA
#### How does TCGA data in cBioPortal compare to TCGA data in Genome Data Commons?
We do not currently load the mutation data from the GDC. Instead, we have the original mutation data generated by the individual TCGA sequencing centers. The source of the data is the Broad Firehose (or the publication pages for data that matches a specific manuscript). These data are usually a combination of two mutation callers, but they differ by center (typically a variant caller like MuTect plus an indel caller), and sequencing centers have modified their mutation calling pipelines over time.
#### What happened to TCGA Provisional datasets?
We renamed TCGA Provisional datasets to TCGA Firehose Legacy to better reflect that this data comes from a legacy processing pipeline. The exact same data is now available in TCGA Firehose Legacy studies.
#### What are TCGA Firehose Legacy datasets and how do they compare to the publication-associated datasets and the PanCancer Atlas datasets?
The Firehose Legacy dataset (formerly Provisional datasets) for each TCGA cancer type contains all data available from the Broad Firehose. The publication datasets reflect the data that were used for each of the publications. The samples in a published dataset are usually a subset of the firehose legacy dataset, since manuscripts were often written before TCGA completed their goal of sequencing 500 tumors.

There can be differences between firehose legacy and published data. For example, the mutation data in the publication usually underwent more QC, and false positives might have been removed or, in rare cases, false negatives added. RNA-Seq and copy-number values may also differ slightly, as different versions of analysis pipelines could have been used. Additionally, due to additional curation during the publication process, the clinical data for the publication may be of higher quality or may contain a few more data elements, sometimes derived from the genomic data (e.g., genomic subtypes).

The TCGA PanCancer Atlas datasets derive from an effort to unify TCGA data across all tumor types. Publications resulting from this effort can be found at the [TCGA PanCancer Atlas site](https://www.cell.com/pb-assets/consortium/pancanceratlas/pancani3/index.html). In the cBioPortal, data from the PanCancer Atlas is divided by tumor type, but these studies have uniform clinical elements, consistent processing and normalization of mutations, copy number, mRNA data and are ideally processed for comparative analyses.
#### Where do the thresholded copy number call in TCGA Firehose Legacy data come from?
Thresholded copy number calls in the TCGA Firehouse Legacy datasets are generated by the GISTIC 2.0 algorithm and obtained from the Broad Firehose.
#### Which studies have MutSig and GISTIC results? How do these results compare to the data in the TCGA publications?
MutSig and GISTIC results about the statistical significance of recurrence of mutations and copy-number alterations in specific genes are available for many TCGA studies. The MutSig and GISTIC results reported in cBioPortal are based on the same mutations and copy number data reported in each TCGA publication, or the Broad Firehose for the firehose legacy data sets. However, the publication may or may not have included the complete MutSig and GISTIC output, and therefore there may be some discrepancies between the publication and the data in cBioPortal.
#### How can I download the PanCancer Atlas data?
PanCancer Atlas data can be downloaded on a study-by-study basis from cBioPortal through the [Datasets page](https://www.cbioportal.org/datasets) or our [DataHub](https://github.com/cBioPortal/datahub/tree/master/public). To download all cancer types together, try the [Genomic Data Commons PanCancer Atlas page](https://gdc.cancer.gov/about-data/publications/pancanatlas).

### DNA (Mutations, Copy Number & Fusions)
#### Does the cBioPortal contain synonymous mutation data?
No, the cBioPortal does not currently support synonymous mutations. This may change in the future, but we have no plans yet to add this feature.
#### What processing or filtering is applied to generate the mutation data?
Within cBioPortal, we utilize the mutation calls as provided by each publication. We do not perform any additional filtering. The only processing we do is to standardize the annotation of the mutations using [Genome Nexus](https://genomenexus.org) (which utilizes [VEP](https://useast.ensembl.org/info/docs/tools/vep/index.html) with the [canonical MSKCC transcript](https://github.com/genome-nexus/genome-nexus-importer/blob/master/data/common_input/isoform_overrides_at_mskcc_grch37.txt)). Read more about the transcript assignments [here](https://docs.cbioportal.org/mutation-data-transcript-annotation/#transcript-assignment). For specifics of which tools were used to call mutations and filters that may have been applied, refer to the publication manuscript.
#### What transcripts are used for annotating mutations?
Prior to loading a study into cBioPortal, we run all mutation data through a standard pipeline (see above), which re-annotates all mutations to the [canonical MSKCC transcript](https://github.com/genome-nexus/genome-nexus-importer/blob/master/data/common_input/isoform_overrides_at_mskcc_grch37.txt). Read more about the transcript assignments [here](https://docs.cbioportal.org/mutation-data-transcript-annotation/#transcript-assignment).
#### How are protein domains in the mutational lollipop diagrams specified?
Protein domain definitions come from [PFAM](https://pfam.xfam.org/).
#### What is the difference between a ‚Äúsplice site‚Äù mutation and a ‚Äúsplice region‚Äù mutation?
A ‚Äúsplice site‚Äù mutation occurs in an intron, in a splice acceptor or donor site (2bp into an intron adjacent to the intron/exon junction), defined by [Sequence Ontology](https://www.sequenceontology.org/browser/current_svn/term/SO:0001629). ‚ÄúSplice region‚Äù mutations are mutations that occur near the intron/exon junction, defined by [Sequence Ontology](https://www.sequenceontology.org/browser/current_svn/term/SO:0001630). While synonymous mutations are generally excluded from cBioPortal, these ‚Äúsplice region‚Äù synonymous mutations are included due to their potential impact on splicing.
#### What do ‚ÄúAmplification‚Äù, ‚ÄúGain‚Äù, ‚ÄúDeep Deletion‚Äù, ‚ÄúShallow Deletion‚Äù and "-2", "-1", "0", "1", and "2" mean in the copy-number data?
These levels are derived from copy-number analysis algorithms like GISTIC or RAE, and indicate the copy-number level per gene:
* -2 or Deep Deletion indicates a deep loss, possibly a homozygous deletion
* -1 or Shallow Deletion indicates a shallow loss, possibley a heterozygous deletion
* 0 is diploid
* 1 or Gain indicates a low-level gain (a few additional copies, often broad)
* 2 or Amplification indicate a high-level amplification (more copies, often focal)

Note that these calls are putative. We consider the deep deletions and amplifications as biologically relevant for individual genes by default. Note that these calls are usually not manually reviewed, and due to differences in purity and ploidy between samples, there may be false positives and false negatives.
#### What is GISTIC? What is RAE?
Copy number data sets within the portal are often generated by the [GISTIC](https://www.ncbi.nlm.nih.gov/sites/entrez?term=18077431) or [RAE](https://www.ncbi.nlm.nih.gov/sites/entrez?term=18784837) algorithms. Both algorithms attempt to identify significantly altered regions of amplification or deletion across sets of patients. Both algorithms also generate putative gene/patient copy number specific calls, which are then input into the portal.

For TCGA studies, the table in allthresholded.bygenes.txt (which is the part of the GISTIC output that is used to determine the copy-number status of each gene in each sample in cBioPortal) is obtained by applying both low- and high-level thresholds to to the gene copy levels of all the samples. The entries with value +/- 2 exceed the high-level thresholds for amplifications/deep deletions, and those with +/- 1 exceed the low-level thresholds but not the high-level thresholds. The low-level thresholds are just the 'ampthresh' and 'delthresh' noise threshold input values to GISTIC (typically 0.1 or 0.3) and are the same for every thresholds.

By contrast, the high-level thresholds are calculated on a sample-by-sample basis and are based on the maximum (or minimum) median arm-level amplification (or deletion) copy number found in the sample. The idea, for deletions anyway, is that this level is a good approximation for hemizygous losses given the purity and ploidy of the sample. The actual cutoffs used for each sample can be found in a table in the output file sample_cutoffs.txt. All GISTIC output files for TCGA are available at: gdac.broadinstitute.org.

### RNA
#### Does the portal store raw or probe-level data?
No, the portal only contains gene-level data. Data for different isoforms of a given gene are merged. Raw and probe-level data for data sets are available via [NCBI GEO](https://www.ncbi.nlm.nih.gov/geo/), [dbGaP](https://www.ncbi.nlm.nih.gov/gap/) or through the [GDC](https://portal.gdc.cancer.gov/). See the cancer type description on the main query page or refer to the original publication for links to the raw data.
#### What are mRNA and microRNA Z-Scores?
For mRNA and microRNA expression data, we typically compute the relative expression of an individual gene in a tumor sample to the gene's expression distribution in a reference population of samples. That reference population is all profiled samples (by default for mRNA), or normal samples (when specified), or all samples that are diploid for the gene in question (discontinued).  The returned value indicates the number of standard deviations away from the mean of expression in the reference population (Z-score). The normalization method is described [here](https://github.com/cBioPortal/cbioportal/blob/master/docs/Z-Score-normalization-script.md). Please note that the expression results by querying a gene with the default setting (z-score threshold of 2) oftentimes are not meaningful. Since the z-scores were usually calculated compared to other tumor samples, high or low expression does not necessarily mean that the gene is expressed irregularly in tumors. The data is useful for correlation analysis, for example, pick a threshold based on overall expression (using Plots tab) and compare survival data between expression high and low groups.
#### Is there any normal RNA-seq data in cBioPortal?
We have RNASeqV2 mRNA expression data for normal samples of 16 TCGA PanCan Atlas Cohorts. The data was curated from [GDC](https://gdc.cancer.gov/about-data/publications/pancanatlas), and can be downloaded from our [Datahub](https://github.com/cBioPortal/datahub) or [Data Set](https://www.cbioportal.org/datasets) page. This data is not directly queriable in portal; they are only used as reference data for calculating the "relavtive to normal expression z-score" profile. [Example: ERBB2 expression z-scores relative to normal expression](https://www.cbioportal.org/results/plots?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&case_set_id=all&data_priority=0&gene_list=ERBB2&geneset_list=%20&plots_coloring_selection=%7B%22colorByCopyNumber%22%3A%22true%22%7D&plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_STUDY%22%7D&plots_vert_selection=%7B%22selectedGeneOption%22%3A2064%2C%22dataType%22%3A%22MRNA_EXPRESSION%22%2C%22selectedDataSourceOption%22%3A%22rna_seq_v2_mrna_median_all_sample_ref_normal_Zscores%22%2C%22logScale%22%3A%22true%22%7D&profileFilter=0&tab_index=tab_visualize). 
#### How is TCGA RNASeqV2 processed? What units are used?
RNASeqV2 from TCGA is processed and normalized using [RSEM](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323). Specifically, the RNASeq V2 data in cBioPortal corresponds to the rsem.genes.normalized_results file from TCGA. A more detailed explanation of RSEM output can be found [here](https://www.biostars.org/p/106127/). cBioPortal then calculates z-scores as described above in [What are mRNA and microRNA Z-Scores?](/user-guide/faq.md#what-are-mrna-and-microrna-z-scores)
#### Is there microRNA data?
We have microRNA data for only a few studies and they are not up to date. To download more updated miRNA data, please go to either [Broad Firehose](https://gdac.broadinstitute.org/), or [GDC](https://portal.gdc.cancer.gov/).
#### How can I query microRNAs in the portal?
You can input either precursor or mature miRNA IDs. Since one precursor ID may correspond to multiple mature IDs and vise versa, the portal creates one internal ID for each pair of precursor ID and mature ID mapping. For example, an internal ID of MIR-29B-1/29B stands for precursor microRNA hsa-mir-29b-1 and mature microRNA hsa-miR-29b. After entering a precursor or mature ID, you will be asked to select one internal ID for query and that internal ID will also be displayed in the Oncoprint.

### Protein
#### How can I query phosphoprotein levels in the portal?
You need to input special IDs for each phosphoprotein/phopshosite such as AKT1_pS473 (which means AKT1 protein phosphorylated at serine residue at position 473). You could also input aliases such as phosphoAKT1 or phosphoprotein, and the portal will ask you to select the phosphoprotein/phosphosite of your interest. Note that phosphoprotein data is only available for select studies and for a limited number of proteins / phosphorylation sites.
#### Why isn‚Äôt there protein data for my gene of interest?
Most of the protein expression data in cBioPortal comes from assays like RPPA which only interrogate a subset of all proteins. TCGA ovarian, breast, and colorectal firehose legacy studies also have mass-spectrometry-based proteomics data from [CPTAC](https://proteomics.cancer.gov/programs/cptac) which cover more genes/proteins.
### DNA Methylation
#### Which methylation probe is used for genes with multiple probes?
For genes with multiple probes (usually from the Infinium arrays), we only include methylation data from the probe with the strongest negative correlation between the methylation signal and the gene's expression in the study (TCGA only).

### Clinical Data
#### What kind of clinical data is stored in the portal?
The portal currently stores de-identified clinical data, such as gender, age, tumor type, tumor grade, overall and disease-free survival data, when available. The available clinical data will differ from study to study.
#### What is the meaning of OS_STATUS / OS_MONTHS, and PFS_STATUS / PFS_MONTHS?
OS_STATUS means overall survival status ("0" -> "living" or "1" -> "deceased") and OS_MONTHS indicates the number of months from time of diagnosis to time of death or last follow up. PFS refers to ‚Äúprogression free survival‚Äù, indicating whether patient‚Äôs disease has recurred/progressed (PFS_STATUS), and at what time the disease recurred or the patient was last seen (PFS_MONTHS).

## Analysis Questions
#### How can I query/explore a select subset of samples?
cBioPortal allows you to run a query or explore study view using a user-specified list of samples/patients.

The first step is to define your sample set. There are two slightly different approaches you can take to defining your sample set, depending on whether you are selecting based on a positive criteria (samples with TP53 mutations) or a negative criteria (samples without a KRAS mutation).

Let‚Äôs take the positive criteria example first. Run a query for TP53 mutations using OQL (TP53: MUT) in your study of interest. Click over to the ‚ÄúDownload‚Äù tab. In the table at the top, find the row that starts with ‚ÄúSamples affected‚Äù, and either Copy or Download that list. This is your list of samples that have a TP53 mutation.

Now for the negative criteria example. This also begins by using OQL to run a query for KRAS mutations (KRAS: MUT) in your study of interest. Click over to the ‚ÄúDownload‚Äù tab. Look at the table at the top again, but this time find the row that starts with ‚ÄúSample matrix‚Äù. Copy or download this data and open it in Excel. You will see a two column table that indicates whether a given sample is altered or not, indicated by 0 or 1. Sort by the second column and then copy all the sample IDs from the first column that have a 0 in the second column. This is your list of samples that do not have a KRAS mutation.

With a sample list in hand, you can now either run a query in just the selected samples (select ‚ÄúUser-defined Case List‚Äù in the ‚ÄúSelect Patient/Case Set:‚Äù dropdown) or explore this set of patients in study view (click ‚ÄúSelect cases by IDs‚Äù and then create a Virtual Study restricted to just those samples).

For more information about OQL, see the [specification page](/user-guide/oql.md) or view the [tutorial slides](https://www.cbioportal.org/tutorials#oql). For more information about virtual studies, read [this FAQ](/user-guide/faq.md#what-is-a-virtual-study) or view the [tutorial slides](https://www.cbioportal.org/tutorials).
#### How can I compare two or more subsets of samples?
cBioPortal has a suite of analysis tools to enable comparisons between user-defined groups of samples/patients. For an overview of this functionality, see our [tutorial on group comparison](https://www.cbioportal.org/tutorials).
#### Is it possible to determine if a particular mutation is heterozygous or homozygous in a sample? When a sample has 2 mutations in one gene, is it possible to determine whether the mutations are in cis or in trans with each other?
There is currently no way to definitively determine whether a mutation is heterozygous/homozygous or in cis/trans with another mutation. However, you can try to infer the status of mutations by noting the copy number status of the gene and the variant allele frequency of the mutation(s) of interest relative to other mutations in the same sample. The cBioPortal patient/sample view can help you accomplish this.

Specifically in the case of TCGA samples with two mutations in the same gene, you can also obtain access to the aligned sequencing reads from the [GDC](https://portal.gdc.cancer.gov/) and check if the mutations are in cis or in trans (if the mutations are close enough to each other).
#### How can I query over/under expression of a gene?
cBioPortal supports Onco Query Language (OQL) which can be used to query over/under expression of a gene. When writing a query, select an mRNA expression profile. By default, samples with expression z-scores >2 or <-2 in any queried genes are considered altered. Alternate cut-offs can be defined using OQL, for example: "EGFR: EXP>2" will query for samples with an EGFR expression z-score >2. Review for the OQL [specification page](/user-guide/oql.md) or [tutorial slides](https://www.cbioportal.org/tutorials#oql) for more specifics and examples.
#### How can I compare outcomes in patients with high vs low expression of a gene?
To compare outcomes in patients with high vs low expression of a gene (excluding those patients with intermediate levels of expression), we will follow a 2 step process that builds on the approach described above in [How can I query/explore a select subset of samples?](/user-guide/faq.md#how-can-i-queryexplore-a-select-subset-of-samples), utilizing [OQL](/user-guide/oql.md) to first identify and then stratify that cases of interest.

First, identify the sample set using OQL. For example, to stratify patients based on expression of EGFR, add an mRNA profile to the query, and write "EGFR: EXP>2 EXP<-2" in the gene set box. After running the query, go to the Download tab and copy/download the ‚ÄúSamples affected‚Äù list.

Second, return to the homepage and paste the list of sample IDs from the previous step into the ‚ÄúUser-defined Case List‚Äù in the ‚ÄúSelect Patient/Case Set:‚Äù dropdown. This query will now only look at samples with high or low expression. To now stratify into high vs low for survival analysis, enter "EGFR: EXP>2" in the gene set box (don‚Äôt forget to select the same mRNA profile). Run the query and click over to the Survival tab. The ‚Äúcases with alteration‚Äù are patients with high expression of EGFR and the cases without alteration are those with low expression of EGFR.

We use 2 and -2 as example thresholds above, but it is also a good idea to look at the distribution of expression data and select a threshold based on that. Plots tab can be useful for analyzing the expression distribution.
## Results View
### OncoPrint
#### What are OncoPrints?
OncoPrints are compact means of visualizing distinct genomic alterations, including somatic mutations, copy number alterations, and mRNA expression changes across a set of cases. They are extremely useful for visualizing gene set and pathway alterations across a set of cases, and for visually identifying trends, such as trends in mutual exclusivity or co-occurrence between gene pairs within a gene set. Individual genes are represented as rows, and individual cases or patients are represented as columns.

![image](https://cloud.githubusercontent.com/assets/840895/24209131/d34e33b2-0efb-11e7-945a-fd2b3c66d195.png)

#### Can I change the order of genes in the OncoPrint?
By default, the order of genes in the OncoPrint will be the same as in your query. You can change the order by (a) clicking on the gene name and dragging it up/down or (b) clicking on the three vertical dots next to the gene name to move the gene up/down.
#### Can I visualize my own data within an OncoPrint?
Yes, check out the OncoPrinter tool on our [Visualize Your Data page](https://www.cbioportal.org/visualize).
#### Why are some samples ‚ÄúNot profiled‚Äù for certain genes?
Some studies include data from one or more targeted sequencing platforms which do not include all genes. For samples sequenced on these smaller panels, cBioPortal will indicate that a particular gene was not included on the sequencing panel used for that sample. Alteration frequency calculations for each gene also take this information into account. Hover over a sample in OncoPrint to see the gene panel name, and click on that gene panel name to view a list of the genes included on that panel.
### Other pages
#### Does the Mutual Exclusivity tab calculate its statistics using all samples/alterations or only a specific subset?
The calculations on the Mutual Exclusivity tab are performed using all samples included in the query. A sample is defined as altered or unaltered for each gene based on the [OQL](/user-guide/oql.md) utilized in the query - by default, this will be non-synonymous mutations, fusions, amplifications and deep deletions.
#### What are the values of the box and whiskers in a boxplot?
In boxplots on cBioPortal, the box is drawn from the 25th percentile (Q1) to the 75th percentile (Q3), with the horizontal line in between representing the median. Whiskers are drawn independently above and below the box, and will extend to the maximum or minimum data values, unless there are outlier values, in which case the whisker will extend to 1.5 * IQR (interquartile range = Q3-Q1). Outliers are defined as values that extend beyond 1.5 * IQR.

## Study View
### How to use filter in the URL of Study View page?
You can filter the study based on values of one attribute in the URL. For example, <https://www.cbioportal.org/study/summary?id=msk_impact_2017#filterJson={"clinicalDataFilters":[{"attributeId":"CANCER_TYPE","values":[{"value":"Melanoma"}]}]}>

filterJson is set in the url hash string. Here are the allowed parameters and format for it in filterJson:
```
{
	"caseLists": [
		["string"]
	],
	"clinicalDataFilters": [{
		"attributeId": "string",
		"values": [{
			"end": 0,
			"start": 0,
			"value": "string"
		}]
	}],
	"geneFilters": [{
		"geneQueries": [
			["string"]
		],
		"molecularProfileIds": ["string"]
	}],
	"genomicDataFilters": [{
		"hugoGeneSymbol": "string",
		"profileType": "string",
		"values": [{
			"end": 0,
			"start": 0,
			"value": "string"
		}]
	}],
	"genomicProfiles": [
		["string"]
	],
	"sampleIdentifiers": [{
		"sampleId": "string",
		"studyId": "string"
	}],
	"studyIds": ["string"]
}
```

## What if I have other questions or comments?
Please contact us at [cbioportal@googlegroups.com](mailto:cbioportal@googlegroups.com). Previous discussions about cBioPortal are available on the [user discussion mailing list](https://groups.google.com/group/cbioportal).
# New Users

Are you new to cBioPortal? Welcome! We have a few options to help you get started.

If you have an hour, we highly recommend watching the recording of our Introduction to cBioPortal webinar ([youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo) or [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp)).

Don't have an hour? Review our tutorial slides for exploring a study ([Google slides](https://docs.google.com/presentation/d/1_OGK69lO4Z62WaxHHkNYmWvY0LQN2v0slfaLyY1_IQ0) or [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%201%20Single%20Study%20Exploration.pdf)) and running a query ([Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) or [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf))

Or, watch two of our short how-to videos which demonstrate how to explore a study ([youtube.com](https://www.youtube.com/watch?v=N8ffDgkqDWc)) and how to run a query ([youtube.com](https://www.youtube.com/watch?v=MH-kY5usA70)).
# Onco Query Language (OQL)

## Introduction to OQL
The Onco Query Language (OQL) is used to define which specific types of alterations are included in a query on the cBioPortal for Cancer Genomics. By default, querying for a gene includes mutations, fusions, amplifications and deep deletions. OQL can be used to specify specific mutations (e.g. BRAF V600E) or types of mutations (e.g. BRCA1 truncating mutations), lower level copy number alterations (e.g. CDKN2A shallow deletions), changes in mRNA or protein expression, and more.

OQL-specified alterations will be reflected on most tabs, including OncoPrint, but are not currently reflected on the Plots, Co-Expression or Expression tabs.

Note that OQL assumes any word that it doesn't recognize is a mutation code.

Additional explanation and examples using OQL are available in the [User Guide](https://docs.cbioportal.org/user-guide/by-page/#oql).

## OQL Keywords
Users can define specific subsets of genetic alterations for five data types:

Data Type | Keywords and Syntax | Default*
--------- | ------------------- | --------
Mutations | `MUT` All non-synonymous mutations <br> `MUT = <protein change>` Specific amino acid changes (e.g. `V600E` or `V600`) <br> `MUT = <mutation type>` Acceptable values are: `MISSENSE, NONSENSE, NONSTART, NONSTOP, FRAMESHIFT, INFRAME, SPLICE, TRUNC` | `MUT`
Fusions | `FUSION` All fusions (note that many studies lack fusion data) | `FUSION`
Copy Number Alterations | `AMP` Amplifications <br> `HOMDEL` Deep Deletions <br> `GAIN` Gains <br> `HETLOSS` Shallow Deletions <br> Comparison operators can also be used with `CNA` (e.g. `CNA >= GAIN` is the same as `AMP GAIN`) | `AMP` <br> `HOMDEL`
mRNA Expression | `EXP < -x` mRNA expression is less than `x` standard deviations (SD) below the mean <br> `EXP > x` mRNA expression is greater than `x` SD above the mean <br> The comparison operators `<=` and `>=` also work | `EXP >= 2` <br> `EXP <= -2`
Protein/phosphoprotein level | `PROT < -x` Protein expression is less than `x` standard deviations (SD) below the mean <br> `PROT > x` Protein expression is greater than `x` SD above the mean <br> The comparison operators `<=` and `>=` also work | `PROT >= 2` <br> `PROT <= -2`

\* These are the default OQL keywords used for each data type when a gene is queried without any explicit OQL.

### OQL modifiers
Mutations and copy number alterations can be further refined using modifiers:

Keyword | Applicable Data Type | Explanation
------- | -------------------- | -----------
`DRIVER` | Mutations <br> Fusions <br> Copy Number Alterations | Include only mutations, fusions and copy number alterations which are driver events, as defined in OncoPrint (default: OncoKB and CancerHotspots).
`GERMLINE` | Mutations | Include only mutations that are defined as germline events by the study.
`SOMATIC` | Mutations | Include all mutations that are not defined as germline.
`(a-b)` (protein position range) | Mutations | Include all mutations that overlap with the protein position range `a-b`, where `a` and `b` are integers. If you add a `*` (i.e. `(a-b*)`) then it will only include those mutations that are fully contained inside `a-b`. The open-ended ranges `(a-)` and `(-b)` are also allowed.

## Basic Usage
When querying a gene without providing any OQL specifications, cBioPortal will default to these OQL terms for a query with Mutation and Copy Number selected in the Genomic Profiles section:
`MUT FUSION AMP HOMDEL`

![image of basic query](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/basic_query.png)

You can see the OQL terms applied by hovering over the gene name in OncoPrint:

![image of basic query oncoprint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/basic_query_oncoprint.png)


If you select RNA and/or Protein in the "Genomic Profiles" section of the query, the default settings are:

RNA: `EXP >= 2 EXP <= -2`

Protein: `PROT >= 2 PROT <= -2`

![image of exp prot query oncoprint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/basic_query_oncoprint_exp_prot.png)

You must select the relevant Genomic Profile in order for OQL to query that data type. For example, you can't add `EXP > 2` to the query without also selecting an RNA profile.


Proper formatting for OQL is straightforward: gene name, followed by a colon, followed by any OQL keywords and ending in a semicolon, an end-of-line, or both.
```
GENE1: OQL KEYWORDS;
GENE2: OQL KEYWORDS
```
In general, any combination of OQL keywords and/or expressions can annotate any gene, and the order of the keywords is immaterial.

Below we will go into greater detail about each data type.

## Mutations
To view cases with specific mutations, provide the specific amino acid change of interest:
```
BRAF: MUT = V600E
```

You can also view all mutations at a particular position:
```
BRAF: MUT = V600
```

Or all mutations of a specific type:
```
TP53: MUT = <mutation type>
```

`<mutation type>` can be one or more of:
* `MISSENSE`
* `NONSENSE`
* `NONSTART`
* `NONSTOP`
* `FRAMESHIFT`
* `INFRAME`
* `SPLICE`
* `TRUNC`

For example, to view TP53 truncating mutations and in-frame insertions/deletions:
```
TP53: MUT = TRUNC INFRAME
```

OQL for mutations can also be written without `MUT =`. The following examples are identical:
```
BRAF: MUT = V600E
BRAF: V600E
TP53: MUT = TRUNC INFRAME
TP53: TRUNC INFRAME
```

OQL can also be used to exclude a specific protein change, position or type of mutation. For example, below are examples to query all EGFR mutations except T790M, all BRAF mutations except those at V600 and all TP53 mutations except missense:
```
EGFR: MUT != T790M
BRAF: MUT != V600
TP53: MUT != MISSENSE
```
Note that this will only work to exclude a single event. Because OQL uses 'OR' logic, excluding multiple mutations or excluding a mutation while including another mutation (e.g. `BRAF: MUT=V600 MUT!=V600E`) will result in querying all mutations.

## Copy Number Alterations
To view cases with specific copy number alterations, provide the appropriate keywords for the copy number alterations of interest. For example, to see amplifications:
```
CCNE1: AMP
```

Or amplified and gained cases:
```
CCNE1: CNA >= GAIN
```

Which can also be written as:
```
CCNE1: GAIN AMP
```

## Expression
High or low mRNA expression of a gene is determined by the number of standard deviations (SD) from the mean. For example, to see cases where mRNA for CCNE1 is greater than 3 SD above the mean:
```
CCNE1: EXP > 3
```

## Protein
High or low protein expression is similarly determined by the number of SD from the mean. For example, to see cases where protein expression is 2 SD above the mean:
```
EGFR: PROT > 2
```

Protein expression can also be queried at the phospho-protein level:
```
EGFR_PY992: PROT > 2
```

## Modifiers
Modifiers can be used on their own or in combination with other OQL terms for mutations, fusions and copy number alterations to further refine the query. Modifiers can be combined with other OQL terms using an underscore. The order in which terms are combined is immaterial.

### Driver
The `DRIVER` modifier applies to mutations, fusions and copy number alterations. The definition of what qualifies as a driver alteration comes from the "Mutation Color" menu in OncoPrint. By default, drivers are defined as mutations, fusions and copy number alterations in <a href="https://www.oncokb.org">OncoKB</a> or <a href="https://www.cancerhotspots.org">CancerHotspots</a>.

On its own, the `DRIVER` modifier includes driver mutations, fusions and copy number alterations:
```
EGFR: DRIVER
```

Or it can be used in combination with another OQL term. For example, to see only driver fusion events:
```
EGFR: FUSION_DRIVER
```

Or driver missense mutations:
```
EGFR: MUT = MISSENSE_DRIVER
```

When combining `DRIVER` with another OQL term, the order doesn't matter: `MUT_DRIVER` and `DRIVER_MUT` are equivalent. `DRIVER` can be combined with:
* `MUT`
* `MUT = <mutation type>` or `MUT = <protein change>`
* `FUSION`
* `CNA`
* `AMP` or `GAIN` or `HETLOSS` or `HOMDEL`
* `GERMLINE` or `SOMATIC` (see below)

### Germline/Somatic
The `GERMLINE` and `SOMATIC` modifiers only apply to mutations. A mutation can be explicitly defined as germline during the data curation process. Note that very few studies on the public cBioPortal contain germline data.

`GERMLINE` or `SOMATIC` can be combined with:
* `MUT`
* `MUT = <mutation type>` or `MUT = <protein change>`
* `DRIVER`

To see all germline BRCA1 mutations:
```
BRCA1: GERMLINE
```

Or to see specifically truncating germline mutations:
```
BRCA1: TRUNC_GERMLINE
BRCA1: GERMLINE_TRUNC
```
The order is immaterial; both options produce identical results.

Or to see somatic missense mutations:
```
BRCA1: MUT = MISSENSE_SOMATIC
``` 

`GERMLINE` or `SOMATIC` can also be combined with `DRIVER` and, optionally, a more specific mutation term (e.g. `NONSENSE`):
```
BRCA1: NONSENSE_GERMLINE_DRIVER
```

## The DATATYPES Command
To save copying and pasting, the `DATATYPES` command sets the genetic annotation for all subsequent genes. Thus,
```
DATATYPES: AMP GAIN HOMDEL EXP > 1.5 EXP < -1.5; CDKN2A MDM2 TP53
```

is equivalent to:
```
CDKN2A: AMP GAIN HOMDEL EXP > 1.5 EXP < -1.5
MDM2: AMP GAIN HOMDEL EXP > 1.5 EXP < -1.5
TP53: AMP GAIN HOMDEL EXP > 1.5 EXP < -1.5
```

## Merged Gene Tracks
OQL can be used to create a merged gene track in OncoPrint, in which alterations in multiple genes appear as a single track. This is done by enclosing a list of genes in square brackets. By default, the track will be labeled by the gene names, separated by '/'. To instead specify a label, type the desired label within double quotes at the beginning of the square brackets. For example:
```
["CDK INHIBITORS" CDKN2A CDKN2B]
[MDM2 MDM4]
```

The resulting merged gene track will be visible in OncoPrint and can be expanded to view the individual gene tracks. For example:

![Image of merged genes in OncoPrint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/merged_genes_oncoprint.png)

<https://www.cbioportal.org/results/oncoprint?session_id=5c1966e2e4b05228701f958e>

It is possible to include OQL for specific alterations in merged gene tracks, as well as querying a combination of single and merged gene tracks.

Note that merged gene tracks only appear in OncoPrint. All other pages show the individual genes.

## Example: RB Pathway Alterations

Provided below is one example of the power of using OQL. Additional examples are available in the [User Guide](https://docs.cbioportal.org/user-guide/by-page/#oql).

### Using the Defaults
Select Ovarian Serous Cystadenocarcinoma (TCGA, Nature 2011) with the following data types:
* Mutations
* Putative copy-number alterations (GISTIC)
* mRNA expression (mRNA expression Z-scores (all genes))

Input the following three genes in the RB pathway:
* CCNE1
* RB1
* CDKN2A

![image of rb query](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/rb_query.png)

Submit this query and note how many samples have alterations in multiple of these genes:

![image of rb oncoprint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/rb_query_oncoprint.png)

<https://www.cbioportal.org/results/oncoprint?session_id=5c1966cee4b05228701f958d>

### Greater Insight with OQL
Given what is known about the RB pathway, the events that are most likely selected for in the tumors are CCNE1 amplification, RB1 deletions or mutations, and loss of expression of CDKN2A. To investigate this hypothesis, we can use OQL to display only these events. Modify the query to reflect this:
```
CCNE1: AMP MUT
RB1: HOMDEL MUT
CDKN2A: HOMDEL EXP < -1
```

Examine the updated OncoPrint:

![image of modified rb oncoprint](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/images/OQL/rb_query_oncoprint_modified.png)

<https://www.cbioportal.org/results/oncoprint?session_id=5c1966aee4b05228701f958c>

This shows that alterations in these genes are almost entirely mutually-exclusive -- no cases are altered in all three genes and only six are altered in two genes. This supports the theory that the tumor has selected for these events.

## Questions? Feedback?
Please share any questions or feedback on OQL with us: <https://groups.google.com/group/cbioportal>

Also note that additional explanation and examples using OQL are available in the [User Guide](https://docs.cbioportal.org/user-guide/by-page/#oql).## Tutorials

### Tutorial Slides
These tutorial slides contain annoted screenshots to walk you through using the cBioPortal site.
1. Single Study Exploration [Google slides](https://docs.google.com/presentation/d/1_OGK69lO4Z62WaxHHkNYmWvY0LQN2v0slfaLyY1_IQ0) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%201%20Single%20Study%20Exploration.pdf)
2. Single Study Query [Google slides](https://docs.google.com/presentation/d/1y9UTIr5vHmsNVWqtGTVGgiuYX9wkK_a_RPNYiR8kYD8) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%202%20Single%20Study%20Query.pdf)
3. Patient View [Google slides](https://docs.google.com/presentation/d/1Jr_2yEfgjKBn4DBiXRk4kmhIbtsRp6gd0iD3k1fIUUk) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%203%20Patient%20View.pdf)
4. Virtual Studies [Google slides](https://docs.google.com/presentation/d/1rQE5rbFNdmup-rAtySHFxlLp3i4qa8SBA7MiQpMdn1I) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%204%20Virtual%20Studies.pdf)
5. Onco Query Language (OQL) [Google slides](https://docs.google.com/presentation/d/1U39xgVujtBodwW20qIfcGu4E5n2zzaKkl2KmzzHqj4A) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
6. Group Comparison [Google slides](https://docs.google.com/presentation/d/1P2boDph8IfpvjxoxDj_496CLHGtshzJnbbZhszPsmf4) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%205%20Onco%20Query%20Language.pdf)
7. Pathways [Google slides](https://docs.google.com/presentation/d/1O5WGucz0lrfdY25b5QS6zaID_26i434EYXBluqZfT2g) | [PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Tutorial%207%20Pathways.pdf)

### Webinar Recordings
Recordings of live webinars from April & May 2020
1. Introduction to cBioPortal [youtube.com](https://www.youtube.com/watch?v=fPIAxH--cSo) | [bilibili.com](https://www.bilibili.com/video/BV1tf4y1m7Lp) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%201%20Introduction%20to%20cBioPortal.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing)
2. Mutation Details & Patient View [youtube.com](https://www.youtube.com/watch?v=uJsp9kd2jIk) | [bilibili.com](https://www.bilibili.com/video/BV1Qf4y1m7Lx) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%202%20Mutation%20Details%20and%20Patient%20View.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing)
3. Expression Data Analysis [youtube.com](https://www.youtube.com/watch?v=YUxVv6pkxD4) | [bilibili.com](https://www.bilibili.com/video/BV1HK4y1t7dE) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%203%20Expression%20Data%20Analysis.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing)
4. Group Comparison [youtube.com](https://www.youtube.com/watch?v=Tx4HZCrIe5c) | [bilibili.com](https://www.bilibili.com/video/BV1VZ4y1W76p) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%204%20Group%20Comparison.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing)
5. API & R Client [youtube.com](https://www.youtube.com/watch?v=Nq12o2i0yaw) | [bilibili.com](https://www.bilibili.com/video/BV1jz4y197iU) | [Download PDF](https://raw.githubusercontent.com/cBioPortal/cbioportal/master/docs/tutorials/cBioPortal%20Webinar%205%20API%20and%20R%20Client.pdf) | [View slides](https://drive.google.com/drive/folders/0B9KTQJAGhFhhRi1qaTdUWmpLQTA?resourcekey=0-9Mbxt9c_wEbaGRLnAaASlg&usp=sharing) | [Workshop code](https://github.com/cBioPortal/2020-cbioportal-r-workshop)

### How-To Videos
Short videos that show how to perform specific analyses or how to use specific pages.
1. Comparing samples based on expression level of a gene [youtube.com](https://www.youtube.com/watch?v=HTiKUXk0j0s)
2. Proteomic profiles in cBioPortal - An example based on cancer cell lines from the Cancer Cell Line Encyclopedia (CCLE) [youtube.com](https://www.youtube.com/watch?v=62qbjQOH9qc) 
3. Filtering and adding clinical data to Mutations tab [youtube.com](https://www.youtube.com/watch?v=q9No2073c5o)
4. Exploring the longitudinal evolution of individual patients [youtube.com](https://www.youtube.com/watch?v=Hbbs-tHh9LQ)
5. Using Onco Query Language (OQL) to query based on the expression level of genes [youtube.com](https://www.youtube.com/watch?v=kHlFXw2TMzc)
6. How to explore the data in a study [youtube.com](https://www.youtube.com/watch?v=N8ffDgkqDWc)
7. How to run a query for genes of interest [youtube.com](https://www.youtube.com/watch?v=MH-kY5usA70)
8. How to download data [youtube.com](https://www.youtube.com/watch?v=JId304dp3tc)
9. Navigating AACR GENIE - Biopharma Collaborative (BPC) dataset [youtube.com](https://www.youtube.com/playlist?list=PLE2tjlUyHL8Gn4ythnuvf2BxWJVpAxkMM)

### Documentation
* Frequently Asked Questions [FAQ](/user-guide/faq.md)
* Onco Query Language [OQL](/user-guide/oql.md)

### Publications
* de Bruijn et al. Cancer Research 2023 [PubMed](https://pubmed.ncbi.nlm.nih.gov/37668528/)
* Cerami et al. Cancer Discovery 2012 [PubMed](http://cancerdiscovery.aacrjournals.org/content/2/5/401.abstract)
* Gao et al. Science Signaling 2013 [PubMed](https://www.ncbi.nlm.nih.gov/pubmed/23550210)

### Tutorials by others
* cBioPortal Tutorial Series by Jackson Laboratory [youtube.com](https://www.youtube.com/watch?v=tWdrcZOiXc0&list=PLWNp6Z5dXDZ7yRyny8aypSwrArs6e0EJi)
* Using the Cancer Digital Slide Archive in cBioPortal by Nicole M. Rivera Acevedo [youtube.com (English)](https://www.youtube.com/watch?v=oP0GbWmfVWM) | [youtube.com (Spanish)](https://www.youtube.com/watch?v=QyOVCxjVIPU)
* Visualizing and Downloading RNASeq data from cBioPortal by Farhan Haq [youtube.com](https://www.youtube.com/watch?v=I3KiLCE35Mk&list=PLvK49Q0ARs91Ilonq_07n0q9Cd5pOsdJW)
# User Guide

The cBioPortal for Cancer Genomics is a resource for interactive exploration of multidimensional cancer genomics data sets. The goal of cBioPortal is to significantly lower the barriers between complex genomic data and cancer researchers by providing rapid, intuitive, and high-quality access to molecular profiles and clinical attributes from large-scale cancer genomics projects, and therefore to empower researchers to translate these rich data sets into biologic insights and clinical applications.

The following section guides users in performing queries and analysis on any instance of cBioportal.
# Using the dataset validator

To facilitate the loading of new studies into its database, cBioPortal [provides a set of staging files formats](/File-Formats.md) for the various data types. To validate your files you can use the dataset validator script. 

## Running the validator

To run the validator first go to the importer folder
`<cbioportal_source_folder>/core/src/main/scripts/importer`
and then run the following command:
```bash
./validateData.py --help
```
This will tell you the parameters you can use: 
```console
usage: validateData.py [-h] -s STUDY_DIRECTORY
                       [-u URL_SERVER | -p PORTAL_INFO_DIR | -n]
                       [-P PORTAL_PROPERTIES] [-html HTML_TABLE]
                       [-e ERROR_FILE] [-v] [-r] [-m]

cBioPortal study validator

optional arguments:
  -h, --help            show this help message and exit
  -s STUDY_DIRECTORY, --study_directory STUDY_DIRECTORY
                        path to directory.
  -u URL_SERVER, --url_server URL_SERVER
                        URL to cBioPortal server. You can set this if your URL
                        is not http://localhost:8080
  -p PORTAL_INFO_DIR, --portal_info_dir PORTAL_INFO_DIR
                        Path to a directory of cBioPortal info files to be
                        used instead of contacting a server
  -n, --no_portal_checks
                        Skip tests requiring information from the cBioPortal
                        installation
  -html HTML_TABLE, --html_table HTML_TABLE
                        path to html report output file
  -e ERROR_FILE, --error_file ERROR_FILE
                        File to which to write line numbers on which errors
                        were found, for scripts
  -v, --verbose         report status info messages in addition to errors and
                        warnings
  -r, --relaxed-clinical_definitions
                        Option to enable relaxed mode for validator when validating
                        clinical data without header definitions
  -m, --strict_maf_checks
                        Option to enable strict mode for validator when validating
                        mutation data
```

For more information on the `--portal_info_dir` option, see [Offline validation](#offline-validation) below. If your cBioPortal is not using `hg19`, 
you have to specify the `reference_genome` field in your `meta_study.txt`.

For more information, see [Validation of non-human data](#validation-of-non-human-data).

When running the validator with parameter `-r` the validator will run the validation of the clinical data it will ignore all failing checks
about values in the headers of the clinical data file.

When running the validator with parameter `-m` the validator will run the validation of the specific [MAF file checks](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/) for the mutation file in strict maf check mode. This means that
when the validator encounters these validation checks it will report them as an error instead of a warning.

### Example 1: test study_es_0
As an example, you can try the validator with one of the test studies found in  `<cbioportal_source_folder>/core/src/test/scripts/test_data`. Example, assuming port 8080 and using -v option to also see the progress:
```bash
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -u http://localhost:8080 -v
```
Results in:
```console
DEBUG: -: Requesting info from portal at 'http://localhost:8080'
DEBUG: -: Requesting cancer-types from portal at 'http://localhost:8080'
DEBUG: -: Requesting genes from portal at 'http://localhost:8080'
DEBUG: -: Requesting genesets from portal at 'http://localhost:8080'
DEBUG: -: Requesting genesets_version from portal at 'http://localhost:8080'
DEBUG: -: Requesting gene-panels from portal at 'http://localhost:8080'

DEBUG: meta_cancer_type.txt: Starting validation of meta file
INFO: meta_cancer_type.txt: Validation of meta file complete

DEBUG: meta_clinical_patients.txt: Starting validation of meta file
INFO: meta_clinical_patients.txt: Validation of meta file complete

DEBUG: meta_clinical_samples.txt: Starting validation of meta file
INFO: meta_clinical_samples.txt: Validation of meta file complete

DEBUG: meta_cna_discrete.txt: Starting validation of meta file
INFO: meta_cna_discrete.txt: Validation of meta file complete

DEBUG: meta_cna_hg19_seg.txt: Starting validation of meta file
INFO: meta_cna_hg19_seg.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_cna_log2.txt: Starting validation of meta file
INFO: meta_cna_log2.txt: Validation of meta file complete

DEBUG: meta_expression_median.txt: Starting validation of meta file
INFO: meta_expression_median.txt: Validation of meta file complete

DEBUG: meta_expression_median_Zscores.txt: Starting validation of meta file
INFO: meta_expression_median_Zscores.txt: Validation of meta file complete

DEBUG: meta_fusions.txt: Starting validation of meta file
INFO: meta_fusions.txt: Validation of meta file complete

DEBUG: meta_gene_panel_matrix.txt: Starting validation of meta file
INFO: meta_gene_panel_matrix.txt: Validation of meta file complete

DEBUG: meta_gistic_genes_amp.txt: Starting validation of meta file
INFO: meta_gistic_genes_amp.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_gsva_pvalues.txt: Starting validation of meta file
INFO: meta_gsva_pvalues.txt: Validation of meta file complete

DEBUG: meta_gsva_scores.txt: Starting validation of meta file
INFO: meta_gsva_scores.txt: Validation of meta file complete

DEBUG: meta_methylation_hm27.txt: Starting validation of meta file
INFO: meta_methylation_hm27.txt: Validation of meta file complete

DEBUG: meta_mutational_signature.txt: Starting validation of meta file
INFO: meta_mutational_signature.txt: Validation of meta file complete

DEBUG: meta_mutations_extended.txt: Starting validation of meta file
INFO: meta_mutations_extended.txt: Validation of meta file complete

DEBUG: meta_resource_definition.txt: Starting validation of meta file
INFO: meta_resource_definition.txt: Validation of meta file complete

DEBUG: meta_resource_patient.txt: Starting validation of meta file
INFO: meta_resource_patient.txt: Validation of meta file complete

DEBUG: meta_resource_sample.txt: Starting validation of meta file
INFO: meta_resource_sample.txt: Validation of meta file complete

DEBUG: meta_resource_study.txt: Starting validation of meta file
INFO: meta_resource_study.txt: Validation of meta file complete

DEBUG: meta_structural_variants.txt: Starting validation of meta file
INFO: meta_structural_variants.txt: Validation of meta file complete

DEBUG: meta_study.txt: Starting validation of meta file
INFO: meta_study.txt: Validation of meta file complete

DEBUG: -: Study Tag file found. It will be validated.

DEBUG: meta_treatment_ec50.txt: Starting validation of meta file
INFO: meta_treatment_ec50.txt: Validation of meta file complete

DEBUG: meta_treatment_ic50.txt: Starting validation of meta file
INFO: meta_treatment_ic50.txt: Validation of meta file complete

DEBUG: data_cancer_type.txt: Starting validation of file
INFO: data_cancer_type.txt: line 1: New disease type will be added to the portal; value encountered: 'brca-es0'
INFO: data_cancer_type.txt: Validation of file complete
INFO: data_cancer_type.txt: Read 1 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_samples.txt: Starting validation of file
INFO: data_clinical_samples.txt: Validation of file complete
INFO: data_clinical_samples.txt: Read 847 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_definition.txt: Starting validation of file
INFO: data_resource_definition.txt: Validation of file complete
INFO: data_resource_definition.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_sample.txt: Starting validation of file
INFO: data_resource_sample.txt: Validation of file complete
INFO: data_resource_sample.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: study_tags.yml: Starting validation of study tags file
INFO: study_tags.yml: Validation of study tags file complete.

DEBUG: -: Validating case lists

DEBUG: case_lists/cases_cnaseq.txt: Starting validation of meta file
INFO: case_lists/cases_cnaseq.txt: Validation of meta file complete

DEBUG: case_lists/cases_test.txt: Starting validation of meta file
INFO: case_lists/cases_test.txt: Validation of meta file complete

DEBUG: case_lists/cases_sequenced.txt: Starting validation of meta file
INFO: case_lists/cases_sequenced.txt: Validation of meta file complete

DEBUG: case_lists/cases_custom.txt: Starting validation of meta file
INFO: case_lists/cases_custom.txt: Validation of meta file complete

DEBUG: case_lists/cases_cna.txt: Starting validation of meta file
INFO: case_lists/cases_cna.txt: Validation of meta file complete

INFO: -: Validation of case list folder complete

DEBUG: data_gene_panel_matrix.txt: Starting validation of file
INFO: data_gene_panel_matrix.txt: line 1: This column can be replaced by a 'gene_panel' property in the respective meta file; value encountered: 'gistic'
INFO: data_gene_panel_matrix.txt: Validation of file complete
INFO: data_gene_panel_matrix.txt: Read 21 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_discrete.txt: Starting validation of file
INFO: data_cna_discrete.txt: Validation of file complete
INFO: data_cna_discrete.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_patients.txt: Starting validation of file
INFO: data_clinical_patients.txt: Validation of file complete
INFO: data_clinical_patients.txt: Read 845 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median.txt: Starting validation of file
INFO: data_expression_median.txt: Validation of file complete
INFO: data_expression_median.txt: Read 7 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median_Zscores.txt: Starting validation of file
INFO: data_expression_median_Zscores.txt: Validation of file complete
INFO: data_expression_median_Zscores.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_fusions.txt: Starting validation of file
INFO: data_fusions.txt: Validation of file complete
INFO: data_fusions.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutational_signature.txt: Starting validation of file
INFO: data_mutational_signature.txt: Validation of file complete
INFO: data_mutational_signature.txt: Read 62 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ec50.txt: Starting validation of file
INFO: data_treatment_ec50.txt: Validation of file complete
INFO: data_treatment_ec50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ic50.txt: Starting validation of file
INFO: data_treatment_ic50.txt: Validation of file complete
INFO: data_treatment_ic50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gistic_genes_amp.txt: Starting validation of file
INFO: data_gistic_genes_amp.txt: Validation of file complete
INFO: data_gistic_genes_amp.txt: Read 13 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_pvalues.txt: Starting validation of file
INFO: data_gsva_pvalues.txt: Validation of file complete
INFO: data_gsva_pvalues.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_scores.txt: Starting validation of file
INFO: data_gsva_scores.txt: Validation of file complete
INFO: data_gsva_scores.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_log2.txt: Starting validation of file
INFO: data_cna_log2.txt: Validation of file complete
INFO: data_cna_log2.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_methylation_hm27.txt: Starting validation of file
INFO: data_methylation_hm27.txt: Validation of file complete
INFO: data_methylation_hm27.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutations_extended.maf: Starting validation of file
INFO: data_mutations_extended.maf: lines [4, 5, 6, (3 more)]: column 164: Values contained in the column cbp_driver_tiers that will appear in the "Mutation Color" menu of the Oncoprint; values encountered: ['Class 2', 'Class 1', 'Class 4', '(1 more)']
INFO: data_mutations_extended.maf: lines [7, 9]: Line will not be loaded due to the variant classification filter. Filtered types: [Silent, Intron, 3'UTR, 3'Flank, 5'UTR, 5'Flank, IGR, RNA]; value encountered: 'Silent'
INFO: data_mutations_extended.maf: Validation of file complete
INFO: data_mutations_extended.maf: Read 35 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_patient.txt: Starting validation of file
INFO: data_resource_patient.txt: Validation of file complete
INFO: data_resource_patient.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_study.txt: Starting validation of file
INFO: data_resource_study.txt: Validation of file complete
INFO: data_resource_study.txt: Read 2 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_hg19.seg: Starting validation of file
INFO: data_cna_hg19.seg: Validation of file complete
INFO: data_cna_hg19.seg: Read 10 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_structural_variants.txt: Starting validation of file
INFO: data_structural_variants.txt: Validation of file complete
INFO: data_structural_variants.txt: Read 46 lines. Lines with warning: 0. Lines with error: 0

INFO: -: Validation complete
Validation of study succeeded.
```

When using the `-html` option, a report will be generated, which looks like this for the previous example:
![Screenshot of a successful validation report](images/scripts/report.png)

### Example 2: test study_es_1
More test studies for trying the validator (`study_es_1` and `study_es_3`) are available in  `<cbioportal_source_folder>/core/src/test/scripts/test_data`. Example, assuming port 8080 and using -v option:
```bash
./validateData.py -s ../../../test/scripts/test_data/study_es_1/ -u http://localhost:8080 -v
```
Results in:
```console
DEBUG: -: Requesting info from portal at 'http://localhost:8081'
DEBUG: -: Requesting cancer-types from portal at 'http://localhost:8081'
DEBUG: -: Requesting genes from portal at 'http://localhost:8081'
DEBUG: -: Requesting genesets from portal at 'http://localhost:8081'
DEBUG: -: Requesting genesets_version from portal at 'http://localhost:8081'
DEBUG: -: Requesting gene-panels from portal at 'http://localhost:8081'

DEBUG: meta_expression_median.txt: Starting validation of meta file
ERROR: meta_expression_median.txt: Invalid stable id for genetic_alteration_type 'MRNA_EXPRESSION', data_type 'Z-SCORE'; expected one of [mrna_U133_Zscores, rna_seq_mrna_median_Zscores, mrna_median_Zscores, rna_seq_v2_mrna_median_Zscores, rna_seq_v2_mrna_median_normals_Zscores, mirna_median_Zscores, mrna_merged_median_Zscores, mrna_zbynorm, mrna_seq_tpm_Zscores, mrna_seq_cpm_Zscores, rna_seq_mrna_capture_Zscores, mrna_seq_fpkm_capture_Zscores, mrna_seq_fpkm_polya_Zscores, mrna_U133_all_sample_Zscores, mrna_all_sample_Zscores, rna_seq_mrna_median_all_sample_Zscores, mrna_median_all_sample_Zscores, rna_seq_v2_mrna_median_all_sample_Zscores, rna_seq_v2_mrna_median_all_sample_ref_normal_Zscores, mrna_seq_cpm_all_sample_Zscores, mrna_seq_tpm_all_sample_Zscores, rna_seq_mrna_capture_all_sample_Zscores, mrna_seq_fpkm_capture_all_sample_Zscores, mrna_seq_fpkm_polya_all_sample_Zscores]; value encountered: 'mrna'

DEBUG: meta_samples.txt: Starting validation of meta file
WARNING: meta_samples.txt: Unrecognized field in meta file; values encountered: ['show_profile_in_analysis_tab', 'profile_description', 'profile_name']
INFO: meta_samples.txt: Validation of meta file complete

DEBUG: meta_study.txt: Starting validation of meta file
INFO: meta_study.txt: Validation of meta file complete
INFO: meta_study.txt: No reference genome specified -- using default (hg19)

DEBUG: meta_treatment_ec50.txt: Starting validation of meta file
INFO: meta_treatment_ec50.txt: Validation of meta file complete

DEBUG: meta_treatment_ic50.txt: Starting validation of meta file
INFO: meta_treatment_ic50.txt: Validation of meta file complete

DEBUG: data_samples.txt: Starting validation of file
INFO: data_samples.txt: Validation of file complete
INFO: data_samples.txt: Read 831 lines. Lines with warning: 0. Lines with error: 0

DEBUG: -: Validating case lists

DEBUG: case_lists/cases_all.txt: Starting validation of meta file
INFO: case_lists/cases_all.txt: Validation of meta file complete
ERROR: case_lists/cases_all.txt: Sample ID not defined in clinical file; value encountered: 'INVALID-A2-A0T2-01'

INFO: -: Validation of case list folder complete

DEBUG: data_treatment_ec50.txt: Starting validation of file
ERROR: data_treatment_ec50.txt: line 2: column 1: Do not use space in the stable id; value encountered: '17 AAG'
ERROR: data_treatment_ec50.txt: line 7: column 5: Blank cell found in column; value encountered: ''' (in column 'TCGA-A1-A0SB-01')'
INFO: data_treatment_ec50.txt: Validation of file complete
INFO: data_treatment_ec50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 2

DEBUG: data_treatment_ic50.txt: Starting validation of file
ERROR: data_treatment_ic50.txt: line 7: column 5: Blank cell found in column; value encountered: ''' (in column 'TCGA-A1-A0SB-01')'
INFO: data_treatment_ic50.txt: Validation of file complete
INFO: data_treatment_ic50.txt: Read 10 lines. Lines with warning: 0. Lines with error: 1

INFO: -: Validation complete
Validation of study failed.
```

And respective HTML report:
![Screenshot of an unsuccessful validation report](images/scripts/report1.png)

## Offline validation ##
The validation script can be used offline, without connecting to a cBioPortal server. The tests that depend on information specific to the portal (which clinical attributes and cancer types have been previously defined, and which Entrez gene identifiers and corresponding symbols are supported), will instead be read from a folder with .json files generated from the portal.

### Example 3: validation with a portal info folder ###
To run the validator with a folder of portal information files, add the `-p/--portal_info_dir` option to the command line, followed by the path to the folder:
```bash
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -p ../../../test/scripts/test_data/api_json_system_tests/ -v
```
```console
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/cancer-types.json
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/genes.json
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/genesets.json
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/genesets_version.json
DEBUG: -: Reading portal information from ../../../test/scripts/test_data/api_json_system_tests/gene-panels.json

DEBUG: meta_cancer_type.txt: Starting validation of meta file
INFO: meta_cancer_type.txt: Validation of meta file complete

DEBUG: meta_clinical_patients.txt: Starting validation of meta file
INFO: meta_clinical_patients.txt: Validation of meta file complete

DEBUG: meta_clinical_samples.txt: Starting validation of meta file
INFO: meta_clinical_samples.txt: Validation of meta file complete

DEBUG: meta_cna_discrete.txt: Starting validation of meta file
INFO: meta_cna_discrete.txt: Validation of meta file complete

DEBUG: meta_cna_hg19_seg.txt: Starting validation of meta file
INFO: meta_cna_hg19_seg.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_cna_log2.txt: Starting validation of meta file
INFO: meta_cna_log2.txt: Validation of meta file complete

DEBUG: meta_expression_median.txt: Starting validation of meta file
INFO: meta_expression_median.txt: Validation of meta file complete

DEBUG: meta_expression_median_Zscores.txt: Starting validation of meta file
INFO: meta_expression_median_Zscores.txt: Validation of meta file complete

DEBUG: meta_fusions.txt: Starting validation of meta file
INFO: meta_fusions.txt: Validation of meta file complete

DEBUG: meta_gene_panel_matrix.txt: Starting validation of meta file
INFO: meta_gene_panel_matrix.txt: Validation of meta file complete

DEBUG: meta_gistic_genes_amp.txt: Starting validation of meta file
INFO: meta_gistic_genes_amp.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_gsva_pvalues.txt: Starting validation of meta file
INFO: meta_gsva_pvalues.txt: Validation of meta file complete

DEBUG: meta_gsva_scores.txt: Starting validation of meta file
INFO: meta_gsva_scores.txt: Validation of meta file complete

DEBUG: meta_methylation_hm27.txt: Starting validation of meta file
INFO: meta_methylation_hm27.txt: Validation of meta file complete

DEBUG: meta_mutational_signature.txt: Starting validation of meta file
INFO: meta_mutational_signature.txt: Validation of meta file complete

DEBUG: meta_mutations_extended.txt: Starting validation of meta file
INFO: meta_mutations_extended.txt: Validation of meta file complete

DEBUG: meta_resource_definition.txt: Starting validation of meta file
INFO: meta_resource_definition.txt: Validation of meta file complete

DEBUG: meta_resource_patient.txt: Starting validation of meta file
INFO: meta_resource_patient.txt: Validation of meta file complete

DEBUG: meta_resource_sample.txt: Starting validation of meta file
INFO: meta_resource_sample.txt: Validation of meta file complete

DEBUG: meta_resource_study.txt: Starting validation of meta file
INFO: meta_resource_study.txt: Validation of meta file complete

DEBUG: meta_structural_variants.txt: Starting validation of meta file
INFO: meta_structural_variants.txt: Validation of meta file complete

DEBUG: meta_study.txt: Starting validation of meta file
INFO: meta_study.txt: Validation of meta file complete

DEBUG: -: Study Tag file found. It will be validated.

DEBUG: meta_treatment_ec50.txt: Starting validation of meta file
INFO: meta_treatment_ec50.txt: Validation of meta file complete

DEBUG: meta_treatment_ic50.txt: Starting validation of meta file
INFO: meta_treatment_ic50.txt: Validation of meta file complete

DEBUG: data_cancer_type.txt: Starting validation of file
INFO: data_cancer_type.txt: Validation of file complete
INFO: data_cancer_type.txt: Read 1 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_samples.txt: Starting validation of file
INFO: data_clinical_samples.txt: Validation of file complete
INFO: data_clinical_samples.txt: Read 847 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_definition.txt: Starting validation of file
INFO: data_resource_definition.txt: Validation of file complete
INFO: data_resource_definition.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_sample.txt: Starting validation of file
INFO: data_resource_sample.txt: Validation of file complete
INFO: data_resource_sample.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: study_tags.yml: Starting validation of study tags file
INFO: study_tags.yml: Validation of study tags file complete.

DEBUG: -: Validating case lists

DEBUG: case_lists/cases_cnaseq.txt: Starting validation of meta file
INFO: case_lists/cases_cnaseq.txt: Validation of meta file complete

DEBUG: case_lists/cases_test.txt: Starting validation of meta file
INFO: case_lists/cases_test.txt: Validation of meta file complete

DEBUG: case_lists/cases_sequenced.txt: Starting validation of meta file
INFO: case_lists/cases_sequenced.txt: Validation of meta file complete

DEBUG: case_lists/cases_custom.txt: Starting validation of meta file
INFO: case_lists/cases_custom.txt: Validation of meta file complete

DEBUG: case_lists/cases_cna.txt: Starting validation of meta file
INFO: case_lists/cases_cna.txt: Validation of meta file complete

INFO: -: Validation of case list folder complete

DEBUG: data_gene_panel_matrix.txt: Starting validation of file
INFO: data_gene_panel_matrix.txt: line 1: This column can be replaced by a 'gene_panel' property in the respective meta file; value encountered: 'gistic'
INFO: data_gene_panel_matrix.txt: Validation of file complete
INFO: data_gene_panel_matrix.txt: Read 21 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_discrete.txt: Starting validation of file
INFO: data_cna_discrete.txt: Validation of file complete
INFO: data_cna_discrete.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_patients.txt: Starting validation of file
INFO: data_clinical_patients.txt: Validation of file complete
INFO: data_clinical_patients.txt: Read 845 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median.txt: Starting validation of file
INFO: data_expression_median.txt: Validation of file complete
INFO: data_expression_median.txt: Read 7 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median_Zscores.txt: Starting validation of file
INFO: data_expression_median_Zscores.txt: Validation of file complete
INFO: data_expression_median_Zscores.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_fusions.txt: Starting validation of file
INFO: data_fusions.txt: Validation of file complete
INFO: data_fusions.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutational_signature.txt: Starting validation of file
INFO: data_mutational_signature.txt: Validation of file complete
INFO: data_mutational_signature.txt: Read 62 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ec50.txt: Starting validation of file
INFO: data_treatment_ec50.txt: Validation of file complete
INFO: data_treatment_ec50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ic50.txt: Starting validation of file
INFO: data_treatment_ic50.txt: Validation of file complete
INFO: data_treatment_ic50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gistic_genes_amp.txt: Starting validation of file
INFO: data_gistic_genes_amp.txt: Validation of file complete
INFO: data_gistic_genes_amp.txt: Read 13 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_pvalues.txt: Starting validation of file
INFO: data_gsva_pvalues.txt: Validation of file complete
INFO: data_gsva_pvalues.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_scores.txt: Starting validation of file
INFO: data_gsva_scores.txt: Validation of file complete
INFO: data_gsva_scores.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_log2.txt: Starting validation of file
INFO: data_cna_log2.txt: Validation of file complete
INFO: data_cna_log2.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_methylation_hm27.txt: Starting validation of file
INFO: data_methylation_hm27.txt: Validation of file complete
INFO: data_methylation_hm27.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutations_extended.maf: Starting validation of file
INFO: data_mutations_extended.maf: lines [4, 5, 6, (3 more)]: column 164: Values contained in the column cbp_driver_tiers that will appear in the "Mutation Color" menu of the Oncoprint; values encountered: ['Class 2', 'Class 1', 'Class 4', '(1 more)']
INFO: data_mutations_extended.maf: lines [7, 9]: Line will not be loaded due to the variant classification filter. Filtered types: [Silent, Intron, 3'UTR, 3'Flank, 5'UTR, 5'Flank, IGR, RNA]; value encountered: 'Silent'
INFO: data_mutations_extended.maf: Validation of file complete
INFO: data_mutations_extended.maf: Read 35 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_patient.txt: Starting validation of file
INFO: data_resource_patient.txt: Validation of file complete
INFO: data_resource_patient.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_study.txt: Starting validation of file
INFO: data_resource_study.txt: Validation of file complete
INFO: data_resource_study.txt: Read 2 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_hg19.seg: Starting validation of file
INFO: data_cna_hg19.seg: Validation of file complete
INFO: data_cna_hg19.seg: Read 10 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_structural_variants.txt: Starting validation of file
INFO: data_structural_variants.txt: Validation of file complete
INFO: data_structural_variants.txt: Read 46 lines. Lines with warning: 0. Lines with error: 0

INFO: -: Validation complete
Validation of study succeeded.
```

### Example 4: generating the portal info folder ###
The portal information files can be generated on the server, using the dumpPortalInfo script. Go to `<cbioportal_source_folder>/core/src/main/scripts`, make sure the environment variables `$JAVA_HOME` and `$PORTAL_HOME` are set, and run dumpPortalInfo.pl with the name of the directory you want to create:
```bash
export JAVA_HOME='/usr/lib/jvm/default-java'
export PORTAL_HOME=<cbioportal_configuration_folder>
./dumpPortalInfo.pl /home/johndoe/my_portal_info_folder/
```

### Example 5: validating without portal-specific information ###
Alternatively, you can run the validation script with the `-n/--no_portal_checks` flag to entirely skip checks relating to installation-specific metadata. Be warned that files succeeding this validation may still fail to load (correctly).

```bash
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -n -v
```
```console
WARNING: -: Skipping validations relating to cancer types defined in the portal
WARNING: -: Skipping validations relating to gene identifiers and aliases defined in the portal
WARNING: -: Skipping validations relating to gene set identifiers
WARNING: -: Skipping validations relating to gene panel identifiers

DEBUG: meta_cancer_type.txt: Starting validation of meta file
INFO: meta_cancer_type.txt: Validation of meta file complete

DEBUG: meta_clinical_patients.txt: Starting validation of meta file
INFO: meta_clinical_patients.txt: Validation of meta file complete

DEBUG: meta_clinical_samples.txt: Starting validation of meta file
INFO: meta_clinical_samples.txt: Validation of meta file complete

DEBUG: meta_cna_discrete.txt: Starting validation of meta file
INFO: meta_cna_discrete.txt: Validation of meta file complete

DEBUG: meta_cna_hg19_seg.txt: Starting validation of meta file
INFO: meta_cna_hg19_seg.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_cna_log2.txt: Starting validation of meta file
INFO: meta_cna_log2.txt: Validation of meta file complete

DEBUG: meta_expression_median.txt: Starting validation of meta file
INFO: meta_expression_median.txt: Validation of meta file complete

DEBUG: meta_expression_median_Zscores.txt: Starting validation of meta file
INFO: meta_expression_median_Zscores.txt: Validation of meta file complete

DEBUG: meta_fusions.txt: Starting validation of meta file
INFO: meta_fusions.txt: Validation of meta file complete

DEBUG: meta_gene_panel_matrix.txt: Starting validation of meta file
INFO: meta_gene_panel_matrix.txt: Validation of meta file complete

DEBUG: meta_gistic_genes_amp.txt: Starting validation of meta file
INFO: meta_gistic_genes_amp.txt: Validation of meta file complete

DEBUG: -: Retrieving chromosome lengths from '/home/sander/git/cbioportal/core/src/main/scripts/importer/chromosome_sizes.json'

DEBUG: meta_gsva_pvalues.txt: Starting validation of meta file
INFO: meta_gsva_pvalues.txt: Validation of meta file complete

DEBUG: meta_gsva_scores.txt: Starting validation of meta file
INFO: meta_gsva_scores.txt: Validation of meta file complete

DEBUG: meta_methylation_hm27.txt: Starting validation of meta file
INFO: meta_methylation_hm27.txt: Validation of meta file complete

DEBUG: meta_mutational_signature.txt: Starting validation of meta file
INFO: meta_mutational_signature.txt: Validation of meta file complete

DEBUG: meta_mutations_extended.txt: Starting validation of meta file
INFO: meta_mutations_extended.txt: Validation of meta file complete

DEBUG: meta_resource_definition.txt: Starting validation of meta file
INFO: meta_resource_definition.txt: Validation of meta file complete

DEBUG: meta_resource_patient.txt: Starting validation of meta file
INFO: meta_resource_patient.txt: Validation of meta file complete

DEBUG: meta_resource_sample.txt: Starting validation of meta file
INFO: meta_resource_sample.txt: Validation of meta file complete

DEBUG: meta_resource_study.txt: Starting validation of meta file
INFO: meta_resource_study.txt: Validation of meta file complete

DEBUG: meta_structural_variants.txt: Starting validation of meta file
INFO: meta_structural_variants.txt: Validation of meta file complete

DEBUG: meta_study.txt: Starting validation of meta file
INFO: meta_study.txt: Validation of meta file complete

DEBUG: -: Study Tag file found. It will be validated.

DEBUG: meta_treatment_ec50.txt: Starting validation of meta file
INFO: meta_treatment_ec50.txt: Validation of meta file complete

DEBUG: meta_treatment_ic50.txt: Starting validation of meta file
INFO: meta_treatment_ic50.txt: Validation of meta file complete

DEBUG: data_cancer_type.txt: Starting validation of file
INFO: data_cancer_type.txt: Validation of file complete
INFO: data_cancer_type.txt: Read 1 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_samples.txt: Starting validation of file
INFO: data_clinical_samples.txt: Validation of file complete
INFO: data_clinical_samples.txt: Read 847 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_definition.txt: Starting validation of file
INFO: data_resource_definition.txt: Validation of file complete
INFO: data_resource_definition.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_sample.txt: Starting validation of file
INFO: data_resource_sample.txt: Validation of file complete
INFO: data_resource_sample.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: study_tags.yml: Starting validation of study tags file
INFO: study_tags.yml: Validation of study tags file complete.

DEBUG: -: Validating case lists

DEBUG: case_lists/cases_cnaseq.txt: Starting validation of meta file
INFO: case_lists/cases_cnaseq.txt: Validation of meta file complete

DEBUG: case_lists/cases_test.txt: Starting validation of meta file
INFO: case_lists/cases_test.txt: Validation of meta file complete

DEBUG: case_lists/cases_sequenced.txt: Starting validation of meta file
INFO: case_lists/cases_sequenced.txt: Validation of meta file complete

DEBUG: case_lists/cases_custom.txt: Starting validation of meta file
INFO: case_lists/cases_custom.txt: Validation of meta file complete

DEBUG: case_lists/cases_cna.txt: Starting validation of meta file
INFO: case_lists/cases_cna.txt: Validation of meta file complete

INFO: -: Validation of case list folder complete

DEBUG: data_gene_panel_matrix.txt: Starting validation of file
INFO: data_gene_panel_matrix.txt: line 1: This column can be replaced by a 'gene_panel' property in the respective meta file; value encountered: 'gistic'
INFO: data_gene_panel_matrix.txt: Validation of file complete
INFO: data_gene_panel_matrix.txt: Read 21 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_discrete.txt: Starting validation of file
INFO: data_cna_discrete.txt: Validation of file complete
INFO: data_cna_discrete.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_clinical_patients.txt: Starting validation of file
INFO: data_clinical_patients.txt: Validation of file complete
INFO: data_clinical_patients.txt: Read 845 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median.txt: Starting validation of file
INFO: data_expression_median.txt: Validation of file complete
INFO: data_expression_median.txt: Read 7 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_expression_median_Zscores.txt: Starting validation of file
INFO: data_expression_median_Zscores.txt: Validation of file complete
INFO: data_expression_median_Zscores.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_fusions.txt: Starting validation of file
INFO: data_fusions.txt: Validation of file complete
INFO: data_fusions.txt: Read 6 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutational_signature.txt: Starting validation of file
INFO: data_mutational_signature.txt: Validation of file complete
INFO: data_mutational_signature.txt: Read 62 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ec50.txt: Starting validation of file
INFO: data_treatment_ec50.txt: Validation of file complete
INFO: data_treatment_ec50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_treatment_ic50.txt: Starting validation of file
INFO: data_treatment_ic50.txt: Validation of file complete
INFO: data_treatment_ic50.txt: Read 11 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gistic_genes_amp.txt: Starting validation of file
INFO: data_gistic_genes_amp.txt: Validation of file complete
INFO: data_gistic_genes_amp.txt: Read 13 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_pvalues.txt: Starting validation of file
INFO: data_gsva_pvalues.txt: Validation of file complete
INFO: data_gsva_pvalues.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_gsva_scores.txt: Starting validation of file
INFO: data_gsva_scores.txt: Validation of file complete
INFO: data_gsva_scores.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_log2.txt: Starting validation of file
INFO: data_cna_log2.txt: Validation of file complete
INFO: data_cna_log2.txt: Read 8 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_methylation_hm27.txt: Starting validation of file
INFO: data_methylation_hm27.txt: Validation of file complete
INFO: data_methylation_hm27.txt: Read 9 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_mutations_extended.maf: Starting validation of file
INFO: data_mutations_extended.maf: lines [4, 5, 6, (3 more)]: column 164: Values contained in the column cbp_driver_tiers that will appear in the "Mutation Color" menu of the Oncoprint; values encountered: ['Class 2', 'Class 1', 'Class 4', '(1 more)']
INFO: data_mutations_extended.maf: lines [7, 9]: Line will not be loaded due to the variant classification filter. Filtered types: [Silent, Intron, 3'UTR, 3'Flank, 5'UTR, 5'Flank, IGR, RNA]; value encountered: 'Silent'
INFO: data_mutations_extended.maf: Validation of file complete
INFO: data_mutations_extended.maf: Read 35 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_patient.txt: Starting validation of file
INFO: data_resource_patient.txt: Validation of file complete
INFO: data_resource_patient.txt: Read 4 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_resource_study.txt: Starting validation of file
INFO: data_resource_study.txt: Validation of file complete
INFO: data_resource_study.txt: Read 2 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_cna_hg19.seg: Starting validation of file
INFO: data_cna_hg19.seg: Validation of file complete
INFO: data_cna_hg19.seg: Read 10 lines. Lines with warning: 0. Lines with error: 0

DEBUG: data_structural_variants.txt: Starting validation of file
INFO: data_structural_variants.txt: Validation of file complete
INFO: data_structural_variants.txt: Read 46 lines. Lines with warning: 0. Lines with error: 0

INFO: -: Validation complete
Validation of study succeeded with warnings.
```

## Validation of non-human data ##
When importing a study with a reference genome other than hg19/GRCh37, this should be specified in the `meta_study.txt` file, next to the `reference_genome` field. Supported values are **hg19**, **hg38** and **mm10**.

cBioPortal is gradually introducing support for mouse. If you want to load mouse studies and you have to [set up your database for mouse](/deployment/deploy-without-docker/Import-the-Seed-Database.md#download-the-cbioportal-seed-database).

As an example, the command for the mouse example using the three parameters is given:
```
./validateData.py -s ../../../test/scripts/test_data/study_es_0/ -P ../../../../../src/main/resources/application.properties -u http://localhost:8080 -v
```

## Running the validator for multiple studies
The importer folder `<cbioportal_source_folder>/core/src/main/scripts/importer` also contains a script for running the validator for multiple studies:
```bash
./validateStudies.py --help
```
The following parameters can be used:
```console
usage: validateStudies.py [-h] [-d ROOT_DIRECTORY] [-l LIST_OF_STUDIES]
                          [-html HTML_FOLDER]
                          [-u URL_SERVER | -p PORTAL_INFO_DIR | -n]
                          [-P PORTAL_PROPERTIES] [-m]

Wrapper where cBioPortal study validator is run for multiple studies

optional arguments:
  -h, --help            show this help message and exit
  -d ROOT_DIRECTORY, --root-directory ROOT_DIRECTORY
                        Path to directory with all studies that should be
                        validated
  -l LIST_OF_STUDIES, --list-of-studies LIST_OF_STUDIES
                        List with paths of studies which should be validated
  -html HTML_FOLDER, --html-folder HTML_FOLDER
                        Path to folder for output HTML reports
  -u URL_SERVER, --url_server URL_SERVER
                        URL to cBioPortal server. You can set this if your URL
                        is not http://localhost:8080
  -p PORTAL_INFO_DIR, --portal_info_dir PORTAL_INFO_DIR
                        Path to a directory of cBioPortal info files to be
                        used instead of contacting a server
  -n, --no_portal_checks
                        Skip tests requiring information from the cBioPortal
                        installation 
  -m, --strict_maf_checks
                        Option to enable strict mode for validator when
                        validating mutation data
```

Parameters `--url_server`, `--portal_info_dir`, `--no_portal_checks` and `--portal_properties` are equal to the parameters with the same name in `validateData.py`. The script will save a log file with validation output (`log-validate-studies.txt`) and output the validation status from the input studies:

```console
=== Validating study ../../../test/scripts/test_data/study_es_0
Result: VALID (WITH WARNINGS)

=== Validating study ../../../test/scripts/test_data/study_es_1
Result: INVALID

=== Validating study ../../../test/scripts/test_data/study_es_invalid
directory cannot be found: ../../../test/scripts/test_data/study_es_invalid
Result: INVALID (PROBLEMS OCCURRED)

```

### Example 1: Root directory parameter
Validation can be run for all studies in a certain directory by using the `--root-directory` parameter. The script will append each folder in the root directory to the study list to validate:
```bash
./validateStudies.py -d ../../../test/scripts/test_data/
```

### Example 2: List of studies parameter
Validation can also be run for specific studies by using the `--list-of-studies` parameter. The paths to the different studies can be defined and seperated by a comma:
```bash
./validateStudies.py -l ../../../test/scripts/test_data/study_es_0,../../../test/scripts/test_data/study_es_1
```

### Example 3: Combination root directory and list of studies parameter
Validation can also be run on specific studies in a certain directory by combining the `--root-directory` and `--list-of-studies` parameter:
```bash
./validateStudies.py -d ../../../test/scripts/test_data/ -l study_es_0,study_es_1
```

### Example 4: HTML folder parameter
When HTML validation reports are desired, an output folder for these HTML files can be specified. This folder does not have to exist, the script can create the folder. The HTML validation reports will get the following name: `<study_name>-validation.html`. To create HTML validation reports for each study the `--html-folder` parameter needs to be defined:
```bash
./validateStudies.py -d ../../../test/scripts/test_data/ -l study_es_0,study_es_1 -html ../../../test/scripts/test_data/validation-reports
```
## Importing Data into cBioPortal
The metaImport script should be used to automate the process of validating and loading datasets. It also has some nice features like an extra option to only load datasets that completely pass validation (i.e. with no errors, while warnings can be explicitly allowed by the user). 

#### Running the metaImport Script
To run the metaImport script first go to the importer folder
`<your_cbioportal_dir>/core/src/main/scripts/importer` 
and then run the following command:
```
./metaImport.py -h
```
This will tell you the parameters you can use:
```
$./metaImport.py -h
usage: metaImport.py [-h] -s STUDY_DIRECTORY
                     [-u URL_SERVER | -p PORTAL_INFO_DIR | -n]
                     [-jar JAR_PATH] [-html HTML_TABLE]
                     [-v] [-o] [-r] [-m]

cBioPortal meta Importer

optional arguments:
  -h, --help            show this help message and exit
  -s STUDY_DIRECTORY, --study_directory STUDY_DIRECTORY
                        path to directory.
  -u URL_SERVER, --url_server URL_SERVER
                        URL to cBioPortal server. You can set this if your URL
                        is not http://localhost/cbioportal
  -p PORTAL_INFO_DIR, --portal_info_dir PORTAL_INFO_DIR
                        Path to a directory of cBioPortal info files to be
                        used instead of contacting the web API
  -n, --no_portal_checks
                        Skip tests requiring information from the cBioPortal
                        installation
  -jar JAR_PATH, --jar_path JAR_PATH
                        Path to scripts JAR file (default: locate it relative
                        to the import script)
  -html HTML_TABLE, --html_table HTML_TABLE
                        path to html report
  -v, --verbose         report status info messages while validating
  -o, --override_warning
                        override warnings and continue importing
  -r, --relaxed_clinical_definitions
                        Option to enable relaxed mode for validator when
                        validating clinical data without header definitions
  -m, --strict_maf_checks
                        Option to enable strict mode for validator when
                        validating mutation data
```

#### Example of Importing a study
Export `PORTAL_HOME` as explained [here](/deployment/deploy-without-docker/Load-Sample-Cancer-Study.md), e.g.

```
export PORTAL_HOME=<cbioportal_configuration_folder>
```

and then run (this simple command only works if your cBioPortal is running at http://localhost/cbioportal - if this is not the case, follow the advanced example):

```
./metaImport.py -s ../../../test/scripts/test_data/study_es_0/
```

#### Advanced Example
This example imports the study to the localhost, creates an html report and shows status messages.
```
./metaImport.py -s ../../../test/scripts/test_data/study_es_0/ -u http://localhost:8080 -html myReport.html -v
```

By adding `-o`, warnings will be overridden and import will start after validation.

## Development / debugging mode
For developers and specific testing purposes, an extra script, cbioportalImporter.py, is available which imports data regardless of validation results. Check [this](Data-Loading-For-Developers.md) page for more information on how to use it.
# API and API Clients

cBioPortal provides a REST API for programmatic access to the data. The visualizations one can see on the website leverage the same API. By connecting to the API directly, anyone can build their own visalizations/reports.

Please see the full reference documentation for the API [here](https://www.cbioportal.org/api/swagger-ui/index.html).

## API Clients

The cBioPortal REST API is described using Swagger/OpenAPI, which allows one to generate a client in most programming languages. One can use the command line tool `curl` for dowloading data on the command line or use another language such as `Python` or `R` to make visualizations. We list some common examples below, but if your language is not listed, there is likely a client generator available elsewhere (see e.g. https://swagger.io/tools/swagger-codegen/). Do reach out if you'd like us to add a language.

### R clients

There are multiple ways to access the API using R. Below are two recommended R packages to access cBioPortal data.

#### cBioPortalData (recommended)

cBioPortalData aims to import all cBioPortal datasets as MultiAssayExperiment objects in Bioconductor. Some of its key features:

1. The MultiAssayExperiment class explicitly links all assays to the patient clinical/pathological data
2. The MultiAssayExperiment class provides a flexible API including harmonized subsetting and reshaping to convenient wide and long formats.
3. It provides complete datasets, not just for subsets of genes
4. It provides automatic local caching, thanks to BiocFileCache.

For a comprehensive user guide to `cBioportalData` see: https://waldronlab.io/cBioPortalData/articles/cBioPortalData.html

See also the workshop materials from our webinar which include an intro to `cBioPortalData`: https://github.com/cBioPortal/2020-cbioportal-r-workshop.

Note that one can point to private authenticated instances like this:

```
cBioPortal(
    hostname = "genie.cbioportal.org",
    token = "~/Downloads/cbioportal_data_access_token.txt"
)
```

#### cbioportalR (recommended)

cbioportalR offers easy-to-use functions that allow users to browse and pull data from public or institutional cBioPortal sites without knowledge of web service or Bioconductor infrastructures. The package is tidyverse-compatible. Key package features include:

1. Comprehensive documentation aimed at helping clinical researchers understand the underlying structure of cBioPortal data
2. [Tutorials]([https://www.karissawhiting.com/cbioportalR/articles/overview-of-workflow.html]) for quick API authentication and set up
3. Functions to pull complete clinical and genomic data by study ID, molecular profile ID, sample list IDs or individual sample ID (e.g. `get_genetics_by_study()`, `get_genetics_by_sample()`)
4. Functions to navigate and identify patient IDs, sample IDs or study IDs as needed, or infer necessary ID information for queries when not supplied by user.
5. Helper functions to pull information on gene panels (`get_gene_panel()`), or lookup entrez ID (`get_entrez_id()`), Hugo Symbol (`get_hugo_symbol()`) or common gene aliases (`get_alias()`) of genes
6. Capability to query multiple sample IDs from different studies concurrently

For a detailed tutorial on `cbioportalR`, see the package website: https://www.karissawhiting.com/cbioportalR/articles/overview-of-workflow.html

#### rapiclient

Although we recommend [cBioPortalData](/#cbioportaldata-recommended) or [cbioportalR](/#cbioportalR-recommended) for most use cases, it is possible to connect to the API directly using [rapiclient](https://github.com/bergant/rapiclient):

```
library(rapiclient)
client <- get_api(url = "https://www.cbioportal.org/api/v2/api-docs")
```

#### CGDSR (will be deprecated)

The CGDS-R package connects an older version of our web API (`webservice.do`). Althought we will continue to keep `webservice.do` running for a while, we can't guarantee the same level of quality as our new API (`cbioportal.org/api`) provides. Therefore we recommend that you use `cBioPortalData` instead.

### Python client

There are multiple ways to access the API using Python. One can use the `bravado` package to access the API directly, or use the `cbio_py` client, which provides a simple wrapper for the API and returns data in a format that is easy to work with.

#### bravado

Generate a client in Python using [bravado](https://github.com/Yelp/bravado) like this:

```python
from bravado.client import SwaggerClient
cbioportal = SwaggerClient.from_url('https://www.cbioportal.org/api/v2/api-docs',
                                    config={"validate_requests":False,"validate_responses":False,"validate_swagger_spec": False})
```

This allows you to access all API endpoints:

```python
>>> dir(cbioportal)
['Cancer Types',
 'Clinical Attributes',
 'Clinical Data',
 'Clinical Events',
 'Copy Number Segments',
 'Discrete Copy Number Alterations',
 'Gene Panels',
 'Genes',
 'Molecular Data',
 'Molecular Profiles',
 'Mutations',
 'Patients',
 'Sample Lists',
 'Samples',
 'Studies']
```

For easy tab completion you can add lower cases and underscores:

```python
for a in dir(cbioportal):
    cbioportal.__setattr__(a.replace(' ', '_').lower(), cbioportal.__getattr__(a))
```

This example gets you all mutation data for the MSK-IMPACT 2017 study:

```python
muts = cbioportal.mutations.getMutationsInMolecularProfileBySampleListIdUsingGET(
    molecularProfileId="msk_impact_2017_mutations", # {study_id}_mutations gives default mutations profile for study 
    sampleListId="msk_impact_2017_all", # {study_id}_all includes all samples
    projection="DETAILED" # include gene info
).result()
```

For a portal that requires authentication one can use (see [Data Access Using Tokens](/deployment/authorization-and-authentication/Authenticating-Users-via-Tokens.md)):

```
from bravado.client import SwaggerClient
from bravado.requests_client import RequestsClient

http_client = RequestsClient()
http_client.set_api_key(
    'genie.cbioportal.org', 'Bearer <TOKEN>',
    param_name='Authorization', param_in='header'
)

cbioportal = SwaggerClient.from_url('https://genie.cbioportal.org/api/v2/api-docs',
                                    http_client=http_client,
                                    config={"validate_requests":False,
                                            "validate_responses":False,
                                            "validate_swagger_spec": False}
)
```

A Jupyter notebook with more examples can be found [here](https://github.com/mskcc/cbsp-hackathon/blob/master/0-introduction/cbsp\_hackathon.ipynb).

#### cbio\_py

See the `cbio_py` documentation: https://pypi.org/project/cbio-py/.
## Introduction
For some data types, when uploading to cBioPortal, it is currently necessary to also provide a z-score transformed version of your input file. The z-score data is essential for the oncoprint functionality. The oncoprint shows high or low mRNA expression of the genes, based on the threshold the user sets when selecting the genomic profile. 

:warning: 
Please keep in mind that the z-scores are calculated using only patient data. Hence, 'mRNA High' in this case implies higher expression than the average patient. Also, the source data on which the z-score data is based does not necessarily follow the normal distribution. If your data does not follow the normal distribution, the z-score threshold is less reliable and will result in more false positives or false negatives. You can consider log transforming your value before calculating z-scores to improve this. However, even the logged values may not follow the normal distribution, especially if the data is bimodal.

## The cBioPortal Z-Score calculation method

cBioPortal currently generates two z-score profiles using two different base populations:
- Distribution based on **diploid** samples only: The expression distribution for unaltered copies of the gene is estimated by calculating the mean and variance of the expression values for samples in which the gene is diploid (i.e. value is "0" as reported by [discrete CNA data](File-Formats.md#discrete-copy-number-data)). We call this the unaltered distribution. If the gene has no diploid samples, then its normalized expression is reported as NA.

- Distribution based on **all** samples: The expression distribution of the gene is estimated by calculating the mean and variance of all samples with expression values. If the gene has samples whose expression values are all zeros or non-numeric, then its normalized expression is reported as NA.

Otherwise for every sample, the gene's normalized expression for both the profiles is reported as

```
(r - mu)/sigma
```
where `r` is the raw expression value, and `mu` and `sigma` are the mean and standard deviation of the base population, respectively.

### How to proceed
cBioPortal expects z-score normalization to take place per gene. You can calculate z-scores with your own preferred method, or use one of the cBioPortal provided approaches:
- [convertExpressionZscores.pl](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/convertExpressionZscores.pl) applies Method 1 (diploid samples as base population)
- [NormalizeExpressionLevels_allsampleref.py](https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/zscores/zscores_relative_allsamples) applies Method 2 (all samples as base population)

Examples of the calculation and running the programs are below.

## convertExpressionZscores method
Given expression and Copy Number Variation data for a set of samples (patients), generate normalized expression values.

### Parameters

`<copy_number_file> <expression_file> <output_file> <normal_sample_suffix> <[min_number_of_diploids]>`

- `<copy_number_file>` : the [discrete copy number (CNA) file](File-Formats.md#discrete-copy-number-data) 
- `<expression_file>` : the [expression (exp) data file](File-Formats.md#expression-data). 
- `<output_file>` : the output file to be generated
- `<normal_sample_suffix>` : use this to identify which of your samples are "normal" samples (if any). E.g. normal TCGA samples have a suffix "-11". Set it to some dummy value, e.g. "NONE", if you have no normal samples in your data.

### Algorithm
Input: [discrete copy number (CNA)](File-Formats.md#discrete-copy-number-data) and [expression (exp)](File-Formats.md#expression-data) files

```
    for each gene{
       identify diploid cases in CNA
       obtain mean and sd of exp of diploid cases
       for each case{
          zScore <- (value - mean)/sd
       }
    }
```
### Example Calculation
<table>
<tr><td>Hugo_Symbol</td><td>Entrez_Gene_Id</td><td>A1-A0SD-01</td><td>A1-A0SE-01</td><td>A1-A0SH-01</td><td>A1-A0SJ-01</td><td>A1-A0SK-01</td><td>A1-A0SM-01</td><td>A1-A0SO-01</td><td>A1-A0SP-01</td><td>A2-A04N-01</td><td>A2-A04P-01</td><td>A2-A04Q-01</td><td>A2-A04R-01</td><td>A2-A04T-01</td><td>A2-A04U-01</td><td>A2-A04V-01</td><td>A2-A04W-01</td><td>A2-A04X-01</td><td>A2-A04Y-01</td></tr>
<tr><td>RPS11 Expr</td><td>6205</td><td>0.765</td><td>0.716</td><td>0.417125</td><td>0.115</td><td>0.492875</td><td>-0.525</td><td>-0.169</td><td>0.396</td><td>0.50475</td><td>0.400875</td><td>0.393125</td><td>0.9165</td><td>0.627125</td><td>0.337125</td><td>0.705</td><td>0.16425</td><td>0.325</td><td>0.11175</td></tr>
<tr><td>RPS11 CNA</td><td>6205</td><td>0</td><td>0</td><td>0</td><td>1</td><td>1</td><td>0</td><td>-1</td><td>0</td><td>0</td><td>2</td><td>0</td><td>0</td><td>1</td><td>-1</td><td>0</td><td>0</td><td>-1</td><td>0</td></tr>
</table>

Calculate mean and stdev where CNA is 0 (=diploid):

<table>
<tr><td>Diploid avg</td><td>Diploid std</td></tr>
<tr><td>---</td><td>---</td></tr>
<tr><td>0.414954545454545</td><td>0.399504498851105</td></tr>
</table>

Calculate the z-scores:
<table>
<tr><td>Hugo_Symbol</td><td>Entrez_Gene_Id</td><td>A1-A0SD-01</td><td>A1-A0SE-01</td><td>A1-A0SH-01</td><td>A1-A0SJ-01</td><td>A1-A0SK-01</td><td>A1-A0SM-01</td><td>A1-A0SO-01</td><td>A1-A0SP-01</td><td>A2-A04N-01</td><td>A2-A04P-01</td><td>A2-A04Q-01</td><td>A2-A04R-01</td><td>A2-A04T-01</td><td>A2-A04U-01</td><td>A2-A04V-01</td><td>A2-A04W-01</td><td>A2-A04X-01</td><td>A2-A04Y-01</td></tr>
<tr><td>RPS11 Expr Output</td><td>6205</td><td>0.8762</td><td>0.7535</td><td>0.0054</td><td>-0.7508</td><td>0.1950</td><td>-2.3528</td><td>-1.4617</td><td>-0.0474</td><td>0.2248</td><td>-0.0352</td><td>-0.0546</td><td>1.2554</td><td>0.5311</td><td>-0.1948</td><td>0.7260</td><td>-0.6275</td><td>-0.2252</td><td>-0.7590</td></tr>
</table>

Note: this implies that your full dataset does not have average=0, std=1

### Running the script


To run the script type the following commands when in the folder `<cbioportal_source_folder>/core/src/main/scripts`:

```
export PORTAL_HOME=<cbioportal_configuration_folder>
```

and then 

```
./convertExpressionZscores.pl <copy_number_file> <expression_file> <output_file> <normal_sample_suffix> <[min_number_of_diploids]>
```

#### Example:

```
./convertExpressionZscores.pl  ../../test/scripts/test_data/study_es_0/data_CNA.txt ../../test/scripts/test_data/study_es_0/data_expression_median.txt data_expression_ZSCORES.txt NONE
```
## NormalizeExpressionLevels_allsampleref method
Given the expression data for a set of samples, generate normalized expression values with the reference population of all samples independent of sample diploid status.

### Parameters
`<expression_file> <output_file> <[log_transform]> <[exclude_zero_negative_values]>`

- `<expression_file>` : the [expression (exp) data file](File-Formats.md#expression-data).
- `<output_file>` : the output file to be generated.
- `<log_transform>` : Use this to log transform the data before calculating z-scores (optional).
- `<exclude_zero_negative_values>` : Use this to exclude zero's or negative counts from the reference population when normalizing the data (optional). 

### Algorithm
Input [expression data file](File-Formats.md#expression-data)
```
for each gene{
  log-transform the raw data, if -l is passed
  identify the base population (if -e is passed, n = # of samples where expression value is not zero or negative)
  compute mean and standard deviation for samples in the base poplulation 
  for each sample in the input dataset{
    Z-Score <- (value - mean)/sd when standard deviation != 0
    Z-Score <- NA when standard deviation = 0
  }
}
```
### Log-transforming the data
Using the `-l` option above calculates log base 2 of the expression values.

Here's how we handle the Negative values when log transforming:
```
Replace the negative values to 0 and add a pseudo-count (+1) to all observed counts prior to applying log transform.
example, if raw value is -1, the log transform would be log(0+1)
         if the value is 0, the log transform would be log(0+1)
         if the value is 1, the log transform would be log(1+1)
```

### Example Calculation:

Log transform and calculate the z-scores (without -e option):
<table>
<tr><td>Hugo_Symbol</td><td>Entrez_Gene_Id</td><td>A1-A0SD-01</td><td>A1-A0SE-01</td><td>A1-A0SH-01</td><td>A1-A0SJ-01</td><td>A1-A0SK-01</td><td>A1-A0SM-01</td><td>A1-A0SO-01</td><td>A1-A0SP-01</td><td>A2-A04N-01</td><td>A2-A04P-01</td><td>A2-A04Q-01</td><td>A2-A04R-01</td><td>A2-A04T-01</td><td>A2-A04U-01</td><td>A2-A04V-01</td><td>A2-A04W-01</td><td>A2-A04X-01</td><td>A2-A04Y-01</td></tr>
<tr><td>RPS11 Expr</td><td>6205</td><td>0.765</td><td>0.716</td><td>0.417125</td><td>0.115</td><td>0.492875</td><td>-0.525</td><td>-0.169</td><td>0.396</td><td>0.50475</td><td>0.400875</td><td>0.393125</td><td>0.9165</td><td>0.627125</td><td>0.337125</td><td>0.705</td><td>0.16425</td><td>0.325</td><td>0.11175</td></tr>
<tr><td>Output</td><td>6205</td><td>1.2879</td><td>1.1378</td><td>0.1177</td><td>-1.1605</td><td>0.3953</td><td>-1.7408</td><td>-1.7408</td><td>0.0376</td><td>0.4375</td><td>0.0562</td><td>0.0266</td><td>1.7268</td><td>0.8543</td><td>-0.1921</td><td>1.1035</td><td>-0.9301</td><td>-0.2407</td><td>-1.1761</td></tr>
  </table>

### Running the script
To run the script clone the datahub-study-curation-tools from [here](https://github.com/cBioPortal/datahub-study-curation-tools) and type the commands when in the folder `zscores/zscores_relative_allsamples`:

```
python NormalizeExpressionLevels_allsampleref.py -i <expression_file> -o <output_file> [-l] [-e]
```

#### Example:

```
python NormalizeExpressionLevels_allsampleref.py -i path/to/your/study/data_expression_median.txt -o path/to/your/study/data_mRNA_median_Zscores.txt -l
```
cBioPortal, 2015# cBioPortal

The cBioPortal for Cancer Genomics provides visualization, analysis, and download of large-scale cancer genomics data sets. For a short intro on cBioPortal, see [these introductory slides](https://docs.google.com/presentation/d/1hm0G77UklZnpQfFvywBfW2ZIsy8deKi5r1RfJarOPLg/edit?usp=sharing).

If you would like to know how to setup a private instance of the portal and/or get set up for developing, see the [documentation](https://docs.cbioportal.org). For details on contributing code changes via pull requests, see our [Contributing document](CONTRIBUTING.md).

If you are interested in coordinating the development of new features, please contact cbioportal@cbioportal.org or reach out on https://slack.cbioportal.org.

## üìò Documentation
See [https://docs.cbioportal.org](https://docs.cbioportal.org)

## ü§ù License
See [LICENSE](./LICENSE)

## üíª Run Backend
cBioPortal consists of several components, please read the [Architecture docs](https://docs.cbioportal.org/architecture-overview/) to figure out what repo would be relevant to edit. If you e.g. only want to make frontend changes, one can directly edit [the frontend repo](https://github.com/cbioportal/cbioportal-frontend) instead. Read the instructions in that repo for more info on how to do frontend development. This repo only contains the backend part. Before editing the backend, it's good to read the [backend code organization](docs/Backend-Code-Organization.md).

### Local Development
#### What MySQL database to use
We recommend to set up a MySQL database using [Docker Compose](https://github.com/cBioPortal/cbioportal-docker-compose). It's useful to know how to do this as it allows you to import any dataset of your choice. For debugging production issues, we also have a database available with all the data on https://cbioportal.org that one can connect to directly. Please reach out on slack to get the credentials.

#### Command Line
If you want to run the cBioPortal web app from the command line please follow these instructions. First, we want to make sure that all ports are open for the services set up through [docker compose](https://github.com/cBioPortal/cbioportal-docker-compose) (i.e. not just accessible to other containers within the same Docker Compose file). To do so, in the [docker compose repo](https://github.com/cBioPortal/cbioportal-docker-compose) run:

```
docker compose -f docker-compose.yml -f open-ports.yml up
```
This should open the ports. Now we are ready to run the cBioPortal web app locally. You can compile the backend code with:

```

java -Xms2g -Xmx4g \
     -Dauthenticate=noauthsessionservice \
     -Dsession.service.url=http://localhost:5000/api/sessions/my_portal/ \
     -Dsession.service.origin='*' \
     -Dspring.datasource.username=cbio_user \
     -Dspring.datasource.password=somepassword \
     -Dspring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver \
     -Dspring.jpa.database-platform=org.hibernate.dialect.MySQL5InnoDBDialect \
     -Dspring.datasource.url='jdbc:mysql://cbio_user:somepassword@localhost:3306/cbioportal?useSSL=false&allowPublicKeyRetrieval=true' \
     -Dshow.civic=true \
     -Dskin.footer='' \
     -Dapp.name='my-portal' \
     -Ddbconnector=dbcp \
     -cp "$PWD:$PWD/BOOT-INF/lib/*" \
     org.cbioportal.PortalApplication
```

The app should now show up at http://localhost:8080.

#### Deploy your development image inside Docker Compose
Another option is to deploy your development image directly into the [docker-compose](https://github.com/cBioPortal/cbioportal-docker-compose/blob/5da068f0eb9b4f42db52ab5e91321b26a1826d7a/docker-compose.yml#L6) file. First build the image like this

```
docker build -t cbioportal/cbioportal:my-dev-cbioportal-image -f docker/web-and-data/Dockerfile .
```

Then change the [env file](https://github.com/cBioPortal/cbioportal-docker-compose/blob/master/.env) to use `cbioportal/cbioportal:my-dev-cbioportal-image`.

### Local Development


Note: internally we have a dev database available with the public data set that one can connect to directly. Please reach out on slack to get the credentials. It is usually best to use a small test dataset, but if a copy of the production database is necessary for e.g. fixing a bug specific to production data that can be useful.

### üïµÔ∏è‚Äç‚ôÄÔ∏è Debugging

If you want to attach a debugger you can change the `docker-compose.yml` file to include the parameters: `-Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=5005` (make sure to expose the debug port by adding `5005:5005` in the ports section of the cbioportal container). If you are running the java app outside of docker you can add the same parameters to the java command line arguments instead.

You can then use a JAVA IDE to connect to that port. E.g. in [VSCode](https://code.visualstudio.com/), one would add the following configuration to `launch.json` to connect:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "java",
            "name": "Debug (Attach)",
            "request": "attach",
            "hostName": "localhost",
            "port": 5005,
            "projectName": "cbioportal"
        }
    ]
}
```

## üå≥ Branch Information

| | main branch | upcoming release branch | later release candidate branch |
| --- | --- | --- | --- |
| Branch name | [`master`](https://github.com/cBioPortal/cbioportal/tree/master) |  -- |  [`rc`](https://github.com/cBioPortal/cbioportal/tree/rc) |
| Description | All bug fixes and features not requiring database migrations go here. This code is either already in production or will be released this week | Next release that requires database migrations. Thorough manual product review often takes place for this branch before release | Later releases with features that require database migrations. This is useful to allow merging in new features without affecting the upcoming release. Could be seen as a development branch, but note that only high quality pull requests are merged. That is the feature should be pretty much ready for release after merge. |
| Live instance | https://www.cbioportal.org / https://master.cbioportal.org | -- | https://rc.cbioportal.org |
| Live instance version | https://www.cbioportal.org/api/info / https://master.cbioportal.org/api/info | -- | https://rc.cbioportal.org/api/info |
| Docker Image | cbioportal/cbioportal:master | --| cbioportal/cbioportal:rc |
| Kubernetes Config | [production](https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/cbioportal/cbioportal_spring_boot.yaml) / [master](https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/cbioportal/cbioportal_backend_master.yaml) | -- | [rc](https://github.com/knowledgesystems/knowledgesystems-k8s-deployment/blob/master/cbioportal/cbioportal_backend_rc.yaml) |
| Status | [![master build status](https://github.com/cbioportal/cbioportal/workflows/Core%20tests/badge.svg)](https://github.com/cBioPortal/cbioportal/actions/workflows/core-test.yml?query=branch%3Amaster) [![master build status](https://github.com/cbioportal/cbioportal/workflows/Integration%20tests/badge.svg)](https://github.com/cBioPortal/cbioportal/actions/workflows/integration-test.yml?query=branch%3Amaster) [![master build status](https://github.com/cbioportal/cbioportal/workflows/Docker%20Image%20CI/badge.svg)](https://github.com/cBioPortal/cbioportal/actions/workflows/dockerimage.yml?query=branch%3Amaster) [![master build status](https://github.com/cbioportal/cbioportal/workflows/Python%20validator/badge.svg)](https://github.com/cBioPortal/cbioportal/actions/workflows/validate-data.yml?query=branch%3Amaster) [![CircleCI](https://circleci.com/gh/cBioPortal/cbioportal/tree/master.svg?style=svg)](https://app.circleci.com/pipelines/github/cBioPortal/cbioportal?branch=master&filter=all) | -- | -- |

## üöÄ Releases
Release Notes on GitHub:

https://github.com/cBioPortal/cbioportal/releases

See also the cBioPortal News section for user focused release information:

https://www.cbioportal.org/news

Docker Images are available for each tag and branch:

https://hub.docker.com/repository/docker/cbioportal/cbioportal/tags

## üëâ Other Repos
Read the [Architecture docs](https://docs.cbioportal.org/2.1-deployment/architecture-overview) to see how these relate:

- https://github.com/cBioPortal/cbioportal-frontend
- https://github.com/cbioportal/session-service
- https://github.com/cBioPortal/datahub/
# Security Policy

## Supported Versions

We release security releases for the following versions:

| Version | Supported          |
| ------- | ------------------ |
| 6.x   | :white_check_mark: |
| <= 5.x   | :x:  |

## Reporting a Vulnerability
If there are any vulnerabilities, don't hesitate to report them.

Use the private contact address cbioportal@cbioportal.org to report the fix. 

Describe the vulnerability.

If you have a fix, that is most welcome -- please attach or summarize it in your message!

We will evaluate the vulnerability and, if necessary, release a fix or mitigating steps to address it. We will contact you to let you know the outcome, and will credit you in the report.

Please do not disclose the vulnerability publicly until a fix is released!

Once we have either a) published a fix, or b) declined to address the vulnerability for whatever reason, you are free to publicly disclose it.
